---
title: "Data integration with LIGER"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Data integration with LIGER}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction

LIGER was initially introduced in [Welch et al. 2019](https://doi.org/10.1016/j.cell.2019.05.006) as a method for integrating single-cell RNA-seq data across multiple technologies, species, and conditions. The method relies on integrative nonnegative matrix factorization (iNMF) to identify shared and dataset-specific factors.

LIGER can be used to compare and contrast experimental datasets in a variety of contexts, for instance:

- Across experimental batches
- Across individuals
- Across sex
- Across tissues
- Across species (e.g., mouse and human)
- Across modalities (e.g., scRNAseq and spatial transcriptomics data, scMethylation, or scATAC-seq)

Once multiple datasets are integrated, the package provides functionality for further data exploration, analysis, and visualization. Users can:

- Identify clusters
- Find significant shared (and dataset-specific) gene markers
- Compare clusters with previously identified cell types
- Visualize clusters and gene expression using t-SNE and UMAP

## Usage

We have now made a [documentation website for rliger 2.0.0](https://welch-lab.github.io/liger/). Please check it out for detailed introduction. 

We have made a number of vignettes for typical types of analysis that can be performed with LIGER.

* [Integrating Multiple Single-Cell RNA-seq Datasets](https://welch-lab.github.io/liger/articles/Integrating_multi_scRNA_data.html)
* [Jointly Defining Cell Types from scRNA-seq and scATAC-seq](https://welch-lab.github.io/liger/articles/Integrating_scRNA_and_scATAC_data.html)
* [Iterative Single-Cell Multi-Omic Integration Using Online iNMF](https://welch-lab.github.io/liger/articles/online_iNMF_tutorial.html)
* [Integrating unshared features with UINMF](https://welch-lab.github.io/liger/articles/UINMF_vignette.html)
* [Integrating spatial transcriptomic and transcriptomic datasets using UINMF](https://welch-lab.github.io/liger/articles/STARmap_dropviz_vig.html)
* [scATAC and scRNA Integration using unshared features (UINMF)](https://welch-lab.github.io/liger/articles/SNAREseq_walkthrough.html)
* [Cross-species Analysis with UINMF](https://welch-lab.github.io/liger/articles/cross_species_vig.html)
* [Jointly Defining Cell Types from Single-Cell RNA-seq and DNA Methylation](https://welch-lab.github.io/liger/articles/rna_methylation.html)

Meanwhile, since version 2.0.0, LIGER is massively updated for usability and interoperability with other packages. Below are links to the introduction of new features.

* [Introduction to new liger object and other related classes](https://welch-lab.github.io/liger/articles/liger_object.html)
* [Running Liger directly on Seurat objects](https://welch-lab.github.io/liger/articles/liger_with_seurat.html)

## Loading the package

```{r setup, message=FALSE}
library(rliger)
```
