<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Changelog"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    <a class="dropdown-item" href="../articles/benchmark.html">Benchmarking integration quality</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Changelog</h1>
      <small>Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.1.0" id="rliger-210">rliger 2.1.0<a class="anchor" aria-label="anchor" href="#rliger-210"></a></h2><p class="text-muted">CRAN release: 2024-10-29</p>
<ul><li>Added <code><a href="../reference/centroidAlign.html">centroidAlign()</a></code> for new cell factor loading alignment method</li>
<li>Added <code><a href="../reference/plotProportionBox.html">plotProportionBox()</a></code> for visualizing compositional analysis</li>
<li>Added <code><a href="../reference/plotClusterGeneViolin.html">plotClusterGeneViolin()</a></code> for visualizing gene expression in clusters</li>
<li>Added <code><a href="../reference/plotBarcodeRank.html">plotBarcodeRank()</a></code> for basic QC visualization</li>
<li>Added <code><a href="../reference/plotPairwiseDEGHeatmap.html">plotPairwiseDEGHeatmap()</a></code> for visualizing pairwise DEG results</li>
<li>Added <code><a href="../reference/plotGODot.html">plotGODot()</a></code> for visualizing GO enrichment results</li>
<li>Added <code><a href="../reference/calcNMI.html">calcNMI()</a></code> for evaluating clustering results against ground truth</li>
<li>Added <code>ligerToH5AD()</code> allowing reticulate/Python free export of liger object to H5AD format. This is presented in extension source code (i.e. not loaded with <code><a href="https://welch-lab.github.io/liger/" class="external-link">library(rliger)</a></code>).</li>
<li>Added organism support in <code><a href="../reference/runGeneralQC.html">runGeneralQC()</a></code> and refined hemoglobin gene matching regex pattern.</li>
<li>Optimized DE test memory usage scalability for both pseudo-bulk method and wilcoxon test</li>
<li>Optimized <code><a href="../reference/plotProportion.html">plotProportionPie()</a></code> by adding argument <code>circleColors</code>
</li>
<li>Optimized <code><a href="../reference/plotVolcano.html">plotVolcano()</a></code> text annotation positioning and gene highlighting logic.</li>
<li>Optimized visualization function additional argument documentation</li>
<li>Changed <code><a href="../reference/liger-DEG.html">runMarkerDEG()</a></code> and <code><a href="../reference/liger-DEG.html">runPairwiseDEG()</a></code> default method from <code>"wilcoxon"</code> to <code>"pseudoBulk"</code>
</li>
<li>Fixed <code>runMarkerDEG(method = "pseudobulk")</code> bug in assigning pseudo-replicates, and optimized error/warning signaling.</li>
<li>Fixed bug in <code><a href="../reference/calcAlignment.html">calcAlignment()</a></code>, <code><a href="../reference/subsetLigerDataset.html">subsetMemLigerDataset()</a></code>, <code><a href="../reference/liger-class.html">cellMeta()</a></code>
</li>
<li>Fixed bug in old version updating functions</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.0.1" id="rliger-201">rliger 2.0.1<a class="anchor" aria-label="anchor" href="#rliger-201"></a></h2><p class="text-muted">CRAN release: 2024-04-04</p>
<ul><li>Fixed wrong UINMF aborting criteria</li>
<li>Fixed example/test skipping criteria for non-existing dependencies</li>
<li>Fixed file access issue when checking on CRAN</li>
<li>Updated installed data file <code>system.file("extdata/ctrl.h5", "extdata/stim.h5")</code> to be of standard 10X H5 format</li>
<li>Updated <code><a href="../reference/quantileNorm.html">quantileNorm()</a></code> automatic reference selection according to <a href="https://github.com/welch-lab/liger/issues/297" class="external-link">#297</a></li>
<li>Other minor fixes (including <a href="https://github.com/welch-lab/liger/issues/308" class="external-link">#308</a>)</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.0.0" id="rliger-200">rliger 2.0.0<a class="anchor" aria-label="anchor" href="#rliger-200"></a></h2><p class="text-muted">CRAN release: 2024-03-20</p>
<ul><li>Added <code>ligerDataset</code> class for per-dataset information storage, with inheritance for specific modalities</li>
<li>Added a number of plotting functions with clear function names and useful functionality</li>
<li>Added Leiden clustering method, now as default rather than Louvain</li>
<li>Added pseudo-bulk DEG method</li>
<li>Added DEG analysis with one-vs-rest marker detection in <code><a href="../reference/liger-DEG.html">runMarkerDEG()</a></code> and pairwise comparison in <code><a href="../reference/liger-DEG.html">runPairwiseDEG()</a></code>
</li>
<li>Added gene name pattern for expression percentage QC metric</li>
<li>Added native Seurat object support for the core integration workflow</li>
<li>Added a documentation website built with pkgdown</li>
<li>Added new iNMF variant method, consensus iNMF (c-iNMF), in <code><a href="../reference/runCINMF.html">runCINMF()</a></code>. Not stable.</li>
<li>Added GO enrichment dowsntream analysis in <code><a href="../reference/runGOEnrich.html">runGOEnrich()</a></code>
</li>
<li>Changed <code>liger</code> object class structure</li>
<li>Moved iNMF (previously <code><a href="../reference/rliger-deprecated.html">optimizeALS()</a></code>), UINMF (previously <code>optimizeALS(unshared = TRUE)</code>) and online iNMF (previously <code><a href="../reference/rliger-deprecated.html">online_iNMF()</a></code>) implementation to new package <em>RcppPlanc</em> with vastly improved performance. Now wrapped in <code><a href="../reference/runINMF.html">runINMF()</a></code>, <code><a href="../reference/runUINMF.html">runUINMF()</a></code> and <code><a href="../reference/runOnlineINMF.html">runOnlineINMF()</a></code> respectively, and all can be invoked with <code><a href="../reference/runIntegration.html">runIntegration()</a></code>.</li>
<li>Updated H5AD support to match up with Python anndata package 0.8.0 specs</li>
<li>Renamed many function/argument names to follow camelCase style, original names are still available while deprecation warnings are issued</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.1" id="rliger-101">rliger 1.0.1<a class="anchor" aria-label="anchor" href="#rliger-101"></a></h2><p class="text-muted">CRAN release: 2023-11-08</p>
<ul><li>Allow setting mito pattern in <code>getMitoProportion()</code> <a href="https://github.com/welch-lab/liger/issues/271" class="external-link">#271</a></li>
<li>Fix efficiency issue when taking the log of norm.data (e.g. <code>runWilcoxon</code>)</li>
<li>Add runable examples to all exported functions when possible</li>
<li>Fix typo in online_iNMF matrix initialization</li>
<li>Adapt to Seurat5</li>
<li>Other minor fixes</li>
</ul></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

