---
title: "Joint definition of cell types from multiple scRNA-seq datasets"
author: "Yichen Wang, Joshua Sodicoff and Joshua Welch"
date: "2024-09-27"
output:
  html_document: 
    toc: 3
    toc_float:
      collapsed: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

This article is an general walkthrough of all main LIGER functionalities. We demonstrate the usage of the *rliger* package in the style of the R Console, which can be accessed through an R development environment (e.g., RStudio) or directly from the R command line.

LIGER relies on iNMF (integrative Non-negative Matrix Factorization, [Welch, JD 2019](https://doi.org/10.1016/j.cell.2019.05.006)) to jointly factorize single-cell data into interpretable low-dimensional representations, and integrated clustering can be obtained upon aligned cell factor loading. The package also provides a variety of utilities for visualization and analysis of clustering, gene expression across datasets, comparisons of cluster assignments and etc.

First, load the necessary libraries that will be used for performing all the analyses. [*dplyr*](https://dplyr.tidyverse.org/) has been widely used for data table manupulation and is famous for its unique user-friendly syntax. [*cowplot*](https://wilkelab.org/cowplot/index.html) is a package that provides a simple and consistent way to combine multiple plots into a single figure.

```{r, results='hide'}
library(rliger)
library(dplyr)
library(cowplot)
```

## Loading data

For the first portion of this protocol, we will be integrating data from control and interferon-stimulated PBMCs from [Kang et al, 2017](https://www.nature.com/articles/nbt.4042). The data can be found in the Gene Expression Omnibus, [Series GSE96583](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE96583). This dataset was originally in the form of output from the 10X Cellranger pipeline. In this tutorial, we prepared a downsampled version of the data.

- [**Click to download `pbmc_ifnb_example.tar.gz`**](https://figshare.com/ndownloader/files/47748940)

After downloading the archive file to your working directory, we need to extract the content to a folder. You can use any unzip tool available on your computer to extract the content. For platform without a graphical UI, operate the following command in its terminal.

```bash
tar -xvzf pbmc_ifnb_example.tar.gz
```

This will result you a folder called `pbmc_ifnb_example` in your working directory, with inner structure like below:

```
pbmc_ifnb_example
├── ctrl
│   └── outs
│       └── filtered_feature_bc_matrix
│           ├── barcodes.tsv.gz
│           ├── features.tsv.gz
│           └── matrix.mtx.gz
└── stim
    └── outs
        └── filtered_feature_bc_matrix
            ├── barcodes.tsv.gz
            ├── features.tsv.gz
            └── matrix.mtx.gz
```

Now we can start to work in R to have the two datasets loaded. When doing your own hands on, **make sure to replace the path to the folder with the actual path on your machine**.

```{r}
# Make sure to modify the following path to reflect the exact location of the folder
pathToFolder <- "pbmc_ifnb_example"
rawList <- read10XRNA(pathToFolder)
pbmcLiger <- createLiger(rawList)
pbmcLiger
```

The loaded raw data `rawList` is a list object containing two matrices for the "ctrl" and "stim" datasets. Then `createLiger()` will take the list of all datasets to be integrated as an input and construct the [liger object](liger_object.html).

For convenience, we have prepared the pre-processed data which are ready to use, easily loaded with `importPBMC()`. 

```{r, eval=FALSE}
# This is not run at rendering time.
pbmcLiger <- importPBMC()
```

For creating a liger object from raw counts data or any other types of source (e.g. import cellranger H5 files, convert from Seurat, SingleCellExperiment, or even H5AD file), please refer to the [detailed tutorial for importing data](import_data.html).

### Quality control

It is always the best practice to check the quality control metrics of the scRNAseq data before performing any other processing. *rliger* by default have it calculated when `createLiger()` is called, with the following variables accessible in `cellMeta(pbmcLiger)` or with the `$` operator:

- `pbmcLiger$nUMI`: The sum of UMI counts for each cell.
- `pbmcLiger$nGene`: The number of genes detected in each cell.
- `pbmcLiger$mito`: The percentage of mitochondrial genes detected in each cell. 
- `pbmcLiger$ribo`: The percentage of ribosomal genes detected in each cell.
- `pbmcLiger$hemo`: The percentage of hemoglobin genes detected in each cell.

> **NOTE** that the example dataset is collected from human samples while the `organism` argument of `createLiger()` is by default "human". If users are working with other species like mouse, they should set the argument to the matching values in order to correctly identify the mitochondrial, ribosomal and hemoglobin genes.

The wrapper function `plotTotalCountViolin()` and `plotGeneDetectedViolin()` show the density distribution of `nUMI` and `nGene`, respectively. 

```{r, fig.width=10, fig.height=4}
plotNUMI <- plotTotalCountViolin(pbmcLiger, dot = TRUE)
plotNGene <- plotGeneDetectedViolin(pbmcLiger, dot = TRUE)
cowplot::plot_grid(plotNUMI, plotNGene, ncol = 2)
```

For plotting the mitochondrial gene expression percentage, please use the following command:

```{r, fig.height=2, fig.width=5}
plotCellViolin(pbmcLiger, "mito", groupBy = "dataset", dot = TRUE)
```

It happens that there is no mitochondrial gene detected in the datasets we use for this tutorial. Users seeing such result with their own data need to pay attention to species setting for identifying the mitochondrial genes and rerun `runGeneralQC()` (click to see usage). If there are mitochondrial genes detected, it is recommended to filter out cells with high mitochondrial gene expression percentage, as they are likely to be dead or dying cells.

There are two approaches of filtering genes and cells in a liger object. The first is to use `removeMissing()` function, which mainly removes non-expressing genes and cells with no counts. This function also allows removing genes that are expressed in too few cells (argument `minCells`) and cells that express too few genes (argument `minFeatures`). The following command removes genes that are detected in less than 3 cells and removes cells that express less than 200 genes.

```{r, results='hide'}
pbmcLiger <- removeMissing(pbmcLiger, minCells = 3, minFeatures = 200)
```

The second way is to use R's native matrix subsetting syntax and use the `cellMeta(pbmcLiger)` variables which are accessible with `$` operator. The following command 

- keeps cells with total counts greater than 500
- keeps cells with more than 200 detected genes
- keeps cells with mitochondrial gene expression percentage less than 5%

```{r, results='hide'}
pbmcLiger <- pbmcLiger[, pbmcLiger$nUMI > 500 & pbmcLiger$nGene > 200 & pbmcLiger$mito < 5]
```

## Preprocessing and Normalization

### Preprocess

Before we can run iNMF on our datasets, we must run several preprocessing steps to normalize expression data to account for differences in sequencing depth and efficiency between cells, identify variably expressed genes, and scale the data so that each gene has the same variance. Note that because nonnegative matrix factorization requires positive values, we do not center the data by subtracting the mean. We also do not log transform the data.

```{r Count-5,results='hide'}
pbmcLiger <- pbmcLiger %>%
    normalize() %>%
    selectGenes() %>%
    scaleNotCenter()
```

For gene selection, we use a customized approach that selects gene for each dataset separately and take the union (or intersection) at last. The selection for each dataset is based on a variance threshold and thus the number of genes selected can be different for each dataset. Users may
see the message log from the run and decide whether to raise or lower the threshold for each dataset to obtain optimal number of genes being selected. 

## Integration with Joint Matrix Factorization

We are now able to perform iNMF on the scaled datasets. The key parameter for this analysis is `k`, the number of ranks each dataset is factorized into (analogous to the number of principal components in PCA). In general, we find that a value of `k` between 20 and 40 is suitable for most analyses and that results are robust for choice of `k`. Because LIGER is an unsupervised, exploratory approach, there is no single "right" value for `k`. In practice, users choose `k` from a combination of biological prior knowledge and other information. A higher `k` value might be considered when there are more distinct composition expected from all datasets involved. For this tutorial, we set `k = 30`.

```{r Count-6, results='hide'}
pbmcLiger <- runIntegration(pbmcLiger, k = 30)
```

>Starting from rliger 2.0.0, we use an optimized implementation of iNMF. Here we deprecated the parameter `thresh` which stands for a convergence detecter in order to speed up each algorithm iteration by omitting the calculation of objective error.

The factorization yields several lower dimension matrices: 

- an $H$ matrix for each dataset, representing the factor loadings for each cell. A factor can be regarded as a metagene.
- the $W$ matrix of shared gene loadings for each factor. 
- A $V$ matrix for each dataset, standing for the dataset-specific gene loadings for each factor. 

Please refer to [liger object documentation](liger_object.html) for how to access them.

The time consumption of this step is linearly dependent of the size of the datasets, in terms of number of cells, number of variable genes selected, and the value of `k`. The implementation supports OpenMP multi-threading on non-Apple-Sillicon machines, and therefore using a machine with a number of cores allocated helps speeding it up.

## Quantile Normalization and Joint Clustering

As introduced above, the $H$ matrices for each dataset will then be used for clustering. These matrices are interpretable in a way that each factor can be considered as a metagene, cell identity program, biological activity program and etc. We need one last step to align the factor loading values across datasets, so that cells voted to be of the same "type" by the factors will have normalized loading values across datasets, before calling a clustering algorithm or generating embeddings for visualization.

### Align the factors

We perform quantile normalization by dataset, factor, and cluster to fully integrate the datasets. All of this functionality is encapsulated within the `quantileNorm()` function, which uses max factor assignment followed by refinement using a k-nearest neighbors graph.

```{r loading-1, results='hide'}
pbmcLiger <- quantileNorm(pbmcLiger)
```

### Clustering

The `quantileNorm()` procedure produces the final low-dimensional representation that integrates the datasets together, namingly `H.norm`. In the mean time, a joint cluster assignment stored as `quantileNorm_cluster` in the object metadata is also generated. This cluster assignment is based on maximum factor voting and is what we used as a guide to align cells from different datasets. For real-life analysis, we recommend running Leiden community detection on SNN graph inferred from the aligned cell factors loading, an approach commonly used for single-cell data. The Leiden algorithm excels at merging small clusters into broad cell classes and thus may be more desirable in some cases than the maximum factor assignments.

```{r loading-2, results='hide'}
pbmcLiger <- runCluster(pbmcLiger, resolution = 0.25, nNeighbors = 30)
```

>Starting from rliger 2.0.0, cluster labeling will be stored in cell metadata, which can be accessed with `cellMeta(pbmcLiger)`. Use argument `clusterName` to specify unique variable names for the result can enable storing multiple cluster labeling variables at the same time.

### Visualize the integration

To visualize the clustering of cells, we can project the normalized cell factors to two or three dimensions. LIGER supports both UMAP and t-SNE for this purpose. 

```{r pre-1, results='hide'}
pbmcLiger <- runUMAP(pbmcLiger, nNeighbors = 30, minDist = 0.3)
```

>Starting from rliger 2.0.0, the slot for storing dimensionality reduction matrices will be renamed to "dimReds". It will be a list that can hold multiple low dimensional matrices that match to all datasets by cell identifiers. Users can access individual matrix with `dimRed(object, "name")`. Use argument `dimredName` to specify unique names for the UMAP/tSNE result so that it allows storing multiple low-dimensional representation matrices at the same time.

`plotByDatasetAndCluster()` returns two graphs, generated by t-SNE or UMAP in the previous step. The first colors cells by dataset of origin, and the second by cluster as determined by previous clustering step. The plots provide visual confirmation that the datasets are well aligned and the clusters are consistent with the shape of the data as revealed by UMAP.

The two subplots can individually be generated with `plotDatasetDimRed()` and `plotClusterDimRed()`, respectively. 

```{r 4-1, fig.align='center', fig.height=6, fig.width=14}
plotByDatasetAndCluster(pbmcLiger)
```

Sometimes it is also important to have a look at the cluster distribution within each dataset, because the UMAP plot colored by dataset can be two messy when different colors are expected to be mixed up. This can be achieved by running `plotClusterDimRed()` with setting `splitBy = "dataset"`. Additionally, we can make a contour density plot of the dimensionality reduction to further assist interpreting the distribution.

```{r, fig.align='center', fig.height=10, fig.width=12}
clusterUmapList <- plotClusterDimRed(pbmcLiger, splitBy = "dataset", title = names(pbmcLiger))
densityList <- plotDensityDimRed(pbmcLiger, splitBy = "dataset", combinePlot = FALSE)
# Use cowplot's functionality to bring individual plots in the list together
plot_grid(plotlist = c(clusterUmapList, densityList), align = "hv")
```

From the plots above, we can visually see that there is a clear difference of density between the two datasets at the position of cluster 8, and a location shift at cluster 0. But otherwise the density pattern in most of the parts is very similar between the two datasets, which is a good sign of successful integration.

## Explore and interpret the factors

To directly study the impact of factors on the clustering and determine what genes load most highly on each factor, we use the `plotGeneLoadings()` function, which returns plots of factor loading on the dimensionality reduction and highly loaded genes by dataset for each factor. Knowing the highly loaded gene of each factor largely helps helps understanding what biological processes are driving the clustering and guides the annotation of cell types. The dataset-specific gene loading further helps to understand the differences between datasets.

In this example, we show the 11'th factor, which is highly loaded in cluster 5. From the top loaded genes, we can easily identify that cluster 5 should be considered as NK cells given the context of PBMCs, as GNLY and NKG7 are well known markers for NK cells. **TODO interpret dataset-specific gene loading**

```{r, results='hide', fig.keep='all', fig.align='center', fig.height=7, fig.width=6}
factorMarkers <- getFactorMarkers(pbmcLiger, dataset1 = "ctrl", dataset2 = "stim")
plotGeneLoadings(pbmcLiger, markerTable = factorMarkers, useFactor = 11)
```

## Explore cell type proportions

*rliger* provides a good number of functions that visualize the cell type proportion within each dataset, which helps understanding the compositional change across conditions of interests, also known as a differential abundance analysis.

`plotProportionBox()` is helpful when condition variable is available and groups multiple samples. It makes boxplot with the proportion of each cluster within each sample and groups boxes by condition. In this example, we don't have multiple samples for each condition, so we manually create a fake pseudo-replicate variable that split each real sample and use the original sample variable as the condition variable. **In real-life analysis, users should use real sample variable at `sampleBy` and real condition variable at `conditionBy`**.

```{r, fig.align='center', fig.height=3, fig.width=7}
pbmcLiger$fake_sample <- factor(sprintf(
    fmt = "%s_%s", 
    pbmcLiger$dataset, 
    sample(1:3, ncol(pbmcLiger), replace = TRUE)
))
print(table(pbmcLiger$dataset, pbmcLiger$fake_sample))
plotProportionBox(pbmcLiger, conditionBy = "dataset", sampleBy = "fake_sample")
```

For the boxplot, given the fact that there are always some clusters of small size and the proportions they take squeeze the boxplots a lot, we can use option `splitByCluster = TRUE` to generate the plot of the same information for each cluster.

```{r, fig.align='center', fig.height=8, fig.width=14}
propBoxList <- plotProportionBox(
  pbmcLiger, 
  conditionBy = "dataset", 
  sampleBy = "fake_sample", 
  splitByCluster = TRUE
)
# Again, use cowplot's functionality to bring individual plots in the list together
plot_grid(plotlist = propBoxList)
```

Please also checkout other functions like `plotProportionDot()` and `plotProportionPie()` that express similar information in different ways.

## Differential expression

Starting from *rliger* 2.0.2, we move the default differential expression test method to pseudo-bulk test performed with [DESeq2](https://bioconductor.org/packages/DESeq2/). This is generally tested to be more robust when comparing across batches than wilcoxon rank-sum test that we previously used. We provide two functions, `runMarkerDEG()` and `runPairwiseDEG()` for different purposes. Please click on the function name to be redirected to detailed usage reference.

- `runMarkerDEG()` always performs tests in a one-vs-rest manner, which is suitable for marker identification. Cluster markers can be identified easily, while dataset-specific markers can be identified within each cluster. See example below.
- `runPairwiseDEG()` performs test between two specific groups of cells, either specified by metadata variable or user customized cell indices with great flexibility. Comparison between two condition can be done within each cluster. See example below.

Pseudo-bulk method works optimally when real replicate information is available, i.e. a pseudo-bulk should better be created with cells that belong to the same biological replicate. We recommend inserting such metadata information into the object beforehand and supply the variable name to argument `useReplicate`. Most of the time, the variable `"dataset"` is used for this purpose if users create the [liger object](liger_object.html) with each matrix representing a single real replicate.

### Marker detection for each cluster

For identifying cluster markers, we can simply call the function with setting `conditionBy = "leiden_cluster"` to allow the test to be built on the cluster labels we just got. The function will return a `data.frame` object containing all the statistics for all clusters, split by `group` column. 

```{r 4-2}
# Run the pseudo-bulk DE test for cluster marker detection
clusterMarkerStats <- runMarkerDEG(pbmcLiger, conditionBy = "leiden_cluster")
# Show the top markers ranked by adjusted p-value and logFC for cluster 5, NK cells
library(dplyr)
clusterMarkerStats %>%
    filter(group == 5, logFC > 0) %>%
    arrange(padj, desc(logFC)) %>%
    head(n = 10)
```

We can visualize cluster markers identified by `runMarkerDEG()` with `plotMarkerHeatmap()`.

```{r 4-3, fig.align='center', fig.height=6, fig.width=8}
plotMarkerHeatmap(pbmcLiger, result = clusterMarkerStats)
```

We can also make use of the user-friendly *dplyr* interface to filter and sort for the top markers for each cluster.

```{r 4-4}
clusterMarkerStats %>%
    filter(padj < 0.05, logFC > 0) %>%
    group_by(group) %>%
    arrange(padj, desc(logFC), .by_group = TRUE) %>%
    slice_head(n = 5)
```

We can also visualize the expression profiles of individual genes, such as the markers that we just identified. This allows us to visually confirm the cluster- or dataset-specific expression patterns of marker genes. `plotGeneDimRed()` by default returns a plot of the dimensionality reduction of all cells, colored by specified genes. Here, we further split the plot by dataset to check if the marker expression is consistent across datasets.

```{r, fig.align='center', fig.width=7, fig.height=6}
# Making a single plot
plotGeneDimRed(pbmcLiger, "GNLY")
```

```{r, fig.align='center', fig.width=14, fig.height=6}
# By splitting the plot by dataset, we get a list of two plots
nkg7PlotList <- plotGeneDimRed(pbmcLiger, "NKG7", splitBy = "dataset", title = names(pbmcLiger))
# Then combine them into one with cowplot function
plot_grid(plotlist = nkg7PlotList)
```

### Dataset marker detection split by cluster

Additionally, by specifying `conditionBy = "dataset"` or other metadata variable, the function will do one-vs-rest marker test basing on the specified variable. So it works for dataset-specific marker detection as well. By adding another argument `splitBy`, the tes can even be done within each level of another variable. For example, `runMarkerDEG(pbmcLiger, conditionBy = "dataset", splitBy = "leiden_cluster")` will perform the test within each cluster and find the marker for each dataset. We skip this demonstration because the example data only contains two datasets which is more suitable for pairwise comparison.

### Pairwise comparison between two specific conditions

In a more complexed study design than what this tutorial has, we may have more combinations of condition, such as gender, treatment, age and etc. Directly comparing two specific conditions can make more sense than performing a marker detection where the control group is a mixture of many conditions. We introduce the `runPairwiseDEG()` function for this purpose. In order to see what changes are happening in the "stim" dataset compared to the "ctrl" dataset, we set `groupTest = "stim"` and `groupCtrl = "ctrl"`. We need to also set `variable1 = "dataset"` to allow the the group setting to be found. We further set `splitBy = "leiden_cluster"` to allow the test to be done within each cluster. For this specific example case where only 2 conditions are being compared, the test setting is actually the same as the previous section.

```{r}
stimVsCtrlStats <- runPairwiseDEG(
  pbmcLiger, 
  groupTest = "stim", 
  groupCtrl = "ctrl", 
  variable1 = "dataset",
  splitBy = "leiden_cluster"
)
```

A volcano plot can be generated from any `data.frame` object returned by LIGER's DEG functions. 

```{r, fig.align='center', fig.height=5, fig.width=6}
plotVolcano(stimVsCtrlStats, group = "5.stim")
```

By specifying `group = "5.stim"`, the function makes a volcano plot for that test that is split for cluster 5, which is identified as NK cell cluster, and compares the "stim" dataset to the control, "ctrl" dataset. From the upper-right part of the volcano plot, we can easily tell that ISGs (Interferon-Stimulated Genes) and IFIs (Interferon-induced proteins) are highly expressed in the NK cells from "stim" dataset compared to the NK cells from "ctrl", which obviously agrees with the biological context of the experiment.

Similar to what we did in marker detection visualization, we also provide a heatmap function for pairwise DE testt. It shows the gene expression in downsampled cells involved in a specified test and divide the heatmap vertically by comparison groups and horizontally by up-/down-regulation.

```{r, fig.align='center'}
plotPairwiseDEGHeatmap(pbmcLiger, result = stimVsCtrlStats, group = "5.stim")
```

## GO term enrichment analysis

After getting the differentially expressed genes, we can submit them to tests that identify enriched GO terms. Function `runGOEnrich()` wraps the functionality of [gprofiler2](https://CRAN.R-project.org/package=gprofiler2) package and can directly be called with the returned `data.frame`s from the DEG functions of *rliger*. Note the active internet connection is required for the function to submit the query to *gprofiler*'s server. 
```{r}
goResult <- runGOEnrich(stimVsCtrlStats)
```

If the input DE analysis result consists of multiple tests (e.g. split by cluster, as we did above), the returned result would be a list for each test group. Each element would be an individual GO enrichment analysis result, represented as a list of one `data.frame` for the statistics (`goResult$group_name$result`) and one list for meta information used by *gprofiler2* (`goResult$group_name$meta`). Here we simply use *dplyr* expression to show the enriched GO terms from the comparison of "stim" against "ctrl" within cluster 5. 

```{r, eval=FALSE}
goResult$`5.stim`$result %>%
    filter(startsWith(term_id, "GO")) %>%
    arrange(p_value) %>%
    select(term_name, term_id, p_value, precision, recall, term_size, intersection_size)
```

```{r, echo=FALSE}
goResult$`5.stim`$result %>%
    filter(startsWith(term_id, "GO")) %>%
    arrange(p_value) %>%
    select(term_name, term_id, p_value, precision, recall, term_size, intersection_size) %>% 
    mutate(
      p_value = signif(p_value, digits = 3),
      precision = signif(precision, digits = 3),
      recall = signif(recall, digits = 3)
    ) %>%
    DT::datatable(options = list(pageLength = 5))
```

Here's also a simple visualization function that generates a dot plot per test, with dot size representing the interaction size between query gene set and each term, and color representing the significance of the term. 

```{r, fig.align='center', fig.height=5, fig.width=9}
plotGODot(goResult, group = "5.stim", n = 20)
```

## R Session Info

```{r rsession, echo=FALSE}
utils::sessionInfo()
```
