<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Uses an extension of nonnegative matrix factorization to identify shared and dataset-specific factors. See Welch J, Kozareva V, et al (2019) &lt;doi:10.1016/j.cell.2019.05.006&gt;, and Liu J, Gao C, Sodicoff J, et al (2020) &lt;doi:10.1038/s41596-020-0391-8&gt; for more details.">
<title>Linked Inference of Genomic Experimental Relationships • rliger</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/Atkinson_Hyperlegible-0.4.8/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Linked Inference of Genomic Experimental Relationships">
<meta property="og:description" content="Uses an extension of nonnegative matrix factorization to identify shared and dataset-specific factors. See Welch J, Kozareva V, et al (2019) &lt;doi:10.1016/j.cell.2019.05.006&gt;, and Liu J, Gao C, Sodicoff J, et al (2020) &lt;doi:10.1038/s41596-020-0391-8&gt; for more details.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">NEWS</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><p><img src="https://github.com/welch-lab/liger/raw/newObj/inst/extdata/logo.png" width="120" style="display: inline;"></p>
<p><a href="https://github.com/welch-lab/liger/actions/workflows/r.yml" class="external-link"><img src="https://github.com/welch-lab/liger/actions/workflows/r.yml/badge.svg?branch=newObj" alt="R" style="display: inline;"></a> <a href="https://app.codecov.io/gh/mvfki/liger" class="external-link"><img src="https://codecov.io/gh/mvfki/liger/graph/badge.svg?token=77TTU4GY8" alt="codecov" style="display: inline;"></a> <a href="https://cran.r-project.org/package=rliger" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/rliger" alt="cran" style="display: inline;"></a></p>
<div class="section level1">
<div class="page-header"><h1 id="liger-linked-inference-of-genomic-experimental-relationships">LIGER (Linked Inference of Genomic Experimental Relationships)<a class="anchor" aria-label="anchor" href="#liger-linked-inference-of-genomic-experimental-relationships"></a>
</h1></div>
<blockquote>
<p><strong>Now we have a comprehensive documentation site for the latest version of <a href="https://welch-lab.github.io/liger/index.html" class="external-link">rliger (2.0)</a>!</strong></p>
</blockquote>
<p>LIGER (installed as <code>rliger</code> ) is a package for integrating and analyzing multiple single-cell datasets, developed by the Macosko lab and maintained/extended by the Welch lab. It relies on integrative non-negative matrix factorization to identify shared and dataset-specific factors.</p>
<p>Check out our <a href="https://doi.org/10.1016/j.cell.2019.05.006" class="external-link">Cell paper</a> for a more complete description of the methods and analyses. To access data used in our SN and BNST analyses, visit our study “SCP466” on the <a href="https://singlecell.broadinstitute.org/single_cell" class="external-link">Single Cell Portal</a>.</p>
<p>LIGER can be used to compare and contrast experimental datasets in a variety of contexts, for instance:</p>
<ul>
<li>Across experimental batches</li>
<li>Across individuals</li>
<li>Across sex</li>
<li>Across tissues</li>
<li>Across species (e.g., mouse and human)</li>
<li>Across modalities (e.g., scRNAseq and spatial transcriptomics data, scMethylation, or scATAC-seq)</li>
</ul>
<p>Once multiple datasets are integrated, the package provides functionality for further data exploration, analysis, and visualization. Users can:</p>
<ul>
<li>Identify clusters</li>
<li>Find significant shared (and dataset-specific) gene markers</li>
<li>Compare clusters with previously identified cell types</li>
<li>Visualize clusters and gene expression using t-SNE and UMAP</li>
</ul>
<p>We have also designed LIGER to interface with existing single-cell analysis packages, including <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>.</p>
<div class="section level2">
<h2 id="feedback">Feedback<a class="anchor" aria-label="anchor" href="#feedback"></a>
</h2>
<p>If you have any questions, comments, or suggestions, you are welcomed to <a href="https://github.com/welch-lab/liger/issues" class="external-link">open an Issue</a>!</p>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>For usage examples and guided walkthroughs, check the <code>vignettes</code> directory of the repo.</p>
<ul>
<li><a href="https://welch-lab.github.io/liger/articles/Integrating_multi_scRNA_data.html" class="external-link">Integrating Multiple Single-Cell RNA-seq Datasets</a></li>
<li><a href="https://welch-lab.github.io/liger/articles/Integrating_scRNA_and_scATAC_data.html" class="external-link">Jointly Defining Cell Types from scRNA-seq and scATAC-seq</a></li>
<li><a href="https://welch-lab.github.io/liger/articles/online_iNMF_tutorial.html" class="external-link">Iterative Single-Cell Multi-Omic Integration Using Online iNMF</a></li>
<li><a href="https://welch-lab.github.io/liger/articles/UINMF_vignette.html" class="external-link">Integrating unshared features with UINMF</a></li>
<li><a href="https://welch-lab.github.io/liger/articles/STARmap_dropviz_vig.html" class="external-link">Integrating spatial transcriptomic and transcriptomic datasets using UINMF</a></li>
<li><a href="https://welch-lab.github.io/liger/articles/SNAREseq_walkthrough.html" class="external-link">scATAC and scRNA Integration using unshared features (UINMF)</a></li>
<li><a href="https://welch-lab.github.io/liger/articles/cross_species_vig.html" class="external-link">Cross-species Analysis with UINMF</a></li>
<li><a href="https://welch-lab.github.io/liger/articles/rna_methylation.html" class="external-link">Jointly Defining Cell Types from Single-Cell RNA-seq and DNA Methylation</a></li>
</ul>
<p>Meanwhile, since version 2.0.0, LIGER is massively updated for usability and interoperability with other packages. Below are links to the introduction of new features.</p>
<ul>
<li><a href="https://welch-lab.github.io/liger/articles/liger_object.html" class="external-link">Introduction to new liger object and other related classes</a></li>
<li><a href="https://welch-lab.github.io/liger/articles/liger_with_seurat.html" class="external-link">Running Liger directly on Seurat objects</a></li>
</ul>
<p>If you need to refer to the tutorials for the old version of rliger, please check the <a href="https://github.com/welch-lab/liger/tree/v1.0.1/vignettes" class="external-link">GitHub archive v1.0.1</a>, download the desired rendered HTML files and open them in your browser.</p>
</div>
<div class="section level2">
<h2 id="sample-datasets">Sample Datasets<a class="anchor" aria-label="anchor" href="#sample-datasets"></a>
</h2>
<p>The <code>rliger</code> package provides different types of small toy dataset for basic demos of the functions. After attaching the package in an R session, you can load them with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pbmc"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pbmcPlot"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"bmmc"</span><span class="op">)</span></span></code></pre></div>
<p>We also provide a set of datasets for real-world style demos, including scRNAseq, scATACseq, spatial transcriptomics and DNA methylation data. They are described in detail in the articles that make use of them. Please check them out from the links above.</p>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=rliger" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/welch-lab/liger/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/welch-lab/liger/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing rliger</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Joshua Welch <br><small class="roles"> Author </small>  </li>
<li>Yichen Wang <br><small class="roles"> Author, maintainer </small>  </li>
<li>Chao Gao <br><small class="roles"> Author </small>  </li>
<li>Jialin Liu <br><small class="roles"> Author </small>  </li>
<li>Joshua Sodicoff <br><small class="roles"> Author, contributor </small>  </li>
<li>Velina Kozareva <br><small class="roles"> Author, contributor </small>  </li>
<li>Evan Macosko <br><small class="roles"> Author, contributor </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
