<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rliger">
<title>Benchmarking • rliger</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Benchmarking">
<meta property="og:description" content="rliger">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    <a class="dropdown-item" href="../articles/benchmark.html">Benchmarking integration quality</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<link href="benchmark_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="benchmark_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="benchmark_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="benchmark_files/datatables-binding-0.33/datatables.js"></script><link href="benchmark_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="benchmark_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="benchmark_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="benchmark_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="benchmark_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Benchmarking</h1>
                        <h4 data-toc-skip class="author">Yichen
Wang</h4>
            
            <h4 data-toc-skip class="date">2024-10-15</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/vignettes/articles/benchmark.Rmd" class="external-link"><code>vignettes/articles/benchmark.Rmd</code></a></small>
      <div class="d-none name"><code>benchmark.Rmd</code></div>
    </div>

    
    
<p>In <em>rliger 2.1.0</em>, we released a new cell factor loading
alignment method which has been outperforming our previous quantile
normalization method as well as many existing well known methods, in
terms of the power of batch effect correction and conservation of
biological information. In this vignette, we demonstrate how the
performance was benchmarked with a number of metrics.</p>
<p>The benchmarking metrics aim at evaluating single-cell data
integration methods and the generalized framework was proposed in
original work in <a href="https://doi.org/10.1038/s41592-021-01336-8" class="external-link">Luecken, M.D., <em>Nat
Methods</em> 2022</a>, by Theis Lab. In this vignette, we will do the
following steps to benchmark our new method and compare it with other
methods:</p>
<ol style="list-style-type: decimal">
<li>Load the dataset that was used in the original work</li>
<li>Run LIGER’s new method</li>
<li>Calculate benchmarking metrics on the new result, using Python
package <a href="https://github.com/theislab/scib" class="external-link"><em>scib</em></a>
from Theis Lab.</li>
<li>Rank the new method with existing methods</li>
</ol>
<div class="section level2">
<h2 id="dataset">Dataset<a class="anchor" aria-label="anchor" href="#dataset"></a>
</h2>
<p>The SCIB (Single-cell integration benchmarking) paper benchmarked
single-cell data integration methods on a number of datasets with
various complexities. Here, we show how to benchmark LIGER’s new method
on the human pancreas dataset.</p>
<p>The pancreas integration task consists of batches sequenced with
different technologies, including CEL-Seq, Fluidigm C1, inDrop,
Smart-Seq and etc. Moreover, the cell composition in each batch varies.
For example, the CEL-Seq and CEL-Seq2 batches contributed by Muraro MJ
2016 and Grün D 2016, respectively, pooled cells using the Islets of
Langerhans from human cadaveric pancreata at atlas level and contains
main and rare pancreatic cell types. While the SMARTer batch contributed
by Xin Y 2016, only contains the main pancreatic cell types.</p>
<p>SCIB has provided the cleaned dataset with cell type annotation
unified across batches. The dataset is ready for download and is
presented in an AnnData object stored in the H5AD file listed below:</p>
<blockquote>
<p>Click to download and <strong>save at working directory</strong>: <a href="https://figshare.com/ndownloader/files/24539828" class="external-link"><strong>human_pancreas_norm_complexBatch.h5ad</strong></a></p>
</blockquote>
</div>
<div class="section level2">
<h2 id="python-environment">Python environment<a class="anchor" aria-label="anchor" href="#python-environment"></a>
</h2>
<p>The SCIB framework is wrapped in a Python package <em>scib</em> and
the metric calculation can be easily done with a function call to a
prepared AnnData object. To install the <em>scib</em> package, we
recommand creating a conda environment and install the package with
<code>pip</code>.</p>
<p>The commands below shows an example of how to create the conda
environment and install the package in system terminal or command-line
prompt (NOT R COMMANDS). Users can freely explore other ways to get the
package installed.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># In terminal/commandline</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-n</span> scib python=3.9.0</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="ex">conda</span> activate scib</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="ex">pip</span> install scib</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h2>
<p>Here we use SCIB’s method to identify the highly variable genes
(HVGs) before running LIGER integration. While LIGER has its own HVG
selection strategy, we decided to use SCIB’s method to ensure a fair
comparison with other methods, which were benchmarked with the same set
of HVGs. This method requests for <code>n</code> (default 2,000) top
variable genes from each batch first. Then, the list of HVGs was ranked
first by the number of batches in which the genes were highly variable
and second by the mean dispersion parameter across batches. Finally, the
top <code>n</code> genes were selected for downstream use.</p>
<p>In a Python session, import the necessary modules and load the
dataset downloaded from the link above.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Python code</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="im">import</span> scib</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">"human_pancreas_norm_complexBatch.h5ad"</span>)</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>adata</span></code></pre></div>
<pre><code><span><span class="co">## AnnData object with n_obs × n_vars = 16382 × 19093</span></span>
<span><span class="co">##     obs: 'tech', 'celltype', 'size_factors'</span></span>
<span><span class="co">##     layers: 'counts'</span></span></code></pre>
<p>From the printed information, the <code>adata.obs</code> column
“tech” represents the batch information.</p>
<p>Next, we run the HVG selection method from <em>scib</em>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Python code</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>hvg <span class="op">=</span> scib.pp.hvg_batch(adata, batch_key<span class="op">=</span><span class="st">'tech'</span>)</span></code></pre></div>
<pre><code><span><span class="co">## Using 65 HVGs from full intersect set</span></span>
<span><span class="co">## Using 221 HVGs from n_batch-1 set</span></span>
<span><span class="co">## Using 274 HVGs from n_batch-2 set</span></span>
<span><span class="co">## Using 269 HVGs from n_batch-3 set</span></span>
<span><span class="co">## Using 349 HVGs from n_batch-4 set</span></span>
<span><span class="co">## Using 435 HVGs from n_batch-5 set</span></span>
<span><span class="co">## Using 387 HVGs from n_batch-6 set</span></span>
<span><span class="co">## Using 2000 HVGs</span></span></code></pre>
<p>The selected HVGs are stored in a <code>list</code> object
<code>hvg</code>. We simply write the <code>list</code> to a text file,
with one gene per line without separators, so that we can load it back
to R later easily.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Python code</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">"pancreas_hvgs.txt"</span>, <span class="st">"w"</span>) <span class="im">as</span> f:</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>    <span class="cf">for</span> gene <span class="kw">in</span> hvg:</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>        f.write(<span class="ss">f"</span><span class="sc">{</span>gene<span class="sc">}</span><span class="ch">\n</span><span class="ss">"</span>)</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>f.close()</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="construct-liger-object">Construct liger object<a class="anchor" aria-label="anchor" href="#construct-liger-object"></a>
</h2>
<p>Back to R, now we can load the H5AD data to R and construct a <a href="liger_object.html">liger object</a>. There is no single function
that can load the H5AD file directly to a liger object due to the fact
that people may have very distinct ways to construct an AnnData object.
Here we show particularly how the pancreas dataset is manually loaded
into a liger object, step by step.</p>
<p>First, load the necessary R libraries</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://hhoeflin.github.io/hdf5r/" class="external-link">hdf5r</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://welch-lab.github.io/liger/" class="external-link">rliger</a></span><span class="op">)</span></span></code></pre></div>
<p>The R package <em>hdf5r</em> provides an interface to interact with
HDF5 based files, which is the base format of an H5AD file. The
<em>rliger</em> package is the R implementation of LIGER.</p>
<p>Secondly, open the H5AD file so we can extract necessary data.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h5ad</span> <span class="op">&lt;-</span> <span class="va"><a href="http://hhoeflin.github.io/hdf5r/reference/H5File-class.html" class="external-link">H5File</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="st">"human_pancreas_norm_complexBatch.h5ad"</span>, mode <span class="op">=</span> <span class="st">"r"</span><span class="op">)</span></span></code></pre></div>
<p>With the connection to the H5AD file open, we can subsequently
extract the raw counts and metadata. Notice that the raw counts are
stored in a dense matrix in the H5AD file, which is very rare in
practice for single-cell data. We need to convert it to sparse matrix
before constructing the liger object. Please view <a href="https://cran.r-project.org/web/packages/hdf5r/vignettes/hdf5r.html" class="external-link"><em>hdf5r</em>
vignette</a> for more information of its usage.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Loading the raw counts from H5AD file</span></span>
<span><span class="va">h5ad.raw</span> <span class="op">&lt;-</span> <span class="va">h5ad</span><span class="op">[[</span><span class="st">'layers/counts'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="op">]</span></span>
<span><span class="va">h5ad.raw.rownames</span> <span class="op">&lt;-</span> <span class="va">h5ad</span><span class="op">[[</span><span class="st">'var/_index'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="op">]</span></span>
<span><span class="va">h5ad.raw.colnames</span> <span class="op">&lt;-</span> <span class="va">h5ad</span><span class="op">[[</span><span class="st">'obs/_index'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dimnames.html" class="external-link">dimnames</a></span><span class="op">(</span><span class="va">h5ad.raw</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">h5ad.raw.rownames</span>, <span class="va">h5ad.raw.colnames</span><span class="op">)</span></span>
<span><span class="co"># Convert it liger compatible sparse matrix</span></span>
<span><span class="va">h5ad.raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">h5ad.raw</span>, <span class="st">"CsparseMatrix"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract necessary metadata</span></span>
<span><span class="va">h5ad.meta.dataset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span></span>
<span>    <span class="va">h5ad</span><span class="op">[[</span><span class="st">'obs/__categories/tech'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="op">]</span><span class="op">[</span><span class="va">h5ad</span><span class="op">[[</span><span class="st">'obs/tech'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="op">]</span> <span class="op">+</span> <span class="fl">1</span><span class="op">]</span>, </span>
<span>    levels <span class="op">=</span> <span class="va">h5ad</span><span class="op">[[</span><span class="st">'obs/__categories/tech'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="va">h5ad.meta.cell_type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span></span>
<span>    <span class="va">h5ad</span><span class="op">[[</span><span class="st">'obs/__categories/celltype'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="op">]</span><span class="op">[</span><span class="va">h5ad</span><span class="op">[[</span><span class="st">'obs/celltype'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="op">]</span> <span class="op">+</span> <span class="fl">1</span><span class="op">]</span>, </span>
<span>    levels <span class="op">=</span> <span class="va">h5ad</span><span class="op">[[</span><span class="st">'obs/__categories/celltype'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="va">pancreas.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    barcode <span class="op">=</span> <span class="va">h5ad.raw.colnames</span>,</span>
<span>    dataset <span class="op">=</span> <span class="va">h5ad.meta.dataset</span>,</span>
<span>    cell_type <span class="op">=</span> <span class="va">h5ad.meta.cell_type</span>,</span>
<span>    row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">h5ad.meta.dataset</span>, <span class="st">"_"</span>, <span class="va">h5ad.raw.colnames</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Now, with all data loaded, we can head to constructing the liger
object. The raw counts data is presented as a single matrix with all
batches merged. In this case, we call <code><a href="../reference/as.liger.html">as.liger()</a></code> function
instead of using <code><a href="../reference/createLiger.html">createLiger()</a></code> which is defined for
creating liger object from a list of matrices, each a batch. Please view
our <a href="import_data.html">guide for importing data</a> for more
guide in detail.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Construct liger object, from the sparse matrix containing all batches</span></span>
<span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.liger.html">as.liger</a></span><span class="op">(</span><span class="va">h5ad.raw</span>, datasetVar <span class="op">=</span> <span class="va">pancreas.meta</span><span class="op">$</span><span class="va">dataset</span><span class="op">)</span></span></code></pre></div>
<p>Lastly, insert other necessary metadata and the previously selected
HVGs into the liger object. Note that we assigned the
<code>rownames</code></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Match the metadata with the liger object</span></span>
<span><span class="va">pancreas.meta.ordered</span> <span class="op">&lt;-</span> <span class="va">pancreas.meta</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># Insert metadata into liger object</span></span>
<span><span class="va">lig</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">&lt;-</span> <span class="va">pancreas.meta.ordered</span><span class="op">$</span><span class="va">cell_type</span></span>
<span><span class="fu"><a href="../reference/liger-class.html">defaultCluster</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"cell_type"</span></span>
<span></span>
<span><span class="co"># Load previously identified HVGs and set to liger object</span></span>
<span><span class="va">hvg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="st">"pancreas_hvgs.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/liger-class.html">varFeatures</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">hvg</span></span>
<span></span>
<span><span class="va">lig</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class liger with 16382 cells</span></span>
<span><span class="co">## datasets(9): celseq (1004 cells), celseq2 (2285 cells), fluidigmc1 (638 cells), ..., smartseq2 (2394 cells) </span></span>
<span><span class="co">## cellMeta(8): dataset, barcode, nUMI, ..., cell_type </span></span>
<span><span class="co">## varFeatures(2000): SPP1, PPY, RBP4, ..., SYN1 </span></span>
<span><span class="co">## dimReds(0):</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="run-liger-integration">Run liger integration<a class="anchor" aria-label="anchor" href="#run-liger-integration"></a>
</h2>
<p>The following code chunk simply run the LIGER integration pipeline.
For a more detailed guidance on running LIGER, please refer to our <a href="Integrating_multi_scRNA_data.html">basic tutorial</a>. A few notes
to mention here. We <strong>don’t</strong> run
<code><a href="../reference/selectGenes.html">selectGenes()</a></code> function as we have already selected HVGs.
For <code><a href="../reference/runIntegration.html">runIntegration()</a></code> function, argument <code>k</code>
represents the number of ranks we factorize the batches into. These can
be interpreted as metagenes that represents biological factors. The
number of factors is arbitrarily chosen and can be adjusted based on the
dataset. Higher <code>k</code> may be selected when complex composition
is expected from the whole dataset. Lastly, we need to set
<code>method = "centroidAlign"</code> in <code><a href="../reference/alignFactors.html">alignFactors()</a></code>
function to use the new alignment method.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="va">lig</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/runIntegration.html">runIntegration</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">40</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/alignFactors.html">alignFactors</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"centroidAlign"</span><span class="op">)</span></span></code></pre></div>
<p>After the integration, we can optionally visualize the data by
creating UMAP dimensionality reduction embedding using the integrated
data. And color the UMAP either by batch source or by cell type
annotation.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">lig</span>, minDist <span class="op">=</span> <span class="fl">.3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>ligerDotSize <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDimRed.html">plotDatasetDimRed</a></span><span class="op">(</span></span>
<span>    <span class="va">lig</span>, </span>
<span>    colorValues <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>         <span class="st">"#BD0026"</span>, <span class="st">"#FC4E2A"</span>,                       <span class="co"># For celseq and celseq2,</span></span>
<span>         <span class="st">"#78C679"</span>,                                  <span class="co"># For fluidigmc1</span></span>
<span>         <span class="st">"#045A8D"</span>, <span class="st">"#0570B0"</span>, <span class="st">"#3690C0"</span>, <span class="st">"#74A9CF"</span>, <span class="co"># For inDrop1-4</span></span>
<span>         <span class="st">"#AA47B9"</span>,                                  <span class="co"># For smarter</span></span>
<span>         <span class="st">"#FE9929"</span>                                   <span class="co"># For smartseq2</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDimRed.html">plotClusterDimRed</a></span><span class="op">(</span><span class="va">lig</span>, legendNCol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="benchmark_files/figure-html/unnamed-chunk-11-1.png" width="1344" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="export-to-h5ad">Export to H5AD<a class="anchor" aria-label="anchor" href="#export-to-h5ad"></a>
</h2>
<p>In order to utilize <em>scib</em> prepared metric calculation
functions, we need to export the result from the liger object to an H5AD
file. This functionality has been implemented in a single function call
to <code>ligerToH5AD()</code> which is <strong>not part of
<em>rliger</em></strong> package, but need to be loaded via source code.
For usage of the function, please call <code>help.ligerToH5AD()</code>
after loading the source code.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the function from source code</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/ligerToH5AD.R"</span>, package <span class="op">=</span> <span class="st">"rliger"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Call the loaded function</span></span>
<span><span class="fu">ligerToH5AD</span><span class="op">(</span><span class="va">lig</span>, <span class="st">"human_pancreas_liger_integrated.h5ad"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="run-scib-metrics">Run SCIB metrics<a class="anchor" aria-label="anchor" href="#run-scib-metrics"></a>
</h2>
<p>Back to a Python session, we can now work on the integrated data and
start to calculate all SCIB metrics.</p>
<p>First, import necessary modules and load the integrated data.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="co"># Python code</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="im">import</span> scib</span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">"human_pancreas_liger_integrated.h5ad"</span>)</span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a>adata</span></code></pre></div>
<pre><code><span><span class="co">## AnnData object with n_obs × n_vars = 16382 × 2000</span></span>
<span><span class="co">##     obs: 'dataset', 'barcode', 'nUMI', 'nGene', 'mito', 'ribo', 'hemo', 'cell_type'</span></span>
<span><span class="co">##     var: 'name'</span></span>
<span><span class="co">##     obsm: 'X_h', 'X_h_norm', 'X_umap'</span></span>
<span><span class="co">##     varm: 'W'</span></span></code></pre>
<p>Prespecify a few variables that will be re-used as parameters for
many of the metric calculation functions. <code>label_key</code> refers
to the column name in <code>adata.obs</code> for the ground truth cell
type annotation. We simply use what was provided in the original data,
as this has been curated by the original SCIB work.
<code>batch_key</code> is the column in <code>adata.obs</code> that
represents the batch information. <code>use_rep</code> and
<code>embed</code> are the same thing and they refer to the
low-dimensional representation of the integrated data, located in
<code>adata.obsm</code>.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="co"># Python code</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>label_key <span class="op">=</span> <span class="st">"cell_type"</span></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>batch_key <span class="op">=</span> <span class="st">"dataset"</span></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>use_rep <span class="op">=</span> <span class="st">"X_h_norm"</span></span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>embed <span class="op">=</span> <span class="st">"X_h_norm"</span></span></code></pre></div>
<p>One more preprocessing that is crucial to calculating part of the
metrics is to obtain a clustering result that is based on the
integration. <em>scib</em> provides a optimized clustering method that
tries a sequence of resolutions and returns the clustering result with
the highest NMI score compared to the ground truth. An NMI score
evalutates the agreement between two categorical variables and will be
introduced later.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="co"># Python code</span></span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a>sc.pp.neighbors(adata, use_rep<span class="op">=</span>use_rep)</span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a>scib.metrics.cluster_optimal_resolution(adata, label_key<span class="op">=</span>label_key, cluster_key<span class="op">=</span><span class="st">'leiden'</span>)</span></code></pre></div>
<pre><code><span><span class="co">## resolution: 0.1, nmi: 0.880045267743085</span></span>
<span><span class="co">## resolution: 0.2, nmi: 0.8888347309477597</span></span>
<span><span class="co">## resolution: 0.3, nmi: 0.87768954528099</span></span>
<span><span class="co">## resolution: 0.4, nmi: 0.8485513798621696</span></span>
<span><span class="co">## resolution: 0.5, nmi: 0.8085522042477449</span></span>
<span><span class="co">## resolution: 0.6, nmi: 0.7474496433636655</span></span>
<span><span class="co">## resolution: 0.7, nmi: 0.739544088692595</span></span>
<span><span class="co">## resolution: 0.8, nmi: 0.7305861030158917</span></span>
<span><span class="co">## resolution: 0.9, nmi: 0.7189902627405381</span></span>
<span><span class="co">## resolution: 1.0, nmi: 0.7002328173180308</span></span>
<span><span class="co">## resolution: 1.1, nmi: 0.6926906848054561</span></span>
<span><span class="co">## resolution: 1.2, nmi: 0.6745390302067559</span></span>
<span><span class="co">## resolution: 1.3, nmi: 0.6771536766291232</span></span>
<span><span class="co">## resolution: 1.4, nmi: 0.6632363867725439</span></span>
<span><span class="co">## resolution: 1.5, nmi: 0.6576928612452401</span></span>
<span><span class="co">## resolution: 1.6, nmi: 0.6397036519791898</span></span>
<span><span class="co">## resolution: 1.7, nmi: 0.6404473674684553</span></span>
<span><span class="co">## resolution: 1.8, nmi: 0.6437690859294083</span></span>
<span><span class="co">## resolution: 1.9, nmi: 0.6295050082838636</span></span>
<span><span class="co">## resolution: 2.0, nmi: 0.6302577443586709</span></span>
<span><span class="co">## optimised clustering against cell_type</span></span>
<span><span class="co">## optimal cluster resolution: 0.2</span></span>
<span><span class="co">## optimal score: 0.8888347309477597</span></span></code></pre>
<p>Lastly, we also need the original unintegrated data, since some of
the metrics works by comparing the integrated data against it. We load
the data into another AnnData object and do some cleaning to match the
integrated data.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="co"># Python code</span></span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>adata_unintegrated <span class="op">=</span> sc.read_h5ad(<span class="st">"human_pancreas_norm_complexBatch.h5ad"</span>)</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a></span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a><span class="co"># Update `adata_unintegrated` .obs_name, by concatenating `adata_unintegrated.obs['tech']` and `adata_unintegrated.obs_names`, sep with '_'</span></span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a>new_obs_names <span class="op">=</span> []</span>
<span id="cb22-6"><a href="#cb22-6" tabindex="-1"></a><span class="cf">for</span> tech, obs_name <span class="kw">in</span> <span class="bu">zip</span>(adata_unintegrated.obs[<span class="st">'tech'</span>], adata_unintegrated.obs_names):</span>
<span id="cb22-7"><a href="#cb22-7" tabindex="-1"></a>    new_obs_names.append(tech <span class="op">+</span> <span class="st">'_'</span> <span class="op">+</span> obs_name)</span>
<span id="cb22-8"><a href="#cb22-8" tabindex="-1"></a></span>
<span id="cb22-9"><a href="#cb22-9" tabindex="-1"></a>adata_unintegrated.obs_names <span class="op">=</span> new_obs_names</span>
<span id="cb22-10"><a href="#cb22-10" tabindex="-1"></a></span>
<span id="cb22-11"><a href="#cb22-11" tabindex="-1"></a><span class="co"># Reorder `adata_unintegrated` to match `adata`</span></span>
<span id="cb22-12"><a href="#cb22-12" tabindex="-1"></a>adata_unintegrated <span class="op">=</span> adata_unintegrated[adata.obs_names,]</span>
<span id="cb22-13"><a href="#cb22-13" tabindex="-1"></a></span>
<span id="cb22-14"><a href="#cb22-14" tabindex="-1"></a><span class="co"># Rename columns in `adata_unintegrated.obs` to match `adata.obs`</span></span>
<span id="cb22-15"><a href="#cb22-15" tabindex="-1"></a>adata_unintegrated.obs <span class="op">=</span> adata_unintegrated.obs.rename(columns<span class="op">=</span>{<span class="st">'tech'</span>: <span class="st">'dataset'</span>, <span class="st">'celltype'</span>: <span class="st">'cell_type'</span>})</span></code></pre></div>
<p>Now we can start to compute each metrics using functions prepared
within <em>scib</em> package. We briefly describe what each metric means
and how it is calculated. For more details of them, please refer to <a href="https://www.nature.com/articles/s41592-021-01336-8#Sec14" class="external-link">the
Method section of the original SCIB paper</a>.</p>
<div class="section level3">
<h3 id="metrics-for-batch-effect-removal">Metrics for batch effect removal<a class="anchor" aria-label="anchor" href="#metrics-for-batch-effect-removal"></a>
</h3>
<div class="section level4">
<h4 id="batch-pcr">Batch PCR<a class="anchor" aria-label="anchor" href="#batch-pcr"></a>
</h4>
<p>PCR (principal component regression) method calculates the <span class="math inline">\(R^2\)</span> from a linear regression of the batch
variable onto each principal component. By multiplying the <span class="math inline">\(R^2\)</span> with the variance explained for each
principal component, PCR can then derive the total variance explained by
the batch variable.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a>pancreas_pcr_batch <span class="op">=</span> scib.metrics.pcr_comparison(</span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>    adata_pre<span class="op">=</span>adata_unintegrated, </span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a>    adata_post<span class="op">=</span>adata, </span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a>    covariate<span class="op">=</span>batch_key, </span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a>    embed<span class="op">=</span>embed</span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a>)</span>
<span id="cb23-7"><a href="#cb23-7" tabindex="-1"></a>pancreas_pcr_batch</span></code></pre></div>
<pre><code><span><span class="co">## 0.9931731661673961</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="batch-asw">Batch ASW<a class="anchor" aria-label="anchor" href="#batch-asw"></a>
</h4>
<p>ASW (average silhouette width) is generally a metric that measures
within-group distances of a cell and the between-group distances of that
cell to the closest group. Averaging over all silhouette widths of a set
of cells yields the ASW, so that a higher ASW indicates better
separation and lower ASW indicates stronger mixing. By grouping cells by
their batch information within each cell type, reversed ASW scores can
be utilized to evaluate how well the batch labels are mixed.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a>pancreas_batch_asw <span class="op">=</span> scib.metrics.silhouette_batch(adata<span class="op">=</span>adata, batch_key<span class="op">=</span>batch_key, label_key<span class="op">=</span>label_key, embed<span class="op">=</span>embed)</span></code></pre></div>
<pre><code><span><span class="co">## mean silhouette per group:                     silhouette_score</span></span>
<span><span class="co">## group                               </span></span>
<span><span class="co">## acinar                      0.817259</span></span>
<span><span class="co">## activated_stellate          0.881205</span></span>
<span><span class="co">## alpha                       0.906360</span></span>
<span><span class="co">## beta                        0.908556</span></span>
<span><span class="co">## delta                       0.892427</span></span>
<span><span class="co">## ductal                      0.868090</span></span>
<span><span class="co">## endothelial                 0.807412</span></span>
<span><span class="co">## epsilon                     0.614904</span></span>
<span><span class="co">## gamma                       0.909407</span></span>
<span><span class="co">## macrophage                  0.709292</span></span>
<span><span class="co">## mast                        0.702113</span></span>
<span><span class="co">## quiescent_stellate          0.741880</span></span>
<span><span class="co">## schwann                     0.704278</span></span>
<span><span class="co">## t_cell                      0.644153</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a>pancreas_batch_asw</span></code></pre></div>
<pre><code><span><span class="co">## 0.7933811059905438</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="graph-ilisi">Graph iLISI<a class="anchor" aria-label="anchor" href="#graph-ilisi"></a>
</h4>
<p>LISI (local inverse Simpson index) scores are computed from
neighborhood lists per node from integrated kNN graphs. It counts the
number of unique labels in the neighborhood of a cell, so that a number
closer to total number of labels indicates better mixing and 1 indicates
perfect separation. By looking at the the batch labels, iLISI scores can
be used to evaluate how well the batch labels are mixed.</p>
<p>Note that since the final iLISI score is averaged for all cells,
integration tasks that contain batches with distinct cell type
composition might not necessarily obtain a high iLISI score.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a>pancreas_graph_ilisi <span class="op">=</span> scib.metrics.ilisi_graph(adata<span class="op">=</span>adata, batch_key<span class="op">=</span>batch_key, type_<span class="op">=</span><span class="st">"embed"</span>, use_rep<span class="op">=</span>use_rep, n_cores<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a>pancreas_graph_ilisi</span></code></pre></div>
<pre><code><span><span class="co">## 0.3073894205861533</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="graph-connectivity">Graph connectivity<a class="anchor" aria-label="anchor" href="#graph-connectivity"></a>
</h4>
<p>This method first subsets the kNN graph of the integrated data to
contain only cells of one cell type, and then compares the size of the
largest connected component against the size of the whole subset graph.
So that a higher connectivity score indicates better mixing with each
cell type.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a>pancreas_graph_connectivity <span class="op">=</span> scib.metrics.graph_connectivity(adata<span class="op">=</span>adata, label_key<span class="op">=</span>label_key)</span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a>pancreas_graph_connectivity</span></code></pre></div>
<pre><code><span><span class="co">## 0.9750180591653</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="kbet">kBET<a class="anchor" aria-label="anchor" href="#kbet"></a>
</h4>
<p>kBET (<a href="https://www.nature.com/articles/s41592-018-0254-1" class="external-link">Büttner, M.
2019</a>) determines whether the label composition of the neighborhood
of a cell is similar to the expected (global) label composition.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a>pancreas_kBET <span class="op">=</span> scib.metrics.kBET(adata<span class="op">=</span>adata, batch_key<span class="op">=</span>batch_key, label_key<span class="op">=</span>label_key, type_<span class="op">=</span><span class="st">'embed'</span>, embed<span class="op">=</span>embed)</span></code></pre></div>
<pre><code><span><span class="co">## R was initialized outside of rpy2 (R_NilValue != NULL). Trying to use it nevertheless.</span></span>
<span><span class="co">## Adding diffusion to step 4</span></span>
<span><span class="co">## Adding diffusion to step 4</span></span>
<span><span class="co">## Adding diffusion to step 4</span></span>
<span><span class="co">## Adding diffusion to step 5</span></span>
<span><span class="co">## Adding diffusion to step 6</span></span>
<span><span class="co">## t_cell consists of a single batch or is too small. Skip.</span></span></code></pre>
<div class="sourceCode" id="cb35"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a>pancreas_kBET</span></code></pre></div>
<pre><code><span><span class="co">## 0.5248958996458675</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="metrics-for-bio-conservation">Metrics for bio conservation<a class="anchor" aria-label="anchor" href="#metrics-for-bio-conservation"></a>
</h3>
<div class="section level4">
<h4 id="nmi-and-ari">NMI and ARI<a class="anchor" aria-label="anchor" href="#nmi-and-ari"></a>
</h4>
<p>NMI (normalized mutual information) and ARI (adjusted Rand index)
both compares the overlap of two clusterings. Here we compare the
clustering result we previously obtained from the integrated data
against the cell type labeling, with higher scores indicating better
agreement.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a>pancreas_nmi <span class="op">=</span> scib.metrics.nmi(adata<span class="op">=</span>adata, cluster_key<span class="op">=</span><span class="st">'leiden'</span>, label_key<span class="op">=</span>label_key)</span>
<span id="cb37-2"><a href="#cb37-2" tabindex="-1"></a>pancreas_nmi</span></code></pre></div>
<pre><code><span><span class="co">## 0.8888347309477597</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a>pancreas_ari <span class="op">=</span> scib.metrics.ari(adata<span class="op">=</span>adata, cluster_key<span class="op">=</span><span class="st">'leiden'</span>, label_key<span class="op">=</span>label_key)</span>
<span id="cb39-2"><a href="#cb39-2" tabindex="-1"></a>pancreas_ari</span></code></pre></div>
<pre><code><span><span class="co">## 0.9284316085505634</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="cell-type-asw">Cell type ASW<a class="anchor" aria-label="anchor" href="#cell-type-asw"></a>
</h4>
<p>Recall that we used a reversed ASW measurement for batch effect
removal, here we group cells by the cell type labels. Therefore, a
higher ASW indicates the original cell types can be properly separated
after integration.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" tabindex="-1"></a>pancreas_cell_type_asw <span class="op">=</span> scib.metrics.silhouette(adata<span class="op">=</span>adata, label_key<span class="op">=</span>label_key, embed<span class="op">=</span>embed)</span>
<span id="cb41-2"><a href="#cb41-2" tabindex="-1"></a>pancreas_cell_type_asw</span></code></pre></div>
<pre><code><span><span class="co">## 0.6222552433843831</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="f1-score-for-isolated-label">F1 score for isolated label<a class="anchor" aria-label="anchor" href="#f1-score-for-isolated-label"></a>
</h4>
<p>To evaluate how well the integration methods deal with cell types
shared by few batches, the SCIB work developed several metrics that
focus on the isolated labels. The isolated label is defined as the label
that presents in the least number of batches.</p>
<p>F1 score is a weighted mean of precision and recall of a clustering.
The cluster assignment of the isolated label is optimized for a range a
resolutions and the optimal F1 score for the isolated label is reported.
A value of 1 1 shows that all of the isolated label cells and no others
are captured in a cluster.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" tabindex="-1"></a>pancreas_isolated_label_F1 <span class="op">=</span> scib.metrics.isolated_labels_f1(adata<span class="op">=</span>adata, label_key<span class="op">=</span>label_key, batch_key<span class="op">=</span>batch_key, embed<span class="op">=</span>embed)</span></code></pre></div>
<pre><code><span><span class="co">## isolated labels: no more than 4 batches per label</span></span>
<span><span class="co">## Compute neighbors on rep X_h_norm</span></span>
<span><span class="co">## t_cell: 0.5185185185185185</span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" tabindex="-1"></a>pancreas_isolated_label_F1</span></code></pre></div>
<pre><code><span><span class="co">## 0.5185185185185185</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="asw-for-isolated-label">ASW for isolated label<a class="anchor" aria-label="anchor" href="#asw-for-isolated-label"></a>
</h4>
<p>For the isolated label as defined in the last metric, the ASW score
for a cell is calculated by comparing its distances within the isolated
label versus in the non-isolated labels. After scaling, a higher
isolated label ASW score indicates better separation of the isolated
label from the rest of the data.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" tabindex="-1"></a>pancreas_isolated_label_silhouette <span class="op">=</span> scib.metrics.isolated_labels_asw(adata<span class="op">=</span>adata, label_key<span class="op">=</span>label_key, batch_key<span class="op">=</span>batch_key, embed<span class="op">=</span>embed)</span></code></pre></div>
<pre><code><span><span class="co">## isolated labels: no more than 4 batches per label</span></span>
<span><span class="co">## t_cell: 0.7445839795169722</span></span></code></pre>
<div class="sourceCode" id="cb49"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" tabindex="-1"></a>pancreas_isolated_label_silhouette</span></code></pre></div>
<pre><code><span><span class="co">## 0.7445839795169722</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="graph-clisi">Graph cLISI<a class="anchor" aria-label="anchor" href="#graph-clisi"></a>
</h4>
<p>Recall that we used iLISI for batch effect removal, here we use look
at the cell type labeling within the neighborhood of each cell. While
less unique labels in the neighbor hood indicates better separation of
cell types, the score is then scaled in a way that a higher cLISI score
indicates better separation of cell types.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" tabindex="-1"></a>pancreas_graph_clisi <span class="op">=</span> scib.metrics.clisi_graph(adata, label_key<span class="op">=</span>label_key, type_<span class="op">=</span><span class="st">'embed'</span>, use_rep<span class="op">=</span>use_rep, n_cores<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb51-2"><a href="#cb51-2" tabindex="-1"></a>pancreas_graph_clisi</span></code></pre></div>
<pre><code><span><span class="co">## 0.9995901383991688</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="cell-cycle-conservation">Cell cycle conservation<a class="anchor" aria-label="anchor" href="#cell-cycle-conservation"></a>
</h4>
<p>A cell-cycle score is first computed for each respective cell-cycle
phase and for each cell. Then for each batch, the method computes the
variance contribution of the resulting S and G2/M phase scores using
PCR. The differences in variance before and after integration were
aggregated into a final score.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" tabindex="-1"></a>pancreas_cell_cycle_conservation <span class="op">=</span> scib.metrics.cell_cycle(adata_pre<span class="op">=</span>adata_unintegrated, adata_post<span class="op">=</span>adata, batch_key<span class="op">=</span>batch_key, embed<span class="op">=</span>embed, organism<span class="op">=</span><span class="st">'human'</span>)</span>
<span id="cb53-2"><a href="#cb53-2" tabindex="-1"></a>pancreas_cell_cycle_conservation</span></code></pre></div>
<pre><code><span><span class="co">## 0.4774870990333798</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="hvg-conservation">HVG conservation<a class="anchor" aria-label="anchor" href="#hvg-conservation"></a>
</h4>
<p>By comparing the overlap between highly variable genes (HVGs) that
can be identified from unintegrated data and those identified from
batch-corrected data, the HVG conservation score can servea as a proxy
for the preservation of the biological signal. However, this metric can
only be applied for integration methods that return feature expression
matrices.</p>
</div>
</div>
<div class="section level3">
<h3 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h3>
<p>Now that we’ve calculated all the metrics, we can print them out to
have a look and save them to a CSV file for further analysis.</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" tabindex="-1"></a><span class="co"># Python code</span></span>
<span id="cb55-2"><a href="#cb55-2" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb55-3"><a href="#cb55-3" tabindex="-1"></a></span>
<span id="cb55-4"><a href="#cb55-4" tabindex="-1"></a>pancreas_all_metrics <span class="op">=</span> {</span>
<span id="cb55-5"><a href="#cb55-5" tabindex="-1"></a>    <span class="st">"pancreas_pcr_batch"</span>: pancreas_pcr_batch,</span>
<span id="cb55-6"><a href="#cb55-6" tabindex="-1"></a>    <span class="st">"pancreas_batch_asw"</span>: pancreas_batch_asw,</span>
<span id="cb55-7"><a href="#cb55-7" tabindex="-1"></a>    <span class="st">"pancreas_graph_ilisi"</span>: pancreas_graph_ilisi,</span>
<span id="cb55-8"><a href="#cb55-8" tabindex="-1"></a>    <span class="st">"pancreas_graph_connectivity"</span>: pancreas_graph_connectivity,</span>
<span id="cb55-9"><a href="#cb55-9" tabindex="-1"></a>    <span class="st">"pancreas_kBET"</span>: pancreas_kBET,</span>
<span id="cb55-10"><a href="#cb55-10" tabindex="-1"></a>    <span class="st">"pancreas_nmi"</span>: pancreas_nmi,</span>
<span id="cb55-11"><a href="#cb55-11" tabindex="-1"></a>    <span class="st">"pancreas_ari"</span>: pancreas_ari,</span>
<span id="cb55-12"><a href="#cb55-12" tabindex="-1"></a>    <span class="st">"pancreas_cell_type_asw"</span>: pancreas_cell_type_asw,</span>
<span id="cb55-13"><a href="#cb55-13" tabindex="-1"></a>    <span class="st">"pancreas_isolated_label_F1"</span>: pancreas_isolated_label_F1,</span>
<span id="cb55-14"><a href="#cb55-14" tabindex="-1"></a>    <span class="st">"pancreas_isolated_label_silhouette"</span>: pancreas_isolated_label_silhouette,</span>
<span id="cb55-15"><a href="#cb55-15" tabindex="-1"></a>    <span class="st">"pancreas_graph_clisi"</span>: pancreas_graph_clisi,</span>
<span id="cb55-16"><a href="#cb55-16" tabindex="-1"></a>    <span class="st">"pancreas_cell_cycle_conservation"</span>: pancreas_cell_cycle_conservation</span>
<span id="cb55-17"><a href="#cb55-17" tabindex="-1"></a>}</span>
<span id="cb55-18"><a href="#cb55-18" tabindex="-1"></a><span class="cf">for</span> metric, score <span class="kw">in</span> pancreas_all_metrics.items():</span>
<span id="cb55-19"><a href="#cb55-19" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss">: </span><span class="sc">{</span>score<span class="sc">}</span><span class="ss">"</span>)</span></code></pre></div>
<pre><code><span><span class="co">## pancreas_pcr_batch: 0.9931731661673961</span></span>
<span><span class="co">## pancreas_batch_asw: 0.7933811059905438</span></span>
<span><span class="co">## pancreas_graph_ilisi: 0.3073894205861533</span></span>
<span><span class="co">## pancreas_graph_connectivity: 0.9750180591653</span></span>
<span><span class="co">## pancreas_kBET: 0.5248958996458675</span></span>
<span><span class="co">## pancreas_nmi: 0.8888347309477597</span></span>
<span><span class="co">## pancreas_ari: 0.9284316085505634</span></span>
<span><span class="co">## pancreas_cell_type_asw: 0.6222552433843831</span></span>
<span><span class="co">## pancreas_isolated_label_F1: 0.5185185185185185</span></span>
<span><span class="co">## pancreas_isolated_label_silhouette: 0.7445839795169722</span></span>
<span><span class="co">## pancreas_graph_clisi: 0.9995901383991688</span></span>
<span><span class="co">## pancreas_cell_cycle_conservation: 0.4774870990333798</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" tabindex="-1"></a><span class="co"># Python code</span></span>
<span id="cb57-2"><a href="#cb57-2" tabindex="-1"></a>pancreas_metrics_df <span class="op">=</span> pd.DataFrame(pancreas_all_metrics, index<span class="op">=</span>[<span class="dv">0</span>])</span>
<span id="cb57-3"><a href="#cb57-3" tabindex="-1"></a>pancreas_metrics_df.to_csv(<span class="st">"pancreas_liger_scores.csv"</span>, index<span class="op">=</span><span class="va">False</span>)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="ranking-with-existing-scores">Ranking with existing scores<a class="anchor" aria-label="anchor" href="#ranking-with-existing-scores"></a>
</h2>
<p>In the original work of SCIB, the scores for a single method may not
be directly useful to see how well the method overally performs compared
to others. Therefore, the authors proposed a way to scale the scores and
calculate an overall score with all methods’ scores together. The
overall score is calculated as a weighted sum of the batch correction
and bio conservation scores. The batch correction score and the bio
conservation score are calculated as the average of the scaled scores of
each single metrics that belong to the category. The authors applied
min-max scaling to the scores of each metric, so that the scores are in
the range of 0 to 1. In this way, the difference of performance between
methods can be better reflected.</p>
<p><span class="math display">\[
\begin{aligned}
score^*_{i,j} &amp;= \frac{score_{i,j} -
min(score_{,j})}{max(score_{,j}) - min(score_{,j})} \\
batch_{i} &amp;= \frac{1}{|BATCH|}\sum_{j}^{BATCH} score^*_{i,j} \\
bio_{i} &amp;= \frac{1}{|BIO|}\sum_{j}^{BIO} score^*_{i,j} \\
Overall_i &amp;= 0.6 \times bio_{i} + 0.4 \times batch_{i}
\end{aligned}
\]</span></p>
<p>Assume we put the scores for all methods and all metrics in a matrix,
where a row is for all metrics of a method, and a column is the scores
of a metric listed with all methods. The above formula can be applied to
calculate the overall score for each method. <span class="math inline">\(score\)</span> is the original score matrix, <span class="math inline">\(i\)</span> is the row index for a method, <span class="math inline">\(j\)</span> is the column index for a metric, <span class="math inline">\(score^*_{i,j}\)</span> is the scaled score for
method <span class="math inline">\(i\)</span> and metric <span class="math inline">\(j\)</span>. <span class="math inline">\(BATCH\)</span> is the set of batch correction
metrics, <span class="math inline">\(BIO\)</span> is the set of bio
conservation metrics. <span class="math inline">\(|BATCH|\)</span> is
the number of batch correction metrics we have in total, for example, 5
for what <em>scib</em> currently provides. Similarly, <span class="math inline">\(|BIO|\)</span> is the number of bio conservation
metrics, 8 in our case. <span class="math inline">\(batch_{i}\)</span>
is the batch correction score for method <span class="math inline">\(i\)</span>, <span class="math inline">\(bio_{i}\)</span> is the bio conservation score for
method <span class="math inline">\(i\)</span>, and <span class="math inline">\(Overall_i\)</span> is the overall score for method
<span class="math inline">\(i\)</span>.</p>
<div class="section level3">
<h3 id="load-all-other-scores">Load all other scores<a class="anchor" aria-label="anchor" href="#load-all-other-scores"></a>
</h3>
<p>Based on what we described above, we need to load the scores the SCIB
work already calculated and insert our new scores into the pool to have
all methods re-ranked.</p>
<p>The full benchmarking scores for the human pancreas dataset task is
available in <a href="https://www.nature.com/articles/s41592-021-01336-8#Sec35" class="external-link">the
Supplementary Information of SCIB’s orginal publication</a>, freely
available as Supplementary Data 1.</p>
<blockquote>
<p>Click to download Supplementary Data 1 and <strong>save to working
directory</strong>: <a href="https://static-content.springer.com/esm/art%3A10.1038%2Fs41592-021-01336-8/MediaObjects/41592_2021_1336_MOESM3_ESM.xlsx" class="external-link"><strong>41592_2021_1336_MOESM3_ESM.xlsx</strong></a></p>
</blockquote>
<p>Now we need to load a few more R libraries for data manipulation.
Given that the supplementary data provided is an Excel file with
multiple sheets, each sheet is for a different dataset task. We need to
also locate which sheet is for the human pancreas dataset task.
<code><a href="https://readxl.tidyverse.org/reference/excel_sheets.html" class="external-link">excel_sheets()</a></code> function from <em>readxl</em> package can
help us with this.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readxl.tidyverse.org" class="external-link">readxl</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://readxl.tidyverse.org/reference/excel_sheets.html" class="external-link">excel_sheets</a></span><span class="op">(</span><span class="st">"41592_2021_1336_MOESM3_ESM.xlsx"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Batch_vs_Bio_RNA_tasks"         "immune_cell_hum"               </span></span>
<span><span class="co">##  [3] "simulations_1_1"                "mouse_brain_atac_peaks_large"  </span></span>
<span><span class="co">##  [5] "mouse_brain_atac_peaks_small"   "mouse_brain_atac_genes_large"  </span></span>
<span><span class="co">##  [7] "mouse_brain_atac_windows_small" "mouse_brain_atac_windows_large"</span></span>
<span><span class="co">##  [9] "mouse_brain_atac_genes_small"   "pancreas"                      </span></span>
<span><span class="co">## [11] "mouse_brain"                    "mouse_brain_atac_large"        </span></span>
<span><span class="co">## [13] "mouse_brain_atac_small"         "lung_atlas"                    </span></span>
<span><span class="co">## [15] "simulations_2"                  "immune_cell_hum_mou"</span></span></code></pre>
<p>The scores for the human pancreas dataset task is stored in sheet 10
from listed above. Now we can load the sheet into one
<code>data.frame</code> object in R. Note that there are <code>NA</code>
values in the result, mainly due to the fact that some metrics are not
applicable to some methods. For example, the HVG conservation metric is
not applicable to LIGER. Be sure to tell the <code><a href="https://readxl.tidyverse.org/reference/read_excel.html" class="external-link">read_excel()</a></code>
function how to handle these values.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pancreas_scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html" class="external-link">read_excel</a></span><span class="op">(</span></span>
<span>    path <span class="op">=</span> <span class="st">"41592_2021_1336_MOESM3_ESM.xlsx"</span>, </span>
<span>    sheet <span class="op">=</span> <span class="fl">10</span>, </span>
<span>    na <span class="op">=</span> <span class="st">"NA"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">pancreas_scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">pancreas_scores</span><span class="op">)</span></span></code></pre></div>
<p>Here is how the table looks like:</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69"],["pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas"],["Harmony","Seurat v3 RPCA","Harmony","Seurat v3 CCA","fastMNN","scGen*","scGen*","Seurat v3 RPCA","scGen*","Seurat v3 CCA","scANVI*","fastMNN","fastMNN","BBKNN","Seurat v3 CCA","scVI","Scanorama","Harmony","Seurat v3 RPCA","trVAE","Scanorama","Seurat v3 RPCA","fastMNN","Harmony","BBKNN","Seurat v3 CCA","ComBat","scVI","Scanorama","fastMNN","scANVI*","Scanorama","ComBat","BBKNN","Conos","MNN","Conos","Conos","fastMNN","MNN","fastMNN","Conos","BBKNN","fastMNN","ComBat","Scanorama","ComBat","Scanorama","Scanorama","DESC","MNN","SAUCIE","SAUCIE","DESC","trVAE","DESC","DESC","SAUCIE","SAUCIE","LIGER","MNN","Scanorama","LIGER","Unintegrated","SAUCIE","SAUCIE","scGen*","SAUCIE","SAUCIE"],["embed","gene","embed","gene","embed","gene","gene","gene","gene","gene","embed","gene","embed","graph","gene","embed","gene","embed","gene","embed","embed","gene","gene","embed","graph","gene","gene","embed","embed","embed","embed","embed","gene","graph","graph","gene","graph","graph","embed","gene","gene","graph","graph","gene","gene","embed","gene","gene","gene","embed","gene","gene","embed","embed","embed","embed","embed","embed","gene","embed","gene","gene","embed","gene","embed","gene","gene","embed","gene"],["HVG","HVG","HVG","HVG","HVG","HVG","HVG","HVG","FULL","HVG","HVG","HVG","HVG","HVG","FULL","HVG","HVG","FULL","FULL","HVG","HVG","FULL","HVG","FULL","FULL","FULL","HVG","FULL","FULL","FULL","FULL","HVG","HVG","HVG","FULL","HVG","FULL","HVG","FULL","HVG","FULL","HVG","FULL","FULL","FULL","FULL","FULL","FULL","HVG","HVG","FULL","HVG","HVG","HVG","FULL","FULL","FULL","HVG","HVG","HVG","FULL","FULL","FULL","FULL","FULL","FULL","FULL","FULL","FULL"],["unscaled","scaled","scaled","scaled","scaled","unscaled","scaled","unscaled","unscaled","unscaled","unscaled","scaled","unscaled","unscaled","scaled","unscaled","scaled","scaled","scaled","unscaled","scaled","unscaled","unscaled","unscaled","unscaled","unscaled","scaled","unscaled","scaled","unscaled","unscaled","unscaled","unscaled","scaled","unscaled","unscaled","scaled","scaled","scaled","scaled","unscaled","unscaled","scaled","scaled","unscaled","unscaled","scaled","scaled","unscaled","unscaled","scaled","unscaled","unscaled","scaled","unscaled","unscaled","scaled","scaled","scaled","unscaled","unscaled","unscaled","unscaled","unscaled","scaled","scaled","scaled","unscaled","unscaled"],[0.7900360679688442,0.781129145712782,0.7746587922502629,0.7709289650483281,0.752087399428877,0.748935268741125,0.7458785649306411,0.741923615586465,0.7348690919539499,0.7279916188864599,0.723884677636787,0.7236909020797591,0.720062557471025,0.716877408319004,0.7156774950994681,0.7134047188995358,0.712660526141463,0.7109212279543679,0.7089738785199889,0.7045645083098591,0.7034837416154639,0.7007014666337851,0.691850644373332,0.6904090332595579,0.690280758696701,0.6896714802526529,0.684223539693684,0.6729758031525142,0.6726879252147939,0.6716565957581488,0.6713116856991459,0.6630131614694179,0.6551205718595791,0.6480344931813979,0.6426782291452789,0.638260231581937,0.637337709026658,0.6346840136326949,0.6345280543046821,0.633429673467783,0.628827999775569,0.6287660074921729,0.6241557408864621,0.603029229588568,0.6003130278293299,0.5870634853127199,0.585429414670055,0.584311694434803,0.579790637676643,0.57324589113112,0.57016860326971,0.558689935045008,0.5573283080592579,0.555621450201603,0.538170454071188,0.53579084462591,0.529183639346401,0.5225152837964759,0.50739293066043,0.479750758417351,0.4784588120968039,0.470742544312876,0.446248226566822,0.431120828284327,0.414497816144272,0.39077737427132,0.3339410784760301,0.324921734186793,0.31356706994327],[0.9250212821004009,0.9150252143186229,0.924731723278992,0.904595650948897,0.9602153908472749,0.8564417051602041,0.899030311582191,0.882995495958573,0.8353590181747,0.8362488044002651,0.823269372366336,0.9616092192682559,0.822177489643386,0.824861908265101,0.863119142008583,0.827936162726147,0.701265633327159,0.859968356631718,0.8310152384303491,0.793508489693577,0.8499982602145458,0.8164413442186491,0.822601431598165,0.79722957473464,0.7802704172236999,0.818570903968977,0.709000812486492,0.7078790378090691,0.7448814135322299,0.688471714732388,0.700382765746179,0.72025849643427,0.70604446294957,0.9039025101426558,0.48106877054924,0.6768971263360761,0.551248676356893,0.537774382599407,0.87914606797565,0.718700931112881,0.6888704683550679,0.451872032418472,0.896405209302126,0.873594695147314,0.6200653769032211,0.592949810243429,0.643076332086611,0.65463423148677,0.640735119932908,0.6087984087888679,0.659826790528718,0.624314919652965,0.625964733888129,0.561363731251603,0.5832455963175179,0.59421979721262,0.5965845142940219,0.598305991790704,0.5983763218240971,0.637409806020193,0.538432687655837,0.477782722530065,0.646623870622931,0.27508540527959,0.484607953466966,0.4860767216544249,0.6232603606013339,0.353536787768401,0.354126947242825],[0.9207507889066251,0.934415099097217,0.9502543092841821,0.9485459054864742,0.992628053680685,0.88418361976556,0.9616585688451948,0.912346604806772,0.898883910134011,0.9191516767860369,0.9174695195723209,0.9926272748537651,0.8499347684495629,null,0.9535777747864591,0.929525587174975,0.917091470045615,0.9647041417429421,0.9407447723807622,0.8835143133559989,0.9363541576606071,0.9267811016886141,0.8499383223813241,0.924309308415366,null,0.931604688484989,1,0.862350614009722,0.9231238712373651,0.664995392543878,0.8592847757856709,0.830140057529084,0.999985097124501,null,null,0.855878437307926,null,null,0.995477577766712,0.931881356544966,0.6649968824489001,null,null,0.9954802244037571,0.999952826970895,0.685529031884872,1,0.8327923651622779,0.8189749642609692,0.8410342129367501,0.9475190930052041,0.7903509111992242,0.7903572966889691,0.8426803885140991,0.649738764344303,0.8869344612178409,0.885149534255942,0.9096096409478579,0.9096113520663041,0.9491740653390609,0.7125446098053699,0.50079641418794,0.9485968294363271,0,0.7742396189704069,0.774238663154949,0.9996110087823841,0.545447833526211,0.5454519414086889],[0.894511369425683,0.830352741727264,0.8682828071792281,0.791238660130418,0.8900114368553191,0.7867018472299409,0.7812526262808991,0.848593198040775,0.7645371062667121,0.7930087360335891,0.8510651793875571,0.8900112438566241,0.8977854125122142,null,0.800717260882654,0.863945401459413,0.7163889233728989,0.8618988047164191,0.8330122782992859,0.872082276249737,0.9132077831855909,0.844258080311638,0.8977854116901529,0.8752655937983671,null,0.793653045863961,0.8629165484350629,0.856719623267588,0.923753592193824,0.8851395475853341,0.8580273944723491,0.923015400485989,0.897589642616812,null,null,0.9057019050549191,null,null,0.8766370081324929,0.891655363823846,0.885139538029635,null,null,0.8766368774402621,0.8773995599969351,0.9075370970552049,0.862004233410549,0.777440726706356,0.7839881551943959,0.520277694088659,0.888926057687407,0.713064112920118,0.7134053300802541,0.463715663608702,0.8593554434528191,0.6682740082372021,0.6872645261494661,0.7205518336529679,0.720167859382302,0.790581911603242,0.8448200803913499,0.8253157844642529,0.802246787096501,0.7702285814202728,0.7518227487864159,0.7516218405720709,0.6867855334771179,0.7407034154344431,0.7404462286383371],[0.367036082891136,0.3843382806103279,0.371651903484589,0.410640737316228,0.397792204350052,0.3868414245549651,0.408808327645455,0.428733003257341,0.349642309418385,0.432617309283705,0.298390493848281,0.401356175881789,0.291011966673681,0.3985925302610949,0.385735692961946,0.304252237012593,0.326013353470828,0.346388322999285,0.32439399417916,0.251749767695298,0.3170986532033739,0.368865149995437,0.291256256294396,0.298450335252708,0.400524797426676,0.3835768488642921,0.096363344854045,0.18144840479935,0.181152893386147,0.157685633682709,0.181487313906364,0.173945575791937,0.10840353385013,0.404664366242914,0,0.07924053136125001,0.031505924490019,0.0032508424915039,0.342386059055253,0.162744822293329,0.158014921893108,0,0.393420430392384,0.333848568323911,0.07773122373431091,0.08280796471408,0.068030329118427,0.207496476358945,0.176542180508463,0.287974637717994,0.085279080319681,0.284840413730502,0.28561878124318,0.241581307255754,0.0054321038700699,0.138642221018394,0.139384329844067,0.311008825147208,0.311268928634215,0.440112154521447,0.004064867867098,0.110113267974042,0.423875104645788,1.39e-16,0.301557735530132,0.299771219728084,0.347378593318301,0.2725262295533989,0.272692316270471],[0.975683999692299,0.9801448072863729,0.970528569381912,0.9826876122830529,0.9859843467077769,0.99151218026432,0.9668994542950159,0.9822476290369929,0.988937932530597,0.9845794790281999,0.9901342649293009,0.9859680975255201,0.979376390999663,0.975431624381728,0.9587294467743058,0.990548145030244,0.9574738019042941,0.9588859051364921,0.9546537986654209,0.9754137804258521,0.9870079187648331,0.9700302922800741,0.9792771069733069,0.954998427344705,0.9322396572839251,0.9672640939169781,0.985943173010672,0.9685477617391931,0.9806229145674791,0.961935902112544,0.9532281842447191,0.9813724345867808,0.97862816408287,0.993727274028197,0.989895857884438,0.9719553452437321,0.9957750369118,0.9959385215446921,0.978249247500479,0.985385865445034,0.9614846903870059,0.989826852289582,0.9832854534661021,0.9754139635264599,0.957349355022956,0.9527370741600258,0.9676092666381819,0.962786669327037,0.958694898692389,0.753305286669516,0.9687006626135541,0.780407164454802,0.780407164454802,0.7942776847012371,0.9789544024133021,0.862306871352902,0.8670100990879219,0.71400294889713,0.71400294889713,0.654438316783443,0.9449895715119619,0.7600025612855831,0.654029948493883,0.7961010610589521,0.575897599398362,0.5816118851126479,0.547182120294373,0.461664788854583,0.461173603849724],[0.5996614124760971,0.6142814460317261,0.612750615378048,0.590520763346827,0.612499215648134,0.523571856669807,0.598024143266249,0.467619053828556,0.532605296187915,0.39863159486603,0.453231066212606,0.611929690780312,0.4518814700998089,0.410417442547723,0.522192465514023,0.437236133016931,0.287857559718044,0.485460006526559,0.485394458578717,0.43919824900201,0.423651260192227,0.374748914523434,0.4527882054357179,0.388823220673077,0.3806229632401189,0.420207115449824,0.342786849255109,0.334852034749818,0.312331400398284,0.38930420378835,0.330336219894599,0.305296785747007,0.288896732617294,0.508585556886334,0.331117056604467,0.319770959195757,0.397540540145497,0.409724340184667,0.488555306102776,0.293141596478703,0.390388008309883,0.285716210455299,0.520307801559875,0.486967461994211,0.145370367678296,0.202691538088101,0.223924808195234,0.276325643337358,0.280516705341841,0.5516811527076221,0.242861322867619,0.378116317832817,0.381091992438199,0.506481373873081,0.316263145552971,0.393344390812548,0.373539306550861,0.273848599583119,0.274156737337568,0.181330287973736,0.218333267419239,0.247330012274959,0.211929947207271,0.138187193126023,0.158938026824837,0.159537658274261,0.449035388027117,0.09518430369088771,0.0972850264515485],[0.7000459252144731,0.6918650999755551,0.674610171564444,0.6818178411146151,0.613335405149944,0.6772643111284061,0.6437774004962751,0.647875695338393,0.6678758078067829,0.6558201618772579,0.6576282144837541,0.565078690620761,0.651985936022785,0.644887741688273,0.6173830638267249,0.637050423015129,0.7202571213509991,0.611556475502801,0.627612971913083,0.6452685207207139,0.605807395882742,0.623541548243875,0.6046834528901109,0.6191953389428361,0.630287653012035,0.603738531108437,0.667705357831812,0.649706980048144,0.6245589330031689,0.66044651644199,0.6519309656677901,0.6248496048261829,0.621171311132918,0.477455815207226,0.7504178682093051,0.612502301745845,0.6947303974731679,0.6992904343215529,0.4714493785240371,0.576582168371051,0.5887996873892359,0.7466953242079741,0.442656095276019,0.4226522525494029,0.587144795113402,0.5831392686922471,0.5469981363923521,0.5374300030668251,0.5391609828391329,0.549544212692622,0.5103964784303721,0.514939945306369,0.5115706908400111,0.551793262834936,0.5081203592403011,0.4968382095681029,0.484249722714654,0.471988145133657,0.446737336551318,0.374644726682123,0.438476228390782,0.466049092168083,0.31266446386275,0.5351444436208179,0.367757724595809,0.327244476015917,0.141061557059162,0.305845031799054,0.286527151743567],[0.903320227308785,0.9102090580759871,0.8734051855884899,0.87811135471224,0.8940041672691059,0.9683188850085129,0.8198004164452108,0.9088929592020341,0.9489065286651079,0.8928547607369449,0.9215370421057031,0.8900024869052661,0.9237232500781,0.8951246083788629,0.8448253282235,0.9144647104516591,0.8961824026731491,0.8380856175468739,0.8833659460414129,0.8815946754467421,0.9085661940507059,0.8685933767491729,0.9208213275090881,0.851039897451092,0.860431090334227,0.8659656801869909,0.9122174888426758,0.874482888741797,0.902521560450612,0.8644503172958929,0.8979026254981701,0.9218217813754971,0.9010552059162241,0.8610936775446729,0.914375731317019,0.9138665032024671,0.8613677410579291,0.857044265209212,0.774494274575386,0.917852543105203,0.8644503172958929,0.907179833969447,0.817019936190885,0.7721838967634269,0.8049258902203199,0.773738642335168,0.8055586040649499,0.7879890529051691,0.8209248896094659,0.7538517744745079,0.7852519348704809,0.7109789234895441,0.690332259525367,0.733757734440333,0.6422760514312249,0.6623606008723461,0.6634468448483161,0.66749620089746,0.676031453972251,0.357791197618315,0.632615412598558,0.693257598442421,0.335312770824209,0.59430029415501,0.4668113781082789,0.452962367019481,0.38284721958606,0.418554543262074,0.425360601369716],[0.935853543835842,0.9387400317732478,0.9090217148460241,0.913093626375544,0.9261574108061491,0.984757098997059,0.862950946406692,0.9441027716270809,0.974399564680736,0.9298666549814429,0.9476104892855741,0.9217813728097449,0.9547840898924109,0.939922490216662,0.8935748330786791,0.9492497974437231,0.934522909010524,0.8612833986070341,0.9194183560129351,0.929957155606444,0.9399865496531891,0.9065988027067741,0.95298193726863,0.867218863924647,0.9201549564417951,0.906947870429289,0.942302458113657,0.924531301553261,0.937025366404205,0.888128110467543,0.9403492307358751,0.954017004862837,0.9419816536547099,0.915275402142792,0.949951107160208,0.951112722518898,0.9023566234172871,0.9016082843351809,0.7197346696759449,0.949079705426828,0.888128110467543,0.9471971497559029,0.882850656783028,0.711702017884835,0.840503094839173,0.596705366056932,0.7195506471796851,0.709964653921349,0.654831476859411,0.472442737589203,0.692630055830857,0.563777353184298,0.5663554335350389,0.398092593028584,0.265595030546838,0.3072428727526419,0.307338686768517,0.5215784058224371,0.516078473293037,0.348859653248523,0.262192519680191,0.4471290957992161,0.340514966092224,0.220806578326864,0.274970499683404,0.25347352335217,0.226542779810754,0.269415343964041,0.211819606861463],[0.66135630363234,0.693136066198349,0.6512615223232679,0.7206991761922841,0.555390309460723,0.6944006532430649,0.643798291683197,0.650083735585213,0.645200282335281,0.6831915825605389,0.629489570856094,0.555390160530806,0.640964698087128,null,0.6745877265930179,0.595530979335308,0.6475326567888261,0.577030595620765,0.6381846070289611,0.5254701729863879,0.565270206101941,0.6069276034832,0.640964716672897,0.582006616845509,null,0.642120033502579,0.6277858018875121,0.5826019421219829,0.5563704028090369,0.5686537902181529,0.5925235748291021,0.6090870733772161,0.598991438746452,null,null,0.617242477834225,null,null,0.514325733797438,0.5957083255052571,0.568653799593449,null,null,0.514325531199575,0.551657270640135,0.558105724960252,0.5681798011064531,0.546423923224211,0.6517546921968461,0.759616315364838,0.5412018820643429,0.623691253364086,0.623559169471264,0.72976952791214,0.5878945291042329,0.702670872211456,0.7013632953166961,0.566439755260944,0.566441215574741,0.5164459633349789,0.564448088407517,0.556303475052118,0.48933357718846,0.499071618833113,0.488922122865915,0.4889212995767589,0.4259597137570379,0.457587536424398,0.457603923976421],[0.08641975308642,0.125,0.073170731707317,0.123893805309734,0.08917197452229303,0.102189781021898,0.110236220472441,0.12962962962963,0.101449275362319,0.095238095238095,0.08974358974359001,0.08974358974359001,0.106060606060606,0.0972222222222219,0.076923076923077,0.098591549295775,0.555555555555556,0.09395973154362401,0.117647058823529,0.114942528735632,0.094594594594595,0.21875,0.106060606060606,0.094594594594595,0.116666666666667,0.09655172413793101,0.128440366972477,0.097902097902098,0.133333333333333,0.106060606060606,0.100719424460432,0.100719424460432,0.106060606060606,0.088607594936709,0.088607594936709,0.107692307692308,0.06278026905829601,0.06451612903225799,0.115384615384615,0.094594594594595,0.106060606060606,0.08805031446540899,0.09032258064516099,0.114285714285714,0.104477611940299,0.106870229007634,0.127272727272727,0.2,0.105263157894737,0.081395348837209,0.135922330097087,0.07339449541284401,0.0547945205479449,0.208955223880597,0.09375,0.09655172413793101,0.09655172413793101,0.0245398773006139,0.0245398773006139,0.14,0.106870229007634,0.175,0.057553956834532,0.108527131782946,0.018575851393189,0.01423487544484,0.03030303030303,0.014388489208633,0.013651877133106],[0.623514565284219,0.8173779547214509,0.609643215061187,0.8298367857933041,0.46981017243252,0.496263781329617,0.819376528263092,0.517560087144375,0.5239315442740921,0.5519008077681059,0.610452234745026,0.469810182228684,0.541079187106712,null,0.7975434064865109,0.5838454514741899,0.6680008769035339,0.5873483846810831,0.7886338233947751,0.780428290367126,0.4433497409916861,0.603610791265965,0.5410791710019111,0.5832385437020721,null,0.648219808936119,0.8162370026111599,0.5760091468691829,0.503627394084121,0.5336192777962171,0.5584926120936869,0.496427363659854,0.486590904183686,null,null,0.520902156829834,null,null,0.544631394089014,0.4207455590367321,0.5336193181574339,null,null,0.544631600379944,0.511869801208377,0.558575004983391,0.766561716794968,0.5124607989564539,0.500734946981538,0.526087565347552,0.4645662792027,0.475069329142571,0.4751207605004311,0.5309048201888801,0.6559278219938279,0.555516790598631,0.538573563098907,0.499710539414082,0.499659434310161,0.890196054476631,0.550404392182827,0.5544933304190639,0.7497285336196751,0.495942037552595,0.491861739195883,0.4918455574661489,0.148293942213058,0.398023776710033,0.398035645484924],[1,0.999768257811407,1,0.999883117491691,0.996425346404487,1,0.999484498814892,1,1,1,0.999794619552544,0.997005943447967,1,0.860580660350907,0.999534033606178,0.999587635359695,0.999570416626533,0.9991641091210319,0.999348126016068,0.996847538736192,0.999419542243468,0.9995436163717859,1,0.999618674880528,0.855748291189953,0.9996212283295959,1,0.998845172330768,0.9988015958300089,0.999625886698975,0.999219481683595,1,0.9995678904081809,0.6630503480982149,0.9977751722239742,1,0.987424388040343,0.995363089832394,0.986932561981122,0.999560077087644,0.99954725822109,0.998146810375707,0.652346977674116,0.9885838107197941,0.9989870229902309,0.99965906821662,0.9993288328416021,0.9994175276521,1,1,0.999237850069408,0.9990097350016099,0.999080609182979,1,0.9994944663392969,0.999450291453146,0.9998669938041371,0.997783746134716,0.9977171095512091,0.8664470046615909,0.9993835933509929,0.999597645005638,0.86611652632404,0.9991650390553141,0.964870280712201,0.963968556532798,0.91975488190975,0.945763044039287,0.94756842359779],[0.6779102937460999,0.5917776819789941,0.6602705937879561,0.537370760233109,0.639216333773048,0.537885971956546,0.583454939689792,0.492602665324469,0.7363731468440121,0.508426571056651,0.496396868874609,0.639216500989236,0.477869996721709,null,0.5846438444334899,0.4862182342249621,0.44666286927646,0.576844026414728,0.6439708531636441,0.5306585822432639,0.556638361309629,0.671780476054546,0.477869497479421,0.587326184996299,null,0.5890571596651,0.643195754619781,0.670155438154198,0.572474665937213,0.8411059677032091,0.627919444190445,0.46201987756582,0.520649941042163,null,null,0.470498471863989,null,null,0.222125626919273,0.5670157489694471,0.8411140700841979,null,null,0.222113277830701,0.9437017698138259,0.822103420266605,0.617965668970258,0.7612319589174241,0.4626764321851989,0.388735352778664,0.785932862020377,0.576389905842928,0.5763895568067869,0.381926657400115,0.7560694492448751,0.461808711719596,0.411711630767818,0.587697970538875,0.5876978419249711,0.447318473818471,0.741363581608263,0.686720428041109,0.475423206447568,1,0.7861060417499509,0.786106474333755,0.37158832173213,0.7340778823052091,0.7340783470754569],[null,0.324337977159138,null,0.300818592115065,null,0.587860660186951,0.363256825118331,0.610236143862915,0.418383905696265,0.6150077933645071,null,0.259244985285162,null,null,0.0786258889935159,null,0.310140137481567,null,0.03763178118843701,null,null,0.161377909768667,0.3018804623973519,null,null,0.172333167897598,0.249845273769315,null,null,null,null,null,0.615128686645938,null,null,0.477220989055308,null,null,null,0.292746036472767,0.114624941091464,null,null,0.119399972636035,0.2071099253393,null,0.029739195562072,0.08085301526748301,0.3624378144955689,null,0.059214757088678,0.490415907227217,null,null,null,null,null,null,0.285761881639227,null,0.202589351840674,0.055423736361612,null,1,null,0.130617234468938,0.07076964459932801,null,0.234625250501002]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Dataset<\/th>\n      <th>Method<\/th>\n      <th>Output<\/th>\n      <th>Features<\/th>\n      <th>Scaling<\/th>\n      <th>Overall Score<\/th>\n      <th>Batch Correction<\/th>\n      <th>PCR batch<\/th>\n      <th>Batch ASW<\/th>\n      <th>graph iLISI<\/th>\n      <th>graph connectivity<\/th>\n      <th>kBET<\/th>\n      <th>Bio conservation<\/th>\n      <th>NMI cluster/label<\/th>\n      <th>ARI cluster/label<\/th>\n      <th>Cell type ASW<\/th>\n      <th>isolated label F1<\/th>\n      <th>isolated label silhouette<\/th>\n      <th>graph cLISI<\/th>\n      <th>CC conservation<\/th>\n      <th>HVG conservation<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"targets":6,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":7,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":8,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":9,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":10,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":11,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":12,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":13,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":14,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":15,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":16,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":17,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":18,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":19,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":20,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":21,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"className":"dt-right","targets":[6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Dataset","targets":1},{"name":"Method","targets":2},{"name":"Output","targets":3},{"name":"Features","targets":4},{"name":"Scaling","targets":5},{"name":"Overall Score","targets":6},{"name":"Batch Correction","targets":7},{"name":"PCR batch","targets":8},{"name":"Batch ASW","targets":9},{"name":"graph iLISI","targets":10},{"name":"graph connectivity","targets":11},{"name":"kBET","targets":12},{"name":"Bio conservation","targets":13},{"name":"NMI cluster/label","targets":14},{"name":"ARI cluster/label","targets":15},{"name":"Cell type ASW","targets":16},{"name":"isolated label F1","targets":17},{"name":"isolated label silhouette","targets":18},{"name":"graph cLISI","targets":19},{"name":"CC conservation","targets":20},{"name":"HVG conservation","targets":21}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.columnDefs.1.render","options.columnDefs.2.render","options.columnDefs.3.render","options.columnDefs.4.render","options.columnDefs.5.render","options.columnDefs.6.render","options.columnDefs.7.render","options.columnDefs.8.render","options.columnDefs.9.render","options.columnDefs.10.render","options.columnDefs.11.render","options.columnDefs.12.render","options.columnDefs.13.render","options.columnDefs.14.render","options.columnDefs.15.render"],"jsHooks":[]}</script>
</div>
<div class="section level3">
<h3 id="insert-scores-for-new-liger">Insert scores for new LIGER<a class="anchor" aria-label="anchor" href="#insert-scores-for-new-liger"></a>
</h3>
<p>Now we load back the scores we just calculated for LIGER’s new method
and insert it to the full table before we can scale the scores and
obtain the final overall scores.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liger_scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"pancreas_liger_scores.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make a one-row data.frame that matches to the columns from the full table</span></span>
<span><span class="va">liger_scores_reformat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    <span class="co"># Dataset,  Method,   Output,  Feaetures, Scaling,  Overall</span></span>
<span>    <span class="st">"pancreas"</span>, <span class="st">"LIGER2"</span>, <span class="st">"embed"</span>, <span class="st">"HVG"</span>,     <span class="st">"scaled"</span>, <span class="cn">NA</span>, </span>
<span>    <span class="co"># Overall score above left as NA, yet to be calculated</span></span>
<span>    <span class="co"># Batch score, and individual batch metrics</span></span>
<span>    <span class="co"># Batch score left as NA, yet to be calculated</span></span>
<span>    <span class="cn">NA</span>, <span class="va">liger_scores</span><span class="op">$</span><span class="va">pancreas_pcr_batch</span>, <span class="va">liger_scores</span><span class="op">$</span><span class="va">pancreas_batch_asw</span>, </span>
<span>    <span class="va">liger_scores</span><span class="op">$</span><span class="va">pancreas_graph_ilisi</span>, <span class="va">liger_scores</span><span class="op">$</span><span class="va">pancreas_graph_connectivity</span>, </span>
<span>    <span class="va">liger_scores</span><span class="op">$</span><span class="va">pancreas_kBET</span>, </span>
<span>    <span class="co"># Bio score, and individual bio metrics</span></span>
<span>    <span class="co"># Bio score left as NA, yet to be calculated</span></span>
<span>    <span class="cn">NA</span>, <span class="va">liger_scores</span><span class="op">$</span><span class="va">pancreas_nmi</span>, <span class="va">liger_scores</span><span class="op">$</span><span class="va">pancreas_ari</span>, </span>
<span>    <span class="va">liger_scores</span><span class="op">$</span><span class="va">pancreas_cell_type_asw</span>, <span class="va">liger_scores</span><span class="op">$</span><span class="va">pancreas_isolated_label_F1</span>, </span>
<span>    <span class="va">liger_scores</span><span class="op">$</span><span class="va">pancreas_isolated_label_silhouette</span>, </span>
<span>    <span class="va">liger_scores</span><span class="op">$</span><span class="va">pancreas_graph_clisi</span>, </span>
<span>    <span class="va">liger_scores</span><span class="op">$</span><span class="va">pancreas_cell_cycle_conservation</span>, <span class="cn">NA</span></span>
<span>    <span class="co"># HVG score left as NA, not applicable</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">liger_scores_reformat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">pancreas_scores</span><span class="op">)</span></span>
<span><span class="va">all_scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">pancreas_scores</span>, <span class="va">liger_scores_reformat</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="calculate-scaled-scores">Calculate scaled scores<a class="anchor" aria-label="anchor" href="#calculate-scaled-scores"></a>
</h3>
<p>Following the formula described earlier, we then calculate the scaled
scores for each metric, derive the average batch score and average bio
score for each method. The <code>scaleMinMax()</code> function is
defined in short below so it can be <code><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply()</a></code>ed to each
column (<code>MARGIN = 2</code>) in a neat way.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># x - expect to be a column from the score table, containing the scores for a </span></span>
<span><span class="co">#     metric, from all methods</span></span>
<span><span class="co"># `na.rm = TRUE` indicates that we want to ignore NA values when calculating min </span></span>
<span><span class="co"># and max</span></span>
<span><span class="va">scaleMinMax</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="op">(</span><span class="va">x</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">x</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">x</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">x</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Calculate batch scores</span></span>
<span><span class="va">batch_scores</span> <span class="op">&lt;-</span> <span class="va">all_scores</span><span class="op">[</span>, <span class="fl">8</span><span class="op">:</span><span class="fl">12</span><span class="op">]</span></span>
<span><span class="va">batch_scores.scaled</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">batch_scores</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>, FUN <span class="op">=</span> <span class="va">scaleMinMax</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">all_scores</span><span class="op">$</span><span class="va">`Batch Correction`</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">batch_scores.scaled</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate bio scores</span></span>
<span><span class="va">bio_scores</span> <span class="op">&lt;-</span> <span class="va">all_scores</span><span class="op">[</span>, <span class="fl">14</span><span class="op">:</span><span class="fl">21</span><span class="op">]</span></span>
<span><span class="va">bio_scores.scaled</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">bio_scores</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>, FUN <span class="op">=</span> <span class="va">scaleMinMax</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">all_scores</span><span class="op">$</span><span class="va">`Bio conservation`</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">bio_scores.scaled</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="overall-ranking">Overall ranking<a class="anchor" aria-label="anchor" href="#overall-ranking"></a>
</h3>
<p>Lastly, we can now reach to the overall score for each method and
have our new method ranked among all others.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate overall scores</span></span>
<span><span class="va">all_scores</span><span class="op">$</span><span class="va">`Overall Score`</span> <span class="op">&lt;-</span> <span class="fl">0.6</span><span class="op">*</span><span class="va">all_scores</span><span class="op">$</span><span class="va">`Bio conservation`</span> <span class="op">+</span> <span class="fl">0.4</span><span class="op">*</span><span class="va">all_scores</span><span class="op">$</span><span class="va">`Batch Correction`</span></span>
<span></span>
<span><span class="co"># Rank all methods by overall score</span></span>
<span><span class="va">all_scores</span> <span class="op">&lt;-</span> <span class="va">all_scores</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">all_scores</span><span class="op">$</span><span class="va">`Overall Score`</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">all_scores</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></code></pre></div>
<p>Now we can see the updated overall scores and the new ranking.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-e5c8c404fe174e4c81bd" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-e5c8c404fe174e4c81bd">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70"],["pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas","pancreas"],["LIGER2","Harmony","Seurat v3 RPCA","Harmony","Seurat v3 CCA","fastMNN","scGen*","scGen*","Seurat v3 RPCA","scGen*","Seurat v3 CCA","scANVI*","fastMNN","fastMNN","BBKNN","Seurat v3 CCA","scVI","Scanorama","Harmony","Seurat v3 RPCA","trVAE","Scanorama","Seurat v3 RPCA","fastMNN","Harmony","BBKNN","Seurat v3 CCA","ComBat","scVI","Scanorama","fastMNN","scANVI*","Scanorama","ComBat","BBKNN","Conos","MNN","Conos","Conos","fastMNN","MNN","fastMNN","Conos","BBKNN","fastMNN","ComBat","Scanorama","ComBat","Scanorama","Scanorama","DESC","MNN","SAUCIE","SAUCIE","DESC","trVAE","DESC","DESC","SAUCIE","SAUCIE","LIGER","MNN","Scanorama","LIGER","Unintegrated","SAUCIE","SAUCIE","scGen*","SAUCIE","SAUCIE"],["embed","embed","gene","embed","gene","embed","gene","gene","gene","gene","gene","embed","gene","embed","graph","gene","embed","gene","embed","gene","embed","embed","gene","gene","embed","graph","gene","gene","embed","embed","embed","embed","embed","gene","graph","graph","gene","graph","graph","embed","gene","gene","graph","graph","gene","gene","embed","gene","gene","gene","embed","gene","gene","embed","embed","embed","embed","embed","embed","gene","embed","gene","gene","embed","gene","embed","gene","gene","embed","gene"],["HVG","HVG","HVG","HVG","HVG","HVG","HVG","HVG","HVG","FULL","HVG","HVG","HVG","HVG","HVG","FULL","HVG","HVG","FULL","FULL","HVG","HVG","FULL","HVG","FULL","FULL","FULL","HVG","FULL","FULL","FULL","FULL","HVG","HVG","HVG","FULL","HVG","FULL","HVG","FULL","HVG","FULL","HVG","FULL","FULL","FULL","FULL","FULL","FULL","HVG","HVG","FULL","HVG","HVG","HVG","FULL","FULL","FULL","HVG","HVG","HVG","FULL","FULL","FULL","FULL","FULL","FULL","FULL","FULL","FULL"],["scaled","unscaled","scaled","scaled","scaled","scaled","unscaled","scaled","unscaled","unscaled","unscaled","unscaled","scaled","unscaled","unscaled","scaled","unscaled","scaled","scaled","scaled","unscaled","scaled","unscaled","unscaled","unscaled","unscaled","unscaled","scaled","unscaled","scaled","unscaled","unscaled","unscaled","unscaled","scaled","unscaled","unscaled","scaled","scaled","scaled","scaled","unscaled","unscaled","scaled","scaled","unscaled","unscaled","scaled","scaled","unscaled","unscaled","scaled","unscaled","unscaled","scaled","unscaled","unscaled","scaled","scaled","scaled","unscaled","unscaled","unscaled","unscaled","unscaled","scaled","scaled","scaled","unscaled","unscaled"],[0.8030982937230094,0.7900360679688441,0.781129145712782,0.774658792250263,0.7709289650483278,0.7520873994288766,0.7489352687411253,0.7458785649306414,0.741923615586465,0.7348690919539502,0.7279916188864606,0.7238846776367868,0.7236909020797589,0.7200625574710253,0.7168774083190042,0.7156774950994682,0.7134047188995362,0.7126605261414629,0.7109212279543677,0.7089738785199895,0.7045645083098593,0.7034837416154639,0.7007014666337852,0.6918506443733323,0.6904090332595578,0.6902807586967008,0.6896714802526528,0.6842235396936835,0.6729758031525137,0.6726879252147936,0.6716565957581493,0.6713116856991459,0.663013161469418,0.6551205718595786,0.6480344931813979,0.6426782291452786,0.6382602315819372,0.637337709026658,0.6346840136326949,0.6345280543046823,0.633429673467783,0.6288279997755692,0.6287660074921732,0.6241557408864619,0.6030292295885673,0.60031302782933,0.5870634853127201,0.5854294146700552,0.5843116944348031,0.5797906376766433,0.5732458911311206,0.5701686032697101,0.5586899350450075,0.5573283080592579,0.5556214502016028,0.5381704540711878,0.5357908446259098,0.5291836393464012,0.5225152837964757,0.5073929306604297,0.479750758417351,0.4784588120968039,0.4707425443128758,0.4462482265668222,0.4311208282843271,0.4144978161442718,0.3907773742713205,0.3339410784760305,0.3249217341867929,0.31356706994327],[0.8393794522980009,0.9250212821004006,0.9150252143186233,0.9247317232789918,0.904595650948897,0.9602153908472746,0.8564417051602037,0.8990303115821912,0.8829954959585733,0.8353590181747004,0.8362488044002649,0.823269372366336,0.9616092192682562,0.8221774896433859,0.8248619082651002,0.8631191420085825,0.8279361627261466,0.7012656333271583,0.8599683566317186,0.8310152384303491,0.7935084896935768,0.8499982602145459,0.8164413442186496,0.8226014315981643,0.7972295747346398,0.7802704172236994,0.818570903968977,0.7090008124864916,0.7078790378090685,0.7448814135322297,0.6884717147323882,0.7003827657461794,0.7202584964342705,0.7060444629495697,0.9039025101426562,0.4810687705492399,0.6768971263360759,0.5512486763568933,0.5377743825994071,0.8791460679756499,0.7187009311128809,0.6888704683550683,0.451872032418472,0.8964052093021263,0.873594695147314,0.6200653769032216,0.592949810243429,0.6430763320866106,0.6546342314867702,0.6407351199329081,0.6087984087888685,0.6598267905287176,0.6243149196529649,0.625964733888129,0.561363731251603,0.5832455963175185,0.5942197972126202,0.5965845142940219,0.5983059917907038,0.5983763218240973,0.6374098060201934,0.5384326876558373,0.4777827225300645,0.6466238706229304,0.2750854052795908,0.4846079534669662,0.4860767216544252,0.6232603606013336,0.3535367877684007,0.3541269472428247],[0.993173166167396,0.9207507889066251,0.934415099097217,0.9502543092841821,0.9485459054864742,0.992628053680685,0.88418361976556,0.9616585688451948,0.912346604806772,0.898883910134011,0.9191516767860369,0.9174695195723209,0.9926272748537651,0.8499347684495629,null,0.9535777747864591,0.929525587174975,0.917091470045615,0.9647041417429421,0.9407447723807622,0.8835143133559989,0.9363541576606071,0.9267811016886141,0.8499383223813241,0.924309308415366,null,0.931604688484989,1,0.862350614009722,0.9231238712373651,0.664995392543878,0.8592847757856709,0.830140057529084,0.999985097124501,null,null,0.855878437307926,null,null,0.995477577766712,0.931881356544966,0.6649968824489001,null,null,0.9954802244037571,0.999952826970895,0.685529031884872,1,0.8327923651622779,0.8189749642609692,0.8410342129367501,0.9475190930052041,0.7903509111992242,0.7903572966889691,0.8426803885140991,0.649738764344303,0.8869344612178409,0.885149534255942,0.9096096409478579,0.9096113520663041,0.9491740653390609,0.7125446098053699,0.50079641418794,0.9485968294363271,0,0.7742396189704069,0.774238663154949,0.9996110087823841,0.545447833526211,0.5454519414086889],[0.7933811059905438,0.894511369425683,0.830352741727264,0.8682828071792281,0.791238660130418,0.8900114368553191,0.7867018472299409,0.7812526262808991,0.848593198040775,0.7645371062667121,0.7930087360335891,0.8510651793875571,0.8900112438566241,0.8977854125122142,null,0.800717260882654,0.863945401459413,0.7163889233728989,0.8618988047164191,0.8330122782992859,0.872082276249737,0.9132077831855909,0.844258080311638,0.8977854116901529,0.8752655937983671,null,0.793653045863961,0.8629165484350629,0.856719623267588,0.923753592193824,0.8851395475853341,0.8580273944723491,0.923015400485989,0.897589642616812,null,null,0.9057019050549191,null,null,0.8766370081324929,0.891655363823846,0.885139538029635,null,null,0.8766368774402621,0.8773995599969351,0.9075370970552049,0.862004233410549,0.777440726706356,0.7839881551943959,0.520277694088659,0.888926057687407,0.713064112920118,0.7134053300802541,0.463715663608702,0.8593554434528191,0.6682740082372021,0.6872645261494661,0.7205518336529679,0.720167859382302,0.790581911603242,0.8448200803913499,0.8253157844642529,0.802246787096501,0.7702285814202728,0.7518227487864159,0.7516218405720709,0.6867855334771179,0.7407034154344431,0.7404462286383371],[0.3073894205861533,0.367036082891136,0.3843382806103279,0.371651903484589,0.410640737316228,0.397792204350052,0.3868414245549651,0.408808327645455,0.428733003257341,0.349642309418385,0.432617309283705,0.298390493848281,0.401356175881789,0.291011966673681,0.3985925302610949,0.385735692961946,0.304252237012593,0.326013353470828,0.346388322999285,0.32439399417916,0.251749767695298,0.3170986532033739,0.368865149995437,0.291256256294396,0.298450335252708,0.400524797426676,0.3835768488642921,0.096363344854045,0.18144840479935,0.181152893386147,0.157685633682709,0.181487313906364,0.173945575791937,0.10840353385013,0.404664366242914,0,0.07924053136125001,0.031505924490019,0.0032508424915039,0.342386059055253,0.162744822293329,0.158014921893108,0,0.393420430392384,0.333848568323911,0.07773122373431091,0.08280796471408,0.068030329118427,0.207496476358945,0.176542180508463,0.287974637717994,0.085279080319681,0.284840413730502,0.28561878124318,0.241581307255754,0.0054321038700699,0.138642221018394,0.139384329844067,0.311008825147208,0.311268928634215,0.440112154521447,0.004064867867098,0.110113267974042,0.423875104645788,1.39e-16,0.301557735530132,0.299771219728084,0.347378593318301,0.2725262295533989,0.272692316270471],[0.9750180591653,0.975683999692299,0.9801448072863729,0.970528569381912,0.9826876122830529,0.9859843467077769,0.99151218026432,0.9668994542950159,0.9822476290369929,0.988937932530597,0.9845794790281999,0.9901342649293009,0.9859680975255201,0.979376390999663,0.975431624381728,0.9587294467743058,0.990548145030244,0.9574738019042941,0.9588859051364921,0.9546537986654209,0.9754137804258521,0.9870079187648331,0.9700302922800741,0.9792771069733069,0.954998427344705,0.9322396572839251,0.9672640939169781,0.985943173010672,0.9685477617391931,0.9806229145674791,0.961935902112544,0.9532281842447191,0.9813724345867808,0.97862816408287,0.993727274028197,0.989895857884438,0.9719553452437321,0.9957750369118,0.9959385215446921,0.978249247500479,0.985385865445034,0.9614846903870059,0.989826852289582,0.9832854534661021,0.9754139635264599,0.957349355022956,0.9527370741600258,0.9676092666381819,0.962786669327037,0.958694898692389,0.753305286669516,0.9687006626135541,0.780407164454802,0.780407164454802,0.7942776847012371,0.9789544024133021,0.862306871352902,0.8670100990879219,0.71400294889713,0.71400294889713,0.654438316783443,0.9449895715119619,0.7600025612855831,0.654029948493883,0.7961010610589521,0.575897599398362,0.5816118851126479,0.547182120294373,0.461664788854583,0.461173603849724],[0.5248958996458675,0.5996614124760971,0.6142814460317261,0.612750615378048,0.590520763346827,0.612499215648134,0.523571856669807,0.598024143266249,0.467619053828556,0.532605296187915,0.39863159486603,0.453231066212606,0.611929690780312,0.4518814700998089,0.410417442547723,0.522192465514023,0.437236133016931,0.287857559718044,0.485460006526559,0.485394458578717,0.43919824900201,0.423651260192227,0.374748914523434,0.4527882054357179,0.388823220673077,0.3806229632401189,0.420207115449824,0.342786849255109,0.334852034749818,0.312331400398284,0.38930420378835,0.330336219894599,0.305296785747007,0.288896732617294,0.508585556886334,0.331117056604467,0.319770959195757,0.397540540145497,0.409724340184667,0.488555306102776,0.293141596478703,0.390388008309883,0.285716210455299,0.520307801559875,0.486967461994211,0.145370367678296,0.202691538088101,0.223924808195234,0.276325643337358,0.280516705341841,0.5516811527076221,0.242861322867619,0.378116317832817,0.381091992438199,0.506481373873081,0.316263145552971,0.393344390812548,0.373539306550861,0.273848599583119,0.274156737337568,0.181330287973736,0.218333267419239,0.247330012274959,0.211929947207271,0.138187193126023,0.158938026824837,0.159537658274261,0.449035388027117,0.09518430369088771,0.0972850264515485],[0.7789108546730151,0.7000459252144731,0.6918650999755546,0.6746101715644438,0.681817841114615,0.6133354051499446,0.6772643111284063,0.6437774004962749,0.6478756953383928,0.6678758078067835,0.6558201618772577,0.657628214483754,0.5650786906207608,0.651985936022785,0.6448877416882735,0.6173830638267251,0.6370504230151292,0.7202571213509994,0.6115564755028006,0.6276129719130831,0.6452685207207142,0.6058073958827425,0.6235415482438755,0.6046834528901111,0.6191953389428365,0.6302876530120349,0.6037385311084367,0.6677053578318115,0.6497069800481438,0.6245589330031696,0.6604465164419899,0.6519309656677903,0.624849604826183,0.6211713111329179,0.4774558152072257,0.7504178682093046,0.6125023017458447,0.6947303974731677,0.6992904343215534,0.4714493785240372,0.576582168371051,0.5887996873892365,0.7466953242079739,0.4426560952760189,0.4226522525494029,0.5871447951134025,0.5831392686922475,0.5469981363923516,0.5374300030668251,0.5391609828391334,0.5495442126926219,0.5103964784303717,0.5149399453063692,0.5115706908400105,0.5517932628349361,0.5081203592403006,0.4968382095681028,0.4842497227146541,0.4719881451336569,0.4467373365513179,0.3746447266821228,0.4384762283907817,0.4660490921680833,0.3126644638627502,0.535144443620818,0.3677577245958089,0.3272444760159174,0.1410615570591617,0.3058450317990543,0.286527151743567],[0.8888347309477597,0.903320227308785,0.9102090580759871,0.8734051855884899,0.87811135471224,0.8940041672691059,0.9683188850085129,0.8198004164452108,0.9088929592020341,0.9489065286651079,0.8928547607369449,0.9215370421057031,0.8900024869052661,0.9237232500781,0.8951246083788629,0.8448253282235,0.9144647104516591,0.8961824026731491,0.8380856175468739,0.8833659460414129,0.8815946754467421,0.9085661940507059,0.8685933767491729,0.9208213275090881,0.851039897451092,0.860431090334227,0.8659656801869909,0.9122174888426758,0.874482888741797,0.902521560450612,0.8644503172958929,0.8979026254981701,0.9218217813754971,0.9010552059162241,0.8610936775446729,0.914375731317019,0.9138665032024671,0.8613677410579291,0.857044265209212,0.774494274575386,0.917852543105203,0.8644503172958929,0.907179833969447,0.817019936190885,0.7721838967634269,0.8049258902203199,0.773738642335168,0.8055586040649499,0.7879890529051691,0.8209248896094659,0.7538517744745079,0.7852519348704809,0.7109789234895441,0.690332259525367,0.733757734440333,0.6422760514312249,0.6623606008723461,0.6634468448483161,0.66749620089746,0.676031453972251,0.357791197618315,0.632615412598558,0.693257598442421,0.335312770824209,0.59430029415501,0.4668113781082789,0.452962367019481,0.38284721958606,0.418554543262074,0.425360601369716],[0.9284316085505634,0.935853543835842,0.9387400317732478,0.9090217148460241,0.913093626375544,0.9261574108061491,0.984757098997059,0.862950946406692,0.9441027716270809,0.974399564680736,0.9298666549814429,0.9476104892855741,0.9217813728097449,0.9547840898924109,0.939922490216662,0.8935748330786791,0.9492497974437231,0.934522909010524,0.8612833986070341,0.9194183560129351,0.929957155606444,0.9399865496531891,0.9065988027067741,0.95298193726863,0.867218863924647,0.9201549564417951,0.906947870429289,0.942302458113657,0.924531301553261,0.937025366404205,0.888128110467543,0.9403492307358751,0.954017004862837,0.9419816536547099,0.915275402142792,0.949951107160208,0.951112722518898,0.9023566234172871,0.9016082843351809,0.7197346696759449,0.949079705426828,0.888128110467543,0.9471971497559029,0.882850656783028,0.711702017884835,0.840503094839173,0.596705366056932,0.7195506471796851,0.709964653921349,0.654831476859411,0.472442737589203,0.692630055830857,0.563777353184298,0.5663554335350389,0.398092593028584,0.265595030546838,0.3072428727526419,0.307338686768517,0.5215784058224371,0.516078473293037,0.348859653248523,0.262192519680191,0.4471290957992161,0.340514966092224,0.220806578326864,0.274970499683404,0.25347352335217,0.226542779810754,0.269415343964041,0.211819606861463],[0.6222552433843831,0.66135630363234,0.693136066198349,0.6512615223232679,0.7206991761922841,0.555390309460723,0.6944006532430649,0.643798291683197,0.650083735585213,0.645200282335281,0.6831915825605389,0.629489570856094,0.555390160530806,0.640964698087128,null,0.6745877265930179,0.595530979335308,0.6475326567888261,0.577030595620765,0.6381846070289611,0.5254701729863879,0.565270206101941,0.6069276034832,0.640964716672897,0.582006616845509,null,0.642120033502579,0.6277858018875121,0.5826019421219829,0.5563704028090369,0.5686537902181529,0.5925235748291021,0.6090870733772161,0.598991438746452,null,null,0.617242477834225,null,null,0.514325733797438,0.5957083255052571,0.568653799593449,null,null,0.514325531199575,0.551657270640135,0.558105724960252,0.5681798011064531,0.546423923224211,0.6517546921968461,0.759616315364838,0.5412018820643429,0.623691253364086,0.623559169471264,0.72976952791214,0.5878945291042329,0.702670872211456,0.7013632953166961,0.566439755260944,0.566441215574741,0.5164459633349789,0.564448088407517,0.556303475052118,0.48933357718846,0.499071618833113,0.488922122865915,0.4889212995767589,0.4259597137570379,0.457587536424398,0.457603923976421],[0.5185185185185185,0.08641975308642,0.125,0.073170731707317,0.123893805309734,0.08917197452229303,0.102189781021898,0.110236220472441,0.12962962962963,0.101449275362319,0.095238095238095,0.08974358974359001,0.08974358974359001,0.106060606060606,0.0972222222222219,0.076923076923077,0.098591549295775,0.555555555555556,0.09395973154362401,0.117647058823529,0.114942528735632,0.094594594594595,0.21875,0.106060606060606,0.094594594594595,0.116666666666667,0.09655172413793101,0.128440366972477,0.097902097902098,0.133333333333333,0.106060606060606,0.100719424460432,0.100719424460432,0.106060606060606,0.088607594936709,0.088607594936709,0.107692307692308,0.06278026905829601,0.06451612903225799,0.115384615384615,0.094594594594595,0.106060606060606,0.08805031446540899,0.09032258064516099,0.114285714285714,0.104477611940299,0.106870229007634,0.127272727272727,0.2,0.105263157894737,0.081395348837209,0.135922330097087,0.07339449541284401,0.0547945205479449,0.208955223880597,0.09375,0.09655172413793101,0.09655172413793101,0.0245398773006139,0.0245398773006139,0.14,0.106870229007634,0.175,0.057553956834532,0.108527131782946,0.018575851393189,0.01423487544484,0.03030303030303,0.014388489208633,0.013651877133106],[0.7445839795169722,0.623514565284219,0.8173779547214509,0.609643215061187,0.8298367857933041,0.46981017243252,0.496263781329617,0.819376528263092,0.517560087144375,0.5239315442740921,0.5519008077681059,0.610452234745026,0.469810182228684,0.541079187106712,null,0.7975434064865109,0.5838454514741899,0.6680008769035339,0.5873483846810831,0.7886338233947751,0.780428290367126,0.4433497409916861,0.603610791265965,0.5410791710019111,0.5832385437020721,null,0.648219808936119,0.8162370026111599,0.5760091468691829,0.503627394084121,0.5336192777962171,0.5584926120936869,0.496427363659854,0.486590904183686,null,null,0.520902156829834,null,null,0.544631394089014,0.4207455590367321,0.5336193181574339,null,null,0.544631600379944,0.511869801208377,0.558575004983391,0.766561716794968,0.5124607989564539,0.500734946981538,0.526087565347552,0.4645662792027,0.475069329142571,0.4751207605004311,0.5309048201888801,0.6559278219938279,0.555516790598631,0.538573563098907,0.499710539414082,0.499659434310161,0.890196054476631,0.550404392182827,0.5544933304190639,0.7497285336196751,0.495942037552595,0.491861739195883,0.4918455574661489,0.148293942213058,0.398023776710033,0.398035645484924],[0.9995901383991688,1,0.999768257811407,1,0.999883117491691,0.996425346404487,1,0.999484498814892,1,1,1,0.999794619552544,0.997005943447967,1,0.860580660350907,0.999534033606178,0.999587635359695,0.999570416626533,0.9991641091210319,0.999348126016068,0.996847538736192,0.999419542243468,0.9995436163717859,1,0.999618674880528,0.855748291189953,0.9996212283295959,1,0.998845172330768,0.9988015958300089,0.999625886698975,0.999219481683595,1,0.9995678904081809,0.6630503480982149,0.9977751722239742,1,0.987424388040343,0.995363089832394,0.986932561981122,0.999560077087644,0.99954725822109,0.998146810375707,0.652346977674116,0.9885838107197941,0.9989870229902309,0.99965906821662,0.9993288328416021,0.9994175276521,1,1,0.999237850069408,0.9990097350016099,0.999080609182979,1,0.9994944663392969,0.999450291453146,0.9998669938041371,0.997783746134716,0.9977171095512091,0.8664470046615909,0.9993835933509929,0.999597645005638,0.86611652632404,0.9991650390553141,0.964870280712201,0.963968556532798,0.91975488190975,0.945763044039287,0.94756842359779],[0.4774870990333798,0.6779102937460999,0.5917776819789941,0.6602705937879561,0.537370760233109,0.639216333773048,0.537885971956546,0.583454939689792,0.492602665324469,0.7363731468440121,0.508426571056651,0.496396868874609,0.639216500989236,0.477869996721709,null,0.5846438444334899,0.4862182342249621,0.44666286927646,0.576844026414728,0.6439708531636441,0.5306585822432639,0.556638361309629,0.671780476054546,0.477869497479421,0.587326184996299,null,0.5890571596651,0.643195754619781,0.670155438154198,0.572474665937213,0.8411059677032091,0.627919444190445,0.46201987756582,0.520649941042163,null,null,0.470498471863989,null,null,0.222125626919273,0.5670157489694471,0.8411140700841979,null,null,0.222113277830701,0.9437017698138259,0.822103420266605,0.617965668970258,0.7612319589174241,0.4626764321851989,0.388735352778664,0.785932862020377,0.576389905842928,0.5763895568067869,0.381926657400115,0.7560694492448751,0.461808711719596,0.411711630767818,0.587697970538875,0.5876978419249711,0.447318473818471,0.741363581608263,0.686720428041109,0.475423206447568,1,0.7861060417499509,0.786106474333755,0.37158832173213,0.7340778823052091,0.7340783470754569],[null,null,0.324337977159138,null,0.300818592115065,null,0.587860660186951,0.363256825118331,0.610236143862915,0.418383905696265,0.6150077933645071,null,0.259244985285162,null,null,0.0786258889935159,null,0.310140137481567,null,0.03763178118843701,null,null,0.161377909768667,0.3018804623973519,null,null,0.172333167897598,0.249845273769315,null,null,null,null,null,0.615128686645938,null,null,0.477220989055308,null,null,null,0.292746036472767,0.114624941091464,null,null,0.119399972636035,0.2071099253393,null,0.029739195562072,0.08085301526748301,0.3624378144955689,null,0.059214757088678,0.490415907227217,null,null,null,null,null,null,0.285761881639227,null,0.202589351840674,0.055423736361612,null,1,null,0.130617234468938,0.07076964459932801,null,0.234625250501002]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Dataset<\/th>\n      <th>Method<\/th>\n      <th>Output<\/th>\n      <th>Features<\/th>\n      <th>Scaling<\/th>\n      <th>Overall Score<\/th>\n      <th>Batch Correction<\/th>\n      <th>PCR batch<\/th>\n      <th>Batch ASW<\/th>\n      <th>graph iLISI<\/th>\n      <th>graph connectivity<\/th>\n      <th>kBET<\/th>\n      <th>Bio conservation<\/th>\n      <th>NMI cluster/label<\/th>\n      <th>ARI cluster/label<\/th>\n      <th>Cell type ASW<\/th>\n      <th>isolated label F1<\/th>\n      <th>isolated label silhouette<\/th>\n      <th>graph cLISI<\/th>\n      <th>CC conservation<\/th>\n      <th>HVG conservation<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"targets":6,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":7,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":8,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":9,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":10,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":11,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":12,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":13,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":14,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":15,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":16,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":17,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":18,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":19,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":20,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"targets":21,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 3, 3, \",\", \".\", null);\n  }"},{"className":"dt-right","targets":[6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Dataset","targets":1},{"name":"Method","targets":2},{"name":"Output","targets":3},{"name":"Features","targets":4},{"name":"Scaling","targets":5},{"name":"Overall Score","targets":6},{"name":"Batch Correction","targets":7},{"name":"PCR batch","targets":8},{"name":"Batch ASW","targets":9},{"name":"graph iLISI","targets":10},{"name":"graph connectivity","targets":11},{"name":"kBET","targets":12},{"name":"Bio conservation","targets":13},{"name":"NMI cluster/label","targets":14},{"name":"ARI cluster/label","targets":15},{"name":"Cell type ASW","targets":16},{"name":"isolated label F1","targets":17},{"name":"isolated label silhouette","targets":18},{"name":"graph cLISI","targets":19},{"name":"CC conservation","targets":20},{"name":"HVG conservation","targets":21}],"order":[],"autoWidth":false,"orderClasses":false,"rowCallback":"function(row, data, displayNum, displayIndex, dataIndex) {\nvar value=data[1]; $(this.api().cell(row, 1).node()).css({'background-color':$.inArray(dataIndex + 1, [1]) >= 0 ? \"#BBED80\" : null});\nvar value=data[2]; $(this.api().cell(row, 2).node()).css({'background-color':$.inArray(dataIndex + 1, [1]) >= 0 ? \"#BBED80\" : null});\nvar value=data[3]; $(this.api().cell(row, 3).node()).css({'background-color':$.inArray(dataIndex + 1, [1]) >= 0 ? \"#BBED80\" : null});\nvar value=data[4]; $(this.api().cell(row, 4).node()).css({'background-color':$.inArray(dataIndex + 1, [1]) >= 0 ? \"#BBED80\" : null});\nvar value=data[5]; $(this.api().cell(row, 5).node()).css({'background-color':$.inArray(dataIndex + 1, [1]) >= 0 ? \"#BBED80\" : null});\nvar value=data[6]; $(this.api().cell(row, 6).node()).css({'background-color':$.inArray(dataIndex + 1, [1]) >= 0 ? \"#BBED80\" : null});\nvar value=data[7]; $(this.api().cell(row, 7).node()).css({'background-color':$.inArray(dataIndex + 1, [1]) >= 0 ? \"#BBED80\" : null});\nvar value=data[8]; $(this.api().cell(row, 8).node()).css({'background-color':$.inArray(dataIndex + 1, [1]) >= 0 ? \"#BBED80\" : null});\n}"}},"evals":["options.columnDefs.0.render","options.columnDefs.1.render","options.columnDefs.2.render","options.columnDefs.3.render","options.columnDefs.4.render","options.columnDefs.5.render","options.columnDefs.6.render","options.columnDefs.7.render","options.columnDefs.8.render","options.columnDefs.9.render","options.columnDefs.10.render","options.columnDefs.11.render","options.columnDefs.12.render","options.columnDefs.13.render","options.columnDefs.14.render","options.columnDefs.15.render","options.rowCallback"],"jsHooks":[]}</script>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
