<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Cross-Species Analysis with UINMF • rliger2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Cross-Species Analysis with UINMF">
<meta property="og:description" content="rliger2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/import_data.html">Import data</a>
    </li>
    <li>
      <a href="../articles/liger_object.html">Introduction to a liger object</a>
    </li>
    <li>
      <a href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Cross-Species Analysis with UINMF</h1>
                        <h4 data-toc-skip class="author">April Kriebel
and Joshua Welch</h4>
            
            <h4 data-toc-skip class="date">12/6/2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/vignettes/articles/cross_species_vig.Rmd" class="external-link"><code>vignettes/articles/cross_species_vig.Rmd</code></a></small>
      <div class="hidden name"><code>cross_species_vig.Rmd</code></div>

    </div>

    
    
<p>In this vignette, we demonstrate how to integrate dataset from
different species.</p>
<div class="section level2">
<h2 id="step-1-load-the-data">Step 1: Load the data<a class="anchor" aria-label="anchor" href="#step-1-load-the-data"></a>
</h2>
<div class="section level3">
<h3 id="load-data-and-create-liger-object">1. Load data and create liger object<a class="anchor" aria-label="anchor" href="#load-data-and-create-liger-object"></a>
</h3>
<p>For this tutorial, we will use two datasets, which can all be
downloaded at <a href="https://www.dropbox.com/sh/y9kjoum8u469nj1/AADik2b2-Qo3os2QSWXdIAbna?dl=0" class="external-link uri">https://www.dropbox.com/sh/y9kjoum8u469nj1/AADik2b2-Qo3os2QSWXdIAbna?dl=0</a>
.</p>
<ul>
<li>The scRNA mouse dataset (Drop_mouse.RDS) is 28,366 genes by 71,639
cells.</li>
<li>The scRNA lizard dataset (RNA_lizard.RDS) is 15,345 genes by 4,202
cells.</li>
</ul>
<p>The input raw count matrices are presented in dgCMatrix class
objects, which is a common form of sparse matrix. Users can then create
a <code>liger</code> object with the two datasets. A named list is
required for submitting the datasets to the object creator.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/welch-lab/liger" class="external-link">rliger2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>ligerDotSize <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mouse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"cross_species_vignette_data/Drop_mouse.RDS"</span><span class="op">)</span></span>
<span><span class="va">lizard</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"cross_species_vignette_data/RNA_lizard.RDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mouse <span class="op">=</span> <span class="va">mouse</span>, lizard <span class="op">=</span> <span class="va">lizard</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="step-2-preprocessing-and-normalization">Step 2: Preprocessing and normalization<a class="anchor" aria-label="anchor" href="#step-2-preprocessing-and-normalization"></a>
</h2>
<div class="section level3">
<h3 id="normalization">2. Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h3>
<p>Liger simply normalizes the matrices by library size, without
multiplying a scale factor or applying “log1p” transformation.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="select-variable-genes">3. Select variable genes<a class="anchor" aria-label="anchor" href="#select-variable-genes"></a>
</h3>
<p>For cross-species analysis, we select shared, homologous genes
between the two species, as well as unshared, non-homologous genes from
the lizard dataset. The default setting of <code><a href="../reference/selectGenes.html">selectGenes()</a></code>
function selects for homologous genes shared between all datasets. To
enable selection of unshared genes, users need to specify the name(s) of
dataset(s) where unshared genes should be chosen from to
<code>useUnsharedDatasets</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="va">lig</span>, var.thres <span class="op">=</span> <span class="fl">0.3</span>, useUnsharedDatasets <span class="op">=</span> <span class="st">"lizard"</span>, unsharedThresh <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Wed Dec 20 16:04:31 2023 ... Selecting variable features for dataset: mouse</span></span></code></pre>
<pre><code><span><span class="co">## Wed Dec 20 16:04:34 2023 ...... 1683 features selected out of 12334 shared features</span></span></code></pre>
<pre><code><span><span class="co">## Wed Dec 20 16:04:34 2023 ... Selecting variable features for dataset: lizard</span></span></code></pre>
<pre><code><span><span class="co">## Wed Dec 20 16:04:34 2023 ...... 612 features selected out of 12334 shared features</span></span></code></pre>
<pre><code><span><span class="co">## Wed Dec 20 16:04:34 2023 ...... 166 features selected out of 3008 unshared features</span></span></code></pre>
<pre><code><span><span class="co">## Wed Dec 20 16:04:34 2023 ... Finally 1976 shared variable features selected.</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="scale-not-center">4. Scale not center<a class="anchor" aria-label="anchor" href="#scale-not-center"></a>
</h3>
<p>Then we scale the dataset. Three new matrices will be created under
the hook, two containing the shared variable features for both datasets,
and one for the unshared variable features of lizard data. Note the
Liger does not center the scaled data because iNMF/UINMF methods require
non-negative input.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="step-3-joint-matrix-factorization">Step 3: Joint Matrix Factorization<a class="anchor" aria-label="anchor" href="#step-3-joint-matrix-factorization"></a>
</h2>
<div class="section level3">
<h3 id="mosaic-integrative-nmf">5. Mosaic integrative NMF<a class="anchor" aria-label="anchor" href="#mosaic-integrative-nmf"></a>
</h3>
<p>Unshared Integrative Non-negative Matrix Factorization (UINMF) can be
applied with <code>runIntegration(..., method = "UINMF")</code>. A
standalone function <code><a href="../reference/runUINMF.html">runUINMF()</a></code> is also provided with more
detailed documentation and initialization setup. This step produces
factor gene loading matrices of the shared genes: <span class="math inline">\(W\)</span> for shared information across datasets,
and <span class="math inline">\(V\)</span> for dataset specific
information. Specific to UINMF method, additional factor gene loading
matrix of unshared genes, <span class="math inline">\(U\)</span> is also
produced. <span class="math inline">\(H\)</span> matrices, the cell
factor loading matrices are produced for each dataset and can be
interpreted as load rank representation of the cells.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runIntegration.html">runIntegration</a></span><span class="op">(</span><span class="va">lig</span>, k <span class="op">=</span> <span class="fl">30</span>, method <span class="op">=</span> <span class="st">"UINMF"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="step-4-quantile-normalization-and-joint-clustering">Step 4: Quantile Normalization and Joint Clustering<a class="anchor" aria-label="anchor" href="#step-4-quantile-normalization-and-joint-clustering"></a>
</h2>
<div class="section level3">
<h3 id="quantile-normalization">6. Quantile normalization<a class="anchor" aria-label="anchor" href="#quantile-normalization"></a>
</h3>
<p>The default reference dataset for quantile normalization is the
largest dataset, but the user should select the higher quality dataset
as the reference dataset, even if it is a smaller dataset. In this case,
the mouse dataset is considered higher quality than the lizard dataset,
so we set the mouse dataset to be the reference dataset.</p>
<p>After this step, the low-rank cell factor loading matrices, <span class="math inline">\(H\)</span>, are aligned and ready for cluster
definition.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantileNorm.html">quantileNorm</a></span><span class="op">(</span><span class="va">lig</span>, reference <span class="op">=</span> <span class="st">"mouse"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="leiden-clustering">7. Leiden clustering<a class="anchor" aria-label="anchor" href="#leiden-clustering"></a>
</h3>
<p>With the aligned cell factor loading information, we next apply
Leiden community detection algorithm to identify cell clusters.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCluster.html">runCluster</a></span><span class="op">(</span><span class="va">lig</span>, nNeighbors <span class="op">=</span> <span class="fl">30</span>, resolution <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="step-5-visualization">Step 5: Visualization<a class="anchor" aria-label="anchor" href="#step-5-visualization"></a>
</h2>
<div class="section level3">
<h3 id="dimensionality-reduction">8. Dimensionality reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction"></a>
</h3>
<p>We create a UMAP with the quantile normalized cell factor
loading.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">lig</span>, nNeighbors <span class="op">=</span> <span class="fl">30</span>, minDist <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plot-umap">9. Plot UMAP<a class="anchor" aria-label="anchor" href="#plot-umap"></a>
</h3>
<p>Next, we can visualize our returned factorized object by dataset to
check the integration between datasets, and also the cluster labeling on
the global population.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimRed.html">plotDatasetDimRed</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span></span></code></pre></div>
<p><img src="cross_species_vig_files/figure-html/visualizations-1.png" width="672"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimRed.html">plotClusterDimRed</a></span><span class="op">(</span><span class="va">lig</span>, legendNCol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="cross_species_vig_files/figure-html/visualizations-2.png" width="672"></p>
<p>We can also use the datasets’ original annotations to check the
correspondence between the cell types of the two species. The
annotations can be downloaded at <a href="https://www.dropbox.com/sh/y9kjoum8u469nj1/AADik2b2-Qo3os2QSWXdIAbna?dl=0" class="external-link uri">https://www.dropbox.com/sh/y9kjoum8u469nj1/AADik2b2-Qo3os2QSWXdIAbna?dl=0</a></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mouse_annies</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"cross_species_vignette_data/Dropviz_general_annotations.RDS"</span><span class="op">)</span></span>
<span><span class="va">lizard_annies</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"cross_species_vignette_data/lizard_labels.RDS"</span><span class="op">)</span></span></code></pre></div>
<p><code>cellMeta()&lt;-</code> method has the feature for partial
insertion implemented and is great for adding dataset specific metadata
to a new variable. If you are unsure about the order between given
variable and the cells in the object, argument <code>cellIdx</code> can
be used for precise locating. Users might use different strategies when
creating the index depending on different situation.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mouse_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"mouse_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mouse_annies</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/liger-class.html">cellMeta</a></span><span class="op">(</span><span class="va">lig</span>, <span class="st">"mouse_ann"</span>, cellIdx <span class="op">=</span> <span class="va">mouse_id</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">mouse_annies</span><span class="op">)</span></span>
<span></span>
<span><span class="va">lizard_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"lizard_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">lizard_annies</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/liger-class.html">cellMeta</a></span><span class="op">(</span><span class="va">lig</span>, <span class="st">"lizard_ann"</span>, cellIdx <span class="op">=</span> <span class="va">lizard_id</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">lizard_annies</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotDimRed.html">plotClusterDimRed</a></span><span class="op">(</span><span class="va">lig</span>, useCluster <span class="op">=</span> <span class="st">"mouse_ann"</span>, legendNCol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="cross_species_vig_files/figure-html/addAnn-1.png" width="672"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimRed.html">plotClusterDimRed</a></span><span class="op">(</span><span class="va">lig</span>, useCluster <span class="op">=</span> <span class="st">"lizard_ann"</span>, legendNCol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="cross_species_vig_files/figure-html/addAnn-2.png" width="672"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
