<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rliger">
<title>Import Data from Various Source • rliger</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Import Data from Various Source">
<meta property="og:description" content="rliger">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Import Data from Various Source</h1>
                        <h4 data-toc-skip class="author">Yichen
Wang</h4>
            
            <h4 data-toc-skip class="date">2024-03-16</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/vignettes/articles/import_data.Rmd" class="external-link"><code>vignettes/articles/import_data.Rmd</code></a></small>
      <div class="d-none name"><code>import_data.Rmd</code></div>
    </div>

    
    
<p>This article only focuses on reading in data presented in various
forms and provides detailed examples and explanations. Currently, we
support importing the following forms of input data into a <a href="../reference/liger-class.html">liger</a> object:</p>
<ul>
<li>
<a href="#data-presented-with-mtx-files">Flat MTX files</a>
containing sparse matrices, together with files containing barcodes and
feature names. These are mostly the 10X cellranger output.</li>
<li>
<a href="#data-presented-with-h5-files">H5 files</a> containing the
sparse data, barcodes, and feature names. These can be found also in 10X
cellranger output, as well as <a href="#importing-from-h5ad-file-python-anndata-object">H5AD files</a>
written from a Python AnnData object.</li>
<li>Loaded R objects of the following classes:
<ul>
<li><a href="#importing-r-object---dgcmatrix">dgCMatrix</a></li>
<li><a href="#importing-r-object---seurat">Seurat, preferably of V5
structure</a></li>
<li><a href="#importing-r-object---singlecellexperiment">SingleCellExperiment</a></li>
</ul>
</li>
</ul>
<p>Whichever type of data you want to bring to a LIGER analysis, we
generally recommend that it represents the raw information, such as RNA
raw counts, peak counts. LIGER performs optimally with its own
normalization (library-size/L1 normalization only) and scaling (no
centering) procedures, and other types of transformation may not be
compatible with LIGER’s assumptions.</p>
<p>The data imported from flat MTX files will be read into memory,
internally converted into dgCMatrix objects. Data from other R object
classes will also be converted into dgCMatrix objects. Data stored in H5
files will not be read into memory, as they are employed for helping
with large datasets that cannot be loaded into memory. The main liger
object constructor function, <code><a href="../reference/createLiger.html">createLiger()</a></code>, requires a
<strong>named list</strong> as its first argument, where each element is
a dataset to be integrated. Each element of the named list can be of
different types from described above and below. However, it is limited
that we don’t currently provide integration method that integrates H5
data with other types of data. That is, either you can integrate H5 data
with H5 data, or you can integrate in-memory data with in-memory data.
If integration under this situation is needed, we still provide some
solution.</p>
<p>If you need to import data presented in other forms but encounter
problems coercing them to supported formats, please feel free to <a href="https://github.com/welch-lab/liger/issues/new" class="external-link">open an Issue on
our GitHub repository</a> to request for support.</p>
<div class="section level2">
<h2 id="importing-10x-cellranger-output">Importing 10X Cellranger Output<a class="anchor" aria-label="anchor" href="#importing-10x-cellranger-output"></a>
</h2>
<p>Given the prevalence of 10X Genomics’ single-cell sequencing
technology, the <a href="https://www.10xgenomics.com/cn/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-overview" class="external-link">cellranger
output</a> has been the most commonly seen data source for real-world
analyses. We provide the following examples showing how to read
multi-sample datasets from cellranger output into a liger object.</p>
<div class="section level3">
<h3 id="data-presented-with-mtx-files">Data presented with MTX files<a class="anchor" aria-label="anchor" href="#data-presented-with-mtx-files"></a>
</h3>
<p>MTX files are proposed for efficiently storing sparse matrix data
into flat text files. Cellranger output MEX format data that contains a
MTX file together with a text file with barcodes and a text file with
feature names. For example, the following is the structure of the
cellranger output directory of one certain sample.</p>
<pre><code># shell commands in unix terminal
$ cd /home/jdoe/runs/sample345/outs
$ tree filtered_feature_bc_matrix
filtered_feature_bc_matrix
    ├── barcodes.tsv.gz
    ├── features.tsv.gz
    └── matrix.mtx.gz
0 directories, 3 files</code></pre>
<p>We provide a function <code><a href="../reference/read10X.html">read10X()</a></code> for identifying proper
files to use and loading them into memory. The function works for either
just one sample or the root directory that conatins multiple
samples.</p>
<div class="section level4">
<h4 id="reading-a-single-sample">Reading a single sample<a class="anchor" aria-label="anchor" href="#reading-a-single-sample"></a>
</h4>
<p>To load a single sample:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read10X.html">read10X</a></span><span class="op">(</span><span class="st">"path/to/filtered_feature_bc_matrix"</span><span class="op">)</span></span></code></pre></div>
<p>Depending on whether the cellranger version is before 3.0 or after,
the returned object <code>sample1</code> varies in structure. For
cellranger version 3.0 and later, the returned object is a named list
where each element is for a sample, even if we opt to only read one
sample. And this element would be another named list for available
feature types. For example when doing scRNAseq, it will be
<code>"Gene Expression"</code>.</p>
<pre><code>&gt; sample1
$sampleName
$sampleName$`Gene Expression`
  [... dgCMatrix]</code></pre>
<p>For older cellranger, <code>sample1</code> will a named list, named
by sample names, where each element is directly a dgCMatrix object.</p>
<pre><code>&gt; sample1
$sampleName
  [... dgCMatrix]</code></pre>
</div>
<div class="section level4">
<h4 id="reading-multiple-samples-from-root-directory">Reading multiple samples from root directory<a class="anchor" aria-label="anchor" href="#reading-multiple-samples-from-root-directory"></a>
</h4>
<p>The root directory that <code><a href="../reference/read10X.html">read10X()</a></code> expects is the
directory where you can see subdirecories named by sample names.</p>
<pre><code># shell commands in unix terminal
$ cd /home/jdoe/
$ tree runs
runs
    ├── sample123
    │   └── outs
    │       └── filtered_feature_bc_matrix
    │           ├── barcodes.tsv.gz
    │           ├── features.tsv.gz
    │           └── matrix.mtx.gz
    └── sample456
        └── outs
            └── filtered_feature_bc_matrix
                ├── barcodes.tsv.gz
                ├── features.tsv.gz
                └── matrix.mtx.gz</code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">multiSample</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read10X.html">read10X</a></span><span class="op">(</span><span class="st">"path/to/runs"</span><span class="op">)</span></span></code></pre></div>
<p>The returned object <code>multiSample</code> is a named list for each
sample available in the <code>runs/</code> folder. Similar to the case
of reading a single sample, the element for each sample will be a list
for available feature types for cellranger&gt;=3.0 or a dgCMatrix object
for older cellranger.</p>
</div>
<div class="section level4">
<h4 id="creating-a-liger-object-from-loaded-data">Creating a liger object from loaded data<a class="anchor" aria-label="anchor" href="#creating-a-liger-object-from-loaded-data"></a>
</h4>
<p>The required structure for the input of <code><a href="../reference/createLiger.html">createLiger()</a></code> is
a named list of matrices. For cellranger&gt;=3.0 case where the loaded
data contains substructure for potentially multiple feature types, we
need to extract the desired data and organize them a bit. The following
example shows how to take only the gene expression out of the loaded
data and organize them into the desired format to finally create a liger
object.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">multiSample</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read10X.html">read10X</a></span><span class="op">(</span><span class="st">"path/to/runs"</span><span class="op">)</span></span>
<span><span class="va">multiSample</span></span>
<span><span class="co">## $sample123</span></span>
<span><span class="co">## $sample123$`Gene Expression`</span></span>
<span><span class="co">##   [... dgCMatrix]</span></span>
<span><span class="co">## $sample123$`Antibody Capture`</span></span>
<span><span class="co">##   [... dgCMatrix]</span></span>
<span><span class="co">## $sample456</span></span>
<span><span class="co">## $sample456$`Gene Expression`</span></span>
<span><span class="co">##   [... dgCMatrix]</span></span>
<span><span class="co">## $sample456$`Antibody Capture`</span></span>
<span><span class="co">##   [... dgCMatrix]</span></span>
<span><span class="va">multiSample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">multiSample</span>, <span class="va">`[[`</span>, i <span class="op">=</span> <span class="st">"Gene Expression"</span><span class="op">)</span></span>
<span><span class="va">multiSample</span></span>
<span><span class="co">## $sample123</span></span>
<span><span class="co">##   [... dgCMatrix]</span></span>
<span><span class="co">## $sample456</span></span>
<span><span class="co">##   [... dgCMatrix]</span></span>
<span><span class="va">ligerObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="va">multiSample</span><span class="op">)</span></span></code></pre></div>
<p>For convenience, if you only want to work on RNA data, you can
directly load with <code><a href="../reference/read10X.html">read10XRNA()</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">multiSample</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read10X.html">read10XRNA</a></span><span class="op">(</span><span class="st">"path/to/runs"</span><span class="op">)</span></span>
<span><span class="va">multiSample</span></span>
<span><span class="co">## $sample123</span></span>
<span><span class="co">##   [... dgCMatrix]</span></span>
<span><span class="co">## $sample456</span></span>
<span><span class="co">##   [... dgCMatrix]</span></span>
<span><span class="va">ligerObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="va">multiSample</span><span class="op">)</span></span></code></pre></div>
<p>We also provide <code><a href="../reference/read10X.html">read10XATAC()</a></code> for loading only the ATAC
data from cellranger output. Note here that 10X provides different
platforms that contain ATAC modality, namingly “cellranger atac” and
“cellranger arc”. Users need to specify argument <code>pipeline</code>
for the function to correctly infer the file structure.</p>
</div>
</div>
<div class="section level3">
<h3 id="data-presented-with-h5-files">Data presented with H5 files<a class="anchor" aria-label="anchor" href="#data-presented-with-h5-files"></a>
</h3>
<p>Cellranger outputs identical information into H5 files, which contain
the same sparse matrix, barcodes and feature names. H5 files are based
on the HDF5 library, which is designed for storing and managing large
and complex data. They can be found at the same folder where you see the
<code>filtered_feature_bc_matrix/</code> folder, and the file would be
named by <code>filtered_feature_bc_matrix.h5</code>.</p>
<pre><code># shell commands in unix terminal
$ cd /home/jdoe/runs/sample345/
$ tree outs
outs
    ├── filtered_feature_bc_matrix
    │   ├── barcodes.tsv.gz
    │   ├── features.tsv.gz
    │   └── matrix.mtx.gz
    └── filtered_feature_bc_matrix.h5
    ...</code></pre>
<p>To create a liger object with H5 files, you simply need to provide
the path to each H5 files in a named list.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">H5PathList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    sample123 <span class="op">=</span> <span class="st">"/home/jdoe/runs/sample123/outs/filtered_feature_bc_matrix.h5"</span>,</span>
<span>    sample456 <span class="op">=</span> <span class="st">"/home/jdoe/runs/sample456/outs/filtered_feature_bc_matrix.h5"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">ligerObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="va">H5PathList</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="importing-r-object---dgcmatrix">Importing R object - dgCMatrix<a class="anchor" aria-label="anchor" href="#importing-r-object---dgcmatrix"></a>
</h2>
<p>“dgCMatrix”, supported by package <a href="https://CRAN.R-project.org/package=Matrix" class="external-link">Matrix</a>, is the core
class we adopt for holding the cell feature matrix in rliger. It
represents CSC (Compressed Sparse Column) matrix, which is the most
common form of sparse matrix in R given its efficiency in both memory
usage and computation.</p>
<p>We expect that users have each dgCMatrix object representing a single
dataset to be integrated. The main object constructor function,
<code><a href="../reference/createLiger.html">createLiger()</a></code>, expects datasets are organized in a
<strong>named list</strong> and being passed to the first argument.</p>
<pre class="importdgcmatrix"><code><span><span class="co"># NOT RUN, for demonstration only</span></span>
<span><span class="va">datasetList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"name1"</span> <span class="op">=</span> <span class="va">dgCMatrix1</span>, </span>
<span>    <span class="st">"name2"</span> <span class="op">=</span> <span class="va">dgCMatrix2</span>,</span>
<span>    <span class="st">"name3"</span> <span class="op">=</span> <span class="va">dgCMatrix3</span></span>
<span><span class="op">)</span></span>
<span><span class="va">ligerObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="va">datasetList</span><span class="op">)</span></span></code></pre>
<p>If your matrix is a combination of multiple datasets to be
integrated, we suggest using <code><a href="../reference/as.liger.html">as.liger()</a></code> with argument
<code>datasetVar</code> supplied with a factor object as the labeling to
indicate the proper splitting behavior.</p>
<pre class="importdgcmatrix2"><code><span><span class="co"># NOT RUN, for demonstration only</span></span>
<span><span class="va">datasetVar</span></span>
<span><span class="co">##  [1] sample1 sample1 sample1 sample1 ...</span></span>
<span><span class="co">## Levels: sample1 sample2 sample3</span></span>
<span><span class="va">ligerObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.liger.html">as.liger</a></span><span class="op">(</span><span class="va">dgCMatrix</span>, datasetVar <span class="op">=</span> <span class="va">datasetVar</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="importing-r-object---seurat">Importing R object - Seurat<a class="anchor" aria-label="anchor" href="#importing-r-object---seurat"></a>
</h2>
<p>Seurat has been the most popular R package for single-cell analysis.
We directly support <a href="liger_with_seurat.html">running rliger
functions on a Seurat object</a> without having to convert a Seurat
object into a liger object. Please click on the link to see a detailed
example, together with many use cases of interoperation between Seurat
and rliger.</p>
</div>
<div class="section level2">
<h2 id="importing-r-object---singlecellexperiment">Importing R object - SingleCellExperiment<a class="anchor" aria-label="anchor" href="#importing-r-object---singlecellexperiment"></a>
</h2>
<p><a href="https://bioconductor.org/packages/SingleCellExperiment/" class="external-link">SingleCellExperiment</a>
(SCE) has also been a popular class for single-cell analysis. Here’s how
we bring raw data from SCEs.</p>
<div class="section level3">
<h3 id="multiple-sces-each-a-dataset">Multiple SCEs, each a dataset<a class="anchor" aria-label="anchor" href="#multiple-sces-each-a-dataset"></a>
</h3>
<p>This is the most preferable case. Most of the times, people do QC and
filterings to remove unexpressing genes and low quality cells. If a SCE
is presented as a combined representation of QC’ed datasets, it’s likely
that unwanted zero entries would be introduced to keep the
dimensionality consistent, and this would unnecessarily increase the
memory usage and computation time. If you have multiple SCEs, each
representing a dataset, we recommend to pass them as a named list to
<code><a href="../reference/createLiger.html">createLiger()</a></code>.</p>
<pre class="importsce"><code><span><span class="co"># NOT RUN, for demonstration only</span></span>
<span><span class="va">sceList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"name1"</span> <span class="op">=</span> <span class="va">sce1</span>, </span>
<span>    <span class="st">"name2"</span> <span class="op">=</span> <span class="va">sce2</span>,</span>
<span>    <span class="st">"name3"</span> <span class="op">=</span> <span class="va">sce3</span></span>
<span><span class="op">)</span></span>
<span><span class="va">ligerObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="va">sceList</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="a-single-sce-including-many-datasets">A single SCE, including many datasets<a class="anchor" aria-label="anchor" href="#a-single-sce-including-many-datasets"></a>
</h3>
<p>If it happens that your single SCE object includes multiple datasets
that need to be integrated. We suggest using <code><a href="../reference/as.liger.html">as.liger()</a></code> to
convert it into a liger object. In this case, the second argument
<code>datasetVar</code> must be provided for inferring how to split the
data. Assume that the example <code>sce</code> object has a column
<code>sample</code> in its <code>colData</code>:</p>
<pre class="importsce2"><code><span><span class="co"># NOT RUN, for demonstration only</span></span>
<span><span class="va">ligerObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.liger.html">as.liger</a></span><span class="op">(</span><span class="va">sce</span>, datasetVar <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="importing-from-h5ad-file-python-anndata-object">Importing from H5AD file, Python AnnData object<a class="anchor" aria-label="anchor" href="#importing-from-h5ad-file-python-anndata-object"></a>
</h2>
<blockquote>
<p><strong>WARNING</strong>: We strongly suggest that you make a COPY of
the H5AD file for rliger analyis. rliger load the HDF5 file with
read-and-write mode and occasionally writes to the file, which might
cause the file unable to be loaded back to Python AnnData object. This
will be improved in the future.</p>
</blockquote>
<p><a href="https://anndata.readthedocs.io/en/latest/" class="external-link">AnnData</a> is a
popular Python class for single-cell analysis. When storing the object,
it is written to an H5AD file, which is also based on HDF5 library, and
is totally supported to be loaded into a liger object. Note that the
H5AD specification has been changed since anndata 0.8.0, and we only
support the new format. If you have H5AD file written by an older
version of anndata, you need to convert it to the new format using an
upgraded version of Python anndata package.</p>
<pre class="importh5ad"><code><span><span class="co"># NOT RUN, for demonstration only</span></span>
<span><span class="va">ligerObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    sample1 <span class="op">=</span> <span class="st">"path/to/sample1.h5ad"</span>,</span>
<span>    sample2 <span class="op">=</span> <span class="st">"path/to/sample2.h5ad"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  formatType <span class="op">=</span> <span class="st">"anndata"</span></span>
<span><span class="op">)</span></span></code></pre>
<p>A few limitations to note here are that:</p>
<ol style="list-style-type: decimal">
<li>We by default read the data located at <code>adata.X</code>.
However, due to the fact that people might be storing data at
<code>adata.layers['key']</code> or <code>adata.raw.X</code>, and LIGER
requires the raw data as the start point, we provide argument
<code>anndataX</code> to allow setting the path to the data. See
<code><a href="../reference/createH5LigerDataset.html">createH5LigerDataset()</a></code> for more details.</li>
<li>LIGER expects that each H5AD file contains only one dataset to be
integrated, because we are not loading the H5 based data into memory and
hence don’t yet have method to split the data by a given variable. This
will be updated in the future.</li>
</ol>
</div>
<div class="section level2">
<h2 id="working-with-mixed-types-of-data">Working with mixed types of data<a class="anchor" aria-label="anchor" href="#working-with-mixed-types-of-data"></a>
</h2>
<p>The <code><a href="../reference/createLiger.html">createLiger()</a></code> constructor function expects a named
list as its first input, while the element in this list can be of
various forms.</p>
<div class="section level3">
<h3 id="mixing-cellranger-output-h5-files-with-anndata-h5ad-files">Mixing cellranger output H5 files with anndata H5AD files<a class="anchor" aria-label="anchor" href="#mixing-cellranger-output-h5-files-with-anndata-h5ad-files"></a>
</h3>
<p>When importing a number of H5-based datasets that are of the same
specification, one can tweak the additional arguments of the function to
apply the same settings to all these datasets. However, if you want to
integrate H5 datasets of different types (e.g. cellranger output and
H5AD), you will need to load each dataset into ligerDataset object
individually, and then call the <code><a href="../reference/createLiger.html">createLiger()</a></code> function with
a list of these ligerDataset objects.</p>
<pre class="importmixed"><code><span><span class="co"># NOT RUN, for demonstration only</span></span>
<span><span class="va">ligerDataset1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createH5LigerDataset.html">createH5LigerDataset</a></span><span class="op">(</span><span class="st">"path/to/runs/sample/outs/filtered_feature_bc_matrix.h5"</span>, formatType <span class="op">=</span> <span class="st">"10x"</span><span class="op">)</span></span>
<span><span class="va">ligerDataset2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createH5LigerDataset.html">createH5LigerDataset</a></span><span class="op">(</span><span class="st">"path/to/sample2.h5ad"</span>, formatType <span class="op">=</span> <span class="st">"anndata"</span>, anndataX <span class="op">=</span> <span class="st">"raw/X"</span><span class="op">)</span></span>
<span><span class="va">ligerObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">ligerDataset1</span>, <span class="va">ligerDataset2</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="mixing-all-kinds-of-r-in-memory-objects">Mixing all kinds of R in-memory objects<a class="anchor" aria-label="anchor" href="#mixing-all-kinds-of-r-in-memory-objects"></a>
</h3>
<p>If each of your R objects contains a single dataset to be integrated,
you can simply pass them to the named list.</p>
<pre class="importmixed2"><code><span><span class="co"># NOT RUN, for demonstration only</span></span>
<span><span class="va">ligerObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    sample1 <span class="op">=</span> <span class="va">dgCMatrix1</span>, </span>
<span>    seurat2 <span class="op">=</span> <span class="va">seuratObj</span>,</span>
<span>    sce3 <span class="op">=</span> <span class="va">sce</span>, </span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
<p>Function <code>as.ligerDataset</code> is called under the hook to
convert each R object to a ligerDataset object, and it directly fetches
the <code>'counts'</code> layer/slot of the default assay from a Seurat
object, or the <code>'counts'</code> assay from an SCE object. If this
is not the desired behavior, you can manually extract the correct
matrix.</p>
<pre class="importmixed3"><code><span><span class="co"># NOT RUN, for demonstration only</span></span>
<span><span class="va">ligerObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    sample1 <span class="op">=</span> <span class="va">dgCMatrix1</span>, </span>
<span>    seurat2 <span class="op">=</span> <span class="fu">SeuratObject</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Layers.html" class="external-link">LayerData</a></span><span class="op">(</span><span class="va">seuratObj</span>, layer <span class="op">=</span> <span class="st">"somelayer"</span>, assay <span class="op">=</span> <span class="st">"someassay"</span><span class="op">)</span>,</span>
<span>    sce3 <span class="op">=</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"someassay"</span><span class="op">)</span>, </span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
<p>If the given Seurat or SCE objects contain multiple datasets, you can
use <code><a href="../reference/as.liger.html">as.liger()</a></code> to convert them into a liger object, with
<code>datasetVar</code> properly specified to have the datasets split.
Then <code><a href="https://rdrr.io/r/base/c.html" class="external-link">c()</a></code> method can be applied to merge multiple liger
objects into one, analogous to how one would concatenate R list
objects.</p>
<pre class="importmixed4"><code><span><span class="co"># NOT RUN, for demonstration only</span></span>
<span><span class="va">ligerObj1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.liger.html">as.liger</a></span><span class="op">(</span><span class="va">seuratObj</span>, datasetVar <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></span>
<span><span class="va">ligerObj1</span></span>
<span><span class="co">## An object of class liger with XXXXX cells</span></span>
<span><span class="co">## datasets(2): patient001 (XXXX cells), patient002 (XXXX cells)</span></span>
<span><span class="co">## cellMeta(10): dataset, barcode, group, nUMI, ..., hemo</span></span>
<span><span class="co">## varFeatures(0): </span></span>
<span><span class="co">## dimReds(0):</span></span>
<span><span class="va">ligerObj2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.liger.html">as.liger</a></span><span class="op">(</span><span class="va">sce</span>, datasetVar <span class="op">=</span> <span class="st">"batch"</span><span class="op">)</span></span>
<span><span class="va">ligerObj2</span></span>
<span><span class="co">## An object of class liger with XXXXX cells</span></span>
<span><span class="co">## datasets(2): someid101 (XXXX cells), someid103 (XXXX cells)</span></span>
<span><span class="co">## cellMeta(10): dataset, barcode, group, nUMI, ..., hemo</span></span>
<span><span class="co">## varFeatures(0):</span></span>
<span><span class="co">## dimReds(0):</span></span>
<span><span class="va">ligerObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">ligerObj1</span>, <span class="va">ligerObj2</span><span class="op">)</span></span>
<span><span class="co"># You can even insert more datasets with `dataset&lt;-()`</span></span>
<span><span class="fu"><a href="../reference/liger-class.html">dataset</a></span><span class="op">(</span><span class="va">ligerObj</span>, <span class="st">"pbmc"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">dgCMatrix1</span></span>
<span><span class="va">ligerObj</span></span>
<span><span class="co">## An object of class liger with XXXXX cells</span></span>
<span><span class="co">## datasets(5): patient001 (XXXX cells), patient002 (XXXX cells), someid101 (XXXX cells), someid103 (XXXX cells), pbmc (XXXX cells)</span></span>
<span><span class="co">## cellMeta(10): dataset, barcode, group, nUMI, ..., hemo</span></span>
<span><span class="co">## varFeatures(0):</span></span>
<span><span class="co">## dimReds(0):</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="mixing-h5-files-with-in-memory-objects">Mixing H5 files with in-memory objects<a class="anchor" aria-label="anchor" href="#mixing-h5-files-with-in-memory-objects"></a>
</h3>
<p>Technically, with the hints provided above, you can create a liger
object with both in-memory and H5-based datasets presented at the same
time. However, the integration methods will not work in this situation.
Generally, users will have to either write the in-memory part to H5
files and replace the datasets with the H5 copies accordingly, or read
subsampled data from H5 files to memory and replace the datasets with
the in-memory copies. Note that we assume that H5 files are employed for
holding large datasets that cannot be fit into the memory and this is
why we suggest subsampling them. If they are indeed okay to be loaded
into memory, you can totally load them with
<code>subsetLigerDataset(..., newH5 = FALSE)</code> and replace the
datasets with the full in-memory copies.</p>
<p>It is becoming more common that users have large datasets stored in
H5 files and are expensive to be all loaded into memory.
<code><a href="../reference/runOnlineINMF.html">runOnlineINMF()</a></code> runs highly efficient methods to integrate
datasets presented in this form. If you want to bring a rather smaller
in-memory dataset to the pool and have it integrated together, you can
write the dataset to an H5 file and then create a liger object with it.
In this way, the integration will have the maximum information captured
from all datasets.</p>
<pre class="writeh5"><code><span><span class="co"># NOT RUN, for demonstration only</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">smallData</span><span class="op">)</span></span>
<span><span class="co">## [1] "dgCMatrix"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/write.html" class="external-link">write</a></span><span class="op">(</span><span class="va">smallData</span>, <span class="st">"path/to/smallData.h5"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">largeLiger</span></span>
<span><span class="co">## An object of class liger with XXXXXX cells</span></span>
<span><span class="co">## datasets(n): sample1 (XXXXX cells), sample2 (XXXXX cells), ...</span></span>
<span><span class="co">## cellMeta(10): dataset, barcode, group, nUMI, ..., hemo </span></span>
<span><span class="co">## varFeatures(6000): AAA1, AAA2, BBB1, BBB2, ...</span></span>
<span><span class="co">## dimReds(0):  </span></span>
<span><span class="va">smallDataset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createH5LigerDataset.html">createH5LigerDataset</a></span><span class="op">(</span><span class="st">"path/to/smallData.h5"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/liger-class.html">dataset</a></span><span class="op">(</span><span class="va">largeLiger</span>, <span class="st">"sampleSmall"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">smallDataset</span></span>
<span></span>
<span><span class="va">largeLiger</span></span>
<span><span class="co">## An object of class liger with XXXXXX cells</span></span>
<span><span class="co">## datasets(n): sample1 (XXXXX cells), sample2 (XXXXX cells), ..., sampleSmall (XXXX cells)</span></span>
<span><span class="co">## cellMeta(10): dataset, barcode, group, nUMI, ..., hemo </span></span>
<span><span class="co">## varFeatures(6000): AAA1, AAA2, BBB1, BBB2, ...</span></span>
<span><span class="co">## dimReds(0):  </span></span></code></pre>
<p>However, after integration, downstream analyses that rely on
gene/feature expression values (e.g. differential expression analysis)
would still need in-memory form of the data. We provide
<code><a href="../reference/downsample.html">downsample()</a></code> to sub-sample data from H5 files and read the
subset to memory, with fraction of dataset source and/or cell type
balanced. This might be improved in the future.</p>
<p>The following example would randomly sample 4,000 cells from each
dataset in <code>largeLiger</code> and load only the normalized data
into memory. The <code>newH5</code> argument is set to
<code>FALSE</code> to load the data into memory, and the
<code>useSlot</code> argument is set to <code>"normData"</code> to load
only the normalized data. The <code>maxCells</code> argument is set to
<code>4000</code> to sample 4,000 cells from each dataset. The resulting
<code>downsampledData</code> is a liger object with the same structure
as <code>largeLiger</code>, but with only the downsampled normalized
in-memory data.</p>
<pre class="downsampleh5"><code><span><span class="co"># NOT RUN, for demonstration only</span></span>
<span><span class="va">downsampledData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/downsample.html">downsample</a></span><span class="op">(</span><span class="va">largeLiger</span>, newH5 <span class="op">=</span> <span class="cn">FALSE</span>, maxCells <span class="op">=</span> <span class="fl">4000</span>, useSlot <span class="op">=</span> <span class="st">"normData"</span><span class="op">)</span></span>
<span><span class="va">downsampledData</span></span>
<span><span class="co">## An object of class liger with XXXXXX cells</span></span>
<span><span class="co">## datasets(n): sample1 (4000 cells), sample2 (4000 cells), ..., sampleSmall (4000 cells)</span></span>
<span><span class="co">## cellMeta(10): dataset, barcode, group, nUMI, ..., leiden_cluster</span></span>
<span><span class="co">## varFeatures(6000): AAA1, AAA2, BBB1, BBB2, ...</span></span>
<span><span class="co">## dimReds(1): UMAP</span></span></code></pre>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
