<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rliger">
<title>Install LIGER with R • rliger</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Install LIGER with R">
<meta property="og:description" content="rliger">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Install LIGER with R</h1>
                        <h4 data-toc-skip class="author">Yichen
Wang</h4>
            
            <h4 data-toc-skip class="date">2024-03-20</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/vignettes/articles/installation.Rmd" class="external-link"><code>vignettes/articles/installation.Rmd</code></a></small>
      <div class="d-none name"><code>installation.Rmd</code></div>
    </div>

    
    
<p>Before setting up the “rliger” package, users should have <a href="https://www.r-project.org/" class="external-link">R</a> version 3.6.0 or higher.</p>
<div class="section level2">
<h2 id="liger-is-available-on-cran">LIGER is available on CRAN<a class="anchor" aria-label="anchor" href="#liger-is-available-on-cran"></a>
</h2>
<p>To install the latest stable release, please run the following
command in R console:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"rliger"</span><span class="op">)</span></span></code></pre></div>
<p>Besides, we now move the core implementation of iNMF related
algorithms to a separate package <a href="https://github.com/welch-lab/RcppPlanc" class="external-link">RcppPlanc</a>, with decent
performance improvement. It <strong>must</strong> be installed for
performing integration analysis. We currently host it on <a href="https://welch-lab.r-universe.dev/RcppPlanc" class="external-link">R-universe</a> as we
are actively submitting it to CRAN. Please run the following command in
R console to install it.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"RcppPlanc"</span>, repos <span class="op">=</span> <span class="st">"https://welch-lab.r-universe.dev"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="building-from-source">Building from source<a class="anchor" aria-label="anchor" href="#building-from-source"></a>
</h2>
<p>The developmental versions as well as source version of the previous
versions are accessible from GitHub. To install the latest version from
GitHub, please run the following command in R console:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span>    <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"welch-lab/liger"</span><span class="op">)</span></span></code></pre></div>
<p>If you still need to work with the previous version (1.0.1):</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"welch-lab/liger"</span>, ref <span class="op">=</span> <span class="st">"v1.0.1"</span><span class="op">)</span></span></code></pre></div>
<p>If you encounter issues when building from source, please see below
sections for details about some dependencies.</p>
<p><strong>Please note that version 2.0.0 is massively updated compared
to prior versions, and functions are not directly compatible with liger
objects created with the old versions which users might possess. Please
use <code><a href="../reference/readLiger.html">readLiger()</a></code> to load old RDS files or
<code><a href="../reference/convertOldLiger.html">convertOldLiger()</a></code> to convert a loaded object into the
up-to-date structure. You might need to be careful when overwriting
existing analysis because we don’t provide methods to convert new the
structure backwards.</strong></p>
<div class="section level3">
<h3 id="compiler-setup">Compiler setup<a class="anchor" aria-label="anchor" href="#compiler-setup"></a>
</h3>
<p>Windows users will need to install <a href="https://cran.r-project.org/bin/windows/Rtools/" class="external-link">Rtools</a> which
can be downloaded from CRAN. After downloading it, open the file and
follow the prompt to install.</p>
<p>MacOS users will need Clang and gfortran, which can also be found on
<a href="https://mac.r-project.org/tools/" class="external-link">CRAN Mac Tools</a>.</p>
</div>
<div class="section level3">
<h3 id="installing-hdf5-library">Installing HDF5 Library<a class="anchor" aria-label="anchor" href="#installing-hdf5-library"></a>
</h3>
<p>HDF5 library is required for interacting with H5 files. It can be
installed via the following commands.</p>
<table class="table">
<colgroup>
<col width="55%">
<col width="44%">
</colgroup>
<thead><tr class="header">
<th align="left">System</th>
<th align="left">Command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">macOS (using Homebrew or Conda)</td>
<td align="left">
<code>brew install hdf5</code> or
<code>conda install -c anaconda hdf5</code>
</td>
</tr>
<tr class="even">
<td align="left">Debian-based systems (including Ubuntu)</td>
<td align="left"><code>sudo apt-get install libhdf5-dev</code></td>
</tr>
<tr class="odd">
<td align="left">Systems supporting yum and RPMs</td>
<td align="left"><code>sudo yum install hdf5-devel</code></td>
</tr>
<tr class="even">
<td align="left">Windows</td>
<td align="left">Go to <a href="https://www.hdfgroup.org/downloads/hdf5/" class="external-link">HDF5 website</a> and
download the proper installer</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="installing-rcppplanc">Installing RcppPlanc<a class="anchor" aria-label="anchor" href="#installing-rcppplanc"></a>
</h3>
<p><a href="https://github.com/welch-lab/RcppPlanc" class="external-link">RcppPlanc</a> is an
extension R package built basing on <a href="https://github.com/ramkikannan/planc" class="external-link">Planc</a>. We implemented
highly optimized iNMF algorithm and its variants here for the new
version upgrade. To install it from source:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"welch-lab/RcppPlanc"</span><span class="op">)</span></span></code></pre></div>
<p>Please refer to <a href="https://github.com/welch-lab/RcppPlanc" class="external-link">RcppPlanc GitHub
repository</a> for detail and more support.</p>
</div>
<div class="section level3">
<h3 id="installing-fit-sne">Installing FIt-SNE<a class="anchor" aria-label="anchor" href="#installing-fit-sne"></a>
</h3>
<p>For dimensionality reduction, we have options of UMAP and t-SNE. For
UMAP, we depend on <em>uwot</em> which is readily scalable with large
datasets. For t-SNE, we by default use <em>Rtsne</em>, which is not
scalable for large datasets. We allow using another implementation of
t-SNE, namingly FIt-SNE, for efficient computation. FIt-SNE is not
distributed with the package or CRAN, thus users need to install it
following steps below.</p>
<div class="section level4">
<h4 id="create-a-directory-fo-organizing-the-source-code">1. Create a directory fo organizing the source code<a class="anchor" aria-label="anchor" href="#create-a-directory-fo-organizing-the-source-code"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="bu">cd</span> path/you/need</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">mkdir</span> FItSNE_local</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="install-fftw-library-">1. Install FFTW library.<a class="anchor" aria-label="anchor" href="#install-fftw-library-"></a>
</h4>
<p>Download FFTW from <a href="http://www.fftw.org/" class="external-link">here</a>. For UNIX
like users, run the following command in terminal. The following
commands in terminal show how to locally install the library of version
3.3.10 as this tutorial is written down. You can check their official
website out if an update is available. You can simply replace the
version number in the URL and the file name to update the
installation.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="bu">cd</span> FItSNE_local</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="ex">curl</span> ftp://ftp.fftw.org/pub/fftw/fftw-3.3.10.tar.gz <span class="at">-o</span> fftw-3.3.10.tar.gz</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xvzf</span>  fftw-3.3.10.tar.gz</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="bu">cd</span> fftw-3.3.10</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="ex">./configure</span> <span class="at">--prefix</span><span class="op">=</span><span class="kw">`</span><span class="bu">pwd</span><span class="kw">`</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="fu">make</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="fu">make</span> install prefix=<span class="kw">`</span><span class="bu">pwd</span><span class="kw">`</span></span></code></pre></div>
<p>Here the FFTW library is installed in the directory where the source
code is located. You can change the part <code>pwd</code> to any other
path (no quotes needed) or even remove the <code>--prefix</code> token
to globally install the library if you do have the root access. A folder
<code>lib</code> with files named by <code>libfftw3*</code> must be
presented at current location.</p>
</div>
<div class="section level4">
<h4 id="install-fit-sne">2. Install FIt-SNE<a class="anchor" aria-label="anchor" href="#install-fit-sne"></a>
</h4>
<p>We next download FIt-SNE from GitHub and compile the part we need for
the R wrapper functionality.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="bu">cd</span> ../ <span class="co"># back to FItSNE_local</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/KlugerLab/FIt-SNE.git</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="bu">cd</span> FIt-SNE</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="ex">g++</span> <span class="at">-std</span><span class="op">=</span>c++11 <span class="at">-O3</span>  src/sptree.cpp src/tsne.cpp src/nbodyfft.cpp <span class="at">-o</span> bin/fast_tsne <span class="at">-pthread</span> <span class="at">-L..</span>/fftw-3.3.10/lib <span class="at">-lfftw3</span> <span class="at">-lm</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre></div>
<p>A folder <code>bin</code> with a file named <code>fast_tsne</code>
must be presented at current location. The command <code>pwd</code>
shows the location to the binary and you will need the printed path for
the next step.</p>
</div>
<div class="section level4">
<h4 id="run-t-sne-with-flt-sne">3. Run t-SNE with Flt-SNE<a class="anchor" aria-label="anchor" href="#run-t-sne-with-flt-sne"></a>
</h4>
<p>Now, we’ve had the binary compiled, and command <code>pwd</code>
shows the location to the binary. Then back to an R LIGER analysis, the
following command can be used to invoke the binary. Just simply pass a
string containing what <code>pwd</code> shows to the argument
<code>fitsnePath</code> in <code><a href="../reference/runTSNE.html">runTSNE()</a></code> function and switch
<code>method = "fftRtsne"</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ligerObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runTSNE.html">runTSNE</a></span><span class="op">(</span><span class="va">ligerObj</span>, method <span class="op">=</span> <span class="st">"fftRtsne"</span>, fitsnePath <span class="op">=</span> <span class="st">"path/you/need/FItSNE_local/FIt-SNE"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
