<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rliger">
<title>Joint definition of cell types from multiple scRNA-seq datasets • rliger</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Joint definition of cell types from multiple scRNA-seq datasets">
<meta property="og:description" content="rliger">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    <a class="dropdown-item" href="../articles/benchmark.html">Benchmarking integration quality</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<link href="Integrating_multi_scRNA_data_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="Integrating_multi_scRNA_data_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="Integrating_multi_scRNA_data_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="Integrating_multi_scRNA_data_files/datatables-binding-0.33/datatables.js"></script><link href="Integrating_multi_scRNA_data_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="Integrating_multi_scRNA_data_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="Integrating_multi_scRNA_data_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="Integrating_multi_scRNA_data_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="Integrating_multi_scRNA_data_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Joint definition of cell types from multiple scRNA-seq datasets</h1>
                        <h4 data-toc-skip class="author">Yichen Wang,
Joshua Sodicoff and Joshua Welch</h4>
            
            <h4 data-toc-skip class="date">2024-10-01</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/vignettes/articles/Integrating_multi_scRNA_data.rmd" class="external-link"><code>vignettes/articles/Integrating_multi_scRNA_data.rmd</code></a></small>
      <div class="d-none name"><code>Integrating_multi_scRNA_data.rmd</code></div>
    </div>

    
    
<p>This article is an general walkthrough of all main LIGER
functionalities. We demonstrate the usage of the <em>rliger</em> package
in the style of the R Console, which can be accessed through an R
development environment (e.g., RStudio) or directly from the R command
line.</p>
<p>LIGER relies on iNMF (integrative Non-negative Matrix Factorization,
<a href="https://doi.org/10.1016/j.cell.2019.05.006" class="external-link">Welch, JD 2019</a>)
to jointly factorize single-cell data into interpretable low-dimensional
representations, and integrated clustering can be obtained upon aligned
cell factor loading. The package also provides a variety of utilities
for visualization and analysis of clustering, gene expression across
datasets, comparisons of cluster assignments and etc.</p>
<div class="section level2">
<h2 id="setup-the-session">Setup the session<a class="anchor" aria-label="anchor" href="#setup-the-session"></a>
</h2>
<p>To get started, please make sure that <em>rliger</em> is successfully
installed in R, as guided on the <a href="installation.html">Installation help page</a>. Then, we load the
package to the current R session, together with a few other helpful
packages. <a href="https://dplyr.tidyverse.org/" class="external-link"><em>dplyr</em></a> has
been widely used for data table manupulation and is famous for its
unique user-friendly syntax. <a href="https://wilkelab.org/cowplot/index.html" class="external-link"><em>cowplot</em></a> is a
package that provides a simple and consistent way to combine multiple
plots into a single figure.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://welch-lab.github.io/liger/" class="external-link">rliger</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-the-data">Load the data<a class="anchor" aria-label="anchor" href="#load-the-data"></a>
</h2>
<p>In this tutorial, we will be integrating data from control and
interferon-stimulated PBMCs from <a href="https://www.nature.com/articles/nbt.4042" class="external-link">Kang et al, 2017</a>.
The data can be found in the Gene Expression Omnibus, <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE96583" class="external-link">Series
GSE96583</a>. This dataset was originally in the form of output from the
10X Cellranger pipeline. In this tutorial, we prepared a downsampled
version of the data.</p>
<ul>
<li><a href="https://figshare.com/ndownloader/files/47748940" class="external-link"><strong>Click to
download <code>pbmc_ifnb_example.tar.gz</code></strong></a></li>
</ul>
<p>After downloading the archive file (.tar.gz) to your working
directory, we need to extract the content to a folder. You can use any
unzip tool available on your computer to extract the content. For
platform without a graphical UI, operate the following command
<strong>in its terminal (not in R)</strong>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xvzf</span> pbmc_ifnb_example.tar.gz</span></code></pre></div>
<p>This will result you a folder called <code>pbmc_ifnb_example</code>
in your working directory, with inner structure like below:</p>
<pre><code>pbmc_ifnb_example
├── ctrl
│   └── outs
│       └── filtered_feature_bc_matrix
│           ├── barcodes.tsv.gz
│           ├── features.tsv.gz
│           └── matrix.mtx.gz
└── stim
    └── outs
        └── filtered_feature_bc_matrix
            ├── barcodes.tsv.gz
            ├── features.tsv.gz
            └── matrix.mtx.gz</code></pre>
<p>Now we can start to work in R to have the two datasets loaded. When
doing your own hands on, <strong>make sure to replace the path to the
folder with the actual path on your machine</strong>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Make sure to modify the following path to reflect the exact location of the folder</span></span>
<span><span class="va">pathToFolder</span> <span class="op">&lt;-</span> <span class="st">"pbmc_ifnb_example"</span></span>
<span><span class="va">rawList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read10X.html">read10XRNA</a></span><span class="op">(</span><span class="va">pathToFolder</span><span class="op">)</span></span>
<span><span class="va">pbmcLiger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="va">rawList</span>, organism <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span>
<span><span class="va">pbmcLiger</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class liger with 13999 cells</span></span>
<span><span class="co">## datasets(2): ctrl (6548 cells), stim (7451 cells) </span></span>
<span><span class="co">## cellMeta(7): dataset, barcode, nUMI, ..., hemo </span></span>
<span><span class="co">## varFeatures(0):  </span></span>
<span><span class="co">## dimReds(0):</span></span></code></pre>
<p>The loaded raw data <code>rawList</code> is a list object containing
two matrices for the “ctrl” and “stim” datasets. Then
<code><a href="../reference/createLiger.html">createLiger()</a></code> will take the list of all datasets to be
integrated as an input and construct the <a href="liger_object.html">liger object</a>.</p>
<details><summary><b>Convenience function for loading the example object</b></summary><p>For convenience, we have also prepared the pre-processed data which
are ready to use, easily loaded with <code><a href="../reference/importVignetteData.html">importPBMC()</a></code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This is not run at rendering time.</span></span>
<span><span class="va">pbmcLiger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importVignetteData.html">importPBMC</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</details><br><p>For creating a liger object from raw counts data or any other types
of source (e.g. import cellranger H5 files, convert from Seurat,
SingleCellExperiment, or even H5AD file), please refer to the <a href="import_data.html">detailed tutorial for importing data</a>.</p>
</div>
<div class="section level2">
<h2 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<p>It is always the best practice to check the quality control metrics
of the scRNAseq data before performing any other processing, in order to
rule out cells of poor quality that might affect the downstream
analysis. <em>rliger</em> by default have the most frequently used
metrics calculated when <code><a href="../reference/createLiger.html">createLiger()</a></code> is called, with the
following variables accessible in <code>cellMeta(pbmcLiger)</code> or
with the <code>$</code> operator:</p>
<ul>
<li>
<code>pbmcLiger$nUMI</code>: The sum of UMI counts for each
cell.</li>
<li>
<code>pbmcLiger$nGene</code>: The number of genes detected in each
cell.</li>
<li>
<code>pbmcLiger$mito</code>: The percentage of mitochondrial genes
detected in each cell.</li>
<li>
<code>pbmcLiger$ribo</code>: The percentage of ribosomal genes
detected in each cell.</li>
<li>
<code>pbmcLiger$hemo</code>: The percentage of hemoglobin genes
detected in each cell.</li>
</ul>
<p>The function <code><a href="../reference/plotViolin.html">plotTotalCountViolin()</a></code> and
<code><a href="../reference/plotViolin.html">plotGeneDetectedViolin()</a></code> show the density distribution of
<code>nUMI</code> and <code>nGene</code>, respectively.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plotNUMI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotViolin.html">plotTotalCountViolin</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, dot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">plotNGene</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotViolin.html">plotGeneDetectedViolin</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, dot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">plotNUMI</span>, <span class="va">plotNGene</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-4-1.png" width="960"></p>
<p>For plotting the mitochondrial gene expression percentage, please use
the following command:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCellViolin.html">plotCellViolin</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, <span class="st">"mito"</span>, groupBy <span class="op">=</span> <span class="st">"dataset"</span>, dot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-5-1.png" width="480"></p>
<p><strong>It happens that there is no mitochondrial gene detected in
the datasets we use for this tutorial.</strong> Users seeing such result
with their own data need to pay attention to species setting for
identifying the mitochondrial genes and rerun
<code><a href="../reference/runGeneralQC.html">runGeneralQC()</a></code> (click to see usage). If there are
mitochondrial genes detected, it is recommended to filter out cells with
high mitochondrial gene expression percentage, as they are likely to be
dead or dying cells.</p>
<p>There are two approaches of filtering genes and cells in a liger
object. The first is to use <code><a href="../reference/removeMissing.html">removeMissing()</a></code> function, which
mainly removes non-expressing genes and cells with no counts. This
function also allows removing genes that are expressed in too few cells
(argument <code>minCells</code>) and cells that express too few genes
(argument <code>minFeatures</code>). The following command removes cells
that express less than 200 genes.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcLiger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeMissing.html">removeMissing</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, minFeatures <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<p>The second way is to use R’s native matrix subsetting syntax and use
the <code>cellMeta(pbmcLiger)</code> variables which are accessible with
<code>$</code> operator. The following command</p>
<ul>
<li>keeps cells with total counts greater than 500</li>
<li>keeps cells with more than 200 detected genes</li>
<li>keeps cells with mitochondrial gene expression percentage less than
5%</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcLiger</span> <span class="op">&lt;-</span> <span class="va">pbmcLiger</span><span class="op">[</span>, <span class="va">pbmcLiger</span><span class="op">$</span><span class="va">nUMI</span> <span class="op">&gt;</span> <span class="fl">500</span> <span class="op">&amp;</span> <span class="va">pbmcLiger</span><span class="op">$</span><span class="va">mito</span> <span class="op">&lt;</span> <span class="fl">5</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preprocess">Preprocess<a class="anchor" aria-label="anchor" href="#preprocess"></a>
</h2>
<p>Before we can run iNMF on our datasets, we must run several
preprocessing steps to normalize expression data to account for
differences in sequencing depth and efficiency between cells, identify
variably expressed genes, and scale the data so that each gene has the
same variance. Note that because nonnegative matrix factorization
requires positive values, we do not center the data by subtracting the
mean. We also do not log transform the data.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcLiger</span> <span class="op">&lt;-</span> <span class="va">pbmcLiger</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>For gene selection, we use a customized approach that selects gene
for each dataset separately and take the union (or intersection) at
last. The selection for each dataset is based on a variance threshold
and thus the number of genes selected can be different for each dataset.
Users may see the message log from the run and decide whether to raise
or lower the threshold for each dataset to obtain optimal number of
genes being selected.</p>
</div>
<div class="section level2">
<h2 id="integration-with-joint-matrix-factorization">Integration with Joint Matrix Factorization<a class="anchor" aria-label="anchor" href="#integration-with-joint-matrix-factorization"></a>
</h2>
<div class="section level3">
<h3 id="joint-factorization">Joint factorization<a class="anchor" aria-label="anchor" href="#joint-factorization"></a>
</h3>
<p>We are now able to perform iNMF on the scaled datasets. The key
parameter for this analysis is <code>k</code>, the number of ranks each
dataset is factorized into (analogous to the number of principal
components in PCA). In general, we find that a value of <code>k</code>
between 20 and 40 is suitable for most analyses and that results are
robust for choice of <code>k</code>. Because LIGER is an unsupervised,
exploratory approach, there is no single “right” value for
<code>k</code>. In practice, users choose <code>k</code> from a
combination of biological prior knowledge and other information. A
higher <code>k</code> value might be considered when there are more
distinct composition expected from all datasets involved. A too-high
<code>k</code> value might result in splitting a distinct cell
population into unnecessary subpopulations, while a too-low
<code>k</code> value might not be sufficient enough for separating a
distinct cell population apart from others. For this tutorial, we set
<code>k = 30</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcLiger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runIntegration.html">runIntegration</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, k <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p>Starting from rliger 2.0.0, we use an optimized implementation of
iNMF. Here we deprecated the parameter <code>thresh</code> which stands
for a convergence detecter in order to speed up each algorithm iteration
by omitting the calculation of objective error.</p>
</blockquote>
<p>The factorization yields several lower dimension matrices:</p>
<ul>
<li>an <span class="math inline">\(H\)</span> matrix for each dataset,
representing the factor loadings for each cell. A factor can be regarded
as a metagene.</li>
<li>the <span class="math inline">\(W\)</span> matrix of shared gene
loadings for each factor.</li>
<li>A <span class="math inline">\(V\)</span> matrix for each dataset,
standing for the dataset-specific gene loadings for each factor.</li>
</ul>
<p>Please refer to <a href="liger_object.html">liger object
documentation</a> for how to access them.</p>
<p>The time consumption of this step is linearly dependent of the size
of the datasets, in terms of number of cells, number of variable genes
selected, and the value of <code>k</code>. The implementation supports
OpenMP multi-threading on non-Apple-Sillicon machines, and therefore
using a machine with a number of cores allocated and setting argument
<code>nCores</code> to a &gt;1 integer helps speeding it up. While
<code>nCores</code> setting is not applicable on Apple Sillicon
machines, the benchmarking there still shows comparable performance.</p>
</div>
<div class="section level3">
<h3 id="factor-alignment">Factor alignment<a class="anchor" aria-label="anchor" href="#factor-alignment"></a>
</h3>
<p>As introduced above, the <span class="math inline">\(H\)</span>
matrices for each dataset can be treated as low-dimensional
representation of cells and can be used for clustering. These matrices
are interpretable in a way that each factor can be considered as a
metagene, cell identity program, biological activity program and etc. We
need one last step to align the factor loading values across datasets,
so that cells voted to be of the same “type” by the factors will have a
aligned loading values across datasets, before calling a clustering
algorithm or generating embeddings for visualization.</p>
<p>From <em>rliger 2.1.0</em>, we introduce a newly developed cell
factor loading alignment method.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcLiger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/alignFactors.html">alignFactors</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, method <span class="op">=</span> <span class="st">"centroidAlign"</span><span class="op">)</span></span></code></pre></div>
<p>This method treats the factor loadings in each cell as
soft-clustering probabilities and therefore can yield centroids for each
cluster per dataset. We then linearly transform the scaled <span class="math inline">\(H\)</span> matrices by moving the centroids of the
same cluster from different datasets towards each other. This method is
benchmarked to show much better overall performance than the previous
quantile normalization method. While having comparable power in removing
batch effect, the centroid alignment method has much improved ability to
conserve biological information, especially when uncommon cell types are
present in part of the datasets.</p>
<details><summary><b>Running conventianal quantile normalization</b></summary><p>Quantile normalization relies on the selection of a reference dataset
and matches the quantiles of the factor loading across a cluster in each
dataset to that of the reference dataset. This method is powerful when
the same cell types are expected to be present in all datasets (e.g. the
PBMC datasets demonstrated in this tutorial) and provides strong batch
effect removal. However, it may not be optimal when the cell types are
not shared across datasets or when the cell types are not evenly
distributed across datasets.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcLiger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/alignFactors.html">alignFactors</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, method <span class="op">=</span> <span class="st">"quantileNorm"</span><span class="op">)</span></span></code></pre></div>
</details><br><p>The factor loading alignment produces matrix <span class="math inline">\(H.norm\)</span>, which is considered as the final
low-dimensional representation that integrates the datasets together,
which can be then used for clustering or visualization.</p>
</div>
<div class="section level3">
<h3 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h3>
<p>We run Leiden community detection on SNN graph inferred from the
aligned cell factors loading, an approach commonly used for single-cell
data. The Leiden algorithm excels at merging small clusters into broad
cell classes and thus may be more desirable in some cases than the
maximum factor assignments.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcLiger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCluster.html">runCluster</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, resolution <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p>Starting from rliger 2.0.0, cluster labeling will be stored in cell
metadata, which can be accessed with <code>cellMeta(pbmcLiger)</code>.
Use argument <code>clusterName</code> in <code><a href="../reference/runCluster.html">runCluster()</a></code> to
specify unique variable names for the result so that multiple cluster
labeling variables can be stored at the same time.</p>
</blockquote>
</div>
<div class="section level3">
<h3 id="visualize-the-integration">Visualize the integration<a class="anchor" aria-label="anchor" href="#visualize-the-integration"></a>
</h3>
<p>To visualize the clustering of cells, we can project the aligned cell
factors to 2D space. LIGER supports both UMAP and t-SNE for this
purpose.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcLiger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, minDist <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p>Starting from rliger 2.0.0, the slot for storing dimensionality
reduction matrices will be renamed to “dimReds”. It will be a list that
can hold multiple low dimensional matrices that match to all datasets by
cell identifiers. Users can access individual matrix with
<code>dimRed(object, "name")</code>. Use argument
<code>dimredName</code> in <code><a href="../reference/runUMAP.html">runUMAP()</a></code> (or
<code><a href="../reference/runTSNE.html">runTSNE()</a></code> to specify unique names for the UMAP/tSNE result
so that it allows storing multiple low-dimensional representation
matrices at the same time.</p>
</blockquote>
<p><code><a href="../reference/plotDimRed.html">plotByDatasetAndCluster()</a></code> returns two graphs, generated
with the UMAP in the previous step. The first colors cells by dataset of
origin, and the second by cluster as determined by previous clustering
step. The plots provide visual confirmation that the datasets are well
aligned and the clusters are consistent with the shape of the data as
revealed by UMAP.</p>
<p>The two subplots can individually be generated with
<code><a href="../reference/plotDimRed.html">plotDatasetDimRed()</a></code> and <code><a href="../reference/plotDimRed.html">plotClusterDimRed()</a></code>,
respectively.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimRed.html">plotByDatasetAndCluster</a></span><span class="op">(</span><span class="va">pbmcLiger</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/4-1-1.png" width="1344" style="display: block; margin: auto;"></p>
<p>Sometimes it is also important to have a look at the cluster
distribution within each dataset, because the UMAP plot colored by
dataset can be too messy when different colors are expected to be mixed
up. This can be achieved by running <code><a href="../reference/plotDimRed.html">plotClusterDimRed()</a></code>
with setting <code>splitBy = "dataset"</code>. Additionally, we can make
a contour density plot of the dimensionality reduction to further assist
interpreting the distribution.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clusterUmapList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDimRed.html">plotClusterDimRed</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, splitBy <span class="op">=</span> <span class="st">"dataset"</span>, title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pbmcLiger</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">densityList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDensityDimRed.html">plotDensityDimRed</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, splitBy <span class="op">=</span> <span class="st">"dataset"</span>, combinePlot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co"># Use cowplot's functionality to bring individual plots in the list together</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">clusterUmapList</span>, <span class="va">densityList</span><span class="op">)</span>, align <span class="op">=</span> <span class="st">"hv"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-10-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>From the plots above, we can visually see that there is a location
shift at cluster 0. But otherwise the density pattern in most of the
parts is very similar between the two datasets, which is a good sign of
successful integration.</p>
</div>
</div>
<div class="section level2">
<h2 id="explore-and-interpret-the-factors">Explore and interpret the factors<a class="anchor" aria-label="anchor" href="#explore-and-interpret-the-factors"></a>
</h2>
<p>Non-negative matrix factorization approach is known for its
interpretability of the information from the low dimensional result. To
directly study the impact of factors on the clustering and determine
what genes load most highly on each factor, we use the
<code><a href="../reference/plotGeneLoadings.html">plotGeneLoadings()</a></code> function, which returns plots of factor
loading on the dimensionality reduction and highly loaded genes by
dataset for each factor. Knowing the highly loaded gene of each factor
largely helps helps understanding what biological processes are driving
the clustering and guides the annotation of cell types. The
dataset-specific gene loading further helps to understand the
differences between datasets.</p>
<p>In this example, we show the 11’th factor, which is highly loaded in
cluster 7. From the top loaded genes, we can easily identify that
cluster 7 should be considered as NK cells given the context of PBMCs,
as GNLY and NKG7 are well known markers for NK cells. From the top
loaded genes in the “stim” dataset (interferon stimulated), we can also
observe IFI6 (interferon-alpha inducible protein 6), which agrees with
the biological context of the experiment.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">factorMarkers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getFactorMarkers.html">getFactorMarkers</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, dataset1 <span class="op">=</span> <span class="st">"ctrl"</span>, dataset2 <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotGeneLoadings.html">plotGeneLoadings</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, markerTable <span class="op">=</span> <span class="va">factorMarkers</span>, useFactor <span class="op">=</span> <span class="fl">11</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-11-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="explore-cell-type-proportions">Explore cell type proportions<a class="anchor" aria-label="anchor" href="#explore-cell-type-proportions"></a>
</h2>
<p><em>rliger</em> provides a good number of functions that visualize
the cell type proportion within each dataset, which helps understanding
the compositional change across conditions of interests, also known as a
differential abundance analysis.</p>
<div class="section level3">
<h3 id="pie-charts">Pie charts<a class="anchor" aria-label="anchor" href="#pie-charts"></a>
</h3>
<p>Here we show a pie chart layered by dataset and sliced by cluster
proportion, allowing us to directly see the proportion of each cluster
within each dataset while being able to compare the differential
abundance of clusters across datasets.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotProportion.html">plotProportionPie</a></span><span class="op">(</span><span class="va">pbmcLiger</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-12-1.png" width="700" style="display: block; margin: auto;"></p>
<p>From above, we can clearly see a difference of proportion in cluster
0 and 1, between the two datasets. Combined with what we observed
earlier from the UMAP, this indicates that the interferon stimulation
may have an impact on the composition of the PBMCs, especially on the
cell type that cluster 0 represents.</p>
</div>
<div class="section level3">
<h3 id="box-plots">Box plots<a class="anchor" aria-label="anchor" href="#box-plots"></a>
</h3>
<p>In the scenario of a more complicated study design, where different
biological conditions consist of multiple samples, it is more
informative to visualize the proportions grouped by the condition
covariate. The function <code><a href="../reference/plotProportionBox.html">plotProportionBox()</a></code> is designed to
group the proportion of each cluster within each sample by condition so
that we can see the differential abundance of clusters across
conditions.</p>
<p>In the example dataset we show in this tutorial, we only have one
sample for each condition. Therefore, to demonstrate how this function
works, we need to manually divide each dataset into three
pseudo-replicates and regard the original dataset covariate as the
condition variable.</p>
<p><strong>In real-life analysis, users should use real sample variable
at <code>sampleBy</code> and real condition variable at
<code>conditionBy</code></strong>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcLiger</span><span class="op">$</span><span class="va">pseudo_replicate</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span></span>
<span>    fmt <span class="op">=</span> <span class="st">"%s_%s"</span>, </span>
<span>    <span class="va">pbmcLiger</span><span class="op">$</span><span class="va">dataset</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pbmcLiger</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">pbmcLiger</span><span class="op">$</span><span class="va">dataset</span>, <span class="va">pbmcLiger</span><span class="op">$</span><span class="va">pseudo_replicate</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       </span></span>
<span><span class="co">##        ctrl_1 ctrl_2 ctrl_3 stim_1 stim_2 stim_3</span></span>
<span><span class="co">##   ctrl   2247   2086   2215      0      0      0</span></span>
<span><span class="co">##   stim      0      0      0   2504   2391   2556</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotProportionBox.html">plotProportionBox</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, conditionBy <span class="op">=</span> <span class="st">"dataset"</span>, sampleBy <span class="op">=</span> <span class="st">"pseudo_replicate"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;"></p>
<p>For the boxplot, given the fact that there are always some clusters
of small size and the proportions they take can be squeezed in the plot
by the larger clusters. To get over this, we can use option
<code>splitByCluster = TRUE</code> to generate the plot of the same
information for each cluster. The return value is now a list of plots,
and we will bring them together with cowplot’s <code><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid()</a></code>
function.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">propBoxList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotProportionBox.html">plotProportionBox</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmcLiger</span>, </span>
<span>  conditionBy <span class="op">=</span> <span class="st">"dataset"</span>, </span>
<span>  sampleBy <span class="op">=</span> <span class="st">"pseudo_replicate"</span>, </span>
<span>  splitByCluster <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Again, use cowplot's functionality to bring individual plots in the list together</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">propBoxList</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-14-1.png" width="1344" style="display: block; margin: auto;"></p>
<p>Please also checkout other functions like
<code><a href="../reference/plotProportion.html">plotProportionDot()</a></code> and <code><a href="../reference/plotProportion.html">plotProportionBar()</a></code>
that express similar information in different ways.</p>
</div>
</div>
<div class="section level2">
<h2 id="differential-expression">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h2>
<p>Starting from <em>rliger</em> 2.1.0, we move the default differential
expression test method to pseudo-bulk test performed with <a href="https://bioconductor.org/packages/DESeq2/" class="external-link">DESeq2</a>. This is
generally tested to be more robust when comparing across batches than
wilcoxon rank-sum test that we previously used. We provide two
functions, <code><a href="../reference/liger-DEG.html">runMarkerDEG()</a></code> and <code><a href="../reference/liger-DEG.html">runPairwiseDEG()</a></code>
for different purposes. Please click on the function name to be
redirected to detailed usage reference.</p>
<ul>
<li>
<code><a href="../reference/liger-DEG.html">runMarkerDEG()</a></code> always performs tests in a one-vs-rest
manner, which is suitable for marker identification. Cluster markers can
be identified easily, while dataset-specific markers can be identified
within each cluster. See example below.</li>
<li>
<code><a href="../reference/liger-DEG.html">runPairwiseDEG()</a></code> performs test between two specific
groups of cells, either specified by metadata variable or user
customized cell indices with great flexibility. Comparison between two
conditions can be done within each cluster. See example below.</li>
</ul>
<p>Pseudo-bulk method works optimally when real replicate information is
available, i.e. a pseudo-bulk should better be created with cells that
belong to the same biological replicate. We recommend inserting such
metadata information into the object beforehand and supply the variable
name to argument <code>useReplicate</code>. Most of the time, the
variable <code>"dataset"</code> is used for this purpose if users create
the <a href="liger_object.html">liger object</a> with each matrix
representing a single real replicate.</p>
<div class="section level3">
<h3 id="marker-detection-for-each-cluster">Marker detection for each cluster<a class="anchor" aria-label="anchor" href="#marker-detection-for-each-cluster"></a>
</h3>
<p>For identifying cluster markers, we can simply call the function with
setting <code>conditionBy = "leiden_cluster"</code> to allow the test to
be built on the cluster labels we just got.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run the pseudo-bulk DE test for cluster marker detection</span></span>
<span><span class="va">clusterMarkerStats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-DEG.html">runMarkerDEG</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, conditionBy <span class="op">=</span> <span class="st">"leiden_cluster"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="manipulate-marker-tables">Manipulate marker tables<a class="anchor" aria-label="anchor" href="#manipulate-marker-tables"></a>
</h3>
<p>The returned object is a <code>data.frame</code> object containing
all the statistics for all clusters. Split by <code>"group"</code>
column, each part of the table represents the result of DEG test
comparing the cluster that the <code>"group"</code> column indicates
against all other clusters. We recommend using package <em>dplyr</em>
for further data manipulation.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Show the top 10 markers ranked by adjusted p-value and logFC for cluster 7, NK cells</span></span>
<span><span class="va">clusterMarkerStats</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">group</span> <span class="op">==</span> <span class="fl">7</span>, <span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">padj</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      feature group    logFC          pval          padj   pct_in    pct_out</span></span>
<span><span class="co">## 1       NKG7     7 4.902003  0.000000e+00  0.000000e+00 95.66667  8.3886857</span></span>
<span><span class="co">## 2       CCL5     7 3.667108 9.945080e-270 5.605047e-266 71.33333 13.5681767</span></span>
<span><span class="co">## 3       GNLY     7 6.228939 1.655173e-172 2.665301e-169 94.50000  5.4630943</span></span>
<span><span class="co">## 4       CTSW     7 5.067694 2.111548e-165 2.644597e-162 57.66667  3.2838272</span></span>
<span><span class="co">## 5      KLRC1     7 6.572406 4.377260e-147 4.111706e-144 37.66667  0.5224270</span></span>
<span><span class="co">## 6  LINC00996     7 5.718599 1.401039e-139 1.214808e-136 20.33333  0.5448168</span></span>
<span><span class="co">## 7      KLRD1     7 4.836627 1.950272e-125 1.570247e-122 52.83333  2.4180909</span></span>
<span><span class="co">## 8       GZMH     7 3.831351 2.170586e-121 1.631123e-118 38.83333  2.8211061</span></span>
<span><span class="co">## 9       AOAH     7 3.210956 2.392265e-117 1.685351e-114 31.16667  3.9853720</span></span>
<span><span class="co">## 10      MATK     7 5.545011 2.734977e-107 1.712703e-104 23.50000  0.6119860</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="make-marker-heatmaps">Make marker heatmaps<a class="anchor" aria-label="anchor" href="#make-marker-heatmaps"></a>
</h3>
<p>We can visualize cluster markers identified by
<code><a href="../reference/liger-DEG.html">runMarkerDEG()</a></code> with <code><a href="../reference/plotMarkerHeatmap.html">plotMarkerHeatmap()</a></code>. A
checkerboard pattern should be observed most of the time. In the
heatmap, each row is a marker gene and each column is a cell. We don’t
by default plot all the cells as it requires too much computation
resource, but instead subsample a sufficient number of cells to show the
pattern. the column annotations on the top edge indicates the metadata
for cells being shown, including the cluster label and dataset source.
These two are by default added. The row annotation of the left edge
indicates which cluster the gene is a marker of. A gene might appear to
be significant for multiple clusters. In this case, we assign it to the
cluster where it shows the highest logFC value and only show it once in
the heatmap, since we can still observe its expression in the other
clusters.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMarkerHeatmap.html">plotMarkerHeatmap</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, result <span class="op">=</span> <span class="va">clusterMarkerStats</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/4-3-1.png" width="768" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="show-top-markers-from-the-table">Show top markers from the table<a class="anchor" aria-label="anchor" href="#show-top-markers-from-the-table"></a>
</h3>
<p>We can also make use of the user-friendly <em>dplyr</em> interface to
filter and sort for the top markers for each cluster.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markerTop3</span> <span class="op">&lt;-</span> <span class="va">clusterMarkerStats</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Exlude insignificant result and use only up-regulated genes for each cluster</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="co"># Allow the following operations to be done within each cluster</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Sort the result by adjusted p-value ascending and logFC descending, and keep </span></span>
<span>    <span class="co"># the sorting happening within each cluster</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">padj</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span>, .by_group <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Show the top three markers. Since we grouped by cluster, that's for each cluster</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">markerTop3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 39 × 7</span></span></span>
<span><span class="co">## <span style="color: #949494;"># Groups:   group [13]</span></span></span>
<span><span class="co">##    feature group logFC      pval      padj pct_in pct_out</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> ATP6V1F 0     2.68  0   <span style="color: #949494;"> </span>     0   <span style="color: #949494;"> </span>       93.8   40.6 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> CFL1    0     1.18  0   <span style="color: #949494;"> </span>     0   <span style="color: #949494;"> </span>       91.1   67.8 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> CD63    0     3.22  8.75<span style="color: #949494;">e</span><span style="color: #BB0000;">-269</span> 2.79<span style="color: #949494;">e</span><span style="color: #BB0000;">-265</span>   99.3   31.8 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> LDHB    1     1.97  0   <span style="color: #949494;"> </span>     0   <span style="color: #949494;"> </span>       79.7   29.8 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> RPL13   1     1.09  0   <span style="color: #949494;"> </span>     0   <span style="color: #949494;"> </span>      100.    98.8 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> RPS27A  1     0.971 5.41<span style="color: #949494;">e</span><span style="color: #BB0000;">-295</span> 6.77<span style="color: #949494;">e</span><span style="color: #BB0000;">-292</span>   99.9   96.3 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CD2     2     1.85  9.15<span style="color: #949494;">e</span><span style="color: #BB0000;">-192</span> 1.10<span style="color: #949494;">e</span><span style="color: #BB0000;">-188</span>   55.7   19.3 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> ZFP36L2 2     1.84  1.02<span style="color: #949494;">e</span><span style="color: #BB0000;">-190</span> 1.11<span style="color: #949494;">e</span><span style="color: #BB0000;">-187</span>   72.2   30.4 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> RPL13A  2     0.879 1.06<span style="color: #949494;">e</span><span style="color: #BB0000;">-168</span> 7.50<span style="color: #949494;">e</span><span style="color: #BB0000;">-166</span>   99.9   98.1 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD79A   3     6.01  1.12<span style="color: #949494;">e</span><span style="color: #BB0000;">-272</span> 3.35<span style="color: #949494;">e</span><span style="color: #BB0000;">-269</span>   66.1    1.64</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 29 more rows</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="make-dot-plots-for-top-markers">Make dot plots for top markers<a class="anchor" aria-label="anchor" href="#make-dot-plots-for-top-markers"></a>
</h3>
<p><code><a href="../reference/plotClusterGeneDot.html">plotClusterGeneDot()</a></code> function produces publication ready
dot plots for showing marker expression and expression percentage in
each cluster. Users can visualize the marker detection result from above
as well as visualizing customized marker list. Here is how we show the
detected markers for each cluster. Make sure to subset the marker stats
table to only include the top ones of each cluster, because the function
itself does not do the selection.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotClusterGeneDot.html">plotClusterGeneDot</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, features <span class="op">=</span> <span class="va">markerTop3</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-17-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="make-dot-plots-from-custom-marker-list">Make dot plots from custom marker list<a class="anchor" aria-label="anchor" href="#make-dot-plots-from-custom-marker-list"></a>
</h3>
<p>The following code demonstrates how to make a dot plot with a custom
marker list. We suggest starting from manually writing down the
information into an R list object, for the sake of better readability.
Then we convert the list to a <code>data.frame</code> so that
<code><a href="../reference/plotClusterGeneDot.html">plotClusterGeneDot()</a></code> can interpret it. When we call the
plotting function, we had two additional arguments this time:</p>
<ul>
<li>
<code>transpose</code> rotates the plot so that we have wider room
for the cell type label (set in the marker list) to be shown</li>
<li>
<code>featureTitleSize</code> sets the font size of the cell type
label.</li>
</ul>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Write down the marker list</span></span>
<span><span class="va">pbmcMarkerList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  `CD14 Mono` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD14"</span>, <span class="st">"LYZ"</span><span class="op">)</span>,</span>
<span>  `T naive` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SELL"</span>, <span class="st">"LEF1"</span><span class="op">)</span>,</span>
<span>  `T memory` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IL7R"</span>, <span class="st">"CCR7"</span><span class="op">)</span>,</span>
<span>  `B` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"CD79A"</span><span class="op">)</span>,</span>
<span>  `CD16 Mono` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FCGR3A"</span>, <span class="st">"MS4A7"</span><span class="op">)</span>,</span>
<span>  `T activated` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD69"</span>, <span class="st">"IL2RA"</span><span class="op">)</span>,</span>
<span>  `CD8_T` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD8A"</span>, <span class="st">"CD8B"</span><span class="op">)</span>,</span>
<span>  `NK` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GNLY"</span>, <span class="st">"NKG7"</span><span class="op">)</span>,</span>
<span>  `DC` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FCER1A"</span>, <span class="st">"CST3"</span><span class="op">)</span>,</span>
<span>  `Mk` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PPBP"</span>, <span class="st">"ITGA2B"</span><span class="op">)</span>,</span>
<span>  `pDC` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CLEC4C"</span>, <span class="st">"IL3RA"</span><span class="op">)</span>,</span>
<span>  `Eryth` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"HBB"</span>, <span class="st">"HBA1"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Convert to acceptable data.frame format</span></span>
<span><span class="va">pbmcMarkerDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">pbmcMarkerList</span>, use.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>  cell_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pbmcMarkerList</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">pbmcMarkerList</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                     levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pbmcMarkerList</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Make the plot</span></span>
<span><span class="fu"><a href="../reference/plotClusterGeneDot.html">plotClusterGeneDot</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, features <span class="op">=</span> <span class="va">pbmcMarkerDF</span>, transpose <span class="op">=</span> <span class="cn">TRUE</span>, featureTitleSize <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-18-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="section level3 tabset">
<h3 class="tabset" id="visualize-individual-gene-expression">Visualize individual gene expression<a class="anchor" aria-label="anchor" href="#visualize-individual-gene-expression"></a>
</h3>
<p>We can also visualize the expression profiles of individual genes,
such as the markers that we just identified. This allows us to visually
confirm the cluster- or dataset-specific expression patterns of marker
genes.</p>



<ul class="nav nav-tabs" id="visualize-individual-gene-expression" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#single-umap-with-all-cells" id="single-umap-with-all-cells-tab" type="button" role="tab" aria-controls="single-umap-with-all-cells" aria-selected="true" class="active nav-link">Single UMAP with all cells</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#umap-view-split-by-dataset" id="umap-view-split-by-dataset-tab" type="button" role="tab" aria-controls="umap-view-split-by-dataset" aria-selected="false" class="nav-link">UMAP view split by dataset</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#violin-plot" id="violin-plot-tab" type="button" role="tab" aria-controls="violin-plot" aria-selected="false" class="nav-link">Violin plot</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="single-umap-with-all-cells" role="tabpanel" aria-labelledby="single-umap-with-all-cells-tab">

<p><code><a href="../reference/plotDimRed.html">plotGeneDimRed()</a></code> by default returns a plot of the
dimensionality reduction of all cells, colored by specified genes.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimRed.html">plotGeneDimRed</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, <span class="st">"GNLY"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-19-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="tab-pane" id="umap-view-split-by-dataset" role="tabpanel" aria-labelledby="umap-view-split-by-dataset-tab">

<p>Here, we further split the plot by dataset to check if the marker
expression is consistent across datasets.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># By splitting the plot by dataset, we get a list of two plots</span></span>
<span><span class="va">nkg7PlotList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDimRed.html">plotGeneDimRed</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, <span class="st">"NKG7"</span>, splitBy <span class="op">=</span> <span class="st">"dataset"</span>, title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pbmcLiger</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Then combine them into one with cowplot function</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">nkg7PlotList</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-20-1.png" width="1344" style="display: block; margin: auto;"></p>
</div>
<div class="tab-pane" id="violin-plot" role="tabpanel" aria-labelledby="violin-plot-tab">

<p><code><a href="../reference/plotViolin.html">plotGeneViolin()</a></code> provides a simple interface for
creating violin plots of gene expression across clusters. Users can
optionally add box plots to the plot to see the quantiles and outliers.
This function by default returns a list of plots, one for each
dataset.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GNLYviolinList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotViolin.html">plotGeneViolin</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, <span class="st">"GNLY"</span>, box <span class="op">=</span> <span class="cn">TRUE</span>, colorBy <span class="op">=</span> <span class="st">"leiden_cluster"</span>, titles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pbmcLiger</span><span class="op">)</span>, legendNCol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">GNLYviolinList</span>, ncol <span class="op">=</span> <span class="fl">1</span>, align <span class="op">=</span> <span class="st">"hv"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-21-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="annotate-clusters-by-cell-type-names">Annotate clusters by cell type names<a class="anchor" aria-label="anchor" href="#annotate-clusters-by-cell-type-names"></a>
</h3>
<p>We provide the function <code><a href="../reference/mapCellMeta.html">mapCellMeta()</a></code> to assign names to
each cluster. We name the functions roughly as shown below, given the
evidence we visualized above. In real life analysis, researchers might
have been through much more work to identify the cell types and finer
sub-types.</p>
<p>Users need to specify <code>from</code> to state where to find the
cluster labels that need to be renamed, and <code>newTo</code> to create
a new column where the new assigned cell types will be stored. Following
these, use a cluster name as an argument and the corresponding cell type
name as the value. If a cluster is not mentioned in the call, it will be
left as it is and stored in <code>newTo</code>.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcLiger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mapCellMeta.html">mapCellMeta</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">pbmcLiger</span>,</span>
<span>    from <span class="op">=</span> <span class="st">"leiden_cluster"</span>, </span>
<span>    newTo <span class="op">=</span> <span class="st">"cell_type"</span>, </span>
<span>    `0` <span class="op">=</span> <span class="st">"CD14 Mono"</span>,</span>
<span>    `1` <span class="op">=</span> <span class="st">"CD4 Naive T"</span>,</span>
<span>    `2` <span class="op">=</span> <span class="st">"CD4 Memory T"</span>,</span>
<span>    `3` <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>    `4` <span class="op">=</span> <span class="st">"CD16 Mono"</span>,</span>
<span>    `5` <span class="op">=</span> <span class="st">"T activated"</span>,</span>
<span>    `6` <span class="op">=</span> <span class="st">"CD8 T"</span>,</span>
<span>    `7` <span class="op">=</span> <span class="st">"NK"</span>,</span>
<span>    `8` <span class="op">=</span> <span class="st">"DC"</span>,</span>
<span>    `9` <span class="op">=</span> <span class="st">"Mk"</span>,</span>
<span>    `10` <span class="op">=</span> <span class="st">"pDC"</span>,</span>
<span>    `11` <span class="op">=</span> <span class="st">"CD14 Mono"</span>,</span>
<span>    `12` <span class="op">=</span> <span class="st">"Eryth"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotDimRed.html">plotClusterDimRed</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, <span class="st">"cell_type"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="dataset-marker-detection-split-by-cluster">Dataset marker detection split by cluster<a class="anchor" aria-label="anchor" href="#dataset-marker-detection-split-by-cluster"></a>
</h3>
<p>Additionally, by specifying <code>conditionBy = "dataset"</code> or
other metadata variable, the function will do one-vs-rest marker test
basing on the specified variable. So it works for dataset-specific
marker detection as well. By adding another argument
<code>splitBy</code>, the tes can even be done within each level of
another variable. For example,
<code>runMarkerDEG(pbmcLiger, conditionBy = "dataset", splitBy = "leiden_cluster")</code>
will perform the test within each cluster and find the marker for each
dataset. We skip this demonstration because the example data only
contains two datasets which is more suitable for pairwise
comparison.</p>
</div>
<div class="section level3">
<h3 id="pairwise-comparison-between-two-specific-conditions">Pairwise comparison between two specific conditions<a class="anchor" aria-label="anchor" href="#pairwise-comparison-between-two-specific-conditions"></a>
</h3>
<p>In a more complexed study design than what this tutorial has, we may
have more combinations of condition, such as gender, treatment, age and
etc. Comparing the appropriate conditions helps to understand the
biological changes happening in the experiment. We introduce the
<code><a href="../reference/liger-DEG.html">runPairwiseDEG()</a></code> function for this purpose. In order to see
what changes are happening in the “stim” dataset compared to the “ctrl”
dataset, we set <code>groupTest = "stim"</code> and
<code>groupCtrl = "ctrl"</code>. We need to also set
<code>variable1 = "dataset"</code> to allow the the group setting to be
found. We further set <code>splitBy = "cell_type"</code> to allow the
test to be done within each cluster, in order to under stand if the
condition effect happens within any certain cell types.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stimVsCtrlStats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-DEG.html">runPairwiseDEG</a></span><span class="op">(</span></span>
<span>  <span class="va">pbmcLiger</span>, </span>
<span>  groupTest <span class="op">=</span> <span class="st">"stim"</span>, </span>
<span>  groupCtrl <span class="op">=</span> <span class="st">"ctrl"</span>, </span>
<span>  variable1 <span class="op">=</span> <span class="st">"dataset"</span>,</span>
<span>  splitBy <span class="op">=</span> <span class="st">"cell_type"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Similar to what <code><a href="../reference/liger-DEG.html">runMarkerDEG()</a></code> returns, the result table
is a combination of results. The column <code>"group"</code> splits the
result by the cell types, as indicated by
<code>splitBy = "cell_type"</code>. For the slice for each cell type in
the result, the statistics are calculated by comparing
<code>groupTest</code> against <code>groupCtrl</code>, which is “stim”
against “ctrl” in this case. For downstream visualization and GO term
enrichment analysis, we can use the result table directly. The
<code>group</code> option from those functions requires an available
group label from this test, which can be chosen from:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">groupAvail</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">stimVsCtrlStats</span><span class="op">$</span><span class="va">group</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">groupAvail</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "CD14 Mono.stim"    "CD4 Naive T.stim"  "CD4 Memory T.stim"</span></span>
<span><span class="co">##  [4] "B.stim"            "CD16 Mono.stim"    "T activated.stim" </span></span>
<span><span class="co">##  [7] "CD8 T.stim"        "NK.stim"           "DC.stim"          </span></span>
<span><span class="co">## [10] "Mk.stim"           "pDC.stim"          "Eryth.stim"</span></span></code></pre>
<p>Recall the compositional difference we identified in cluster 0, which
is later labeled as CD14 Monocytes, here we visualize the DEG result
tested within this cluster to see what genes are differentially
expressed between the two datasets.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>ggrepel.max.overlaps <span class="op">=</span> <span class="fl">40</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotVolcano.html">plotVolcano</a></span><span class="op">(</span><span class="va">stimVsCtrlStats</span>, group <span class="op">=</span> <span class="st">"CD14 Mono.stim"</span>, labelTopN <span class="op">=</span> <span class="fl">80</span>, labelTextSize <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-25-1.png" width="768" style="display: block; margin: auto;"></p>
<p>By specifying <code>group = "CD14 Mono.stim"</code>, the function
makes a volcano plot for that test that happened within the CD14
Monocytes, and compares the “stim” dataset to the “ctrl” dataset. From
the upper-right part of the volcano plot, we can easily tell that ISGs
(Interferon-Stimulated Genes) and IFIs (Interferon-induced proteins) are
highly expressed in the CD14 monocytes from “stim” dataset compared to
those from “ctrl”, which obviously agrees with the biological context of
the experiment.</p>
<p>Similar to what we did in marker detection visualization, we also
provide a heatmap function for pairwise DE testt. It shows the gene
expression in downsampled cells involved in a specified test and divide
the heatmap vertically by comparison groups and horizontally by
up-/down-regulation.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotPairwiseDEGHeatmap.html">plotPairwiseDEGHeatmap</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, result <span class="op">=</span> <span class="va">stimVsCtrlStats</span>, group <span class="op">=</span> <span class="st">"CD14 Mono.stim"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-26-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="go-term-enrichment-analysis">GO term enrichment analysis<a class="anchor" aria-label="anchor" href="#go-term-enrichment-analysis"></a>
</h2>
<p>After getting the differentially expressed genes, we can submit them
to tests that identify enriched GO terms. Function
<code><a href="../reference/runGOEnrich.html">runGOEnrich()</a></code> wraps the functionality of <a href="https://CRAN.R-project.org/package=gprofiler2" class="external-link">gprofiler2</a>
package and can directly be called with the returned
<code>data.frame</code>s from the DEG functions of <em>rliger</em>. Note
the active internet connection is required for the function to submit
the query to <em>gprofiler</em>’s server.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">goResult</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runGOEnrich.html">runGOEnrich</a></span><span class="op">(</span><span class="va">stimVsCtrlStats</span>, group <span class="op">=</span> <span class="st">"CD14 Mono.stim"</span>, splitReg <span class="op">=</span> <span class="cn">TRUE</span>, organism <span class="op">=</span> <span class="st">"hsapiens"</span>, evcodes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>
<code>splitReg = TRUE</code> requires sending separate queries for
both up- and down-regulated genes. This is helpful when working on
pairwise comparisons, while specifying <code>splitReg = FALSE</code> for
marker detection makes more sense.</li>
<li>
<code>organism</code> specifies the species for the GO term
database. <code>"hsapiens"</code> for human and <code>"mmusculus"</code>
for mouse.</li>
<li>
<code>evcodes = TRUE</code> is optional. It requests the overlapping
genes between the query gene set and each term. This is by default not
turned on because it slows down the analysis. If set, an additional
column <code>"intersection"</code> will be added to the result
table.</li>
</ul>
<p>If the input DE analysis result consists of multiple tests
(e.g. split by cluster, as we did above, or marker detection), the
returned result would be a list for each test group. Each element would
be an individual GO enrichment analysis result, represented as a list of
one <code>data.frame</code> for the statistics
(<code>goResult$group_name$result</code>) and one list for meta
information used by <em>gprofiler2</em>
(<code>goResult$group_name$meta</code>). Here we simply use
<em>dplyr</em> expression to show the enriched GO terms from the
up-regulated genes in the comparison of “stim” against “ctrl”, within
CD14 monocytes.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">goResult</span><span class="op">[[</span><span class="st">"CD14 Mono.stim"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">result</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">query</span> <span class="op">==</span> <span class="st">"Up"</span>, <span class="fu"><a href="https://rdrr.io/r/base/startsWith.html" class="external-link">startsWith</a></span><span class="op">(</span><span class="va">term_id</span>, <span class="st">"GO"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">p_value</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">query</span>, <span class="va">term_name</span>, <span class="va">term_id</span>, <span class="va">p_value</span>, <span class="va">intersection</span>, <span class="va">term_size</span>, <span class="va">intersection_size</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142"],["Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up","Up"],["defense response to virus","response to virus","defense response to symbiont","innate immune response","biological process involved in interspecies interaction between organisms","response to other organism","response to external biotic stimulus","response to biotic stimulus","defense response to other organism","defense response","immune response","regulation of response to biotic stimulus","negative regulation of viral process","regulation of viral process","viral process","negative regulation of viral genome replication","viral life cycle","viral genome replication","response to cytokine","regulation of defense response","regulation of viral life cycle","regulation of viral genome replication","immune system process","regulation of innate immune response","regulation of immune response","response to external stimulus","regulation of response to external stimulus","response to type II interferon","cellular response to cytokine stimulus","regulation of cytokine production","cytokine production","positive regulation of defense response","negative regulation of response to biotic stimulus","positive regulation of response to external stimulus","cytokine-mediated signaling pathway","positive regulation of immune response","response to stress","positive regulation of innate immune response","positive regulation of response to biotic stimulus","positive regulation of cytokine production","positive regulation of immune system process","activation of innate immune response","negative regulation of innate immune response","response to interferon-beta","positive regulation of interleukin-1 beta production","regulation of immune system process","negative regulation of immune response","response to bacterium","response to type I interferon","activation of immune response","pattern recognition receptor signaling pathway","interleukin-1 beta production","regulation of interleukin-1 beta production","regulation of response to stress","cellular response to type II interferon","positive regulation of interleukin-1 production","cellular response to type I interferon","TAP binding","regulation of interleukin-1 production","interleukin-1 production","pyroptosis","immune response-activating signaling pathway","innate immune response-activating signaling pathway","immune response-regulating signaling pathway","interferon-mediated signaling pathway","type I interferon-mediated signaling pathway","response to lipopolysaccharide","response to molecule of bacterial origin","type I interferon production","regulation of type I interferon production","cytoplasmic pattern recognition receptor signaling pathway","response to interferon-alpha","cellular response to lipopolysaccharide","response to organic substance","regulation of defense response to virus","inflammatory response","cellular response to organic substance","interleukin-27-mediated signaling pathway","positive regulation of canonical NF-kappaB signal transduction","cellular response to molecule of bacterial origin","regulation of canonical NF-kappaB signal transduction","positive regulation of pattern recognition receptor signaling pathway","negative regulation of defense response","positive regulation of defense response to virus by host","negative regulation of immune system process","biological process involved in interaction with host","symbiont entry into host cell","cellular response to biotic stimulus","antiviral innate immune response","regulation of interleukin-12 production","interleukin-12 production","biological process involved in symbiotic interaction","cytolysis in another organism","regulation of defense response to virus by host","symbiont entry into host","MHC class I protein complex","interferon-beta production","regulation of interferon-beta production","antigen processing and presentation","response to tumor necrosis factor","positive regulation of interferon-beta production","positive regulation of type I interferon production","response to interleukin-1","canonical NF-kappaB signal transduction","cellular response to interferon-beta","ficolin-1-rich granule lumen","positive regulation of inflammatory response","positive regulation of pyroptosis","positive regulation of nucleotide-binding oligomerization domain containing 2 signaling pathway","regulation of adaptive immune response","tolerance induction","phagocytic vesicle","regulation of immune effector process","positive regulation of response to stimulus","antigen processing and presentation of endogenous peptide antigen via MHC class I","death-inducing signaling complex","antigen processing and presentation of endogenous peptide antigen","cytolysis","regulation of viral entry into host cell","negative regulation of viral life cycle","cellular response to tumor necrosis factor","negative regulation of natural killer cell mediated cytotoxicity","regulation of lymphocyte proliferation","defense response to bacterium","lumenal side of endoplasmic reticulum membrane","negative regulation of natural killer cell mediated immunity","positive regulation of lymphocyte proliferation","regulation of response to cytokine stimulus","regulation of mononuclear cell proliferation","positive regulation of interleukin-12 production","antigen processing and presentation of peptide antigen","positive regulation of tumor necrosis factor superfamily cytokine production","cell surface pattern recognition receptor signaling pathway","antigen processing and presentation of endogenous peptide antigen via MHC class I via ER pathway, TAP-independent","regulation of pattern recognition receptor signaling pathway","positive regulation of leukocyte proliferation","positive regulation of cell activation","antigen processing and presentation of peptide antigen via MHC class I","2'-5'-oligoadenylate synthetase activity","chemokine-mediated signaling pathway","positive regulation of nucleotide-binding domain, leucine rich repeat containing receptor signaling pathway","TAP1 binding"],["GO:0051607","GO:0009615","GO:0140546","GO:0045087","GO:0044419","GO:0051707","GO:0043207","GO:0009607","GO:0098542","GO:0006952","GO:0006955","GO:0002831","GO:0048525","GO:0050792","GO:0016032","GO:0045071","GO:0019058","GO:0019079","GO:0034097","GO:0031347","GO:1903900","GO:0045069","GO:0002376","GO:0045088","GO:0050776","GO:0009605","GO:0032101","GO:0034341","GO:0071345","GO:0001817","GO:0001816","GO:0031349","GO:0002832","GO:0032103","GO:0019221","GO:0050778","GO:0006950","GO:0045089","GO:0002833","GO:0001819","GO:0002684","GO:0002218","GO:0045824","GO:0035456","GO:0032731","GO:0002682","GO:0050777","GO:0009617","GO:0034340","GO:0002253","GO:0002221","GO:0032611","GO:0032651","GO:0080134","GO:0071346","GO:0032732","GO:0071357","GO:0046977","GO:0032652","GO:0032612","GO:0070269","GO:0002757","GO:0002758","GO:0002764","GO:0140888","GO:0060337","GO:0032496","GO:0002237","GO:0032606","GO:0032479","GO:0002753","GO:0035455","GO:0071222","GO:0010033","GO:0050688","GO:0006954","GO:0071310","GO:0070106","GO:0043123","GO:0071219","GO:0043122","GO:0062208","GO:0031348","GO:0002230","GO:0002683","GO:0051701","GO:0046718","GO:0071216","GO:0140374","GO:0032655","GO:0032615","GO:0044403","GO:0051715","GO:0050691","GO:0044409","GO:0042612","GO:0032608","GO:0032648","GO:0019882","GO:0034612","GO:0032728","GO:0032481","GO:0070555","GO:0007249","GO:0035458","GO:1904813","GO:0050729","GO:0140639","GO:0070434","GO:0002819","GO:0002507","GO:0045335","GO:0002697","GO:0048584","GO:0019885","GO:0031264","GO:0002483","GO:0019835","GO:0046596","GO:1903901","GO:0071356","GO:0045953","GO:0050670","GO:0042742","GO:0098553","GO:0002716","GO:0050671","GO:0060759","GO:0032944","GO:0032735","GO:0048002","GO:1903557","GO:0002752","GO:0002486","GO:0062207","GO:0070665","GO:0050867","GO:0002474","GO:0001730","GO:0070098","GO:0070426","GO:0046978"],[2.37e-31,1.98e-29,8.19e-29,9.39e-29,1.19e-28,2.02e-28,2.7e-28,4.28e-28,1.62e-27,1.92e-26,1.49e-20,3.97e-20,7.82e-20,5.84e-19,9.1e-19,1.21e-18,3.04e-17,1.95e-16,3.11e-16,4.66e-16,7.73e-16,1.02e-15,5.39e-15,1.37e-14,3.91e-14,2.62e-13,3.94e-13,3.67e-12,7.32e-12,3.63e-11,5.77e-11,9.579999999999999e-11,2.69e-10,2.73e-10,3.86e-10,8.13e-10,1.16e-09,2.05e-09,2.34e-09,2.34e-09,6.01e-09,2.37e-08,2.54e-08,3.62e-08,1.55e-07,1.88e-07,2.33e-07,3.39e-07,4.01e-07,4.3e-07,5.630000000000001e-07,6.41e-07,6.41e-07,6.58e-07,6.82e-07,1.22e-06,1.23e-06,1.39e-06,1.89e-06,1.89e-06,2.34e-06,3.59e-06,5.16e-06,8.29e-06,8.600000000000001e-06,1.04e-05,1.32e-05,1.37e-05,3.62e-05,3.62e-05,6.05e-05,7.9e-05,8.279999999999999e-05,0.000123,0.000143,0.000193,0.000219,0.000225,0.000315,0.000346,0.000385,0.000564,0.000569,0.000629,0.000695,0.000945,0.00104,0.00111,0.00115,0.00119,0.00119,0.00133,0.00187,0.00195,0.00221,0.00292,0.00302,0.00302,0.00322,0.00336,0.00364,0.00418,0.00433,0.00441,0.00452,0.00484,0.00646,0.00706,0.00716,0.00728,0.008059999999999999,0.008059999999999999,0.00825,0.00842,0.009310000000000001,0.0134,0.0159,0.0168,0.0183,0.0192,0.0203,0.0206,0.0231,0.0233,0.029,0.0329,0.0337,0.0337,0.035,0.0369,0.0375,0.0402,0.0407,0.0408,0.0413,0.0413,0.0419,0.0428,0.0447,0.0459,0.047,0.0473],["ISG15,IFI6,GBP1,PLSCR1,IFIT2,IFIT3,IFIT1,IFITM2,IFITM3,IRF7,OAS1,OASL,ISG20,BST2,APOBEC3A,MX1,MX2,IFI44L,NT5C3A,HSP90AA1,IFIH1,APOBEC3B,HERC5,EIF2AK2,IFI16,FGL2,OAS2,GBP5,OAS3,IRF2,CXCL10,RSAD2,STAT1,RTP4,TRIM22,APOBEC3G,PARP9,CD40,IL27,CASP1,GBP3,IFI27,GBP2,DHX58,AIM2,ADAR,STAT2,DTX3L,PML,IFIT5,DDX60L,IFITM1,ZNFX1,GBP7,UNC93B1,TANK,TRIM5,TRIM38,SAMHD1,IL15,NT5C2,ZC3HAV1,IRF1,DDX60,MOV10,MLKL,RNASE2,APOBEC3F,CXCL9,TRIM21,MYD88","ISG15,IFI6,GBP1,PLSCR1,HSPB1,IFIT2,IFIT3,IFIT1,IFITM2,IFITM3,IRF7,OAS1,OASL,ISG20,CCL8,BST2,APOBEC3A,MX1,MX2,IFI44L,NT5C3A,HSP90AA1,NMI,IFIH1,APOBEC3B,HERC5,EIF2AK2,IFI44,IFI16,FGL2,LGALS9,OAS2,GBP5,OAS3,IRF2,CXCL10,RSAD2,STAT1,RTP4,TRIM22,APOBEC3G,PARP9,CD40,IL27,CASP1,GBP3,IFI27,CCL4,GBP2,DHX58,AIM2,ADAR,STAT2,DTX3L,PML,IFIT5,DDX60L,IFITM1,ZNFX1,GBP7,UNC93B1,TANK,TRIM5,TRIM38,SAMHD1,IL15,NT5C2,ZC3HAV1,IRF1,DDX60,JAK2,MOV10,MLKL,RNASE2,APOBEC3F,CXCL9,TRIM21,MYD88","ISG15,IFI6,GBP1,PLSCR1,CXCL11,HSPA1A,SAMD9,IFIT2,IFIT3,IFIT1,IFITM2,IFITM3,IRF7,UBE2L6,OAS1,OASL,ISG20,CCL2,CCL8,IFI35,BST2,APOBEC3A,MX1,MNDA,MX2,HLA-B,GBP4,HLA-E,IFI44L,SLAMF7,HLA-C,NT5C3A,HSP90AA1,NMI,HLA-A,IFIH1,APOBEC3B,HERC5,EIF2AK2,IFI16,NAGK,CD47,FGL2,LGALS9,NINJ1,CTSS,ANXA1,OAS2,DEFB1,GBP5,CASP4,FCN1,OPTN,OAS3,GCH1,IRF2,PARP14,CXCL10,LYN,SERPING1,RSAD2,STAT1,RTP4,TRIM22,NCF1,APOBEC3G,PARP9,CD40,IL27,HLA-DPA1,NUB1,CCL7,CASP1,GBP3,SP100,IFI27,CCL4,GBP2,DHX58,HAVCR2,AIM2,GSDMD,ADAR,CCL3,STAT2,DTX3L,PML,IFIT5,DDX60L,IFITM1,ZNFX1,RIPK2,LILRA5,SERPINB9,GBP7,UNC93B1,TRAFD1,TANK,CFD,IRF8,TRIM5,TRIM38","ISG15,IFI6,GBP1,PLSCR1,HSPA1A,SAMD9,IFIT2,IFIT3,IFIT1,IFITM2,IFITM3,IRF7,UBE2L6,OAS1,OASL,ISG20,CCL2,CCL8,IFI35,BST2,APOBEC3A,MX1,MNDA,MX2,HLA-B,GBP4,HLA-E,SLAMF7,HLA-C,HSP90AA1,NMI,HLA-A,IFIH1,APOBEC3B,HERC5,EIF2AK2,IFI16,NAGK,CD47,LGALS9,NINJ1,CTSS,ANXA1,OAS2,DEFB1,GBP5,CASP4,FCN1,OPTN,OAS3,GCH1,PARP14,CXCL10,LYN,SERPING1,RSAD2,STAT1,TRIM22,NCF1,APOBEC3G,PARP9,CD40,IL27,HLA-DPA1,NUB1,CCL7,CASP1,GBP3,SP100,IFI27,CCL4,GBP2,DHX58,HAVCR2,AIM2,GSDMD,ADAR,CCL3,STAT2,DTX3L,PML,IFIT5,IFITM1,ZNFX1,RIPK2,LILRA5,SERPINB9,GBP7,UNC93B1,TRAFD1,TANK,CFD,IRF8,TRIM5,TRIM38","ISG15,IFI6,GBP1,PLSCR1,CXCL11,HSPA1A,DYNLT1,HSPB1,SAMD9,IDO1,LY6E,IFIT2,IFIT3,IFIT1,IFITM2,IFITM3,IRF7,UBE2L6,OAS1,OASL,ISG20,CCL2,CCL8,IFI35,BST2,APOBEC3A,MX1,MNDA,MX2,HLA-B,SOD2,GBP4,HLA-E,IFI44L,SLAMF7,HLA-C,NT5C3A,HSP90AA1,NMI,HLA-A,IFIH1,APOBEC3B,CTSL,HERC5,EIF2AK2,IFI44,IFI16,NAGK,CD47,FGL2,LGALS9,NINJ1,CTSS,ANXA1,CMPK2,OAS2,DEFB1,CTSB,GBP5,CASP4,FCN1,OPTN,OAS3,RNF213,GCH1,RBCK1,IRF2,SCARB2,PARP14,HSP90AB1,CXCL10,LYN,SERPING1,RSAD2,STAT1,RTP4,TRIM22,NCF1,APOBEC3G,PARP9,CD40,SNX3,IL27,CARD16,STOM,PLAC8,HLA-DPA1,NUB1,CCL7,CASP1,CCR5,GBP3,SP100,CHMP5,IFI27,CCL4,GBP2,DHX58,HAVCR2","ISG15,IFI6,GBP1,PLSCR1,CXCL11,HSPA1A,HSPB1,SAMD9,IDO1,IFIT2,IFIT3,IFIT1,IFITM2,IFITM3,IRF7,UBE2L6,OAS1,OASL,ISG20,CCL2,CCL8,IFI35,BST2,APOBEC3A,MX1,MNDA,MX2,HLA-B,SOD2,GBP4,HLA-E,IFI44L,SLAMF7,HLA-C,NT5C3A,HSP90AA1,NMI,HLA-A,IFIH1,APOBEC3B,HERC5,EIF2AK2,IFI44,IFI16,NAGK,CD47,FGL2,LGALS9,NINJ1,CTSS,ANXA1,CMPK2,OAS2,DEFB1,GBP5,CASP4,FCN1,OPTN,OAS3,RNF213,GCH1,RBCK1,IRF2,PARP14,CXCL10,LYN,SERPING1,RSAD2,STAT1,RTP4,TRIM22,NCF1,APOBEC3G,PARP9,CD40,SNX3,IL27,CARD16,PLAC8,HLA-DPA1,NUB1,CCL7,CASP1,CCR5,GBP3,SP100,CHMP5,IFI27,CCL4,GBP2,DHX58,HAVCR2,AIM2,GSDMD,ADAR,CCL3,STAT2,DTX3L,ZFP36,PML,IFIT5,DDX60L,IFITM1,HERC6,ZNFX1,RIPK2,LILRA5,SERPINB9,GBP7,UNC93B1,TRAFD1,CD86,TANK,CNP,CFD,IRF8,IL10RA,TRIM5,TRIM38,CD274,SAMHD1,PTGER2,TAP2,CLEC4E,IL15,NT5C2,CD80,USP15,LILRB2,TRIM14,ARL8B,HLA-F,ZC3HAV1,CAPN2,WIPF1,IRF1,DDX60,JAK2,HSPA1B,BIRC3,MOV10,CCR7,MLKL,FFAR2,SCIMP,RNF19B,RNASET2,RNASE2,APOBEC3F,CXCL9,TRIM21,MYD88,CASP8,WDFY1,APOL1,IKBKE,KMO,PIK3AP1,CASP7,AXL,CLEC4D,PMAIP1,CASP3,N4BP1,TMEM106A,ZBP1,CCL5,RAB20,MSRB1,CALCOCO2,TLR4,APPL1,BATF2,MICB,CDC73,TRIM25,FKBP5,CDC42EP2,PDCD1LG2,TGFB1,LAG3,MEFV,IL7R,PELI1,MCOLN2,EDN1,CCL3L1,TRIM26,PRF1,FCGR1A,SPN,P2RX7,EXT1,BAK1,APOBEC3C,UBA7,SLC15A4,LY86,ACTA2,GJB2,PDE4B,PTGER4,NCK1,ENDOD1,IL6,CD1D,CLEC7A,CCL13,CRTAM,NLRC5,IRF9","ISG15,IFI6,GBP1,PLSCR1,CXCL11,HSPA1A,HSPB1,SAMD9,IDO1,IFIT2,IFIT3,IFIT1,IFITM2,IFITM3,IRF7,UBE2L6,OAS1,OASL,ISG20,CCL2,CCL8,IFI35,BST2,APOBEC3A,MX1,MNDA,MX2,HLA-B,SOD2,GBP4,HLA-E,IFI44L,SLAMF7,HLA-C,NT5C3A,HSP90AA1,NMI,HLA-A,IFIH1,APOBEC3B,HERC5,EIF2AK2,IFI44,IFI16,NAGK,CD47,FGL2,LGALS9,NINJ1,CTSS,ANXA1,CMPK2,OAS2,DEFB1,GBP5,CASP4,FCN1,OPTN,OAS3,RNF213,GCH1,RBCK1,IRF2,PARP14,CXCL10,LYN,SERPING1,RSAD2,STAT1,RTP4,TRIM22,NCF1,APOBEC3G,PARP9,CD40,SNX3,IL27,CARD16,PLAC8,HLA-DPA1,NUB1,CCL7,CASP1,CCR5,GBP3,SP100,CHMP5,IFI27,CCL4,GBP2,DHX58,HAVCR2,AIM2,GSDMD,ADAR,CCL3,STAT2,DTX3L,ZFP36,PML,IFIT5,DDX60L,IFITM1,HERC6,ZNFX1,RIPK2,LILRA5,SERPINB9,GBP7,UNC93B1,TRAFD1,CD86,TANK,CNP,CFD,IRF8,IL10RA,TRIM5,TRIM38,CD274,SAMHD1,PTGER2,TAP2,CLEC4E,IL15,NT5C2,CD80,USP15,LILRB2,TRIM14,ARL8B,HLA-F,ZC3HAV1,CAPN2,WIPF1,IRF1,DDX60,JAK2,HSPA1B,BIRC3,MOV10,CCR7,MLKL,FFAR2,SCIMP,RNF19B,RNASET2,RNASE2,APOBEC3F,CXCL9,TRIM21,MYD88,CASP8,WDFY1,APOL1,IKBKE,KMO,PIK3AP1,CASP7,AXL,CLEC4D,PMAIP1,CASP3,N4BP1,TMEM106A,ZBP1,CCL5,RAB20,MSRB1,CALCOCO2,TLR4,APPL1,BATF2,MICB,CDC73,TRIM25,FKBP5,CDC42EP2,PDCD1LG2,TGFB1,LAG3,MEFV,IL7R,PELI1,MCOLN2,EDN1,CCL3L1,TRIM26,PRF1,FCGR1A,SPN,P2RX7,EXT1,BAK1,APOBEC3C,UBA7,SLC15A4,LY86,ACTA2,GJB2,PDE4B,PTGER4,NCK1,ENDOD1,IL6,CD1D,CLEC7A,CCL13,CRTAM,NLRC5,IRF9","ISG15,IFI6,GBP1,PLSCR1,CXCL11,HSPA1A,HSPB1,SAMD9,IDO1,IFIT2,IFIT3,IFIT1,IFITM2,IFITM3,IRF7,UBE2L6,OAS1,OASL,ISG20,CCL2,CCL8,IFI35,BST2,IL4I1,APOBEC3A,MX1,MNDA,MX2,HLA-B,SOD2,GBP4,HLA-E,IFI44L,SLAMF7,HLA-C,NT5C3A,HSP90AA1,NMI,HLA-A,IFIH1,APOBEC3B,HERC5,EIF2AK2,IFI44,IFI16,NAGK,CD47,FGL2,LGALS9,NINJ1,CTSS,ANXA1,CMPK2,OAS2,DEFB1,GBP5,CASP4,FCN1,OPTN,OAS3,RNF213,GCH1,RBCK1,IRF2,PARP14,CXCL10,LYN,SERPING1,RSAD2,STAT1,RTP4,TRIM22,NCF1,APOBEC3G,PARP9,CD40,SNX3,IL27,CARD16,PLAC8,HLA-DPA1,NUB1,CCL7,CASP1,CCR5,GBP3,SP100,CHMP5,IFI27,CCL4,GBP2,DHX58,HAVCR2,AIM2,GSDMD,ADAR,CCL3,STAT2,DTX3L,ZFP36,PML,IFIT5,DDX60L,IFITM1,HERC6,ZNFX1,RIPK2,LILRA5,SERPINB9,GBP7,UNC93B1,TRAFD1,CD86,TANK,CNP,TXNIP,CFD,IRF8,IL10RA,TRIM5,TRIM38","ISG15,IFI6,GBP1,PLSCR1,CXCL11,HSPA1A,SAMD9,IFIT2,IFIT3,IFIT1,IFITM2,IFITM3,IRF7,UBE2L6,OAS1,OASL,ISG20,CCL2,CCL8,IFI35,BST2,APOBEC3A,MX1,MNDA,MX2,HLA-B,GBP4,HLA-E,IFI44L,SLAMF7,HLA-C,NT5C3A,HSP90AA1,NMI,HLA-A,IFIH1,APOBEC3B,HERC5,EIF2AK2,IFI16,NAGK,CD47,FGL2,LGALS9,NINJ1,CTSS,ANXA1,OAS2,DEFB1,GBP5,CASP4,FCN1,OPTN,OAS3,RNF213,GCH1,RBCK1,IRF2,PARP14,CXCL10,LYN,SERPING1,RSAD2,STAT1,RTP4,TRIM22,NCF1,APOBEC3G,PARP9,CD40,IL27,PLAC8,HLA-DPA1,NUB1,CCL7,CASP1,GBP3,SP100,IFI27,CCL4,GBP2,DHX58,HAVCR2,AIM2,GSDMD,ADAR,CCL3,STAT2,DTX3L,PML,IFIT5,DDX60L,IFITM1,ZNFX1,RIPK2,LILRA5,SERPINB9,GBP7,UNC93B1,TRAFD1,TANK,CFD,IRF8,TRIM5,TRIM38","ISG15,IFI6,GBP1,IL1RN,PLSCR1,CXCL11,HSPA1A,SAMD9,IDO1,IFIT2,IFIT3,IFIT1,IFITM2,IFITM3,IRF7,UBE2L6,OAS1,OASL,ISG20,CCL2,CCL8,IFI35,BST2,APOBEC3A,MX1,MNDA,MX2,HLA-B,GBP4,HLA-E,IFI44L,SLAMF7,HLA-C,NT5C3A,HSP90AA1,NMI,HLA-A,IFIH1,APOBEC3B,HERC5,EIF2AK2,TAP1,CST3,IFI16,NAGK,CD47,FGL2,CD48,LGALS9,NINJ1,CTSS,ANXA1,OAS2,DEFB1,TNFAIP6,GBP5,CASP4,FCN1,OPTN,OAS3,RNF213,GCH1,CCR1,RBCK1,IRF2,APOL3,PARP14,CXCL10,LILRB4,LYN,CTSC,SERPING1,RSAD2,STAT1,RTP4,TRIM22,NCF1,APOBEC3G,PARP9,C3AR1,CD40,IL27,CARD16,PLAC8,HLA-DPA1,NUB1,CCL7,CASP1,CCR5,TRADD,GBP3,SP100,IFI27,CCL4,GBP2,DHX58,LGALS2,HAVCR2","ISG15,IFI6,GBP1,IL1RN,PLSCR1,TNFSF10,CXCL11,HSPA1A,SAMD9,IDO1,IFIT2,IFIT3,IFIT1,IFITM2,IFITM3,IRF7,UBE2L6,OAS1,OASL,TNFSF13B,ISG20,CCL2,CCL8,IFI35,BST2,IL4I1,APOBEC3A,MX1,MNDA,MX2,HLA-B,GBP4,HLA-E,IFI44L,SLAMF7,HLA-C,HSP90AA1,NMI,HLA-A,IFIH1,APOBEC3B,CTSL,HERC5,EIF2AK2,TAP1,IFI44,IFI16,NAGK,CD47,FGL2,IGSF6,LGALS9,NINJ1,CTSS,ANXA1,OAS2,DEFB1,GBP5,CASP4,FCN1,OPTN,OAS3,GCH1,CCR1,RBCK1,PARP14,CXCL10,LILRB4,LYN,CTSC,SERPING1,RSAD2,STAT1,TRIM22,NCF1,APOBEC3G,PARP9,C3AR1,CD40,IL27,CD164,HLA-DPA1,SAMSN1,NUB1,CD38,CCL7,CASP1,CCR5,GBP3,SP100,IFI27,CCL4,GBP2,DHX58,HAVCR2","ISG15,PLSCR1,HSPA1A,IRF7,OAS1,OASL,IFI35,IL4I1,MNDA,HLA-B,HLA-E,HSP90AA1,NMI,HLA-A,IFIH1,HERC5,EIF2AK2,IFI16,NAGK,FGL2,LGALS9,NINJ1,CTSS,GBP5,FCN1,OPTN,OAS3,PARP14,LYN,SERPING1,RSAD2,STAT1,TRIM22,NCF1,APOBEC3G,PARP9,CD40,IL27,CARD16,CASP1,GBP2,DHX58,HAVCR2,AIM2,ADAR,STAT2,DTX3L,PML,ZNFX1,RIPK2,SERPINB9,UNC93B1,TRAFD1,TRIM5,TRIM38,CD274,SAMHD1,CLEC4E,IL15,NT5C2,USP15,HLA-F,ZC3HAV1,IRF1,DDX60,HSPA1B,BIRC3,FFAR2,SCIMP,APOBEC3F,TRIM21,MYD88,CASP8,WDFY1,IKBKE,PIK3AP1,N4BP1,ZBP1,CCL5,TLR4,APPL1,MICB,TRIM25,LAG3,MEFV,PELI1,SPN,SLC15A4,LY86,CD1D,CLEC7A,CRTAM,NLRC5,USP18,TICAM1,TLR7,CEP63,TRIM56,MR1","ISG15,PLSCR1,LY6E,IFIT1,IFITM2,IFITM3,OAS1,OASL,ISG20,BST2,APOBEC3A,MX1,IFIH1,APOBEC3B,EIF2AK2,IFI16,OAS2,FCN1,OAS3,RSAD2,STAT1,APOBEC3G,SNX3,SP100,ZFP36,IFIT5,IFITM1,ZNFX1","ISG15,PLSCR1,DYNLT1,LY6E,IFIT1,IFITM2,IFITM3,OAS1,OASL,ISG20,BST2,APOBEC3A,MX1,IFIH1,APOBEC3B,EIF2AK2,IFI16,LGALS9,OAS2,FCN1,OAS3,RSAD2,STAT1,TRIM22,APOBEC3G,SNX3,STOM,SP100,ADAR,ZFP36,IFIT5,IFITM1,ZNFX1,GBP7,TRIM5,TRIM38","ISG15,SSB,PLSCR1,HSPA1A,DYNLT1,LY6E,IFIT1,IFITM2,IFITM3,IRF7,OAS1,OASL,ISG20,CCL2,CCL8,BST2,APOBEC3A,MX1,IFIH1,APOBEC3B,CTSL,EIF2AK2,IFI16,LGALS9,OAS2,CTSB,FCN1,OAS3,SCARB2,HSP90AB1,RSAD2,STAT1,TRIM22,APOBEC3G,PARP9,SNX3,STOM,DEK,CCR5,SP100,CHMP5,IFI27,CCL4,ADAR,CCL3,ZFP36,IFIT5,IFITM1,ZNFX1,GTF2B,GBP7,CD86,TRIM5,TRIM38","ISG15,PLSCR1,IFIT1,IFITM2,IFITM3,OAS1,OASL,ISG20,BST2,APOBEC3A,MX1,IFIH1,APOBEC3B,EIF2AK2,IFI16,OAS2,OAS3,RSAD2,APOBEC3G","ISG15,PLSCR1,HSPA1A,DYNLT1,LY6E,IFIT1,IFITM2,IFITM3,OAS1,OASL,ISG20,CCL2,CCL8,BST2,APOBEC3A,MX1,IFIH1,APOBEC3B,CTSL,EIF2AK2,IFI16,LGALS9,OAS2,CTSB,FCN1,OAS3,SCARB2,HSP90AB1,RSAD2,TRIM22,APOBEC3G,SNX3,STOM,DEK,CCR5,CHMP5,IFI27","ISG15,PLSCR1,IFIT1,IFITM2,IFITM3,OAS1,OASL,ISG20,CCL2,CCL8,BST2,APOBEC3A,MX1,IFIH1,APOBEC3B,EIF2AK2,IFI16,OAS2,OAS3,RSAD2,APOBEC3G,STOM,DEK,IFI27,ADAR,IFIT5,IFITM1,ZNFX1,GBP7,TRIM38","ISG15,GBP1,IL1RN,PLSCR1,CXCL11,HSPA1A,IFIT1,IFITM2,IFITM3,IRF7,OAS1,OASL,TNFSF13B,MT2A,CCL2,CCL8,BST2,MX1,MNDA,MX2,GBP4,IFIH1,XAF1,EIF2AK2,IFI16,CD47,LGALS9,SNX10,ANXA1,OAS2,GBP5,CASP4,OAS3,GCH1,CCR1,PARP14,HSP90AB1,CXCL10,LILRB4,LYN,SOCS1,STAT1,PARP9,CD40,CARD16,HLA-DPA1,NUB1,CD38,CCL7,CASP1,CCR5,TRADD,GBP3,SP100,IFI27,CCL4,GBP2,CCRL2,AIM2,ADAR,CCL3,STAT2,ZFP36,PML,MCL1,IFITM1,SELL,PNPT1","ISG15,PLSCR1,HSPA1A,IDO1,IRF7,OAS1,OASL,IFI35,MNDA,HLA-B,HLA-E,HSP90AA1,NMI,HLA-A,IFIH1,HERC5,EIF2AK2,IFI16,NAGK,CD47,FGL2,LGALS9,NINJ1,CTSS,ANXA1,TNFAIP6,GBP5,CASP4,FCN1,OPTN,OAS3,PARP14,LYN,CTSC,SERPING1,RSAD2,STAT1,TRIM22,NCF1,APOBEC3G,PARP9,CD40,IL27,CARD16,CASP1,TRADD,GBP2,DHX58,LGALS2,HAVCR2,AIM2,GSDMD,ADAR,CCL3,STAT2,DTX3L,ZFP36,PML,ZNFX1,RIPK2,LILRA5,SERPINB9,UNC93B1,TRAFD1,IL10RA,TRIM5,TRIM38,SAMHD1,CLEC4E,IL15,NT5C2,USP15,MGLL,HLA-F,ZC3HAV1,IRF1,DDX60,JAK2,HSPA1B,BIRC3,CCR7,FFAR2,SCIMP,APOBEC3F,TRIM21,MYD88,CASP8,WDFY1,MDK,IKBKE,PIK3AP1,STAT3","ISG15,PLSCR1,LY6E,IFIT1,IFITM2,IFITM3,OAS1,OASL,ISG20,BST2,APOBEC3A,MX1,IFIH1,APOBEC3B,EIF2AK2,IFI16,LGALS9,OAS2,FCN1,OAS3,RSAD2,TRIM22,APOBEC3G,SNX3,ADAR,IFIT5,IFITM1,ZNFX1,GBP7,TRIM5,TRIM38","ISG15,PLSCR1,IFIT1,IFITM2,IFITM3,OAS1,OASL,ISG20,BST2,APOBEC3A,MX1,IFIH1,APOBEC3B,EIF2AK2,IFI16,OAS2,OAS3,RSAD2,APOBEC3G,ADAR,IFIT5,IFITM1,ZNFX1,GBP7,TRIM38","ISG15,IFI6,GBP1,IL1RN,PLSCR1,TNFSF10,CXCL11,HSPA1A,PSMB9,SAMD9,IDO1,IFIT2,IFIT3,IFIT1,IFITM2,IFITM3,IRF7,UBE2L6,OAS1,OASL,TNFSF13B,ISG20,CCL2,CCL8,IFI35,BST2,IL4I1,APOBEC3A,MX1,MNDA,MX2,HLA-B,GBP4,HLA-E,IFI44L,SLAMF7,HLA-C,HSP90AA1,NMI,HLA-A,IFIH1,APOBEC3B,CTSL,HERC5,EIF2AK2,TAP1,IFI44,IFI16,NAGK,KLF6,CD47,CFLAR,FGL2,IGSF6,LGALS9,NINJ1,CTSS,ANXA1,OAS2,DEFB1,TNFAIP6,GBP5,CASP4,FCN1,OPTN,OAS3,RNF213,GCH1,CCR1,RBCK1,IRF2,PARP14,CXCL10,LILRB4,LYN,CTSC,SERPING1,RSAD2,SOCS1,STAT1,TRIM22,NCF1,APOBEC3G,PARP9,C3AR1,CD40,IL27,CD164,HLA-DPA1,SAMSN1,NUB1,CD38,CCL7,CASP1,CCR5,GBP3,SP100,CHMP5,IFI27,CCL4,GBP2,DHX58,LGALS2,HAVCR2","ISG15,PLSCR1,HSPA1A,IRF7,OAS1,OASL,IFI35,MNDA,HLA-B,HLA-E,HSP90AA1,NMI,HLA-A,IFIH1,EIF2AK2,IFI16,NAGK,LGALS9,NINJ1,CTSS,GBP5,FCN1,OAS3,PARP14,LYN,SERPING1,RSAD2,NCF1,PARP9,CD40,CASP1,GBP2,DHX58,HAVCR2,AIM2,ADAR,STAT2,ZNFX1,RIPK2,SERPINB9,UNC93B1,TRAFD1,TRIM5,SAMHD1,CLEC4E,USP15,HLA-F,ZC3HAV1,IRF1,DDX60,HSPA1B,BIRC3,FFAR2,SCIMP,TRIM21,MYD88,CASP8,WDFY1,IKBKE,PIK3AP1,N4BP1,ZBP1,CCL5,TLR4,APPL1,TRIM25,LAG3,MEFV,PELI1,SLC15A4,CD1D,CLEC7A,CRTAM,NLRC5,USP18,TICAM1,TLR7,CEP63,TRIM56","ISG15,GBP1,PLSCR1,HSPA1A,IDO1,IRF7,OAS1,OASL,TNFSF13B,IFI35,IL4I1,MNDA,HLA-B,HLA-E,HLA-C,HSP90AA1,NMI,HLA-A,IFIH1,EIF2AK2,IFI16,NAGK,CD47,FGL2,LGALS9,NINJ1,CTSS,ANXA1,GBP5,FCN1,OAS3,RBCK1,PARP14,LILRB4,LYN,SERPING1,RSAD2,NCF1,PARP9,C3AR1,CD40,IL27,HLA-DPA1,SAMSN1,CD38,CASP1,GBP2,DHX58,HAVCR2,AIM2,ADAR,STAT2,ZNFX1,DENND1B,RIPK2,SERPINB9,UNC93B1,TRAFD1,CD86,CFD,ELF1,TRIM5,ASCL2,CD274,SAMHD1,TAP2,CLEC4E,IL15,CD80,USP15,LILRB2,SPPL2A,CMKLR1,HLA-F,ZC3HAV1,IRF1,DDX60,JAK2,HSPA1B,BIRC3,HCST,CCR7,FFAR2,C9orf72,SCIMP,ITGB2,FPR3,TRIM21,NR4A3,MYD88,CASP8,WDFY1,IKBKE,PIK3AP1,CLEC4D,N4BP1,ZBP1,CCL5,TLR4,APPL1,MICB,TRIM25,LAG3,MEFV,IL7R,PELI1,FCGR1A,SPN,P2RX7,LCP2,SLC15A4,BLNK,WDR41,PDE4B,IL6,CD1D,CLEC7A,CRTAM,NLRC5,USP18,TICAM1,TLR7,C2,CEP63,TRIM56,MR1,PRKD2,ADCY7,STAP1,BIRC2","ISG15,IFI6,GBP1,PLSCR1,CXCL11,HSPA1A,HSPB1,SAMD9,IDO1,IFIT2,IFIT3,IFIT1,IFITM2,IFITM3,IRF7,UBE2L6,OAS1,OASL,ISG20,CCL2,CCL8,IFI35,BST2,APOBEC3A,TYMP,MX1,MNDA,MX2,HLA-B,SOD2,GBP4,HLA-E,IFI44L,SLAMF7,HLA-C,NT5C3A,HSP90AA1,NMI,HLA-A,IFIH1,APOBEC3B,HERC5,EIF2AK2,IFI44,IFI16,NAGK,CD47,FGL2,LGALS9,NINJ1,CTSS,ANXA1,HPSE,CMPK2,OAS2,DEFB1,TNFAIP6,GBP5,CASP4,FCN1,OPTN,OAS3,RNF213,GCH1,CCR1,RBCK1,IRF2,PARP14,CXCL10,LYN,CTSC,SERPING1,RSAD2,STAT1,RTP4,TRIM22,NCF1,APOBEC3G,PARP9,C3AR1,CD40,SNX3,IL27,CARD16,PLAC8,HLA-DPA1,NUB1,CCL7,CASP1,CCR5,TRADD,GBP3,SP100,CHMP5,IFI27,CCL4,GBP2,DHX58,LGALS2,HAVCR2","ISG15,PLSCR1,HSPA1A,HSPB1,IDO1,IRF7,OAS1,OASL,CCL2,IFI35,MNDA,HLA-B,HLA-E,HSP90AA1,NMI,HLA-A,IFIH1,HERC5,EIF2AK2,IFI16,NAGK,CD47,FGL2,LGALS9,NINJ1,CTSS,ANXA1,HPSE,TNFAIP6,GBP5,CASP4,FCN1,OPTN,OAS3,CCR1,PARP14,CXCL10,LYN,CTSC,SERPING1,RSAD2,STAT1,TRIM22,NCF1,APOBEC3G,PARP9,C3AR1,CD40,IL27,CARD16,CCL7,CASP1,TRADD,CCL4,GBP2,DHX58,LGALS2,HAVCR2,AIM2,GSDMD,ADAR,CCL3,STAT2,DTX3L,ZFP36,PML","GBP1,IFITM2,IFITM3,CCL2,CCL8,BST2,GBP4,CD47,LGALS9,GBP5,GCH1,PARP14,STAT1,PARP9,CD40,HLA-DPA1,NUB1,CCL7,CASP1,GBP3,SP100,CCL4,GBP2,CCL3,IFITM1,GBP7,IRF8,IRF1,JAK2,TRIM21,CCL5,RAB20,CALCOCO2,TLR4,CDC42EP2,MEFV,EDN1,CCL3L1","ISG15,GBP1,IL1RN,CXCL11,HSPA1A,IFIT1,IFITM2,IFITM3,IRF7,OAS1,OASL,TNFSF13B,MT2A,CCL2,CCL8,MX1,MNDA,GBP4,IFIH1,IFI16,CD47,LGALS9,SNX10,OAS2,GBP5,CASP4,OAS3,CCR1,PARP14,HSP90AB1,CXCL10,LILRB4,LYN,SOCS1,STAT1,PARP9,CD40,CARD16,HLA-DPA1,CCL7,CASP1,CCR5,TRADD,GBP3,SP100,IFI27,CCL4,GBP2,CCRL2,AIM2,ADAR,CCL3,STAT2,ZFP36,PML,IFITM1,PNPT1,RIPK2,LILRA5,GBP7,TANK,ACSL1,IRF8,IL10RA,SAMHD1,IL15RA,DOCK8,IL15,LILRB2,SPPL2A,CMKLR1,CAPN2,IRF1,JAK2,HSPA1B,BIRC3,CCR7,RNF138,CXCL9,MYD88,GCLM,B3GNT2,CASP8,FAS,HK2,IKBKE,KMO,STAT3,AXL","ISG15,GBP1,HSPA1A,HSPB1,IDO1,IRF7,OAS1,BST2,HLA-B,HLA-E,HSP90AA1,NMI,HLA-A,IFIH1,XAF1,EIF2AK2,IFI16,CD47,LGALS9,ANXA1,HPSE,OAS2,GBP5,FCN1,SERPINB1,OAS3,HSP90AB1,LILRB4,LYN,RSAD2,SOCS1,STAT1,C3AR1,CD40,IL27,CARD16,HLA-DPA1,CASP1,DHX58,HAVCR2,AIM2,GSDMD,CCL3,ZFP36,PML,DENND1B,RIPK2,LILRA5,GBP7,UNC93B1,CD86,TANK,ELF1,ANXA4,IRF8,TRIM38,CD274,SIGLEC1,CLEC4E,CD83,IL15,CD80,LILRB2,CMKLR1,HLA-F,ZC3HAV1,CAPN2,IRF1,JAK2,HSPA1B,CCR7,CD2AP,FFAR2,SCIMP,TRIM21,NR4A3,MYD88,CASP8,MDK,IKBKE,STAT3,AXL,CASP3,N4BP1,TMEM106A","ISG15,GBP1,HSPA1A,HSPB1,IDO1,IRF7,OAS1,BST2,HLA-B,HLA-E,HSP90AA1,NMI,HLA-A,IFIH1,XAF1,EIF2AK2,IFI16,CD47,LGALS9,ANXA1,HPSE,OAS2,GBP5,FCN1,SERPINB1,OAS3,HSP90AB1,LILRB4,LYN,RSAD2,SOCS1,STAT1,C3AR1,CD40,IL27,CARD16,HLA-DPA1,CASP1,DHX58,HAVCR2,AIM2,GSDMD,CCL3,ZFP36,PML,DENND1B,RIPK2,LILRA5,GBP7,UNC93B1,CD86,TANK,ELF1,ANXA4,IRF8,TRIM38,CD274,SIGLEC1,CLEC4E,CD83,IL15,CD80,LILRB2,CMKLR1,HLA-F,ZC3HAV1,CAPN2,IRF1,JAK2,HSPA1B,CCR7,CD2AP,FFAR2,SCIMP,TRIM21,NR4A3,MYD88,CASP8,MDK,IKBKE,STAT3,AXL,CASP3,N4BP1,TMEM106A","PLSCR1,HSPA1A,IDO1,IRF7,OAS1,OASL,IFI35,MNDA,HLA-E,HSP90AA1,NMI,IFIH1,EIF2AK2,IFI16,NAGK,CD47,NINJ1,CTSS,GBP5,CASP4,FCN1,OPTN,OAS3,LYN,CTSC,RSAD2,PARP9,CD40,CASP1,TRADD,GBP2,DHX58,LGALS2,HAVCR2,AIM2,GSDMD,CCL3,ZNFX1,RIPK2,LILRA5,UNC93B1,TRIM5,CLEC4E,IL15,USP15,HLA-F,ZC3HAV1,IRF1,DDX60,HSPA1B,BIRC3,CCR7,FFAR2,SCIMP,MYD88,WDFY1,MDK,IKBKE,PIK3AP1,ZBP1,CCL5,TLR4,APPL1,TRIM25,LAG3,MEFV,PELI1,FCGR1A,SLC15A4,PTGER4,IL6,CD1D,CLEC7A,CRTAM,NLRC5,TICAM1,TLR7,TRIM56","ISG15,OAS1,IL4I1,HLA-B,HLA-E,NMI,HLA-A,IFI16,FGL2,LGALS9,OAS3,PARP14,SERPING1,CARD16,DHX58,HAVCR2,ADAR,STAT2,SERPINB9,TRAFD1,TRIM38,SAMHD1,NT5C2,USP15,HLA-F","PLSCR1,HSPA1A,HSPB1,IDO1,IRF7,OAS1,OASL,IFI35,MNDA,HLA-E,HSP90AA1,NMI,IFIH1,EIF2AK2,IFI16,NAGK,CD47,LGALS9,NINJ1,CTSS,GBP5,CASP4,FCN1,OPTN,OAS3,CCR1,CXCL10,LYN,CTSC,RSAD2,PARP9,C3AR1,CD40,CCL7,CASP1,TRADD,CCL4,GBP2,DHX58,LGALS2,HAVCR2,AIM2,GSDMD,CCL3","ISG15,IL1RN,CXCL11,HSPA1A,IFITM2,IFITM3,IRF7,OAS1,OASL,TNFSF13B,CCL2,CCL8,MX1,IFIH1,OAS2,CASP4,OAS3,CCR1,PARP14,CXCL10,LILRB4,LYN,SOCS1,STAT1,PARP9,CARD16,CCL7,CASP1,CCR5,TRADD,SP100,IFI27,CCL4,CCRL2,AIM2,ADAR,CCL3,STAT2,IFITM1,RIPK2,LILRA5,TANK,ACSL1,IL10RA,SAMHD1,IL15RA,IL15,LILRB2,SPPL2A,CMKLR1,IRF1,JAK2,HSPA1B,BIRC3,CCR7","GBP1,PLSCR1,HSPA1A,IDO1,IRF7,OAS1,OASL,TNFSF13B,IFI35,MNDA,HLA-B,HLA-E,HLA-C,HSP90AA1,NMI,HLA-A,IFIH1,EIF2AK2,IFI16,NAGK,CD47,LGALS9,NINJ1,CTSS,ANXA1,GBP5,FCN1,OAS3,RBCK1,LILRB4,LYN,SERPING1,RSAD2,PARP9,C3AR1,CD40,HLA-DPA1,CD38,CASP1,GBP2,DHX58,HAVCR2,AIM2,ZNFX1,DENND1B,RIPK2,UNC93B1,CD86,CFD,ELF1,TRIM5,CD274,TAP2,CLEC4E,IL15,CD80,USP15,CMKLR1,HLA-F,ZC3HAV1,IRF1,DDX60,JAK2,HSPA1B,BIRC3,CCR7,FFAR2,SCIMP,ITGB2,FPR3,NR4A3,MYD88","ISG15,IFI6,GBP1,IL1RN,PLSCR1,CXCL11,HSPA1A,HSPB1,SAMD9,IDO1,TXN,IFIT2,IFIT3,IFIT1,IFITM2,IFITM3,IRF7,UBE2L6,OAS1,OASL,ISG20,MT2A,CCL2,CCL8,IFI35,MYL12A,BST2,APOBEC3A,MX1,MNDA,MX2,HLA-B,SOD2,GBP4,HLA-E,IFI44L,SLAMF7,PLEK,HLA-C,NT5C3A,HSP90AA1,NMI,HLA-A,IFIH1,APOBEC3B,HERC5,EIF2AK2,TAP1,CST3,IFI16,NAGK,CD47,CFLAR,FGL2,CD48,LGALS9,NINJ1,CTSS,ANXA1,DNAJA1,HPSE,OAS2,CLIC1,DEFB1,TNFAIP6,GBP5,CASP4,FCN1,OPTN,OAS3,RNF213,GCH1,CCR1,RBCK1,IRF2,APOL3,PARP14,HSP90AB1,CXCL10,LILRB4,LYN,CTSC,SERPING1,RSAD2,STAT1,RTP4,TRIM22,NCF1,APOBEC3G,SMCHD1,PARP9,C3AR1,MARCKS,CD40,IL27,CARD16,PLAC8,HLA-DPA1,DEK,FBXO6,NUB1,CD38,CCL7,CASP1,CCR5,TRADD,GBP3,SP100,CHMP5,IFI27,CCL4,GBP2,DHX58,LGALS2,HAVCR2","PLSCR1,HSPA1A,IRF7,OAS1,OASL,IFI35,MNDA,HLA-E,HSP90AA1,NMI,IFIH1,EIF2AK2,IFI16,NAGK,NINJ1,CTSS,GBP5,FCN1,OAS3,LYN,RSAD2,PARP9,CD40,CASP1,GBP2,DHX58,HAVCR2,AIM2,ZNFX1,RIPK2,UNC93B1,TRIM5,CLEC4E,USP15,HLA-F,ZC3HAV1,IRF1,DDX60,HSPA1B,BIRC3,FFAR2,SCIMP,MYD88,WDFY1,IKBKE,PIK3AP1,ZBP1,CCL5,TLR4,APPL1,TRIM25,LAG3,MEFV,PELI1,SLC15A4,CD1D,CLEC7A,CRTAM,NLRC5,TICAM1,TLR7,TRIM56","PLSCR1,HSPA1A,IRF7,OAS1,OASL,IFI35,MNDA,HLA-E,HSP90AA1,NMI,IFIH1,EIF2AK2,IFI16,NAGK,NINJ1,CTSS,GBP5,FCN1,OAS3,LYN,RSAD2,PARP9,CD40,CASP1,GBP2,DHX58,HAVCR2,AIM2,ZNFX1,RIPK2,UNC93B1,TRIM5,CD274,CLEC4E,USP15,HLA-F,ZC3HAV1,IRF1,DDX60,HSPA1B,BIRC3,FFAR2,SCIMP,MYD88,WDFY1,IKBKE,PIK3AP1,ZBP1,CCL5,TLR4,APPL1,TRIM25,LAG3,MEFV,PELI1,SLC15A4,LY86,CD1D,CLEC7A,CRTAM,NLRC5,TICAM1,TLR7,TRIM56,MR1,BIRC2,RELA,TNF","ISG15,HSPA1A,HSPB1,IDO1,IRF7,OAS1,HLA-E,HSP90AA1,HLA-A,IFIH1,EIF2AK2,IFI16,LGALS9,ANXA1,HPSE,OAS2,GBP5,FCN1,OAS3,RSAD2,STAT1,C3AR1,CD40,IL27,HLA-DPA1,CASP1,DHX58,HAVCR2,AIM2,GSDMD,CCL3,DENND1B,RIPK2,LILRA5,UNC93B1,CD86,TANK,IRF8,CD274,CLEC4E,CD83,IL15,CD80,LILRB2,HLA-F,ZC3HAV1,IRF1,JAK2,HSPA1B,CCR7,FFAR2,SCIMP,NR4A3,MYD88,CASP8,MDK,IKBKE,STAT3,TMEM106A,TLR4,APPL1,PANX1,TGFB1,MEFV,PELI1,MCOLN2,SPN,P2RX7,PDE4B,PTGER4,IL6,CLEC7A,CRTAM,TICAM1,TLR7,TRIM56,PRKD2,RELA,TNF,EPHB2,IRF5,RIPK1,HEG1","GBP1,PLSCR1,HSPA1A,IDO1,IRF7,OAS1,OASL,TNFSF13B,CCL2,CCL8,IFI35,IL4I1,MNDA,HLA-B,HLA-E,HLA-C,HSP90AA1,NMI,HLA-A,IFIH1,EIF2AK2,IFI16,NAGK,CD47,LGALS9,NINJ1,CTSS,ANXA1,GBP5,FCN1,OAS3,CCR1,RBCK1,CXCL10,LILRB4,LYN,CTSC,SERPING1,RSAD2,SOCS1,PARP9,C3AR1,CD40,HLA-DPA1,CD38,CCL7,CASP1,CCL4,GBP2,DHX58,HAVCR2,AIM2,CCL3,ZNFX1,DENND1B,RIPK2,UNC93B1,CD86,CFD,ELF1,TRIM5,ASCL2,CD274,TAP2,IL15RA,DOCK8,CLEC4E,CD83,IL15,CD80,USP15,LILRB2,CMKLR1,HLA-F,ZC3HAV1,IRF1,DDX60,JAK2,HSPA1B,BIRC3,CCR7,P2RX4,FFAR2,SCIMP,ITGB2,MIA3,FPR3,NR4A3,MYD88,CASP8,WDFY1,MDK,IKBKE,PIK3AP1,AXL,CLEC4D,ZBP1,CCL5,FUT4,TLR4,APPL1,PANX1,MICB,TRIM25,PDCD1LG2,TGFB1,LAG3,MEFV,IL7R,PELI1,EDN1,FCGR1A,SPN,P2RX7,LCP2,SLC15A4,BLNK,PDE4B,NCK1,IL6,CD1D,CLEC7A,CRTAM,NLRC5","HSPA1A,IRF7,OAS1,OASL,IFI35,MNDA,HSP90AA1,NMI,IFIH1,EIF2AK2,IFI16,NAGK,NINJ1,CTSS,GBP5,FCN1,OAS3,LYN,RSAD2,CD40,CASP1,GBP2,DHX58,HAVCR2,AIM2,ZNFX1,RIPK2,UNC93B1,TRIM5,CLEC4E,USP15,ZC3HAV1,IRF1,DDX60,HSPA1B,BIRC3,FFAR2,SCIMP,MYD88,WDFY1,PIK3AP1,ZBP1,TLR4,APPL1,TRIM25,MEFV,PELI1","ISG15,OAS1,HLA-B,HLA-E,NMI,HLA-A,IFI16,LGALS9,OAS3,PARP14,SERPING1,DHX58,HAVCR2,ADAR,STAT2,SERPINB9,TRAFD1,SAMHD1,USP15,HLA-F","PLSCR1,IFITM2,IFITM3,OAS1,BST2,MNDA,XAF1,IFI16,STAT1,AIM2,IFITM1,PNPT1,CAPN2,IRF1","HSPB1,IFI16,LGALS9,GBP5,CASP1,AIM2,GSDMD,CCL3,RIPK2,LILRA5,JAK2,MYD88,CASP8,STAT3,TMEM106A,TLR4,PANX1,MEFV,P2RX7,IL6,CLEC7A,RELA,TNF","ISG15,GBP1,PLSCR1,HSPA1A,IDO1,IRF7,OAS1,OASL,TNFSF13B,CCL2,CCL8,IFI35,BST2,IL4I1,MNDA,HLA-B,HLA-E,HLA-C,HSP90AA1,NMI,HLA-A,IFIH1,EIF2AK2,IFI16,NAGK,CD47,FGL2,LGALS9,NINJ1,CTSS,ANXA1,TNFAIP6,GBP5,FCN1,OAS3,CCR1,RBCK1,PARP14,CXCL10,LILRB4,LYN,CTSC,SERPING1,RSAD2,SOCS1,STAT1,NCF1,PARP9,C3AR1,CD40,IL27,HLA-DPA1,SAMSN1,CD38,CCL7,CASP1,CCL4,GBP2,DHX58,HAVCR2,AIM2,ADAR,CCL3,STAT2,ZFP36,ZNFX1,DENND1B,RIPK2,SERPINB9,UNC93B1,TRAFD1,CD86,CFD,ELF1,TRIM5,ASCL2,CD274,SAMHD1,TAP2,IL15RA,DOCK8,CLEC4E,CD83,IL15,CD80,USP15,LILRB2,SPPL2A,FBXO7,CMKLR1,HLA-F,ZC3HAV1,IRF1,DDX60,JAK2,HSPA1B,BIRC3,HCST,CCR7,P2RX4,FFAR2,C9orf72,SCIMP,SCIN,ITGB2,MIA3,FPR3,TRIM21,NR4A3,MYD88","ISG15,OAS1,IL4I1,HLA-B,HLA-E,NMI,HLA-A,IFI16,FGL2,LGALS9,ANXA1,OAS3,PARP14,LILRB4,LYN,SERPING1,SAMSN1,DHX58,HAVCR2,ADAR,STAT2,SERPINB9,TRAFD1,ASCL2,SAMHD1,CD80,USP15","ISG15,GBP1,CXCL11,IDO1,OAS1,CCL2,HLA-B,SOD2,GBP4,HLA-E,HLA-A,IFI44,NAGK,LGALS9,CMPK2,OAS2,DEFB1,GBP5,CASP4,OPTN,OAS3,RNF213,GCH1,RBCK1,CXCL10,LYN,CD40,SNX3,IL27,CARD16,PLAC8,CASP1,CCR5,GBP3,CHMP5,GBP2,DHX58,HAVCR2,GSDMD,ZFP36,HERC6,ZNFX1,RIPK2,GBP7,CD86,CNP,CFD,IRF8,IL10RA,TRIM5,CD274,PTGER2,TAP2,CLEC4E","ISG15,IFIT1,IFITM2,IFITM3,IRF7,OAS1,MX1,IFIH1,OAS2,OAS3,STAT1,SP100,IFI27,ADAR,STAT2,IFITM1,TANK,SAMHD1","GBP1,PLSCR1,HSPA1A,IRF7,OAS1,OASL,IFI35,MNDA,HSP90AA1,NMI,HLA-A,IFIH1,EIF2AK2,IFI16,NAGK,CD47,NINJ1,CTSS,GBP5,FCN1,OAS3,RBCK1,LILRB4,LYN,SERPING1,RSAD2,C3AR1,CD40,CD38,CASP1,GBP2,DHX58,HAVCR2,AIM2,ZNFX1,DENND1B,RIPK2,UNC93B1,CFD,ELF1,TRIM5,CLEC4E,USP15,CMKLR1,ZC3HAV1,IRF1,DDX60,HSPA1B,BIRC3,CCR7,FFAR2,SCIMP,FPR3,NR4A3,MYD88","HSPA1A,IRF7,OAS1,OASL,IFI35,NMI,IFIH1,EIF2AK2,NAGK,NINJ1,CTSS,GBP5,FCN1,OAS3,LYN,RSAD2,CD40,CASP1,GBP2,DHX58,HAVCR2,AIM2,RIPK2,UNC93B1,CLEC4E,USP15,ZC3HAV1,IRF1,DDX60,HSPA1B,BIRC3,FFAR2,SCIMP,MYD88,WDFY1,PIK3AP1,TLR4,APPL1,TRIM25,MEFV,PELI1","HSPB1,IFI16,LGALS9,GBP5,SERPINB1,LILRB4,CARD16,CASP1,AIM2,GSDMD,CCL3,PML,RIPK2,LILRA5,JAK2,CCR7,MYD88,CASP8,STAT3,TMEM106A,TLR4,PANX1,MEFV,P2RX7,IL6,CLEC7A,RELA,TNF","HSPB1,IFI16,LGALS9,GBP5,SERPINB1,LILRB4,CARD16,CASP1,AIM2,GSDMD,CCL3,PML,RIPK2,LILRA5,JAK2,CCR7,MYD88,CASP8,STAT3,TMEM106A,TLR4,PANX1,MEFV,P2RX7,IL6,CLEC7A,RELA,TNF","ISG15,PLSCR1,HSPA1A,IDO1,IRF7,OAS1,OASL,IFI35,MNDA,HLA-B,HLA-E,HSP90AA1,NMI,HLA-A,IFIH1,HERC5,EIF2AK2,IFI16,NAGK,CD47,FGL2,LGALS9,NINJ1,CTSS,ANXA1,DNAJA1,HPSE,TNFAIP6,GBP5,CASP4,FCN1,OPTN,OAS3,GCH1,PARP14,LYN,CTSC,SERPING1,RSAD2,STAT1,TRIM22,NCF1,APOBEC3G,SMCHD1,PARP9,CD40,IL27,CARD16,DEK,CASP1,TRADD,GBP2,DHX58,LGALS2,HAVCR2,AIM2,GSDMD,ADAR,CCL3,STAT2,DTX3L,ZFP36,PML","GBP1,CCL2,CCL8,GBP4,CD47,LGALS9,GBP5,PARP14,STAT1,PARP9,HLA-DPA1,CCL7,CASP1,GBP3,SP100,CCL4,GBP2","HSPB1,IFI16,LGALS9,GBP5,CASP1,HAVCR2,AIM2,GSDMD,CCL3,RIPK2,LILRA5,JAK2,MYD88,CASP8,STAT3,TMEM106A,TLR4,PANX1,MEFV,P2RX7,IL6,CLEC7A,RELA,TNF","ISG15,IFIT1,IFITM2,IFITM3,IRF7,OAS1,IFIH1,OAS2,OAS3,STAT1,SP100,IFI27,ADAR,STAT2,IFITM1,TANK,SAMHD1","HLA-B,HLA-C,HLA-A,TAP1,TAP2,TAPBP,HLA-F","HSPB1,IFI16,LGALS9,ANXA1,GBP5,SERPINB1,LILRB4,CARD16,CASP1,HAVCR2,AIM2,GSDMD,CCL3,PML,RIPK2,LILRA5,JAK2,CCR7,MYD88,CASP8,STAT3,TMEM106A,TLR4,PANX1,MEFV,P2RX7,IL6,CLEC7A","HSPB1,IFI16,LGALS9,ANXA1,GBP5,SERPINB1,LILRB4,CARD16,CASP1,HAVCR2,AIM2,GSDMD,CCL3,PML,RIPK2,LILRA5,JAK2,CCR7,MYD88,CASP8,STAT3,TMEM106A,TLR4,PANX1,MEFV,P2RX7,IL6,CLEC7A","GBP1,NINJ1,GBP5,CASP4,CASP1,GBP3,IFI27,GBP2,AIM2,GSDMD,TRIM21,CASP8,CASP3,ZBP1,MEFV","GBP1,PLSCR1,HSPA1A,IRF7,OAS1,OASL,IFI35,MNDA,NMI,HLA-A,IFIH1,EIF2AK2,NAGK,CD47,NINJ1,CTSS,GBP5,FCN1,OAS3,RBCK1,LILRB4,LYN,RSAD2,C3AR1,CD40,CD38,CASP1,GBP2,DHX58,HAVCR2,AIM2,DENND1B,RIPK2,UNC93B1,ELF1,CLEC4E,USP15,CMKLR1,ZC3HAV1,IRF1,DDX60,HSPA1B,BIRC3,CCR7,FFAR2,SCIMP,FPR3,NR4A3,MYD88","HSPA1A,IRF7,OAS1,OASL,IFI35,NMI,IFIH1,EIF2AK2,NAGK,NINJ1,CTSS,GBP5,FCN1,OAS3,LYN,RSAD2,CD40,CASP1,GBP2,DHX58,HAVCR2,AIM2,RIPK2,UNC93B1,CLEC4E,USP15,ZC3HAV1,IRF1,DDX60,HSPA1B,BIRC3,FFAR2,SCIMP,MYD88,WDFY1,PIK3AP1,TLR4,APPL1,TRIM25,MEFV,PELI1","GBP1,PLSCR1,HSPA1A,IRF7,OAS1,OASL,IFI35,MNDA,NMI,HLA-A,IFIH1,EIF2AK2,NAGK,CD47,NINJ1,CTSS,GBP5,FCN1,OAS3,RBCK1,LILRB4,LYN,RSAD2,C3AR1,CD40,CD38,CASP1,GBP2,DHX58,HAVCR2,AIM2,DENND1B,RIPK2,UNC93B1,ELF1,CLEC4E,USP15,LILRB2,CMKLR1,ZC3HAV1,IRF1,DDX60,HSPA1B,BIRC3,CCR7,FFAR2,SCIMP,FPR3,NR4A3,MYD88","ISG15,IFITM2,IFITM3,IRF7,OAS1,IFIH1,OAS2,OAS3,PARP14,STAT1,PARP9,SP100,IFI27,ADAR,STAT2,IFITM1,TANK,SAMHD1,IRF1,JAK2,MYD88,IKBKE,ZBP1","ISG15,IFITM2,IFITM3,IRF7,OAS1,IFIH1,OAS2,OAS3,STAT1,SP100,IFI27,ADAR,STAT2,IFITM1,TANK,SAMHD1","CXCL11,IDO1,CCL2,SOD2,LGALS9,CMPK2,GBP5,GCH1,CXCL10,LYN,CD40,CARD16,CASP1,CCR5,GBP3,CHMP5,GBP2,HAVCR2,ZFP36,RIPK2,CD86,CNP,IRF8,IL10RA,TRIM5,CD274,PTGER2,CD80,LILRB2,CAPN2,JAK2,CCR7,SCIMP,CXCL9,MYD88,CASP8,KMO,CASP7,AXL,CASP3,TLR4,CDC73,PDCD1LG2,PELI1,EDN1,P2RX7,LY86,GJB2,PDE4B,PTGER4,IL6,TICAM1,C2,STAP1,RELA,TNF,EPHB2,PLAA","CXCL11,IDO1,CCL2,SOD2,LGALS9,CMPK2,GBP5,GCH1,CXCL10,LYN,CD40,CARD16,CASP1,CCR5,GBP3,CHMP5,GBP2,HAVCR2,ZFP36,RIPK2,CD86,CNP,IRF8,IL10RA,TRIM5,CD274,PTGER2,TAP2,CD80,LILRB2,CAPN2,JAK2,CCR7,SCIMP,CXCL9,MYD88,CASP8,KMO,CASP7,AXL,CASP3,TLR4,CDC73,PDCD1LG2,PELI1,EDN1,P2RX7,LY86,GJB2,PDE4B,PTGER4,IL6,TICAM1,C2,STAP1,RELA,TNF,EPHB2,IRF5,PLAA","ISG15,IRF7,OAS1,HSP90AA1,NMI,IFIH1,XAF1,OAS2,OAS3,STAT1,DHX58,HAVCR2,RIPK2,GBP7,TANK,IRF8,TRIM38,SIGLEC1,ZC3HAV1,IRF1,TRIM21,MYD88,IKBKE","ISG15,IRF7,OAS1,HSP90AA1,NMI,IFIH1,XAF1,OAS2,OAS3,STAT1,DHX58,HAVCR2,RIPK2,GBP7,TANK,IRF8,TRIM38,SIGLEC1,ZC3HAV1,IRF1,TRIM21,MYD88,IKBKE","HSPA1A,IRF7,OAS1,OASL,IFIH1,EIF2AK2,NAGK,GBP5,OAS3,RSAD2,GBP2,DHX58,HAVCR2,RIPK2,UNC93B1,USP15,ZC3HAV1,DDX60,HSPA1B,BIRC3,SCIMP,MYD88,WDFY1,PIK3AP1,TLR4,TRIM25,MEFV,PELI1,SLC15A4,TICAM1,TLR7,BIRC2,RELA,TNF,SLC15A3","IFITM2,IFITM3,OAS1,BST2,MX2,EIF2AK2","CXCL11,CCL2,CMPK2,GBP5,CXCL10,LYN,CD40,CARD16,CASP1,CCR5,GBP3,CHMP5,GBP2,HAVCR2,ZFP36,RIPK2,CD86,IRF8,TRIM5,CD274,CD80,LILRB2,CAPN2,JAK2,SCIMP,CXCL9,MYD88,KMO,CASP7,AXL,TLR4,CDC73,PDCD1LG2,LY86,PDE4B,IL6,TICAM1,STAP1,RELA,TNF,EPHB2,PLAA","ISG15,GBP1,IL1RN,PLSCR1,TNFSF10,CXCL11,HSPA1A,HSPB1,IDO1,IFIT1,IFITM2,IFITM3,IRF7,OAS1,OASL,TNFSF13B,MT2A,CCL2,CCL8,BST2,MX1,MNDA,MX2,SOD2,GBP4,HSP90AA1,IFIH1,XAF1,CTSL,EIF2AK2,IFI16,NAGK,CD47,CFLAR,LGALS9,CTSS,SNX10,ANXA1,DNAJA1,HPSE,CMPK2,OAS2,CTSB,GBP5,CASP4,OPTN,OAS3,GCH1,CCR1,PARP14,HSP90AB1,CXCL10,LILRB4,LYN,CTSC,SOCS1,STAT1,NCF1,PARP9,CD40,CARD16,RAB8A,NADK,HLA-DPA1,FBXO6,NUB1,CD38,XRN1,CCL7,CASP1,CCR5,TRADD,GBP3,SP100,CHMP5,IFI27,CCL4,GBP2,HAVCR2,CCRL2,AIM2,PIM3,ADAR,CCL3,STAT2,ZFP36,ATF3,PML,MCL1,IFITM1,SELL,PNPT1,SNX6,RIPK2,LILRA5,GLUL,TRAM1,RAB31,SERPINB9,RRAGC,GBP7,CD86,MSR1,TANK,CNP,ACSL1,TXNIP,IRF8,IL10RA,TRIM5","HSP90AA1,HERC5,FGL2,STAT1,TRIM22,APOBEC3G,PARP9,IL27,AIM2,DTX3L,PML,TRIM38,IL15,NT5C2","IL1RN,PLSCR1,CXCL11,IDO1,CCL2,CCL8,IFI35,HLA-E,NMI,EIF2AK2,IFI16,CD47,LGALS9,NINJ1,ANXA1,TNFAIP6,GBP5,CASP4,CCR1,APOL3,CXCL10,LILRB4,LYN,CTSC,NCF1,C3AR1,CD40,IL27,CARD16,CCL7,CASP1,CCR5,TRADD,CCL4,GBP2,LGALS2,HAVCR2,CCRL2,AIM2,GSDMD,CCL3,ZFP36,RIPK2,LILRA5,TXNIP,IL10RA,PTGER2,SIGLEC1,IL15,TRIM14,CMKLR1,MGLL,CD163,JAK2,BIRC3,APOL2,CCR7,FFAR2,ITGB2,CXCL9,FPR3,MYD88,MDK,PIK3AP1,STAT3,AXL,ZBP1,CCL5,FUT4,TLR4,APPL1,TGFB1,MEFV,CCL3L1,FCGR1A,SPN,P2RX7,EXT1,ABCD1,BLNK,LY86,PLGRKT,PTGER4,IL6,CLEC7A,CCL13,USP18,TICAM1,TLR7,RICTOR,BPGM,ADCY7,STAP1,BIRC2,RELA,AOAH,RHBDF2,TNF,RB1,PLD4,EPHB2,IRF5,RIPK1,PARP4,PLAA","ISG15,GBP1,IL1RN,CXCL11,HSPA1A,IFIT1,IFITM2,IFITM3,IRF7,OAS1,OASL,TNFSF13B,MT2A,CCL2,CCL8,MX1,MNDA,SOD2,GBP4,IFIH1,CTSL,EIF2AK2,IFI16,CD47,CFLAR,LGALS9,CTSS,SNX10,ANXA1,DNAJA1,CMPK2,OAS2,CTSB,GBP5,CASP4,OPTN,OAS3,CCR1,PARP14,HSP90AB1,CXCL10,LILRB4,LYN,SOCS1,STAT1,NCF1,PARP9,CD40,CARD16,RAB8A,NADK,HLA-DPA1,XRN1,CCL7,CASP1,CCR5,TRADD,GBP3,SP100,CHMP5,IFI27,CCL4,GBP2,HAVCR2,CCRL2,AIM2,PIM3,ADAR,CCL3,STAT2,ZFP36,ATF3,PML,IFITM1,PNPT1,SNX6,RIPK2,LILRA5,RAB31,SERPINB9,RRAGC,GBP7,CD86,MSR1,TANK,ACSL1,IRF8,IL10RA,TRIM5","OAS1,OASL,MX1,OAS2,STAT1","TNFSF10,BST2,CFLAR,LGALS9,RBCK1,APOL3,TRIM22,CD40,CARD16,CASP1,TRADD,IFIT5,RIPK2,TRIM5,TRIM38,CUL1,ZC3HAV1,BIRC3,CASP10,CCR7,TRIM21,MYD88,CASP8,IKBKE,CLEC4D,SECTM1,TMEM106A,TFG,TLR4,TRIM25,MIER1,SHISA5,PELI1,CLEC7A,TICAM1,BIRC2,RELA,TNF,RIPK1,CD4,SLC44A2,TAB2,TRIM58,PRKCE,ALPK1,NOD1","CXCL11,CCL2,CMPK2,GBP5,CXCL10,LYN,CD40,CARD16,CASP1,CCR5,GBP3,CHMP5,GBP2,HAVCR2,ZFP36,RIPK2,CD86,IRF8,TRIM5,CD274,CD80,LILRB2,CAPN2,JAK2,SCIMP,CXCL9,MYD88,KMO,CASP7,AXL,TLR4,CDC73,PDCD1LG2,LY86,PDE4B,IL6,TICAM1,STAP1,RELA,TNF,EPHB2,PLAA","TNFSF10,HSPB1,BST2,CFLAR,LGALS9,OPTN,RBCK1,APOL3,LILRB4,STAT1,TRIM22,CD40,CARD16,CASP1,TRADD,IFIT5,RIPK2,GBP7,TANK,TRIM5,TRIM38,CUL1,ZC3HAV1,BIRC3,CASP10,CCR7,TRIM21,MYD88,CASP8,IKBKE,CLEC4D,SECTM1,TMEM106A,TFG,TLR4,TRIM25,MIER1,SHISA5,PELI1","HSPA1A,OASL,IFI35,NAGK,NINJ1,GBP5,RSAD2,GBP2,DHX58,USP15,ZC3HAV1,DDX60,HSPA1B,MYD88,WDFY1,TLR4,PELI1,SLC15A4","ISG15,OAS1,HLA-B,HLA-E,NMI,HLA-A,IFI16,FGL2,LGALS9,TNFAIP6,OAS3,PARP14,LYN,SERPING1,DHX58,HAVCR2,ADAR,STAT2,ZFP36,SERPINB9,TRAFD1,IL10RA,TRIM38,SAMHD1,NT5C2,USP15","HSP90AA1,STAT1,TRIM22,APOBEC3G,PARP9,IL27,AIM2,DTX3L,PML","ISG15,GBP1,IDO1,OAS1,CCL2,BST2,IL4I1,MNDA,HLA-B,HLA-E,NMI,HLA-A,IFI16,FGL2,LGALS9,ANXA1,TNFAIP6,OAS3,PARP14,LILRB4,LYN,SERPING1,SOCS1","PLSCR1,HSPA1A,DYNLT1,LY6E,IFIT1,IFITM2,IFITM3,CTSL,LGALS9,CTSB,FCN1,SCARB2,HSP90AB1,TRIM22,SNX3,CCR5,CHMP5,IFITM1,CD86,TRIM5,TRIM38,SIGLEC1,CD80,ARL8B","PLSCR1,HSPA1A,DYNLT1,LY6E,IFITM2,IFITM3,CTSL,LGALS9,CTSB,FCN1,SCARB2,TRIM22,SNX3,CCR5,IFITM1,CD86,TRIM5,TRIM38,SIGLEC1,CD80","CXCL11,CCL2,CMPK2,GBP5,CXCL10,LYN,CD40,CARD16,CASP1,CCR5,GBP3,CHMP5,GBP2,HAVCR2,ZFP36,RIPK2,CD86,TXNIP,IRF8,TRIM5,CD274,CD80,LILRB2,CAPN2,JAK2,SCIMP,CXCL9,MYD88,KMO,CASP7,AXL,TLR4,CDC73,PDCD1LG2,LY86,PDE4B,IL6,CLEC7A,TICAM1,STAP1,RELA,TNF,EPHB2,PLAA","IFIT2,IFIT3,IFIT1,OAS1,MX1,IFIH1,EIF2AK2","IDO1,HLA-B,CD47,LGALS9,CD40,RIPK2,LILRA5,UNC93B1,IRF8,CMKLR1,IRF1,CCR7,SCIMP,MDK,TLR4,MEFV,CLEC7A,RELA,IRF5","IDO1,HLA-B,CD47,LGALS9,CD40,RIPK2,LILRA5,UNC93B1,IRF8,CMKLR1,IRF1,CCR7,SCIMP,MDK,TLR4,MEFV,CLEC7A,RELA,IRF5","PLSCR1,HSPA1A,DYNLT1,LY6E,IFIT1,IFITM2,IFITM3,CCL8,CTSL,LGALS9,CTSB,FCN1,SCARB2,HSP90AB1,TRIM22,SNX3,STOM,CCR5,CHMP5,IFI27,CCL4,CCL3,IFITM1,CD86,TRIM5,TRIM38,SIGLEC1,RAB9A,CD80,ARL8B","GBP1,GBP5,GBP3,GBP2,GBP7","HSP90AA1,STAT1,TRIM22,APOBEC3G,PARP9,IL27,AIM2,DTX3L,PML,IL15,NT5C2","PLSCR1,HSPA1A,DYNLT1,LY6E,IFITM2,IFITM3,CTSL,LGALS9,CTSB,FCN1,SCARB2,TRIM22,SNX3,CCR5,IFITM1,CD86,TRIM5,TRIM38,SIGLEC1,CD80","HLA-B,HLA-E,HLA-C,HLA-A","ISG15,IRF7,OAS1,HSP90AA1,NMI,IFIH1,OAS2,OAS3,RIPK2,TRIM38,ZC3HAV1,IRF1,TLR4,TICAM1,TLR7,TRIM56,MORC3,IRF5","ISG15,IRF7,OAS1,HSP90AA1,NMI,IFIH1,OAS2,OAS3,RIPK2,TRIM38,ZC3HAV1,IRF1,TLR4,TICAM1,TLR7,TRIM56,MORC3,IRF5","HLA-B,HLA-E,HLA-C,HLA-A,CTSL,TAP1,FGL2,CTSS,HLA-DPA1,PSMB8,UNC93B1,TAP2,TAPBP,RAB8B,LILRB2,ARL8B,HLA-F,CCR7","GBP1,HSPA1A,TNFSF13B,CCL2,CCL8,CASP4,GCH1,STAT1,CD40,CARD16,NUB1,CCL7,CASP1,TRADD,GBP3,CCL4,GBP2,AIM2,CCL3,ZFP36","ISG15,IRF7,OAS1,HSP90AA1,IFIH1,OAS2,OAS3,RIPK2,ZC3HAV1,IRF1,TLR4,TICAM1,TLR7,TRIM56,IRF5","ISG15,IRF7,OAS1,HSP90AA1,IFIH1,OAS2,OAS3,STAT1,DHX58,RIPK2,TANK,ZC3HAV1,IRF1,MYD88,IKBKE,TLR4,TICAM1,TLR7,TRIM56,IRF5","GBP1,IL1RN,CCL2,CCL8,CD47,LGALS9,ANXA1,CD40,CD38,CCL7,GBP3,CCL4,GBP2,CCL3,RIPK2,TANK","TNFSF10,HSPB1,BST2,CFLAR,LGALS9,OPTN,RBCK1,APOL3,LILRB4,STAT1,TRIM22,CD40,CARD16,CASP1,TRADD,IFIT5,RIPK2,GBP7,TANK,TRIM5,TRIM38,CUL1,ZC3HAV1,BIRC3,CASP10,CCR7,TRIM21,MYD88,CASP8,IKBKE,CLEC4D,SECTM1,TMEM106A,TFG,TLR4,TRIM25,MIER1,SHISA5,PELI1,EDN1","IFITM2,OAS1,MNDA,IFI16,STAT1,AIM2,PNPT1,CAPN2,IRF1","HSPA1A,DYNLT1,MNDA,HSP90AA1,CST3,FGL2,CTSS,CTSB,TNFAIP6,FCN1,HSP90AB1","IDO1,IFI35,HLA-E,NMI,CD47,NINJ1,CASP4,CTSC,CASP1,TRADD,LGALS2,AIM2,GSDMD,CCL3,LILRA5,IL15,CCR7,FFAR2,MDK,ZBP1,TLR4,MEFV,FCGR1A,PTGER4,IL6,TLR7,STAP1,TNF,RIPK1","GBP1,GBP5,GBP3,GBP2,CASP3","HSPA1A,NAGK,HSPA1B,TLR4,SLC15A4,SLC15A3","IRF7,TNFSF13B,IL4I1,HLA-B,HLA-E,HLA-C,HLA-A,ANXA1,LILRB4,RSAD2,CD40,IL27,SAMSN1,HAVCR2,DENND1B,RIPK2,ASCL2,CD274,TAP2,CD80,HLA-F,IRF1,JAK2,MICB,IL7R,FCGR1A,SPN,P2RX7,SLC15A4,IL6,CD1D,CLEC7A,MR1,ADCY7,TNF,EPHB2,CD4","IDO1,HLA-B,HLA-E,LGALS9,LILRB4,LYN,HAVCR2,CD274,LILRB2","HLA-B,HLA-E,HLA-C,HLA-A,TAP1,CTSS,NCF1,SNX3,RAB8A,RAB31,UNC93B1,TAP2,TAPBP,CLEC4E,RAB9A,RAB8B,TRIM14,HLA-F,SCIMP,SRGAP2","BST2,IL4I1,HLA-B,HLA-E,HLA-C,HLA-A,FGL2,LGALS9,ANXA1,FCN1,LILRB4,LYN,SERPING1,RSAD2,NCF1,CD40,IL27,DHX58,HAVCR2,DENND1B,RIPK2,SERPINB9,CD86,ASCL2,TAP2,CD80,HLA-F,DDX60,FFAR2,SCIMP,ITGB2,NR4A3,MYD88,AXL,TLR4,APPL1,PANX1,MICB,LAG3,IL7R,FCGR1A,SPN,P2RX7,SLC15A4,IL6,CD1D,CLEC7A,CRTAM,TICAM1,TLR7,MR1,STAP1,TNF,EPHB2,IRF5","GBP1,PLSCR1,TNFSF10,HSPA1A,HSPB1,IDO1,TXN,IRF7,OAS1,OASL,TNFSF13B,CCL2,CCL8,IFI35,BST2,MNDA,HLA-B,HLA-E,HLA-C,HSP90AA1,NMI,HLA-A,IFIH1,CTSL,EIF2AK2,IFI16,NAGK,CD47,CFLAR,LGALS9,NINJ1,CTSS,ANXA1,HPSE,GBP5,CASP4,FCN1,OPTN,OAS3,CCR1,RBCK1,APOL3,PARP14,HSP90AB1,CXCL10,LILRB4,LYN,CTSC,SERPING1,RSAD2,TRIM22,NCF1,SMCHD1,PARP9,C3AR1","HLA-B,HLA-E,HLA-C,HLA-A,TAP1,TAP2,TAPBP,HLA-F","CFLAR,TRADD,CASP10,CASP8,FAS,CASP3,RIPK1","HLA-B,HLA-E,HLA-C,HLA-A,TAP1","GBP1,GBP5,GBP3,GBP2,GBP7","LY6E,IFITM2,IFITM3,LGALS9,FCN1,TRIM22,SNX3,IFITM1,TRIM5,TRIM38","LY6E,IFITM2,IFITM3,BST2,FCN1,SNX3,IFITM1,TRIM5","GBP1,HSPA1A,TNFSF13B,CCL2,CCL8,CASP4,STAT1,CD40,CARD16,CCL7,CASP1,TRADD,GBP3,CCL4,GBP2,AIM2,CCL3,ZFP36","HLA-B,HLA-E,HLA-A,LGALS9,HAVCR2,SERPINB9,HLA-F","IDO1,TNFSF13B,IL4I1,MNDA,HLA-E,HLA-A,LGALS9,ANXA1,LILRB4,LYN,CD40,IL27,HLA-DPA1,CD38,HAVCR2,RIPK2,CD86,CD274,IL15,CD80,LILRB2,IRF1,JAK2,MYD88,CASP3,CCL5,TLR4,PDCD1LG2,TGFB1,PELI1,SPN,NCK1,IL6,CD1D,CRTAM,TICAM1","ISG15,GBP1,OAS1,GBP4,HLA-E,HLA-A,NAGK,OAS2,DEFB1,GBP5,CASP4,OPTN,OAS3,RNF213,RBCK1,PLAC8,CASP1,GBP3,GBP2,HAVCR2,GSDMD","HLA-B,HLA-E,HLA-C,HLA-A,HLA-DPA1,TAPBP,SPPL2A,HLA-F","HLA-B,HLA-E,HLA-A,LGALS9,HAVCR2,SERPINB9,HLA-F","TNFSF13B,HLA-E,HLA-A,LGALS9,ANXA1,CD40,HLA-DPA1,CD38,HAVCR2,RIPK2,CD86,CD274,IL15,CD80,LILRB2,JAK2,MYD88,CCL5,TLR4,PDCD1LG2,PELI1,SPN,NCK1,IL6,CD1D,TICAM1","ISG15,IL1RN,HSPA1A,IRF7,OAS1,IFIH1,CASP4,OAS3,PARP14,PARP9,CARD16,CASP1,ADAR,STAT2,RIPK2,SAMHD1,SPPL2A,HSPA1B,CASP8,IKBKE,AXL,ZBP1,CCL5,TLR4,EDN1,IL6,NLRC5,USP18,TRIM56,STAP1","IDO1,TNFSF13B,IL4I1,MNDA,HLA-E,HLA-A,LGALS9,ANXA1,LILRB4,LYN,CD40,IL27,HLA-DPA1,CD38,HAVCR2,RIPK2,CD86,CD274,IL15,CD80,LILRB2,IRF1,JAK2,MYD88,CASP3,CCL5,TLR4,PDCD1LG2,TGFB1,PELI1,SPN,NCK1,IL6,CD1D,CRTAM,TICAM1","IDO1,LGALS9,CD40,RIPK2,UNC93B1,IRF8,IRF1,CCR7,SCIMP,MDK,TLR4,CLEC7A,RELA,IRF5","HLA-B,HLA-E,HLA-C,HLA-A,CTSL,TAP1,CTSS","HSPB1,OAS1,HLA-E,IFIH1,LGALS9,OAS2,OAS3,HAVCR2,CCL3,RIPK2,LILRA5,CD86,JAK2,MYD88,STAT3,TMEM106A,TLR4,SPN,IL6,CLEC7A,TICAM1,EPHB2,RIPK1","OAS1,IFI35,NMI,NINJ1,FCN1,LYN,RIPK2,FFAR2,SCIMP,MYD88,WDFY1,PIK3AP1,TLR4,APPL1,PELI1","HLA-B,HLA-E,HLA-C,HLA-A","HSPA1A,IRF7,OASL,IFI35,EIF2AK2,NAGK,NINJ1,GBP5,LYN,RSAD2,GBP2,DHX58,USP15,ZC3HAV1,IRF1,DDX60,HSPA1B,BIRC3,MYD88,WDFY1,TLR4,APPL1,MEFV,PELI1","TNFSF13B,BST2,HLA-E,HLA-A,LGALS9,ANXA1,LYN,CD40,HLA-DPA1,CD38,HAVCR2,RIPK2,CD86,CD274,IL15,CD80,LILRB2,JAK2,MYD88,CCL5,TLR4,PDCD1LG2,PELI1,SPN,NCK1,IL6,CD1D,TICAM1","TNFSF13B,CCL2,IL4I1,HLA-E,PLEK,HLA-A,CD47,LGALS9,ANXA1,LILRB4,LYN,CTSC,SOCS1,CD40,HLA-DPA1,CD38,HAVCR2,CCL3,RIPK2,LILRA5,CD86,CD274,IL15RA,DOCK8,CD83,IL15,CD80,LILRB2,JAK2,CCR7,NR4A3,MYD88,MDK,AXL,CLEC4D,CCL5,TLR4,PDCD1LG2,TGFB1,IL7R,PELI1,SPN,ACTA2,NCK1,IL6,CD1D,CLEC7A,TICAM1","HLA-B,HLA-E,HLA-C,HLA-A,TAP1,TAP2,TAPBP,HLA-F,MICB,FCGR1A,MR1,LNPEP,ERAP2,TAPBPL","OAS1,OAS2,OAS3","CXCL11,CCL2,CCL8,CCR1,CXCL10,LYN,CCL7,CCR5,CCL4,CCRL2,CCL3","HSPA1A,NAGK,HSPA1B,TLR4,SLC15A4,SLC15A3","TAP1,TAP2,TAPBP,HLA-F"],[247,325,746,655,1138,1003,1004,1024,803,1146,1187,396,73,130,345,44,243,109,651,561,109,71,1744,333,664,1647,739,109,583,566,569,353,93,444,347,545,2578,275,287,364,787,224,68,27,51,1108,145,400,64,412,191,74,74,982,93,57,60,7,85,85,28,375,206,394,78,59,235,247,109,109,126,20,160,1692,54,550,1284,7,171,168,230,54,176,24,352,148,104,189,41,56,56,223,5,34,110,8,53,53,94,181,38,64,89,259,21,122,102,5,6,161,27,122,276,1482,15,9,17,6,31,20,162,14,183,144,23,15,113,132,186,37,55,84,51,8,118,127,295,28,3,54,6,4],[71,78,102,95,99,211,211,121,105,98,95,99,28,36,54,19,37,30,68,92,31,25,104,79,130,100,66,38,89,85,85,78,25,44,55,72,115,62,68,83,124,47,20,14,23,110,27,54,18,55,41,28,28,63,17,24,17,7,28,28,15,49,41,50,23,16,58,60,23,23,35,6,42,110,14,105,89,5,46,42,39,18,26,9,23,24,20,44,7,19,19,30,5,11,20,4,18,18,18,20,15,20,16,40,9,11,29,5,6,37,9,20,55,55,8,7,5,5,10,8,18,7,36,21,8,7,26,30,36,14,7,23,15,4,24,28,48,14,3,11,6,4]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>query<\/th>\n      <th>term_name<\/th>\n      <th>term_id<\/th>\n      <th>p_value<\/th>\n      <th>intersection<\/th>\n      <th>term_size<\/th>\n      <th>intersection_size<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":5,"columnDefs":[{"className":"dt-right","targets":[4,6,7]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"query","targets":1},{"name":"term_name","targets":2},{"name":"term_id","targets":3},{"name":"p_value","targets":4},{"name":"intersection","targets":5},{"name":"term_size","targets":6},{"name":"intersection_size","targets":7}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><p>And similarly, show the GO terms enriched within the down-regulated
genes.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">goResult</span><span class="op">[[</span><span class="st">"CD14 Mono.stim"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">result</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">query</span> <span class="op">==</span> <span class="st">"Down"</span>, <span class="fu"><a href="https://rdrr.io/r/base/startsWith.html" class="external-link">startsWith</a></span><span class="op">(</span><span class="va">term_id</span>, <span class="st">"GO"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">p_value</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">query</span>, <span class="va">term_name</span>, <span class="va">term_id</span>, <span class="va">p_value</span>, <span class="va">intersection</span>, <span class="va">term_size</span>, <span class="va">intersection_size</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-e5c8c404fe174e4c81bd" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-e5c8c404fe174e4c81bd">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6"],["Down","Down","Down","Down","Down","Down"],["cytosolic large ribosomal subunit","large ribosomal subunit","cytosolic ribosome","negative regulation of cysteine-type endopeptidase activity involved in apoptotic signaling pathway","cytoplasmic translation","regulation of apoptotic signaling pathway"],["GO:0022625","GO:0015934","GO:0022626","GO:2001268","GO:0002181","GO:2001233"],[6.23e-05,0.00497,0.00497,0.021,0.0263,0.0319],["RPL6,RPL7,RPL10A,RPL23,RPL4,RPL5","RPL6,RPL7,RPL10A,RPL23,RPL4,RPL5","RPL6,RPL7,RPL10A,RPL23,RPL4,RPL5","PLAUR,PARK7,MMP9","RPL6,RPL7,RPL10A,RPL23,RPL4,RPL5,EIF3L","S100A8,IER3,PLAUR,ENO1,THBS1,PPIF,IL1B,PARK7,G0S2,MMP9"],[55,110,109,5,143,292],[6,6,6,3,7,10]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>query<\/th>\n      <th>term_name<\/th>\n      <th>term_id<\/th>\n      <th>p_value<\/th>\n      <th>intersection<\/th>\n      <th>term_size<\/th>\n      <th>intersection_size<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":5,"columnDefs":[{"className":"dt-right","targets":[4,6,7]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"query","targets":1},{"name":"term_name","targets":2},{"name":"term_id","targets":3},{"name":"p_value","targets":4},{"name":"intersection","targets":5},{"name":"term_size","targets":6},{"name":"intersection_size","targets":7}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><p>Here’s also a simple visualization function that generates a dot plot
per test, with dot size representing the interaction size between query
gene set and each term, and color representing the significance of the
term. Use <code>group = "CD14 Mono.stim"</code> to specify the test
group, and <code>query = "Up"</code> or <code>query = "Down"</code> to
specify the if the test uses up- or down-regulated genes,
respectively.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGODot.html">plotGODot</a></span><span class="op">(</span><span class="va">goResult</span>, group <span class="op">=</span> <span class="st">"CD14 Mono.stim"</span>, query <span class="op">=</span> <span class="st">"Up"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-32-1.png" width="864" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGODot.html">plotGODot</a></span><span class="op">(</span><span class="va">goResult</span>, group <span class="op">=</span> <span class="st">"CD14 Mono.stim"</span>, query <span class="op">=</span> <span class="st">"Down"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-33-1.png" width="864" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="r-session-info">R Session Info<a class="anchor" aria-label="anchor" href="#r-session-info"></a>
</h2>
<pre><code><span><span class="co">## R version 4.4.0 (2024-04-24)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS Sonoma 14.5</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/Detroit</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] cowplot_1.1.3 dplyr_1.1.4   rliger_2.1.0 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] bitops_1.0-9                gridExtra_2.3              </span></span>
<span><span class="co">##   [3] rlang_1.1.4                 magrittr_2.0.3             </span></span>
<span><span class="co">##   [5] clue_0.3-65                 GetoptLong_1.0.5           </span></span>
<span><span class="co">##   [7] RcppAnnoy_0.0.22            matrixStats_1.4.1          </span></span>
<span><span class="co">##   [9] compiler_4.4.0              sccore_1.0.5               </span></span>
<span><span class="co">##  [11] gprofiler2_0.2.3            png_0.1-8                  </span></span>
<span><span class="co">##  [13] systemfonts_1.1.0           vctrs_0.6.5                </span></span>
<span><span class="co">##  [15] pkgconfig_2.0.3             shape_1.4.6.1              </span></span>
<span><span class="co">##  [17] crayon_1.5.3                fastmap_1.2.0              </span></span>
<span><span class="co">##  [19] magick_2.8.3                XVector_0.44.0             </span></span>
<span><span class="co">##  [21] labeling_0.4.3              utf8_1.2.4                 </span></span>
<span><span class="co">##  [23] rmarkdown_2.28              UCSC.utils_1.0.0           </span></span>
<span><span class="co">##  [25] ragg_1.3.2                  purrr_1.0.2                </span></span>
<span><span class="co">##  [27] xfun_0.48                   zlibbioc_1.50.0            </span></span>
<span><span class="co">##  [29] cachem_1.1.0                GenomeInfoDb_1.40.0        </span></span>
<span><span class="co">##  [31] jsonlite_1.8.9              highr_0.11                 </span></span>
<span><span class="co">##  [33] DelayedArray_0.30.1         BiocParallel_1.38.0        </span></span>
<span><span class="co">##  [35] cluster_2.1.6               irlba_2.3.5.1              </span></span>
<span><span class="co">##  [37] parallel_4.4.0              R6_2.5.1                   </span></span>
<span><span class="co">##  [39] bslib_0.8.0                 RColorBrewer_1.1-3         </span></span>
<span><span class="co">##  [41] leidenAlg_1.1.3             GenomicRanges_1.56.0       </span></span>
<span><span class="co">##  [43] jquerylib_0.1.4             scattermore_1.2            </span></span>
<span><span class="co">##  [45] Rcpp_1.0.13                 SummarizedExperiment_1.34.0</span></span>
<span><span class="co">##  [47] iterators_1.0.14            knitr_1.48                 </span></span>
<span><span class="co">##  [49] IRanges_2.38.0              Matrix_1.7-0               </span></span>
<span><span class="co">##  [51] igraph_2.0.3                tidyselect_1.2.1           </span></span>
<span><span class="co">##  [53] rstudioapi_0.16.0           abind_1.4-8                </span></span>
<span><span class="co">##  [55] yaml_2.3.10                 viridis_0.6.5              </span></span>
<span><span class="co">##  [57] doParallel_1.0.17           codetools_0.2-20           </span></span>
<span><span class="co">##  [59] lattice_0.22-6              tibble_3.2.1               </span></span>
<span><span class="co">##  [61] Biobase_2.64.0              withr_3.0.1                </span></span>
<span><span class="co">##  [63] evaluate_1.0.1              desc_1.4.3                 </span></span>
<span><span class="co">##  [65] isoband_0.2.7               circlize_0.4.16            </span></span>
<span><span class="co">##  [67] pillar_1.9.0                MatrixGenerics_1.16.0      </span></span>
<span><span class="co">##  [69] DT_0.33                     foreach_1.5.2              </span></span>
<span><span class="co">##  [71] stats4_4.4.0                plotly_4.10.4              </span></span>
<span><span class="co">##  [73] generics_0.1.3              RCurl_1.98-1.14            </span></span>
<span><span class="co">##  [75] S4Vectors_0.42.0            ggplot2_3.5.1              </span></span>
<span><span class="co">##  [77] munsell_0.5.1               scales_1.3.0               </span></span>
<span><span class="co">##  [79] glue_1.8.0                  lazyeval_0.2.2             </span></span>
<span><span class="co">##  [81] tools_4.4.0                 data.table_1.16.2          </span></span>
<span><span class="co">##  [83] locfit_1.5-9.9              RANN_2.6.2                 </span></span>
<span><span class="co">##  [85] fs_1.6.4                    Cairo_1.6-2                </span></span>
<span><span class="co">##  [87] grid_4.4.0                  tidyr_1.3.1                </span></span>
<span><span class="co">##  [89] crosstalk_1.2.1             colorspace_2.1-1           </span></span>
<span><span class="co">##  [91] GenomeInfoDbData_1.2.12     cli_3.6.3                  </span></span>
<span><span class="co">##  [93] textshaping_0.4.0           fansi_1.0.6                </span></span>
<span><span class="co">##  [95] S4Arrays_1.4.0              viridisLite_0.4.2          </span></span>
<span><span class="co">##  [97] ComplexHeatmap_2.20.0       uwot_0.2.2                 </span></span>
<span><span class="co">##  [99] gtable_0.3.5                DESeq2_1.44.0              </span></span>
<span><span class="co">## [101] sass_0.4.9                  digest_0.6.37              </span></span>
<span><span class="co">## [103] BiocGenerics_0.50.0         RcppPlanc_1.0.0            </span></span>
<span><span class="co">## [105] SparseArray_1.4.3           ggrepel_0.9.6              </span></span>
<span><span class="co">## [107] rjson_0.2.21                htmlwidgets_1.6.4          </span></span>
<span><span class="co">## [109] farver_2.1.2                memoise_2.0.1              </span></span>
<span><span class="co">## [111] htmltools_0.5.8.1           pkgdown_2.0.9              </span></span>
<span><span class="co">## [113] lifecycle_1.0.4             httr_1.4.7                 </span></span>
<span><span class="co">## [115] GlobalOptions_0.1.2         MASS_7.3-60.2</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
