<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Joint definition of cell types from multiple scRNA-seq datasets • rliger</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Joint definition of cell types from multiple scRNA-seq datasets">
<meta property="og:description" content="rliger">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown-header">Installation</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Introductory Vignettes</li>
    <li class="dropdown-header">Integrating Multiple scRNAseq Data</li>
    <li class="dropdown-header">Integrating On-Disk (HDF5 Based) Data</li>
    <li class="divider">
    </li>
<li class="dropdown-header">Data structure and access</li>
    <li class="dropdown-header">Outer container - liger object</li>
    <li class="dropdown-header">Dataset container - ligerDataset object and subclasses</li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Import Data</li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Joint definition of cell types from multiple
scRNA-seq datasets</h1>
                        <h4 data-toc-skip class="author">Joshua Sodicoff
and Joshua Welch</h4>
            
            <h4 data-toc-skip class="date">2023-02-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/vignettes/Integrating_multi_scRNA_data.rmd" class="external-link"><code>vignettes/Integrating_multi_scRNA_data.rmd</code></a></small>
      <div class="hidden name"><code>Integrating_multi_scRNA_data.rmd</code></div>

    </div>

    
    
<p>This guide will demonstrate the usage of the Liger package in the
style of the R Console, which can be accessed through an R development
environment (e.g., RStudio) or directly from the R command line.</p>
<div class="section level2">
<h2 id="stage-i-preprocessing-and-normalization-3---5-seconds">Stage I: Preprocessing and Normalization (3 - 5 seconds)<a class="anchor" aria-label="anchor" href="#stage-i-preprocessing-and-normalization-3---5-seconds"></a>
</h2>
<div class="section level3">
<h3 id="read-data-into-memory">1. Read data into memory<a class="anchor" aria-label="anchor" href="#read-data-into-memory"></a>
</h3>
<p>For the first portion of this protocol, we will be integrating data
from control and interferon-stimulated PBMCs from <a href="https://www.nature.com/articles/nbt.4042" class="external-link">Kang et al, 2017</a>.
The data can be found in the Gene Expression Omnibus, <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE96583" class="external-link">Series
GSE96583</a>. This dataset was originally in the form of output from the
10X Cellranger pipeline, though we will directly load downsampled
versions of the control and stimulated DGEs here.</p>
<p><strong>For convenience, we have prepared the pre-processed data
which are ready to use.</strong> There are three datasets:
“PBMC_control.RDS” and “PBMC_interferon-stimulated.RDS”, which
correspond to control and interferon-stimulated PBMCs individually. The
data can be downloaded <a href="https://www.dropbox.com/sh/u94ib3dkf9pb6nd/AABemvnxDgKDGRs8Ek5QGlXWa?dl=0" class="external-link">here</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/welch-lab/liger" class="external-link">rliger</a></span><span class="op">)</span></span>
<span><span class="va">ctrl_dge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../pbmcs_ctrl.RDS"</span><span class="op">)</span>;</span>
<span><span class="va">stim_dge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../pbmcs_stim.RDS"</span><span class="op">)</span>;</span></code></pre></div>
<p>For 10X CellRanger output, we can instead use the
<code><a href="../reference/read10X.html">read10X()</a></code> function, which generates a matrix or list of
matrices directly from the output directories.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">matrix_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read10X.html">read10X</a></span><span class="op">(</span>sample.dirs <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"10x_ctrl_outs"</span>, <span class="st">"10x_stim_outs"</span><span class="op">)</span>, sample.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ctrl"</span>, <span class="st">"stim"</span><span class="op">)</span>, merge <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>;</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-liger-obejct">2. Create liger obejct<a class="anchor" aria-label="anchor" href="#create-liger-obejct"></a>
</h3>
<p>With the digital gene expression matrices for both datasets, we can
initialize a Liger object using the <code><a href="../reference/createLiger.html">createLiger()</a></code>
function.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb_liger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ctrl <span class="op">=</span> <span class="va">ctrl_dge</span>, stim <span class="op">=</span> <span class="va">stim_dge</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><code>ifnb_liger</code> now contains two datasets in its raw.data
slot, ctrl and stim. We can run the rest of the analysis on this Liger
object.</p>
</div>
<div class="section level3">
<h3 id="preprocess">3. Preprocess<a class="anchor" aria-label="anchor" href="#preprocess"></a>
</h3>
<p>Before we can run iNMF on our datasets, we must run several
preprocessing steps to normalize expression data to account for
differences in sequencing depth and efficiency between cells, identify
variably expressed genes, and scale the data so that each gene has the
same variance. Note that because nonnegative matrix factorization
requires positive values, we do not center the data by subtracting the
mean. We also do not log transform the data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb_liger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="va">ifnb_liger</span><span class="op">)</span></span>
<span><span class="va">ifnb_liger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="va">ifnb_liger</span><span class="op">)</span></span>
<span><span class="va">ifnb_liger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="va">ifnb_liger</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="determine-the-parameters">4. Determine the parameters<a class="anchor" aria-label="anchor" href="#determine-the-parameters"></a>
</h3>
<p>We are now able to run integrative non-negative matrix factorization
on the normalized and scaled datasets. The key parameter for this
analysis is k, the number of matrix factors (analogous to the number of
principal components in PCA). In general, we find that a value of k
between 20 and 40 is suitable for most analyses and that results are
robust for choice of k. Because LIGER is an unsupervised, exploratory
approach, there is no single “right” value for k, and in practice, users
choose k from a combination of biological prior knowledge and other
information.</p>
</div>
</div>
<div class="section level2">
<h2 id="stage-ii-joint-matrix-factorization-3---10-minutes">Stage II: Joint Matrix Factorization (3 - 10 minutes)<a class="anchor" aria-label="anchor" href="#stage-ii-joint-matrix-factorization-3---10-minutes"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb_liger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/optimizeALS.html">optimizeALS</a></span><span class="op">(</span><span class="va">ifnb_liger</span>, k <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p>Important parameters are as follows:</p>
<ul>
<li>
<em>k</em> - Integer value specifying the inner dimension of
factorization, or number of factors. Higher k is recommended for
datasets with more substructure. We find that a value of k in the range
20 - 40 works well for most datasets. Because this is an unsupervised,
exploratory analysis, there is no single “right” value for k, and in
practice, users choose k from a combination of biological prior
knowledge and other information.</li>
<li>
<em>lambda</em> - This is a regularization parameter. Larger values
penalize dataset-specific effects more strongly, causing the datasets to
be better aligned, but possibly at the cost of higher reconstruction
error. Default <code>5</code>. We recommend using this value for most
analyses, but find that it can be lowered to 1 in cases where the
dataset differences are expected to be relatively small, such as
scRNA-seq data from the same tissue but different individuals.</li>
<li>
<em>thresh</em> - This sets the convergence threshold. Lower values
cause the algorithm to run longer. Default <code>1e-6</code>.</li>
<li>
<em>maxIter</em> - This variable sets the maximum number of
iterations to perform. Default <code>30</code>.</li>
</ul>
<p>The optimization yields several lower dimension matrices, including
the <span class="math inline">\(H\)</span> matrix of metagene loadings
for each cell, the <span class="math inline">\(W\)</span> matrix of
shared factor loadings and the <span class="math inline">\(V\)</span>
matrices of dataset-specific factor loadings.</p>
<p>Please note that the time required of this step is highly dependent
on the size of the datasets being used. In most cases, this step should
not take much longer than 30 minutes.</p>
</div>
<div class="section level2">
<h2 id="stage-iii-quantile-normalization-and-joint-clustering-1-minute">Stage III: Quantile Normalization and Joint Clustering (1
minute)<a class="anchor" aria-label="anchor" href="#stage-iii-quantile-normalization-and-joint-clustering-1-minute"></a>
</h2>
<div class="section level3">
<h3 id="align-the-factors">5. Align the factors<a class="anchor" aria-label="anchor" href="#align-the-factors"></a>
</h3>
<p>We can now use the resulting factors to jointly cluster cells and
perform quantile normalization by dataset, factor, and cluster to fully
integrate the datasets. All of this functionality is encapsulated within
the <code>quantileNorm()</code> function, which uses max factor
assignment followed by refinement using a k-nearest neighbors graph.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb_liger</span> <span class="op">&lt;-</span> <span class="fu">quantileNorm</span><span class="op">(</span><span class="va">ifnb_liger</span><span class="op">)</span></span></code></pre></div>
<p>Important parameters of <code>quantileNorm()</code> are as
follows:</p>
<ul>
<li>
<em><code>nNeighbors</code></em> - This sets the number of nearest
neighbors for within-dataset KNN graph. Default <code>20</code>.</li>
<li>
<em><code>quantiles</code></em> - This sets the number of quantiles
to use for quantile normalization. Default <code>50</code>.</li>
<li>
<em><code>minCells</code></em> - This indicates the minimum number
of cells to consider a cluster as shared across datasets. Default
<code>20</code>.</li>
<li>
<em><code>useDims</code></em> - This sets the indices of factors to
use for quantile normalization. The user can pass in a vector of indices
indicating specific factors. This is helpful for excluding factors
capturing biological signals such as the cell cycle or technical signals
such as mitochondrial genes. The default is all k of the factors.</li>
<li>
<em><code>center</code></em> - This indicates whether to center the
data when scaling factors. Default <code>FALSE</code>. This option
should be set to <code>TRUE</code> when metagene loadings have a mean
above zero, as with dense data such as DNA methylation.</li>
<li>
<em><code>maxSample</code></em> - This sets the maximum number of
cells used for quantile normalization of each cluster and factor.
Default <code>1000</code>.</li>
<li>
<em><code>refineKNN</code></em> - This indicates whether to increase
robustness of cluster assignments using KNN graph. Default
<code>TRUE</code>.</li>
<li>
<em><code>eps</code></em> - This sets the error bound of the nearest
neighbor search. The default is <code>0.9</code>. Lower values give more
accurate nearest neighbor graphs but take much longer to computer.</li>
<li>
<em><code>reference</code></em> - This indicates the name of the
dataset to be used as a reference for quantile normalization. By
default, the dataset with the largest number of cells is used.</li>
</ul>
</div>
<div class="section level3">
<h3 id="clustering">6. Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h3>
<p>The <code>quantileNorm()</code> procedure produces joint clustering
assignments and a low-dimensional representation that integrates the
datasets together. These joint clusters directly from iNMF can be used
for downstream analyses (see below). Alternatively, you can also run
Louvain community detection, an algorithm commonly used for single-cell
data, on the normalized cell factors. The Louvain algorithm excels at
merging small clusters into broad cell classes and thus may be more
desirable in some cases than the maximum factor assignments produced
directly by iNMF.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb_liger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/louvainCluster.html">louvainCluster</a></span><span class="op">(</span><span class="va">ifnb_liger</span>, resolution <span class="op">=</span> <span class="fl">0.25</span>, nNeighbors <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="stage-iv-visualization-2---3-minutes-and-downstream-analysis-25---40-seconds">Stage IV: Visualization (2 - 3 minutes) and Downstream Analysis (25
- 40 seconds)<a class="anchor" aria-label="anchor" href="#stage-iv-visualization-2---3-minutes-and-downstream-analysis-25---40-seconds"></a>
</h2>
<div class="section level3">
<h3 id="generate-dimensionality-reduced-embedding">7. Generate dimensionality reduced embedding<a class="anchor" aria-label="anchor" href="#generate-dimensionality-reduced-embedding"></a>
</h3>
<p>To visualize the clustering of cells graphically, we can project the
normalized cell factors to two or three dimensions. Liger supports both
t-SNE and UMAP for this purpose. Note that if both techniques are run,
the object will only hold the results from the most recent.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb_liger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">ifnb_liger</span>, distance <span class="op">=</span> <span class="st">'cosine'</span>, n_neighbors <span class="op">=</span> <span class="fl">30</span>, min_dist <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<p>The <code>rliger</code> package implements a variety of utilities for
visualization and analysis of clustering, gene expression across
datasets, and comparisons of cluster assignments. We will summarize
several here.</p>
</div>
<div class="section level3">
<h3 id="create-plots">8. Create plots<a class="anchor" aria-label="anchor" href="#create-plots"></a>
</h3>
<p><code><a href="../reference/plotClusterDimRed.html">plotByDatasetAndCluster()</a></code> returns two graphs, generated
by t-SNE or UMAP in the previous step. The first colors cells by dataset
of origin, and the second by cluster as determined by Liger. The plots
provide visual confirmation that the datasets are well aligned and the
clusters are consistent with the shape of the data as revealed by
UMAP.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotClusterDimRed.html">plotByDatasetAndCluster</a></span><span class="op">(</span><span class="va">ifnb_liger</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/4-1-1.png" width="960" style="display: block; margin: auto;"></p>
<p>To directly study the impact of factors on the clustering and
determine what genes load most highly on each factor, we use the
plotGeneLoadings function, which returns plots of factor loading on the
dimensionally reduced graphs and highly loaded genes by dataset for each
factor.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">factorMarkers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getFactorMarkers.html">getFactorMarkers</a></span><span class="op">(</span><span class="va">ifnb_liger</span>, dataset1 <span class="op">=</span> <span class="st">"ctrl"</span>, dataset2 <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotGeneLoadings.html">plotGeneLoadings</a></span><span class="op">(</span><span class="va">ifnb_liger</span>, markerTable <span class="op">=</span> <span class="va">factorMarkers</span>, useFactor <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-2-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Using the <code><a href="../reference/runWilcoxon.html">runWilcoxon()</a></code> function, we can next identify
gene markers for all clusters. We can also compare expression within
each cluster across datasets, which in this case reveals markers of
interferon-beta stimulation. The function returns a table of data that
allows us to determine the significance of each gene’s differential
expression, including log fold change, area under the curve and
p-value.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cluster.results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runWilcoxon.html">runWilcoxon</a></span><span class="op">(</span><span class="va">ifnb_liger</span>, method <span class="op">=</span> <span class="st">"clusters"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cluster.results</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     feature group    avgExpr        logFC statistic       auc          pval</span></span>
<span><span class="co">## 1 LINC00115     0 -22.959298 -0.005233673  18052529 0.4998531  8.249561e-01</span></span>
<span><span class="co">## 2     NOC2L     0 -22.625381 -0.920638905  16984918 0.4702922  4.454753e-32</span></span>
<span><span class="co">## 3    KLHL17     0 -23.012477 -0.007396182  18048826 0.4997506  4.669214e-01</span></span>
<span><span class="co">## 4   PLEKHN1     0 -22.822091  0.158611126  18247210 0.5052436  5.942716e-13</span></span>
<span><span class="co">## 5      HES4     0 -17.954723  3.788720176  22389518 0.6199392 1.786389e-267</span></span>
<span><span class="co">## 6     ISG15     0  -9.042043  2.638307374  24507522 0.6785842 3.058849e-225</span></span>
<span><span class="co">##            padj pct_in pct_out</span></span>
<span><span class="co">## 1  8.538120e-01    100     100</span></span>
<span><span class="co">## 2  3.865938e-31    100     100</span></span>
<span><span class="co">## 3  5.269389e-01    100     100</span></span>
<span><span class="co">## 4  2.641506e-12    100     100</span></span>
<span><span class="co">## 5 8.052636e-266    100     100</span></span>
<span><span class="co">## 6 1.211855e-223    100     100</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">datasets.results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runWilcoxon.html">runWilcoxon</a></span><span class="op">(</span><span class="va">ifnb_liger</span>, method <span class="op">=</span> <span class="st">"datasets"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">datasets.results</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     feature  group   avgExpr        logFC statistic         auc          pval</span></span>
<span><span class="co">## 1 LINC00115 0-ctrl -22.95345   0.01028626 1428236.0 0.500314922  7.831783e-01</span></span>
<span><span class="co">## 2     NOC2L 0-ctrl -22.40323   0.39100147 1464462.5 0.513005163  3.037979e-06</span></span>
<span><span class="co">## 3    KLHL17 0-ctrl -22.99488   0.03097077 1430244.0 0.501018330  4.692700e-02</span></span>
<span><span class="co">## 4   PLEKHN1 0-ctrl -22.92373  -0.17889581 1410486.0 0.494097049  3.071954e-03</span></span>
<span><span class="co">## 5      HES4 0-ctrl -22.85106  -8.61786261  639402.5 0.223984420 5.497616e-243</span></span>
<span><span class="co">## 6     ISG15 0-ctrl -16.86357 -13.76636462     664.0 0.000232601  0.000000e+00</span></span>
<span><span class="co">##            padj pct_in pct_out</span></span>
<span><span class="co">## 1  8.890833e-01    100     100</span></span>
<span><span class="co">## 2  1.699057e-05    100     100</span></span>
<span><span class="co">## 3  1.038107e-01    100     100</span></span>
<span><span class="co">## 4  9.846462e-03    100     100</span></span>
<span><span class="co">## 5 1.016422e-240    100     100</span></span>
<span><span class="co">## 6  0.000000e+00    100     100</span></span></code></pre>
<p>The number of significant genes identified by
<code><a href="../reference/runWilcoxon.html">runWilcoxon()</a></code> varies and depends on the datasets used. You
can then filter the markers which are statistically and biologically
significant. For example, one strategy is to filter the output by taking
markers which have padj (Benjamini-Hochberg adjusted p-value) less than
0.05 and logFC (log fold change between observations in group versus
out) larger than 3:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cluster.results</span> <span class="op">&lt;-</span> <span class="va">cluster.results</span><span class="op">[</span><span class="va">cluster.results</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span>,<span class="op">]</span></span>
<span><span class="va">cluster.results</span> <span class="op">&lt;-</span> <span class="va">cluster.results</span><span class="op">[</span><span class="va">cluster.results</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">3</span>,<span class="op">]</span></span></code></pre></div>
<p>You can then re-sort the markers by its padj value in ascending order
and choose the top 100 for each cell type. For example, we can subset
and re-sort the output for Cluster 3 and take the top 20 markers by
typing these commands:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wilcoxon.cluster_3</span> <span class="op">&lt;-</span> <span class="va">cluster.results</span><span class="op">[</span><span class="va">cluster.results</span><span class="op">$</span><span class="va">group</span> <span class="op">==</span> <span class="fl">3</span>, <span class="op">]</span></span>
<span><span class="va">wilcoxon.cluster_3</span> <span class="op">&lt;-</span> <span class="va">wilcoxon.cluster_3</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">wilcoxon.cluster_3</span><span class="op">$</span><span class="va">padj</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="va">wilcoxon.cluster_3</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">markers</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        feature group    avgExpr    logFC statistic       auc pval padj pct_in</span></span>
<span><span class="co">## 38467      ID3     3 -15.618658 6.032842  12236370 0.6919701    0    0    100</span></span>
<span><span class="co">## 41422    BANK1     3 -19.908153 3.069597  10555352 0.5969081    0    0    100</span></span>
<span><span class="co">## 42024     CD74     3  -5.159748 6.902373  16141072 0.9127821    0    0    100</span></span>
<span><span class="co">## 42218     CD83     3 -13.554142 7.258636  13119884 0.7419330    0    0    100</span></span>
<span><span class="co">## 42378  HLA-DRA     3  -5.789938 8.307215  14084679 0.7964925    0    0    100</span></span>
<span><span class="co">## 42381 HLA-DQA1     3 -13.694802 6.955797  12896722 0.7293132    0    0    100</span></span>
<span><span class="co">##       pct_out</span></span>
<span><span class="co">## 38467     100</span></span>
<span><span class="co">## 41422     100</span></span>
<span><span class="co">## 42024     100</span></span>
<span><span class="co">## 42218     100</span></span>
<span><span class="co">## 42378     100</span></span>
<span><span class="co">## 42381     100</span></span></code></pre>
<p>We can then visualize the expression profiles of individual genes,
such as the differentially expressed genes that we just identified. This
allows us to visually confirm the cluster- or dataset-specific
expression patterns of marker genes. <code><a href="../reference/plotClusterDimRed.html">plotGeneDimRed()</a></code>
returns graphs of gene loading on the dimensionality reduced graph for
each dataset.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PRF1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotClusterDimRed.html">plotGeneDimRed</a></span><span class="op">(</span><span class="va">ifnb_liger</span>, <span class="st">"PRF1"</span>, splitBy <span class="op">=</span> <span class="st">"dataset"</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">PRF1</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ifnb_liger</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-5-1.png" width="960" style="display: block; margin: auto;"></p>
<p>We can also use <code><a href="../reference/plotClusterDimRed.html">plotGeneDimRed()</a></code> to compare the loading
of cluster markers within and between datasets.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">IFIT3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotClusterDimRed.html">plotGeneDimRed</a></span><span class="op">(</span><span class="va">ifnb_liger</span>, <span class="st">"IFIT3"</span>, splitBy <span class="op">=</span> <span class="st">"dataset"</span><span class="op">)</span></span>
<span><span class="va">IFITM3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotClusterDimRed.html">plotGeneDimRed</a></span><span class="op">(</span><span class="va">ifnb_liger</span>, <span class="st">"IFITM3"</span>, splitBy <span class="op">=</span> <span class="st">"dataset"</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">IFIT3</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">IFIT3</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="va">IFITM3</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">IFITM3</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, ncol <span class="op">=</span> <span class="fl">2</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ctrl"</span>, <span class="st">"stim"</span>, <span class="st">"ctrl"</span>, <span class="st">"stim"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-6-1.png" width="864" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="r-session-info">R Session Info<a class="anchor" aria-label="anchor" href="#r-session-info"></a>
</h2>
<pre><code><span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Ventura 13.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] rliger_1.2.0 Matrix_1.5-3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] bitops_1.0-7                matrixStats_0.63.0         </span></span>
<span><span class="co">##  [3] fs_1.5.2                    RcppAnnoy_0.0.20           </span></span>
<span><span class="co">##  [5] doParallel_1.0.17           RColorBrewer_1.1-3         </span></span>
<span><span class="co">##  [7] rprojroot_2.0.3             GenomeInfoDb_1.34.7        </span></span>
<span><span class="co">##  [9] tools_4.2.1                 bslib_0.4.2                </span></span>
<span><span class="co">## [11] utf8_1.2.2                  R6_2.5.1                   </span></span>
<span><span class="co">## [13] irlba_2.3.5.1               uwot_0.1.14                </span></span>
<span><span class="co">## [15] DBI_1.1.3                   BiocGenerics_0.44.0        </span></span>
<span><span class="co">## [17] colorspace_2.0-3            withr_2.5.0                </span></span>
<span><span class="co">## [19] sp_1.6-0                    tidyselect_1.2.0           </span></span>
<span><span class="co">## [21] compiler_4.2.1              progressr_0.13.0           </span></span>
<span><span class="co">## [23] textshaping_0.3.6           cli_3.6.0                  </span></span>
<span><span class="co">## [25] Biobase_2.58.0              desc_1.4.2                 </span></span>
<span><span class="co">## [27] DelayedArray_0.24.0         labeling_0.4.2             </span></span>
<span><span class="co">## [29] sass_0.4.4                  scales_1.2.1               </span></span>
<span><span class="co">## [31] pkgdown_2.0.7               systemfonts_1.0.4          </span></span>
<span><span class="co">## [33] stringr_1.5.0               digest_0.6.31              </span></span>
<span><span class="co">## [35] rmarkdown_2.20              XVector_0.38.0             </span></span>
<span><span class="co">## [37] pkgconfig_2.0.3             htmltools_0.5.4            </span></span>
<span><span class="co">## [39] parallelly_1.34.0           MatrixGenerics_1.10.0      </span></span>
<span><span class="co">## [41] highr_0.10                  fastmap_1.1.0              </span></span>
<span><span class="co">## [43] rlang_1.0.6                 rstudioapi_0.14            </span></span>
<span><span class="co">## [45] FNN_1.1.3.1                 farver_2.1.1               </span></span>
<span><span class="co">## [47] jquerylib_0.1.4             generics_0.1.3             </span></span>
<span><span class="co">## [49] riverplot_0.10              jsonlite_1.8.4             </span></span>
<span><span class="co">## [51] dplyr_1.0.10                RCurl_1.98-1.9             </span></span>
<span><span class="co">## [53] magrittr_2.0.3              GenomeInfoDbData_1.2.9     </span></span>
<span><span class="co">## [55] Rcpp_1.0.10                 munsell_0.5.0              </span></span>
<span><span class="co">## [57] S4Vectors_0.36.1            fansi_1.0.3                </span></span>
<span><span class="co">## [59] lifecycle_1.0.3             stringi_1.7.12             </span></span>
<span><span class="co">## [61] yaml_2.3.6                  SummarizedExperiment_1.28.0</span></span>
<span><span class="co">## [63] zlibbioc_1.44.0             plyr_1.8.8                 </span></span>
<span><span class="co">## [65] grid_4.2.1                  parallel_4.2.1             </span></span>
<span><span class="co">## [67] listenv_0.9.0               ggrepel_0.9.2              </span></span>
<span><span class="co">## [69] lattice_0.20-45             cowplot_1.1.1              </span></span>
<span><span class="co">## [71] knitr_1.41                  pillar_1.8.1               </span></span>
<span><span class="co">## [73] GenomicRanges_1.50.2        future.apply_1.10.0        </span></span>
<span><span class="co">## [75] codetools_0.2-18            stats4_4.2.1               </span></span>
<span><span class="co">## [77] glue_1.6.2                  evaluate_0.20              </span></span>
<span><span class="co">## [79] SeuratObject_4.1.3          vctrs_0.5.1                </span></span>
<span><span class="co">## [81] foreach_1.5.2               gtable_0.3.1               </span></span>
<span><span class="co">## [83] RANN_2.6.1                  purrr_1.0.1                </span></span>
<span><span class="co">## [85] future_1.30.0               assertthat_0.2.1           </span></span>
<span><span class="co">## [87] cachem_1.0.6                ggplot2_3.4.0              </span></span>
<span><span class="co">## [89] xfun_0.36                   viridisLite_0.4.1          </span></span>
<span><span class="co">## [91] ragg_1.2.5                  SingleCellExperiment_1.20.0</span></span>
<span><span class="co">## [93] tibble_3.1.8                iterators_1.0.14           </span></span>
<span><span class="co">## [95] memoise_2.0.1               IRanges_2.32.0             </span></span>
<span><span class="co">## [97] globals_0.16.2</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Joshua Welch, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko, Yichen Wang.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
