<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rliger">
<title>Joint definition of cell types from multiple scRNA-seq datasets • rliger</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Joint definition of cell types from multiple scRNA-seq datasets">
<meta property="og:description" content="rliger">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Joint definition of cell types from multiple scRNA-seq datasets</h1>
                        <h4 data-toc-skip class="author">Joshua Sodicoff
and Joshua Welch</h4>
            
            <h4 data-toc-skip class="date">2023-02-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/vignettes/articles/Integrating_multi_scRNA_data.rmd" class="external-link"><code>vignettes/articles/Integrating_multi_scRNA_data.rmd</code></a></small>
      <div class="d-none name"><code>Integrating_multi_scRNA_data.rmd</code></div>
    </div>

    
    
<p>This guide will demonstrate the usage of the Liger package in the
style of the R Console, which can be accessed through an R development
environment (e.g., RStudio) or directly from the R command line.</p>
<div class="section level2">
<h2 id="stage-i-preprocessing-and-normalization">Stage I: Preprocessing and Normalization<a class="anchor" aria-label="anchor" href="#stage-i-preprocessing-and-normalization"></a>
</h2>
<div class="section level3">
<h3 id="loading-data">1. Loading data<a class="anchor" aria-label="anchor" href="#loading-data"></a>
</h3>
<p>For the first portion of this protocol, we will be integrating data
from control and interferon-stimulated PBMCs from <a href="https://www.nature.com/articles/nbt.4042" class="external-link">Kang et al, 2017</a>.
The data can be found in the Gene Expression Omnibus, <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE96583" class="external-link">Series
GSE96583</a>. This dataset was originally in the form of output from the
10X Cellranger pipeline, though we will directly load downsampled
versions of the control and stimulated DGEs here.</p>
<p><strong>For convenience, we have prepared the pre-processed data
which are ready to use.</strong> There are two datasets: PBMC control
and PBMC interferon-stimulated. We provided ready-to-use liger object,
which can be easily loaded with <code><a href="../reference/importVignetteData.html">importPBMC()</a></code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://welch-lab.github.io/liger/" class="external-link">rliger</a></span><span class="op">)</span></span>
<span><span class="va">pbmcLiger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importVignetteData.html">importPBMC</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>For creating a liger object from raw counts data or any other types
of source, please refer to the <a href="#">detailed tutorial for
importing data</a>.</p>
</div>
<div class="section level3">
<h3 id="preprocess">2. Preprocess<a class="anchor" aria-label="anchor" href="#preprocess"></a>
</h3>
<p>Before we can run iNMF on our datasets, we must run several
preprocessing steps to normalize expression data to account for
differences in sequencing depth and efficiency between cells, identify
variably expressed genes, and scale the data so that each gene has the
same variance. Note that because nonnegative matrix factorization
requires positive values, we do not center the data by subtracting the
mean. We also do not log transform the data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcLiger</span> <span class="op">&lt;-</span> <span class="va">pbmcLiger</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="stage-ii-integration-with-joint-matrix-factorization">Stage II: Integration with Joint Matrix Factorization<a class="anchor" aria-label="anchor" href="#stage-ii-integration-with-joint-matrix-factorization"></a>
</h2>
<div class="section level3">
<h3 id="determine-parameters-and-perform-inmf-integration">3. Determine parameters and perform iNMF integration<a class="anchor" aria-label="anchor" href="#determine-parameters-and-perform-inmf-integration"></a>
</h3>
<p>We are now able to run integrative non-negative matrix factorization
(iNMF) on the normalized and scaled datasets. The key parameter for this
analysis is <code>k</code>, the number of matrix factors (analogous to
the number of principal components in PCA). In general, we find that a
value of <code>k</code> between 20 and 40 is suitable for most analyses
and that results are robust for choice of <code>k</code>. Because LIGER
is an unsupervised, exploratory approach, there is no single “right”
value for <code>k</code>. In practice, users choose <code>k</code> from
a combination of biological prior knowledge and other information. For
this tutorial, we set <code>k = 20</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcLiger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runIntegration.html">runIntegration</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, k <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p>Starting from rliger 2.0.0, we use an optimized implementation of
iNMF. Here we deprecated the parameter <code>thresh</code> which stands
for a convergence detecter in order to speed up each algorithm iteration
by omitting the calculation of objective error.</p>
</blockquote>
<p>The factorization yields several lower dimension matrices, including
the <span class="math inline">\(H\)</span> matrices of metagene loadings
for each cell, the <span class="math inline">\(W\)</span> matrix of
shared factor loadings and the <span class="math inline">\(V\)</span>
matrices of dataset-specific factor loadings. Please refer to <a href="#">liger object documentation</a> for how to access them.</p>
<p>The time consumption of this step is dependent of the size of the
datasets, in terms of number of cells, number of variable genes
selected, and the value of <code>k</code>. The implementation supports
OpenMP multi-threading, and there for using a machine with a number of
cores allocated helps speeding it up.</p>
</div>
</div>
<div class="section level2">
<h2 id="stage-iii-quantile-normalization-and-joint-clustering">Stage III: Quantile Normalization and Joint Clustering<a class="anchor" aria-label="anchor" href="#stage-iii-quantile-normalization-and-joint-clustering"></a>
</h2>
<div class="section level3">
<h3 id="align-the-factors">4. Align the factors<a class="anchor" aria-label="anchor" href="#align-the-factors"></a>
</h3>
<p>We can now use the resulting factors to jointly cluster cells and
perform quantile normalization by dataset, factor, and cluster to fully
integrate the datasets. All of this functionality is encapsulated within
the <code><a href="../reference/quantileNorm.html">quantileNorm()</a></code> function, which uses max factor
assignment followed by refinement using a k-nearest neighbors graph.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcLiger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantileNorm.html">quantileNorm</a></span><span class="op">(</span><span class="va">pbmcLiger</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="clustering">5. Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h3>
<p>The <code><a href="../reference/quantileNorm.html">quantileNorm()</a></code> procedure produces joint clustering
assignments and a low-dimensional representation that integrates the
datasets together. These joint clusters directly from iNMF can be used
for downstream analyses (see below). Alternatively, you can also run
Louvain community detection, an algorithm commonly used for single-cell
data, on the normalized cell factors. The Louvain algorithm excels at
merging small clusters into broad cell classes and thus may be more
desirable in some cases than the maximum factor assignments produced
directly by iNMF.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcLiger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCluster.html">runCluster</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, resolution <span class="op">=</span> <span class="fl">0.25</span>, nNeighbors <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p>Starting from rliger 2.0.0, cluster labeling will be stored in cell
metadata, which can be accessed with <code>cellMeta(pbmcLiger)</code>.
Use argument <code>clusterName</code> to specify unique variable names
for the result can enable storing multiple cluster labeling variables at
the same time.</p>
</blockquote>
</div>
</div>
<div class="section level2">
<h2 id="stage-iv-visualization-and-downstream-analysis">Stage IV: Visualization and Downstream Analysis<a class="anchor" aria-label="anchor" href="#stage-iv-visualization-and-downstream-analysis"></a>
</h2>
<div class="section level3">
<h3 id="generate-dimensionality-reduced-embedding">6. Generate dimensionality reduced embedding<a class="anchor" aria-label="anchor" href="#generate-dimensionality-reduced-embedding"></a>
</h3>
<p>To visualize the clustering of cells graphically, we can project the
normalized cell factors to two or three dimensions. LIGER supports both
UMAP and t-SNE for this purpose.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcLiger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, n_neighbors <span class="op">=</span> <span class="fl">30</span>, min_dist <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p>Starting from rliger 2.0.0, the slot for storing dimensionality
reduction matrices will be renamed to “dimReds”. It will be a list that
can hold multiple low dimensional matrices that match to the dataset by
cell identifiers. Users can access individual matrix with
<code>dimRed(object, "name")</code>. Use argument
<code>dimredName</code> to specify unique names for the UMAP result so
that it allows storing multiple low-dimensional representation matrices
at the same time.</p>
</blockquote>
</div>
<div class="section level3">
<h3 id="create-plots">7. Create plots<a class="anchor" aria-label="anchor" href="#create-plots"></a>
</h3>
<p>We provide a variety of utilities for visualization and analysis of
clustering, gene expression across datasets, and comparisons of cluster
assignments. Here we demonstrate several commonly used examples.</p>
<p><code><a href="../reference/plotDimRed.html">plotByDatasetAndCluster()</a></code> returns two graphs, generated
by t-SNE or UMAP in the previous step. The first colors cells by dataset
of origin, and the second by cluster as determined by previous
clustering step. The plots provide visual confirmation that the datasets
are well aligned and the clusters are consistent with the shape of the
data as revealed by UMAP.</p>
<p>The two subplots can individually be generated with
<code><a href="../reference/plotDimRed.html">plotDatasetDimRed()</a></code> and <code><a href="../reference/plotDimRed.html">plotClusterDimRed()</a></code>,
respectively.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimRed.html">plotByDatasetAndCluster</a></span><span class="op">(</span><span class="va">pbmcLiger</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/4-1-1.png" width="960" style="display: block; margin: auto;"></p>
<p>To directly study the impact of factors on the clustering and
determine what genes load most highly on each factor, we use the
<code><a href="../reference/plotGeneLoadings.html">plotGeneLoadings()</a></code> function, which returns plots of factor
loading on the dimensionality reduction and highly loaded genes by
dataset for each factor.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">factorMarkers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getFactorMarkers.html">getFactorMarkers</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, dataset1 <span class="op">=</span> <span class="st">"ctrl"</span>, dataset2 <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotGeneLoadings.html">plotGeneLoadings</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, markerTable <span class="op">=</span> <span class="va">factorMarkers</span>, useFactor <span class="op">=</span> <span class="fl">11</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-2-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="differential-expression">8. Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h3>
<p>Using the <code><a href="../reference/liger-DEG.html">runMarkerDEG()</a></code> function, we can next identify
gene markers for all clusters. We can also compare expression within
each cluster across datasets, which in this case reveals markers of
interferon-beta stimulation. The function returns a table of data that
allows us to determine the significance of each gene’s differential
expression, including log fold change, area under the curve (auc) and
p-value.</p>
<p>The default parameters performs Wilcoxon rank-sum test at a cluster
level:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cluster.results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-DEG.html">runMarkerDEG</a></span><span class="op">(</span><span class="va">pbmcLiger</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cluster.results</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     feature group     avgExpr        logFC statistic       auc          pval</span></span>
<span><span class="co">## 1 LINC00115     0  0.06557615 -0.006707732  19046596 0.4998280  7.902339e-01</span></span>
<span><span class="co">## 2     NOC2L     0  0.42228382 -0.916821744  17926394 0.4704313  1.876299e-33</span></span>
<span><span class="co">## 3    KLHL17     0  0.01232972 -0.009023670  19042684 0.4997253  4.105767e-01</span></span>
<span><span class="co">## 4   PLEKHN1     0  0.18811948  0.141802251  19232439 0.5047050  3.185813e-11</span></span>
<span><span class="co">## 5      HES4     0  4.78053287  3.500041958  23275316 0.6107997 4.655312e-241</span></span>
<span><span class="co">## 6     ISG15     0 13.53789841  2.106235323  25048607 0.6573351 7.640631e-185</span></span>
<span><span class="co">##            padj      pct_in    pct_out</span></span>
<span><span class="co">## 1  8.228727e-01  0.43243243  0.4660647</span></span>
<span><span class="co">## 2  1.603211e-32  2.81081081  8.5542286</span></span>
<span><span class="co">## 3  4.708915e-01  0.08108108  0.1359355</span></span>
<span><span class="co">## 4  1.232679e-10  1.24324324  0.3010001</span></span>
<span><span class="co">## 5 1.816141e-239 30.48648649  8.1270026</span></span>
<span><span class="co">## 6 2.512153e-183 71.37837838 64.4819885</span></span></code></pre>
<p>Alternatively, it is also helpful to identify dataset specific
markers within each cluster. For example in this tutorial, we can split
data by cluster labeling, and within a cluster, find the markers from
interferon-stimulated cells.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">datasets.results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-DEG.html">runMarkerDEG</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, conditionBy <span class="op">=</span> <span class="st">"dataset"</span>, splitBy <span class="op">=</span> <span class="st">"leiden_cluster"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">datasets.results</span><span class="op">$</span><span class="va">`0`</span><span class="op">)</span> <span class="co"># Note that the first cluster is "0"</span></span></code></pre></div>
<pre><code><span><span class="co">##     feature group    avgExpr        logFC statistic          auc          pval</span></span>
<span><span class="co">## 1 LINC00115  ctrl 0.07183401   0.01158862 1701494.0 0.5003517031  7.453285e-01</span></span>
<span><span class="co">## 2     NOC2L  ctrl 0.66393901   0.44750960 1750817.0 0.5148559253  5.085806e-08</span></span>
<span><span class="co">## 3    KLHL17  ctrl 0.02680373   0.02680373 1703295.0 0.5008813161  6.054223e-02</span></span>
<span><span class="co">## 4   PLEKHN1  ctrl 0.09765942  -0.16751863 1681486.5 0.4944681756  2.474522e-03</span></span>
<span><span class="co">## 5      HES4  ctrl 0.18904920  -8.50274754  775284.5 0.2279848885 3.942972e-269</span></span>
<span><span class="co">## 6     ISG15  ctrl 6.05199165 -13.86279031    2251.0 0.0006619428  0.000000e+00</span></span>
<span><span class="co">##            padj     pct_in     pct_out</span></span>
<span><span class="co">## 1  8.278741e-01  0.4700353   0.4004004</span></span>
<span><span class="co">## 2  3.231322e-07  4.4065805   1.4514515</span></span>
<span><span class="co">## 3  1.220948e-01  0.1762632   0.0000000</span></span>
<span><span class="co">## 4  7.671268e-03  0.6462985   1.7517518</span></span>
<span><span class="co">## 5 7.147217e-267  1.2338425  55.4054054</span></span>
<span><span class="co">## 6  0.000000e+00 37.7790834 100.0000000</span></span></code></pre>
<p>The number of significant genes identified by
<code><a href="../reference/liger-DEG.html">runMarkerDEG()</a></code> varies and depends on the datasets used. And
the raw output of the function contains the statistics of all tested
genes in all groups (clusters). In order to pick out the top markers for
each cluster, we strongly suggest using package “dplyr”, which provides
a user-friendly interface for data table manipulation. The following
code chunk first filters the markers which are statistically and
biologically significant. For example, we filter the output by taking
markers which have padj (Benjamini-Hochberg adjusted p-value) less than
0.05 and logFC (log fold change between observations in group versus
out) larger than 3. Then for each cluster, we sort the markers primarily
by its padj value in ascending order. Given that mathematically, the
lowest padj values are rounded to 0 as they are too small, for genes
tying on this metric, we then sort the markers by logFC in descending
order. Finally, we select the top 20 markers for each cluster.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">cluster.results.sort</span> <span class="op">&lt;-</span> <span class="va">cluster.results</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">3</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">padj</span>, <span class="op">-</span><span class="va">logFC</span>, .by_group <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html" class="external-link">top_n</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span><span class="co"># rank by logFC from high to low</span></span>
<span></span>
<span><span class="co"># Show the markers for cluster 3</span></span>
<span><span class="va">cluster.results.sort</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">group</span> <span class="op">==</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 20 × 10</span></span></span>
<span><span class="co">## <span style="color: #949494;"># Groups:   group [1]</span></span></span>
<span><span class="co">##    feature  group avgExpr logFC statistic   auc      pval      padj pct_in</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> HLA-DRA  3        17.3  8.42 14<span style="text-decoration: underline;">081</span>152  0.800 0   <span style="color: #949494;"> </span>     0   <span style="color: #949494;"> </span>       97.9</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> CD74     3        18.0  7.01 16<span style="text-decoration: underline;">134</span>037  0.917 0   <span style="color: #949494;"> </span>     0   <span style="color: #949494;"> </span>       99.1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> HLA-DPA1 3        14.5  7.76 13<span style="text-decoration: underline;">686</span>101  0.778 3.53<span style="color: #949494;">e</span><span style="color: #BB0000;">-302</span> 2.50<span style="color: #949494;">e</span><span style="color: #BB0000;">-299</span>   87.2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> RPS5     3        17.0  3.20 13<span style="text-decoration: underline;">721</span>636  0.780 7.87<span style="color: #949494;">e</span><span style="color: #BB0000;">-260</span> 4.19<span style="color: #949494;">e</span><span style="color: #BB0000;">-257</span>   98.2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> HLA-DPB1 3        13.6  6.73 13<span style="text-decoration: underline;">249</span>646. 0.753 2.68<span style="color: #949494;">e</span><span style="color: #BB0000;">-251</span> 1.27<span style="color: #949494;">e</span><span style="color: #BB0000;">-248</span>   81.3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> HLA-DRB1 3        15.3  7.22 13<span style="text-decoration: underline;">192</span>870. 0.750 3.92<span style="color: #949494;">e</span><span style="color: #BB0000;">-231</span> 1.47<span style="color: #949494;">e</span><span style="color: #BB0000;">-228</span>   90.1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CCR7     3        14.0  7.14 12<span style="text-decoration: underline;">975</span>000. 0.737 5.12<span style="color: #949494;">e</span><span style="color: #BB0000;">-223</span> 1.72<span style="color: #949494;">e</span><span style="color: #BB0000;">-220</span>   83.0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> RPL10A   3        16.2  3.85 12<span style="text-decoration: underline;">668</span>400  0.720 6.70<span style="color: #949494;">e</span><span style="color: #BB0000;">-163</span> 1.40<span style="color: #949494;">e</span><span style="color: #BB0000;">-160</span>   95.0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> EEF1B2   3        13.1  4.34 12<span style="text-decoration: underline;">392</span>988. 0.704 1.99<span style="color: #949494;">e</span><span style="color: #BB0000;">-150</span> 3.57<span style="color: #949494;">e</span><span style="color: #BB0000;">-148</span>   80.0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> RPLP0    3        15.5  3.55 12<span style="text-decoration: underline;">437</span>076. 0.707 1.40<span style="color: #949494;">e</span><span style="color: #BB0000;">-144</span> 2.23<span style="color: #949494;">e</span><span style="color: #BB0000;">-142</span>   92.4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">11</span> HSP90AB1 3        12.4  3.63 12<span style="text-decoration: underline;">280</span>914. 0.698 5.75<span style="color: #949494;">e</span><span style="color: #BB0000;">-142</span> 8.63<span style="color: #949494;">e</span><span style="color: #BB0000;">-140</span>   74.6</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">12</span> RPL5     3        14.5  4.01 12<span style="text-decoration: underline;">272</span>426. 0.698 8.83<span style="color: #949494;">e</span><span style="color: #BB0000;">-135</span> 1.24<span style="color: #949494;">e</span><span style="color: #BB0000;">-132</span>   87.2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">13</span> RPSA     3        14.8  3.96 12<span style="text-decoration: underline;">242</span>276. 0.696 1.30<span style="color: #949494;">e</span><span style="color: #BB0000;">-131</span> 1.71<span style="color: #949494;">e</span><span style="color: #BB0000;">-129</span>   89.0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">14</span> CXCR4    3        12.9  3.83 12<span style="text-decoration: underline;">106</span>296. 0.688 1.21<span style="color: #949494;">e</span><span style="color: #BB0000;">-127</span> 1.47<span style="color: #949494;">e</span><span style="color: #BB0000;">-125</span>   76.2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">15</span> RPL7A    3        14.9  3.03 12<span style="text-decoration: underline;">167</span>762. 0.692 2.55<span style="color: #949494;">e</span><span style="color: #BB0000;">-124</span> 2.93<span style="color: #949494;">e</span><span style="color: #BB0000;">-122</span>   89.7</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">16</span> RAN      3        10.9  3.41 11<span style="text-decoration: underline;">805</span>844. 0.671 1.96<span style="color: #949494;">e</span><span style="color: #BB0000;">-112</span> 1.94<span style="color: #949494;">e</span><span style="color: #BB0000;">-110</span>   66.5</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">17</span> RPL4     3        13.8  3.24 11<span style="text-decoration: underline;">883</span>184. 0.675 1.05<span style="color: #949494;">e</span><span style="color: #BB0000;">-106</span> 8.96<span style="color: #949494;">e</span><span style="color: #BB0000;">-105</span>   83.2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">18</span> NPM1     3        12.2  3.74 11<span style="text-decoration: underline;">778</span>232. 0.669 7.79<span style="color: #949494;">e</span><span style="color: #BB0000;">-106</span> 6.58<span style="color: #949494;">e</span><span style="color: #BB0000;">-104</span>   73.6</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">19</span> CREM     3        10.1  3.38 11<span style="text-decoration: underline;">163</span>252. 0.634 5.25<span style="color: #949494;">e</span><span style="color: #BB0000;">- 74</span> 3.00<span style="color: #949494;">e</span><span style="color: #BB0000;">- 72</span>   61.7</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">20</span> GLTSCR2  3        10.2  3.26 11<span style="text-decoration: underline;">106</span>918. 0.631 3.24<span style="color: #949494;">e</span><span style="color: #BB0000;">- 69</span> 1.71<span style="color: #949494;">e</span><span style="color: #BB0000;">- 67</span>   63.4</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1 more variable: pct_out &lt;dbl&gt;</span></span></span></code></pre>
<p>We can then visualize the expression profiles of individual genes,
such as the differentially expressed genes that we just identified. This
allows us to visually confirm the cluster- or dataset-specific
expression patterns of marker genes. <code><a href="../reference/plotDimRed.html">plotGeneDimRed()</a></code>
returns graphs of gene loading on the dimensionality reduced graph for
each dataset.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimRed.html">plotGeneDimRed</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, <span class="st">"PRF1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;"></p>
<p>We can also plot the gene expression by dataset.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prf1List</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDimRed.html">plotGeneDimRed</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, <span class="st">"PRF1"</span>, splitBy <span class="op">=</span> <span class="st">"dataset"</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">prf1List</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">prf1List</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-5-1.png" width="960" style="display: block; margin: auto;"></p>
<p>We can also use <code><a href="../reference/plotDimRed.html">plotGeneDimRed()</a></code> to compare the loading
of cluster markers within and between datasets.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">IFIT3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDimRed.html">plotGeneDimRed</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, <span class="st">"IFIT3"</span>, splitBy <span class="op">=</span> <span class="st">"dataset"</span><span class="op">)</span></span>
<span><span class="va">IFITM3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDimRed.html">plotGeneDimRed</a></span><span class="op">(</span><span class="va">pbmcLiger</span>, <span class="st">"IFITM3"</span>, splitBy <span class="op">=</span> <span class="st">"dataset"</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">IFIT3</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">IFIT3</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="va">IFITM3</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">IFITM3</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, ncol <span class="op">=</span> <span class="fl">2</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ctrl"</span>, <span class="st">"stim"</span>, <span class="st">"ctrl"</span>, <span class="st">"stim"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integrating_multi_scRNA_data_files/figure-html/unnamed-chunk-6-1.png" width="864" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="r-session-info">R Session Info<a class="anchor" aria-label="anchor" href="#r-session-info"></a>
</h2>
<pre><code><span><span class="co">## R version 4.3.2 RC (2023-10-30 r85440)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Sonoma 14.4.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/Detroit</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] dplyr_1.1.4  rliger_2.0.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] sass_0.4.9          utf8_1.2.4          generics_0.1.3     </span></span>
<span><span class="co">##  [4] lattice_0.22-6      digest_0.6.35       magrittr_2.0.3     </span></span>
<span><span class="co">##  [7] evaluate_0.23       grid_4.3.2          fastmap_1.1.1      </span></span>
<span><span class="co">## [10] jsonlite_1.8.8      Matrix_1.6-5        ggrepel_0.9.5      </span></span>
<span><span class="co">## [13] scattermore_1.2     purrr_1.0.2         fansi_1.0.6        </span></span>
<span><span class="co">## [16] viridisLite_0.4.2   scales_1.3.0        codetools_0.2-20   </span></span>
<span><span class="co">## [19] textshaping_0.3.7   jquerylib_0.1.4     cli_3.6.2          </span></span>
<span><span class="co">## [22] rlang_1.1.3         RcppAnnoy_0.0.22    uwot_0.1.16        </span></span>
<span><span class="co">## [25] cowplot_1.1.3       munsell_0.5.1       withr_3.0.0        </span></span>
<span><span class="co">## [28] RANN_2.6.1          cachem_1.0.8        yaml_2.3.8         </span></span>
<span><span class="co">## [31] tools_4.3.2         parallel_4.3.2      memoise_2.0.1      </span></span>
<span><span class="co">## [34] colorspace_2.1-0    ggplot2_3.5.0       sccore_1.0.5       </span></span>
<span><span class="co">## [37] BiocGenerics_0.48.1 vctrs_0.6.5         R6_2.5.1           </span></span>
<span><span class="co">## [40] stats4_4.3.2        lifecycle_1.0.4     S4Vectors_0.40.2   </span></span>
<span><span class="co">## [43] fs_1.6.3            irlba_2.3.5.1       ragg_1.3.0         </span></span>
<span><span class="co">## [46] pkgconfig_2.0.3     leidenAlg_1.1.3     desc_1.4.3         </span></span>
<span><span class="co">## [49] pkgdown_2.0.7       pillar_1.9.0        bslib_0.7.0        </span></span>
<span><span class="co">## [52] gtable_0.3.4        glue_1.7.0          Rcpp_1.0.12        </span></span>
<span><span class="co">## [55] systemfonts_1.0.6   highr_0.10          xfun_0.43          </span></span>
<span><span class="co">## [58] tibble_3.2.1        tidyselect_1.2.1    rstudioapi_0.16.0  </span></span>
<span><span class="co">## [61] knitr_1.45          farver_2.1.1        htmltools_0.5.8    </span></span>
<span><span class="co">## [64] igraph_2.0.3        labeling_0.4.3      rmarkdown_2.26     </span></span>
<span><span class="co">## [67] RcppPlanc_1.0.0     compiler_4.3.2</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
