<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rliger">
<title>Interact with A Liger Object • rliger</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Interact with A Liger Object">
<meta property="og:description" content="rliger">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Interact with A Liger Object</h1>
                        <h4 data-toc-skip class="author">Yichen
Wang</h4>
            
            <h4 data-toc-skip class="date">2023-11-06</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/vignettes/articles/liger_object.Rmd" class="external-link"><code>vignettes/articles/liger_object.Rmd</code></a></small>
      <div class="d-none name"><code>liger_object.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="structure">Structure<a class="anchor" aria-label="anchor" href="#structure"></a>
</h2>
<p>Starting from rliger 2.0.0, we introduced a newly designed structure
for the main data container object class. The figure below brings a
overall idea about the design.</p>
<p><img src="img/liger_class_structure.png" alt="liger_object_structure"><br></p>
<p>On the left hand side of the figure, we briefly illustrate how the
old structure was like. It has slots for data at each processing stage,
and each slot is a named list object containing matrices.</p>
<p>In the new version, on the right side, we first introduce another new
class, <a href="../reference/ligerDataset-class.html">ligerDataset</a>,
to serve as a container for all matrices belonging to the same specific
dataset. In this way, we have easy and safe control on data matching
thanks to a number of object oriented accessor methods and validity
checks. As for the <a href="../reference/liger-class.html">liger</a>
class, there are some main differences:</p>
<ul>
<li>Feature (gene expression) matrices are now wrapped in ligerDataset
objects, and all put in the <code>datasets(obj)</code> slot. Click here
for how to <a href="#access-feature-matrices">access the raw counts,
normalized counts, scaled data</a>.</li>
<li>Cell metadata including dataset belonging, study design conditions,
quality control (QC) metrics are now stored in slot
<code>cellMeta(obj)</code>. Meanwhile, we moved cluster labeling result
also to this slot, in order to have multiple cluster variables existing
at the same time. Additionally, by introducing
<code><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">S4Vectors::DataFrame</a></code> class for the metadata for flexibility
and tidy display. Click here for how to <a href="#access-cell-metadata">access cell metadata</a>.</li>
<li>Dimensionality reductions are now expanded to a list of
low-dimensional representation (e.g. UMAP, t-SNE) in slot
<code>dimReds(obj)</code>. Click here for how to <a href="#access-dimensionality-reductions">access dimensionality
reductions</a>.</li>
<li>Variable features identified are now stored in
<code>varFeatures(obj)</code>, with no structural change.</li>
<li>A new slot <code>@uns</code> is added for storing miscellaneous
unstructured information, including default setting for cluster labeling
and dimension reduction, for faster visualization calls.</li>
<li>We also added the feature to record the commands applied to the
liger object and allow retrieving the records with
<code>commands(obj)</code>. See more about liger <a href="#check-the-records-of-run-commands">command recording and
retrieving</a>.</li>
</ul>
<p>We demonstrate examples below with example dataset “<a href="../reference/pbmc.html">pbmc</a>”, which is a minimal subset from
the study conducted in Hyun Min Kang and et. al., Nature Biotechnology,
2018. The data is a ready to use (new) liger object with only raw counts
data. We quickly process it here so that we can show how to retrieve all
kinds of data in sections below.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://welch-lab.github.io/liger/" class="external-link">rliger</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pbmc"</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="va">pbmc</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/runIntegration.html">runIntegration</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/quantileNorm.html">quantileNorm</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/runCluster.html">runCluster</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="access-a-dataset">Access a dataset<a class="anchor" aria-label="anchor" href="#access-a-dataset"></a>
</h2>
<p>As introduced above, the dataset-specific information is contained in
a <a href="../reference/ligerDataset-class.html">ligerDataset</a>
object.</p>
<ul>
<li>To get the names of all datasets</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ctrl" "stim"</span></span></code></pre>
<ul>
<li>To get the number of all datasets</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 2</span></span></code></pre>
<ul>
<li>To list out number of cells in each dataset</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ctrl stim </span></span>
<span><span class="co">##  300  300</span></span></code></pre>
<ul>
<li>To access the <a href="../reference/ligerDataset-class.html">ligerDataset</a> object for
a specific dataset</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctrlLD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-class.html">dataset</a></span><span class="op">(</span><span class="va">pbmc</span>, dataset <span class="op">=</span> <span class="st">"ctrl"</span><span class="op">)</span></span>
<span><span class="co"># Alternatively, using numeric index</span></span>
<span><span class="va">ctrlLD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-class.html">dataset</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>In any other <em>rliger</em> functions where the argument
<code>useDatasets</code> is exposed, users can always use the exact
character name(s) or the numeric index to specify the datasets to be
involved in the analysis. Moreoever, a logical vector of index is also
allowed and could ease the usage in some cases.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Not run, just for example, assuming we've got the clustering for such an object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ligerObj</span><span class="op">)</span></span>
<span><span class="co">## [1] female-1  female-2  male-3  male-4  female-5 ......</span></span>
<span><span class="va">femaleIdx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/startsWith.html" class="external-link">startsWith</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ligerObj</span><span class="op">)</span>, <span class="st">"fe"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/liger-DEG.html">runMarkerDEG</a></span><span class="op">(</span><span class="va">ligerObj</span>, conditionBy <span class="op">=</span> <span class="st">"dataset"</span>, splitBy <span class="op">=</span> <span class="st">"leiden_cluster"</span>, </span>
<span>             useDatasets <span class="op">=</span> <span class="va">femaleIdx</span><span class="op">)</span></span></code></pre></div>
<p>In the example above, the <code><a href="../reference/liger-DEG.html">runMarkerDEG()</a></code> funcion is
parametered for detecting dataset specific markers within each cluster,
and only within the female samples. For example, cells from condition
“female-1 and cluster 1” will be tested against cells belonging to
condition “cluster 1 and all other female datasets”. Can be use</p>
<ul>
<li>To access multiple datasets, returned in a list</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ldList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-class.html">datasets</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="access-feature-matrices">Access feature matrices<a class="anchor" aria-label="anchor" href="#access-feature-matrices"></a>
</h2>
<p>We have three main generics for accessing feature matrices, namingly
<code><a href="../reference/ligerDataset-class.html">rawData()</a></code>, <code><a href="../reference/ligerDataset-class.html">normData()</a></code> and
<code><a href="../reference/ligerDataset-class.html">scaleData()</a></code>. For scaled unshared features, used for UINMF,
we also have <code><a href="../reference/ligerDataset-class.html">scaleUnsharedData()</a></code>. Additionally, we provide
<code><a href="../reference/peak.html">rawPeak()</a></code> and <code><a href="../reference/peak.html">normPeak()</a></code> for accessing the
peak counts in a ATACseq dataset. The logistics of the accessor to all
these feature matrices are the same, so we only present the case for raw
counts.</p>
<ul>
<li>To get a list of the raw counts from all datasets:</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rawList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ligerDataset-class.html">rawData</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">rawList</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">rawList</span><span class="op">)</span></span>
<span><span class="co"># Alternatively</span></span>
<span><span class="va">rawList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ligerDataset-class.html">getMatrix</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"rawData"</span>, returnList <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "list"</span></span>
<span><span class="co">## [1] 2</span></span></code></pre>
<ul>
<li>To get the raw counts from a specific dataset:</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stimRaw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ligerDataset-class.html">rawData</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"stim"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">stimRaw</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">stimRaw</span><span class="op">)</span></span>
<span><span class="co"># Alternatively, get the `ligerDataset` object of the dataset first</span></span>
<span><span class="co"># and then fetch from there</span></span>
<span><span class="va">stim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-class.html">dataset</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"stim"</span><span class="op">)</span></span>
<span><span class="va">stimRaw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ligerDataset-class.html">rawData</a></span><span class="op">(</span><span class="va">stim</span><span class="op">)</span></span>
<span><span class="co"># Alternatively</span></span>
<span><span class="va">stimRaw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ligerDataset-class.html">getMatrix</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"rawData"</span>, dataset <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "dgCMatrix"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "Matrix"</span></span>
<span><span class="co">## [1] 262 300</span></span></code></pre>
<ul>
<li>To replace the raw counts with a new matrix object:</li>
</ul>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctrlRaw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ligerDataset-class.html">rawData</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"ctrl"</span><span class="op">)</span></span>
<span><span class="co"># Assume we do some operation on it here</span></span>
<span><span class="fu"><a href="../reference/ligerDataset-class.html">rawData</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"ctrl"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">ctrlRaw</span></span></code></pre></div>
<p>In the new version, strict validity checks have been put upon
modification in object content. Replacement with unmatching feature
names or barcodes will be rejected. In the case where there is a need to
replace the dataset with a different set of barcodes or features, we
suggest recreating a new <a href="../reference/ligerDataset-class.html">ligerDataset</a> object with
the new raw counts (or other feature matrix), and then replace the whole
dataset with it.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctrlRaw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ligerDataset-class.html">rawData</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"ctrl"</span><span class="op">)</span></span>
<span><span class="va">ctrlRawSubset</span> <span class="op">&lt;-</span> <span class="va">ctrlRaw</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">200</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">200</span><span class="op">]</span></span>
<span><span class="co">## Not Run, will raise error</span></span>
<span><span class="co"># rawData(pbmc, "ctrl") &lt;- ctrlRawSubset</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctrlNew</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLigerDataset.html">createLigerDataset</a></span><span class="op">(</span>rawData <span class="op">=</span> <span class="va">ctrlRawSubset</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/liger-class.html">dataset</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"ctrl"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">ctrlNew</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="co">## [1] NA 500</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="access-cell-metadata">Access cell metadata<a class="anchor" aria-label="anchor" href="#access-cell-metadata"></a>
</h2>
<p>As previously descibed at the top of this page, cell metadata
including dataset origin, study metadata, QC metrics and cluster
labeling are all stored in <code>cellMeta(obj)</code>.</p>
<ul>
<li>To have a look at the full metadata table:</li>
</ul>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/liger-class.html">cellMeta</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 600 rows and 9 columns</span></span>
<span><span class="co">##                        dataset               barcode      nUMI     nGene</span></span>
<span><span class="co">##                       &lt;factor&gt;           &lt;character&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## ctrl_AAACATACCTCGCT.1     ctrl ctrl_AAACATACCTCGCT.1      2151       102</span></span>
<span><span class="co">## ctrl_AAACGGCTCTTCGC.1     ctrl ctrl_AAACGGCTCTTCGC.1      1916       103</span></span>
<span><span class="co">## ctrl_AACACTCTAAGTAG.1     ctrl ctrl_AACACTCTAAGTAG.1      1869        95</span></span>
<span><span class="co">## ctrl_AACCGCCTCAGGAG.1     ctrl ctrl_AACCGCCTCAGGAG.1       733        84</span></span>
<span><span class="co">## ctrl_AACGTTCTTCCGTC.1     ctrl ctrl_AACGTTCTTCCGTC.1       573        76</span></span>
<span><span class="co">## ...                        ...                   ...       ...       ...</span></span>
<span><span class="co">## stim_TTCATGACTTATCC.1     stim stim_TTCATGACTTATCC.1      1091       101</span></span>
<span><span class="co">## stim_TTCATGACTTCAGG.1     stim stim_TTCATGACTTCAGG.1       431        86</span></span>
<span><span class="co">## stim_TTCGGAGATTTCAC.1     stim stim_TTCGGAGATTTCAC.1      1344       102</span></span>
<span><span class="co">## stim_TTGACACTTCCTGC.1     stim stim_TTGACACTTCCTGC.1      1423       105</span></span>
<span><span class="co">## stim_TTTGCATGAACGAA.1     stim stim_TTTGCATGAACGAA.1      2818       120</span></span>
<span><span class="co">##                            mito      ribo      hemo quantileNorm_cluster</span></span>
<span><span class="co">##                       &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;             &lt;factor&gt;</span></span>
<span><span class="co">## ctrl_AAACATACCTCGCT.1         0  10.83217         0                   12</span></span>
<span><span class="co">## ctrl_AAACGGCTCTTCGC.1         0  20.25052         0                   5 </span></span>
<span><span class="co">## ctrl_AACACTCTAAGTAG.1         0   5.08293         0                   3 </span></span>
<span><span class="co">## ctrl_AACCGCCTCAGGAG.1         0  42.70123         0                   2 </span></span>
<span><span class="co">## ctrl_AACGTTCTTCCGTC.1         0  38.04538         0                   19</span></span>
<span><span class="co">## ...                         ...       ...       ...                  ...</span></span>
<span><span class="co">## stim_TTCATGACTTATCC.1         0   8.61595         0                   11</span></span>
<span><span class="co">## stim_TTCATGACTTCAGG.1         0  41.29930         0                   14</span></span>
<span><span class="co">## stim_TTCGGAGATTTCAC.1         0   8.85417         0                   6 </span></span>
<span><span class="co">## stim_TTGACACTTCCTGC.1         0  11.94659         0                   6 </span></span>
<span><span class="co">## stim_TTTGCATGAACGAA.1         0   9.43932         0                   5 </span></span>
<span><span class="co">##                       leiden_cluster</span></span>
<span><span class="co">##                             &lt;factor&gt;</span></span>
<span><span class="co">## ctrl_AAACATACCTCGCT.1              0</span></span>
<span><span class="co">## ctrl_AAACGGCTCTTCGC.1              0</span></span>
<span><span class="co">## ctrl_AACACTCTAAGTAG.1              0</span></span>
<span><span class="co">## ctrl_AACCGCCTCAGGAG.1              3</span></span>
<span><span class="co">## ctrl_AACGTTCTTCCGTC.1              1</span></span>
<span><span class="co">## ...                              ...</span></span>
<span><span class="co">## stim_TTCATGACTTATCC.1              0</span></span>
<span><span class="co">## stim_TTCATGACTTCAGG.1              1</span></span>
<span><span class="co">## stim_TTCGGAGATTTCAC.1              7</span></span>
<span><span class="co">## stim_TTGACACTTCCTGC.1              7</span></span>
<span><span class="co">## stim_TTTGCATGAACGAA.1              6</span></span></code></pre>
<ul>
<li>To retrieve one of the variables:</li>
</ul>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nUMI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-class.html">cellMeta</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"nUMI"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">nUMI</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">nUMI</span><span class="op">)</span></span>
<span><span class="co"># Alternatively</span></span>
<span><span class="va">nUMI</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">nUMI</span></span>
<span><span class="va">nUMI</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">[[</span><span class="st">"nUMI"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "numeric"</span></span>
<span><span class="co">## [1] 600</span></span></code></pre>
<ul>
<li>To retrieve multiple variables:</li>
</ul>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/liger-class.html">cellMeta</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nUMI"</span>, <span class="st">"nGene"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 600 rows and 2 columns</span></span>
<span><span class="co">##                            nUMI     nGene</span></span>
<span><span class="co">##                       &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## ctrl_AAACATACCTCGCT.1      2151       102</span></span>
<span><span class="co">## ctrl_AAACGGCTCTTCGC.1      1916       103</span></span>
<span><span class="co">## ctrl_AACACTCTAAGTAG.1      1869        95</span></span>
<span><span class="co">## ctrl_AACCGCCTCAGGAG.1       733        84</span></span>
<span><span class="co">## ctrl_AACGTTCTTCCGTC.1       573        76</span></span>
<span><span class="co">## ...                         ...       ...</span></span>
<span><span class="co">## stim_TTCATGACTTATCC.1      1091       101</span></span>
<span><span class="co">## stim_TTCATGACTTCAGG.1       431        86</span></span>
<span><span class="co">## stim_TTCGGAGATTTCAC.1      1344       102</span></span>
<span><span class="co">## stim_TTGACACTTCCTGC.1      1423       105</span></span>
<span><span class="co">## stim_TTTGCATGAACGAA.1      2818       120</span></span></code></pre>
<ul>
<li>To get a variable for only a subset of cells (e.g. from a specific
dataset), use the argument <code>useDatasets</code>, or alternatively
<code>cellIdx</code> that subscribes cells explicitly:</li>
</ul>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nUMI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-class.html">cellMeta</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"nUMI"</span>, useDatasets <span class="op">=</span> <span class="st">"ctrl"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">nUMI</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 300</span></span></code></pre>
<ul>
<li>Add or replace a variable. If the new variable being added has a
matching size (<code><a href="https://rdrr.io/r/base/length.html" class="external-link">length()</a></code> or <code><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow()</a></code>) with the
number of all cells (<code>ncol(pbmc)</code>):</li>
</ul>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">foo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pbmc</span><span class="op">$</span><span class="va">foo</span> <span class="op">&lt;-</span> <span class="va">foo</span></span>
<span><span class="co"># Alternatively</span></span>
<span><span class="fu"><a href="../reference/liger-class.html">cellMeta</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"foo"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">foo</span></span></code></pre></div>
<p>If the new variable is only for a subset of cells (e.g. the original
clustering derived from an individual dataset).</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctrlBar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="../reference/liger-class.html">dataset</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"ctrl"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/liger-class.html">cellMeta</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"ctrl_bar"</span>, useDataset <span class="op">=</span> <span class="st">"ctrl"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">ctrlBar</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="access-dimensionality-reductions">Access dimensionality reductions<a class="anchor" aria-label="anchor" href="#access-dimensionality-reductions"></a>
</h2>
<ul>
<li>To get a list of all dimensionality reductions:</li>
</ul>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">allDimReds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-class.html">dimReds</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">allDimReds</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">allDimReds</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "list"</span></span>
<span><span class="co">## [1] 1</span></span></code></pre>
<ul>
<li>To get a specific dimensionality reduction:</li>
</ul>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-class.html">dimRed</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"UMAP"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">umap</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">umap</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "matrix" "array" </span></span>
<span><span class="co">## [1] 600   2</span></span></code></pre>
<ul>
<li>To get a specific dimensionality reduction for a specific dataset,
again, use the argument <code>useDatasets</code>.</li>
</ul>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctrlUMAP</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-class.html">dimRed</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"UMAP"</span>, useDatasets <span class="op">=</span> <span class="st">"ctrl"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">ctrlUMAP</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 300   2</span></span></code></pre>
<ul>
<li>Setting an existing dimensionality reduction as the default for
visualization</li>
</ul>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/liger-class.html">defaultDimRed</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"UMAP"</span></span></code></pre></div>
<p>Every time when <code><a href="../reference/runUMAP.html">runUMAP()</a></code> or <code><a href="../reference/runTSNE.html">runTSNE()</a></code> is
called, the new result will be set as default. When default
dimensionality reduction is set, any plotting function that shall work
with it will use it by default without the need to specify it
explicitly. And the dimensionality reduction accessor function also
returns the default one if no specific one is requested.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-class.html">dimRed</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>Add a new matrix into the object</li>
</ul>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/liger-class.html">dimRed</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"newUMAP"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">umap</span></span></code></pre></div>
<p>Cell identifiers on <code>rownames(value)</code> will be checked for
matching is present. The check is aware of that a dataset name prefix is
added to the object cell IDs.</p>
<ul>
<li>Adding a dimensionality reduction matrix for only one certain
dataset</li>
</ul>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctrlUMAP</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-class.html">dimRed</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"UMAP"</span>, useDatasets <span class="op">=</span> <span class="st">"ctrl"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/liger-class.html">dimRed</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"ctrlUMAP"</span>, useDatasets <span class="op">=</span> <span class="st">"ctrl"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">ctrlUMAP</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="access-factorization-result">Access factorization result<a class="anchor" aria-label="anchor" href="#access-factorization-result"></a>
</h2>
<p>We suggest using <code><a href="../reference/ligerDataset-class.html">getMatrix()</a></code> for all matrices involved
in the factorization, including:</p>
<ul>
<li>
<span class="math inline">\(H\)</span>, <span class="math inline">\(V\)</span> matrices produced for each dataset,
involved in all iNMF variant algorithms</li>
<li>
<span class="math inline">\(W\)</span> matrix shared for all
datasets, involved in all iNMF variant algorithms and NMF dimension
reduction</li>
<li>
<span class="math inline">\(H.norm\)</span>, the aligned factor
loading matrix, produced downstream of iNMF integration by
<code><a href="../reference/quantileNorm.html">quantileNorm()</a></code>
</li>
<li>
<span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span> matrices produced as intermediate
information for each dataset during online iNMF interations</li>
<li>
<span class="math inline">\(U\)</span> matrices produced for each
dataset, involved in UINMF</li>
</ul>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ligerDataset-class.html">getMatrix</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"H"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">HList</span>, <span class="va">dim</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $ctrl</span></span>
<span><span class="co">## [1]  20 300</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $stim</span></span>
<span><span class="co">## [1]  20 300</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="subsetting-the-data">Subsetting the data<a class="anchor" aria-label="anchor" href="#subsetting-the-data"></a>
</h2>
<ul>
<li>A <a href="../reference/liger-class.html">liger</a> object can be
subset by both cells and genes.</li>
</ul>
<p>For cell level subsetting, any indexing method among barcode names,
numeric or logical index can do the job. Cells are indexed by
<code>rownames(cellMeta(object))</code>, which is a concatenation of the
barcodes from each dataset, and datasets are ordered as
<code>names(object)</code> shows.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcSmall</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Subsetting dataset: <span style="color: #0000BB;">"ctrl"</span></span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> Subsetting dataset: <span style="color: #0000BB;">"ctrl"</span> ... done</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcCluster1</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">[</span>, <span class="va">pbmc</span><span class="op">$</span><span class="va">leiden_cluster</span> <span class="op">==</span> <span class="fl">1</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Subsetting dataset: <span style="color: #0000BB;">"ctrl"</span></span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Subsetting dataset: <span style="color: #0000BB;">"stim"</span></span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> Subsetting dataset: <span style="color: #0000BB;">"stim"</span> ... done</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Subsetting dataset: <span style="color: #0000BB;">"ctrl"</span><span style="color: #00BB00;">✔</span> Subsetting dataset: <span style="color: #0000BB;">"ctrl"</span> ... done</span></span></code></pre>
<p>For gene level subsetting, we only allow using gene names, because it
is assumed that different datasets can have different set of genes. And
only genes shared by all datasets can be used.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcVarOnly</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">[</span><span class="fu"><a href="../reference/liger-class.html">varFeatures</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Subsetting dataset: <span style="color: #0000BB;">"ctrl"</span></span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Subsetting dataset: <span style="color: #0000BB;">"stim"</span></span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> Subsetting dataset: <span style="color: #0000BB;">"stim"</span> ... done</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Subsetting dataset: <span style="color: #0000BB;">"ctrl"</span><span style="color: #00BB00;">✔</span> Subsetting dataset: <span style="color: #0000BB;">"ctrl"</span> ... done</span></span></code></pre>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctrlUnsharedGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P2RY1"</span>, <span class="st">"GFI1B"</span>, <span class="st">"HDGFRP2"</span>, <span class="st">"TUBGCP6"</span>, <span class="st">"CELA1"</span><span class="op">)</span></span>
<span><span class="co"># Not run, will raise error</span></span>
<span><span class="co"># pbmc[ctrl.unshared.genes,]</span></span></code></pre></div>
<ul>
<li>A <a href="../reference/liger-class.html">ligerDataset object</a>
can be subset by both cells and genes.</li>
</ul>
<p>Cell level subsetting works in the exactly same way as a <a href="../reference/liger-class">liger object</a>.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctrlLD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-class.html">dataset</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"ctrl"</span><span class="op">)</span></span>
<span><span class="va">ctrlLDSmall</span> <span class="op">&lt;-</span> <span class="va">ctrlLD</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span></code></pre></div>
<p>Gene level subsetting on a <a href="../reference/liger-class.html">ligerDataset object</a> can
achieved with any type of index.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctrlLDsmall</span> <span class="op">&lt;-</span> <span class="va">ctrlLD</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="op">]</span></span>
<span><span class="va">ctrlLDsmall</span> <span class="op">&lt;-</span> <span class="va">ctrlLD</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span></code></pre></div>
<p>Note that, <code>scaleData(ctrlLD)</code> and
<code>scaleUnsharedData(ctrlLD)</code> comes with only variable genes
identified upstream. Subsetting genes on a <a href="../reference/liger-class.html">ligerDataset object</a> is based on
its raw input data. Therefore, we only take the user specification
available in scaled data into the subset of scaled data.</p>
</div>
<div class="section level2">
<h2 id="check-the-records-of-run-commands">Check the records of run commands<a class="anchor" aria-label="anchor" href="#check-the-records-of-run-commands"></a>
</h2>
<p>We implemented a analysis tracking feature in order to keep a record
of what functions are called and what parameters are used.</p>
<ul>
<li>To show a list of function names applied to the <a href="../reference/liger-class.html">liger object</a> in time order</li>
</ul>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/liger-class.html">commands</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "normalize.liger_9e0805533b"      "selectGenes.liger_1b8286227c"   </span></span>
<span><span class="co">## [3] "scaleNotCenter.liger_0d016f7878" "runINMF.liger_b8a640ceac"       </span></span>
<span><span class="co">## [5] "quantileNorm.liger_d46431a24f"   "runCluster_d820badc15"          </span></span>
<span><span class="co">## [7] "runUMAP_94564eef9a"</span></span></code></pre>
<p>A unique suffix is added to each function name to keep track of calls
of the same function with different parameters.</p>
<ul>
<li>Detailed a function call information can be retrieved with partial
matching.</li>
</ul>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/liger-class.html">commands</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"runINMF"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## A liger command record, performed at 04-04-2024 21:18:19 EDT </span></span>
<span><span class="co">## Call: runINMF.liger(object, k = k, lambda = lambda, seed = seed, verbose = verbose,      ...) </span></span>
<span><span class="co">## Parameters:</span></span>
<span><span class="co">##     k : 20 </span></span>
<span><span class="co">##     lambda : 5 </span></span>
<span><span class="co">##     nIteration : 30 </span></span>
<span><span class="co">##     nRandomStarts : 1 </span></span>
<span><span class="co">##     HInit : NULL </span></span>
<span><span class="co">##     WInit : NULL </span></span>
<span><span class="co">##     VInit : NULL </span></span>
<span><span class="co">##     seed : 1 </span></span>
<span><span class="co">##     nCores : 2 </span></span>
<span><span class="co">##     verbose : TRUE</span></span></code></pre>
<p>A function can be applied to an object several times with parameter
tweaks. For example, different <code>lambda</code> for iNMF integration.
If <code><a href="../reference/runINMF.html">runINMF()</a></code> is called several times, calling
<code>commands(pbmc, "runINMF")</code> returns a list of records of all
such calls, as all record names starting with <code>"runINMF"</code> are
matched. So listing names first and using the unique record name will be
required for getting the information of one specific call among all of
such. For another example, given that <code><a href="../reference/runCluster.html">runCluster()</a></code> and
<code><a href="../reference/runUMAP.html">runUMAP()</a></code> are also in the record, the following result
would be returned if we do matching with only <code>"run"</code></p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/liger-class.html">commands</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"run"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $runINMF.liger_b8a640ceac</span></span>
<span><span class="co">## A liger command record, performed at 04-04-2024 21:18:19 EDT </span></span>
<span><span class="co">## Call: runINMF.liger(object, k = k, lambda = lambda, seed = seed, verbose = verbose,      ...) </span></span>
<span><span class="co">## Parameters:</span></span>
<span><span class="co">##     k : 20 </span></span>
<span><span class="co">##     lambda : 5 </span></span>
<span><span class="co">##     nIteration : 30 </span></span>
<span><span class="co">##     nRandomStarts : 1 </span></span>
<span><span class="co">##     HInit : NULL </span></span>
<span><span class="co">##     WInit : NULL </span></span>
<span><span class="co">##     VInit : NULL </span></span>
<span><span class="co">##     seed : 1 </span></span>
<span><span class="co">##     nCores : 2 </span></span>
<span><span class="co">##     verbose : TRUE </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $runCluster_d820badc15</span></span>
<span><span class="co">## A liger command record, performed at 04-04-2024 21:18:19 EDT </span></span>
<span><span class="co">## Call: runCluster(.) </span></span>
<span><span class="co">## Parameters:</span></span>
<span><span class="co">##     resolution : 1 </span></span>
<span><span class="co">##     nNeighbors : 20 </span></span>
<span><span class="co">##     prune : 0.0666666666666667 </span></span>
<span><span class="co">##     eps : 0.1 </span></span>
<span><span class="co">##     nRandomStarts : 10 </span></span>
<span><span class="co">##     nIterations : 5 </span></span>
<span><span class="co">##     method : "leiden" </span></span>
<span><span class="co">##     useRaw : NULL </span></span>
<span><span class="co">##     useDims : NULL </span></span>
<span><span class="co">##     groupSingletons : TRUE </span></span>
<span><span class="co">##     saveSNN : FALSE </span></span>
<span><span class="co">##     clusterName : "leiden_cluster" </span></span>
<span><span class="co">##     seed : 1 </span></span>
<span><span class="co">##     verbose : TRUE </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $runUMAP_94564eef9a</span></span>
<span><span class="co">## A liger command record, performed at 04-04-2024 21:18:19 EDT </span></span>
<span><span class="co">## Call: runUMAP(.) </span></span>
<span><span class="co">## Parameters:</span></span>
<span><span class="co">##     useRaw : NULL </span></span>
<span><span class="co">##     useDims : NULL </span></span>
<span><span class="co">##     nDims : 2 </span></span>
<span><span class="co">##     distance : "cosine" </span></span>
<span><span class="co">##     nNeighbors : 20 </span></span>
<span><span class="co">##     minDist : 0.1 </span></span>
<span><span class="co">##     dimredName : "UMAP" </span></span>
<span><span class="co">##     seed : 42 </span></span>
<span><span class="co">##     verbose : TRUE</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
