<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rliger">
<title>Using LIGER to integrate datasets stored in Seurat objects • rliger</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.8/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Using LIGER to integrate datasets stored in Seurat objects">
<meta property="og:description" content="rliger">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Using LIGER to integrate datasets stored in Seurat objects</h1>
                        <h4 data-toc-skip class="author">Yichen
Wang</h4>
            
            <h4 data-toc-skip class="date">2024-03-11</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/vignettes/articles/liger_with_seurat.Rmd" class="external-link"><code>vignettes/articles/liger_with_seurat.Rmd</code></a></small>
      <div class="d-none name"><code>liger_with_seurat.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="goal-of-this-article">Goal of this article<a class="anchor" aria-label="anchor" href="#goal-of-this-article"></a>
</h2>
<p>We have introduced the basic usage of LIGER throughout many other
vignettes, together with various use cases. In this article, we bring
LIGER to the Seurat users, showing how to run rliger functions on Seurat
objects to integrate the datasets.</p>
<p>With old version (&lt; 1.99.0), we provided wrapper functions in <a href="https://github.com/satijalab/seurat-wrappers" class="external-link">SeuratWrappers</a>
that allow rliger functions to be run on Seurat objects. With the
current new version (&gt;= 1.99.0), we now provide native support for
Seurat objects in rliger. The functions applied to a <a href="../reference/liger-class.html">liger</a> object can now be
directly applied to a Seurat object.</p>
<p>This article will demonstrate the workflow from loading a standard
Seurat data to obtaining the integration labeling.</p>
</div>
<div class="section level2">
<h2 id="example-datasets-with-v5-style">Example datasets, with V5 style<a class="anchor" aria-label="anchor" href="#example-datasets-with-v5-style"></a>
</h2>
<p>This vignette uses the example shown on <a href="https://satijalab.org/seurat/articles/integration_introduction.html" class="external-link">Seurat
official integration tutorial</a>. A package called <a href="https://github.com/satijalab/seurat-data" class="external-link">SeuratData</a> would be
needed for loading the example datasets.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.satijalab.org/seurat" class="external-link">SeuratData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/welch-lab/liger" class="external-link">rliger</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/InstallData.html" class="external-link">InstallData</a></span><span class="op">(</span><span class="st">"ifnb"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/LoadData.html" class="external-link">LoadData</a></span><span class="op">(</span><span class="st">"ifnb"</span><span class="op">)</span></span>
<span><span class="va">ifnb</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 14053 features across 13999 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (14053 features, 0 variable features)</span></span>
<span><span class="co">##  2 layers present: counts, data</span></span></code></pre>
<p>With the release of Seurat v5, it is now recommended to have the gene
expression data, namingly “counts”, “data” and “scale.data” slots
previously in a Seurat Assay, splitted by batches. A metadata variable
is required to be presented that indicates the batch information. In the
example dataset, the batch information is stored in the
<code>stim</code> variable. This would be the analogous to the
<code>ligerObj$dataset</code> variable with a liger object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>, f <span class="op">=</span> <span class="va">ifnb</span><span class="op">$</span><span class="va">stim</span><span class="op">)</span></span>
<span><span class="va">ifnb</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 14053 features across 13999 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (14053 features, 0 variable features)</span></span>
<span><span class="co">##  4 layers present: counts.CTRL, counts.STIM, data.CTRL, data.STIM</span></span></code></pre>
<p>From the printed output, we can see that the the expression data of
the <code>ifnb</code> object now is presented in layers by
conditions.</p>
</div>
<div class="section level2">
<h2 id="integration-with-liger">Integration with LIGER<a class="anchor" aria-label="anchor" href="#integration-with-liger"></a>
</h2>
<div class="section level3">
<h3 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h3>
<p>The very first step is to go through the QC and remove non-expressing
cells and genes. LIGER’s primary strategy for integration is to perform
integrative non-negative matrix factorization (iNMF). Due to the nature
of the method, it is important that all-zero rows or columns are removed
from the input data. Other QC criteria can also be optionally added,
such as minimum number of genes expressed, minimum number of UMIs,
mitochondrial gene percentage and etc. The QC techniques are simple and
similar across all kinds of toolkits and platforms, so we will not go
into detail here. Given that the metrics are already calculated when
creating the Seurat object, here we directly go with Seurat suggested
syntax to obtain the subset.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">ifnb</span>, subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">200</span> <span class="op">&amp;</span> <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">2500</span><span class="op">)</span></span></code></pre></div>
<p>We next perform preprocessing steps specific to LIGER:</p>
<ol style="list-style-type: decimal">
<li>library size normalization, without CPM or log1p transformation</li>
<li>Variable gene selection, done individually for each dataset and then
combined</li>
<li>Scaling the normalized expression of variable genes, without
centering to address the non-negative requirement of iNMF</li>
</ol>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="va">ifnb</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">ifnb</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 14053 features across 13997 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (14053 features, 4233 variable features)</span></span>
<span><span class="co">##  8 layers present: counts.CTRL, counts.STIM, data.CTRL, data.STIM, ligerNormData.CTRL, ligerNormData.STIM, ligerScaleData.CTRL, ligerScaleData.STIM</span></span></code></pre>
<p>Note that new layers created for the normalized and scaled data are
named different to the Seurat general convention. Instead of
<code>"data"</code> and <code>"scale.data"</code> as you would see in a
Seurat object, we name them <code>"ligerNormData"</code> and
<code>"ligerScaleData"</code>, respectively, in order to avoid data
processed with LIGER-specific approach being misused in general scRNAseq
analysis, or other Seurat general production being misused for LIGER
integration.</p>
</div>
<div class="section level3">
<h3 id="integration">Integration<a class="anchor" aria-label="anchor" href="#integration"></a>
</h3>
<p>Now, we are ready for performing iNMF integration on the two
datasets.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="va">ifnb</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/runINMF.html">runINMF</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/quantileNorm.html">quantileNorm</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">ifnb</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 14053 features across 13997 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (14053 features, 4233 variable features)</span></span>
<span><span class="co">##  8 layers present: counts.CTRL, counts.STIM, data.CTRL, data.STIM, ligerNormData.CTRL, ligerNormData.STIM, ligerScaleData.CTRL, ligerScaleData.STIM</span></span>
<span><span class="co">##  2 dimensional reductions calculated: inmf, inmfNorm</span></span></code></pre>
<p>As the object information shown above, <code><a href="../reference/runINMF.html">runINMF()</a></code>
produces two low-dimensional representations, which can be accessed with
<code>ifnb[["inmf"]]</code> and <code>ifnb[["inmfNorm"]]</code>. We call
them cell factor loading matrices. The first one is the raw
factorization result from iNMF, and the second one is the quantile
normalized version of the first one. The quantile normalization is
performed for suppressing the variance between the cell datasets so that
the datasets can be well aligned.</p>
</div>
<div class="section level3">
<h3 id="visualizing-the-integration-result">Visualizing the integration result<a class="anchor" aria-label="anchor" href="#visualizing-the-integration-result"></a>
</h3>
<p>Given that the example data provided by Seurat already has annotation
of cell types, we can now examine visually how the integration result
looks like. We will create a UMAP embedding using the cell factor
loading, and then make scatter plot colored by dataset source and the
cell type annotation.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">ifnb</span>, reduction <span class="op">=</span> <span class="st">"inmfNorm"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">gg.byDataset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, group.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span>
<span><span class="va">gg.byCelltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, group.by <span class="op">=</span> <span class="st">"seurat_annotations"</span><span class="op">)</span></span>
<span><span class="va">gg.byDataset</span> <span class="op">+</span> <span class="va">gg.byCelltype</span></span></code></pre></div>
<p><img src="liger_with_seurat_files/figure-html/umap-1.png" width="1152"></p>
<p>To visualize the two conditions side-by-side, we can use the
<code>split.by</code> argument to show each condition colored by
cluster.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, group.by <span class="op">=</span> <span class="st">"seurat_annotations"</span>, split.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span></code></pre></div>
<p><img src="liger_with_seurat_files/figure-html/splitumap-1.png" width="1152"></p>
</div>
<div class="section level3">
<h3 id="joint-clustering">Joint clustering<a class="anchor" aria-label="anchor" href="#joint-clustering"></a>
</h3>
<p>Despite of the annotation provided above, most of the time, people
would need to cluster the cells basing on the integration and then work
on the annotation. Here we demonstrate how to use Seurat functions to
create cluster labels basing on the cell factor loading obtained with
<code><a href="../reference/runINMF.html">runINMF()</a></code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="va">ifnb</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"inmfNorm"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 13997</span></span>
<span><span class="co">## Number of edges: 475772</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.8865</span></span>
<span><span class="co">## Number of communities: 17</span></span>
<span><span class="co">## Elapsed time: 1 seconds</span></span></code></pre>
<p>Again, we can also visualize the clustering label on the UMAP
previously created.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, group.by <span class="op">=</span> <span class="st">"seurat_clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="liger_with_seurat_files/figure-html/clusterumap-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="class-conversion-in-various-use-cases">Class conversion in various use cases<a class="anchor" aria-label="anchor" href="#class-conversion-in-various-use-cases"></a>
</h2>
<p>Examples above is based on the latest Seurat architecture, where
layers of data can be split by dataset source, which makes it
computationally efficient for all kinds of integration tool, not only
LIGER but also other methods introduced in <a href="https://satijalab.org/seurat/articles/integration_introduction.html" class="external-link">Seurat
official integration vignette</a>. We highly recommend that users follow
up with the syntax. However, if you are working on existing projects
where old version of Seurat object is presence, or you want to move on
to Seurat downstream analysis after LIGER integration, we show some
potential work arounds in the section below.</p>
<div class="section level3">
<h3 id="update-seurat-object-to-its-latest-structure">Update Seurat object to its latest structure<a class="anchor" aria-label="anchor" href="#update-seurat-object-to-its-latest-structure"></a>
</h3>
<p>Seurat provides <code><a href="https://rdrr.io/pkg/SeuratObject/man/UpdateSeuratObject.html" class="external-link">UpdateSeuratObject()</a></code> function to
automatically convert Seurat objects of old versions to the latest
version.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seuratObj.new</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/UpdateSeuratObject.html" class="external-link">UpdateSeuratObject</a></span><span class="op">(</span><span class="va">seuratObj.old</span><span class="op">)</span></span></code></pre></div>
<p>After this, all the previously introduced analysis can be applied to
the updated object.</p>
</div>
<div class="section level3">
<h3 id="create-a-liger-object-from-multiple-seurat-objects">Create a liger object from multiple Seurat objects<a class="anchor" aria-label="anchor" href="#create-a-liger-object-from-multiple-seurat-objects"></a>
</h3>
<p>In the old days, Seurat recommended that datasets to be integrated
should be stored separately in individual Seurat objects. If you have
data in this form, we suggest using <code>createLigerObject()</code>
function with a named list of Seurat objects.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>dataName1 <span class="op">=</span> <span class="va">seuratObj1</span>, dataName2 <span class="op">=</span> <span class="va">seuratObj2</span><span class="op">)</span></span>
<span><span class="va">ligerObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="va">seurat.list</span><span class="op">)</span></span></code></pre></div>
<p>Note that the metadata from individual objects is not copied to the
liger object, because there is no guarantee of field consistency and can
bring a lot of mess, and LIGER’s core integration functionality does not
rely on metadata for now. However, if you need to keep the metadata for
quick visualization after integration, you can manually bring them to
the liger object with <code>cellMeta()&lt;-</code> method. Setting
<code>useDataset</code> argument ensures that the metadata extracted
from a single dataset is assigned to the matching dataset; argument
<code>inplace</code> indicates that the values are replaced only for the
specified dataset instead of erasing existing values for all cells in
the variable first.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/liger-class.html">cellMeta</a></span><span class="op">(</span><span class="va">ligerObj</span>, <span class="st">"newVarName"</span>, useDataset <span class="op">=</span> <span class="st">"dataName1"</span>, inplace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">seuratObj1</span><span class="op">$</span><span class="va">oldVarName</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="convert-a-single-seurat-object-to-a-liger-object">Convert a single Seurat object to a liger object<a class="anchor" aria-label="anchor" href="#convert-a-single-seurat-object-to-a-liger-object"></a>
</h3>
<p>It is expected that there can be scenario where datasets to be
integrated are stored in a single Seurat object. In this case, the best
practice should be to update the object to the latest version if it is
not yet, and then do <code><a href="https://rdrr.io/r/base/split.html" class="external-link">split()</a></code> as shown in the previous
example. However, if users are interested in a full LIGER style
analysis, we provide <code><a href="../reference/as.liger.html">as.liger()</a></code> method for the Seurat
object. Importantly, the <code>datasetVar</code> argument must be
properly provided for LIGER to infer the dataset source. Optionally, if
annotation and dimensionality reduction is presented, we can set the
default value to the liger object for quick visualization</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb.liger</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.liger.html">as.liger</a></span><span class="op">(</span><span class="va">ifnb</span>, datasetVar <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/liger-class.html">defaultCluster</a></span><span class="op">(</span><span class="va">ifnb.liger</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"seurat_clusters"</span></span>
<span><span class="fu"><a href="../reference/liger-class.html">defaultDimRed</a></span><span class="op">(</span><span class="va">ifnb.liger</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"umap"</span></span>
<span><span class="fu"><a href="../reference/plotDimRed.html">plotByDatasetAndCluster</a></span><span class="op">(</span><span class="va">ifnb.liger</span><span class="op">)</span></span></code></pre></div>
<p><img src="liger_with_seurat_files/figure-html/asliger-1.png" width="1152"></p>
<p>Note that when converting a Seurat object to a liger object, only the
raw gene expression counts are copied to the output object. As mentioned
in the preprocessing step, normalized and scaled data produced by other
tools are generally not suitable for LIGER analysis.</p>
</div>
<div class="section level3">
<h3 id="move-on-to-seurat-downstream-analysis-after-liger">Move on to Seurat downstream analysis after LIGER<a class="anchor" aria-label="anchor" href="#move-on-to-seurat-downstream-analysis-after-liger"></a>
</h3>
<p>Downstream analysis after obtaining joint clustering labels from the
integration is always a common use case. LIGER provides a wide range of
methods for this purpose, such as <code><a href="../reference/liger-DEG.html">runMarkerDEG()</a></code> and
<code><a href="../reference/liger-DEG.html">runPairwiseDEG()</a></code> that help with across-cluster or
across-condition-per-cluster comparison; <code><a href="../reference/imputeKNN.html">imputeKNN()</a></code> and
<code><a href="../reference/linkGenesAndPeaks.html">linkGenesAndPeaks()</a></code> for cross comparison between dual-omics
data; <code><a href="../reference/runGOEnrich.html">runGOEnrich()</a></code> and <code><a href="../reference/runGSEA.html">runGSEA()</a></code> for
functional enrichment analysis. For extending the analysis with as much
possibility as possible, we provide conversion methods for the exporting
a liger object to a more publically known structure.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pbmc"</span>, package <span class="op">=</span> <span class="st">"rliger"</span><span class="op">)</span></span>
<span><span class="co"># Go through a full regular LIGER analysis</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="va">pbmc</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/runINMF.html">runINMF</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/quantileNorm.html">quantileNorm</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/runCluster.html">runCluster</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc.seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ligerToSeurat.html">ligerToSeurat</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc.seurat</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 279 features across 600 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (279 features, 173 variable features)</span></span>
<span><span class="co">##  6 layers present: counts.ctrl, counts.stim, ligerNormData.ctrl, ligerNormData.stim, ligerScaleData.ctrl, ligerScaleData.stim</span></span>
<span><span class="co">##  2 dimensional reductions calculated: inmf, UMAP</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc.seurat</span><span class="op">)</span></span></code></pre></div>
<p><img src="liger_with_seurat_files/figure-html/asseurat-1.png" width="700"></p>
<p>As shown in the printed information above, normalized and scaled data
are explicitly named to avoid confusion with general scRNAseq use case.
The quantile-normalized cell factor loading is now stored in reduction
<code>pbmc.seurat[["inmf"]]</code>.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
