<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rliger">
<title>Iterative single-cell multi-omic integration using online learning • rliger</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.8/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Iterative single-cell multi-omic integration using online learning">
<meta property="og:description" content="rliger">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Iterative single-cell multi-omic integration using online learning</h1>
                        <h4 data-toc-skip class="author">Chao Gao and
Joshua Welch</h4>
            
            <h4 data-toc-skip class="date">9/7/2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/vignettes/articles/online_iNMF_tutorial.Rmd" class="external-link"><code>vignettes/articles/online_iNMF_tutorial.Rmd</code></a></small>
      <div class="d-none name"><code>online_iNMF_tutorial.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="load-data-and-preprocess">Load data and preprocess<a class="anchor" aria-label="anchor" href="#load-data-and-preprocess"></a>
</h2>
<p>LIGER supports integrating datasets based on HDF5 files, for
examples, the 10X CellRanger output file located at
<code>sample/outs/filtered_feature_bc_matrix.h5</code> can be used with
LIGER. Here’s the link to learn more about <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/h5_matrices" class="external-link">CellRanger
output HDF5 files</a>. With the feature that we read one chunk of data
into memory at a time, large-scale datasets stored on disk can be
analyzed with a rather small memory specification.</p>
<blockquote>
<p>Large datasets are often generated over multiple 10X runs (for
example, multiple biological replicates). In such cases it may be
necessary to merge the HDF5 files from each run into a single HDF5 file.
We provide the <code><a href="../reference/mergeH5.html">mergeH5()</a></code> function for this purpose.</p>
</blockquote>
<p>To integrate datasets stored in HDF5 files, we first create a <a href="../reference/liger-class.html">liger object</a> using the file
names of the HDF5 files. We prepared the following example datasets for
demonstration:</p>
<ul>
<li><a href="https://figshare.com/ndownloader/files/43108963" class="external-link">pbmcs_ctrl.h5</a></li>
<li><a href="https://figshare.com/ndownloader/files/43108966" class="external-link">pbmcs_stim.h5</a></li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/welch-lab/liger" class="external-link">rliger</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">dataSource</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    filenames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pbmcs_ctrl_vignette.h5"</span>, </span>
<span>                  <span class="st">"pbmcs_stim_vignette.h5"</span><span class="op">)</span>,</span>
<span>    urls <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"https://figshare.com/ndownloader/files/43108963"</span>,</span>
<span>             <span class="st">"https://figshare.com/ndownloader/files/43108966"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">dataSource</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">dataSource</span><span class="op">$</span><span class="va">filenames</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">dataSource</span><span class="op">$</span><span class="va">urls</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, destfile <span class="op">=</span> <span class="va">dataSource</span><span class="op">$</span><span class="va">filenames</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">pbmcs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ctrl <span class="op">=</span> <span class="va">dataSource</span><span class="op">$</span><span class="va">filenames</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                          stim <span class="op">=</span> <span class="va">dataSource</span><span class="op">$</span><span class="va">filenames</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Similar to other iNMF integration tutorial, we also need to preprocee
the data by 1. normalizing each cell by its library size, 2. select
variable genes for each dataset, and 3. scale the gene expression within
each dataset in a non-negative way.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcs</span> <span class="op">&lt;-</span> <span class="va">pbmcs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/selectGenes.html">selectGenes</a></span><span class="op">(</span>var.thresh <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="scenario-1-sampling-minibatches-from-fully-observed-datasets">Scenario 1: sampling minibatches from fully observed datasets<a class="anchor" aria-label="anchor" href="#scenario-1-sampling-minibatches-from-fully-observed-datasets"></a>
</h2>
<div class="section level3">
<h3 id="online-integrative-nonnegative-matrix-factorization">Online Integrative Nonnegative Matrix Factorization<a class="anchor" aria-label="anchor" href="#online-integrative-nonnegative-matrix-factorization"></a>
</h3>
<p>Now we can use online iNMF to factorize the data, again using only
minibatches that we read from the HDF5 files on demand (default
mini-batch size = 5000). Sufficient number of iterations is crucial for
obtaining ideal factorization result. If the size of the mini-batch is
set to be close to the size of the whole dataset (i.e. an epoch only
contains one iteration), <code>max.epochs</code> needs to be increased
accordingly for more iterations.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runIntegration.html">runIntegration</a></span><span class="op">(</span><span class="va">pbmcs</span>, k <span class="op">=</span> <span class="fl">20</span>, method <span class="op">=</span> <span class="st">"online"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="joint-clustering-and-visualization">Joint clustering and visualization<a class="anchor" aria-label="anchor" href="#joint-clustering-and-visualization"></a>
</h3>
<p>After performing the factorization, we can perform quantile
normalization to align the datasets, and then use graph based community
detection algorithm to find clusters.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantileNorm.html">quantileNorm</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">)</span></span>
<span><span class="va">pbmcs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCluster.html">runCluster</a></span><span class="op">(</span><span class="va">pbmcs</span>, nNeighbors <span class="op">=</span> <span class="fl">30</span>, resolution <span class="op">=</span> <span class="fl">.6</span>, nRandomStarts <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p>We can also visualize the cell factor loadings in two dimensions
using t-SNE or UMAP.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">pbmcs</span>, nNeighbors <span class="op">=</span> <span class="fl">30</span>, minDist <span class="op">=</span> <span class="fl">.3</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotDimRed.html">plotByDatasetAndCluster</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">)</span></span></code></pre></div>
<p><img src="online_iNMF_tutorial_files/figure-html/unnamed-chunk-5-1.png" width="960"></p>
</div>
<div class="section level3">
<h3 id="downstream-analysis">Downstream analysis<a class="anchor" aria-label="anchor" href="#downstream-analysis"></a>
</h3>
<p>For downstream analysis that requires expression data, such as
differential expression test, we currently do not support direct
computation using on-disk HDF5 data. In order to conduct such analyses,
users need to create downsampled subset from the large-scale on-disk
data and read it into memory. By default, <code><a href="../reference/downsample.html">downsample()</a></code>
function randomly samples <code>maxCells</code> cells. Alternatively, we
can set balanced sampling basing on given categorical variables. Note
that <code>newH5 = FALSE</code> has to be set in order to take the
subsample into memory, other wise another downsampled HDF5 file would be
created and still cannot be used for those downstream analysis
method.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcs.sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/downsample.html">downsample</a></span><span class="op">(</span><span class="va">pbmcs</span>, maxCells <span class="op">=</span> <span class="fl">5000</span>, useSlot <span class="op">=</span> <span class="st">"normData"</span>, newH5 <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Using the sampled data stored in memory, we can now compare clusters
or datasets to identify differentially expressed genes. The
<code><a href="../reference/liger-DEG.html">runMarkerDEG()</a></code> function performs differential expression
analysis, by default with Wilcoxon rank-sum test. The default mode of
mark detection goes for finding the marker of each cluster by performing
the test between a cluster against all the rest. Here, we demonstrate
finding the dataset-specific markers, within each cluster.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markerTableList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-DEG.html">runMarkerDEG</a></span><span class="op">(</span><span class="va">pbmcs.sample</span>, conditionBy <span class="op">=</span> <span class="st">"dataset"</span>, splitBy <span class="op">=</span> <span class="st">"leiden_cluster"</span><span class="op">)</span></span></code></pre></div>
<p>The returned object is a list where each element is a
dataset-specific marker table derived within each cluster and contains
all stats for all the genes without filtering or sorting. We strongly
suggest using package “dplyr” for data frame operation due to the
simplicity. Here, we show an example of getting the top 10 markers for
“stim” within cluster 2.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">markerTable2</span> <span class="op">&lt;-</span> <span class="va">markerTableList</span><span class="op">$</span><span class="va">`2`</span></span>
<span><span class="va">markerTable2</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">1</span>, <span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="va">group</span> <span class="op">==</span> <span class="st">"stim"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">padj</span>, <span class="op">-</span><span class="va">logFC</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html" class="external-link">top_n</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">feature</span>, <span class="va">logFC</span>, <span class="va">padj</span><span class="op">)</span></span></code></pre></div>
<p>We can show UMAP coordinates of sampled cells by their loadings on
each factor (Factor 4 as an example). The lower part of the figure shows
the gene loading ranking of the factor, with the middle showing the
ranking of the shared gene loading, and the dataset-specific gene
loading ranking on two sides.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">factorMarker</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getFactorMarkers.html">getFactorMarkers</a></span><span class="op">(</span><span class="va">pbmcs.sample</span>, dataset1 <span class="op">=</span> <span class="st">"ctrl"</span>, dataset2 <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotGeneLoadings.html">plotGeneLoadings</a></span><span class="op">(</span><span class="va">pbmcs.sample</span>, markerTable <span class="op">=</span> <span class="va">factorMarker</span>, useFactor <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="online_iNMF_tutorial_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualization-with-expression">Visualization with expression<a class="anchor" aria-label="anchor" href="#visualization-with-expression"></a>
</h3>
<p>With the subsampled data loaded into memory, we can then also
generate plots of dimensional reduction coordinates colored by
expression of specified gene. For example, the gene ISG15 has expression
visualized below.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimRed.html">plotGeneDimRed</a></span><span class="op">(</span><span class="va">pbmcs.sample</span>, <span class="st">"ISG15"</span><span class="op">)</span></span></code></pre></div>
<p><img src="online_iNMF_tutorial_files/figure-html/unnamed-chunk-10-1.png" width="576"></p>
<p>Furthermore, we can make violin plots of expression of specified
gene. Taking ISG15 as an example again, we can make a violin plot for
each dataset and group each by cluster.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotViolin.html">plotGeneViolin</a></span><span class="op">(</span><span class="va">pbmcs.sample</span>, <span class="st">"ISG15"</span>, byDataset <span class="op">=</span> <span class="cn">TRUE</span>, title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pbmcs.sample</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="online_iNMF_tutorial_files/figure-html/unnamed-chunk-11-1.png" width="700"><img src="online_iNMF_tutorial_files/figure-html/unnamed-chunk-11-2.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="scenario-2-iterative-refinement-by-incorporating-new-datasets">Scenario 2: iterative refinement by incorporating new datasets<a class="anchor" aria-label="anchor" href="#scenario-2-iterative-refinement-by-incorporating-new-datasets"></a>
</h2>
<p>The second scenario of online learning algorithm is where we have
performed factorization on existing dataset(s), and we add on
continually arriving datasets to the pool and update the factorization
with existing result.</p>
<p>For example, still with the same PBMC example, we first perform
online iNMF on the control dataset.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ctrl <span class="op">=</span> <span class="st">"pbmcs_ctrl_vignette.h5"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/runIntegration.html">runIntegration</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"online"</span><span class="op">)</span></span></code></pre></div>
<p>Then, we obtain the new dataset, and then run online iNMF with adding
it as a new arriving dataset. There are a couple of points to note.</p>
<ol style="list-style-type: decimal">
<li>There are different ways of specifying new datasets. We allow using
either a named list of raw data. If users work with HDF5 based data all
the way, then it’s best to have a named list of HDF5 filenames. If users
work with in-memory matrices, then a named list of matrix objects
(dgCMatrix class).</li>
<li>New datasets are always regarded as raw count input, and will be
preprocessed internally. The variable genes identified using the
original existing datasets will always be used for taking the scaled
data. Therefore, users need to make sure that the newly arriving
datasets have them available.</li>
<li>Alternatively, a <a href="../references/liger-class.html">liger
object</a> can also be used as an input for the argument
<code>newDatasets</code>. Users need to be aware of the second point
above, because we remove missing features by default and can cause error
when variable features obtained from existing dataserts is identified as
missing when creating the object for new datasets.</li>
<li>For HDF5 filename inputs, it is assumed that the program can find
raw counts data with 10X CellRanger format. If the HDF5 file stores raw
counts data in a different way (e.g. H5AD data), users will need to
create a separate <a href="../references/liger-class.html">liger
object</a> with specification, and then use the constructed object as
input while being aware of the 2nd and 3rd point above.</li>
<li>After seccessfully integrating the new datasets with the existing
factorization, the returned <a href="../references/liger-class.html">liger obejct</a> will have them
inserted.</li>
</ol>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcs.new</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runOnlineINMF.html">runOnlineINMF</a></span><span class="op">(</span><span class="va">pbmcs</span>, newDatasets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>stim <span class="op">=</span> <span class="st">"pbmcs_stim_vignette.h5"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>After the factorization, we can go through the same procedure to
align the factor loading and jointly cluster the cells, as well as
visualize the integration result.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcs.new</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantileNorm.html">quantileNorm</a></span><span class="op">(</span><span class="va">pbmcs.new</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/runCluster.html">runCluster</a></span><span class="op">(</span>nNeighbors <span class="op">=</span> <span class="fl">30</span>, resolution <span class="op">=</span> <span class="fl">.6</span>, nRandomStarts <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span>nNeighbors <span class="op">=</span> <span class="fl">30</span>, minDist <span class="op">=</span> <span class="fl">.3</span><span class="op">)</span> </span>
<span><span class="fu"><a href="../reference/plotDimRed.html">plotByDatasetAndCluster</a></span><span class="op">(</span><span class="va">pbmcs.new</span><span class="op">)</span></span></code></pre></div>
<p><img src="online_iNMF_tutorial_files/figure-html/unnamed-chunk-14-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="scenario-3-projecting-new-datasets">Scenario 3: Projecting new datasets<a class="anchor" aria-label="anchor" href="#scenario-3-projecting-new-datasets"></a>
</h2>
<p>The third scenario serves as a fast inspection for new datasets. We
do not update the factorization with newly arriving data, but instead,
compute the factor loading in cells (i.e. <span class="math inline">\(H\)</span> matrices for new datasets) using the
existing gene loading in the factors (<span class="math inline">\(W\)</span> matrix from previous factorization).
Therefore, no dataset-specific gene loading in factors (<span class="math inline">\(V\)</span> matrices) can be produced for the new
datasets in this approach. Scenario 3 can be triggered with specifying
the new datasets, following the same guidance as in scenario 2 above,
and set <code>projection = TRUE</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmcs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ctrl <span class="op">=</span> <span class="st">"pbmcs_ctrl_vignette.h5"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/runIntegration.html">runIntegration</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"online"</span><span class="op">)</span></span>
<span><span class="va">pbmcs.new</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runOnlineINMF.html">runOnlineINMF</a></span><span class="op">(</span><span class="va">pbmcs</span>, newDatasets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>stim <span class="op">=</span> <span class="st">"pbmcs_stim_vignette.h5"</span><span class="op">)</span>, projection <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">pbmcs.new</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantileNorm.html">quantileNorm</a></span><span class="op">(</span><span class="va">pbmcs.new</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/runCluster.html">runCluster</a></span><span class="op">(</span>nNeighbors <span class="op">=</span> <span class="fl">30</span>, resolution <span class="op">=</span> <span class="fl">.6</span>, nRandomStarts <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span>nNeighbors <span class="op">=</span> <span class="fl">30</span>, minDist <span class="op">=</span> <span class="fl">.3</span><span class="op">)</span> </span>
<span><span class="fu"><a href="../reference/plotDimRed.html">plotByDatasetAndCluster</a></span><span class="op">(</span><span class="va">pbmcs.new</span><span class="op">)</span></span></code></pre></div>
<p><img src="online_iNMF_tutorial_files/figure-html/unnamed-chunk-15-1.png" width="960"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
