<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Jointly Defining Cell Types from Single-Cell Gene Expression and Methylation Data Using LIGER • rliger2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Jointly Defining Cell Types from Single-Cell Gene Expression and Methylation Data Using LIGER">
<meta property="og:description" content="rliger2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown-header">Installation</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Introductory Vignettes</li>
    <li class="dropdown-header">Integrating Multiple scRNAseq Data</li>
    <li class="dropdown-header">Integrating On-Disk (HDF5 Based) Data</li>
    <li class="divider">
    </li>
<li class="dropdown-header">Data structure and access</li>
    <li class="dropdown-header">Outer container - liger object</li>
    <li class="dropdown-header">Dataset container - ligerDataset object and subclasses</li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Import Data</li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Jointly Defining Cell Types from Single-Cell
Gene Expression and Methylation Data Using LIGER</h1>
                        <h4 data-toc-skip class="author">Joshua
Welch</h4>
            
            <h4 data-toc-skip class="date">4/19/2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/vignettes/rna_methylation.Rmd" class="external-link"><code>vignettes/rna_methylation.Rmd</code></a></small>
      <div class="hidden name"><code>rna_methylation.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette shows how to use LIGER to jointly define cell types
from single-cell gene expression and DNA methylation. We will be using
scRNA-seq and single-nucleus DNA methylation data from mouse cortex.
These are the same datasets used in our previous <a href="https://www.cell.com/cell/fulltext/S0092-8674(19)30504-5" class="external-link">paper</a>.
The steps involved are quite similar to those for integrating multiple
RNA datasets. The only differences are (1) we process the methylation
data into gene-level values comparable to gene expression, (2) we
perform gene selection using the RNA data only, and (3) we both scale
and center the factor loadings before performing quantile
normalization.</p>
<p>The full dataset used in the paper is quite large, so in this
example, we restrict our attention to inhibitory interneurons derived
from the caudal ganglionic eminence (CGE). We downloaded the gene-level
methylation <a href="https://brainome.ucsd.edu/annoj/brain_single_nuclei/" class="external-link">data</a>
provided by the Ecker lab. They generated these values by dividing the
numbers of methylated (non-CG methylation, mCH) and detected positions
across the gene body within each cell. They and we found these gene body
mCH proportions to be negatively correlated with gene expression level
in neurons. The gene expression data is from Saunders et al. 2018. The
full dataset is available for download through <a href="http://dropviz.org/" class="external-link">Dropviz</a>. For convenience, we provide the
cells x genes count matrices for only CGE interneurons as well as the
original cluster assignment obtained from the study:</p>
<ul>
<li><a href="https://figshare.com/ndownloader/files/40222699" class="external-link">RNA
data</a></li>
<li><a href="https://figshare.com/ndownloader/files/40223842" class="external-link">RNA data
cluster assignment</a></li>
<li><a href="https://figshare.com/ndownloader/files/40222702" class="external-link">Methylation
data</a></li>
<li><a href="https://figshare.com/ndownloader/files/40223854" class="external-link">Methylation data
cluster assignment</a></li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"mouse_frontal_cortex_cge_rna.RDS"</span><span class="op">)</span></span>
<span><span class="va">met</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'mouse_frontal_cortex_cge_met.RDS'</span><span class="op">)</span></span>
<span><span class="va">rna_clusts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"rna_clusters.RDS"</span><span class="op">)</span></span>
<span><span class="va">met_clusts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"met_clusters.RDS"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-liger-object-and-preprocess-data">Create liger object and preprocess data<a class="anchor" aria-label="anchor" href="#create-liger-object-and-preprocess-data"></a>
</h2>
<p>We then create a liger object using the methylation and expression
data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/welch-lab/liger" class="external-link">rliger2</a></span><span class="op">)</span></span>
<span><span class="va">rna.met</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>rna <span class="op">=</span> <span class="va">rna</span>, met <span class="op">=</span> <span class="va">met</span><span class="op">)</span>, addPrefix <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, we build the importing function to directly pull
datasets from online storage and build the liger object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna.met</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importVignetteData.html">importVignetteData</a></span><span class="op">(</span><span class="st">"cge"</span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/selectGenes.html">selectGenes()</a></code> function performs variable gene
selection on each of the datasets separately, then takes the union of
the result. The variable genes are selected by comparing the variance of
each gene’s expression to its mean expression. The
<code><a href="../reference/selectGenes.html">selectGenes()</a></code> function was written primarily scRNA-seq in
mind, and the methylation data distribution is quite different. So
instead of taking the union of variable genes from RNA and methylation,
we set <code>datasets.use = "rna"</code> in the function to perform gene
selection using only the RNA dataset. Because gene body mCH proportions
are negatively correlated with gene expression level in neurons, we need
to reverse the direction of the methylation data, with function
<code><a href="../reference/reverseMethData.html">reverseMethData()</a></code> with <code>useDatasets = "met"</code>.
This function simply subtracts all values from the maximum methylation
value per selected dataset. The resulting values are positively
correlated with gene expression. In addition, the proportional nature of
the gene body methylation makes it unnecessary to normalize and scale
the methylation data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna.met</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">)</span></span>
<span><span class="va">rna.met</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="va">rna.met</span>, datasets.use <span class="op">=</span> <span class="st">"rna"</span><span class="op">)</span></span>
<span><span class="va">rna.met</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="va">rna.met</span>, useDatasets <span class="op">=</span> <span class="st">"rna"</span><span class="op">)</span></span>
<span><span class="va">rna.met</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/reverseMethData.html">reverseMethData</a></span><span class="op">(</span><span class="va">rna.met</span>, useDatasets <span class="op">=</span> <span class="st">"met"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="factorize-and-perform-quantile-normalization">Factorize and perform quantile normalization<a class="anchor" aria-label="anchor" href="#factorize-and-perform-quantile-normalization"></a>
</h2>
<p>Next we perform integrative non-negative matrix factorization (iNMF)
in order to identify shared and distinct metagenes across the datasets
and the corresponding metagene loadings for each cell. The most
important parameters in the factorization are <code>k</code> (the number
of factors) and <code>lambda</code> (the penalty parameter which limits
the dataset-specific component of the factorization). The default value
<code>lambda = 5</code> usually provides reasonable results for most
analyses. For this analysis, we simply use <code>k = 20</code> and the
default value of lambda.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna.met</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/optimizeALS.html">optimizeALS</a></span><span class="op">(</span><span class="va">rna.met</span>, k <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p>Using the metagene factors calculated by iNMF, we then assign each
cell to the factor on which it has the highest loading, giving joint
clusters that correspond across datasets. We then perform quantile
normalization by dataset, factor, and cluster to fully integrate the
datasets. To perform this analysis, typing in:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna.met</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantileNorm.html">quantileNorm</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">)</span></span></code></pre></div>
<p>Important parameters of <code><a href="../reference/quantileNorm.html">quantileNorm()</a></code> are as
follows:</p>
<ul>
<li>
<code>nNeighbors</code>. This sets the number of nearest neighbors
for within-dataset KNN graph. The default is <code>20</code>.</li>
<li>
<code>quantiles</code>. This sets the number of quantiles to use for
quantile normalization. The default is <code>50</code>.</li>
<li>
<code>minCells</code>. This indicates the minimum number of cells to
consider a cluster as shared across datasets. The default is
<code>20</code>.</li>
<li>
<code>useDims</code>. This sets the indices of factors to use for
quantile normalization. The user can pass in a vector of indices
indicating specific factors. This is helpful for excluding factors
capturing biological signals such as the cell cycle or technical signals
such as mitochondrial genes. The default is all <code>k</code> of the
factors.</li>
<li>
<code>center</code>. This indicates whether to center the data when
scaling factors. The default is <code>FALSE</code>. This option should
be set to <code>TRUE</code> when metagene loadings have a mean above
zero, as with dense data such as DNA methylation.</li>
<li>
<code>maxSample</code>. This sets the maximum number of cells used
for quantile normalization of each cluster and factor. The default is
<code>1000</code>.</li>
<li>
<code>refineKNN</code>. This indicates whether to increase
robustness of cluster assignments using KNN graph. The default is
<code>TRUE</code>.</li>
<li>
<code>eps</code>. This sets the error bound of the nearest neighbor
search. The default is <code>0.9</code>. Lower values give more accurate
nearest neighbor graphs but take much longer to computer.</li>
<li>
<code>reference</code>. This indicates the name of the dataset to be
used as a reference for quantile normalization. By default, the dataset
with the largest number of cells is used.</li>
</ul>
<p>The <code><a href="../reference/quantileNorm.html">quantileNorm()</a></code> function gives joint clusters that
correspond across datasets, which are often completely satisfactory and
sufficient for downstream analyses. However, if desired, after quantile
normalization, users can additionally run the Leiden algorithm for
community detection, which is widely used in single-cell analysis and
excels at merging small clusters into broad cell classes. This can be
achieved by running the <code><a href="../reference/clustering.html">runLeidenCluster()</a></code> function.
Several tuning parameters, including <code>resolution</code>,
<code>nNeighbors</code>, and <code>prune</code> control the number of
clusters produced by this function. For this dataset, we use a
resolution of <code>0.2</code>, which yields 26 clusters (see
below).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna.met</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clustering.html">runLeidenCluster</a></span><span class="op">(</span><span class="va">rna.met</span>, nNeighbors <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualize-results">Visualize results<a class="anchor" aria-label="anchor" href="#visualize-results"></a>
</h2>
<p>We run t-SNE on the normalized factors, then color the t-SNE
coordinates by dataset and cluster.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna.met</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runTSNE.html">runTSNE</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotDimRed.html">plotByDatasetAndCluster</a></span><span class="op">(</span><span class="va">rna.met</span>, useDimRed <span class="op">=</span> <span class="st">"TSNE"</span>, legendNCol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="rna_methylation_files/figure-html/unnamed-chunk-8-1.png" width="960" style="display: block; margin: auto;"></p>
<p>The t-SNE plot shows that the datasets align well and indicates the
jointly inferred clusters. Using the original RNA and methylation
cluster assignments, we can visually confirm that the joint analysis is
highly consistent with the single-modality analyses.</p>
<p>First, we insert the original cluster assignment into “cellMeta”
variables.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna.met</span><span class="op">$</span><span class="va">rna_cluster</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">$</span><span class="va">rna_cluster</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">)</span></span>
<span><span class="va">rna_clusts</span> <span class="op">&lt;-</span> <span class="va">rna_clusts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rna_clusts</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">rna.met</span><span class="op">$</span><span class="va">rna_cluster</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rna_clusts</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">rna_clusts</span><span class="op">)</span></span>
<span><span class="va">rna.met</span><span class="op">$</span><span class="va">rna_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">$</span><span class="va">rna_cluster</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rna.met</span><span class="op">$</span><span class="va">met_cluster</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">$</span><span class="va">met_cluster</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">)</span></span>
<span><span class="va">met_clusts</span> <span class="op">&lt;-</span> <span class="va">met_clusts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">met_clusts</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">rna.met</span><span class="op">$</span><span class="va">met_cluster</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">met_clusts</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">met_clusts</span><span class="op">)</span></span>
<span><span class="va">rna.met</span><span class="op">$</span><span class="va">met_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">$</span><span class="va">met_cluster</span><span class="op">)</span></span></code></pre></div>
<p>Then, we can visualize with these variables. Note that “NA” values on
the plot indicate the cells belong to the other dataset.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rnaPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDimRed.html">plotClusterDimRed</a></span><span class="op">(</span><span class="va">rna.met</span>, useCluster <span class="op">=</span> <span class="st">"rna_cluster"</span>, useDimRed <span class="op">=</span> <span class="st">"TSNE"</span>, legendNCol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">metPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDimRed.html">plotClusterDimRed</a></span><span class="op">(</span><span class="va">rna.met</span>, useCluster <span class="op">=</span> <span class="st">"met_cluster"</span>, useDimRed <span class="op">=</span> <span class="st">"TSNE"</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">rnaPlot</span>, <span class="va">metPlot</span><span class="op">)</span></span></code></pre></div>
<p><img src="rna_methylation_files/figure-html/unnamed-chunk-10-1.png" width="960" style="display: block; margin: auto;"></p>
<p>Plotting marker genes for subtypes of CGE interneurons confirms that
the data types are properly aligned, with the expected inverse
relationship between gene body mCH and gene expression.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDimRed.html">plotGeneDimRed</a></span><span class="op">(</span><span class="va">rna.met</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Vip"</span>, <span class="st">"Lamp5"</span><span class="op">)</span>, useDimRed <span class="op">=</span> <span class="st">"TSNE"</span>, splitBy <span class="op">=</span> <span class="st">"dataset"</span>,</span>
<span>                        titles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plots</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="rna_methylation_files/figure-html/unnamed-chunk-11-1.png" width="960" style="display: block; margin: auto;"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Joshua Welch, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko, Yichen Wang.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
