<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rliger">
<title>Jointly Defining Cell Types from Single-Cell Gene Expression and Methylation Data Using LIGER • rliger</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Jointly Defining Cell Types from Single-Cell Gene Expression and Methylation Data Using LIGER">
<meta property="og:description" content="rliger">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Jointly Defining Cell Types from Single-Cell Gene Expression and Methylation Data Using LIGER</h1>
                        <h4 data-toc-skip class="author">Joshua
Welch</h4>
            
            <h4 data-toc-skip class="date">4/19/2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/vignettes/articles/rna_methylation.Rmd" class="external-link"><code>vignettes/articles/rna_methylation.Rmd</code></a></small>
      <div class="d-none name"><code>rna_methylation.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette shows how to use LIGER to jointly define cell types
from single-cell gene expression and DNA methylation. We will be using
scRNA-seq and single-nucleus DNA methylation data from mouse cortex.
These are the same datasets used in our previous <a href="https://www.cell.com/cell/fulltext/S0092-8674(19)30504-5" class="external-link">paper</a>.
The steps involved are quite similar to those for integrating multiple
RNA datasets. The only differences are (1) we process the methylation
data into gene-level values comparable to gene expression, (2) we
perform gene selection using the RNA data only, and (3) we both scale
and center the factor loadings before performing quantile
normalization.</p>
<p>The full dataset used in the paper is quite large, so in this
example, we restrict our attention to inhibitory interneurons derived
from the caudal ganglionic eminence (CGE). We downloaded the gene-level
methylation <a href="https://brainome.ucsd.edu/annoj/brain_single_nuclei/" class="external-link">data</a>
provided by the Ecker lab. They generated these values by dividing the
numbers of methylated (non-CG methylation, mCH) and detected positions
across the gene body within each cell. They and we found these gene body
mCH proportions to be negatively correlated with gene expression level
in neurons. The gene expression data is from Saunders et al. 2018. The
full dataset is available for download through <a href="http://dropviz.org/" class="external-link">Dropviz</a>. For convenience, we provide the
gene x cell count matrices for only CGE interneurons as well as the
original cluster assignment obtained from the study:</p>
<ul>
<li>RNA data with 18,455 genes by 2,927 cells <a href="https://figshare.com/ndownloader/files/40222699" class="external-link">mouse_frontal_cortex_cge_rna.RDS</a>
</li>
<li>RNA data cluster assignment <a href="https://figshare.com/ndownloader/files/40223842" class="external-link">rna_clusters.RDS</a>
</li>
<li>Methylation data with 31,928 genes by 184 cells <a href="https://figshare.com/ndownloader/files/40222702" class="external-link">mouse_frontal_cortex_cge_met.RDS</a>
</li>
<li>Methylation data cluster assignment <a href="https://figshare.com/ndownloader/files/40223854" class="external-link">met_clusters.RDS</a>
</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"meth_data/mouse_frontal_cortex_cge_rna.RDS"</span><span class="op">)</span></span>
<span><span class="va">met</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"meth_data/mouse_frontal_cortex_cge_met.RDS"</span><span class="op">)</span></span>
<span><span class="va">rna_clusts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"meth_data/rna_clusters.RDS"</span><span class="op">)</span></span>
<span><span class="va">met_clusts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"meth_data/met_clusters.RDS"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-liger-object-and-preprocess-data">Create liger object and preprocess data<a class="anchor" aria-label="anchor" href="#create-liger-object-and-preprocess-data"></a>
</h2>
<p>We then create a liger object using the methylation and expression
data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://welch-lab.github.io/liger/" class="external-link">rliger</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">rna.met</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>rna <span class="op">=</span> <span class="va">rna</span>, met <span class="op">=</span> <span class="va">met</span><span class="op">)</span>, modal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rna"</span>, <span class="st">"meth"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, we build the importing function to directly pull
datasets from online storage and build the liger object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># NOT RUN</span></span>
<span><span class="va">rna.met</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importVignetteData.html">importCGE</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/selectGenes.html">selectGenes()</a></code> function performs variable gene
selection on each of the datasets separately, then takes the union of
the result. The variable genes are selected by comparing the variance of
each gene’s expression to its mean expression. The
<code><a href="../reference/selectGenes.html">selectGenes()</a></code> function was written primarily scRNA-seq in
mind, and the methylation data distribution is quite different. So
instead of taking the union of variable genes from RNA and methylation,
we set <code>useDatasets = "rna"</code> in the function to perform gene
selection using only the RNA dataset.</p>
<p>Because gene body mCH proportions are negatively correlated with gene
expression level in neurons, we need to reverse the direction of the
methylation data, by simply subtracting all values from the maximum
methylation value per selected dataset. The resulting values are
positively correlated with gene expression. In addition, the
proportional nature of the gene body methylation makes it unnecessary to
normalize and scale the methylation data. By setting argument
<code>modal</code> in <code><a href="../reference/createLiger.html">createLiger()</a></code>, we’ve already marked
that we need to take such actions on the methylation dataset, and the
function <code><a href="../reference/scaleNotCenter.html">scaleNotCenter()</a></code> will automatically have this done
properly.</p>
<p>Optionally, <code><a href="../reference/reverseMethData.html">reverseMethData()</a></code> explicitly does the
reversing operation on specified datasets if the datasets are not
initially marked as methylation data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna.met</span> <span class="op">&lt;-</span> <span class="va">rna.met</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/selectGenes.html">selectGenes</a></span><span class="op">(</span>useDatasets <span class="op">=</span> <span class="st">"rna"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="factorize-and-perform-quantile-normalization">Factorize and perform quantile normalization<a class="anchor" aria-label="anchor" href="#factorize-and-perform-quantile-normalization"></a>
</h2>
<p>Next we perform integrative non-negative matrix factorization (iNMF)
in order to identify shared and distinct metagenes across the datasets
and the corresponding metagene loadings for each cell. The most
important parameters in the factorization are <code>k</code> (the number
of factors) and <code>lambda</code> (the penalty parameter which limits
the dataset-specific component of the factorization). The default value
<code>lambda = 5</code> usually provides reasonable results for most
analyses. For this analysis, we simply use <code>k = 20</code> and the
default value of lambda.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna.met</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runIntegration.html">runIntegration</a></span><span class="op">(</span><span class="va">rna.met</span>, k <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p>Using the metagene factors calculated by iNMF, we then assign each
cell to the factor on which it has the highest loading, giving joint
clusters that correspond across datasets. We then perform quantile
normalization by dataset, factor, and cluster to fully integrate the
datasets. To perform this analysis, typing in:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna.met</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantileNorm.html">quantileNorm</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/quantileNorm.html">quantileNorm()</a></code> function gives joint clusters that
correspond across datasets. However, if desired, after quantile
normalization, users can additionally run the Leiden algorithm for
community detection, which is widely used in single-cell analysis and
excels at merging small clusters into broad cell classes. This can be
achieved by running the <code><a href="../reference/runCluster.html">runCluster()</a></code> function. Several
tuning parameters, including <code>resolution</code>,
<code>nNeighbors</code>, and <code>prune</code> control the number of
clusters produced by this function.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna.met</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCluster.html">runCluster</a></span><span class="op">(</span><span class="va">rna.met</span>, nNeighbors <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualize-results">Visualize results<a class="anchor" aria-label="anchor" href="#visualize-results"></a>
</h2>
<p>We run t-SNE on the normalized factors, then color the t-SNE
coordinates by dataset and cluster.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna.met</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runTSNE.html">runTSNE</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotDimRed.html">plotDatasetDimRed</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">)</span></span></code></pre></div>
<p><img src="rna_methylation_files/figure-html/unnamed-chunk-9-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimRed.html">plotClusterDimRed</a></span><span class="op">(</span><span class="va">rna.met</span>, legendNCol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="rna_methylation_files/figure-html/unnamed-chunk-9-2.png" width="576" style="display: block; margin: auto;"></p>
<p>The t-SNE plot shows that the datasets align well and indicates the
jointly inferred clusters. Using the original RNA and methylation
cluster assignments, we can visually confirm that the joint analysis is
highly consistent with the single-modality analyses.</p>
<p>First, we insert the original cluster assignment into “cellMeta”
variables. The cell metadata table contains variables that apply to all
datasets together. Partial insertion of values for part of all datasets
requires cell index specification to ensure the correctness.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># `rna_clusts` is a named factor object, and the names match with `colnames(rna)`</span></span>
<span><span class="co"># However, a prefix of `"datasetName_"` is appended when creating the liger object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rna_clusts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"rna_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rna_clusts</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># `rna_clusts` contains all cells from the original study while we only deal</span></span>
<span><span class="co"># with the CGE interneurons in this vignette.</span></span>
<span><span class="co"># Use `drop = TRUE` to omit unexisting categories in the resulting subset.</span></span>
<span><span class="va">rna_clusts</span> <span class="op">&lt;-</span> <span class="va">rna_clusts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rna_clusts</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">)</span>, drop <span class="op">=</span> <span class="cn">TRUE</span><span class="op">]</span></span>
<span><span class="co"># Use `columns` to name the new variable in metadata</span></span>
<span><span class="fu"><a href="../reference/liger-class.html">cellMeta</a></span><span class="op">(</span><span class="va">rna.met</span>, columns <span class="op">=</span> <span class="st">"rna_cluster"</span>, cellIdx <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rna_clusts</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">rna_clusts</span></span>
<span></span>
<span><span class="co"># Similarly</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">met_clusts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"met_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">met_clusts</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">met_clusts</span> <span class="op">&lt;-</span> <span class="va">met_clusts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">met_clusts</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">)</span>, drop <span class="op">=</span> <span class="cn">TRUE</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/liger-class.html">cellMeta</a></span><span class="op">(</span><span class="va">rna.met</span>, columns <span class="op">=</span> <span class="st">"met_cluster"</span>, cellIdx <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">met_clusts</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">met_clusts</span></span></code></pre></div>
<p>Then, we can visualize with these variables. Note that “NA” values on
the plot indicate the cells belong to the other dataset.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rnaPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDimRed.html">plotClusterDimRed</a></span><span class="op">(</span><span class="va">rna.met</span>, useCluster <span class="op">=</span> <span class="st">"rna_cluster"</span>, legendNCol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">metPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDimRed.html">plotClusterDimRed</a></span><span class="op">(</span><span class="va">rna.met</span>, useCluster <span class="op">=</span> <span class="st">"met_cluster"</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">rnaPlot</span>, <span class="va">metPlot</span><span class="op">)</span></span></code></pre></div>
<p><img src="rna_methylation_files/figure-html/unnamed-chunk-11-1.png" width="1056" style="display: block; margin: auto;"></p>
<p>Plotting marker genes for subtypes of CGE interneurons confirms that
the data types are properly aligned, with the expected inverse
relationship between gene body mCH and gene expression.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDimRed.html">plotGeneDimRed</a></span><span class="op">(</span><span class="va">rna.met</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Vip"</span>, <span class="st">"Lamp5"</span><span class="op">)</span>, splitBy <span class="op">=</span> <span class="st">"dataset"</span>,</span>
<span>                        titles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rna.met</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plots</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="rna_methylation_files/figure-html/unnamed-chunk-12-1.png" width="1056" style="display: block; margin: auto;"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
