<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>UINMF integration of Dual-omics data • rliger2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="UINMF integration of Dual-omics data">
<meta property="og:description" content="rliger2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown-header">Installation</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Introductory Vignettes</li>
    <li class="dropdown-header">Integrating Multiple scRNAseq Data</li>
    <li class="dropdown-header">Integrating On-Disk (HDF5 Based) Data</li>
    <li class="divider">
    </li>
<li class="dropdown-header">Data structure and access</li>
    <li class="dropdown-header">Outer container - liger object</li>
    <li class="dropdown-header">Dataset container - ligerDataset object and subclasses</li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Import Data</li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>UINMF integration of Dual-omics data</h1>
                        <h4 data-toc-skip class="author">April Kriebel
and Joshua Welch</h4>
            
            <h4 data-toc-skip class="date">04/25/2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/vignettes/SNAREseq_walkthrough.Rmd" class="external-link"><code>vignettes/SNAREseq_walkthrough.Rmd</code></a></small>
      <div class="hidden name"><code>SNAREseq_walkthrough.Rmd</code></div>

    </div>

    
    
<p>Here we integrate the scATAC and scRNA reads from the dual-omics
dataset SNARE-seq as an illustration of how UINMF can be used to
integrate cross-modality data.</p>
<div class="section level2">
<h2 id="download-the-data">1. Download the data<a class="anchor" aria-label="anchor" href="#download-the-data"></a>
</h2>
<p>First, read in your datasets. For this tutorial, we will use three
matrices, which can all be downloaded at <a href="https://www.dropbox.com/sh/y9kjoum8u469nj1/AADik2b2-Qo3os2QSWXdIAbna?dl=0" class="external-link uri">https://www.dropbox.com/sh/y9kjoum8u469nj1/AADik2b2-Qo3os2QSWXdIAbna?dl=0</a>
.</p>
<p>The transcriptomic measures <code>SNAREseq_RNA.RDS</code> is the
SNARE-seq scRNA dataset (31,367 genes by 10,309 cells).</p>
<p>For the shared epigenomic features
<code>SNARE_seq_shared_chromatin_features.RDS</code>, we create a
gene-centric matrix, such that we sum of the number of accessibiltiy
peaks that occur over the gene body and promoter regions for each gene.
For a detailed walkthough of how to generate such a matrix, please see
our <a href="Integrating_scRNA_and_scATAC_data.html">Integrating scRNA
and scATAC data vignette</a>. The resulting matrix of gene-centric
chromatin accessibility is 22,379 genes by 10,309 cells</p>
<p>For the unshared epigenomic features, we binned the genome into bins
of 100,000 bp, and summed the number of peaks occuring in each bin. We
then filtered this matrix for all bins that overlapped with ENCODE
Blacklist regions, ,genes, and promoters. Our filtered matrix
<code>SNARE_seq_unshared_chromatin_features.RDS</code> is 10,309 cells
by 7,437 bins.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/welch-lab/liger" class="external-link">rliger2</a></span><span class="op">)</span></span>
<span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"SNAREseq_RNA.RDS"</span><span class="op">)</span></span>
<span><span class="va">atac_shared</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"SNAREseq_chromatin_accessibility_shared.RDS"</span><span class="op">)</span></span>
<span><span class="va">atac_unshared</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"SNARE_seq_unshared_chromatin_features.RDS"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="selecting-the-unshared-features">2. Selecting the unshared features<a class="anchor" aria-label="anchor" href="#selecting-the-unshared-features"></a>
</h2>
<p>When selecting unshared features for the UINMF integration, it is
critical to consider the type of data you are working with. For unshared
features that gene-centric, the user should follow the feature selection
process outlined in the <a href="UINMF_vignette.html">‘Integrating
unshared features with UINMF’ tutorial</a>.</p>
<p>However, when dealing with features that are <strong>not
gene-centric</strong> (as is the case with our binned intergenic peak
counts), we must select the most variable features using an appropriate
method. Here, we opt to use Seurat’s <code><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures()</a></code>
function.</p>
<p>First, we normalize the unshared features:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">unshareNormed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="va">atac_unshared</span><span class="op">)</span></span></code></pre></div>
<p>Then we select the top 2,000 variable features with Seurat VST
method:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">unshareNormed</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">se</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="va">top2000</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre></div>
<p>Then we scale, but do not center the unshared features</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">unshareScaled</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="va">unshareNormed</span><span class="op">[</span><span class="va">top2000</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preprocessing-and-normalization">3. Preprocessing and normalization<a class="anchor" aria-label="anchor" href="#preprocessing-and-normalization"></a>
</h2>
<p>Create a LIGER object and normalize the shared data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>rna <span class="op">=</span> <span class="va">rna</span>, atac <span class="op">=</span> <span class="va">atac_shared</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span></span></code></pre></div>
<p>Note that when we select the variable genes between the shared
features, we use the RNA dataset to select variable shared features.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="va">lig</span>, var.thresh <span class="op">=</span> <span class="fl">0.1</span>, datasets.use <span class="op">=</span> <span class="st">"rna"</span>, </span>
<span>                   unshared <span class="op">=</span> <span class="cn">TRUE</span>, unshared.datasets <span class="op">=</span> <span class="st">"atac"</span>,</span>
<span>                   unshared.thresh <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span></code></pre></div>
<p>Scale the data, still, not centered.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span></span></code></pre></div>
<p>Add the unshared features that have been properly selected, such that
they are added as a genes by cells matrix.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Fetch the dataset specific object</span></span>
<span><span class="va">ld.atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liger-class.html">dataset</a></span><span class="op">(</span><span class="va">lig</span>, <span class="st">"atac"</span><span class="op">)</span></span>
<span><span class="co"># Insert prepared information</span></span>
<span><span class="va">ld.atac</span><span class="op">@</span><span class="va">varUnsharedFeatures</span> <span class="op">&lt;-</span> <span class="va">top2000</span></span>
<span><span class="fu"><a href="../reference/ligerDataset-class.html">scaleUnsharedData</a></span><span class="op">(</span><span class="va">ld.atac</span>, check <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">unshareScaled</span></span>
<span><span class="co"># Put the updated dataset back</span></span>
<span><span class="fu"><a href="../reference/liger-class.html">datasets</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span><span class="op">[[</span><span class="st">"atac"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">ld.atac</span></span></code></pre></div>
<blockquote>
<p>The latest version of rliger package is equipped with strict
checkpoints on object validity, including the matching of features and
cells between matrices belonging to the same dataset. Given that the
special processing method we are introducing in this vignette, that we
use the intergenic bins as unshared features, we have to turn the checks
off with <code>check = FALSE</code> when inserting the scaled unshared
features into the object.</p>
</blockquote>
</div>
<div class="section level2">
<h2 id="joint-matrix-factorization-20-30-min">4. Joint Matrix Factorization (20-30 min)<a class="anchor" aria-label="anchor" href="#joint-matrix-factorization-20-30-min"></a>
</h2>
<p>To factorize the datasets and include the unshared datasets, set
<code>useUnshared = TRUE</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/optimizeALS.html">optimizeALS</a></span><span class="op">(</span><span class="va">lig</span>, k <span class="op">=</span> <span class="fl">30</span>, maxIter <span class="op">=</span> <span class="fl">30</span>, useUnshared <span class="op">=</span> <span class="cn">TRUE</span>, thresh <span class="op">=</span> <span class="fl">1e-10</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quantile-normalization-and-joint-clustering-5-min">5. Quantile Normalization and Joint Clustering (~ 5 min)<a class="anchor" aria-label="anchor" href="#quantile-normalization-and-joint-clustering-5-min"></a>
</h2>
<p>After factorization, the resulting Liger object can be used in all
downstream LIGER functions without adjustment. The default reference
dataset for quantile normalization is the larger dataset, but the user
should select the higher quality dataset as the reference dataset, even
if it is the smaller dataset.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantileNorm.html">quantileNorm</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span></span>
<span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clustering.html">runLeidenCluster</a></span><span class="op">(</span><span class="va">lig</span>, resolution <span class="op">=</span> <span class="fl">0.8</span>, nNeighbors <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualizations-and-downstream-processing-1-min">6. Visualizations and Downstream processing (&lt; 1 min)<a class="anchor" aria-label="anchor" href="#visualizations-and-downstream-processing-1-min"></a>
</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">lig</span>, nNeighbors <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p>Next, we can visualize our returned factorized object by dataset to
check the alignment between datasets, as well as by cluster determined
in the factorization.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimRed.html">plotByDatasetAndCluster</a></span><span class="op">(</span><span class="va">lig</span>, dotSize <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="SNAREseq_walkthrough_files/figure-html/visualizations-1.png" width="960"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Joshua Welch, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko, Yichen Wang.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
