<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rliger">
<title>Unshared Features UINMF • rliger</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.8/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Unshared Features UINMF">
<meta property="og:description" content="rliger">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Unshared Features UINMF</h1>
                        <h4 data-toc-skip class="author">Joshua Welch
and April Kriebel</h4>
            
            <h4 data-toc-skip class="date">12/03/2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/vignettes/articles/UINMF_vignette.Rmd" class="external-link"><code>vignettes/articles/UINMF_vignette.Rmd</code></a></small>
      <div class="d-none name"><code>UINMF_vignette.Rmd</code></div>
    </div>

    
    
<p>Unshared integrative non-negative matrix factorization (UINMF) [<a href="https://doi.org/10.1038/s41467-022-28431-4" class="external-link">Kriebel and Welch,
2022</a>] is an extensive algorithm of our original iNMF algorithm [<a href="https://doi.org/10.1016/j.cell.2019.05.006" class="external-link">Welch J D, 2019</a>].
It allows integrating datasets using shared and unshared features at the
same time. We demonstrated in the publication that:</p>
<ol style="list-style-type: decimal">
<li>UINMF can be used to incorporate intergenic information when
integrating transcriptomic and epigenomic datasets.</li>
<li>UINMF can incorporate genes not measured in targeted spatial
transcriptomic datasets, allowing better resolution of fine cellular
subtypes within a spatial coordinate frame.</li>
<li>UINMF can utilize all of the information present in single-cell
multimodal datasets when integrating with single-modality datasets.</li>
<li>UINMF can accommodate non-orthologous genes in cross-species
integration analyses.</li>
</ol>
<p>For this tutorial, we prepared example input data which can be found
in the <a href="https://www.dropbox.com/scl/fo/npj3x6ygp5bkl702gycnu/h?rlkey=llaana00wg81tgzp3oodhhe7p&amp;dl=0" class="external-link">Dropbox
folder</a></p>
<ul>
<li>osmFISH dataset <code>OSMFISH.vin.RDS</code>, 33 genes by 5,219
cells, in dense matrix form, rom the mouse somatosensory cortex</li>
<li>A downsampled single-cell RNA-seq dataset
<code>DROPVIZ.vin.RDS</code>, 29,463 genes by 15,000 cells, from the
frontal cortex.</li>
</ul>
<div class="section level2">
<h2 id="step-1-preprocessing-and-normalization">Step 1: Preprocessing and Normalization<a class="anchor" aria-label="anchor" href="#step-1-preprocessing-and-normalization"></a>
</h2>
<div class="section level3">
<h3 id="loading-the-data">1.1 Loading the data<a class="anchor" aria-label="anchor" href="#loading-the-data"></a>
</h3>
<p>The prepared RDS files can be directly loaded into the R environment.
They are presented in dense matrix (R default) form. We recommend
converting them to sparse matrix form (“dgCMatrix” class supported by
“Matrix” package) in order to reduce memory usage and speed up
downstream analysis.</p>
<p>Next, create your Liger object, submitting the datasets in list
format. The unshared features should not be subsetted out, or submitted
separately. Rather, they should be included in the matrix submitted for
that dataset. For example,the scRNA-seq data is submitted in its
entirety, the unshared features are not submitted separately. This helps
ensure proper normalization.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/welch-lab/liger" class="external-link">rliger</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>ligerDotSize <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">osmFISH</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"uinmf_vignette_data/OSMFISH.vin.RDS"</span><span class="op">)</span></span>
<span><span class="va">osmFISH</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">osmFISH</span>, <span class="st">"CsparseMatrix"</span><span class="op">)</span></span>
<span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"uinmf_vignette_data/DROPVIZ.vin.RDS"</span><span class="op">)</span></span>
<span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createLiger.html">createLiger</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>osmFISH <span class="op">=</span> <span class="va">osmFISH</span>, rna <span class="op">=</span> <span class="va">rna</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="preprocess-the-datasets">1.2 Preprocess the datasets<a class="anchor" aria-label="anchor" href="#preprocess-the-datasets"></a>
</h3>
<p>The normalization is applied to the datasets in their entirety.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize.html">normalize</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span></span></code></pre></div>
<p>The normalized data for each dataset can be accessed with
<code><a href="../reference/ligerDataset-class.html">normData()</a></code>.</p>
<p>Next, we select variable genes to be used for integration. To include
unshared features in your analysis, set the
<code>useUnsharedDatasets</code> with a vector of names of the datasets
where unshared features should be included. For instance, in this case,
we wish to include the unshared features from the RNA dataset, the
second dataset, in our analysis. We provide an individual tuning
threshold for unshared features. If a single value is submitted, that
threshold is applied to all datasets’ unshared features. If multiple
datasets include unshared features, the user can specify an individual
thresholds for each dataset by submitting a list of thresholds the same
length as the number of datasets with unshared datasets.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="va">lig</span>, useUnsharedDatasets <span class="op">=</span> <span class="st">"rna"</span>, unsharedThresh <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Selecting variable features for dataset <span style="color: #0000BB;">"osmFISH"</span></span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> ... 33 features selected out of 33 shared features.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Selecting variable features for dataset <span style="color: #0000BB;">"rna"</span></span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> ... 29 features selected out of 33 shared features.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> ... 1327 features selected out of 24520 unshared features.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> Finally 33 shared variable features are selected.</span></span></code></pre>
<p>The selected shared features, all 33 osmFISH genes in this case, can
be accessed with <code>varFeatures(lig)</code>. The unshared variable
features of a specific dataset, namingly <code>"rna"</code> in this
case, can be accessed with
<code>varUnsharedFeatures(lig, "rna")</code>.</p>
<p>We scale the selected gene expression matrix before the integration.
The datasets are not centered before scaling so the non-negative
constraint for the factorization is addressed.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span></span></code></pre></div>
<p>The scaled matrix for shared features can be accessed with
<code><a href="../reference/ligerDataset-class.html">scaleData()</a></code>. The scaled matrix of the unshared features of
the RNA dataset is a separate matrix and can be accessed with
<code>scaleUnsharedData(lig, "rna")</code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="step-2-joint-matrix-factorization">Step 2: Joint Matrix Factorization<a class="anchor" aria-label="anchor" href="#step-2-joint-matrix-factorization"></a>
</h2>
<p>We provide function <code><a href="../reference/runIntegration.html">runIntegration()</a></code> for a general
wrapper of all integration method we developed. Users can do
<code>runIntegration(..., method = "UINMF")</code> to perform the
analysis. Alternatively, stand-alone function <code><a href="../reference/runUINMF.html">runUINMF()</a></code> is
also provided with more specific parameter setitng guide.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runIntegration.html">runIntegration</a></span><span class="op">(</span><span class="va">lig</span>, k <span class="op">=</span> <span class="fl">30</span>, method <span class="op">=</span> <span class="st">"UINMF"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-3-quantile-normalization-and-joint-clustering">Step 3: Quantile Normalization and Joint Clustering<a class="anchor" aria-label="anchor" href="#step-3-quantile-normalization-and-joint-clustering"></a>
</h2>
<div class="section level3">
<h3 id="quantile-normalize-the-factor-loading">3.1 Quantile normalize the factor loading<a class="anchor" aria-label="anchor" href="#quantile-normalize-the-factor-loading"></a>
</h3>
<p>After factorization, the resulting liger object can be used in all
downstream LIGER functions without adjustment. The default reference
dataset for quantile normalization is the larger dataset, but the user
should select the higher quality dataset as the reference dataset, even
if it is the smaller dataset.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantileNorm.html">quantileNorm</a></span><span class="op">(</span><span class="va">lig</span>, reference <span class="op">=</span> <span class="st">"rna"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="leiden-clustering">3.2 Leiden clustering<a class="anchor" aria-label="anchor" href="#leiden-clustering"></a>
</h3>
<p>With the quantile normalized cell factor loading, we can then apply
Leiden graph-based community detection method on it.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCluster.html">runCluster</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="step-4-visualizations-and-downstream-processing">Step 4: Visualizations and Downstream processing<a class="anchor" aria-label="anchor" href="#step-4-visualizations-and-downstream-processing"></a>
</h2>
<div class="section level3">
<h3 id="create-dimension-reduction-embedding">4.1 Create dimension reduction embedding<a class="anchor" aria-label="anchor" href="#create-dimension-reduction-embedding"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualize">4.2 Visualize<a class="anchor" aria-label="anchor" href="#visualize"></a>
</h3>
<p>Next, we can visualize our returned factorized object by dataset to
check the alignment between datasets, as well as by cluster determined
in the factorization.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimRed.html">plotDatasetDimRed</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span></span></code></pre></div>
<p><img src="UINMF_vignette_files/figure-html/visualizations-1.png" width="576"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimRed.html">plotClusterDimRed</a></span><span class="op">(</span><span class="va">lig</span>, legendNCol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="UINMF_vignette_files/figure-html/visualizations-2.png" width="576"></p>
<p>We can also examine features such as gene expression.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pdgfra</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDimRed.html">plotGeneDimRed</a></span><span class="op">(</span><span class="va">lig</span>, <span class="st">"Pdgfra"</span>, splitBy <span class="op">=</span> <span class="st">"dataset"</span>, title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bmp4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDimRed.html">plotGeneDimRed</a></span><span class="op">(</span><span class="va">lig</span>, <span class="st">"Bmp4"</span>, splitBy <span class="op">=</span> <span class="st">"dataset"</span>, title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">lig</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>    <span class="va">pdgfra</span><span class="op">$</span><span class="va">Pdgfra.rna</span>, <span class="va">pdgfra</span><span class="op">$</span><span class="va">Pdgfra.osmFISH</span>,</span>
<span>    <span class="va">bmp4</span><span class="op">$</span><span class="va">Bmp4.rna</span>, <span class="va">bmp4</span><span class="op">$</span><span class="va">Bmp4.osmFISH</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="UINMF_vignette_files/figure-html/gene%20visualizations-1.png" width="768"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
