<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='This function is a wrapper to switch between alternative factor loading
alignment methods that LIGER provides, which is a required step for producing
the final integrated result. Two methods are provided (click on options for
more details):

method = "quantileNorm": Previously published quantile
normalization method. (default)
method = "centroidAlign": Newly developed centroid
alignment method. 

'><title>Align factor loadings to get final integration — alignFactors • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Align factor loadings to get final integration — alignFactors"><meta property="og:description" content='This function is a wrapper to switch between alternative factor loading
alignment methods that LIGER provides, which is a required step for producing
the final integrated result. Two methods are provided (click on options for
more details):

method = "quantileNorm": Previously published quantile
normalization method. (default)
method = "centroidAlign": Newly developed centroid
alignment method. 

'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    <a class="dropdown-item" href="../articles/benchmark.html">Benchmarking integration quality</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Align factor loadings to get final integration</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/integration.R" class="external-link"><code>R/integration.R</code></a></small>
      <div class="d-none name"><code>alignFactors.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function is a wrapper to switch between alternative factor loading
alignment methods that LIGER provides, which is a required step for producing
the final integrated result. Two methods are provided (click on options for
more details):</p>
<ul><li><p><code>method = "<a href="quantileNorm.html">quantileNorm</a>"</code>: Previously published quantile
normalization method. (default)</p></li>
<li><p><code>method = "<a href="centroidAlign.html">centroidAlign</a>"</code>: Newly developed centroid
alignment method. <a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="figures/lifecycle-experimental.svg" alt="[Experimental]"></a></p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">alignFactors</span><span class="op">(</span><span class="va">object</span>, method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"quantileNorm"</span>, <span class="st">"centroidAlign"</span><span class="op">)</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for liger</span></span>
<span><span class="fu">alignFactors</span><span class="op">(</span><span class="va">object</span>, method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"quantileNorm"</span>, <span class="st">"centroidAlign"</span><span class="op">)</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for Seurat</span></span>
<span><span class="fu">alignFactors</span><span class="op">(</span><span class="va">object</span>, method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"quantileNorm"</span>, <span class="st">"centroidAlign"</span><span class="op">)</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> or Seurat object with valid factorization
result available (i.e. <code><a href="runIntegration.html">runIntegration</a></code> performed in advance).</p></dd>


<dt>method</dt>
<dd><p>Character, method to align factors. Default
<code>"centroidAlign"</code>. Optionally <code>"quantileNorm"</code>.</p></dd>


<dt>...</dt>
<dd><p>Additional arguments passed to selected methods.
For <code>"quantileNorm"</code>:</p><dl><dt><code>quantiles</code></dt>
<dd><p>Number of quantiles to use for quantile
normalization. Default <code>50</code>.</p></dd>

<dt><code>reference</code></dt>
<dd><p>Character, numeric or logical selection of one
dataset, out of all available datasets in <code>object</code>, to use as a
"reference" for quantile normalization. Default <code>NULL</code> tries to find
an RNA dataset with the largest number of cells; if no RNA dataset
available, use the globally largest dataset.</p></dd>

<dt><code>minCells</code></dt>
<dd><p>Minimum number of cells to consider a cluster
shared across datasets. Default <code>20</code>.</p></dd>

<dt><code>nNeighbors</code></dt>
<dd><p>Number of nearest neighbors for within-dataset
knn graph. Default <code>20</code>.</p></dd>

<dt><code>useDims</code></dt>
<dd><p>Indices of factors to use for shared nearest factor
determination. Default <code>NULL</code> uses all factors.</p></dd>

<dt><code>center</code></dt>
<dd><p>Whether to center the data when scaling factors.
Could be useful for less sparse modalities like methylation data.
Default <code>FALSE</code>.</p></dd>

<dt><code>maxSample</code></dt>
<dd><p>Maximum number of cells used for quantile
normalization of each cluster and factor. Default <code>1000</code>.</p></dd>

<dt><code>eps</code></dt>
<dd><p>The error bound of the nearest neighbor search. Lower
values give more accurate nearest neighbor graphs but take much longer to
compute. Default <code>0.9</code>.</p></dd>

<dt><code>refineKNN</code></dt>
<dd><p>Whether to increase robustness of cluster
assignments using KNN graph. Default <code>TRUE</code>.</p></dd>

<dt><code>clusterName</code></dt>
<dd><p>Variable name that will store the clustering
result in metadata of a <a href="liger-class.html">liger</a> object or a <code>Seurat</code>
object. Default <code>"quantileNorm_cluster"</code>.</p></dd>

<dt><code>seed</code></dt>
<dd><p>Random seed to allow reproducible results. Default
<code>1</code>.</p></dd>

<dt><code>verbose</code></dt>
<dd><p>Logical. Whether to show information of the
progress. Default <code>getOption("ligerVerbose")</code> or <code>TRUE</code> if
users have not set.</p></dd>


</dl><p>For <code>"centroidAlign"</code> <a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="figures/lifecycle-experimental.svg" alt="[Experimental]"></a>:</p><dl><dt><code>lambda</code></dt>
<dd><p>Ridge regression penalty applied to each dataset.
Can be one number that applies to all datasets, or a numeric vector with
length equal to the number of datasets. Default <code>1</code>.</p></dd>

<dt><code>useDims</code></dt>
<dd><p>Indices of factors to use considered for the
alignment. Default <code>NULL</code> uses all factors.</p></dd>

<dt><code>scaleEmb</code></dt>
<dd><p>Logical, whether to scale the factor loading being
considered as the embedding. Default <code>TRUE</code>.</p></dd>

<dt><code>centerEmb</code></dt>
<dd><p>Logical, whether to center the factor loading
being considered as the embedding before scaling it. Default <code>TRUE</code>.</p></dd>

<dt><code>scaleCluster</code></dt>
<dd><p>Logical, whether to scale the factor loading
being considered as the cluster assignment probability. Default
<code>FALSE</code>.</p></dd>

<dt><code>centerCluster</code></dt>
<dd><p>Logical, whether to center the factor loading
being considered as the cluster assignment probability before scaling it.
Default <code>FALSE</code>.</p></dd>

<dt><code>shift</code></dt>
<dd><p>Logical, whether to shift the factor loading being
considered as the cluster assignment probability after centered scaling.
Default <code>FALSE</code>.</p></dd>

<dt><code>diagnosis</code></dt>
<dd><p>Logical, whether to return cell metadata variables
with diagnostic information. Default <code>FALSE</code>.</p></dd>


</dl></dd>

</dl></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="quantileNorm.html">quantileNorm</a></code>, <code><a href="centroidAlign.html">centroidAlign</a></code></p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

