<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function converts data stored in SingleCellExperiment (SCE), Seurat
object or a merged sparse matrix (dgCMatrix) into a liger object. This is
designed for a container object or matrix that already contains multiple
datasets to be integerated with LIGER. For individual datasets, please use
createLiger instead."><title>Converting other classes of data to a liger object — as.liger.dgCMatrix • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Converting other classes of data to a liger object — as.liger.dgCMatrix"><meta property="og:description" content="This function converts data stored in SingleCellExperiment (SCE), Seurat
object or a merged sparse matrix (dgCMatrix) into a liger object. This is
designed for a container object or matrix that already contains multiple
datasets to be integerated with LIGER. For individual datasets, please use
createLiger instead."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    <a class="dropdown-item" href="../articles/benchmark.html">Benchmarking integration quality</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Converting other classes of data to a liger object</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/classConversion.R" class="external-link"><code>R/classConversion.R</code></a>, <a href="https://github.com/welch-lab/liger/blob/HEAD/R/generics.R" class="external-link"><code>R/generics.R</code></a></small>
      <div class="d-none name"><code>as.liger.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function converts data stored in SingleCellExperiment (SCE), Seurat
object or a merged sparse matrix (dgCMatrix) into a liger object. This is
designed for a container object or matrix that already contains multiple
datasets to be integerated with LIGER. For individual datasets, please use
<code><a href="createLiger.html">createLiger</a></code> instead.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S3 method for dgCMatrix</span></span>
<span><span class="fu">as.liger</span><span class="op">(</span><span class="va">object</span>, datasetVar <span class="op">=</span> <span class="cn">NULL</span>, modal <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for SingleCellExperiment</span></span>
<span><span class="fu">as.liger</span><span class="op">(</span><span class="va">object</span>, datasetVar <span class="op">=</span> <span class="cn">NULL</span>, modal <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for Seurat</span></span>
<span><span class="fu">as.liger</span><span class="op">(</span><span class="va">object</span>, datasetVar <span class="op">=</span> <span class="cn">NULL</span>, modal <span class="op">=</span> <span class="cn">NULL</span>, assay <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">seuratToLiger</span><span class="op">(</span><span class="va">object</span>, datasetVar <span class="op">=</span> <span class="cn">NULL</span>, modal <span class="op">=</span> <span class="cn">NULL</span>, assay <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">as.liger</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>Object.</p></dd>


<dt>datasetVar</dt>
<dd><p>Specify the dataset belonging by: 1. Select a variable from
existing metadata in the object (e.g. colData column); 2. Specify a
vector/factor that assign the dataset belonging. 3. Give a single character
string which means that all data is from one dataset (must not be a metadata
variable, otherwise it is understood as 1.). Default <code>NULL</code> gathers
things into one dataset and names it "sample" for dgCMatrix, attempts
to find variable "sample" from SCE or "orig.ident" from Seurat.</p></dd>


<dt>modal</dt>
<dd><p>Modality setting for each dataset. See
<code><a href="createLiger.html">createLiger</a></code>.</p></dd>


<dt>...</dt>
<dd><p>Additional arguments passed to <code><a href="createLiger.html">createLiger</a></code></p></dd>


<dt>assay</dt>
<dd><p>Name of assay to use. Default <code>NULL</code> uses current active
assay.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>a <a href="liger-class.html">liger</a> object.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>For Seurat V5 structure, it is highly recommended that users make use of its
split layer feature, where things like "counts", "data", and "scale.data"
can be held for each dataset in the same Seurat object, e.g. with
"count.ctrl", "count.stim", not merged. If a Seurat object with split layers
is given, <code>datasetVar</code> will be ignored and the layers will be directly
used.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># dgCMatrix (common sparse matrix class), usually obtained from other</span></span></span>
<span class="r-in"><span><span class="co"># container object, and contains multiple samples merged in one.</span></span></span>
<span class="r-in"><span><span class="va">matList</span> <span class="op">&lt;-</span> <span class="fu"><a href="ligerDataset-class.html">rawData</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">multiSampleMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="mergeSparseAll.html">mergeSparseAll</a></span><span class="op">(</span><span class="va">matList</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># The `datasetVar` argument expects the variable assigning the sample source</span></span></span>
<span class="r-in"><span><span class="va">pbmc2</span> <span class="op">&lt;-</span> <span class="fu">as.liger</span><span class="op">(</span><span class="va">multiSampleMatrix</span>, datasetVar <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">dataset</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> No human mitochondrial gene found in the union of dataset <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> calculating QC for dataset <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Updated QC variables: <span style="color: #0000BB;">"nUMI"</span>, <span style="color: #0000BB;">"nGene"</span>, <span style="color: #0000BB;">"mito"</span>, <span style="color: #0000BB;">"ribo"</span>, and <span style="color: #0000BB;">"hemo"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> calculating QC for dataset <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> calculating QC for dataset <span style="color: #0000BB;">"ctrl"</span> ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> No human mitochondrial gene found in the union of dataset <span style="color: #0000BB;">"stim"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> calculating QC for dataset <span style="color: #0000BB;">"stim"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Updated QC variables: <span style="color: #0000BB;">"nUMI"</span>, <span style="color: #0000BB;">"nGene"</span>, <span style="color: #0000BB;">"mito"</span>, <span style="color: #0000BB;">"ribo"</span>, and <span style="color: #0000BB;">"hemo"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> calculating QC for dataset <span style="color: #0000BB;">"stim"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> calculating QC for dataset <span style="color: #0000BB;">"stim"</span> ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">pbmc2</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class liger with 600 cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> datasets(2): ctrl (300 cells), stim (300 cells) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> cellMeta(7): dataset, barcode, nUMI, ..., hemo </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> varFeatures(0):  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dimReds(0):  </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"SingleCellExperiment"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>    <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">SingleCellExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>        assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">multiSampleMatrix</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">sce</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">dataset</span></span></span>
<span class="r-in"><span>    <span class="va">pbmc3</span> <span class="op">&lt;-</span> <span class="fu">as.liger</span><span class="op">(</span><span class="va">sce</span>, datasetVar <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">pbmc3</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> No human mitochondrial gene found in the union of dataset <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> calculating QC for dataset <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Updated QC variables: <span style="color: #0000BB;">"nUMI"</span>, <span style="color: #0000BB;">"nGene"</span>, <span style="color: #0000BB;">"mito"</span>, <span style="color: #0000BB;">"ribo"</span>, and <span style="color: #0000BB;">"hemo"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> calculating QC for dataset <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> calculating QC for dataset <span style="color: #0000BB;">"ctrl"</span> ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> No human mitochondrial gene found in the union of dataset <span style="color: #0000BB;">"stim"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> calculating QC for dataset <span style="color: #0000BB;">"stim"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Updated QC variables: <span style="color: #0000BB;">"nUMI"</span>, <span style="color: #0000BB;">"nGene"</span>, <span style="color: #0000BB;">"mito"</span>, <span style="color: #0000BB;">"ribo"</span>, and <span style="color: #0000BB;">"hemo"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> calculating QC for dataset <span style="color: #0000BB;">"stim"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> calculating QC for dataset <span style="color: #0000BB;">"stim"</span> ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class liger with 600 cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> datasets(2): ctrl (300 cells), stim (300 cells) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> cellMeta(8): dataset, barcode, nUMI, ..., sample </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> varFeatures(0):  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dimReds(0):  </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"Seurat"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>    <span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">SeuratObject</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">multiSampleMatrix</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="co"># Seurat creates variable "orig.ident" by identifying the cell barcode</span></span></span>
<span class="r-in"><span>    <span class="co"># prefixes, which is indeed what we need in this case. Users might need</span></span></span>
<span class="r-in"><span>    <span class="co"># to be careful and have it confirmed first.</span></span></span>
<span class="r-in"><span>    <span class="va">pbmc4</span> <span class="op">&lt;-</span> <span class="fu">as.liger</span><span class="op">(</span><span class="va">seu</span>, datasetVar <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">pbmc4</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="co"># As per Seurat V5 updates with layered data, specifically helpful udner the</span></span></span>
<span class="r-in"><span>    <span class="co"># scenario of dataset integration. "counts" and etc for each datasets can be</span></span></span>
<span class="r-in"><span>    <span class="co"># split into layers.</span></span></span>
<span class="r-in"><span>    <span class="va">seu5</span> <span class="op">&lt;-</span> <span class="va">seu</span></span></span>
<span class="r-in"><span>    <span class="va">seu5</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">seu5</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>, <span class="va">pbmc</span><span class="op">$</span><span class="va">dataset</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">SeuratObject</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Layers.html" class="external-link">Layers</a></span><span class="op">(</span><span class="va">seu5</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">pbmc5</span> <span class="op">&lt;-</span> <span class="fu">as.liger</span><span class="op">(</span><span class="va">seu5</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">pbmc5</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> No human mitochondrial gene found in the union of dataset <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> calculating QC for dataset <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Updated QC variables: <span style="color: #0000BB;">"nUMI"</span>, <span style="color: #0000BB;">"nGene"</span>, <span style="color: #0000BB;">"mito"</span>, <span style="color: #0000BB;">"ribo"</span>, and <span style="color: #0000BB;">"hemo"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> calculating QC for dataset <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> calculating QC for dataset <span style="color: #0000BB;">"ctrl"</span> ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> No human mitochondrial gene found in the union of dataset <span style="color: #0000BB;">"stim"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> calculating QC for dataset <span style="color: #0000BB;">"stim"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Updated QC variables: <span style="color: #0000BB;">"nUMI"</span>, <span style="color: #0000BB;">"nGene"</span>, <span style="color: #0000BB;">"mito"</span>, <span style="color: #0000BB;">"ribo"</span>, and <span style="color: #0000BB;">"hemo"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> calculating QC for dataset <span style="color: #0000BB;">"stim"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> calculating QC for dataset <span style="color: #0000BB;">"stim"</span> ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "counts.ctrl" "counts.stim"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> No human mitochondrial gene found in the union of dataset <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> calculating QC for dataset <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Updated QC variables: <span style="color: #0000BB;">"nUMI"</span>, <span style="color: #0000BB;">"nGene"</span>, <span style="color: #0000BB;">"mito"</span>, <span style="color: #0000BB;">"ribo"</span>, and <span style="color: #0000BB;">"hemo"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> calculating QC for dataset <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> calculating QC for dataset <span style="color: #0000BB;">"ctrl"</span> ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> No human mitochondrial gene found in the union of dataset <span style="color: #0000BB;">"stim"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> calculating QC for dataset <span style="color: #0000BB;">"stim"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Updated QC variables: <span style="color: #0000BB;">"nUMI"</span>, <span style="color: #0000BB;">"nGene"</span>, <span style="color: #0000BB;">"mito"</span>, <span style="color: #0000BB;">"ribo"</span>, and <span style="color: #0000BB;">"hemo"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> calculating QC for dataset <span style="color: #0000BB;">"stim"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> calculating QC for dataset <span style="color: #0000BB;">"stim"</span> ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class liger with 600 cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> datasets(2): ctrl (300 cells), stim (300 cells) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> cellMeta(10): dataset, barcode, nUMI, ..., nFeature_RNA </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> varFeatures(0):  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dimReds(0):  </span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

