<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>[Deprecated] Calculate agreement metric — calcAgreement-deprecated • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="[Deprecated] Calculate agreement metric — calcAgreement-deprecated"><meta property="og:description" content="This metric quantifies how much the factorization and alignment distorts the geometry of the
original datasets. The greater the agreement, the less distortion of geometry there is. This is
calculated by performing dimensionality reduction on the original and quantile aligned (or just
factorized) datasets, and measuring similarity between the k nearest neighbors for each cell in
original and aligned datasets. The Jaccard index is used to quantify similarity, and is the final
metric averages across all cells.
Note that for most datasets, the greater the chosen k, the greater the agreement in general.
There are several options for dimensionality reduction, with the default being 'NMF' as it is
expected to be most similar to iNMF. Although agreement can theoretically approach 1, in practice
it is usually no higher than 0.2-0.3 (particularly for non-deterministic approaches like NMF)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/import_data.html">Import data</a>
    </li>
    <li>
      <a href="../articles/liger_object.html">Introduction to a liger object</a>
    </li>
    <li>
      <a href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>[Deprecated] Calculate agreement metric</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/metrics.R" class="external-link"><code>R/metrics.R</code></a></small>
    <div class="hidden name"><code>calcAgreement-deprecated.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This metric quantifies how much the factorization and alignment distorts the geometry of the
original datasets. The greater the agreement, the less distortion of geometry there is. This is
calculated by performing dimensionality reduction on the original and quantile aligned (or just
factorized) datasets, and measuring similarity between the k nearest neighbors for each cell in
original and aligned datasets. The Jaccard index is used to quantify similarity, and is the final
metric averages across all cells.</p>
<p>Note that for most datasets, the greater the chosen k, the greater the agreement in general.
There are several options for dimensionality reduction, with the default being 'NMF' as it is
expected to be most similar to iNMF. Although agreement can theoretically approach 1, in practice
it is usually no higher than 0.2-0.3 (particularly for non-deterministic approaches like NMF).</p>
    </div>


    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p><code>liger</code> object. Should call quantile_norm before calling.</p></dd>


<dt>dr.method</dt>
<dd><p>Dimensionality reduction method to use for assessing pre-alignment geometry
(either "PCA", "NMF", or "ICA"). (default "NMF")</p></dd>


<dt>ndims</dt>
<dd><p>Number of dimensions to use in dimensionality reduction (recommended to use the
same as number of factors) (default 40).</p></dd>


<dt>k</dt>
<dd><p>Number of nearest neighbors to use in calculating Jaccard index (default 15).</p></dd>


<dt>use.aligned</dt>
<dd><p>Whether to use quantile aligned or unaligned cell factor loadings (default
TRUE).</p></dd>


<dt>rand.seed</dt>
<dd><p>Random seed for reproducibility (default 42).</p></dd>


<dt>by.dataset</dt>
<dd><p>Return agreement calculated for each dataset (default FALSE).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Agreement metric (or vector of agreement per dataset).</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="rliger2-deprecated.html">rliger2-deprecated</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># ligerex (liger object based on in-memory datasets), factorization complete</span></span></span>
<span class="r-in"><span><span class="co"># generate H.norm by quantile normalizig factor loadings</span></span></span>
<span class="r-in"><span><span class="va">ligerex</span> <span class="op">&lt;-</span> <span class="fu"><a href="rliger2-deprecated.html">quantile_norm</a></span><span class="op">(</span><span class="va">ligerex</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">agreement</span> <span class="op">&lt;-</span> <span class="fu"><a href="rliger2-deprecated.html">calcAgreement</a></span><span class="op">(</span><span class="va">ligerex</span>, dr.method <span class="op">=</span> <span class="st">"NMF"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># ligerex (liger object based on datasets in HDF5 format), factorization complete</span></span></span>
<span class="r-in"><span><span class="va">ligerex</span> <span class="op">&lt;-</span> <span class="fu"><a href="rliger2-deprecated.html">quantile_norm</a></span><span class="op">(</span><span class="va">ligerex</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ligerex</span> <span class="op">&lt;-</span> <span class="fu"><a href="readSubset.html">readSubset</a></span><span class="op">(</span><span class="va">ligerex</span>, slot.use <span class="op">=</span> <span class="st">"scale.data"</span>, max.cells <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">agreement</span> <span class="op">&lt;-</span> <span class="fu"><a href="rliger2-deprecated.html">calcAgreement</a></span><span class="op">(</span><span class="va">ligerex</span>, dr.method <span class="op">=</span> <span class="st">"NMF"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

