<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This process treats the factor loading of each dataset as the low dimensional
embedding as well as the cluster assignment probability, i.e. the soft
clustering result. Then the method aligns the embedding by linearly moving
the centroids of the same cluster but within each dataset towards each other.
ATTENTION: This method is still under development while has shown
encouraging results in benchmarking tests. The arguments and their default
values reflect the best scored parameters in the tests and some of them may
be subject to change in the future."><title> Align factor loading by centroid alignment (beta) — centroidAlign • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content=" Align factor loading by centroid alignment (beta) — centroidAlign"><meta property="og:description" content="This process treats the factor loading of each dataset as the low dimensional
embedding as well as the cluster assignment probability, i.e. the soft
clustering result. Then the method aligns the embedding by linearly moving
the centroids of the same cluster but within each dataset towards each other.
ATTENTION: This method is still under development while has shown
encouraging results in benchmarking tests. The arguments and their default
values reflect the best scored parameters in the tests and some of them may
be subject to change in the future."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    <a class="dropdown-item" href="../articles/benchmark.html">Benchmarking integration quality</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="figures/lifecycle-experimental.svg" alt="[Experimental]"></a> Align factor loading by centroid alignment (beta)</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/integration.R" class="external-link"><code>R/integration.R</code></a></small>
      <div class="d-none name"><code>centroidAlign.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This process treats the factor loading of each dataset as the low dimensional
embedding as well as the cluster assignment probability, i.e. the soft
clustering result. Then the method aligns the embedding by linearly moving
the centroids of the same cluster but within each dataset towards each other.</p>
<p><b>ATTENTION: This method is still under development while has shown
encouraging results in benchmarking tests. The arguments and their default
values reflect the best scored parameters in the tests and some of them may
be subject to change in the future.</b></p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">centroidAlign</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for liger</span></span>
<span><span class="fu">centroidAlign</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  lambda <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  useDims <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scaleEmb <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  centerEmb <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scaleCluster <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  centerCluster <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  shift <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  diagnosis <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for Seurat</span></span>
<span><span class="fu">centroidAlign</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  reduction <span class="op">=</span> <span class="st">"inmf"</span>,</span>
<span>  lambda <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  useDims <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scaleEmb <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  centerEmb <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scaleCluster <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  centerCluster <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  shift <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  diagnosis <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> or Seurat object with valid factorization
result available (i.e. <code><a href="runIntegration.html">runIntegration</a></code> performed in advance).</p></dd>


<dt>...</dt>
<dd><p>Arguments passed to other S3 methods of this function.</p></dd>


<dt>lambda</dt>
<dd><p>Ridge regression penalty applied to each dataset. Can be one
number that applies to all datasets, or a numeric vector with length equal to
the number of datasets. Default <code>1</code>.</p></dd>


<dt>useDims</dt>
<dd><p>Indices of factors to use considered for the alignment.
Default <code>NULL</code> uses all factors.</p></dd>


<dt>scaleEmb</dt>
<dd><p>Logical, whether to scale the factor loading being considered
as the embedding. Default <code>TRUE</code>.</p></dd>


<dt>centerEmb</dt>
<dd><p>Logical, whether to center the factor loading being
considered as the embedding before scaling it. Default <code>TRUE</code>.</p></dd>


<dt>scaleCluster</dt>
<dd><p>Logical, whether to scale the factor loading being
considered as the cluster assignment probability. Default <code>FALSE</code>.</p></dd>


<dt>centerCluster</dt>
<dd><p>Logical, whether to center the factor loading being
considered as the cluster assignment probability before scaling it. Default
<code>FALSE</code>.</p></dd>


<dt>shift</dt>
<dd><p>Logical, whether to shift the factor loading being considered as
the cluster assignment probability after centered scaling. Default
<code>FALSE</code>.</p></dd>


<dt>diagnosis</dt>
<dd><p>Logical, whether to return cell metadata variables with
diagnostic information. See Details. Default <code>FALSE</code>.</p></dd>


<dt>reduction</dt>
<dd><p>Name of the reduction where LIGER integration result is
stored. Default <code>"inmf"</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Returns the updated input object</p><ul><li><p>liger method
</p><ul><li><p>Update the <code>H.norm</code> slot for the aligned cell factor
loading, ready for running graph based community detection clustering
or dimensionality reduction for visualization.</p></li>
<li><p>Update the <code>cellMata</code> slot with diagnostic information if
<code>diagnosis = TRUE</code>.</p></li>
</ul></li>
<li><p>Seurat method
</p><ul><li><p>Update the <code>reductions</code> slot with a new <code>DimReduc</code>
object containing the aligned cell factor loading.</p></li>
<li><p>Update the metadata with diagnostic information if
<code>diagnosis = TRUE</code>.</p></li>
</ul></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Diagnostic information include:</p>
<ul><li><p>object$raw_which.max: The index of the factor with the maximum value
in the raw factor loading.</p></li>
<li><p>object$R_which.max: The index of the factor with the maximum value in
the soft clustering probability matrix used for correction.</p></li>
<li><p>object$Z_which.max: The index of the factor with the maximum value in
the aligned factor loading.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">centroidAlign</span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

