<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='For convenience, the default formatType = "10x" directly fits the
structure of cellranger output. formatType = "anndata" works for
current AnnData H5AD file specification (see Details). If a customized H5
file structure is presented, any of the rawData,
indicesName, indptrName, genesName, barcodesName
should be specified accordingly to override the formatType preset.
DO make a copy of the H5AD files because rliger functions write to
the files and they will not be able to be read back to Python. This will be
fixed in the future.'><title>Create on-disk ligerDataset Object — createH5LigerDataset • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Create on-disk ligerDataset Object — createH5LigerDataset"><meta property="og:description" content='For convenience, the default formatType = "10x" directly fits the
structure of cellranger output. formatType = "anndata" works for
current AnnData H5AD file specification (see Details). If a customized H5
file structure is presented, any of the rawData,
indicesName, indptrName, genesName, barcodesName
should be specified accordingly to override the formatType preset.
DO make a copy of the H5AD files because rliger functions write to
the files and they will not be able to be read back to Python. This will be
fixed in the future.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Create on-disk ligerDataset Object</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/import.R" class="external-link"><code>R/import.R</code></a></small>
      <div class="d-none name"><code>createH5LigerDataset.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>For convenience, the default <code>formatType = "10x"</code> directly fits the
structure of cellranger output. <code>formatType = "anndata"</code> works for
current AnnData H5AD file specification (see Details). If a customized H5
file structure is presented, any of the <code>rawData</code>,
<code>indicesName</code>, <code>indptrName</code>, <code>genesName</code>, <code>barcodesName</code>
should be specified accordingly to override the <code>formatType</code> preset.</p>
<p><b>DO</b> make a copy of the H5AD files because rliger functions write to
the files and they will not be able to be read back to Python. This will be
fixed in the future.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">createH5LigerDataset</span><span class="op">(</span></span>
<span>  <span class="va">h5file</span>,</span>
<span>  formatType <span class="op">=</span> <span class="st">"10x"</span>,</span>
<span>  rawData <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normData <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scaleData <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  barcodesName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genesName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  indicesName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  indptrName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  anndataX <span class="op">=</span> <span class="st">"X"</span>,</span>
<span>  modal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"default"</span>, <span class="st">"rna"</span>, <span class="st">"atac"</span>, <span class="st">"spatial"</span>, <span class="st">"meth"</span><span class="op">)</span>,</span>
<span>  featureMeta <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>h5file</dt>
<dd><p>Filename of an H5 file</p></dd>


<dt>formatType</dt>
<dd><p>Select preset of H5 file structure. Default <code>"10X"</code>.
Alternatively, we also support <code>"anndata"</code> for H5AD files.</p></dd>


<dt>rawData, indicesName, indptrName</dt>
<dd><p>The path in a H5 file for the raw
sparse matrix data. These three types of data stands for the <code>x</code>,
<code>i</code>, and <code>p</code> slots of a <code><a href="https://rdrr.io/pkg/Matrix/man/dgCMatrix-class.html" class="external-link">dgCMatrix-class</a></code>
object. Default <code>NULL</code> uses <code>formatType</code> preset.</p></dd>


<dt>normData</dt>
<dd><p>The path in a H5 file for the "x" vector of the normalized
sparse matrix. Default <code>NULL</code>.</p></dd>


<dt>scaleData</dt>
<dd><p>The path in a H5 file for the Group that contains the sparse
matrix constructing information for the scaled data. Default <code>NULL</code>.</p></dd>


<dt>genesName, barcodesName</dt>
<dd><p>The path in a H5 file for the gene names and
cell barcodes. Default <code>NULL</code> uses <code>formatType</code> preset.</p></dd>


<dt>anndataX</dt>
<dd><p>The HDF5 path to the raw count data in an H5AD file. See
Details. Default <code>"X"</code>.</p></dd>


<dt>modal</dt>
<dd><p>Name of modality for this dataset. Currently options of
<code>"default"</code>, <code>"rna"</code>, <code>"atac"</code>, <code>"spatial"</code> and
<code>"meth"</code> are supported. Default <code>"default"</code>.</p></dd>


<dt>featureMeta</dt>
<dd><p>Data frame for feature metadata. Default <code>NULL</code>.</p></dd>


<dt>...</dt>
<dd><p>Additional slot data. See <a href="ligerDataset-class.html">ligerDataset</a> for detail.
Given values will be directly placed at corresponding slots.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>H5-based <a href="ligerDataset-class.html">ligerDataset</a> object</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>For H5AD file written from an AnnData object, we allow using
<code>formatType = "anndata"</code> for the function to infer the proper structure.
However, while a typical AnnData-based analysis tends to in-place update the
<code>adata.X</code> attribute and there is no standard/forced convention for where
the raw count data, as needed from LIGER, is stored. Therefore, we expose
argument <code>anndataX</code> for specifying this information. The default value
<code>"X"</code> looks for <code>adata.X</code>. If the raw data is stored in a layer,
e.g. <code>adata.layers['count']</code>, then <code>anndataX = "layers/count"</code>.
If it is stored to <code>adata.raw.X</code>, then <code>anndataX = "raw/X"</code>. If
your AnnData object does not have the raw count retained, you will have to
go back to the Python work flow to have it inserted at desired object space
and re-write the H5AD file, or just go from upstream source files with which
the AnnData was originally created.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">h5Path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/ctrl.h5"</span>, package <span class="op">=</span> <span class="st">"rliger"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tempPath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".h5"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">h5Path</span>, to <span class="op">=</span> <span class="va">tempPath</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span><span class="va">ld</span> <span class="op">&lt;-</span> <span class="fu">createH5LigerDataset</span><span class="op">(</span><span class="va">tempPath</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "H5D" in cache; using the first, from namespace 'rliger'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘hdf5r’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "H5D" in cache; using the first, from namespace 'rliger'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘hdf5r’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "H5D" in cache; using the first, from namespace 'rliger'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘hdf5r’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "H5D" in cache; using the first, from namespace 'rliger'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘hdf5r’</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

