<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>General heatmap plotting with prepared matrix and data.frames — .plotHeatmap • rliger</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="General heatmap plotting with prepared matrix and data.frames — .plotHeatmap"><meta property="og:description" content="This is not an exported function. This documentation just
serves for a manual of extra arguments that users can use when generating
heatmaps with plotGeneHeatmap or
plotFactorHeatmap.
Note that the following arguments are pre-occupied by upstream wrappers so
users should not include them in a function call: dataMatrix,
dataName, cellDF, featureDF, cellSplitVar,
featureSplitVar.
The following arguments of Heatmap is occupied
by this function, so users should include them in a function call as well:
matrix, name, col, heatmap_legend_param,
top_annotation, column_title_gp, column_names_gp,
show_column_names, column_split, column_gap,
left_annotation, row_title_gp, row_names_gp,
show_row_names, row_split, row_gap."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/import_data.html">Import data</a>
    </li>
    <li>
      <a href="../articles/liger_object.html">Introduction to a liger object</a>
    </li>
    <li>
      <a href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://welch-lab.github.io/" class="external-link">Welch Lab</a>
</li>
<li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/LabWelch" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>General heatmap plotting with prepared matrix and data.frames</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/heatmap.R" class="external-link"><code>R/heatmap.R</code></a></small>
    <div class="hidden name"><code>dot-plotHeatmap.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This is not an exported function. This documentation just
serves for a manual of extra arguments that users can use when generating
heatmaps with <code><a href="plotHeatmap.html">plotGeneHeatmap</a></code> or
<code><a href="plotHeatmap.html">plotFactorHeatmap</a></code>.</p>
<p>Note that the following arguments are pre-occupied by upstream wrappers so
users should not include them in a function call: <code>dataMatrix</code>,
<code>dataName</code>, <code>cellDF</code>, <code>featureDF</code>, <code>cellSplitVar</code>,
<code>featureSplitVar</code>.</p>
<p>The following arguments of <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></code> is occupied
by this function, so users should include them in a function call as well:
<code>matrix</code>, <code>name</code>, <code>col</code>, <code>heatmap_legend_param</code>,
<code>top_annotation</code>, <code>column_title_gp</code>, <code>column_names_gp</code>,
<code>show_column_names</code>, <code>column_split</code>, <code>column_gap</code>,
<code>left_annotation</code>, <code>row_title_gp</code>, <code>row_names_gp</code>,
<code>show_row_names</code>, <code>row_split</code>, <code>row_gap</code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">.plotHeatmap</span><span class="op">(</span></span>
<span>  <span class="va">dataMatrix</span>,</span>
<span>  dataName <span class="op">=</span> <span class="st">"Value"</span>,</span>
<span>  cellDF <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  featureDF <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transpose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cellSplitVar <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  featureSplitVar <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dataScaleFunc <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  showCellLabel <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  showCellLegend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  showFeatureLabel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  showFeatureLegend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cellAnnColList <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  featureAnnColList <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  trim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>  baseSize <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  cellTextSize <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  featureTextSize <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cellTitleSize <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  featureTitleSize <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legendTextSize <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legendTitleSize <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  viridisOption <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>  viridisDirection <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>  RColorBrewerOption <span class="op">=</span> <span class="st">"RdBu"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>dataMatrix</dt>
<dd><p>Matrix object with features/factors as rows and cells as
columns.</p></dd>


<dt>dataName</dt>
<dd><p>Text for heatmap color bar title. Default <code>Value</code>.</p></dd>


<dt>cellDF</dt>
<dd><p>data.frame object. Number of rows must match with number of
columns of <code>dataMatrix</code>.</p></dd>


<dt>featureDF</dt>
<dd><p>data.frame object. Number of columns must match with number
of rows of <code>dataMatrix</code>.</p></dd>


<dt>transpose</dt>
<dd><p>Logical, whether to "rotate" the heatmap by 90 degrees so
that cell information is displayed by row. Default <code>FALSE</code>.</p></dd>


<dt>cellSplitVar, featureSplitVar</dt>
<dd><p>Subset columns of <code>cellDF</code> or
<code>featureDF</code>, respectively.</p></dd>


<dt>dataScaleFunc</dt>
<dd><p>A function object, applied to <code>dataMatrix</code>.</p></dd>


<dt>showCellLabel, showFeatureLabel</dt>
<dd><p>Logical, whether to show cell barcodes,
gene symbols or factor names. Default <code>TRUE</code> for gene/factors but
<code>FALSE</code> for cells.</p></dd>


<dt>showCellLegend, showFeatureLegend</dt>
<dd><p>Logical, whether to show cell or
feature legends. Default <code>TRUE</code>. Can be a scalar for overall control
or a vector matching with each given annotation variable.</p></dd>


<dt>cellAnnColList, featureAnnColList</dt>
<dd><p>List object, with each element a
named vector of R-interpretable color code. The names of the list elements
are used for matching the annotation variable names. The names of the colors
in the vectors are used for matching the levels of a variable (factor object,
categorical). Default <code>NULL</code> generates ggplot-flavor categorical colors.</p></dd>


<dt>scale</dt>
<dd><p>Logical, whether to take z-score to scale and center gene
expression. Applied after <code>dataScaleFunc</code>. Default <code>FALSE</code>.</p></dd>


<dt>trim</dt>
<dd><p>Numeric vector of two values. Limit the z-score value into this
range when <code>scale = TRUE</code>. Default <code>c(-2, 2)</code>.</p></dd>


<dt>baseSize</dt>
<dd><p>One-parameter control of all text sizes. Individual text
element sizes can be controlled by other size arguments. "Title" sizes are
2 points larger than "text" sizes when being controlled by this.</p></dd>


<dt>cellTextSize, featureTextSize, legendTextSize</dt>
<dd><p>Size of cell barcode
labels, gene/factor labels, or legend values. Default <code>NULL</code>.</p></dd>


<dt>cellTitleSize, featureTitleSize, legendTitleSize</dt>
<dd><p>Size of titles of the
cell slices, gene/factor slices, or the legends. Default <code>NULL</code>.</p></dd>


<dt>viridisOption, viridisDirection</dt>
<dd><p>See argument <code>option</code> and
<code>direction</code> of <code><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></code>. Default <code>"A"</code>
and <code>-1</code>.</p></dd>


<dt>RColorBrewerOption</dt>
<dd><p>When <code>scale = TRUE</code>, heatmap color will be
mapped with <code><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></code>. This is passed to
<code>name</code>. Default <code>"RdBu"</code>.</p></dd>


<dt>...</dt>
<dd><p>Additional arguments to be passed to
<code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p><code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapList-class.html" class="external-link">HeatmapList-class</a></code> object</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

