<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Find shared and dataset-specific markers — getFactorMarkers • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Find shared and dataset-specific markers — getFactorMarkers"><meta property="og:description" content="Applies various filters to genes on the shared (\(W\)) and
dataset-specific (\(V\)) components of the factorization, before selecting
those which load most significantly on each factor (in a shared or
dataset-specific way)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Import data</li>
    <li class="dropdown-header">Introduction to a liger object</li>
    <li class="dropdown-header">Interoperate with Seurat</li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Find shared and dataset-specific markers</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/wilcoxon.R" class="external-link"><code>R/wilcoxon.R</code></a></small>
    <div class="hidden name"><code>getFactorMarkers.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Applies various filters to genes on the shared (\(W\)) and
dataset-specific (\(V\)) components of the factorization, before selecting
those which load most significantly on each factor (in a shared or
dataset-specific way).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getFactorMarkers</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">dataset1</span>,</span>
<span>  <span class="va">dataset2</span>,</span>
<span>  factorShareThresh <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  datasetSpecificity <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  logFCThresh <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  pvalThresh <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  nGenes <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  printGenes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span>,</span>
<span>  factor.share.thresh <span class="op">=</span> <span class="va">factorShareThresh</span>,</span>
<span>  dataset.specificity <span class="op">=</span> <span class="va">datasetSpecificity</span>,</span>
<span>  log.fc.thresh <span class="op">=</span> <span class="va">logFCThresh</span>,</span>
<span>  pval.thresh <span class="op">=</span> <span class="va">pvalThresh</span>,</span>
<span>  num.genes <span class="op">=</span> <span class="va">nGenes</span>,</span>
<span>  print.genes <span class="op">=</span> <span class="va">printGenes</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p><a href="liger-class.html">liger</a> object with factorization results.</p></dd>


<dt>dataset1</dt>
<dd><p>Name of first dataset. Required.</p></dd>


<dt>dataset2</dt>
<dd><p>Name of second dataset. Required</p></dd>


<dt>factorShareThresh</dt>
<dd><p>Numeric. Only factors with a dataset specificity
less than or equal to this threshold will be used. Default <code>10</code>.</p></dd>


<dt>datasetSpecificity</dt>
<dd><p>Numeric vector. Pre-calculated dataset specificity
if available. Length should match number of all factors available. Default
<code>NULL</code> automatically calculates with
<code><a href="calcDatasetSpecificity.html">calcDatasetSpecificity</a></code>.</p></dd>


<dt>logFCThresh</dt>
<dd><p>Numeric. Lower log-fold change threshold for differential
expression in markers. Default <code>1</code>.</p></dd>


<dt>pvalThresh</dt>
<dd><p>Numeric. Upper p-value threshold for Wilcoxon rank test for
gene expression. Default <code>0.05</code>.</p></dd>


<dt>nGenes</dt>
<dd><p>Integer. Max number of genes to report for each dataset.
Default <code>30</code>.</p></dd>


<dt>printGenes</dt>
<dd><p>Logical. Whether to print ordered markers passing logFC,
UMI and frac thresholds, when <code>verbose = TRUE</code>. Default <code>FALSE</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> which is <code>TRUE</code> if users have not set.</p></dd>


<dt>factor.share.thresh, dataset.specificity, log.fc.thresh, pval.thresh, num.genes, print.genes</dt>
<dd><p><b>Deprecated</b>. See Usage section for replacement.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list object consisting of the following entries:</p>
<dl><dt>[value of `dataset1`]</dt>
<dd><p>data.frame of dataset1-specific markers</p></dd>

<dt>shared</dt>
<dd><p>data.frame of shared markers</p></dd>

<dt>[value of `dataset1`]</dt>
<dd><p>data.frame of dataset2-specific markers</p></dd>

<dt>num_factors_V1</dt>
<dd><p>A frequency table indicating the number of factors each
marker appears, in dataset1</p></dd>

<dt>num_factors_V2</dt>
<dd><p>A frequency table indicating the number of factors each
marker appears, in dataset2</p></dd>

</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">getFactorMarkers</span><span class="op">(</span><span class="va">pbmcPlot</span>, dataset1 <span class="op">=</span> <span class="st">"ctrl"</span>, dataset2 <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 17:13:10 2023 ... Performing wilcoxon test between datasets "ctrl" and "stim", </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>                              basing on factor loading.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                                              |                                                                      |   0%  |                                                                              |===================================                                   |  50%  |                                                                              |======================================================================| 100%</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "list"</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ctrl"           "shared"         "stim"           "num_factors_V1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5] "num_factors_V2"</span>
<span class="r-in"><span><span class="va">result</span><span class="op">$</span><span class="va">shared</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">factor_num</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html" class="external-link">top_n</a></span><span class="op">(</span><span class="fl">2</span>, <span class="va">logFC</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 4 × 4</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># Groups:   factor_num [2]</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   feature factor_num logFC     pval</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> GPX1             1  3.05 8.08<span style="color: #949494;">e</span><span style="color: #BB0000;">-50</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> RPS8             1  2.99 3.39<span style="color: #949494;">e</span><span style="color: #BB0000;">- 4</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> CXCR4            2  2.09 8.41<span style="color: #949494;">e</span><span style="color: #BB0000;">- 4</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> S100A6           2  3.40 1.52<span style="color: #949494;">e</span><span style="color: #BB0000;">- 6</span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

