<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/import_data.html">Import data</a>
    </li>
    <li>
      <a href="../articles/liger_object.html">Introduction to a liger object</a>
    </li>
    <li>
      <a href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="read-data">Read Data <a href="#read-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="createH5LigerDataset.html">createH5LigerDataset()</a></code> </p>
        </td>
        <td><p>Create on-disk ligerDataset Object</p></td>
      </tr><tr><td>
          <p><code><a href="createLiger.html">createLiger()</a></code> </p>
        </td>
        <td><p>Create liger object</p></td>
      </tr><tr><td>
          <p><code><a href="createLigerDataset.html">createLigerDataset()</a></code> </p>
        </td>
        <td><p>Create in-memory ligerDataset object</p></td>
      </tr><tr><td>
          <p><code><a href="readLiger.html">readLiger()</a></code> </p>
        </td>
        <td><p>Read liger object from RDS file</p></td>
      </tr><tr><td>
          <p><code><a href="read10X.html">read10X()</a></code> <code><a href="read10X.html">read10XRNA()</a></code> <code><a href="read10X.html">read10XATAC()</a></code> </p>
        </td>
        <td><p>Load in data from 10X</p></td>
      </tr><tr><td>
          <p><code><a href="restoreH5Liger.html">restoreH5Liger()</a></code> <code><a href="restoreH5Liger.html">restoreOnlineLiger()</a></code> </p>
        </td>
        <td><p>Restore links (to HDF5 files) for reloaded liger/ligerDataset object</p></td>
      </tr><tr><td>
          <p><code><a href="importVignetteData.html">importVignetteData()</a></code> </p>
        </td>
        <td><p>Import prepared dataset publically available</p></td>
      </tr><tr><td>
          <p><code><a href="as.ligerDataset.html">as.ligerDataset()</a></code> </p>
        </td>
        <td><p>Converting other classes of data to a as.ligerDataset object</p></td>
      </tr><tr><td>
          <p><code><a href="as.liger.html">as.liger()</a></code> </p>
        </td>
        <td><p>Converting other classes of data to a liger object</p></td>
      </tr><tr><td>
          <p><code><a href="pbmc.html">pbmc</a></code> </p>
        </td>
        <td><p>liger object of PBMC subsample data with Control and Stimulated datasets</p></td>
      </tr><tr><td>
          <p><code><a href="pbmcPlot.html">pbmcPlot</a></code> </p>
        </td>
        <td><p>liger object of PBMC subsample data with plotting information available</p></td>
      </tr><tr><td>
          <p><code><a href="bmmc.html">bmmc</a></code> </p>
        </td>
        <td><p>liger object of bone marrow subsample data with RNA and ATAC modality</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="preprocessing">Preprocessing <a href="#preprocessing" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="runGeneralQC.html">runGeneralQC()</a></code> </p>
        </td>
        <td><p>General QC for liger object</p></td>
      </tr><tr><td>
          <p><code><a href="removeMissing.html">removeMissing()</a></code> <code><a href="removeMissing.html">removeMissingObs()</a></code> </p>
        </td>
        <td><p>Remove missing cells or features from liger object</p></td>
      </tr><tr><td>
          <p><code><a href="normalize.html">normalize()</a></code> <code><a href="normalize.html">normalizePeak()</a></code> </p>
        </td>
        <td><p>Normalize raw counts data</p></td>
      </tr><tr><td>
          <p><code><a href="selectGenes.html">selectGenes()</a></code> </p>
        </td>
        <td><p>Select a subset of informative genes</p></td>
      </tr><tr><td>
          <p><code><a href="scaleNotCenter.html">scaleNotCenter()</a></code> </p>
        </td>
        <td><p>Scale genes by root-mean-square across cells</p></td>
      </tr><tr><td>
          <p><code><a href="reverseMethData.html">reverseMethData()</a></code> </p>
        </td>
        <td><p>Create "scaled data" for DNA methylation datasets</p></td>
      </tr><tr><td>
          <p><code><a href="downsample.html">downsample()</a></code> </p>
        </td>
        <td><p>Downsample datasets</p></td>
      </tr><tr><td>
          <p><code><a href="selectGenesVST.html">selectGenesVST()</a></code> </p>
        </td>
        <td><p>Select variable gene from one dataset with VST method</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="integration">Integration <a href="#integration" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="runIntegration.html">runIntegration()</a></code> </p>
        </td>
        <td><p>Integrate scaled datasets with iNMF or variant methods</p></td>
      </tr><tr><td>
          <p><code><a href="quantileNorm.html">quantileNorm()</a></code> </p>
        </td>
        <td><p>Quantile Align (Normalize) Factor Loadings</p></td>
      </tr><tr><td>
          <p><code><a href="runINMF.html">runINMF()</a></code> </p>
        </td>
        <td><p>Perform iNMF on scaled datasets</p></td>
      </tr><tr><td>
          <p><code><a href="runOnlineINMF.html">runOnlineINMF()</a></code> </p>
        </td>
        <td><p>Perform online iNMF on scaled datasets</p></td>
      </tr><tr><td>
          <p><code><a href="runUINMF.html">runUINMF()</a></code> </p>
        </td>
        <td><p>Perform Mosaic iNMF (UINMF) on scaled datasets with unshared features</p></td>
      </tr><tr><td>
          <p><code><a href="optimizeNewData.html">optimizeNewData()</a></code> </p>
        </td>
        <td><p>Perform factorization for new data</p></td>
      </tr><tr><td>
          <p><code><a href="optimizeNewK.html">optimizeNewK()</a></code> </p>
        </td>
        <td><p>Perform factorization for new value of k</p></td>
      </tr><tr><td>
          <p><code><a href="optimizeNewLambda.html">optimizeNewLambda()</a></code> </p>
        </td>
        <td><p>Perform factorization for new lambda value</p></td>
      </tr><tr><td>
          <p><code><a href="optimizeSubset.html">optimizeSubset()</a></code> </p>
        </td>
        <td><p>Perform factorization for subset of data</p></td>
      </tr><tr><td>
          <p><code><a href="suggestK.html">suggestK()</a></code> </p>
        </td>
        <td><p>Visually suggest appropiate k value</p></td>
      </tr><tr><td>
          <p><code><a href="suggestLambda.html">suggestLambda()</a></code> </p>
        </td>
        <td><p>Visually suggest appropriate lambda value</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="downstream-analysis">Downstream Analysis <a href="#downstream-analysis" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="runUMAP.html">runUMAP()</a></code> </p>
        </td>
        <td><p>Perform UMAP Dimensionality Reduction</p></td>
      </tr><tr><td>
          <p><code><a href="runTSNE.html">runTSNE()</a></code> </p>
        </td>
        <td><p>Perform t-SNE dimensionality reduction</p></td>
      </tr><tr><td>
          <p><code><a href="runCluster.html">runCluster()</a></code> </p>
        </td>
        <td><p>SNN Graph Based Community Detection</p></td>
      </tr><tr><td>
          <p><code><a href="liger-DEG.html">runPairwiseDEG()</a></code> <code><a href="liger-DEG.html">runMarkerDEG()</a></code> </p>
        </td>
        <td><p>Find DEG between two groups</p></td>
      </tr><tr><td>
          <p><code><a href="runWilcoxon.html">runWilcoxon()</a></code> </p>
        </td>
        <td><p>Perform Wilcoxon rank-sum test</p></td>
      </tr><tr><td>
          <p><code><a href="runPseudoBulkDEG.html">runPseudoBulkDEG()</a></code> </p>
        </td>
        <td><p>Run DESeq2 differential expression test on pseudo-bulk expression</p></td>
      </tr><tr><td>
          <p><code><a href="getFactorMarkers.html">getFactorMarkers()</a></code> </p>
        </td>
        <td><p>Find shared and dataset-specific markers</p></td>
      </tr><tr><td>
          <p><code><a href="runDoubletFinder.html">runDoubletFinder()</a></code> </p>
        </td>
        <td><p>Doublet detection with DoubletFinder</p></td>
      </tr><tr><td>
          <p><code><a href="runGSEA.html">runGSEA()</a></code> </p>
        </td>
        <td><p>Analyze biological interpretations of metagene</p></td>
      </tr><tr><td>
          <p><code><a href="runGOEnrich.html">runGOEnrich()</a></code> </p>
        </td>
        <td><p>Run Gene Ontology enrichment analysis on metagenes</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="atac-specific-analysis">ATAC Specific Analysis <a href="#atac-specific-analysis" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="imputeKNN.html">imputeKNN()</a></code> </p>
        </td>
        <td><p>Impute the query cell expression matrix</p></td>
      </tr><tr><td>
          <p><code><a href="linkGenesAndPeaks.html">linkGenesAndPeaks()</a></code> </p>
        </td>
        <td><p>Linking genes to putative regulatory elements</p></td>
      </tr><tr><td>
          <p><code><a href="exportInteractTrack.html">exportInteractTrack()</a></code> </p>
        </td>
        <td><p>Export predicted gene-pair interaction</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="visualization">Visualization <a href="#visualization" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotCellScatter.html">plotCellScatter()</a></code> </p>
        </td>
        <td><p>Generate scatter plot(s) using liger object</p></td>
      </tr><tr><td>
          <p><code><a href="plotCellViolin.html">plotCellViolin()</a></code> </p>
        </td>
        <td><p>Generate violin/box plot(s) using liger object</p></td>
      </tr><tr><td>
          <p><code><a href="plotClusterFactorDot.html">plotClusterFactorDot()</a></code> </p>
        </td>
        <td><p>Make dot plot of factor loading in cell groups</p></td>
      </tr><tr><td>
          <p><code><a href="plotClusterGeneDot.html">plotClusterGeneDot()</a></code> </p>
        </td>
        <td><p>Make dot plot of gene expression in cell groups</p></td>
      </tr><tr><td>
          <p><code><a href="plotDensityDimRed.html">plotDensityDimRed()</a></code> </p>
        </td>
        <td><p>Create density plot basing on specified coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="plotDimRed.html">plotClusterDimRed()</a></code> <code><a href="plotDimRed.html">plotDatasetDimRed()</a></code> <code><a href="plotDimRed.html">plotByDatasetAndCluster()</a></code> <code><a href="plotDimRed.html">plotGeneDimRed()</a></code> <code><a href="plotDimRed.html">plotPeakDimRed()</a></code> <code><a href="plotDimRed.html">plotFactorDimRed()</a></code> </p>
        </td>
        <td><p>Generate Dimensionality Reduction Plot with Coloring</p></td>
      </tr><tr><td>
          <p><code><a href="plotGeneLoadings.html">plotGeneLoadings()</a></code> <code><a href="plotGeneLoadings.html">plotGeneLoadingRank()</a></code> </p>
        </td>
        <td><p>Visualize factor expression and gene loading</p></td>
      </tr><tr><td>
          <p><code><a href="plotHeatmap.html">plotGeneHeatmap()</a></code> <code><a href="plotHeatmap.html">plotFactorHeatmap()</a></code> </p>
        </td>
        <td><p>Plot Heatmap of Gene Expression or Factor Loading</p></td>
      </tr><tr><td>
          <p><code><a href="plotProportion.html">plotProportion()</a></code> <code><a href="plotProportion.html">plotProportionDot()</a></code> <code><a href="plotProportion.html">plotProportionBar()</a></code> <code><a href="plotProportion.html">plotClusterProportions()</a></code> <code><a href="plotProportion.html">plotProportionPie()</a></code> </p>
        </td>
        <td><p>Visualize proportion across two categorical variables</p></td>
      </tr><tr><td>
          <p><code><a href="plotSuggestK.html">plotSuggestK()</a></code> </p>
        </td>
        <td><p>Visualize suggestK result statistics</p></td>
      </tr><tr><td>
          <p><code><a href="plotSuggestLambda.html">plotSuggestLambda()</a></code> </p>
        </td>
        <td><p>Visualize suggestLambda result statistics</p></td>
      </tr><tr><td>
          <p><code><a href="plotVarFeatures.html">plotVarFeatures()</a></code> </p>
        </td>
        <td><p>Plot the variance vs mean of feature expression</p></td>
      </tr><tr><td>
          <p><code><a href="plotViolin.html">plotGeneViolin()</a></code> <code><a href="plotViolin.html">plotTotalCountViolin()</a></code> <code><a href="plotViolin.html">plotGeneDetectedViolin()</a></code> </p>
        </td>
        <td><p>Visualize gene expression or cell metadata with violin plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotVolcano.html">plotVolcano()</a></code> <code><a href="plotVolcano.html">plotEnhancedVolcano()</a></code> </p>
        </td>
        <td><p>Create volcano plot for Wilcoxon test result</p></td>
      </tr><tr><td>
          <p><code><a href="plotWordClouds.html">plotWordClouds()</a></code> </p>
        </td>
        <td><p>Generate word clouds and t-SNE plots</p></td>
      </tr><tr><td>
          <p><code><a href="makeRiverplot.html">makeRiverplot()</a></code> </p>
        </td>
        <td><p>Generate a river (Sankey) plot</p></td>
      </tr><tr><td>
          <p><code><a href="dot-ggCellViolin.html">.ggCellViolin()</a></code> </p>
        </td>
        <td><p>Produce single violin plot with data frame passed from upstream</p></td>
      </tr><tr><td>
          <p><code><a href="dot-ggScatter.html">.ggScatter()</a></code> </p>
        </td>
        <td><p>Produce single scatter plot with data frame passed from upstream</p></td>
      </tr><tr><td>
          <p><code><a href="dot-ggplotLigerTheme.html">.ggplotLigerTheme()</a></code> </p>
        </td>
        <td><p>Generic ggplot theme setting for rliger package</p></td>
      </tr><tr><td>
          <p><code><a href="dot-plotHeatmap.html">.plotHeatmap()</a></code> </p>
        </td>
        <td><p>General heatmap plotting with prepared matrix and data.frames</p></td>
      </tr><tr><td>
          <p><code><a href="dot-complexHeatmapDotPlot.html">.complexHeatmapDotPlot()</a></code> </p>
        </td>
        <td><p>Generate dot plot from input matrix with ComplexHeatmap</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="evaluation-metrics">Evaluation Metrics <a href="#evaluation-metrics" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="calcARI-deprecated.html">calcARI-deprecated</a></code> </p>
        </td>
        <td><p>[Deprecated] Calculate adjusted Rand index</p></td>
      </tr><tr><td>
          <p><code><a href="calcAgreement-deprecated.html">calcAgreement-deprecated</a></code> </p>
        </td>
        <td><p>[Deprecated] Calculate agreement metric</p></td>
      </tr><tr><td>
          <p><code><a href="calcAlignment.html">calcAlignment()</a></code> </p>
        </td>
        <td><p>Calculate alignment metric</p></td>
      </tr><tr><td>
          <p><code><a href="calcAlignmentPerCluster-deprecated.html">calcAlignmentPerCluster-deprecated</a></code> </p>
        </td>
        <td><p>[Deprecated] Calculate alignment for each cluster</p></td>
      </tr><tr><td>
          <p><code><a href="calcDatasetSpecificity.html">calcDatasetSpecificity()</a></code> </p>
        </td>
        <td><p>Calculate a dataset-specificity score for each factor</p></td>
      </tr><tr><td>
          <p><code><a href="calcNormLoadings.html">calcNormLoadings()</a></code> </p>
        </td>
        <td><p>Calculate loadings for each factor</p></td>
      </tr><tr><td>
          <p><code><a href="calcPurity-deprecated.html">calcPurity-deprecated</a></code> </p>
        </td>
        <td><p>[Deprecated] Calculate purity</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="object-class">Object Class <a href="#object-class" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="liger-class.html">liger-class</a></code> <code><a href="liger-class.html">liger</a></code> <code><a href="liger-class.html">show,liger-method</a></code> <code><a href="liger-class.html">dim,liger-method</a></code> <code><a href="liger-class.html">dimnames,liger-method</a></code> <code><a href="liger-class.html">dimnames</a></code> <code><a href="liger-class.html">[,liger,character,missing,ANY-method</a></code> <code><a href="liger-class.html">[,liger,missing,index,ANY-method</a></code> <code><a href="liger-class.html">[,liger,character,index,ANY-method</a></code> <code><a href="liger-class.html">datasets</a></code> <code><a href="liger-class.html">datasets</a></code> <code><a href="liger-class.html">datasets,liger-method</a></code> <code><a href="liger-class.html">datasets</a></code> <code><a href="liger-class.html">datasets</a></code> <code><a href="liger-class.html">dataset</a></code> <code><a href="liger-class.html">dataset</a></code> <code><a href="liger-class.html">dataset,liger,character_OR_NULL-method</a></code> <code><a href="liger-class.html">dataset,liger,missing-method</a></code> <code><a href="liger-class.html">dataset,liger,numeric-method</a></code> <code><a href="liger-class.html">dataset</a></code> <code><a href="liger-class.html">dataset</a></code> <code><a href="liger-class.html">dataset</a></code> <code><a href="liger-class.html">names,liger-method</a></code> <code><a href="liger-class.html">names</a></code> <code><a href="liger-class.html">length,liger-method</a></code> <code><a href="liger-class.html">cellMeta</a></code> <code><a href="liger-class.html">cellMeta</a></code> <code><a href="liger-class.html">cellMeta,liger,NULL-method</a></code> <code><a href="liger-class.html">cellMeta,liger,character-method</a></code> <code><a href="liger-class.html">cellMeta,liger,missing-method</a></code> <code><a href="liger-class.html">cellMeta</a></code> <code><a href="liger-class.html">cellMeta</a></code> <code><a href="liger-class.html">[[,liger,ANY,missing-method</a></code> <code><a href="liger-class.html">[[</a></code> <code><a href="liger-class.html">$,liger-method</a></code> <code><a href="liger-class.html">$</a></code> <code><a href="liger-class.html">defaultCluster</a></code> <code><a href="liger-class.html">defaultCluster</a></code> <code><a href="liger-class.html">defaultCluster,liger-method</a></code> <code><a href="liger-class.html">defaultCluster</a></code> <code><a href="liger-class.html">defaultCluster</a></code> <code><a href="liger-class.html">defaultCluster</a></code> <code><a href="liger-class.html">dimRed</a></code> <code><a href="liger-class.html">dimRed</a></code> <code><a href="liger-class.html">dimRed,liger,missing-method</a></code> <code><a href="liger-class.html">dimRed,liger,character-method</a></code> <code><a href="liger-class.html">dimRed</a></code> <code><a href="liger-class.html">defaultDimRed</a></code> <code><a href="liger-class.html">defaultDimRed</a></code> <code><a href="liger-class.html">defaultDimRed,liger-method</a></code> <code><a href="liger-class.html">defaultDimRed</a></code> <code><a href="liger-class.html">defaultDimRed</a></code> <code><a href="liger-class.html">varFeatures</a></code> <code><a href="liger-class.html">varFeatures</a></code> <code><a href="liger-class.html">varFeatures,liger-method</a></code> <code><a href="liger-class.html">varFeatures</a></code> <code><a href="liger-class.html">fortify.liger</a></code> <code><a href="liger-class.html">c.liger</a></code> <code><a href="liger-class.html">commands</a></code> <code><a href="liger-class.html">commands,liger-method</a></code> <code><a href="liger-class.html">rawData,liger-method</a></code> <code><a href="liger-class.html">rawData</a></code> <code><a href="liger-class.html">rawData</a></code> <code><a href="liger-class.html">normData,liger-method</a></code> <code><a href="liger-class.html">normData</a></code> <code><a href="liger-class.html">normData</a></code> <code><a href="liger-class.html">scaleData,liger,ANY-method</a></code> <code><a href="liger-class.html">scaleData</a></code> <code><a href="liger-class.html">scaleData</a></code> <code><a href="liger-class.html">scaleData</a></code> <code><a href="liger-class.html">scaleUnsharedData,liger,character-method</a></code> <code><a href="liger-class.html">scaleUnsharedData,liger,numeric-method</a></code> <code><a href="liger-class.html">getMatrix,liger,ANY,ANY,ANY-method</a></code> <code><a href="liger-class.html">getH5File,liger,ANY-method</a></code> </p>
        </td>
        <td><p>liger class</p></td>
      </tr><tr><td>
          <p><code><a href="ligerATACDataset-class.html">ligerATACDataset-class</a></code> <code><a href="ligerATACDataset-class.html">ligerATACDataset</a></code> </p>
        </td>
        <td><p>Subclass of ligerDataset for ATAC modality</p></td>
      </tr><tr><td>
          <p><code><a href="ligerCommand-class.html">show(<i>&lt;ligerCommand&gt;</i>)</a></code> </p>
        </td>
        <td><p>ligerCommand object: Record the input and time of a LIGER function call</p></td>
      </tr><tr><td>
          <p><code><a href="ligerDataset-class.html">show(<i>&lt;ligerDataset&gt;</i>)</a></code> <code><a href="ligerDataset-class.html">dim(<i>&lt;ligerDataset&gt;</i>)</a></code> <code><a href="ligerDataset-class.html">dimnames(<i>&lt;ligerDataset&gt;</i>)</a></code> <code><a href="ligerDataset-class.html">`dimnames&lt;-`(<i>&lt;ligerDataset&gt;</i>,<i>&lt;list&gt;</i>)</a></code> <code><a href="ligerDataset-class.html">`[`(<i>&lt;ligerDataset&gt;</i>,<i>&lt;index&gt;</i>,<i>&lt;missing&gt;</i>,<i>&lt;ANY&gt;</i>)</a></code> <code><a href="ligerDataset-class.html">`[`(<i>&lt;ligerDataset&gt;</i>,<i>&lt;missing&gt;</i>,<i>&lt;index&gt;</i>,<i>&lt;ANY&gt;</i>)</a></code> <code><a href="ligerDataset-class.html">`[`(<i>&lt;ligerDataset&gt;</i>,<i>&lt;index&gt;</i>,<i>&lt;index&gt;</i>,<i>&lt;ANY&gt;</i>)</a></code> <code><a href="ligerDataset-class.html">rawData()</a></code> <code><a href="ligerDataset-class.html">`rawData&lt;-`()</a></code> <code><a href="ligerDataset-class.html">normData()</a></code> <code><a href="ligerDataset-class.html">`normData&lt;-`()</a></code> <code><a href="ligerDataset-class.html">scaleData()</a></code> <code><a href="ligerDataset-class.html">`scaleData&lt;-`()</a></code> <code><a href="ligerDataset-class.html">scaleUnsharedData()</a></code> <code><a href="ligerDataset-class.html">`scaleUnsharedData&lt;-`()</a></code> <code><a href="ligerDataset-class.html">getMatrix()</a></code> <code><a href="ligerDataset-class.html">h5fileInfo()</a></code> <code><a href="ligerDataset-class.html">`h5fileInfo&lt;-`()</a></code> <code><a href="ligerDataset-class.html">getH5File()</a></code> <code><a href="ligerDataset-class.html">featureMeta()</a></code> <code><a href="ligerDataset-class.html">`featureMeta&lt;-`()</a></code> <code><a href="ligerDataset-class.html">cbind(<i>&lt;ligerDataset&gt;</i>)</a></code> </p>
        </td>
        <td><p>ligerDataset class</p></td>
      </tr><tr><td>
          <p><code><a href="ligerRNADataset-class.html">ligerRNADataset-class</a></code> <code><a href="ligerRNADataset-class.html">ligerRNADataset</a></code> </p>
        </td>
        <td><p>Subclass of ligerDataset for RNA modality</p></td>
      </tr><tr><td>
          <p><code><a href="peak.html">rawPeak()</a></code> <code><a href="peak.html">`rawPeak&lt;-`()</a></code> <code><a href="peak.html">normPeak()</a></code> <code><a href="peak.html">`normPeak&lt;-`()</a></code> </p>
        </td>
        <td><p>Access ligerATACDataset peak data</p></td>
      </tr><tr><td>
          <p><code><a href="retrieveCellFeature.html">retrieveCellFeature()</a></code> </p>
        </td>
        <td><p>Retrieve a single matrix of cells from a slot</p></td>
      </tr><tr><td>
          <p><code><a href="subsetLiger.html">subsetLiger()</a></code> </p>
        </td>
        <td><p>Subset liger object</p></td>
      </tr><tr><td>
          <p><code><a href="subsetLigerDataset.html">subsetLigerDataset()</a></code> <code><a href="subsetLigerDataset.html">subsetH5LigerDataset()</a></code> <code><a href="subsetLigerDataset.html">subsetMemLigerDataset()</a></code> </p>
        </td>
        <td><p>Subset ligerDataset object</p></td>
      </tr><tr><td>
          <p><code><a href="convertOldLiger.html">convertOldLiger()</a></code> </p>
        </td>
        <td><p>Convert old liger object to latest version</p></td>
      </tr><tr><td>
          <p><code><a href="ligerToSeurat.html">ligerToSeurat()</a></code> </p>
        </td>
        <td><p>Convert between liger and Seurat object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="misc">Misc <a href="#misc" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="H5Apply.html">H5Apply()</a></code> </p>
        </td>
        <td><p>Apply function to chunks of H5 data in ligerDataset object</p></td>
      </tr><tr><td>
          <p><code><a href="commandDiff.html">commandDiff()</a></code> </p>
        </td>
        <td><p>Check difference of two liger command</p></td>
      </tr><tr><td>
          <p><code><a href="isH5Liger.html">isH5Liger()</a></code> </p>
        </td>
        <td><p>Check if a liger or ligerDataset object is made of HDF5 file</p></td>
      </tr><tr><td>
          <p><code><a href="makeFeatureMatrix.html">makeFeatureMatrix()</a></code> </p>
        </td>
        <td><p>Fast calculation of feature count matrix</p></td>
      </tr><tr><td>
          <p><code><a href="mergeH5.html">mergeH5()</a></code> </p>
        </td>
        <td><p>Merge hdf5 files</p></td>
      </tr><tr><td>
          <p><code><a href="mergeSparseAll.html">mergeSparseAll()</a></code> <code><a href="mergeSparseAll.html">mergeDenseAll()</a></code> </p>
        </td>
        <td><p>Merge matrices while keeping the union of rows</p></td>
      </tr><tr><td>
          <p><code><a href="modalOf.html">modalOf()</a></code> </p>
        </td>
        <td><p>Return preset modality of a ligerDataset object or that of all datasets in a
liger object</p></td>
      </tr><tr><td>
          <p><code><a href="closeAllH5.html">closeAllH5()</a></code> </p>
        </td>
        <td><p>Close all links (to HDF5 files) of a liger object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="deprecated">Deprecated <a href="#deprecated" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="calcARI-deprecated.html">calcARI-deprecated</a></code> </p>
        </td>
        <td><p>[Deprecated] Calculate adjusted Rand index</p></td>
      </tr><tr><td>
          <p><code><a href="calcAgreement-deprecated.html">calcAgreement-deprecated</a></code> </p>
        </td>
        <td><p>[Deprecated] Calculate agreement metric</p></td>
      </tr><tr><td>
          <p><code><a href="calcAlignmentPerCluster-deprecated.html">calcAlignmentPerCluster-deprecated</a></code> </p>
        </td>
        <td><p>[Deprecated] Calculate alignment for each cluster</p></td>
      </tr><tr><td>
          <p><code><a href="calcPurity-deprecated.html">calcPurity-deprecated</a></code> </p>
        </td>
        <td><p>[Deprecated] Calculate purity</p></td>
      </tr><tr><td>
          <p><code><a href="louvainCluster-deprecated.html">louvainCluster-deprecated</a></code> </p>
        </td>
        <td><p>[Deprecated] Louvain algorithm for community detection</p></td>
      </tr><tr><td>
          <p><code><a href="makeInteractTrack-deprecated.html">makeInteractTrack-deprecated</a></code> </p>
        </td>
        <td><p>[Deprecated] Export predicted gene-pair interaction</p></td>
      </tr><tr><td>
          <p><code><a href="online_iNMF-deprecated.html">online_iNMF-deprecated</a></code> </p>
        </td>
        <td><p>[Deprecated] Perform online iNMF on scaled datasets</p></td>
      </tr><tr><td>
          <p><code><a href="optimizeALS-deprecated.html">optimizeALS-deprecated</a></code> </p>
        </td>
        <td><p>[Deprecated] Perform iNMF on scaled datasets</p></td>
      </tr><tr><td>
          <p><code><a href="quantile_norm-deprecated.html">quantile_norm-deprecated</a></code> </p>
        </td>
        <td><p>[Deprecated] Quantile align (normalize) factor loading</p></td>
      </tr><tr><td>
          <p><code><a href="readSubset.html">readSubset()</a></code> </p>
        </td>
        <td><p>[Deprecated] See <code><a href="https://rdrr.io/pkg/rliger2/man/downsample.html" class="external-link">downsample</a></code></p></td>
      </tr><tr><td>
          <p><code><a href="quantileAlignSNF.html">quantileAlignSNF()</a></code> </p>
        </td>
        <td><p>Quantile align (normalize) factor loadings</p></td>
      </tr><tr><td>
          <p><code><a href="getProportionMito.html">getProportionMito()</a></code> </p>
        </td>
        <td><p>Calculate proportion mitochondrial contribution</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

