<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.8/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="read-data">Read Data<a class="anchor" aria-label="anchor" href="#read-data"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="createH5LigerDataset.html">createH5LigerDataset()</a></code> 
        </dt>
        <dd>Create on-disk ligerDataset Object</dd>
      </dl><dl><dt>
          
          <code><a href="createLiger.html">createLiger()</a></code> 
        </dt>
        <dd>Create liger object</dd>
      </dl><dl><dt>
          
          <code><a href="createLigerDataset.html">createLigerDataset()</a></code> 
        </dt>
        <dd>Create in-memory ligerDataset object</dd>
      </dl><dl><dt>
          
          <code><a href="readLiger.html">readLiger()</a></code> 
        </dt>
        <dd>Read liger object from RDS file</dd>
      </dl><dl><dt>
          
          <code><a href="read10X.html">read10X()</a></code> <code><a href="read10X.html">read10XRNA()</a></code> <code><a href="read10X.html">read10XATAC()</a></code> 
        </dt>
        <dd>Load in data from 10X</dd>
      </dl><dl><dt>
          
          <code><a href="as.ligerDataset.html">as.ligerDataset()</a></code> 
        </dt>
        <dd>Converting other classes of data to a ligerDataset object</dd>
      </dl><dl><dt>
          
          <code><a href="as.liger.html">seuratToLiger()</a></code> <code><a href="as.liger.html">as.liger()</a></code> 
        </dt>
        <dd>Converting other classes of data to a liger object</dd>
      </dl></div><div class="section level2">
      <h2 id="example-data">Example Data<a class="anchor" aria-label="anchor" href="#example-data"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="importVignetteData.html">importPBMC()</a></code> <code><a href="importVignetteData.html">importBMMC()</a></code> <code><a href="importVignetteData.html">importCGE()</a></code> 
        </dt>
        <dd>Import prepared dataset publically available</dd>
      </dl><dl><dt>
          
          <code><a href="pbmc.html">pbmc</a></code> 
        </dt>
        <dd>liger object of PBMC subsample data with Control and Stimulated datasets</dd>
      </dl><dl><dt>
          
          <code><a href="pbmcPlot.html">pbmcPlot</a></code> 
        </dt>
        <dd>liger object of PBMC subsample data with plotting information available</dd>
      </dl><dl><dt>
          
          <code><a href="bmmc.html">bmmc</a></code> 
        </dt>
        <dd>liger object of bone marrow subsample data with RNA and ATAC modality</dd>
      </dl></div><div class="section level2">
      <h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="runGeneralQC.html">runGeneralQC()</a></code> 
        </dt>
        <dd>General QC for liger object</dd>
      </dl><dl><dt>
          
          <code><a href="runDoubletFinder.html">runDoubletFinder()</a></code> 
        </dt>
        <dd>Doublet detection with DoubletFinder</dd>
      </dl><dl><dt>
          
          <code><a href="removeMissing.html">removeMissing()</a></code> <code><a href="removeMissing.html">removeMissingObs()</a></code> 
        </dt>
        <dd>Remove missing cells or features from liger object</dd>
      </dl><dl><dt>
          
          <code><a href="normalize.html">normalize()</a></code> <code><a href="normalize.html">normalizePeak()</a></code> 
        </dt>
        <dd>Normalize raw counts data</dd>
      </dl><dl><dt>
          
          <code><a href="selectGenes.html">selectGenes()</a></code> 
        </dt>
        <dd>Select a subset of informative genes</dd>
      </dl><dl><dt>
          
          <code><a href="scaleNotCenter.html">scaleNotCenter()</a></code> 
        </dt>
        <dd>Scale genes by root-mean-square across cells</dd>
      </dl><dl><dt>
          
          <code><a href="reverseMethData.html">reverseMethData()</a></code> 
        </dt>
        <dd>Create "scaled data" for DNA methylation datasets</dd>
      </dl><dl><dt>
          
          <code><a href="downsample.html">downsample()</a></code> 
        </dt>
        <dd>Downsample datasets</dd>
      </dl><dl><dt>
          
          <code><a href="selectGenesVST.html">selectGenesVST()</a></code> 
        </dt>
        <dd>Select variable genes from one dataset with Seurat VST method</dd>
      </dl></div><div class="section level2">
      <h2 id="integration">Integration<a class="anchor" aria-label="anchor" href="#integration"></a></h2>
      
      

      
    </div><div class="section level2">
      
      <h3 id="perform-inmf-and-variants">Perform iNMF and Variants<a class="anchor" aria-label="anchor" href="#perform-inmf-and-variants"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="runIntegration.html">runIntegration()</a></code> 
        </dt>
        <dd>Integrate scaled datasets with iNMF or variant methods</dd>
      </dl><dl><dt>
          
          <code><a href="runINMF.html">runINMF()</a></code> 
        </dt>
        <dd>Perform iNMF on scaled datasets</dd>
      </dl><dl><dt>
          
          <code><a href="runOnlineINMF.html">runOnlineINMF()</a></code> 
        </dt>
        <dd>Perform online iNMF on scaled datasets</dd>
      </dl><dl><dt>
          
          <code><a href="runUINMF.html">runUINMF()</a></code> 
        </dt>
        <dd>Perform Mosaic iNMF (UINMF) on scaled datasets with unshared features</dd>
      </dl><dl><dt>
          
          <code><a href="runCINMF.html">runCINMF()</a></code> 
        </dt>
        <dd>Perform consensus iNMF on scaled datasets</dd>
      </dl></div><div class="section level2">
      
      <h3 id="post-factorization-alignment">Post-Factorization Alignment<a class="anchor" aria-label="anchor" href="#post-factorization-alignment"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="quantileNorm.html">quantileNorm()</a></code> 
        </dt>
        <dd>Quantile Align (Normalize) Factor Loadings</dd>
      </dl></div><div class="section level2">
      
      <h3 id="evaluation">Evaluation<a class="anchor" aria-label="anchor" href="#evaluation"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="calcAlignment.html">calcAlignment()</a></code> 
        </dt>
        <dd>Calculate alignment metric after integration</dd>
      </dl><dl><dt>
          
          <code><a href="calcAgreement.html">calcAgreement()</a></code> 
        </dt>
        <dd>Calculate agreement metric after integration</dd>
      </dl><dl><dt>
          
          <code><a href="calcDatasetSpecificity.html">calcDatasetSpecificity()</a></code> 
        </dt>
        <dd>Calculate a dataset-specificity score for each factor</dd>
      </dl></div><div class="section level2">
      
      <h3 id="optimization-with-new-parameter">Optimization with New Parameter<a class="anchor" aria-label="anchor" href="#optimization-with-new-parameter"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="optimizeNewData.html">optimizeNewData()</a></code> 
        </dt>
        <dd>Perform factorization for new data</dd>
      </dl><dl><dt>
          
          <code><a href="optimizeNewK.html">optimizeNewK()</a></code> 
        </dt>
        <dd>Perform factorization for new value of k</dd>
      </dl><dl><dt>
          
          <code><a href="optimizeNewLambda.html">optimizeNewLambda()</a></code> 
        </dt>
        <dd>Perform factorization for new lambda value</dd>
      </dl><dl><dt>
          
          <code><a href="optimizeSubset.html">optimizeSubset()</a></code> 
        </dt>
        <dd>Perform factorization for subset of data</dd>
      </dl></div><div class="section level2">
      <h2 id="downstream-analysis">Downstream Analysis<a class="anchor" aria-label="anchor" href="#downstream-analysis"></a></h2>
      
      

      
    </div><div class="section level2">
      
      <h3 id="dimensionality-reduction">Dimensionality Reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="runUMAP.html">runUMAP()</a></code> 
        </dt>
        <dd>Perform UMAP Dimensionality Reduction</dd>
      </dl><dl><dt>
          
          <code><a href="runTSNE.html">runTSNE()</a></code> 
        </dt>
        <dd>Perform t-SNE dimensionality reduction</dd>
      </dl></div><div class="section level2">
      
      <h3 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="runCluster.html">runCluster()</a></code> 
        </dt>
        <dd>SNN Graph Based Community Detection</dd>
      </dl><dl><dt>
          
          <code><a href="mapCellMeta.html">mapCellMeta()</a></code> 
        </dt>
        <dd>Create new variable from categories in cellMeta</dd>
      </dl><dl><dt>
          
          <code><a href="calcARI.html">calcARI()</a></code> 
        </dt>
        <dd>Calculate adjusted Rand index (ARI) by comparing two cluster labeling variables</dd>
      </dl><dl><dt>
          
          <code><a href="calcPurity.html">calcPurity()</a></code> 
        </dt>
        <dd>Calculate purity by comparing two cluster labeling variables</dd>
      </dl></div><div class="section level2">
      
      <h3 id="differential-expression">Differential Expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="liger-DEG.html">runPairwiseDEG()</a></code> <code><a href="liger-DEG.html">runMarkerDEG()</a></code> <code><a href="liger-DEG.html">runWilcoxon()</a></code> 
        </dt>
        <dd>Find DEG between two groups</dd>
      </dl><dl><dt>
          
          <code><a href="getFactorMarkers.html">getFactorMarkers()</a></code> 
        </dt>
        <dd>Find shared and dataset-specific markers</dd>
      </dl></div><div class="section level2">
      
      <h3 id="go-and-gsea">GO and GSEA<a class="anchor" aria-label="anchor" href="#go-and-gsea"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="runGOEnrich.html">runGOEnrich()</a></code> 
        </dt>
        <dd>Run Gene Ontology enrichment analysis on differentially expressed genes.</dd>
      </dl><dl><dt>
          
          <code><a href="runGSEA.html">runGSEA()</a></code> 
        </dt>
        <dd>Analyze biological interpretations of metagene</dd>
      </dl></div><div class="section level2">
      
      <h3 id="atac-specific-analysis">ATAC Specific Analysis<a class="anchor" aria-label="anchor" href="#atac-specific-analysis"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="imputeKNN.html">imputeKNN()</a></code> 
        </dt>
        <dd>Impute the peak counts from gene expression data referring to an ATAC dataset
after integration</dd>
      </dl><dl><dt>
          
          <code><a href="linkGenesAndPeaks.html">linkGenesAndPeaks()</a></code> 
        </dt>
        <dd>Linking genes to putative regulatory elements</dd>
      </dl><dl><dt>
          
          <code><a href="exportInteractTrack.html">exportInteractTrack()</a></code> 
        </dt>
        <dd>Export predicted gene-pair interaction</dd>
      </dl></div><div class="section level2">
      <h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="plotCellViolin.html">plotCellViolin()</a></code> 
        </dt>
        <dd>Generate violin/box plot(s) using liger object</dd>
      </dl><dl><dt>
          
          <code><a href="plotClusterFactorDot.html">plotClusterFactorDot()</a></code> 
        </dt>
        <dd>Make dot plot of factor loading in cell groups</dd>
      </dl><dl><dt>
          
          <code><a href="plotClusterGeneDot.html">plotClusterGeneDot()</a></code> 
        </dt>
        <dd>Make dot plot of gene expression in cell groups</dd>
      </dl><dl><dt>
          
          <code><a href="plotDensityDimRed.html">plotDensityDimRed()</a></code> 
        </dt>
        <dd>Create density plot basing on specified coordinates</dd>
      </dl><dl><dt>
          
          <code><a href="plotDimRed.html">plotDimRed()</a></code> <code><a href="plotDimRed.html">plotClusterDimRed()</a></code> <code><a href="plotDimRed.html">plotDatasetDimRed()</a></code> <code><a href="plotDimRed.html">plotByDatasetAndCluster()</a></code> <code><a href="plotDimRed.html">plotGeneDimRed()</a></code> <code><a href="plotDimRed.html">plotPeakDimRed()</a></code> <code><a href="plotDimRed.html">plotFactorDimRed()</a></code> 
        </dt>
        <dd>Generate scatter plot(s) using liger object</dd>
      </dl><dl><dt>
          
          <code><a href="plotGeneLoadings.html">plotGeneLoadings()</a></code> <code><a href="plotGeneLoadings.html">plotGeneLoadingRank()</a></code> 
        </dt>
        <dd>Visualize factor expression and gene loading</dd>
      </dl><dl><dt>
          
          <code><a href="plotGroupClusterDimRed.html">plotGroupClusterDimRed()</a></code> 
        </dt>
        <dd>Comprehensive group splited cluster plot on dimension reduction with
proportion</dd>
      </dl><dl><dt>
          
          <code><a href="plotHeatmap.html">plotGeneHeatmap()</a></code> <code><a href="plotHeatmap.html">plotFactorHeatmap()</a></code> 
        </dt>
        <dd>Plot Heatmap of Gene Expression or Factor Loading</dd>
      </dl><dl><dt>
          
          <code><a href="plotMarkerHeatmap.html">plotMarkerHeatmap()</a></code> 
        </dt>
        <dd>Create heatmap for showing top marker expression in conditions</dd>
      </dl><dl><dt>
          
          <code><a href="plotProportion.html">plotProportion()</a></code> <code><a href="plotProportion.html">plotProportionDot()</a></code> <code><a href="plotProportion.html">plotProportionBar()</a></code> <code><a href="plotProportion.html">plotClusterProportions()</a></code> <code><a href="plotProportion.html">plotProportionPie()</a></code> 
        </dt>
        <dd>Visualize proportion across two categorical variables</dd>
      </dl><dl><dt>
          
          <code><a href="plotSankey.html">plotSankey()</a></code> 
        </dt>
        <dd>Make Riverplot/Sankey diagram that shows label mapping across datasets</dd>
      </dl><dl><dt>
          
          <code><a href="plotSpatial.html">plotSpatial2D()</a></code> 
        </dt>
        <dd>Visualize a spatial dataset</dd>
      </dl><dl><dt>
          
          <code><a href="plotVarFeatures.html">plotVarFeatures()</a></code> 
        </dt>
        <dd>Plot the variance vs mean of feature expression</dd>
      </dl><dl><dt>
          
          <code><a href="plotViolin.html">plotGeneViolin()</a></code> <code><a href="plotViolin.html">plotTotalCountViolin()</a></code> <code><a href="plotViolin.html">plotGeneDetectedViolin()</a></code> 
        </dt>
        <dd>Visualize gene expression or cell metadata with violin plot</dd>
      </dl><dl><dt>
          
          <code><a href="plotVolcano.html">plotVolcano()</a></code> <code><a href="plotVolcano.html">plotEnhancedVolcano()</a></code> 
        </dt>
        <dd>Create volcano plot for Wilcoxon test result</dd>
      </dl><dl><dt>
          
          <code><a href="rliger-deprecated.html">makeInteractTrack()</a></code> <code><a href="rliger-deprecated.html">louvainCluster()</a></code> <code><a href="rliger-deprecated.html">optimizeALS()</a></code> <code><a href="rliger-deprecated.html">online_iNMF()</a></code> <code><a href="rliger-deprecated.html">quantile_norm()</a></code> <code><a href="rliger-deprecated.html">makeRiverplot()</a></code> 
        </dt>
        <dd>Deprecated functions in package <span class="pkg">rliger</span>.</dd>
      </dl><dl><dt>
          
          <code><a href="dot-ggCellViolin.html">.ggCellViolin()</a></code> 
        </dt>
        <dd>Produce single violin plot with data frame passed from upstream</dd>
      </dl><dl><dt>
          
          <code><a href="dot-ggScatter.html">.ggScatter()</a></code> 
        </dt>
        <dd>Produce single scatter plot with data frame passed from upstream</dd>
      </dl><dl><dt>
          
          <code><a href="dot-ggplotLigerTheme.html">.ggplotLigerTheme()</a></code> 
        </dt>
        <dd>Generic ggplot theme setting for rliger package</dd>
      </dl><dl><dt>
          
          <code><a href="dot-plotHeatmap.html">.plotHeatmap()</a></code> 
        </dt>
        <dd>General heatmap plotting with prepared matrix and data.frames</dd>
      </dl><dl><dt>
          
          <code><a href="dot-complexHeatmapDotPlot.html">.complexHeatmapDotPlot()</a></code> 
        </dt>
        <dd>Generate dot plot from input matrix with ComplexHeatmap</dd>
      </dl></div><div class="section level2">
      <h2 id="object-class">Object Class<a class="anchor" aria-label="anchor" href="#object-class"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="liger-class.html">datasets()</a></code> <code><a href="liger-class.html">`datasets&lt;-`()</a></code> <code><a href="liger-class.html">dataset()</a></code> <code><a href="liger-class.html">`dataset&lt;-`()</a></code> <code><a href="liger-class.html">cellMeta()</a></code> <code><a href="liger-class.html">`cellMeta&lt;-`()</a></code> <code><a href="liger-class.html">defaultCluster()</a></code> <code><a href="liger-class.html">`defaultCluster&lt;-`()</a></code> <code><a href="liger-class.html">dimReds()</a></code> <code><a href="liger-class.html">`dimReds&lt;-`()</a></code> <code><a href="liger-class.html">dimRed()</a></code> <code><a href="liger-class.html">`dimRed&lt;-`()</a></code> <code><a href="liger-class.html">defaultDimRed()</a></code> <code><a href="liger-class.html">`defaultDimRed&lt;-`()</a></code> <code><a href="liger-class.html">varFeatures()</a></code> <code><a href="liger-class.html">`varFeatures&lt;-`()</a></code> <code><a href="liger-class.html">varUnsharedFeatures()</a></code> <code><a href="liger-class.html">`varUnsharedFeatures&lt;-`()</a></code> <code><a href="liger-class.html">commands()</a></code> <code><a href="liger-class.html">show(<i>&lt;liger&gt;</i>)</a></code> <code><a href="liger-class.html">dim(<i>&lt;liger&gt;</i>)</a></code> <code><a href="liger-class.html">dimnames(<i>&lt;liger&gt;</i>)</a></code> <code><a href="liger-class.html">`dimnames&lt;-`(<i>&lt;liger&gt;</i>,<i>&lt;list&gt;</i>)</a></code> <code><a href="liger-class.html">names(<i>&lt;liger&gt;</i>)</a></code> <code><a href="liger-class.html">`names&lt;-`(<i>&lt;liger&gt;</i>)</a></code> <code><a href="liger-class.html">length(<i>&lt;liger&gt;</i>)</a></code> <code><a href="liger-class.html">lengths(<i>&lt;liger&gt;</i>)</a></code> <code><a href="liger-class.html">rawData(<i>&lt;liger&gt;</i>)</a></code> <code><a href="liger-class.html">`rawData&lt;-`(<i>&lt;liger&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;matrixLike_OR_NULL&gt;</i>)</a></code> <code><a href="liger-class.html">`rawData&lt;-`(<i>&lt;liger&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;H5D&gt;</i>)</a></code> <code><a href="liger-class.html">normData(<i>&lt;liger&gt;</i>)</a></code> <code><a href="liger-class.html">`normData&lt;-`(<i>&lt;liger&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;matrixLike_OR_NULL&gt;</i>)</a></code> <code><a href="liger-class.html">`normData&lt;-`(<i>&lt;liger&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;H5D&gt;</i>)</a></code> <code><a href="liger-class.html">scaleData(<i>&lt;liger&gt;</i>,<i>&lt;ANY&gt;</i>)</a></code> <code><a href="liger-class.html">`scaleData&lt;-`(<i>&lt;liger&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;matrixLike_OR_NULL&gt;</i>)</a></code> <code><a href="liger-class.html">`scaleData&lt;-`(<i>&lt;liger&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;H5D&gt;</i>)</a></code> <code><a href="liger-class.html">`scaleData&lt;-`(<i>&lt;liger&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;H5Group&gt;</i>)</a></code> <code><a href="liger-class.html">scaleUnsharedData(<i>&lt;liger&gt;</i>,<i>&lt;character&gt;</i>)</a></code> <code><a href="liger-class.html">scaleUnsharedData(<i>&lt;liger&gt;</i>,<i>&lt;numeric&gt;</i>)</a></code> <code><a href="liger-class.html">`scaleUnsharedData&lt;-`(<i>&lt;liger&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;matrixLike_OR_NULL&gt;</i>)</a></code> <code><a href="liger-class.html">`scaleUnsharedData&lt;-`(<i>&lt;liger&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;H5D&gt;</i>)</a></code> <code><a href="liger-class.html">`scaleUnsharedData&lt;-`(<i>&lt;liger&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;H5Group&gt;</i>)</a></code> <code><a href="liger-class.html">getMatrix(<i>&lt;liger&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>)</a></code> <code><a href="liger-class.html">getH5File(<i>&lt;liger&gt;</i>,<i>&lt;ANY&gt;</i>)</a></code> <code><a href="liger-class.html">`[[&lt;-`(<i>&lt;liger&gt;</i>)</a></code> <code><a href="liger-class.html">`$`(<i>&lt;liger&gt;</i>)</a></code> <code><a href="liger-class.html">`$&lt;-`(<i>&lt;liger&gt;</i>)</a></code> <code><a href="liger-class.html">fortify(<i>&lt;liger&gt;</i>)</a></code> <code><a href="liger-class.html">c(<i>&lt;liger&gt;</i>)</a></code> 
        </dt>
        <dd>liger class</dd>
      </dl><dl><dt>
          
          <code><a href="ligerATACDataset-class.html">ligerATACDataset-class</a></code> <code><a href="ligerATACDataset-class.html">ligerATACDataset</a></code> 
        </dt>
        <dd>Subclass of ligerDataset for ATAC modality</dd>
      </dl><dl><dt>
          
          <code><a href="ligerCommand-class.html">show(<i>&lt;ligerCommand&gt;</i>)</a></code> 
        </dt>
        <dd>ligerCommand object: Record the input and time of a LIGER function call</dd>
      </dl><dl><dt>
          
          <code><a href="ligerDataset-class.html">rawData()</a></code> <code><a href="ligerDataset-class.html">`rawData&lt;-`()</a></code> <code><a href="ligerDataset-class.html">normData()</a></code> <code><a href="ligerDataset-class.html">`normData&lt;-`()</a></code> <code><a href="ligerDataset-class.html">scaleData()</a></code> <code><a href="ligerDataset-class.html">`scaleData&lt;-`()</a></code> <code><a href="ligerDataset-class.html">scaleUnsharedData()</a></code> <code><a href="ligerDataset-class.html">`scaleUnsharedData&lt;-`()</a></code> <code><a href="ligerDataset-class.html">getMatrix()</a></code> <code><a href="ligerDataset-class.html">h5fileInfo()</a></code> <code><a href="ligerDataset-class.html">`h5fileInfo&lt;-`()</a></code> <code><a href="ligerDataset-class.html">getH5File()</a></code> <code><a href="ligerDataset-class.html">featureMeta()</a></code> <code><a href="ligerDataset-class.html">`featureMeta&lt;-`()</a></code> <code><a href="ligerDataset-class.html">show(<i>&lt;ligerDataset&gt;</i>)</a></code> <code><a href="ligerDataset-class.html">dim(<i>&lt;ligerDataset&gt;</i>)</a></code> <code><a href="ligerDataset-class.html">dimnames(<i>&lt;ligerDataset&gt;</i>)</a></code> <code><a href="ligerDataset-class.html">`dimnames&lt;-`(<i>&lt;ligerDataset&gt;</i>,<i>&lt;list&gt;</i>)</a></code> <code><a href="ligerDataset-class.html">cbind(<i>&lt;ligerDataset&gt;</i>)</a></code> 
        </dt>
        <dd>ligerDataset class</dd>
      </dl><dl><dt>
          
          <code><a href="ligerMethDataset-class.html">ligerMethDataset-class</a></code> <code><a href="ligerMethDataset-class.html">ligerMethDataset</a></code> 
        </dt>
        <dd>Subclass of ligerDataset for Methylation modality</dd>
      </dl><dl><dt>
          
          <code><a href="ligerRNADataset-class.html">ligerRNADataset-class</a></code> <code><a href="ligerRNADataset-class.html">ligerRNADataset</a></code> 
        </dt>
        <dd>Subclass of ligerDataset for RNA modality</dd>
      </dl><dl><dt>
          
          <code><a href="ligerSpatialDataset-class.html">ligerSpatialDataset-class</a></code> <code><a href="ligerSpatialDataset-class.html">ligerSpatialDataset</a></code> 
        </dt>
        <dd>Subclass of ligerDataset for Spatial modality</dd>
      </dl><dl><dt>
          
          <code><a href="sub-liger.html">`[`(<i>&lt;liger&gt;</i>)</a></code> 
        </dt>
        <dd>Subset liger with brackets</dd>
      </dl><dl><dt>
          
          <code><a href="sub-sub-liger.html">`[[`(<i>&lt;liger&gt;</i>)</a></code> 
        </dt>
        <dd>Get cell metadata variable</dd>
      </dl><dl><dt>
          
          <code><a href="sub-ligerDataset.html">`[`(<i>&lt;ligerDataset&gt;</i>)</a></code> 
        </dt>
        <dd>Subset ligerDataset object</dd>
      </dl><dl><dt>
          
          <code><a href="peak.html">rawPeak()</a></code> <code><a href="peak.html">`rawPeak&lt;-`()</a></code> <code><a href="peak.html">normPeak()</a></code> <code><a href="peak.html">`normPeak&lt;-`()</a></code> 
        </dt>
        <dd>Access ligerATACDataset peak data</dd>
      </dl><dl><dt>
          
          <code><a href="coordinate.html">coordinate()</a></code> <code><a href="coordinate.html">`coordinate&lt;-`()</a></code> 
        </dt>
        <dd>Access ligerSpatialDataset coordinate data</dd>
      </dl><dl><dt>
          
          <code><a href="retrieveCellFeature.html">retrieveCellFeature()</a></code> 
        </dt>
        <dd>Retrieve a single matrix of cells from a slot</dd>
      </dl><dl><dt>
          
          <code><a href="subsetLiger.html">subsetLiger()</a></code> 
        </dt>
        <dd>Subset liger object</dd>
      </dl><dl><dt>
          
          <code><a href="subsetLigerDataset.html">subsetLigerDataset()</a></code> <code><a href="subsetLigerDataset.html">subsetH5LigerDataset()</a></code> <code><a href="subsetLigerDataset.html">subsetMemLigerDataset()</a></code> 
        </dt>
        <dd>Subset ligerDataset object</dd>
      </dl><dl><dt>
          
          <code><a href="convertOldLiger.html">convertOldLiger()</a></code> 
        </dt>
        <dd>Convert old liger object to latest version</dd>
      </dl><dl><dt>
          
          <code><a href="ligerToSeurat.html">ligerToSeurat()</a></code> 
        </dt>
        <dd>Convert between liger and Seurat object</dd>
      </dl><dl><dt>
          
          <code><a href="is.newLiger.html">is.newLiger()</a></code> 
        </dt>
        <dd>Check if given liger object if under new implementation</dd>
      </dl><dl><dt>
          
          <code><a href="modalOf.html">modalOf()</a></code> 
        </dt>
        <dd>Return preset modality of a ligerDataset object or that of all datasets in a
liger object</dd>
      </dl><dl><dt>
          
          <code><a href="isH5Liger.html">isH5Liger()</a></code> 
        </dt>
        <dd>Check if a liger or ligerDataset object is made of HDF5 file</dd>
      </dl><dl><dt>
          
          <code><a href="restoreH5Liger.html">restoreH5Liger()</a></code> <code><a href="restoreH5Liger.html">restoreOnlineLiger()</a></code> 
        </dt>
        <dd>Restore links (to HDF5 files) for reloaded liger/ligerDataset object</dd>
      </dl><dl><dt>
          
          <code><a href="closeAllH5.html">closeAllH5()</a></code> 
        </dt>
        <dd>Close all links (to HDF5 files) of a liger object</dd>
      </dl></div><div class="section level2">
      <h2 id="misc">Misc<a class="anchor" aria-label="anchor" href="#misc"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="H5Apply.html">H5Apply()</a></code> 
        </dt>
        <dd>Apply function to chunks of H5 data in ligerDataset object</dd>
      </dl><dl><dt>
          
          <code><a href="commandDiff.html">commandDiff()</a></code> 
        </dt>
        <dd>Check difference of two liger command</dd>
      </dl><dl><dt>
          
          <code><a href="makeFeatureMatrix.html">makeFeatureMatrix()</a></code> 
        </dt>
        <dd>Fast calculation of feature count matrix</dd>
      </dl><dl><dt>
          
          <code><a href="mergeH5.html">mergeH5()</a></code> 
        </dt>
        <dd>Merge hdf5 files</dd>
      </dl><dl><dt>
          
          <code><a href="mergeSparseAll.html">mergeSparseAll()</a></code> <code><a href="mergeSparseAll.html">mergeDenseAll()</a></code> 
        </dt>
        <dd>Merge matrices while keeping the union of rows</dd>
      </dl><dl><dt>
          
          <code><a href="writeH5.html">writeH5()</a></code> 
        </dt>
        <dd>Write in-memory data into H5 file</dd>
      </dl></div><div class="section level2">
      <h2 id="deprecated">Deprecated<a class="anchor" aria-label="anchor" href="#deprecated"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="louvainCluster-deprecated.html">louvainCluster-deprecated</a></code> 
        </dt>
        <dd>[Deprecated] Louvain algorithm for community detection</dd>
      </dl><dl><dt>
          
          <code><a href="makeInteractTrack-deprecated.html">makeInteractTrack-deprecated</a></code> 
        </dt>
        <dd>[Deprecated] Export predicted gene-pair interaction</dd>
      </dl><dl><dt>
          
          <code><a href="makeRiverplot-deprecated.html">makeRiverplot-deprecated</a></code> 
        </dt>
        <dd>[Deprecated] Generate a river (Sankey) plot</dd>
      </dl><dl><dt>
          
          <code><a href="online_iNMF-deprecated.html">online_iNMF-deprecated</a></code> 
        </dt>
        <dd>[Deprecated] Perform online iNMF on scaled datasets</dd>
      </dl><dl><dt>
          
          <code><a href="optimizeALS-deprecated.html">optimizeALS-deprecated</a></code> 
        </dt>
        <dd>[Deprecated] Perform iNMF on scaled datasets</dd>
      </dl><dl><dt>
          
          <code><a href="quantile_norm-deprecated.html">quantile_norm-deprecated</a></code> 
        </dt>
        <dd>[Deprecated] Quantile align (normalize) factor loading</dd>
      </dl><dl><dt>
          
          <code><a href="readSubset.html">readSubset()</a></code> 
        </dt>
        <dd>[Deprecated] See <code><a href="https://welch-lab.github.io/liger/reference/downsample.html" class="external-link">downsample</a></code></dd>
      </dl><dl><dt>
          
          <code><a href="quantileAlignSNF.html">quantileAlignSNF()</a></code> 
        </dt>
        <dd>Quantile align (normalize) factor loadings</dd>
      </dl><dl><dt>
          
          <code><a href="getProportionMito.html">getProportionMito()</a></code> 
        </dt>
        <dd>Calculate proportion mitochondrial contribution</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

