<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='Two methods are supported: "pseudoBulk" and
"wilcoxon". Pseudo-bulk method aggregates cells basing on biological
replicates and calls bulk RNAseq DE methods, DESeq2 wald test, while
Wilcoxon rank sum test is performed on single-cell level.
runPairwiseDEG() is generally used for flexibly comparing two specific
groups of cells, while runMarkerDEG() is used for a one-vs-rest marker
test strategy.
While using pseudo-bulk method, it is generally recommended that you have
these variables available in your object:

The cell type or cluster labeling. This can be obtained from prior
study or computed with runCluster
The biological replicate labeling, most of the time the
"dataset" variable automatically generated when the
liger object is created. Users may use other variables if
a "dataset" is merged from multiple replicates.
The condition labeling that reflects the study design, such as the
treatment or disease status for each sample/dataset.


Please see below for detailed scenarios.'><title>Find DEG between groups — runPairwiseDEG • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Find DEG between groups — runPairwiseDEG"><meta property="og:description" content='Two methods are supported: "pseudoBulk" and
"wilcoxon". Pseudo-bulk method aggregates cells basing on biological
replicates and calls bulk RNAseq DE methods, DESeq2 wald test, while
Wilcoxon rank sum test is performed on single-cell level.
runPairwiseDEG() is generally used for flexibly comparing two specific
groups of cells, while runMarkerDEG() is used for a one-vs-rest marker
test strategy.
While using pseudo-bulk method, it is generally recommended that you have
these variables available in your object:

The cell type or cluster labeling. This can be obtained from prior
study or computed with runCluster
The biological replicate labeling, most of the time the
"dataset" variable automatically generated when the
liger object is created. Users may use other variables if
a "dataset" is merged from multiple replicates.
The condition labeling that reflects the study design, such as the
treatment or disease status for each sample/dataset.


Please see below for detailed scenarios.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    <a class="dropdown-item" href="../articles/benchmark.html">Benchmarking integration quality</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Find DEG between groups</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/DEG_marker.R" class="external-link"><code>R/DEG_marker.R</code></a></small>
      <div class="d-none name"><code>liger-DEG.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Two methods are supported: <code>"pseudoBulk"</code> and
<code>"wilcoxon"</code>. Pseudo-bulk method aggregates cells basing on biological
replicates and calls bulk RNAseq DE methods, DESeq2 wald test, while
Wilcoxon rank sum test is performed on single-cell level.
<code>runPairwiseDEG()</code> is generally used for flexibly comparing two specific
groups of cells, while <code>runMarkerDEG()</code> is used for a one-vs-rest marker
test strategy.</p>
<p>While using pseudo-bulk method, it is generally recommended that you have
these variables available in your object:</p>
<ol><li><p>The cell type or cluster labeling. This can be obtained from prior
study or computed with <code><a href="runCluster.html">runCluster</a></code></p></li>
<li><p>The biological replicate labeling, most of the time the
<code>"dataset"</code> variable automatically generated when the
<a href="liger-class.html">liger</a> object is created. Users may use other variables if
a "dataset" is merged from multiple replicates.</p></li>
<li><p>The condition labeling that reflects the study design, such as the
treatment or disease status for each sample/dataset.</p></li>
</ol><p>Please see below for detailed scenarios.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runPairwiseDEG</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">groupTest</span>,</span>
<span>  <span class="va">groupCtrl</span>,</span>
<span>  variable1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  variable2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  splitBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pseudoBulk"</span>, <span class="st">"wilcoxon"</span><span class="op">)</span>,</span>
<span>  usePeak <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  useReplicate <span class="op">=</span> <span class="st">"dataset"</span>,</span>
<span>  nPsdRep <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minCellPerRep <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  printDiagnostic <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  chunk <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span>, <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">runMarkerDEG</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  conditionBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  splitBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pseudoBulk"</span>, <span class="st">"wilcoxon"</span><span class="op">)</span>,</span>
<span>  useDatasets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  usePeak <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  useReplicate <span class="op">=</span> <span class="st">"dataset"</span>,</span>
<span>  nPsdRep <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minCellPerRep <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  printDiagnostic <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  chunk <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span>, <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">runWilcoxon</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  data.use <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  compare.method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clusters"</span>, <span class="st">"datasets"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object, with normalized data available</p></dd>


<dt>groupTest, groupCtrl, variable1, variable2</dt>
<dd><p>Condition specification. See
<code>?runPairwiseDEG</code> section <b>Pairwise DEG Scenarios</b> for detail.</p></dd>


<dt>splitBy</dt>
<dd><p>Name(s) of the variable(s) in <code>cellMeta</code> to split the
comparison. See Details. Default <code>NULL</code>.</p></dd>


<dt>method</dt>
<dd><p>DEG test method to use. Choose from <code>"pseudoBulk"</code> or
<code>"wilcoxon"</code>. Default <code>"pseudoBulk"</code></p></dd>


<dt>usePeak</dt>
<dd><p>Logical. Whether to use peak count instead of gene count.
Only supported when ATAC datasets are involved. Default <code>FALSE</code>.</p></dd>


<dt>useReplicate</dt>
<dd><p><code>cellMeta</code> variable of biological replicate
annotation. Only used with <code>method = "pseudoBulk"</code>. Default
<code>"dataset"</code>.</p></dd>


<dt>nPsdRep</dt>
<dd><p>Number of pseudo-replicates to create. Only used when
<code>method = "pseudoBulk"</code>. Default <code>NULL</code>. See Details.</p></dd>


<dt>minCellPerRep</dt>
<dd><p>Numeric, will not make pseudo-bulk for replicate with
less than this number of cells. Default <code>3</code>.</p></dd>


<dt>printDiagnostic</dt>
<dd><p>Logical. Whether to show more detail when
<code>verbose = TRUE</code>. Default <code>FALSE</code>.</p></dd>


<dt>chunk</dt>
<dd><p>Number of features to process at a time during Wilcoxon test.
Useful when memory is limited. Default <code>NULL</code> will process all features
at once.</p></dd>


<dt>seed</dt>
<dd><p>Random seed to use for pseudo-replicate generation. Default
<code>1</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> or <code>TRUE</code> if users have not set.</p></dd>


<dt>conditionBy</dt>
<dd><p><code>cellMeta</code> variable(s). Marker detection will be
performed for each level of this variable. Multiple variables will be
combined. Default <code>NULL</code> uses default cluster.</p></dd>


<dt>useDatasets</dt>
<dd><p>Datasets to perform marker detection within. Default
<code>NULL</code> will use all datasets.</p></dd>


<dt>data.use</dt>
<dd><p>Same as <code>useDatasets</code>.</p></dd>


<dt>compare.method</dt>
<dd><p>Choose from <code>"clusters"</code> (default) or
<code>"datasets"</code>. <code>"clusters"</code> compares each cluster against all other
cells, while <code>"datasets"</code> run within each cluster and compare each
dataset against all other datasets.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A data.frame with DEG information with the all or some of the
following fields:</p>
<dl><dt>feature</dt>
<dd><p>Gene names</p></dd>

<dt>group</dt>
<dd><p>Test group name. Multiple tests might be present for each
function call. This is the main variable to distinguish the tests. For a
pairwise test, a row with a certain group name represents the test result
between the this group against the other control group; When split by a
variable, it would be presented in "split.group" format, meaning the stats
is by comparing the group in the split level against the control group in
the same split level. When running marker detection without splitting,
a row with group "a" represents the stats of the gene in group "a" against
all other cells. When running split marker detection, the group name would
be in "split.group" format, meaning the stats is by comparing the group in
the split level against all other cells in the same split level.</p></dd>

<dt>logFC</dt>
<dd><p>Log fold change</p></dd>

<dt>pval</dt>
<dd><p>P-value</p></dd>

<dt>padj</dt>
<dd><p>Adjusted p-value</p></dd>

<dt>avgExpr</dt>
<dd><p>Mean expression in the test group indicated by the "group"
field. Only available for wilcoxon tests.</p></dd>

<dt>statistic</dt>
<dd><p>Wilcoxon rank-sum test statistic. Only available for
wilcoxon tests.</p></dd>

<dt>auc</dt>
<dd><p>Area under the ROC curve. Only available for wilcoxon tests.</p></dd>

<dt>pct_in</dt>
<dd><p>Percentage of cells in the test group, indicated by the
"group" field, that express the feature. Only available for wilcoxon
tests.</p></dd>

<dt>pct_out</dt>
<dd><p>Percentage of cells in the control group or other cells, as
explained for the "group" field, that express the feature. Only available
for wilcoxon tests.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="using-wilcoxon-rank-sum-test">Using Wilcoxon rank-sum test<a class="anchor" aria-label="anchor" href="#using-wilcoxon-rank-sum-test"></a></h2>
    

<p>Wilcoxon rank-sum test works for each gene and is based on the rank of the
expression in each cell. LIGER provides dataset integration but does not
"correct" the expression values. Projects with strong batch effects or
integrate drastically different modalities should be cautious when using
this method.</p>
    </div>
    <div class="section level2">
    <h2 id="comparing-difference-between-across-cell-types">Comparing difference between/across cell types<a class="anchor" aria-label="anchor" href="#comparing-difference-between-across-cell-types"></a></h2>
    

<p>Most of times, people would want to know what cell types are for each cluster
after clustering. This can be done with a marker detection method that test
each cluster against all the other cells. This can be done with a command
like <code>runMarkerDEG(object, conditionBy = "cluster_var")</code>. When using
default pseudo-bulk method, users should additionaly determine the
pseudo-bulk setup parameters. If the real biological replicate variable is
available, it should be supplied to argument <code>useReplicate</code>, otherwise,
pseudo-replicates should be created. See "Pseudo-Replicate" section for more.</p>
    </div>
    <div class="section level2">
    <h2 id="compare-between-conditions">Compare between conditions<a class="anchor" aria-label="anchor" href="#compare-between-conditions"></a></h2>
    

<p>It is frequently needed to identify the difference between conditions. Users
can simply set <code>conditionBy = "condition_var"</code>. However, most of time,
such comparisons should be ideally done in a per-cluster manner. This can be
done by setting <code>splitBy = "cluster_var"</code>. This will run a loop for each
cluster, and within the group of cells, compare each condition against all
other cells in the cluster.</p>
<p>In the scenario when users only need to compare two conditions for each
cluster, running <code>runPairwiseDEG(object, groupTest = "condition1",
groupCtrl = "condition2", variable1 = "condition_var",
splitBy = "cluster_var")</code> would address the need.</p>
<p>For both use case, if pseudo-bulk (default) method is used, users should
determine the pseudo-bulk setup parameters as mentioned in the previous
section.</p>
    </div>
    <div class="section level2">
    <h2 id="detailed-runmarkerdeg-usage">Detailed <code>runMarkerDEG</code> usage<a class="anchor" aria-label="anchor" href="#detailed-runmarkerdeg-usage"></a></h2>
    

<p>Marker detection is performed in a one vs. rest manner. The grouping of such
condition is specified by <code>conditionBy</code>, which should be a column name
in <code>cellMeta</code>. When <code>splitBy</code> is specified as another variable
name in <code>cellMeta</code>, the marker detection will be iteratively done for
within each level of <code>splitBy</code> variable.</p>
<p>For example, when <code>conditionBy = "celltype"</code> and <code>splitBy = NULL</code>,
marker detection will be performed by comparing all cells of "celltype_i"
against all other cells, and etc. This is analogous to the old version when
running <code>runWilcoxon(method = "cluster")</code>.</p>
<p>When <code>conditionBy = "gender"</code> and <code>splitBy = "leiden_cluster"</code>,
marker detection will be performed by comparing "gender_i" cells from "cluster_j"
against other cells from "cluster_j", and etc. This is analogous to the old
version when running <code>runWilcoxon(method = "dataset")</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="detailed-runpairwisedeg-usage">Detailed <code>runPairwiseDEG</code> usage<a class="anchor" aria-label="anchor" href="#detailed-runpairwisedeg-usage"></a></h2>
    

<p>Users can select classes of cells from a variable in <code>cellMeta</code>.
<code>variable1</code> and <code>variable2</code> are used to specify a column in
<code>cellMeta</code>, and <code>groupTest</code> and <code>groupCtrl</code> are used to specify
existing classes from <code>variable1</code> and <code>variable2</code>, respectively.
When <code>variable2</code> is missing, <code>groupCtrl</code> will be considered from
<code>variable1</code>.</p>
<p>For example, when <code>variable1 = "celltype"</code> and <code>variable2 = NULL</code>,
<code>groupTest</code> and <code>groupCtrl</code> should be valid cell types in
<code>object$celltype</code>.</p>
<p>When <code>variable1</code> is "celltype" and <code>variable2</code> is "gender",
<code>groupTest</code> should be a valid cell type from <code>object$celltype</code> and
<code>groupCtrl</code> should be a valid class from <code>object$gender</code>.</p>
<p>When both <code>variable1</code> and <code>variable2</code> are missing, <code>groupTest</code>
and <code>groupCtrl</code> should be valid index of cells in <code>object</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="pseudo-replicate">Pseudo-Replicate<a class="anchor" aria-label="anchor" href="#pseudo-replicate"></a></h2>
    

<p>Pseudo-replicate assignment is a technique to complement the lack of real
biological replicates when using pseudo-bulk DE methods. LIGER's pseudo-bulk
method generally requires that each comparison group has at least 3
replicates each composed of at least 3 cells, in order to ensure the
statistic power. When less than 3 real replicates are found for a comparison,
the default setting (<code>nPsdRep = NULL</code>) splits each into 3
pseudo-replicates, otherwise no pseudo-replicates are automatically
generated. When <code>nPsdRep</code> is given a number, LIGER will always go
through each comparison group and split each real replicate into the given
number of pseudo-replicates.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="va">pbmc</span><span class="op">$</span><span class="va">leiden_cluster</span> <span class="op">&lt;-</span> <span class="va">pbmcPlot</span><span class="op">$</span><span class="va">leiden_cluster</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Identify cluster markers</span></span></span>
<span class="r-in"><span><span class="va">degStats1</span> <span class="op">&lt;-</span> <span class="fu">runMarkerDEG</span><span class="op">(</span><span class="va">pbmc</span>, conditionBy <span class="op">=</span> <span class="st">"leiden_cluster"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DESeq2 Wald test <span style="color: #00BB00;">■■■■■■■■■                       </span>  25% | ETA:  4s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DESeq2 Wald test <span style="color: #00BB00;">■■■■■■■■■■■■■■■■■■■■■■■■■■■     </span>  88% | ETA:  1s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> Ignoring replicates (size in bracket) with too few cells: <span style="color: #0000BB;">"7.ctrl.rep1 (2)"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DESeq2 Wald test <span style="color: #00BB00;">■■■■■■■■■■■■■■■■■■■■■■■■■■■     </span>  88% | ETA:  1s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Consider decrease <span style="color: #00BB00;">minCellPerRep</span> to exclude less replicates or/and lower <span style="color: #00BB00;">nPsdRep</span> to generate larger pseudo-replicates.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DESeq2 Wald test <span style="color: #00BB00;">■■■■■■■■■■■■■■■■■■■■■■■■■■■     </span>  88% | ETA:  1s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DESeq2 Wald test <span style="color: #00BB00;">■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ </span> 100% | ETA:  0s</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Compare "stim" data against "ctrl" data within each cluster</span></span></span>
<span class="r-in"><span><span class="va">degStats3</span> <span class="op">&lt;-</span> <span class="fu">runPairwiseDEG</span><span class="op">(</span><span class="va">pbmc</span>, groupTest <span class="op">=</span> <span class="st">"stim"</span>, groupCtrl <span class="op">=</span> <span class="st">"ctrl"</span>,</span></span>
<span class="r-in"><span>                            variable1 <span class="op">=</span> <span class="st">"dataset"</span>,</span></span>
<span class="r-in"><span>                            splitBy <span class="op">=</span> <span class="st">"leiden_cluster"</span>,</span></span>
<span class="r-in"><span>                            minCellPerRep <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running DEG within: <span style="color: #0000BB;">"0"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Calling pairwise DESeq2 Wald test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running DEG within: <span style="color: #0000BB;">"1"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> -- note: fitType='parametric', but the dispersion trend was not well captured by the</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>    function: y = a/x + b, and a local regression fit was automatically substituted.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>    specify fitType='local' or 'mean' to avoid this message next time.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Calling pairwise DESeq2 Wald test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running DEG within: <span style="color: #0000BB;">"2"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Calling pairwise DESeq2 Wald test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running DEG within: <span style="color: #0000BB;">"3"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Calling pairwise DESeq2 Wald test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running DEG within: <span style="color: #0000BB;">"4"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Calling pairwise DESeq2 Wald test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running DEG within: <span style="color: #0000BB;">"5"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Calling pairwise DESeq2 Wald test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running DEG within: <span style="color: #0000BB;">"6"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Calling pairwise DESeq2 Wald test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running DEG within: <span style="color: #0000BB;">"7"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> Ignoring replicates (size in bracket) with too few cells: <span style="color: #0000BB;">"others.ctrl.rep1 (2)"</span>, <span style="color: #0000BB;">"others.ctrl.rep2 (3)"</span>, and <span style="color: #0000BB;">"others.ctrl.rep3 (3)"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Consider decrease <span style="color: #00BB00;">minCellPerRep</span> to exclude less replicates or/and lower <span style="color: #00BB00;">nPsdRep</span> to generate larger pseudo-replicates.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> Error when computing on <span style="color: #0000BB;">"7.stim"</span>: Too few replicates with more than <span style="color: #0000BB;">4</span> cells (`minCellPerRep`) for condition <span style="color: #0000BB;">"others"</span>.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> Empty result returned for this test.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> Calling pairwise DESeq2 Wald test ... failed</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

