<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>liger class — liger-class • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="liger class — liger-class"><meta property="og:description" content="liger object is the main data container for LIGER
analysis in R. The slot datasets is a list where each element should
be a ligerDataset object containing dataset specific
information, such as the expression matrices. The other parts of liger object
stores information that can be shared across the analysis, such as the cell
metadata and factorization result matrices.
This manual provides explanation to the liger object structure as well
as usage of class-specific methods. Please see detail sections for more
information.
For liger objects created with older versions of rliger package,
please try updating the objects individually with
convertOldLiger."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/import_data.html">Import data</a>
    </li>
    <li>
      <a href="../articles/liger_object.html">Introduction to a liger object</a>
    </li>
    <li>
      <a href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>liger class</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/liger-class.R" class="external-link"><code>R/liger-class.R</code></a>, <a href="https://github.com/welch-lab/liger/blob/HEAD/R/ligerCommand-class.R" class="external-link"><code>R/ligerCommand-class.R</code></a>, <a href="https://github.com/welch-lab/liger/blob/HEAD/R/ligerDataset-class.R" class="external-link"><code>R/ligerDataset-class.R</code></a></small>
    <div class="hidden name"><code>liger-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>liger</code> object is the main data container for LIGER
analysis in R. The slot <code>datasets</code> is a list where each element should
be a <a href="ligerDataset-class.html">ligerDataset</a> object containing dataset specific
information, such as the expression matrices. The other parts of liger object
stores information that can be shared across the analysis, such as the cell
metadata and factorization result matrices.</p>
<p>This manual provides explanation to the <code>liger</code> object structure as well
as usage of class-specific methods. Please see detail sections for more
information.</p>
<p>For <code>liger</code> objects created with older versions of rliger package,
please try updating the objects individually with
<code><a href="convertOldLiger.html">convertOldLiger</a></code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre><code># S4 method for liger
show(object)

# S4 method for liger
dim(x)

# S4 method for liger
dimnames(x)

# S4 method for liger,list
dimnames(x) &lt;- value

# S4 method for liger,character,missing,ANY
[(x, i, j, ..., drop = TRUE)

# S4 method for liger,missing,index,ANY
[(x, i, j, ..., drop = TRUE)

# S4 method for liger,character,index,ANY
[(x, i, j, ..., drop = TRUE)

datasets(x, check = NULL)

datasets(x, check = TRUE) &lt;- value

# S4 method for liger
datasets(x, check = NULL)

# S4 method for liger,logical
datasets(x, check = TRUE) &lt;- value

# S4 method for liger,missing
datasets(x, check = TRUE) &lt;- value

dataset(x, dataset = NULL)

dataset(x, dataset, type = NULL, qc = TRUE) &lt;- value

# S4 method for liger,character_OR_NULL
dataset(x, dataset = NULL)

# S4 method for liger,missing
dataset(x, dataset = NULL)

# S4 method for liger,numeric
dataset(x, dataset = NULL)

# S4 method for liger,character,missing,ANY,ligerDataset
dataset(x, dataset, type = NULL, qc = TRUE) &lt;- value

# S4 method for liger,character,ANY,ANY,matrixLike
dataset(x, dataset, type = c("rawData", "normData", "scaleData"), qc = FALSE) &lt;- value

# S4 method for liger,character,missing,ANY,`NULL`
dataset(x, dataset, type = NULL, qc = TRUE) &lt;- value

# S4 method for liger
names(x)

# S4 method for liger,character
names(x) &lt;- value

# S4 method for liger
length(x)

cellMeta(
  x,
  columns = NULL,
  useDataset = NULL,
  cellIdx = NULL,
  as.data.frame = FALSE,
  ...
)

cellMeta(x, columns = NULL, useDataset = NULL, cellIdx = NULL, check = FALSE) &lt;- value

# S4 method for liger,`NULL`
cellMeta(
  x,
  columns = NULL,
  useDataset = NULL,
  cellIdx = NULL,
  as.data.frame = FALSE,
  ...
)

# S4 method for liger,character
cellMeta(
  x,
  columns = NULL,
  useDataset = NULL,
  cellIdx = NULL,
  as.data.frame = FALSE,
  ...
)

# S4 method for liger,missing
cellMeta(
  x,
  columns = NULL,
  useDataset = NULL,
  cellIdx = NULL,
  as.data.frame = FALSE,
  ...
)

# S4 method for liger,missing
cellMeta(x, columns = NULL, useDataset = NULL, cellIdx = NULL, check = FALSE) &lt;- value

# S4 method for liger,character
cellMeta(x, columns = NULL, useDataset = NULL, cellIdx = NULL, check = FALSE) &lt;- value

# S4 method for liger,ANY,missing
[[(x, i, j, ...)

# S4 method for liger,ANY,missing
[[(x, i, j, ...) &lt;- value

# S4 method for liger
$(x, name)

# S4 method for liger
$(x, name) &lt;- value

defaultCluster(x, useDatasets = NULL, ...)

defaultCluster(x, name = NULL, useDatasets = NULL, ...) &lt;- value

# S4 method for liger
defaultCluster(x, useDatasets = NULL, droplevels = FALSE, ...)

# S4 method for liger,ANY,ANY,character
defaultCluster(x, name = NULL, useDatasets = NULL, ...) &lt;- value

# S4 method for liger,ANY,ANY,factor
defaultCluster(x, name = NULL, useDatasets = NULL, droplevels = TRUE, ...) &lt;- value

# S4 method for liger,ANY,ANY,`NULL`
defaultCluster(x, name = NULL, useDatasets = NULL, ...) &lt;- value

dimRed(x, name = NULL, useDatasets = NULL, ...)

dimRed(x, name = NULL, useDatasets = NULL, ...) &lt;- value

# S4 method for liger,missing
dimRed(x, name = NULL, useDatasets = NULL, ...)

# S4 method for liger,character
dimRed(x, name = NULL, useDatasets = NULL, ...)

# S4 method for liger,character,ANY,matrixLike
dimRed(x, name = NULL, useDatasets = NULL, asDefault = NULL, ...) &lt;- value

defaultDimRed(x, useDatasets = NULL)

defaultDimRed(x, name, useDatasets = NULL) &lt;- value

# S4 method for liger
defaultDimRed(x, useDatasets = NULL)

# S4 method for liger,missing,ANY,character
defaultDimRed(x, name = NULL, useDatasets = NULL) &lt;- value

# S4 method for liger,character,ANY,matrixLike
defaultDimRed(x, name, useDatasets = NULL) &lt;- value

varFeatures(x)

varFeatures(x, check = TRUE) &lt;- value

# S4 method for liger
varFeatures(x)

# S4 method for liger,ANY,character
varFeatures(x, check = TRUE) &lt;- value

# S3 method for liger
fortify(model, data, ...)

# S3 method for liger
c(...)

commands(x, funcName = NULL, arg = NULL)

# S4 method for liger
commands(x, funcName = NULL, arg = NULL)

# S4 method for liger
rawData(x, dataset = NULL)

# S4 method for liger,ANY,ANY,matrixLike_OR_NULL
rawData(x, dataset = NULL, check = TRUE) &lt;- value

# S4 method for liger,ANY,ANY,H5D
rawData(x, dataset = NULL, check = TRUE) &lt;- value

# S4 method for liger
normData(x, dataset = NULL)

# S4 method for liger,ANY,ANY,matrixLike_OR_NULL
normData(x, dataset = NULL, check = TRUE) &lt;- value

# S4 method for liger,ANY,ANY,H5D
normData(x, dataset = NULL, check = TRUE) &lt;- value

# S4 method for liger,ANY
scaleData(x, dataset = NULL)

# S4 method for liger,ANY,ANY,matrixLike_OR_NULL
scaleData(x, dataset = NULL, check = TRUE) &lt;- value

# S4 method for liger,ANY,ANY,H5D
scaleData(x, dataset = NULL, check = TRUE) &lt;- value

# S4 method for liger,ANY,ANY,H5Group
scaleData(x, dataset = NULL, check = TRUE) &lt;- value

# S4 method for liger,character
scaleUnsharedData(x, dataset = NULL)

# S4 method for liger,numeric
scaleUnsharedData(x, dataset = NULL)

# S4 method for liger,ANY,ANY,ANY
getMatrix(
  x,
  slot = c("rawData", "normData", "scaleData", "scaleUnsharedData", "H", "V", "U", "A",
    "B", "W", "H.norm"),
  dataset = NULL,
  returnList = FALSE
)

# S4 method for liger,ANY
getH5File(x, dataset = NULL)</code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x, object, model</dt>
<dd><p>A liger object</p></dd>


<dt>value</dt>
<dd><p>Check detail sections for requirements.</p></dd>


<dt>i, j</dt>
<dd><p>Feature and cell index for <code>`[`</code> method. For <code>`[[`</code>
method, use a single variable name with <code>i</code> while <code>j</code> is not
applicable.</p></dd>


<dt>...</dt>
<dd><p>See detailed sections for explanation.</p></dd>


<dt>drop</dt>
<dd><p>Not applicable.</p></dd>


<dt>check</dt>
<dd><p>Logical, whether to perform object validity check on setting new
value. Users are not supposed to set <code>FALSE</code> here.</p></dd>


<dt>dataset</dt>
<dd><p>Name or numeric index of a dataset</p></dd>


<dt>type</dt>
<dd><p>When using <code>dataset&lt;-</code> with a matrix like <code>value</code>,
specify what type the matrix is. Choose from <code>"rawData"</code>,
<code>"normData"</code> or <code>"scaleData"</code>.</p></dd>


<dt>qc</dt>
<dd><p>Logical, whether to perform general qc on added new dataset.</p></dd>


<dt>columns</dt>
<dd><p>The names of available variables in <code>cellMeta</code> slot. When
<code>as.data.frame = TRUE</code>, please use variable names after coercion.</p></dd>


<dt>cellIdx</dt>
<dd><p>Valid cell subscription to subset retrieved variables. Default
<code>NULL</code> uses all cells.</p></dd>


<dt>as.data.frame</dt>
<dd><p>Logical, whether to apply
<code><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></code> on the subscription. Default <code>FALSE</code>.</p></dd>


<dt>name</dt>
<dd><p>The name of available variables in <code>cellMeta</code> slot or the
name of a new variable to store.</p></dd>


<dt>useDatasets</dt>
<dd><p>Setter or getter method should only apply on cells in
specified datasets. Any valid character, numeric or logical subscriber is
acceptable. Default <code>NULL</code> works with all datasets.</p></dd>


<dt>data</dt>
<dd><p>fortify method required argument. Not used.</p></dd>


<dt>funcName, arg</dt>
<dd><p>See Command records section.</p></dd>


<dt>slot</dt>
<dd><p>Name of slot to retrieve matrix from. Options shown in Usage.</p></dd>


<dt>returnList</dt>
<dd><p>Logical, whether to force return a list even when only one
dataset-specific matrix (i.e. expression matrices, H, V or U) is requested.
Default <code>FALSE</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>See detailed sections for explanetion.</p>
    </div>
    <div id="slots">
    <h2>Slots</h2>
    

<dl><dt><code>datasets</code></dt>
<dd><p>list of <a href="ligerDataset-class.html">ligerDataset</a> objects. Use generic
<code>dataset</code>, <code>dataset&lt;-</code>, <code>datasets</code> or <code>datasets&lt;-</code> to
interact with. See detailed section accordingly.</p></dd>


<dt><code>cellMeta</code></dt>
<dd><p>DFrame object for cell metadata. Pre-existing
metadata, QC metrics, cluster labeling, low-dimensional embedding and etc.
are all stored here. Use generic <code>cellMeta</code>, <code>cellMeta&lt;-</code>,
<code>$</code>, <code>[[]]</code> or <code>[[]]&lt;-</code> to interact with. See detailed section
accordingly.</p></dd>


<dt><code>varFeatures</code></dt>
<dd><p>Character vector of feature names. Use generic
<code>varFeatures</code> or <code>varFeatures&lt;-</code> to interact with. See detailed
section accordingly.</p></dd>


<dt><code>W</code></dt>
<dd><p>Matrix of gene loading for each factor. See
<code><a href="runIntegration.html">runIntegration</a></code>.</p></dd>


<dt><code>H.norm</code></dt>
<dd><p>Matrix of aligned factor loading for each cell. See
<code><a href="quantileNorm.html">quantileNorm</a></code> and <code><a href="runIntegration.html">runIntegration</a></code>.</p></dd>


<dt><code>commands</code></dt>
<dd><p>List of <a href="ligerCommand-class.html">ligerCommand</a> objects. Record of
analysis. Use <code>commands</code> to retrieve information. See detailed section
accordingly.</p></dd>


<dt><code>uns</code></dt>
<dd><p>List for unstructured meta-info of analyses or presets.</p></dd>


<dt><code>version</code></dt>
<dd><p>Record of version of rliger2 package</p></dd>


</dl></div>
    <div id="dimensionality">
    <h2>Dimensionality</h2>
    

<p>For a <code>liger</code> object, the column orientation is assigned for
cells. Due to the data structure, it is hard to define a row index for the
<code>liger</code> object, which might contain datasets that vary in number of
genes.</p>
<p>Therefore, for <code>liger</code> objects, <code>dim</code> and <code>dimnames</code> returns
<code>NA</code>/<code>NULL</code> for rows and total cell counts/barcodes for the
columns.</p>
<p>For direct call of <code>dimnames&lt;-</code> method, <code>value</code> should be a list
with <code>NULL</code> as the first element and valid cell identifiers as the
second element. For <code>colnames&lt;-</code> method, the character vector of cell
identifiers. <code>rownames&lt;-</code> method is not applicable.</p>
    </div>
    <div id="subsetting">
    <h2>Subsetting</h2>
    

<p>For more detail of subsetting a <code>liger</code> object or a
<a href="ligerDataset-class.html">ligerDataset</a> object, please check out <code><a href="subsetLiger.html">subsetLiger</a></code>
and <code><a href="subsetLigerDataset.html">subsetLigerDataset</a></code>. Here, we set the S4 method
"single-bracket" <code>[</code> as a quick wrapper to subset a <code>liger</code> object.
Note that <code>j</code> serves as cell subscriptor which can be any valid index
refering the collection of all cells (i.e. <code>rownames(cellMeta(obj))</code>).
While <code>i</code>, the feature subscriptor can only be character vector because
the features for each dataset can vary. <code>...</code> arugments are passed to
<code>subsetLiger</code> so that advanced options are allowed.</p>
    </div>
    <div id="dataset-access">
    <h2>Dataset access</h2>
    

<p><code>datasets()</code> method only accesses the <code>datasets</code> slot, the list of
<a href="ligerDataset-class.html">ligerDataset</a> objects. <code>dataset()</code> method accesses a single
dataset, with subsequent cell metadata updates and checks bonded when adding
or modifying a dataset. Therefore, when users want to modify something inside
a <code>ligerDataset</code> while no cell metadata change should happen, it is
recommended to use: <code>datasets(x)[[name]] &lt;- ligerD</code> for efficiency,
though the result would be the same as <code>dataset(x, name) &lt;- ligerD</code>.</p>
<p><code><a href="https://rdrr.io/r/base/length.html" class="external-link">length()</a></code> and <code><a href="https://rdrr.io/r/base/names.html" class="external-link">names()</a></code> methods are implemented to access the
number and names of datasets. <code>names&lt;-</code> method is supported for
modifying dataset names, with taking care of the "dataset" variable in cell
metadata.</p>
    </div>
    <div id="matrix-access">
    <h2>Matrix access</h2>
    

<p>For <code>liger</code> object, <code><a href="ligerDataset-class.html">rawData()</a></code>, <code>normData</code>,
<code><a href="ligerDataset-class.html">scaleData()</a></code> and <code><a href="ligerDataset-class.html">scaleUnsharedData()</a></code> methods are exported for
users to access the corresponding feature expression matrix with
specification of one dataset. For retrieving a type of matrix from multiple
datasets, please use <code><a href="ligerDataset-class.html">getMatrix()</a></code> method.</p>
<p>When only one matrix is expected to be retrieved by <code><a href="ligerDataset-class.html">getMatrix()</a></code>, the
matrix itself will be returned. A list will be returned if multiple matrices
is requested (by querying multiple datasets) or <code>returnList</code> is set to
<code>TRUE</code>.</p>
    </div>
    <div id="cell-metadata-access">
    <h2>Cell metadata access</h2>
    

<p>Three approaches are provided for access of cell metadata. A generic function
<code>cellMeta</code> is implemented with plenty of options and multi-variable
accessibility. Besides, users can use double-bracket (e.g.
<code>ligerObj[[varName]]</code>) or dollor-sign (e.g. <code>ligerObj$nUMI</code>) to
access or modify single variables.</p>
<p>For users' convenience of generating a customized ggplot with available cell
metadata, the S3 method <code>fortify.liger</code> is implemented. With this under
the hook, users can create simple ggplots by directly starting with
<code>ggplot(ligerObj, aes(...))</code> where cell metadata variables can be
directly thrown into <code><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes()</a></code>.</p>
<p>The generic <code>defaultCluster</code> works as both getter and setter. As a
setter, users can do <code>defaultCluster(obj) &lt;- "existingVariableName"</code> to
set a categorical variable as default cluster used for visualization or
downstream analysis. Users can also do <code>defaultCluster(obj,
"newVarName") &lt;- factorOfLabels</code> to push new labeling into the object and set
as default. For getter method, the function returns a factor object of the
default cluster labeling. Argument <code>useDatasets</code> can be used for
requiring that given or retrieved labeling should match with cells in
specified datasets. We generally don't recommend setting <code>"dataset"</code> as
a default cluster because it is a preserved (always existing) field in
metadata and can lead to meaningless result when running analysis that
utilizes both clustering information and the dataset source information.</p>
    </div>
    <div id="dimension-reduction-access">
    <h2>Dimension reduction access</h2>
    

<p>Currently, low-dimensional representaion of cells, presented as dense
matrices, are all stored in <code>cellMeta</code> slot, and can totally be accessed
with generics <code>cellMeta</code> and <code>cellMeta&lt;-</code>. In addition to that,
we provide specific generics <code>dimRed</code> and <code>dimRed&lt;-</code> for getting
and setting matrix like cell metadata, respectively. Adding a matrix to the
object looks as simple as <code>dimRed(obj, "name") &lt;- matrixLike</code>. It can
be retrived back with <code>dimRed(obj, "name")</code>. Similar to having a default
cluster labeling, we also constructed the feature of default dimRed. It can
be set with <code>defaultDimRed(obj) &lt;- "existingMatLikeVar"</code> and the matrix
can be retrieved with <code>defaultDimRed(obj)</code>.</p>
    </div>
    <div id="variable-feature-access">
    <h2>Variable feature access</h2>
    

<p>The <code>varFeatures</code> slot allows for character vectors of gene names.
<code>varFeatures(x)</code> returns this vector and <code>value</code> for
<code>varFeatures&lt;-</code> method has to be a character vector or <code>NULL</code>.
The replacement method, when <code>check = TRUE</code> performs checks on gene
name consistency check across the <code>scaleData</code>, <code>H</code>, <code>V</code> slots
of inner <code>ligerDataset</code> objects as well as the <code>W</code> and
<code>H.norm</code> slots of the input <code>liger</code> object.</p>
    </div>
    <div id="combining-multiple-liger-object">
    <h2>Combining multiple liger object</h2>
    <p>The list of <code>datasets</code> slot,
the rows of <code>cellMeta</code> slot and the list of <code>commands</code> slot will
be simply concatenated. Variable features in <code>varFeatures</code> slot will be
taken a union. The \(W\) and \(H.norm\) matrices are not taken into
account for now.</p>
    </div>
    <div id="command-records">
    <h2>Command records</h2>
    

<p>rliger functions, that perform calculation and update the <code>liger</code>
object, will be recorded in a <code>ligerCommand</code> object and stored in the
<code>commands</code> slot, a list, of <code>liger</code> object. Method
<code>commands()</code> is implemented to retrieve or show the log history.
Running with <code>funcName = NULL</code> (default) returns all command labels.
Specifying <code>funcName</code> allows partial matching to all command labels
and returns a subset list (of <code>ligerCommand</code> object) of matches (or
the <code>ligerCommand</code> object if only one match found). If <code>arg</code> is
further specified, a subset list of parameters from the matches will be
returned. For example, requesting a list of resolution values used in
all louvain cluster attempts: <code>commands(ligerObj, "louvainCluster",
"resolution")</code></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Methods for base generics</span></span></span>
<span class="r-in"><span><span class="va">pbmcPlot</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class liger with 600 cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> datasets(2): ctrl (300 cells), stim (300 cells) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> cellMeta(4): dataset, nUMI, leiden_cluster, UMAP </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> varFeatures(50): ISG15, ID3, RPL11, ..., HIST1H2AC </span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class liger with 600 cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> datasets(2): ctrl (300 cells), stim (300 cells) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> cellMeta(4): dataset, nUMI, leiden_cluster, UMAP </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> varFeatures(50): ISG15, ID3, RPL11, ..., HIST1H2AC </span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]  NA 600</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 600</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ctrl_AAACATACCTCGCT.1" "ctrl_AAACGGCTCTTCGC.1" "ctrl_AACACTCTAAGTAG.1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4] "ctrl_AACCGCCTCAGGAG.1" "ctrl_AACGTTCTTCCGTC.1"</span>
<span class="r-in"><span><span class="va">pbmcPlot</span><span class="op">[</span><span class="fu">varFeatures</span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Nov  6 16:33:41 2023 ... Subsetting dataset: ctrl</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class liger with 10 cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> datasets(1): ctrl (10 cells) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> cellMeta(4): dataset, nUMI, leiden_cluster, UMAP </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> varFeatures(10): ISG15, ID3, RPL11, ..., S100A8 </span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ctrl" "stim"</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 2</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># rliger2 generics</span></span></span>
<span class="r-in"><span><span class="co">## Retrieving dataset(s), replacement methods available</span></span></span>
<span class="r-in"><span><span class="fu">datasets</span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $ctrl</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class ligerDataset with 300 cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> normData: 50 features</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $stim</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class ligerDataset with 300 cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> normData: 50 features</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu">dataset</span><span class="op">(</span><span class="va">pbmcPlot</span>, <span class="st">"ctrl"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class ligerDataset with 300 cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> normData: 50 features</span>
<span class="r-in"><span><span class="fu">dataset</span><span class="op">(</span><span class="va">pbmcPlot</span>, <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class ligerDataset with 300 cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> normData: 50 features</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Retrieving cell metadata, replacement methods available</span></span></span>
<span class="r-in"><span><span class="fu">cellMeta</span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 600 rows and 4 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                        dataset      nUMI leiden_cluster              UMAP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       &lt;factor&gt; &lt;numeric&gt;       &lt;factor&gt;          &lt;matrix&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AAACATACCTCGCT.1     ctrl      2151              1  6.70697:-4.70087</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AAACGGCTCTTCGC.1     ctrl      1916              1  6.11953:-3.43883</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AACACTCTAAGTAG.1     ctrl      1869              1  9.62906:-5.76486</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AACCGCCTCAGGAG.1     ctrl       733              0 -7.23223: 1.09265</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AACGTTCTTCCGTC.1     ctrl       573              0 -4.69289: 2.38874</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ...                        ...       ...            ...               ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> stim_TTCATGACTTATCC.1     stim      1091              3 11.50498:-3.79783</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> stim_TTCATGACTTCAGG.1     stim       431              0 -6.23811: 1.46327</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> stim_TTCGGAGATTTCAC.1     stim      1344              5  7.60062: 1.41078</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> stim_TTGACACTTCCTGC.1     stim      1423              5  7.42693: 1.01122</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> stim_TTTGCATGAACGAA.1     stim      2818              6  4.27674:-3.05098</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">[[</span><span class="st">"nUMI"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 2151 1916 1869  733  573 1872</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">$</span><span class="va">UMAP</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                               1          2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AAACATACCTCGCT.1  6.706970 -4.7008704</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AAACGGCTCTTCGC.1  6.119526 -3.4388267</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AACACTCTAAGTAG.1  9.629061 -5.7648618</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AACCGCCTCAGGAG.1 -7.232232  1.0926543</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AACGTTCTTCCGTC.1 -4.692893  2.3887372</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AAGAACGAAACGAA.1  7.474532 -0.7516093</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Retrieving variable features, replacement methods available</span></span></span>
<span class="r-in"><span><span class="fu">varFeatures</span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "ISG15"     "ID3"       "RPL11"     "MARCKSL1"  "RPS8"      "GBP1"     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7] "S100A10"   "S100A11"   "S100A9"    "S100A8"    "S100A6"    "S100A4"   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13] "RPS27"     "FCER1G"    "FCGR3A"    "XCL2"      "XCL1"      "SELL"     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [19] "RSAD2"     "RPS27A"    "GNLY"      "DUSP2"     "RPL31"     "IL1B"     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [25] "CXCR4"     "PTMA"      "RPL32"     "RPL15"     "RPL14"     "GPX1"     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [31] "TEX264"    "FGFBP2"    "RPL9"      "IL8"       "PPBP"      "CXCL3"    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [37] "CXCL10"    "PLAC8"     "H2AFZ"     "RPL34"     "ANXA5"     "RPS3A"    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [43] "GZMK"      "RPS23"     "CD14"      "CD74"      "RPS14"     "NPM1"     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [49] "CD83"      "HIST1H2AC"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Command record/history</span></span></span>
<span class="r-in"><span><span class="va">pbmcPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Nov  6 16:33:41 2023 ... Scaling dataset: ctrl</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Nov  6 16:33:41 2023 ... Scaling dataset: stim</span>
<span class="r-in"><span><span class="fu">commands</span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "scaleNotCenter.liger_eab47af92a"</span>
<span class="r-in"><span><span class="fu">commands</span><span class="op">(</span><span class="va">pbmcPlot</span>, funcName <span class="op">=</span> <span class="st">"scaleNotCenter"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> A liger command record, performed at 11-06-2023 16:33:41 EST </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Call: scaleNotCenter.liger(pbmcPlot) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Parameters:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     useDatasets : "ctrl", "stim" </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     features : "Long character with 50 elements: ISG15, ID3, RPL11, ..., HIST1H2AC" </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     verbose : TRUE </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># S3 methods</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">pbmcPlot</span>, <span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class liger with 1200 cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> datasets(4): ctrl (300 cells), stim (300 cells), ctrl (300 cells), stim (300 cells) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> cellMeta(4): dataset, nUMI, leiden_cluster, UMAP </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> varFeatures(50): ISG15, ID3, RPL11, ..., HIST1H2AC </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>package ‘ggplot2’ was built under R version 4.3.1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Need help getting started? Try the R Graphics Cookbook:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> https://r-graphics.org</span>
<span class="r-in"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">pbmcPlot</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">UMAP.1</span>, y <span class="op">=</span> <span class="va">UMAP.2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="liger-class-1.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

