<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>ligerDataset class — ligerDataset-class • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="ligerDataset class — ligerDataset-class"><meta property="og:description" content="Object for storing dastaset specific information. Will be embedded within a
higher level liger object"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/import_data.html">Import data</a>
    </li>
    <li>
      <a href="../articles/liger_object.html">Introduction to a liger object</a>
    </li>
    <li>
      <a href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>ligerDataset class</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/ligerDataset-class.R" class="external-link"><code>R/ligerDataset-class.R</code></a></small>
    <div class="hidden name"><code>ligerDataset-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Object for storing dastaset specific information. Will be embedded within a
higher level <a href="liger-class.html">liger</a> object</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre><code># S4 method for ligerDataset
show(object)

# S4 method for ligerDataset
dim(x)

# S4 method for ligerDataset
dimnames(x)

# S4 method for ligerDataset,list
dimnames(x) &lt;- value

# S4 method for ligerDataset,index,missing,ANY
[(x, i, j, ..., drop = TRUE)

# S4 method for ligerDataset,missing,index,ANY
[(x, i, j, ..., drop = TRUE)

# S4 method for ligerDataset,index,index,ANY
[(x, i, j, ..., drop = TRUE)

rawData(x, dataset = NULL)

rawData(x, dataset = NULL, check = TRUE) &lt;- value

# S4 method for ligerDataset
rawData(x, dataset = NULL)

# S4 method for ligerDataset,ANY,ANY,matrixLike_OR_NULL
rawData(x, dataset = NULL, check = TRUE) &lt;- value

# S4 method for ligerDataset,ANY,ANY,H5D
rawData(x, dataset = NULL, check = TRUE) &lt;- value

normData(x, dataset = NULL)

normData(x, dataset = NULL, check = TRUE) &lt;- value

# S4 method for ligerDataset
normData(x, dataset = NULL)

# S4 method for ligerDataset,ANY,ANY,matrixLike_OR_NULL
normData(x, dataset = NULL, check = TRUE) &lt;- value

# S4 method for ligerDataset,ANY,ANY,H5D
normData(x, dataset = NULL, check = TRUE) &lt;- value

scaleData(x, dataset = NULL)

scaleData(x, dataset = NULL, check = TRUE) &lt;- value

# S4 method for ligerDataset,missing
scaleData(x, dataset = NULL)

# S4 method for ligerDataset,ANY,ANY,matrixLike_OR_NULL
scaleData(x, dataset = NULL, check = TRUE) &lt;- value

# S4 method for ligerDataset,ANY,ANY,H5D
scaleData(x, dataset = NULL, check = TRUE) &lt;- value

# S4 method for ligerDataset,ANY,ANY,H5Group
scaleData(x, dataset = NULL, check = TRUE) &lt;- value

scaleUnsharedData(x, dataset = NULL)

scaleUnsharedData(x, check = TRUE) &lt;- value

# S4 method for ligerDataset,missing
scaleUnsharedData(x, dataset = NULL)

# S4 method for ligerDataset,ANY,matrixLike_OR_NULL
scaleUnsharedData(x, check = TRUE) &lt;- value

# S4 method for ligerDataset,ANY,H5D
scaleUnsharedData(x, check = TRUE) &lt;- value

# S4 method for ligerDataset,ANY,H5Group
scaleUnsharedData(x, check = TRUE) &lt;- value

getMatrix(x, slot = "rawData", dataset = NULL, returnList = FALSE)

# S4 method for ligerDataset,ANY,missing,missing
getMatrix(
  x,
  slot = c("rawData", "normData", "scaleData", "scaleUnsharedData", "H", "V", "U", "A",
    "B"),
  dataset = NULL
)

h5fileInfo(x, info = NULL)

h5fileInfo(x, info = NULL, check = TRUE) &lt;- value

# S4 method for ligerDataset
h5fileInfo(x, info = NULL)

# S4 method for ligerDataset
h5fileInfo(x, info = NULL, check = TRUE) &lt;- value

getH5File(x, dataset = NULL)

# S4 method for ligerDataset,missing
getH5File(x, dataset = NULL)

featureMeta(x, check = NULL)

featureMeta(x, check = TRUE) &lt;- value

# S4 method for ligerDataset
featureMeta(x, check = NULL)

# S4 method for ligerDataset
featureMeta(x, check = TRUE) &lt;- value

# S3 method for ligerDataset
cbind(x, ..., deparse.level = 1)</code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x, object</dt>
<dd><p>A <code>ligerDataset</code> object.</p></dd>


<dt>value</dt>
<dd><p>See detail sections for requirements</p></dd>


<dt>i, j</dt>
<dd><p>Feature and cell index for <code>`[`</code> method. For <code>`[[`</code>
method, use a single variable name with <code>i</code> and <code>j</code> is not
applicable.</p></dd>


<dt>...</dt>
<dd><p>See detailed sections for explanation.</p></dd>


<dt>drop</dt>
<dd><p>Not applicable.</p></dd>


<dt>dataset</dt>
<dd><p>Not applicable for <code>ligerDataset</code> methods.</p></dd>


<dt>check</dt>
<dd><p>Whether to perform object validity check on setting new value.</p></dd>


<dt>slot</dt>
<dd><p>The slot name when using <code>getMatrix</code>.</p></dd>


<dt>returnList</dt>
<dd><p>Not applicable for <code>ligerDataset</code> methods.</p></dd>


<dt>info</dt>
<dd><p>Name of the entry in <code>h5fileInfo</code> slot.</p></dd>


<dt>deparse.level</dt>
<dd><p>Not used here.</p></dd>

</dl></div>
    <div id="slots">
    <h2>Slots</h2>
    

<dl><dt><code>rawData</code></dt>
<dd><p>Raw data.</p></dd>


<dt><code>normData</code></dt>
<dd><p>Normalized data</p></dd>


<dt><code>scaleData</code></dt>
<dd><p>Scaled data, usually with subset variable features</p></dd>


<dt><code>scaleUnsharedData</code></dt>
<dd><p>Scaled data of features not shared with other
datasets</p></dd>


<dt><code>varUnsharedFeatures</code></dt>
<dd><p>Variable features not shared with other datasets</p></dd>


<dt><code>V</code></dt>
<dd><p>matrix</p></dd>


<dt><code>A</code></dt>
<dd><p>matrix</p></dd>


<dt><code>B</code></dt>
<dd><p>matrix</p></dd>


<dt><code>H</code></dt>
<dd><p>matrix</p></dd>


<dt><code>U</code></dt>
<dd><p>matrix</p></dd>


<dt><code>h5fileInfo</code></dt>
<dd><p>list</p></dd>


<dt><code>featureMeta</code></dt>
<dd><p>Feature metadata, DataFrame</p></dd>


<dt><code>colnames</code></dt>
<dd><p>character</p></dd>


<dt><code>rownames</code></dt>
<dd><p>character</p></dd>


</dl></div>
    <div id="dimensionality">
    <h2>Dimensionality</h2>
    

<p>For a <code>ligerDataset</code> object, the column orientation is assigned for
cells and rows are for features. Therefore, for <code>ligerDataset</code> objects,
<code><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim()</a></code> returns a numeric vector of two numbers which are number of
features and number of cells. <code><a href="https://rdrr.io/r/base/dimnames.html" class="external-link">dimnames()</a></code> returns a list of two
character vectors, which are the feature names and the cell barcodes.</p>
<p>For direct call of <code>dimnames&lt;-</code> method, <code>value</code> should be a list
with a character vector of feature names as the first element and cell
identifiers as the second element. For <code>colnames&lt;-</code> method, the
character vector of cell identifiers. For <code>rownames&lt;-</code> method, the
character vector of feature names.</p>
    </div>
    <div id="subsetting">
    <h2>Subsetting</h2>
    

<p>For more detail of subsetting a <code>liger</code> object or a
ligerDataset object, please check out <code><a href="subsetLiger.html">subsetLiger</a></code>
and <code><a href="subsetLigerDataset.html">subsetLigerDataset</a></code>. Here, we set the S4 method
"single-bracket" <code>[</code> as a quick wrapper to subset a <code>ligerDataset</code>
object. <code>i</code> and <code>j</code> serves as feature and cell subscriptor,
respectively, which can be any valid index refering the available features
and cells in a dataset. <code>...</code> arugments are passed to
<code>subsetLigerDataset</code> so that advanced options are allowed.</p>
    </div>
    <div id="matrix-access">
    <h2>Matrix access</h2>
    

<p>For <code>ligerDataset</code> object, <code>rawData()</code>, <code>normData</code>,
<code>scaleData()</code> and <code>scaleUnsharedData()</code> methods are exported for
users to access the corresponding feature expression matrix. Replacement
methods are also available to modify the slots.</p>
<p>For other matrices, such as the \(H\) and \(V\), which are dataset
specific, please use <code>getMatrix()</code> method with specifying slot name.
Directly accessing slot with <code>@</code> is generally not recommended.</p>
    </div>
    <div id="h-file-and-information-access">
    <h2>H5 file and information access</h2>
    

<p>A <code>ligerDataset</code> object has a slot called <code>h5fileInfo</code>, which is a
list object. The first element is called <code>$H5File</code>, which is an
<code>H5File</code> class object and is the connection to the input file. The
second element is <code>$filename</code> which stores the absolute path of the H5
file in the current machine. The third element <code>$formatType</code> stores the
name of preset being used, if applicable. The other following keys pair with
paths in the H5 file that point to specific data for constructing a feature
expression matrix.</p>
<p><code>h5fileInfo()</code> method access the list described above and simply
retrieves the corresponding value. When <code>info = NULL</code>, returns the whole
list. When <code>length(info) == 1</code>, returns the requested list value. When
more info requested, returns a subset list.</p>
<p>The replacement method modifies the list elements and corresponding slot
value (if applicable) at the same time. For example, running
<code>h5fileInfo(obj, "rawData") &lt;- newPath</code> not only updates the list, but
also updates the <code>rawData</code> slot with the <code>H5D</code> class data at
"newPath" in the <code>H5File</code> object.</p>
<p><code>getH5File()</code> is a wrapper and is equivalent to
<code>h5fileInfo(obj, "H5File")</code>.</p>
    </div>
    <div id="feature-metadata-access">
    <h2>Feature metadata access</h2>
    

<p>A slot <code>featureMeta</code> is included for each <code>ligerDataset</code> object.
This slot requires a <code><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame-class</a></code> object, which
is the same as <code>cellMeta</code> slot of a <a href="liger-class.html">liger</a> object. However,
the associated S4 methods only include access to the whole table for now.
Internal information access follows the same way as data.frame operation.
For example, <code>featureMeta(ligerD)$nCell</code> or
<code>featureMeta(ligerD)[varFeatures(ligerObj), "gene_var"]</code>.</p>
    </div>
    <div id="concatenate-ligerdataset">
    <h2>Concatenate ligerDataset</h2>
    

<p><code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind()</a></code> method is implemented for concatenating <code>ligerDataset</code>
objects by cells. When applying, all feature expression matrix will be merged
with taking a union of all features for the rows.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">ctrl</span> <span class="op">&lt;-</span> <span class="fu"><a href="liger-class.html">dataset</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"ctrl"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Methods for base generics</span></span></span>
<span class="r-in"><span><span class="va">ctrl</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class ligerDataset with 300 cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rawData: 266 features</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">ctrl</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class ligerDataset with 300 cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rawData: 266 features</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">ctrl</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 266 300</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">ctrl</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 300</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">ctrl</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 266</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">ctrl</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ctrl_AAACATACCTCGCT.1" "ctrl_AAACGGCTCTTCGC.1" "ctrl_AACACTCTAAGTAG.1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4] "ctrl_AACCGCCTCAGGAG.1" "ctrl_AACGTTCTTCCGTC.1"</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ctrl</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ISG15"    "ID3"      "RPL11"    "MARCKSL1" "RPS8"    </span>
<span class="r-in"><span><span class="va">ctrl</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class ligerDataset with 5 cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rawData: 5 features</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># rliger generics</span></span></span>
<span class="r-in"><span><span class="co">## raw data</span></span></span>
<span class="r-in"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">rawData</span><span class="op">(</span><span class="va">ctrl</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "dgCMatrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> attr(,"package")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Matrix"</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 266 300</span>
<span class="r-in"><span><span class="co">## normalized data</span></span></span>
<span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="normalize.html">normalize</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:10:55 2023 ... Normalizing dataset: ctrl</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:10:55 2023 ... Normalizing dataset: stim</span>
<span class="r-in"><span><span class="va">ctrl</span> <span class="op">&lt;-</span> <span class="fu"><a href="liger-class.html">dataset</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"ctrl"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">normData</span><span class="op">(</span><span class="va">ctrl</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "dgCMatrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> attr(,"package")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Matrix"</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 266 300</span>
<span class="r-in"><span><span class="co">## scaled data</span></span></span>
<span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:10:55 2023 ... Selecting variable features for dataset: ctrl</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:10:55 2023 ...... 168 features selected out of 249 shared features</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:10:55 2023 ... Selecting variable features for dataset: stim</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:10:55 2023 ...... 166 features selected out of 249 shared features</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:10:55 2023 ... Finally 173 shared variable features selected.</span>
<span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:10:56 2023 ... Scaling dataset: ctrl</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:10:56 2023 ... Scaling dataset: stim</span>
<span class="r-in"><span><span class="va">ctrl</span> <span class="op">&lt;-</span> <span class="fu"><a href="liger-class.html">dataset</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"ctrl"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">scaleData</span><span class="op">(</span><span class="va">ctrl</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "dgCMatrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> attr(,"package")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Matrix"</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 173 300</span>
<span class="r-in"><span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu">scaleData</span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"ctrl"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">m</span>, <span class="va">n</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span><span class="co">## Any other matrices</span></span></span>
<span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="rliger2-deprecated.html">online_iNMF</a></span><span class="op">(</span><span class="va">pbmc</span>, k <span class="op">=</span> <span class="fl">20</span>, miniBatch_size <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>`online_iNMF()` was deprecated in rliger2 1.99.0.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Please use the `method` argument of `runIntegration()` instead.</span>
<span class="r-in"><span><span class="va">ctrl</span> <span class="op">&lt;-</span> <span class="fu"><a href="liger-class.html">dataset</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"ctrl"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">V</span> <span class="op">&lt;-</span> <span class="fu">getMatrix</span><span class="op">(</span><span class="va">ctrl</span>, <span class="st">"V"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">V</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              Factor_1     Factor_2     Factor_3     Factor_4 Factor_5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ISG15    1.000000e-16 1.000000e-16 1.000000e-16 7.374195e-04    1e-16</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ID3      1.000000e-16 3.322680e-03 5.289525e-03 1.000000e-16    1e-16</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> RPL11    1.000000e-16 8.240158e-03 1.000000e-16 4.101785e-03    1e-16</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MARCKSL1 1.000000e-16 1.000000e-16 1.000000e-16 1.000000e-16    1e-16</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> RPS8     5.607769e-03 1.000000e-16 1.000000e-16 7.515360e-05    1e-16</span>
<span class="r-in"><span><span class="va">Vs</span> <span class="op">&lt;-</span> <span class="fu">getMatrix</span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"V"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">Vs</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 2</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">Vs</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ctrl" "stim"</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">Vs</span><span class="op">$</span><span class="va">ctrl</span>, <span class="va">V</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

