<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='Find DEG between two groups. Two methods are supported:
"wilcoxon" and "pseudoBulk". Wilcoxon rank sum test is
performed on single-cell level, while pseudo-bulk method aggregates cells
basing on biological replicates and calls bulk RNAseq DE methods, DESeq2 wald
test. When real biological replicates are not available, pseudo replicates
can be generated. Please see below for detailed scenario usage.'><title>Find DEG between two groups — runPairwiseDEG • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.8/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Find DEG between two groups — runPairwiseDEG"><meta property="og:description" content='Find DEG between two groups. Two methods are supported:
"wilcoxon" and "pseudoBulk". Wilcoxon rank sum test is
performed on single-cell level, while pseudo-bulk method aggregates cells
basing on biological replicates and calls bulk RNAseq DE methods, DESeq2 wald
test. When real biological replicates are not available, pseudo replicates
can be generated. Please see below for detailed scenario usage.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Find DEG between two groups</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/DEG_marker.R" class="external-link"><code>R/DEG_marker.R</code></a></small>
      <div class="d-none name"><code>liger-DEG.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Find DEG between two groups. Two methods are supported:
<code>"wilcoxon"</code> and <code>"pseudoBulk"</code>. Wilcoxon rank sum test is
performed on single-cell level, while pseudo-bulk method aggregates cells
basing on biological replicates and calls bulk RNAseq DE methods, DESeq2 wald
test. When real biological replicates are not available, pseudo replicates
can be generated. Please see below for detailed scenario usage.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runPairwiseDEG</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">groupTest</span>,</span>
<span>  <span class="va">groupCtrl</span>,</span>
<span>  variable1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  variable2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"wilcoxon"</span>, <span class="st">"pseudoBulk"</span><span class="op">)</span>,</span>
<span>  usePeak <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  useReplicate <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nPsdRep <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  minCellPerRep <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span>, <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">runMarkerDEG</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  conditionBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  splitBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"wilcoxon"</span>, <span class="st">"pseudoBulk"</span><span class="op">)</span>,</span>
<span>  useDatasets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  usePeak <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  useReplicate <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nPsdRep <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  minCellPerRep <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span>, <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">runWilcoxon</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  data.use <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  compare.method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clusters"</span>, <span class="st">"datasets"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object, with normalized data available</p></dd>


<dt>groupTest, groupCtrl, variable1, variable2</dt>
<dd><p>Condition specification. See
<code>?runPairwiseDEG</code> section <b>Pairwise DEG Scenarios</b> for detail.</p></dd>


<dt>method</dt>
<dd><p>DEG test method to use. Choose from <code>"wilcoxon"</code> or
<code>"pseudoBulk"</code>. Default <code>"wilcoxon"</code></p></dd>


<dt>usePeak</dt>
<dd><p>Logical. Whether to use peak count instead of gene count.
Only supported when ATAC datasets are involved. Default <code>FALSE</code>.</p></dd>


<dt>useReplicate</dt>
<dd><p><code>cellMeta</code> variable of biological replicate
annotation. Only used with <code>method = "pseudoBulk"</code>. Default <code>NULL</code>
will create <code>nPsdRep</code> pseudo replicates per group.</p></dd>


<dt>nPsdRep</dt>
<dd><p>Number of pseudo replicates to create. Only used when
<code>method = "pseudoBulk", useReplicate = NULL</code>. Default <code>5</code>.</p></dd>


<dt>minCellPerRep</dt>
<dd><p>Numeric, will not make pseudo-bulk for replicate with
less than this number of cells. Default <code>10</code>.</p></dd>


<dt>seed</dt>
<dd><p>Random seed to use for pseudo-replicate generation. Default
<code>1</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> or <code>TRUE</code> if users have not set.</p></dd>


<dt>conditionBy</dt>
<dd><p><code>cellMeta</code> variable(s). Marker detection will be
performed for each level of this variable. Multiple variables will be
combined. Default <code>NULL</code> uses default cluster.</p></dd>


<dt>splitBy</dt>
<dd><p>Split data by <code>cellMeta</code> variable(s) here and identify
markers for <code>conditionBy</code> within each chunk. Default <code>NULL</code>.</p></dd>


<dt>useDatasets</dt>
<dd><p>Datasets to perform marker detection within. Default
<code>NULL</code> will use all datasets.</p></dd>


<dt>data.use</dt>
<dd><p>Same as <code>useDatasets</code>.</p></dd>


<dt>compare.method</dt>
<dd><p>Choose from <code>"clusters"</code> (default) or
<code>"datasets"</code>. <code>"clusters"</code> compares each cluster against all other
cells, while <code>"datasets"</code> run within each cluster and compare each
dataset against all other datasets.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A data.frame with DEG information</p>
    </div>
    <div class="section level2">
    <h2 id="pairwise-deg-scenarios">Pairwise DEG Scenarios<a class="anchor" aria-label="anchor" href="#pairwise-deg-scenarios"></a></h2>
    

<p>Users can select classes of cells from a variable in <code>cellMeta</code>.
<code>variable1</code> and <code>variable2</code> are used to specify a column in
<code>cellMeta</code>, and <code>groupTest</code> and <code>groupCtrl</code> are used to specify
existing classes from <code>variable1</code> and <code>variable2</code>, respectively.
When <code>variable2</code> is missing, <code>groupCtrl</code> will be considered from
<code>variable1</code>.</p>
<p>For example, when <code>variable1 = "celltype"</code> and <code>variable2 = NULL</code>,
<code>groupTest</code> and <code>groupCtrl</code> should be valid cell types in
<code>object$celltype</code>.</p>
<p>When <code>variable1</code> is "celltype" and <code>variable2</code> is "gender",
<code>groupTest</code> should be a valid cell type from <code>object$celltype</code> and
<code>groupCtrl</code> should be a valid class from <code>object$gender</code>.</p>
<p>When both <code>variable1</code> and <code>variable2</code> are missing, <code>groupTest</code>
and <code>groupCtrl</code> should be valid index of cells in <code>object</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="marker-detection-scenarios">Marker Detection Scenarios<a class="anchor" aria-label="anchor" href="#marker-detection-scenarios"></a></h2>
    

<p>Marker detection is generally performed in a one vs. rest manner. The
grouping of such condition is specified by <code>conditionBy</code>, which should
be a column name in <code>cellMeta</code>. When <code>splitBy</code> is specified as
another variable name in <code>cellMeta</code>, the marker detection will be
iteratively done for within each level of <code>splitBy</code> variable.</p>
<p>For example, when <code>conditionBy = "celltype"</code> and <code>splitBy = NULL</code>,
marker detection will be performed by comparing all cells of "celltype_i"
against all other cells, and etc.</p>
<p>When <code>conditionBy = "celltype"</code> and <code>splitBy = "gender"</code>, marker
detection will be performed by comparing "celltype_i" cells from "gender_j"
against other cells from "gender_j", and etc.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Compare between cluster "0" and cluster "1"</span></span></span>
<span class="r-in"><span><span class="va">degStats</span> <span class="op">&lt;-</span> <span class="fu">runPairwiseDEG</span><span class="op">(</span><span class="va">pbmcPlot</span>, groupTest <span class="op">=</span> <span class="fl">0</span>, groupCtrl <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>                           variable1 <span class="op">=</span> <span class="st">"leiden_cluster"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running Wilcoxon rank-sum test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Running Wilcoxon rank-sum test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co"># Compare between all cells from cluster "5" and</span></span></span>
<span class="r-in"><span><span class="co"># all cells from dataset "stim"</span></span></span>
<span class="r-in"><span><span class="va">degStats</span> <span class="op">&lt;-</span> <span class="fu">runPairwiseDEG</span><span class="op">(</span><span class="va">pbmcPlot</span>, groupTest <span class="op">=</span> <span class="st">"5"</span>, groupCtrl <span class="op">=</span> <span class="st">"stim"</span>,</span></span>
<span class="r-in"><span>                           variable1 <span class="op">=</span> <span class="st">"leiden_cluster"</span>,</span></span>
<span class="r-in"><span>                           variable2 <span class="op">=</span> <span class="st">"dataset"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running Wilcoxon rank-sum test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Running Wilcoxon rank-sum test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co"># Identify markers for each cluster. Equivalent to old version</span></span></span>
<span class="r-in"><span><span class="co"># `runWilcoxon(method = "cluster")`</span></span></span>
<span class="r-in"><span><span class="va">markerStats</span> <span class="op">&lt;-</span> <span class="fu">runMarkerDEG</span><span class="op">(</span><span class="va">pbmcPlot</span>, conditionBy <span class="op">=</span> <span class="st">"leiden_cluster"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running Wilcoxon rank-sum test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Running Wilcoxon rank-sum test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co"># Identify dataset markers within each cluster. Equivalent to old version</span></span></span>
<span class="r-in"><span><span class="co"># `runWilcoxon(method = "dataset")`.</span></span></span>
<span class="r-in"><span><span class="va">markerStatsList</span> <span class="op">&lt;-</span> <span class="fu">runMarkerDEG</span><span class="op">(</span><span class="va">pbmcPlot</span>, conditionBy <span class="op">=</span> <span class="st">"dataset"</span>,</span></span>
<span class="r-in"><span>                                splitBy <span class="op">=</span> <span class="st">"leiden_cluster"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running Wilcoxon rank-sum test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Running Wilcoxon rank-sum test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running Wilcoxon rank-sum test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Running Wilcoxon rank-sum test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running Wilcoxon rank-sum test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Running Wilcoxon rank-sum test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running Wilcoxon rank-sum test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Running Wilcoxon rank-sum test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running Wilcoxon rank-sum test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Running Wilcoxon rank-sum test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running Wilcoxon rank-sum test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Running Wilcoxon rank-sum test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running Wilcoxon rank-sum test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Running Wilcoxon rank-sum test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running Wilcoxon rank-sum test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Running Wilcoxon rank-sum test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

