<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Find DEG between two groups — runPairwiseDEG • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Find DEG between two groups — runPairwiseDEG"><meta property="og:description" content='Find DEG between two groups. Two methods are supported:
"wilcoxon" and "pseudoBulk". Wilcoxon rank sum test is
performed on single-cell level, while pseudo-bulk method aggregates cells
basing on biological replicates and calls bulk RNAseq DE methods, DESeq2 wald
test. When real biological replicates are not available, pseudo replicates
can be generated. Please see below for detailed scenario usage.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown-header">Installation</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Import data</li>
    <li class="dropdown-header">Introduction to a liger object</li>
    <li class="dropdown-header">Interoperate with Seurat</li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Find DEG between two groups</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/DEG_marker.R" class="external-link"><code>R/DEG_marker.R</code></a></small>
    <div class="hidden name"><code>liger-DEG.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Find DEG between two groups. Two methods are supported:
<code>"wilcoxon"</code> and <code>"pseudoBulk"</code>. Wilcoxon rank sum test is
performed on single-cell level, while pseudo-bulk method aggregates cells
basing on biological replicates and calls bulk RNAseq DE methods, DESeq2 wald
test. When real biological replicates are not available, pseudo replicates
can be generated. Please see below for detailed scenario usage.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runPairwiseDEG</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">group1</span>,</span>
<span>  <span class="va">group2</span>,</span>
<span>  variable1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  variable2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"wilcoxon"</span>, <span class="st">"pseudoBulk"</span><span class="op">)</span>,</span>
<span>  usePeak <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  useReplicate <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nPsdRep <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">runMarkerDEG</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  conditionBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  splitBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"wilcoxon"</span>, <span class="st">"pseudoBulk"</span><span class="op">)</span>,</span>
<span>  useDatasets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  usePeak <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  useReplicate <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nPsdRep <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object, with normalized data available</p></dd>


<dt>group1, group2, variable1, variable2</dt>
<dd><p>Condition specification. See
<code>?runPairwiseDEG</code> section <b>Pairwise DEG Scenarios</b> for detail.</p></dd>


<dt>method</dt>
<dd><p>DEG test method to use. Choose from <code>"wilcoxon"</code> or
<code>"pseudoBulk"</code>. Default <code>"wilcoxon"</code></p></dd>


<dt>usePeak</dt>
<dd><p>Logical. Whether to use peak count instead of gene count.
Only supported when ATAC datasets are involved. Default <code>FALSE</code>.</p></dd>


<dt>useReplicate</dt>
<dd><p><code>cellMeta</code> variable of biological replicate
annotation. Only used with <code>method = "pseudoBulk"</code>. Default <code>NULL</code>
will create <code>nPsdRep</code> pseudo replicates per group.</p></dd>


<dt>nPsdRep</dt>
<dd><p>Number of pseudo replicates to create. Only used when
<code>method = "pseudoBulk", useReplicate = NULL</code>. Default <code>5</code>.</p></dd>


<dt>seed</dt>
<dd><p>Random seed to use for pseudo-replicate generation. Default
<code>1</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> which is <code>TRUE</code> if users have not set.</p></dd>


<dt>conditionBy</dt>
<dd><p><code>cellMeta</code> variable(s). Marker detection will be
performed for each level of this variable. Multiple variables will be
combined. Default <code>NULL</code> uses default cluster.</p></dd>


<dt>splitBy</dt>
<dd><p>Split data by <code>cellMeta</code> variable(s) here and identify
markers for <code>conditionBy</code> within each chunk. Default <code>NULL</code>.</p></dd>


<dt>useDatasets</dt>
<dd><p>Datasets to perform marker detection within. Default
<code>NULL</code> will use all datasets.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A data.frame with DEG information</p>
    </div>
    <div id="pairwise-deg-scenarios">
    <h2>Pairwise DEG Scenarios</h2>
    

<p>Users can select classes of cells from a variable in <code>cellMeta</code>.
<code>variable1</code> and <code>variable2</code> are used to specify a column in
<code>cellMeta</code>, and <code>group1</code> and <code>group2</code> are used to specify
existing classes from <code>variable1</code> and <code>variable2</code>, respectively.
When <code>variable2</code> is missing, <code>group2</code> will be considered from
<code>variable1</code>.</p>
<p>For example, when <code>variable1 = "celltype"</code> and <code>variable2 = NULL</code>,
<code>group1</code> and <code>group2</code> should be valid cell types in
<code>object$celltype</code>.</p>
<p>When <code>variable1</code> is "celltype" and <code>variable2</code> is "gender",
<code>group1</code> should be a valid cell type from <code>object$celltype</code> and
<code>group2</code> should be a valid class from <code>object$gender</code>.</p>
<p>When both <code>variable1</code> and <code>variable2</code> are missing, <code>group1</code>
and <code>group2</code> should be valid index of cells in <code>object</code>.</p>
    </div>
    <div id="marker-detection-scenarios">
    <h2>Marker Detection Scenarios</h2>
    

<p>Marker detection is generally performed in a one vs. rest manner. The
grouping of such condition is specified by <code>conditionBy</code>, which should
be a column name in <code>cellMeta</code>. When <code>splitBy</code> is specified as
another variable name in <code>cellMeta</code>, the marker detection will be
iteratively done for each level of <code>splitBy</code> variable.</p>
<p>For example, when <code>conditionBy = "celltype"</code> and <code>splitBy = NULL</code>,
marker detection will be performed by comparing all cells of "celltype_i"
against all other cells, and etc.</p>
<p>When <code>conditionBy = "celltype"</code> and <code>splitBy = "gender"</code>, marker
detection will be performed by comparing "celltype_i" cells from "gender_i"
against other cells from "gender_i", and etc.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Compare between cluster "0" and cluster "1"</span></span></span>
<span class="r-in"><span><span class="va">degStats</span> <span class="op">&lt;-</span> <span class="fu">runPairwiseDEG</span><span class="op">(</span><span class="va">pbmcPlot</span>, group1 <span class="op">=</span> <span class="fl">0</span>, group2 <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>                           variable1 <span class="op">=</span> <span class="st">"leiden_cluster"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Compare between all cells from cluster "5" and</span></span></span>
<span class="r-in"><span><span class="co"># all cells from dataset "stim"</span></span></span>
<span class="r-in"><span><span class="va">degStats</span> <span class="op">&lt;-</span> <span class="fu">runPairwiseDEG</span><span class="op">(</span><span class="va">pbmcPlot</span>, group1 <span class="op">=</span> <span class="st">"5"</span>, group2 <span class="op">=</span> <span class="st">"stim"</span>,</span></span>
<span class="r-in"><span>                           variable1 <span class="op">=</span> <span class="st">"leiden_cluster"</span>,</span></span>
<span class="r-in"><span>                           variable2 <span class="op">=</span> <span class="st">"dataset"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Identify markers for each cluster</span></span></span>
<span class="r-in"><span><span class="va">markerStats</span> <span class="op">&lt;-</span> <span class="fu">runMarkerDEG</span><span class="op">(</span><span class="va">pbmcPlot</span>, conditionBy <span class="op">=</span> <span class="st">"leiden_cluster"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Identify dataset markers within each cluster</span></span></span>
<span class="r-in"><span><span class="va">markerStatsList</span> <span class="op">&lt;-</span> <span class="fu">runMarkerDEG</span><span class="op">(</span><span class="va">pbmcPlot</span>, conditionBy <span class="op">=</span> <span class="st">"dataset"</span>,</span></span>
<span class="r-in"><span>                                splitBy <span class="op">=</span> <span class="st">"leiden_cluster"</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

