<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Linking genes to putative regulatory elements — linkGenesAndPeaks • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Linking genes to putative regulatory elements — linkGenesAndPeaks"><meta property="og:description" content="Evaluate the relationships between pairs of genes and peaks
based on specified distance metric. Usually used for inferring the
correlation between gene expression and imputed peak counts for datasets
without the modality originally (i.e. applied to imputeKNN
result)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown-header">Installation</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Import data</li>
    <li class="dropdown-header">Introduction to a liger object</li>
    <li class="dropdown-header">Interoperate with Seurat</li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Linking genes to putative regulatory elements</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/ATAC.R" class="external-link"><code>R/ATAC.R</code></a></small>
    <div class="hidden name"><code>linkGenesAndPeaks.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Evaluate the relationships between pairs of genes and peaks
based on specified distance metric. Usually used for inferring the
correlation between gene expression and imputed peak counts for datasets
without the modality originally (i.e. applied to <code><a href="imputeKNN.html">imputeKNN</a></code>
result).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">linkGenesAndPeaks</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">useDataset</span>,</span>
<span>  <span class="va">pathToCoords</span>,</span>
<span>  useGenes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"spearman"</span>, <span class="st">"pearson"</span>, <span class="st">"kendall"</span><span class="op">)</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span>,</span>
<span>  path_to_coords <span class="op">=</span> <span class="va">pathToCoords</span>,</span>
<span>  genes.list <span class="op">=</span> <span class="va">useGenes</span>,</span>
<span>  dist <span class="op">=</span> <span class="va">method</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object, with datasets that is of
<a href="ligerATACDataset-class.html">ligerATACDataset</a> class in the <code>datasets</code> slot.</p></dd>


<dt>useDataset</dt>
<dd><p>Name of one dataset, with both normalized gene expression
and normalized peak counts available.</p></dd>


<dt>pathToCoords</dt>
<dd><p>Path tothe gene coordinates file, usually a BED file.</p></dd>


<dt>useGenes</dt>
<dd><p>Character vector of gene names to be tested. Default
<code>NULL</code> uses all genes available in <code>useDataset</code>.</p></dd>


<dt>method</dt>
<dd><p>Choose the type of correlation to calculate, from
<code>"spearman"</code>, <code>"pearson"</code> and <code>"kendall"</code>. Default
<code>"spearman"</code></p></dd>


<dt>alpha</dt>
<dd><p>Numeric, significance threshold for correlation p-value.
Peak-gene correlations with p-values below this threshold are considered
significant. Default <code>0.05</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> which is <code>TRUE</code> if users have not set.</p></dd>


<dt>path_to_coords, genes.list, dist</dt>
<dd><p><b>Deprecated</b>. See Usage section
for replacement.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A sparse matrix with peak names as rows and gene names as columns,
with each element indicating the correlation between peak i and gene j, 0 if
the gene and peak are not significantly linked.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="imputeKNN.html">imputeKNN</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="normalize.html">normalize</a></span><span class="op">(</span><span class="va">bmmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:37:42 2023 ... Normalizing dataset: rna</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:37:42 2023 ... Normalizing dataset: atac</span>
<span class="r-in"><span><span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="va">bmmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:37:42 2023 ... Selecting variable features for dataset: rna</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:37:42 2023 ...... 83 features selected out of 172 shared features</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:37:42 2023 ... Selecting variable features for dataset: atac</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:37:42 2023 ...... 126 features selected out of 172 shared features</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:37:42 2023 ... Finally 135 shared variable features selected.</span>
<span class="r-in"><span><span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="va">bmmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:37:42 2023 ... Scaling dataset: rna</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:37:42 2023 ... Scaling dataset: atac</span>
<span class="r-in"><span><span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="rliger2-deprecated.html">online_iNMF</a></span><span class="op">(</span><span class="va">bmmc</span>, miniBatch_size <span class="op">=</span> <span class="fl">80</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="quantileNorm.html">quantileNorm</a></span><span class="op">(</span><span class="va">bmmc</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="normalize.html">normalizePeak</a></span><span class="op">(</span><span class="va">bmmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:37:42 2023 ... Normalizing rawPeak counts in dataset: atac</span>
<span class="r-in"><span><span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="imputeKNN.html">imputeKNN</a></span><span class="op">(</span><span class="va">bmmc</span>, reference <span class="op">=</span> <span class="st">"atac"</span>, queries <span class="op">=</span> <span class="st">"rna"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:37:42 2023 ... Imputing all the datasets exept the reference dataset</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>                              Reference dataset: atac</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>                              Query datasets: rna</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:37:42 2023 ... Normalizing rawPeak counts in dataset: rna</span>
<span class="r-in"><span><span class="va">corr</span> <span class="op">&lt;-</span> <span class="fu">linkGenesAndPeaks</span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">bmmc</span>, useDataset <span class="op">=</span> <span class="st">"rna"</span>,</span></span>
<span class="r-in"><span>    pathToCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/hg19_genes.bed"</span>, package <span class="op">=</span> <span class="st">"rliger2"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:37:42 2023 ... 172 genes to be tested against 995 peaks</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:37:42 2023 ... Calculating correlation for gene-peak pairs...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                                              |                                                                      |   0%  |                                                                              |=                                                                     |   2%  |                                                                              |==                                                                    |   2%  |                                                                              |==                                                                    |   3%  |                                                                              |====                                                                  |   5%  |                                                                              |====                                                                  |   6%  |                                                                              |=====                                                                 |   8%  |                                                                              |======                                                                |   8%  |                                                                              |=======                                                               |   9%  |                                                                              |========                                                              |  11%  |                                                                              |==========                                                            |  14%  |                                                                              |===========                                                           |  15%  |                                                                              |===========                                                           |  16%  |                                                                              |============                                                          |  17%  |                                                                              |=============                                                         |  18%  |                                                                              |=============                                                         |  19%  |                                                                              |===============                                                       |  22%  |                                                                              |================                                                      |  23%  |                                                                              |=================                                                     |  24%  |                                                                              |==================                                                    |  26%  |                                                                              |=====================                                                 |  30%  |                                                                              |======================                                                |  31%  |                                                                              |======================                                                |  32%  |                                                                              |========================                                              |  34%  |                                                                              |=========================                                             |  36%  |                                                                              |===========================                                           |  38%  |                                                                              |===========================                                           |  39%  |                                                                              |============================                                          |  40%  |                                                                              |============================                                          |  41%  |                                                                              |===============================                                       |  44%  |                                                                              |================================                                      |  46%  |                                                                              |=================================                                     |  47%  |                                                                              |=================================                                     |  48%  |                                                                              |==================================                                    |  48%  |                                                                              |==================================                                    |  49%  |                                                                              |=====================================                                 |  52%  |                                                                              |=====================================                                 |  53%  |                                                                              |======================================                                |  55%  |                                                                              |=======================================                               |  56%  |                                                                              |========================================                              |  57%  |                                                                              |========================================                              |  58%  |                                                                              |=========================================                             |  59%  |                                                                              |===========================================                           |  62%  |                                                                              |============================================                          |  62%  |                                                                              |==============================================                        |  65%  |                                                                              |==============================================                        |  66%  |                                                                              |===============================================                       |  67%  |                                                                              |================================================                      |  68%  |                                                                              |==================================================                    |  72%  |                                                                              |====================================================                  |  74%  |                                                                              |====================================================                  |  75%  |                                                                              |======================================================                |  77%  |                                                                              |=======================================================               |  78%  |                                                                              |=======================================================               |  79%  |                                                                              |========================================================              |  80%  |                                                                              |=========================================================             |  81%  |                                                                              |=========================================================             |  82%  |                                                                              |==========================================================            |  83%  |                                                                              |===========================================================           |  84%  |                                                                              |============================================================          |  85%  |                                                                              |=============================================================         |  87%  |                                                                              |==============================================================        |  89%  |                                                                              |=================================================================     |  93%  |                                                                              |==================================================================    |  94%  |                                                                              |====================================================================  |  98%  |                                                                              |===================================================================== |  98%  |                                                                              |======================================================================| 100%</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

