<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Perform online iNMF on scaled datasets — online_iNMF • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Perform online iNMF on scaled datasets — online_iNMF"><meta property="og:description" content="Perform online integrative non-negative matrix factorization to
represent multiple single-cell datasets in terms of \(H\), \(W\), and
\(V\) matrices. It optimizes the iNMF objective function using online
learning (non-negative least squares for H matrix, hierarchical alternating
least squares for \(W\) and \(V\) matrices), where the number of factors
is set by k. The function allows online learning in 3 scenarios:

Fully observed datasets;
Iterative refinement using continually arriving datasets;
Projection of new datasets without updating the existing factorization


All three scenarios require fixed memory independent of the number of cells.
For each dataset, this factorization produces an \(H\) matrix (k by cell),
a \(V\) matrix (genes by k), and a shared \(W\) matrix (genes by k). The
\(H\) matrices represent the cell factor loadings. \(W\) is identical
among all datasets, as it represents the shared components of the metagenes
across datasets. The \(V\) matrices represent the dataset-specific
components of the metagenes."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown-header">Installation</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Introductory Vignettes</li>
    <li class="dropdown-header">Integrating Multiple scRNAseq Data</li>
    <li class="dropdown-header">Integrating On-Disk (HDF5 Based) Data</li>
    <li class="divider">
    <li class="dropdown-header">Data structure and access</li>
    <li class="dropdown-header">Outer container - liger object</li>
    <li class="dropdown-header">Dataset container - ligerDataset object and subclasses</li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Import Data</li>
      </ul></li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform online iNMF on scaled datasets</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/online_iNMF.R" class="external-link"><code>R/online_iNMF.R</code></a></small>
    <div class="hidden name"><code>online_iNMF.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Perform online integrative non-negative matrix factorization to
represent multiple single-cell datasets in terms of \(H\), \(W\), and
\(V\) matrices. It optimizes the iNMF objective function using online
learning (non-negative least squares for H matrix, hierarchical alternating
least squares for \(W\) and \(V\) matrices), where the number of factors
is set by <code>k</code>. The function allows online learning in 3 scenarios:</p>
<ol><li><p>Fully observed datasets;</p></li>
<li><p>Iterative refinement using continually arriving datasets;</p></li>
<li><p>Projection of new datasets without updating the existing factorization</p></li>
</ol><p>All three scenarios require fixed memory independent of the number of cells.</p>
<p>For each dataset, this factorization produces an \(H\) matrix (k by cell),
a \(V\) matrix (genes by k), and a shared \(W\) matrix (genes by k). The
\(H\) matrices represent the cell factor loadings. \(W\) is identical
among all datasets, as it represents the shared components of the metagenes
across datasets. The \(V\) matrices represent the dataset-specific
components of the metagenes.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">online_iNMF</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  X_new <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  projection <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  W.init <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  V.init <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  A.init <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  B.init <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  lambda <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  max.epochs <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  miniBatch_max_iters <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  miniBatch_size <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  h5_chunk_size <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">123</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p><a href="liger-class.html">liger</a> object. Scaled data required.</p></dd>


<dt>X_new</dt>
<dd><p>New datasets for scenario 2 or scenario 3. See detail for usage.</p></dd>


<dt>projection</dt>
<dd><p>Perform data integration with scenario 3. See description.
Default <code>FALSE</code>.</p></dd>


<dt>W.init</dt>
<dd><p>Optional initialization for W. See detail. Default <code>NULL</code>.</p></dd>


<dt>V.init</dt>
<dd><p>Optional initialization for V. See detail. Default <code>NULL</code>.</p></dd>


<dt>A.init</dt>
<dd><p>Optional initialization for A. See detail. Default <code>NULL</code>.</p></dd>


<dt>B.init</dt>
<dd><p>Optional initialization for B. See detail. Default <code>NULL</code>.</p></dd>


<dt>k</dt>
<dd><p>Inner dimension of factorization--number of metagenes. A value in
the range 20-50 works well for most analyses. Default <code>20</code>.</p></dd>


<dt>lambda</dt>
<dd><p>Regularization parameter. Larger values penalize
dataset-specific effects more strongly (i.e. alignment should increase as
lambda increases). We recommend always using the default value except
possibly for analyses with relatively small differences (biological
replicates, male/female comparisons, etc.) in which case a lower value such
as 1.0 may improve reconstruction quality. Default <code>5.0</code>.</p></dd>


<dt>max.epochs</dt>
<dd><p>The number of epochs to iterate through. See detail.
Default <code>5</code>.</p></dd>


<dt>miniBatch_max_iters</dt>
<dd><p>Maximum number of block coordinate descent (HALS
algorithm) iterations to perform for each update of \(W\) and \(V\).
Default <code>1</code>. Changing this parameter is not recommended.</p></dd>


<dt>miniBatch_size</dt>
<dd><p>Total number of cells in each minibatch. See detail.
Default <code>5000</code>.</p></dd>


<dt>h5_chunk_size</dt>
<dd><p>Chunk size of input hdf5 files. Default <code>1000</code>. The
chunk size should be no larger than the batch size.</p></dd>


<dt>seed</dt>
<dd><p>Random seed to allow reproducible results. Default <code>123</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> which is <code>TRUE</code> if users have not set.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p><code>object</code> with <code>W</code> slot updated with resulting \(W\)</p>


<p>matrix; the <code>H</code>, <code>V</code>, <code>A</code> and <code>B</code> slots of each
<a href="ligerDataset-class.html">ligerDataset</a> object in <code>datasets</code> slot is updated with the
corresponding result matrices.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>For optional initialization, <code>W.init</code> must be a matrix object with
number of rows equal to number of variable genes (denoted as <code>g</code>) and
number of columns equal to <code>k</code>. Any of <code>V.init</code>, <code>A</code> and
<code>B</code> must be a list object of n matrices where n is the number of
datasets in <code>object</code>. For <code>V.init</code>, each matrix should be of size g
x k. For <code>A.init</code>, each matrix should be k x k and for <code>B.init</code>,
each matrix should be g x k.</p>
<p>Minibatch iterations is performed on small subset of cells. The exact
minibatch size applied on each dataset is <code>miniBatch_size</code> multiplied by
the proportion of cells in this dataset out of all cells. The setting of
<code>miniBatch_size</code> is by default <code>5000</code>, which is reasonable.
However, a smaller value such as <code>1000</code> may be necessary for analyzing
very small datasets. In general, <code>miniBatch_size</code> should be no larger
than the number of cells in the smallest dataset. An epoch is one completion
of calculation on all cells after a number of iterations of minibatches.
Therefore, the total number of iterations is determined by the setting of
<code>max.epochs</code>, total number of cells, and <code>miniBatch_size</code>.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="normalize.html">normalize</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Normalizing dataset: ctrl</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Normalizing dataset: stim</span>
<span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Selecting HVG for dataset: ctrl</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ...   168 shared variable features selected</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Selecting HVG for dataset: stim</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ...   166 shared variable features selected</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Finally 173 shared variable features selected.</span>
<span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Scaling dataset: ctrl</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Scaling dataset: stim</span>
<span class="r-in"><span><span class="co"># Minibatch size has to be less than number of cell in the smallest dataset</span></span></span>
<span class="r-in"><span><span class="co"># Scenario 1</span></span></span>
<span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">online_iNMF</span><span class="op">(</span><span class="va">pbmc</span>, miniBatch_size <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... 2 new datasets detected.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Scenario 1, init arguments ignored.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Initialization done.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Starting Online iNMF...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                                              |==                                                                    |   3%  |                                                                              |=====                                                                 |   7%  |                                                                              |=======                                                               |  10%  |                                                                              |=========                                                             |  13%  |                                                                              |============                                                          |  17%  |                                                                              |==============                                                        |  20%  |                                                                              |================                                                      |  23%  |                                                                              |===================                                                   |  27%  |                                                                              |=====================                                                 |  30%  |                                                                              |=======================                                               |  33%  |                                                                              |==========================                                            |  37%  |                                                                              |============================                                          |  40%  |                                                                              |==============================                                        |  43%  |                                                                              |=================================                                     |  47%  |                                                                              |===================================                                   |  50%  |                                                                              |=====================================                                 |  53%  |                                                                              |========================================                              |  57%  |                                                                              |==========================================                            |  60%  |                                                                              |============================================                          |  63%  |                                                                              |===============================================                       |  67%  |                                                                              |=================================================                     |  70%  |                                                                              |===================================================                   |  73%  |                                                                              |======================================================                |  77%  |                                                                              |========================================================              |  80%  |                                                                              |==========================================================            |  83%  |                                                                              |=============================================================         |  87%  |                                                                              |===============================================================       |  90%  |                                                                              |=================================================================     |  93%  |                                                                              |====================================================================  |  97%  |                                                                              |======================================================================| 100%</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Calculating metagene loadings...</span>
<span class="r-in"><span><span class="co"># Scenario 2</span></span></span>
<span class="r-in"><span><span class="co"># Fake new dataset by increasing all non-zero value in "ctrl" by 1</span></span></span>
<span class="r-in"><span><span class="va">ctrl2</span> <span class="op">&lt;-</span> <span class="fu"><a href="ligerDataset-class.html">rawData</a></span><span class="op">(</span><span class="fu"><a href="liger-class.html">dataset</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"ctrl"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ctrl2</span><span class="op">@</span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">ctrl2</span><span class="op">@</span><span class="va">x</span> <span class="op">+</span> <span class="fl">1</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">ctrl2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">ctrl2</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pbmc2</span> <span class="op">&lt;-</span> <span class="fu">online_iNMF</span><span class="op">(</span><span class="va">pbmc</span>, k <span class="op">=</span> <span class="fl">20</span>, X_new <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ctrl2 <span class="op">=</span> <span class="va">ctrl2</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                     miniBatch_size <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking in memory ligerDataset validity</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking liger object validity</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Normalizing dataset: ctrl2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Scaling dataset: ctrl2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... 1 new datasets detected.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Scenario 2, initiating parameters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Initialization done.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Starting Online iNMF...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                                              |=====                                                                 |   7%  |                                                                              |=========                                                             |  13%  |                                                                              |==============                                                        |  20%  |                                                                              |===================                                                   |  27%  |                                                                              |=======================                                               |  33%  |                                                                              |============================                                          |  40%  |                                                                              |=================================                                     |  47%  |                                                                              |=====================================                                 |  53%  |                                                                              |==========================================                            |  60%  |                                                                              |===============================================                       |  67%  |                                                                              |===================================================                   |  73%  |                                                                              |========================================================              |  80%  |                                                                              |=============================================================         |  87%  |                                                                              |=================================================================     |  93%  |                                                                              |======================================================================| 100%</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Calculating metagene loadings...</span>
<span class="r-in"><span><span class="co"># Scenario 3</span></span></span>
<span class="r-in"><span><span class="va">pbmc3</span> <span class="op">&lt;-</span> <span class="fu">online_iNMF</span><span class="op">(</span><span class="va">pbmc</span>, k <span class="op">=</span> <span class="fl">20</span>, X_new <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ctrl2 <span class="op">=</span> <span class="va">ctrl2</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                     miniBatch_size <span class="op">=</span> <span class="fl">100</span>, projection <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking in memory ligerDataset validity</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking liger object validity</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Normalizing dataset: ctrl2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Scaling dataset: ctrl2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... 1 new datasets detected.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Scenario 3, metagene projection</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko, Yichen Wang.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

