<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>[Deprecated] Perform online iNMF on scaled datasets — online_iNMF-deprecated • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="[Deprecated] Perform online iNMF on scaled datasets — online_iNMF-deprecated"><meta property="og:description" content="Please turn to runOnlineINMF or
runIntegration.
Perform online integrative non-negative matrix factorization to represent
multiple single-cell datasets in terms of H, W, and V matrices. It optimizes
the iNMF objective function using online learning (non-negative least squares
for H matrix, hierarchical alternating least squares for W and V matrices),
where the number of factors is set by k. The function allows online learning
in 3 scenarios: (1) fully observed datasets; (2) iterative refinement using
continually arriving datasets; and (3) projection of new datasets without
updating the existing factorization. All three scenarios require fixed memory
independent of the number of cells.
For each dataset, this factorization produces an H matrix (cells by k), a V
matrix (k by genes), and a shared W matrix (k by genes). The H matrices
represent the cell factor loadings. W is identical among all datasets, as it
represents the shared components of the metagenes across datasets. The V
matrices represent the dataset-specific components of the metagenes."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/import_data.html">Import data</a>
    </li>
    <li>
      <a href="../articles/liger_object.html">Introduction to a liger object</a>
    </li>
    <li>
      <a href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>[Deprecated] Perform online iNMF on scaled datasets</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/integration.R" class="external-link"><code>R/integration.R</code></a></small>
    <div class="hidden name"><code>online_iNMF-deprecated.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><b>Please turn to <code><a href="runOnlineINMF.html">runOnlineINMF</a></code> or
<code><a href="runIntegration.html">runIntegration</a></code></b>.</p>
<p>Perform online integrative non-negative matrix factorization to represent
multiple single-cell datasets in terms of H, W, and V matrices. It optimizes
the iNMF objective function using online learning (non-negative least squares
for H matrix, hierarchical alternating least squares for W and V matrices),
where the number of factors is set by k. The function allows online learning
in 3 scenarios: (1) fully observed datasets; (2) iterative refinement using
continually arriving datasets; and (3) projection of new datasets without
updating the existing factorization. All three scenarios require fixed memory
independent of the number of cells.</p>
<p>For each dataset, this factorization produces an H matrix (cells by k), a V
matrix (k by genes), and a shared W matrix (k by genes). The H matrices
represent the cell factor loadings. W is identical among all datasets, as it
represents the shared components of the metagenes across datasets. The V
matrices represent the dataset-specific components of the metagenes.</p>
    </div>


    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p><code>liger</code> object with data stored in HDF5 files. Should
normalize, select genes, and scale before calling.</p></dd>


<dt>X_new</dt>
<dd><p>List of new datasets for scenario 2 or scenario 3. Each list
element should be the name of an HDF5 file.</p></dd>


<dt>projection</dt>
<dd><p>Perform data integration by shared metagene (W) projection
(scenario 3). (default FALSE)</p></dd>


<dt>W.init</dt>
<dd><p>Optional initialization for W. (default NULL)</p></dd>


<dt>V.init</dt>
<dd><p>Optional initialization for V (default NULL)</p></dd>


<dt>H.init</dt>
<dd><p>Optional initialization for H (default NULL)</p></dd>


<dt>A.init</dt>
<dd><p>Optional initialization for A (default NULL)</p></dd>


<dt>B.init</dt>
<dd><p>Optional initialization for B (default NULL)</p></dd>


<dt>k</dt>
<dd><p>Inner dimension of factorization--number of metagenes (default 20).
A value in the range 20-50 works well for most analyses.</p></dd>


<dt>lambda</dt>
<dd><p>Regularization parameter. Larger values penalize
dataset-specific effects more
  strongly (ie. alignment should increase as lambda increases). We recommend
  always using the default value except
  possibly for analyses with relatively small differences (biological
  replicates, male/female comparisons, etc.)
  in which case a lower value such as 1.0 may improve reconstruction quality.
  (default 5.0).</p></dd>


<dt>max.epochs</dt>
<dd><p>Maximum number of epochs (complete passes through the
data). (default 5)</p></dd>


<dt>miniBatch_max_iters</dt>
<dd><p>Maximum number of block coordinate descent (HALS
algorithm) iterations to perform for each update of W and V (default 1).
Changing this parameter is not  recommended.</p></dd>


<dt>miniBatch_size</dt>
<dd><p>Total number of cells in each minibatch (default 5000).
This is a reasonable default, but a smaller value such as 1000 may be
necessary for analyzing very small datasets. In general, minibatch size
should be no larger than the number of cells in the smallest dataset.</p></dd>


<dt>h5_chunk_size</dt>
<dd><p>Chunk size of input hdf5 files (default 1000). The chunk
size should be no larger than the batch size.</p></dd>


<dt>seed</dt>
<dd><p>Random seed to allow reproducible results (default 123).</p></dd>


<dt>verbose</dt>
<dd><p>Print progress bar/messages (TRUE by default)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p><code>liger</code> object with H, W, V, A and B slots set.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

