<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Please turn to runOnlineINMF or
runIntegration.
Perform online integrative non-negative matrix factorization to represent
multiple single-cell datasets in terms of H, W, and V matrices. It optimizes
the iNMF objective function using online learning (non-negative least squares
for H matrix, hierarchical alternating least squares for W and V matrices),
where the number of factors is set by k. The function allows online learning
in 3 scenarios: (1) fully observed datasets; (2) iterative refinement using
continually arriving datasets; and (3) projection of new datasets without
updating the existing factorization. All three scenarios require fixed memory
independent of the number of cells.
For each dataset, this factorization produces an H matrix (cells by k), a V
matrix (k by genes), and a shared W matrix (k by genes). The H matrices
represent the cell factor loadings. W is identical among all datasets, as it
represents the shared components of the metagenes across datasets. The V
matrices represent the dataset-specific components of the metagenes."><title> Perform online iNMF on scaled datasets — online_iNMF-deprecated • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content=" Perform online iNMF on scaled datasets — online_iNMF-deprecated"><meta property="og:description" content="Please turn to runOnlineINMF or
runIntegration.
Perform online integrative non-negative matrix factorization to represent
multiple single-cell datasets in terms of H, W, and V matrices. It optimizes
the iNMF objective function using online learning (non-negative least squares
for H matrix, hierarchical alternating least squares for W and V matrices),
where the number of factors is set by k. The function allows online learning
in 3 scenarios: (1) fully observed datasets; (2) iterative refinement using
continually arriving datasets; and (3) projection of new datasets without
updating the existing factorization. All three scenarios require fixed memory
independent of the number of cells.
For each dataset, this factorization produces an H matrix (cells by k), a V
matrix (k by genes), and a shared W matrix (k by genes). The H matrices
represent the cell factor loadings. W is identical among all datasets, as it
represents the shared components of the metagenes across datasets. The V
matrices represent the dataset-specific components of the metagenes."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    <a class="dropdown-item" href="../articles/benchmark.html">Benchmarking integration quality</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1><a href="https://lifecycle.r-lib.org/articles/stages.html#deprecated" class="external-link"><img src="figures/lifecycle-deprecated.svg" alt="[Deprecated]"></a> Perform online iNMF on scaled datasets</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/integration.R" class="external-link"><code>R/integration.R</code></a></small>
      <div class="d-none name"><code>online_iNMF-deprecated.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><b>Please turn to <code><a href="runOnlineINMF.html">runOnlineINMF</a></code> or
<code><a href="runIntegration.html">runIntegration</a></code></b>.</p>
<p>Perform online integrative non-negative matrix factorization to represent
multiple single-cell datasets in terms of H, W, and V matrices. It optimizes
the iNMF objective function using online learning (non-negative least squares
for H matrix, hierarchical alternating least squares for W and V matrices),
where the number of factors is set by k. The function allows online learning
in 3 scenarios: (1) fully observed datasets; (2) iterative refinement using
continually arriving datasets; and (3) projection of new datasets without
updating the existing factorization. All three scenarios require fixed memory
independent of the number of cells.</p>
<p>For each dataset, this factorization produces an H matrix (cells by k), a V
matrix (k by genes), and a shared W matrix (k by genes). The H matrices
represent the cell factor loadings. W is identical among all datasets, as it
represents the shared components of the metagenes across datasets. The V
matrices represent the dataset-specific components of the metagenes.</p>
    </div>


    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p><code>liger</code> object with data stored in HDF5 files. Should
normalize, select genes, and scale before calling.</p></dd>


<dt>X_new</dt>
<dd><p>List of new datasets for scenario 2 or scenario 3. Each list
element should be the name of an HDF5 file.</p></dd>


<dt>projection</dt>
<dd><p>Perform data integration by shared metagene (W) projection
(scenario 3). (default FALSE)</p></dd>


<dt>W.init</dt>
<dd><p>Optional initialization for W. (default NULL)</p></dd>


<dt>V.init</dt>
<dd><p>Optional initialization for V (default NULL)</p></dd>


<dt>H.init</dt>
<dd><p>Optional initialization for H (default NULL)</p></dd>


<dt>A.init</dt>
<dd><p>Optional initialization for A (default NULL)</p></dd>


<dt>B.init</dt>
<dd><p>Optional initialization for B (default NULL)</p></dd>


<dt>k</dt>
<dd><p>Inner dimension of factorization--number of metagenes (default 20).
A value in the range 20-50 works well for most analyses.</p></dd>


<dt>lambda</dt>
<dd><p>Regularization parameter. Larger values penalize
dataset-specific effects more
strongly (ie. alignment should increase as lambda increases). We recommend
always using the default value except
possibly for analyses with relatively small differences (biological
replicates, male/female comparisons, etc.)
in which case a lower value such as 1.0 may improve reconstruction quality.
(default 5.0).</p></dd>


<dt>max.epochs</dt>
<dd><p>Maximum number of epochs (complete passes through the
data). (default 5)</p></dd>


<dt>miniBatch_max_iters</dt>
<dd><p>Maximum number of block coordinate descent (HALS
algorithm) iterations to perform for each update of W and V (default 1).
Changing this parameter is not  recommended.</p></dd>


<dt>miniBatch_size</dt>
<dd><p>Total number of cells in each minibatch (default 5000).
This is a reasonable default, but a smaller value such as 1000 may be
necessary for analyzing very small datasets. In general, minibatch size
should be no larger than the number of cells in the smallest dataset.</p></dd>


<dt>h5_chunk_size</dt>
<dd><p>Chunk size of input hdf5 files (default 1000). The chunk
size should be no larger than the batch size.</p></dd>


<dt>seed</dt>
<dd><p>Random seed to allow reproducible results (default 123).</p></dd>


<dt>verbose</dt>
<dd><p>Print progress bar/messages (TRUE by default)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p><code>liger</code> object with H, W, V, A and B slots set.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

