<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Perform iNMF on scaled datasets — optimizeALS • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Perform iNMF on scaled datasets — optimizeALS"><meta property="og:description" content="Performs integrative non-negative matrix (iNMF) factorization to return
factorized \(H\), \(W\), and \(V\) matrices. It optimizes the iNMF
objective function using block coordinate descent (alternating non-negative
least squares), where the number of factors is set by k. TODO: include
objective function equation here in documentation (using deqn)
For each dataset, this factorization produces an \(H\) matrix (cells by k),
a \(V\) matrix (k by genes), and a shared \(W\) matrix (k by genes). The
\(H\) matrices represent the cell factor loadings. \(W\) is held
consistent among all datasets, as it represents the shared components of the
metagenes across datasets. The \(V\) matrices represent the
dataset-specific components of the metagenes."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown-header">Installation</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Introductory Vignettes</li>
    <li class="dropdown-header">Integrating Multiple scRNAseq Data</li>
    <li class="dropdown-header">Integrating On-Disk (HDF5 Based) Data</li>
    <li class="divider">
    <li class="dropdown-header">Data structure and access</li>
    <li class="dropdown-header">Outer container - liger object</li>
    <li class="dropdown-header">Dataset container - ligerDataset object and subclasses</li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Import Data</li>
      </ul></li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform iNMF on scaled datasets</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/optimizeALS.R" class="external-link"><code>R/optimizeALS.R</code></a></small>
    <div class="hidden name"><code>optimizeALS.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Performs integrative non-negative matrix (iNMF) factorization to return
factorized \(H\), \(W\), and \(V\) matrices. It optimizes the iNMF
objective function using block coordinate descent (alternating non-negative
least squares), where the number of factors is set by <code>k</code>. TODO: include
objective function equation here in documentation (using deqn)</p>
<p>For each dataset, this factorization produces an \(H\) matrix (cells by k),
a \(V\) matrix (k by genes), and a shared \(W\) matrix (k by genes). The
\(H\) matrices represent the cell factor loadings. \(W\) is held
consistent among all datasets, as it represents the shared components of the
metagenes across datasets. The \(V\) matrices represent the
dataset-specific components of the metagenes.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">optimizeALS</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">k</span>,</span>
<span>  lambda <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  thresh <span class="op">=</span> <span class="fl">1e-06</span>,</span>
<span>  maxIter <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  nrep <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  H.init <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  W.init <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  V.init <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useUnshared <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  readH5 <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span>,</span>
<span>  max.iters <span class="op">=</span> <span class="va">maxIter</span>,</span>
<span>  use.unshared <span class="op">=</span> <span class="va">useUnshared</span>,</span>
<span>  rand.seed <span class="op">=</span> <span class="va">seed</span>,</span>
<span>  print.obj <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for liger</span></span>
<span><span class="fu">optimizeALS</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">k</span>,</span>
<span>  lambda <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  thresh <span class="op">=</span> <span class="fl">1e-06</span>,</span>
<span>  maxIter <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  nrep <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  H.init <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  W.init <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  V.init <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useUnshared <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  readH5 <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span>,</span>
<span>  max.iters <span class="op">=</span> <span class="va">maxIter</span>,</span>
<span>  use.unshared <span class="op">=</span> <span class="va">useUnshared</span>,</span>
<span>  rand.seed <span class="op">=</span> <span class="va">seed</span>,</span>
<span>  print.obj <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for list</span></span>
<span><span class="fu">optimizeALS</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">k</span>,</span>
<span>  lambda <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  thresh <span class="op">=</span> <span class="fl">1e-06</span>,</span>
<span>  maxIter <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  nrep <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  H.init <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  W.init <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  V.init <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useUnshared <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  readH5 <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span>,</span>
<span>  max.iters <span class="op">=</span> <span class="va">maxIter</span>,</span>
<span>  use.unshared <span class="op">=</span> <span class="va">useUnshared</span>,</span>
<span>  rand.seed <span class="op">=</span> <span class="va">seed</span>,</span>
<span>  print.obj <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object or a named list of matrix object,
where the names represents dataset names and matrices are scaled on the same
set of variable features, with rows as features and columns as cells.</p></dd>


<dt>k</dt>
<dd><p>Inner dimension of factorization (number of factors). Run
<code><a href="suggestK.html">suggestK</a></code> to determine appropriate value; a general rule of
thumb is that a higher <code>k</code> will be needed for datasets with more
sub-structure.</p></dd>


<dt>lambda</dt>
<dd><p>Regularization parameter. Larger values penalize
dataset-specific effects more strongly (i.e. alignment should increase as
<code>lambda</code> increases). Default <code>5</code>.</p></dd>


<dt>thresh</dt>
<dd><p>Convergence threshold. Convergence occurs when
\(|obj_0-obj|/(mean(obj_0,obj)) &lt; thresh\). Default <code>1e-6</code>.</p></dd>


<dt>maxIter</dt>
<dd><p>Maximum number of block coordinate descent iterations to
perform. Default <code>30</code>.</p></dd>


<dt>nrep</dt>
<dd><p>Number of restarts to perform (iNMF objective function is
non-convex, so taking the best objective from multiple successive
initialization is recommended). For easier reproducibility, this increments
the random seed by 1 for each consecutive restart, so future factorization
of the same dataset can be run with one rep if necessary. Default <code>1</code>.</p></dd>


<dt>H.init</dt>
<dd><p>Initial values to use for \(H\) matrices. A list object where
each element is the initial \(H\) matrix of each dataset. Default
<code>NULL</code>.</p></dd>


<dt>W.init</dt>
<dd><p>Initial values to use for \(W\) matrix. A matrix object.
Default <code>NULL</code>.</p></dd>


<dt>V.init</dt>
<dd><p>Initial values to use for \(V\) matrices. A list object where
each element is the initial \(V\) matrix of each dataset. Default
<code>NULL</code>.</p></dd>


<dt>useUnshared</dt>
<dd><p>Logical, whether to include unshared variable features and
run optimizeUANLS algorithm. Defaul <code>FALSE</code>. Running
<code><a href="selectGenes.html">selectGenes</a></code> with <code>unshared = TRUE</code> and then running
<code><a href="scaleNotCenter.html">scaleNotCenter</a></code> is required.</p></dd>


<dt>seed</dt>
<dd><p>Random seed to allow reproducible results. Default <code>1</code>.</p></dd>


<dt>readH5</dt>
<dd><p><code>TRUE</code> to force reading H5 based data into memory and
conduct factorization. <code>"auto"</code> reads H5 dataset with less than 8000
cells. <code>FALSE</code> will stop users from running if H5 data presents.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> which is <code>TRUE</code> if users have not set.</p></dd>


<dt>max.iters, use.unshared, rand.seed</dt>
<dd><p><b>Deprecated</b>. See Usage section
for replacement.</p></dd>


<dt>print.obj</dt>
<dd><p><b>Defunct</b>. Whether to print objective function values
after convergence when <code>verbose = TRUE</code>. Now always print when verbose.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p><code>object</code> with <code>W</code> slot updated with the result \(W\)</p>


<p>matrix, and the <code>H</code> and <code>V</code> slots of each
<a href="ligerDataset-class.html">ligerDataset</a> object in the <code>datasets</code> slot updated with
the dataset specific \(H\) and \(V\) matrix, respectively.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="normalize.html">normalize</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Normalizing dataset: ctrl</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Normalizing dataset: stim</span>
<span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Selecting HVG for dataset: ctrl</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ...   168 shared variable features selected</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Selecting HVG for dataset: stim</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ...   166 shared variable features selected</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Finally 173 shared variable features selected.</span>
<span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Scaling dataset: ctrl</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:35 2023 ... Scaling dataset: stim</span>
<span class="r-in"><span><span class="co"># Only running a few iterations for fast examples</span></span></span>
<span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">optimizeALS</span><span class="op">(</span><span class="va">pbmc</span>, k <span class="op">=</span> <span class="fl">20</span>, maxIter <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:37 2023 ... Start iNMF with seed: 1...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                                              |                                                                      |   0%  |                                                                              |===================================                                   |  50%  |                                                                              |======================================================================| 100%</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:40 2023 ... Finished in 3.6766140460968 secs, 2 iterations. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>                              Max iterations set: 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>                              Final objective delta: 0.278833145358608</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:40 2023 ... Objective: 44307.2622916006</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Apr 24 14:49:40 2023 ... Best results with seed 1</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko, Yichen Wang.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

