<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Generate t-SNE plots and gene loading plots — plotGeneLoadingsOld • rliger</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Generate t-SNE plots and gene loading plots — plotGeneLoadingsOld"><meta property="og:description" content="Plots t-SNE coordinates of all cells by their loadings on each factor. Underneath it displays the
most highly loading shared and dataset-specific genes, along with the overall gene loadings
for each dataset.
It is recommended to call this function into a PDF due to the large number of
plots produced."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown-header">Installation</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Introductory Vignettes</li>
    <li class="dropdown-header">PBMC guided tutorial</li>
    <li class="dropdown-header">Analysis with on-disk (HDF5 based) data</li>
    <li class="divider">
    <li class="dropdown-header">Data structure and access</li>
    <li class="dropdown-header">Outer container - liger object</li>
    <li class="dropdown-header">Dataset container - ligerDataset object and subclasses</li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Import Data</li>
      </ul></li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate t-SNE plots and gene loading plots</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/visualization.R" class="external-link"><code>R/visualization.R</code></a></small>
    <div class="hidden name"><code>plotGeneLoadingsOld.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plots t-SNE coordinates of all cells by their loadings on each factor. Underneath it displays the
most highly loading shared and dataset-specific genes, along with the overall gene loadings
for each dataset.</p>
<p>It is recommended to call this function into a PDF due to the large number of
plots produced.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotGeneLoadingsOld</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  dataset1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dataset2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  num.genes.show <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  num.genes <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  mark.top.genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  factor.share.thresh <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  log.fc.thresh <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  umi.thresh <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  frac.thresh <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  pval.thresh <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  do.spec.plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  max.val <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  pt.size <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>  option <span class="op">=</span> <span class="st">"plasma"</span>,</span>
<span>  zero.color <span class="op">=</span> <span class="st">"#F5F5F5"</span>,</span>
<span>  return.plots <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  axis.labels <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  do.title <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  raster <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p><code>liger</code> object. Should call runTSNE before calling.</p></dd>


<dt>dataset1</dt>
<dd><p>Name of first dataset (by default takes first two datasets for dataset1 and 2)</p></dd>


<dt>dataset2</dt>
<dd><p>Name of second dataset</p></dd>


<dt>num.genes.show</dt>
<dd><p>Number of genes displayed as y-axis labels in the gene loading plots at
the bottom (default 12)</p></dd>


<dt>num.genes</dt>
<dd><p>Number of genes to show in word clouds (default 30).</p></dd>


<dt>mark.top.genes</dt>
<dd><p>Plot points corresponding to top loading genes in different color (default
TRUE).</p></dd>


<dt>factor.share.thresh</dt>
<dd><p>Use only factors with a dataset specificity less than or equal to
threshold (default 10).</p></dd>


<dt>log.fc.thresh</dt>
<dd><p>Lower log-fold change threshold for differential expression in markers
(default 1).</p></dd>


<dt>umi.thresh</dt>
<dd><p>Lower UMI threshold for markers (default 30).</p></dd>


<dt>frac.thresh</dt>
<dd><p>Lower threshold for fraction of cells expressing marker (default 0).</p></dd>


<dt>pval.thresh</dt>
<dd><p>Upper p-value threshold for Wilcoxon rank test for gene expression
(default 0.05).</p></dd>


<dt>do.spec.plot</dt>
<dd><p>Include dataset specificity plot in printout (default TRUE).</p></dd>


<dt>max.val</dt>
<dd><p>Value between 0 and 1 at which color gradient should saturate to max color. Set to
NULL to revert to default gradient scaling. (default 0.1)</p></dd>


<dt>pt.size</dt>
<dd><p>Point size for plots (default 0.4).</p></dd>


<dt>option</dt>
<dd><p>Colormap option to use for ggplot2's scale_color_viridis (default 'plasma').</p></dd>


<dt>zero.color</dt>
<dd><p>Color to use for zero values (no expression) (default '#F5F5F5').</p></dd>


<dt>return.plots</dt>
<dd><p>Return ggplot objects instead of printing directly (default FALSE).</p></dd>


<dt>axis.labels</dt>
<dd><p>Vector of two strings to use as x and y labels respectively (default NULL).</p></dd>


<dt>do.title</dt>
<dd><p>Include top title with cluster and Dataset Specificity (default FALSE).</p></dd>


<dt>verbose</dt>
<dd><p>Print progress bar/messages (TRUE by default)</p></dd>


<dt>raster</dt>
<dd><p>Rasterization of points (default NULL). Automatically convert to raster format if
there are over 100,000 cells to plot.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>List of ggplot plot objects (only if return.plots TRUE, otherwise prints plots to
  console).</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># ligerex (liger object based on in-memory datasets), factorization complete</span></span></span>
<span class="r-in"><span><span class="va">ligerex</span> <span class="op">&lt;-</span> <span class="fu">quantile_norm</span><span class="op">(</span><span class="va">ligerex</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ligerex</span> <span class="op">&lt;-</span> <span class="fu"><a href="runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">ligerex</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># pdf("gene_loadings.pdf")</span></span></span>
<span class="r-in"><span><span class="fu">plotGeneLoadings</span><span class="op">(</span><span class="va">ligerex</span>, num.genes <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># dev.off()</span></span></span>
<span class="r-in"><span><span class="co"># ligerex (liger object based on datasets in HDF5 format), factorization complete input</span></span></span>
<span class="r-in"><span><span class="va">ligerex</span> <span class="op">&lt;-</span> <span class="fu"><a href="readSubset.html">readSubset</a></span><span class="op">(</span><span class="va">ligerex</span>, slot.use <span class="op">=</span> <span class="st">"norm.data"</span>, max.cells <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">plotGeneLoadings</span><span class="op">(</span><span class="va">ligerex</span>, num.genes <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko, Yichen Wang.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

