<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='plotVolcano is a simple implementation and shares most of arguments
with other rliger plotting functions. plotEnhancedVolcano is a
wrapper function of EnhancedVolcano::EnhancedVolcano(), which has
provides substantial amount of arguments for graphical control. However, that
requires the installation of package "EnhancedVolcano".
highlight and labelTopN both controls the feature name
labeling, whereas highlight is considered first. If both are as
default (NULL), all significant features will be labeled.'><title>Create volcano plot for Wilcoxon test result — plotVolcano • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Create volcano plot for Wilcoxon test result — plotVolcano"><meta property="og:description" content='plotVolcano is a simple implementation and shares most of arguments
with other rliger plotting functions. plotEnhancedVolcano is a
wrapper function of EnhancedVolcano::EnhancedVolcano(), which has
provides substantial amount of arguments for graphical control. However, that
requires the installation of package "EnhancedVolcano".
highlight and labelTopN both controls the feature name
labeling, whereas highlight is considered first. If both are as
default (NULL), all significant features will be labeled.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    <a class="dropdown-item" href="../articles/benchmark.html">Benchmarking integration quality</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Create volcano plot for Wilcoxon test result</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/visualization.R" class="external-link"><code>R/visualization.R</code></a></small>
      <div class="d-none name"><code>plotVolcano.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>plotVolcano</code> is a simple implementation and shares most of arguments
with other rliger plotting functions. <code>plotEnhancedVolcano</code> is a
wrapper function of <code><a href="https://rdrr.io/pkg/EnhancedVolcano/man/EnhancedVolcano.html" class="external-link">EnhancedVolcano::EnhancedVolcano()</a></code>, which has
provides substantial amount of arguments for graphical control. However, that
requires the installation of package "EnhancedVolcano".</p>
<p><code>highlight</code> and <code>labelTopN</code> both controls the feature name
labeling, whereas <code>highlight</code> is considered first. If both are as
default (<code>NULL</code>), all significant features will be labeled.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotVolcano</span><span class="op">(</span></span>
<span>  <span class="va">result</span>,</span>
<span>  group <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  logFCThresh <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  padjThresh <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  highlight <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  labelTopN <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dotSize <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  dotAlpha <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  legendPosition <span class="op">=</span> <span class="st">"top"</span>,</span>
<span>  labelSize <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>result</dt>
<dd><p>Data frame table returned by <code><a href="liger-DEG.html">runMarkerDEG</a></code> or
<code><a href="liger-DEG.html">runPairwiseDEG</a></code>.</p></dd>


<dt>group</dt>
<dd><p>Selection of one group available from <code>result$group</code>. If
only one group is available from <code>result</code>, default <code>NULL</code> uses it.</p></dd>


<dt>logFCThresh</dt>
<dd><p>Number for the threshold on the absolute value of the log2
fold change statistics. Default <code>1</code>.</p></dd>


<dt>padjThresh</dt>
<dd><p>Number for the threshold on the adjusted p-value
statistics. Default <code>0.01</code>.</p></dd>


<dt>highlight</dt>
<dd><p>A character vector of feature names to be highlighted.
Default <code>NULL</code>.</p></dd>


<dt>labelTopN</dt>
<dd><p>Number of top differential expressed features to be labeled
on the top of the dots. Ranked by adjusted p-value first and absolute value
of logFC next. Default <code>NULL</code>.</p></dd>


<dt>dotSize, dotAlpha</dt>
<dd><p>Numbers for universal aesthetics control of dots.
Default <code>2</code> and <code>0.8</code>.</p></dd>


<dt>legendPosition</dt>
<dd><p>Text indicating where to place the legend. Choose from
<code>"top"</code>, <code>"bottom"</code>, <code>"left"</code> or <code>"right"</code>. Default
<code>"top"</code>.</p></dd>


<dt>labelSize</dt>
<dd><p>Size of labeled top features and line annotations. Default
<code>4</code>.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed on to <code><a href="dot-ggScatter.html">.ggScatter</a></code>, <code><a href="dot-ggplotLigerTheme.html">.ggplotLigerTheme</a></code></p><dl><dt><code>dotOrder</code></dt>
<dd><p>Controls the order that each dot is added to the plot. Choose
from <code>"shuffle"</code>, <code>"ascending"</code>, or <code>"descending"</code>. Default
<code>"shuffle"</code>, useful when coloring by categories that overlaps (e.g.
"dataset"), <code>"ascending"</code> can be useful when coloring by a continuous
variable (e.g. gene expression) where high values needs more
highlight. <code>NULL</code> use default order.</p></dd>

    <dt><code>raster</code></dt>
<dd><p>Logical, whether to rasterize the plot. Default <code>NULL</code>
automatically rasterize the plot when number of total dots to be plotted
exceeds 100,000.</p></dd>

    <dt><code>labelText</code></dt>
<dd><p>Logical, whether to show text label at the median position
of each categorical group specified by <code>colorBy</code>. Default <code>TRUE</code>.
Does not work when continuous coloring is specified.</p></dd>

    <dt><code>labelTextSize</code></dt>
<dd><p>Numeric, controls the size of label size when
<code>labelText = TRUE</code>. Default <code>4</code>.</p></dd>

    <dt><code>seed</code></dt>
<dd><p>Random seed for reproducibility. Default <code>1</code>.</p></dd>

    <dt><code>legendColorTitle</code></dt>
<dd><p>Legend title text for color aesthetics, often used
for categorical or continuous coloring of dots. Default <code>NULL</code> shows the
original variable name.</p></dd>

    <dt><code>showLegend</code></dt>
<dd><p>Whether to show the legend. Default <code>TRUE</code>.</p></dd>

    <dt><code>baseSize</code></dt>
<dd><p>One-parameter control of all text sizes. Individual text
element sizes can be controlled by other size arguments. "Title" sizes are
2 points larger than "text" sizes when being controlled by this.</p></dd>

    <dt><code>titleSize,xTitleSize,yTitleSize,legendTitleSize</code></dt>
<dd><p>Size of main title,
axis titles and legend title. Default <code>NULL</code> controls by
<code>baseSize + 2</code>.</p></dd>

    <dt><code>subtitleSize,xTextSize,yTextSize,legendTextSize</code></dt>
<dd><p>Size of subtitle text,
axis texts and legend text. Default <code>NULL</code> controls by <code>baseSize</code>.</p></dd>

    <dt><code>panelBorder</code></dt>
<dd><p>Whether to show rectangle border of the panel instead of
using ggplot classic bottom and left axis lines. Default <code>FALSE</code>.</p></dd>

  
</dl></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>ggplot</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="liger-class.html">defaultCluster</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pbmcPlot</span><span class="op">$</span><span class="va">leiden_cluster</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Storing given cluster labels to `cellMeta(x)` field: <span style="color: #0000BB;">"defaultCluster"</span>.</span>
<span class="r-in"><span><span class="co"># Test the DEG between "stim" and "ctrl", within each cluster</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="liger-DEG.html">runPairwiseDEG</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">pbmc</span>,</span></span>
<span class="r-in"><span>    groupTest <span class="op">=</span> <span class="st">"stim"</span>,</span></span>
<span class="r-in"><span>    groupCtrl <span class="op">=</span> <span class="st">"ctrl"</span>,</span></span>
<span class="r-in"><span>    variable1 <span class="op">=</span> <span class="st">"dataset"</span>,</span></span>
<span class="r-in"><span>    splitBy <span class="op">=</span> <span class="st">"defaultCluster"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running DEG within: <span style="color: #0000BB;">"0"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Calling pairwise DESeq2 Wald test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running DEG within: <span style="color: #0000BB;">"1"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> -- note: fitType='parametric', but the dispersion trend was not well captured by the</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>    function: y = a/x + b, and a local regression fit was automatically substituted.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>    specify fitType='local' or 'mean' to avoid this message next time.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Calling pairwise DESeq2 Wald test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running DEG within: <span style="color: #0000BB;">"2"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Calling pairwise DESeq2 Wald test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running DEG within: <span style="color: #0000BB;">"3"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Calling pairwise DESeq2 Wald test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running DEG within: <span style="color: #0000BB;">"4"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Calling pairwise DESeq2 Wald test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running DEG within: <span style="color: #0000BB;">"5"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Calling pairwise DESeq2 Wald test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running DEG within: <span style="color: #0000BB;">"6"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Calling pairwise DESeq2 Wald test ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running DEG within: <span style="color: #0000BB;">"7"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> Ignoring replicates (size in bracket) with too few cells: <span style="color: #0000BB;">"others.ctrl.rep1 (2)"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Consider decrease <span style="color: #00BB00;">minCellPerRep</span> to exclude less replicates or/and lower <span style="color: #00BB00;">nPsdRep</span> to generate larger pseudo-replicates.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> Error when computing on <span style="color: #0000BB;">"7.stim"</span>: Too few replicates with more than <span style="color: #0000BB;">3</span> cells (`minCellPerRep`) for condition <span style="color: #0000BB;">"others"</span>.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> Empty result returned for this test.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calling pairwise DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> Calling pairwise DESeq2 Wald test ... failed</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu">plotVolcano</span><span class="op">(</span><span class="va">result</span>, <span class="st">"0.stim"</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>ggrepel: 6 unlabeled data points (too many overlaps). Consider increasing max.overlaps</span>
<span class="r-plt img"><img src="plotVolcano-1.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

