<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Visualize factor expression and gene loading"><title>Visualize factor expression and gene loading — plotGeneLoadings • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Visualize factor expression and gene loading — plotGeneLoadings"><meta property="og:description" content="Visualize factor expression and gene loading"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    <a class="dropdown-item" href="../articles/benchmark.html">Benchmarking integration quality</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Visualize factor expression and gene loading</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/visualization.R" class="external-link"><code>R/visualization.R</code></a></small>
      <div class="d-none name"><code>plotGeneLoadings.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Visualize factor expression and gene loading</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotGeneLoadings</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">markerTable</span>,</span>
<span>  <span class="va">useFactor</span>,</span>
<span>  useDimRed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nLabel <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  nPlot <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotGeneLoadingRank</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">markerTable</span>,</span>
<span>  <span class="va">useFactor</span>,</span>
<span>  nLabel <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  nPlot <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object with valid factorization result.</p></dd>


<dt>markerTable</dt>
<dd><p>Returned result of <code><a href="getFactorMarkers.html">getFactorMarkers</a></code>.</p></dd>


<dt>useFactor</dt>
<dd><p>Integer index for which factor to visualize.</p></dd>


<dt>useDimRed</dt>
<dd><p>Name of the variable storing dimensionality reduction result
in the <code>cellMeta</code> slot. Default <code>"UMAP"</code>.</p></dd>


<dt>nLabel</dt>
<dd><p>Integer, number of top genes to be shown with text labels.
Default <code>15</code>.</p></dd>


<dt>nPlot</dt>
<dd><p>Integer, number of top genes to be shown in the loading rank
plot. Default <code>30</code>.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed on to <code><a href="plotDimRed.html">plotDimRed</a></code>, <code><a href="dot-ggScatter.html">.ggScatter</a></code>, <code><a href="dot-ggplotLigerTheme.html">.ggplotLigerTheme</a></code></p><dl><dt><code>colorByFunc</code></dt>
<dd><p>Default <code>NULL</code>. A function object that expects a
vector/factor/data.frame retrieved by <code>colorBy</code> as the only input, and
returns an object of the same size, so that the all color "aes" are replaced
by this output. Useful when, for example, users need to scale the gene
expression shown on plot.</p></dd>

    <dt><code>cellIdx</code></dt>
<dd><p>Character, logical or numeric index that can subscribe cells.
Missing or <code>NULL</code> for all cells.</p></dd>

    <dt><code>shapeBy</code></dt>
<dd><p>Available variable name in <code>cellMeta</code> slot to look for
categorical annotation to be reflected by dot shapes. Default <code>NULL</code>.</p></dd>

    <dt><code>titles</code></dt>
<dd><p>Title text. A character scalar or a character vector with as
many elements as multiple plots are supposed to be generated. Default
<code>NULL</code>.</p></dd>

    <dt><code>dotSize,dotAlpha</code></dt>
<dd><p>Numeric, controls the size or transparency of all
dots. Default <code>getOption("ligerDotSize")</code> (1) and <code>0.9</code>.</p></dd>

    <dt><code>trimHigh,trimLow</code></dt>
<dd><p>Numeric, limit the largest or smallest value of
continuous <code>colorBy</code> variable. Default <code>NULL</code>.</p></dd>

    <dt><code>raster</code></dt>
<dd><p>Logical, whether to rasterize the plot. Default <code>NULL</code>
automatically rasterize the plot when number of total dots to be plotted
exceeds 100,000.</p></dd>

    <dt><code>legendColorTitle</code></dt>
<dd><p>Legend title text for color aesthetics, often used
for categorical or continuous coloring of dots. Default <code>NULL</code> shows the
original variable name.</p></dd>

    <dt><code>legendShapeTitle</code></dt>
<dd><p>Legend title text for shape aesthetics, often used
for shaping dots by categorical variable. Default <code>NULL</code> shows the
original variable name.</p></dd>

    <dt><code>showLegend</code></dt>
<dd><p>Whether to show the legend. Default <code>TRUE</code>.</p></dd>

    <dt><code>legendPosition</code></dt>
<dd><p>Text indicating where to place the legend. Choose from
<code>"top"</code>, <code>"bottom"</code>, <code>"left"</code> or <code>"right"</code>. Default
<code>"right"</code>.</p></dd>

    <dt><code>baseSize</code></dt>
<dd><p>One-parameter control of all text sizes. Individual text
element sizes can be controlled by other size arguments. "Title" sizes are
2 points larger than "text" sizes when being controlled by this.</p></dd>

    <dt><code>titleSize,xTitleSize,yTitleSize,legendTitleSize</code></dt>
<dd><p>Size of main title,
axis titles and legend title. Default <code>NULL</code> controls by
<code>baseSize + 2</code>.</p></dd>

    <dt><code>subtitleSize,xTextSize,yTextSize,legendTextSize</code></dt>
<dd><p>Size of subtitle text,
axis texts and legend text. Default <code>NULL</code> controls by <code>baseSize</code>.</p></dd>

    <dt><code>panelBorder</code></dt>
<dd><p>Whether to show rectangle border of the panel instead of
using ggplot classic bottom and left axis lines. Default <code>FALSE</code>.</p></dd>

    <dt><code>colorPalette</code></dt>
<dd><p>For continuous coloring, an index or a palette name to
select from available options from ggplot
<code>scale_brewer</code> or <code><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></code>.
Default <code>"magma"</code>.</p></dd>

    <dt><code>colorDirection</code></dt>
<dd><p>Choose <code>1</code> or <code>-1</code>. Applied when
<code>colorPalette</code> is from Viridis options. Default <code>-1</code> use darker
color for higher value, while <code>1</code> reverses this direction.</p></dd>

    <dt><code>naColor</code></dt>
<dd><p>The color code for <code>NA</code> values. Default <code>"#DEDEDE"</code>.
<code><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradient2</a></code>. Default <code>NULL</code>.</p></dd>

  
</dl></dd>

</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="getFactorMarkers.html">getFactorMarkers</a></span><span class="op">(</span><span class="va">pbmcPlot</span>, <span class="st">"ctrl"</span>, <span class="st">"stim"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> Factor 7 did not appear as max in any cell in either dataset</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu">plotGeneLoadings</span><span class="op">(</span><span class="va">pbmcPlot</span>, <span class="va">result</span>, useFactor <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Plotting feature <span style="color: #0000BB;">"Factor_2"</span> on <span style="color: #0000BB;">600</span> cells</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Plotting feature <span style="color: #0000BB;">"Factor_2"</span> on <span style="color: #0000BB;">600</span> cells ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="plotGeneLoadings-1.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

