<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function calculate the proportion of each category (e.g. cluster, cell
type) within each dataset, and then make box plot grouped by condition. The
proportion of all categories within one dataset sums up to 1. The condition
variable must be a variable of dataset, i.e. each dataset must belong to only
one condition."><title>Box plot of cluster proportion in each dataset, grouped by condition — plotProportionBox • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Box plot of cluster proportion in each dataset, grouped by condition — plotProportionBox"><meta property="og:description" content="This function calculate the proportion of each category (e.g. cluster, cell
type) within each dataset, and then make box plot grouped by condition. The
proportion of all categories within one dataset sums up to 1. The condition
variable must be a variable of dataset, i.e. each dataset must belong to only
one condition."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    <a class="dropdown-item" href="../articles/benchmark.html">Benchmarking integration quality</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Box plot of cluster proportion in each dataset, grouped by condition</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/visualization.R" class="external-link"><code>R/visualization.R</code></a></small>
      <div class="d-none name"><code>plotProportionBox.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function calculate the proportion of each category (e.g. cluster, cell
type) within each dataset, and then make box plot grouped by condition. The
proportion of all categories within one dataset sums up to 1. The condition
variable must be a variable of dataset, i.e. each dataset must belong to only
one condition.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotProportionBox</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  useCluster <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  conditionBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sampleBy <span class="op">=</span> <span class="st">"dataset"</span>,</span>
<span>  splitByCluster <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  dot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  dotSize <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerDotSize"</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  dotJitter <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object.</p></dd>


<dt>useCluster</dt>
<dd><p>Name of variable in <code>cellMeta(object)</code>. Default
<code>NULL</code> uses default cluster.</p></dd>


<dt>conditionBy</dt>
<dd><p>Name of the variable in <code>cellMeta(object)</code> that
represents the condition. Must be a high level variable of the
<code>sampleBy</code> variable, i.e. each sample must belong to only one condition.
Default <code>NULL</code> does not group samples by condition.</p></dd>


<dt>sampleBy</dt>
<dd><p>Name of the variable in <code>cellMeta(object)</code> that
represents individual samples. Default <code>"dataset"</code>.</p></dd>


<dt>splitByCluster</dt>
<dd><p>Logical, whether to split the wide grouped box plot by
cluster, into a list of boxplots for each cluster. Default <code>FALSE</code>.</p></dd>


<dt>dot</dt>
<dd><p>Logical, whether to add dot plot on top of the box plot. Default
<code>FALSE</code>.</p></dd>


<dt>dotSize</dt>
<dd><p>Size of the dot. Default uses user option "ligerDotSize", or
<code>1</code> if not set.</p></dd>


<dt>dotJitter</dt>
<dd><p>Logical, whether to jitter the dot to avoid overlapping
within a box when many dots are presented. Default <code>FALSE</code>.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed on to <code><a href="dot-ggplotLigerTheme.html">.ggplotLigerTheme</a></code></p><dl><dt><code>title,subtitle,xlab,ylab</code></dt>
<dd><p>Main title, subtitle or X/Y axis title text.
By default, no main title or subtitle will be set, and X/Y axis title will be
the names of variables used for plotting. Use <code>NULL</code> to hide elements.
<code>TRUE</code> for <code>xlab</code> or <code>ylab</code> shows default values.</p></dd>

    <dt><code>legendFillTitle</code></dt>
<dd><p>Legend title text for fill aesthetics, often used for
violin, box, bar plots. Default <code>NULL</code> shows the original variable name.</p></dd>

    <dt><code>showLegend</code></dt>
<dd><p>Whether to show the legend. Default <code>TRUE</code>.</p></dd>

    <dt><code>legendPosition</code></dt>
<dd><p>Text indicating where to place the legend. Choose from
<code>"top"</code>, <code>"bottom"</code>, <code>"left"</code> or <code>"right"</code>. Default
<code>"right"</code>.</p></dd>

    <dt><code>baseSize</code></dt>
<dd><p>One-parameter control of all text sizes. Individual text
element sizes can be controlled by other size arguments. "Title" sizes are
2 points larger than "text" sizes when being controlled by this.</p></dd>

    <dt><code>titleSize,xTitleSize,yTitleSize,legendTitleSize</code></dt>
<dd><p>Size of main title,
axis titles and legend title. Default <code>NULL</code> controls by
<code>baseSize + 2</code>.</p></dd>

    <dt><code>subtitleSize,xTextSize,yTextSize,legendTextSize</code></dt>
<dd><p>Size of subtitle text,
axis texts and legend text. Default <code>NULL</code> controls by <code>baseSize</code>.</p></dd>

    <dt><code>panelBorder</code></dt>
<dd><p>Whether to show rectangle border of the panel instead of
using ggplot classic bottom and left axis lines. Default <code>FALSE</code>.</p></dd>

    <dt><code>colorLabels</code></dt>
<dd><p>Character vector for modifying category names in a
color legend. Passed to <code>ggplot2::scale_color_manual(labels)</code>. Default
<code>NULL</code> uses original levels of the factor.</p></dd>

    <dt><code>colorValues</code></dt>
<dd><p>Character vector of colors for modifying category colors
in a color legend. Passed to <code>ggplot2::scale_color_manual(values)</code>.
Default <code>NULL</code> uses internal selected palette when &lt;= 26 categories are
presented, otherwise ggplot hues.</p></dd>

    <dt><code>legendNRow,legendNCol</code></dt>
<dd><p>Integer, when too many categories in one
variable, arranges number of rows or columns. Default <code>NULL</code>,
automatically split to <code>ceiling(levels(variable)/10)</code> columns.</p></dd>

    <dt><code>colorPalette</code></dt>
<dd><p>For continuous coloring, an index or a palette name to
select from available options from ggplot
<code>scale_brewer</code> or <code><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></code>.
Default <code>"magma"</code>.</p></dd>

    <dt><code>colorDirection</code></dt>
<dd><p>Choose <code>1</code> or <code>-1</code>. Applied when
<code>colorPalette</code> is from Viridis options. Default <code>-1</code> use darker
color for higher value, while <code>1</code> reverses this direction.</p></dd>

    <dt><code>colorLow,colorMid,colorHigh,colorMidPoint</code></dt>
<dd><p>All four of these must be
specified to customize palette with</p></dd>

    <dt><code>naColor</code></dt>
<dd><p>The color code for <code>NA</code> values. Default <code>"#DEDEDE"</code>.
<code><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradient2</a></code>. Default <code>NULL</code>.</p></dd>

    <dt><code>plotly</code></dt>
<dd><p>Whether to use plotly to enable web based interactive browsing
for the plot. Requires installation of package "plotly". Default
<code>FALSE</code>.</p></dd>

  
</dl></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A ggplot object or a list of ggplot objects if
<code>splitByCluster = TRUE</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># "boxes" are expected to appear as horizontal lines, because there's no</span></span></span>
<span class="r-in"><span><span class="co"># "condition" variable that groups the datasets in the example object, and</span></span></span>
<span class="r-in"><span><span class="co"># thus only one value exists for each "box".</span></span></span>
<span class="r-in"><span><span class="fu">plotProportionBox</span><span class="op">(</span><span class="va">pbmcPlot</span>, conditionBy <span class="op">=</span> <span class="st">"dataset"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotProportionBox-1.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

