<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Generate scatter plot(s) using liger object — plotCellScatter • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Generate scatter plot(s) using liger object — plotCellScatter"><meta property="og:description" content="This function allows for using available cell metadata to build
the x-/y-axis. Available per-cell data can be used to form the color/shape
annotation, including cell metadata, raw or processed gene expression, and
unnormalized or aligned factor loading. Multiple coloring variable is allowed
from the same specification of slot, and this returns a list of plots
with different coloring values. Users can further split the plot(s) by
grouping on cells (e.g. datasets)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/import_data.html">Import data</a>
    </li>
    <li>
      <a href="../articles/liger_object.html">Introduction to a liger object</a>
    </li>
    <li>
      <a href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate scatter plot(s) using liger object</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/ggplotting.R" class="external-link"><code>R/ggplotting.R</code></a></small>
    <div class="hidden name"><code>plotCellScatter.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function allows for using available cell metadata to build
the x-/y-axis. Available per-cell data can be used to form the color/shape
annotation, including cell metadata, raw or processed gene expression, and
unnormalized or aligned factor loading. Multiple coloring variable is allowed
from the same specification of <code>slot</code>, and this returns a list of plots
with different coloring values. Users can further split the plot(s) by
grouping on cells (e.g. datasets).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotCellScatter</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">y</span>,</span>
<span>  colorBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  slot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cellMeta"</span>, <span class="st">"rawData"</span>, <span class="st">"normData"</span>, <span class="st">"scaleData"</span>, <span class="st">"H.norm"</span>, <span class="st">"H"</span>, <span class="st">"normPeak"</span>,</span>
<span>    <span class="st">"rawPeak"</span><span class="op">)</span>,</span>
<span>  colorByFunc <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cellIdx <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  splitBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  shapeBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  titles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object</p></dd>


<dt>x, y</dt>
<dd><p>Available variable name in <code>cellMeta</code> slot to look for
the dot coordinates. See details.</p></dd>


<dt>colorBy</dt>
<dd><p>Available variable name in specified <code>slot</code> to look for
color annotation information. See details. Default <code>NULL</code> generates
all-black dots.</p></dd>


<dt>slot</dt>
<dd><p>Choose the slot to find the <code>colorBy</code> variable. See details.
Default <code>"cellMeta"</code>.</p></dd>


<dt>colorByFunc</dt>
<dd><p>Default <code>NULL</code>. A function object that expects a
vector/factor/data.frame retrieved by <code>colorBy</code> as the only input, and
returns an object of the same size, so that the all color "aes" are replaced
by this output. Useful when, for example, users need to scale the gene
expression shown on plot.</p></dd>


<dt>cellIdx</dt>
<dd><p>Character, logical or numeric index that can subscribe cells.
Missing or <code>NULL</code> for all cells.</p></dd>


<dt>splitBy</dt>
<dd><p>Character vector of categorical variable names in
<code>cellMeta</code> slot. Split all cells by groupings on this/these variable(s)
to produce a scatter plot containing only the cells in each group. Default
<code>NULL</code>.</p></dd>


<dt>shapeBy</dt>
<dd><p>Available variable name in <code>cellMeta</code> slot to look for
categorical annotation to be reflected by dot shapes. Default <code>NULL</code>.</p></dd>


<dt>titles</dt>
<dd><p>Title text. A character scalar or a character vector with as
many elements as multiple plots are supposed to be generated. Default
<code>NULL</code>.</p></dd>


<dt>...</dt>
<dd><p>More plot setting arguments. See <code><a href="dot-ggScatter.html">.ggScatter</a></code> and
<code><a href="dot-ggplotLigerTheme.html">.ggplotLigerTheme</a></code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A ggplot object when a single plot is intended. A list of ggplot
objects, when multiple <code>colorBy</code> variables and/or <code>splitBy</code> are
set. When <code>plotly = TRUE</code>, all ggplot objects become plotly (htmlwidget)
objects.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Available option for <code>slot</code> include: <code>"cellMeta"</code>,
<code>"rawData"</code>, <code>"normData"</code>, <code>"scaleData"</code>, <code>"H.norm"</code>
and <code>"H"</code>. When <code>"rawData"</code>, <code>"normData"</code> or
<code>"scaleData"</code>, <code>colorBy</code> has to be a character vector of feature
names. When <code>"H.norm"</code> or <code>"H"</code>, <code>colorBy</code> can be any valid
index to select one factor of interests. Note that character index follows
<code>"Factor_[k]"</code> format, with replacing <code>[k]</code> with an integer.</p>
<p>When <code>"cellMeta"</code>, <code>colorBy</code> has to be an available column name in
the table. Note that, for <code>colorBy</code> as well as <code>x</code>, <code>y</code>,
<code>shapeBy</code> and <code>splitBy</code>, since a matrix object is feasible in
<code>cellMeta</code> table, using a column (e.g. named as <code>"column1"</code> in a
certain matrix (e.g. named as <code>"matrixVar"</code>) should follow the syntax of
<code>"matrixVar.column1"</code>. When the matrix does not have a "colname"
attribute, the subscription goes with <code>"matrixVar.V1"</code>,
<code>"matrixVar.V2"</code> and etc. Use <code>"UMAP.1"</code>, <code>"UMAP.2"</code>,
<code>"TSNE.1"</code> or <code>"TSNE.2"</code> for the 2D embeddings generated with
rliger package. These are based on the nature of <code>as.data.frame</code> method
on a <code><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></code> object.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu">plotCellScatter</span><span class="op">(</span><span class="va">pbmcPlot</span>, x <span class="op">=</span> <span class="st">"UMAP.1"</span>, y <span class="op">=</span> <span class="st">"UMAP.2"</span>,</span></span>
<span class="r-in"><span>                colorBy <span class="op">=</span> <span class="st">"dataset"</span>, slot <span class="op">=</span> <span class="st">"cellMeta"</span>,</span></span>
<span class="r-in"><span>                labelText <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Dec 20 15:50:48 2023 ... Plotting feature: dataset, on 600 cells...</span>
<span class="r-plt img"><img src="plotCellScatter-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu">plotCellScatter</span><span class="op">(</span><span class="va">pbmcPlot</span>, x <span class="op">=</span> <span class="st">"UMAP.1"</span>, y <span class="op">=</span> <span class="st">"UMAP.2"</span>,</span></span>
<span class="r-in"><span>                colorBy <span class="op">=</span> <span class="st">"S100A8"</span>, slot <span class="op">=</span> <span class="st">"normData"</span>,</span></span>
<span class="r-in"><span>                dotOrder <span class="op">=</span> <span class="st">"ascending"</span>, dotSize <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Dec 20 15:50:48 2023 ... Plotting feature: S100A8, on 600 cells...</span>
<span class="r-plt img"><img src="plotCellScatter-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu">plotCellScatter</span><span class="op">(</span><span class="va">pbmcPlot</span>, x <span class="op">=</span> <span class="st">"UMAP.1"</span>, y <span class="op">=</span> <span class="st">"UMAP.2"</span>,</span></span>
<span class="r-in"><span>                colorBy <span class="op">=</span> <span class="fl">2</span>, slot <span class="op">=</span> <span class="st">"H.norm"</span>,</span></span>
<span class="r-in"><span>                dotOrder <span class="op">=</span> <span class="st">"ascending"</span>, dotSize <span class="op">=</span> <span class="fl">2</span>,</span></span>
<span class="r-in"><span>                colorPalette <span class="op">=</span> <span class="st">"viridis"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Dec 20 15:50:48 2023 ... Plotting feature: Factor_19, on 600 cells...</span>
<span class="r-plt img"><img src="plotCellScatter-3.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

