<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Create density plot basing on specified coordinates — plotDensityDimRed • rliger</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Create density plot basing on specified coordinates — plotDensityDimRed"><meta property="og:description" content='This function shows the cell density presented in a 2D
dimensionality reduction coordinates. Density is shown with coloring and
contour lines. A scatter plot of the dimensionality reduction is added as
well. The density plot can be splitted by categorical variables (e.g.
"dataset"), while the scatter plot will always be shown for all cells
in subplots as a reference of the global structure.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/import_data.html">Import data</a>
    </li>
    <li>
      <a href="../articles/liger_object.html">Introduction to a liger object</a>
    </li>
    <li>
      <a href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://welch-lab.github.io/" class="external-link">Welch Lab</a>
</li>
<li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/LabWelch" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create density plot basing on specified coordinates</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/visualization.R" class="external-link"><code>R/visualization.R</code></a></small>
    <div class="hidden name"><code>plotDensityDimRed.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function shows the cell density presented in a 2D
dimensionality reduction coordinates. Density is shown with coloring and
contour lines. A scatter plot of the dimensionality reduction is added as
well. The density plot can be splitted by categorical variables (e.g.
<code>"dataset"</code>), while the scatter plot will always be shown for all cells
in subplots as a reference of the global structure.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotDensityDimRed</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  useDimRed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  splitBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  combinePlot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  minDensity <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  contour <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  contourLineWidth <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  contourBins <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  dot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  dotColor <span class="op">=</span> <span class="st">"grey"</span>,</span>
<span>  dotSize <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>  dotAlpha <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  dotRaster <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legendFillTitle <span class="op">=</span> <span class="st">"Density"</span>,</span>
<span>  colorPalette <span class="op">=</span> <span class="st">"magma"</span>,</span>
<span>  colorDirection <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object</p></dd>


<dt>useDimRed</dt>
<dd><p>Name of the variable storing dimensionality reduction result
in the <code>cellMeta</code> slot. Default uses default dimension reduction.</p></dd>


<dt>splitBy</dt>
<dd><p>Character vector of categorical variable names in
<code>cellMeta</code> slot. Split all cells by groupings on this/these variable(s)
to produce a density plot containing only the cells in each group. Default
<code>NULL</code>.</p></dd>


<dt>combinePlot</dt>
<dd><p>Logical, whether to utilize
<code><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></code> to combine multiple plots into one. Default
<code>TRUE</code> returns combined ggplot. <code>FALSE</code> returns a list of ggplot
or a single ggplot when only one plot is requested.</p></dd>


<dt>minDensity</dt>
<dd><p>A positive number to filter out low density region colored
on plot. Default <code>8</code>. Setting zero will show density on the whole panel.</p></dd>


<dt>contour</dt>
<dd><p>Logical, whether to draw the contour line. Default
<code>TRUE</code>.</p></dd>


<dt>contourLineWidth</dt>
<dd><p>Numeric, the width of the contour line. Default
<code>0.3</code>.</p></dd>


<dt>contourBins</dt>
<dd><p>Number of contour bins. Higher value generates more
contour lines. Default <code>5</code>.</p></dd>


<dt>dot</dt>
<dd><p>Logical, whether to add scatter plot of all cells, even when
density plot is splitted with <code>splitBy</code>. Default <code>TRUE</code>.</p></dd>


<dt>dotColor, dotSize, dotAlpha</dt>
<dd><p>Numeric, controls the appearance of all
dots. Default <code>"grey"</code>, <code>0.6</code> and <code>0.3</code>, respectively.</p></dd>


<dt>dotRaster</dt>
<dd><p>Logical, whether to rasterize the scatter plot. Default
<code>NULL</code> automatically rasterizes the dots when number of total cells to
be plotted exceeds 100,000.</p></dd>


<dt>title</dt>
<dd><p>Text of main title of the plots. Default <code>NULL</code>. Length
of character vector input should match with number of plots generated.</p></dd>


<dt>legendFillTitle</dt>
<dd><p>Text of legend title. Default <code>"Density"</code>.</p></dd>


<dt>colorPalette</dt>
<dd><p>Name of the option for
<code><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></code>. Default <code>"magma"</code>.</p></dd>


<dt>colorDirection</dt>
<dd><p>Color gradient direction for
<code><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></code>. Default <code>-1</code>.</p></dd>


<dt>...</dt>
<dd><p>More theme setting arguments passed to
<code><a href="dot-ggplotLigerTheme.html">.ggplotLigerTheme</a></code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A ggplot object when only one plot is generated, A ggplot object
combined with <code><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></code> when multiple plots and
<code>combinePlot = TRUE</code>. A list of ggplot when multiple plots and
<code>combinePlot = FALSE</code>.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Example dataset has small number of cells, thus cutoff adjusted.</span></span></span>
<span class="r-in"><span><span class="fu">plotDensityDimRed</span><span class="op">(</span><span class="va">pbmcPlot</span>, minDensity <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotDensityDimRed-1.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

