<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Make dot plot of gene expression in cell groups — plotClusterGeneDot • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Make dot plot of gene expression in cell groups — plotClusterGeneDot"><meta property="og:description" content="This function produces dot plots. Each column represent a group
of cells specified by groupBy, each row is a gene specified by
features. The color of dots reflects mean of normalized expression of
specified genes in each cell group and sizes reflects the percentage of cells
expressing each gene in a group. We utilize
ComplexHeatmap
for simplified management of adding annotation and slicing subplots. This was
inspired by the implementation in
scCustomize."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Import data</li>
    <li class="dropdown-header">Introduction to a liger object</li>
    <li class="dropdown-header">Interoperate with Seurat</li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make dot plot of gene expression in cell groups</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/dotplot.R" class="external-link"><code>R/dotplot.R</code></a></small>
    <div class="hidden name"><code>plotClusterGeneDot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function produces dot plots. Each column represent a group
of cells specified by <code>groupBy</code>, each row is a gene specified by
<code>features</code>. The color of dots reflects mean of normalized expression of
specified genes in each cell group and sizes reflects the percentage of cells
expressing each gene in a group. We utilize
<a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/index.html" class="external-link">ComplexHeatmap</a>
for simplified management of adding annotation and slicing subplots. This was
inspired by the implementation in
<a href="https://samuel-marsh.github.io/scCustomize/reference/Clustered_DotPlot.html" class="external-link">scCustomize</a>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotClusterGeneDot</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  groupBy <span class="op">=</span> <span class="st">"leiden_cluster"</span>,</span>
<span>  splitBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  featureScaleFunc <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">10000</span> <span class="op">*</span> <span class="va">x</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>  cellIdx <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legendColorTitle <span class="op">=</span> <span class="st">"Mean\nExpression"</span>,</span>
<span>  legendSizeTitle <span class="op">=</span> <span class="st">"Percent\nExpressed"</span>,</span>
<span>  viridisOption <span class="op">=</span> <span class="st">"magma"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object</p></dd>


<dt>features</dt>
<dd><p>Character vector of gene names <code>object</code>. Alternatively,
a data.frame where the first column is gene names and second column is
grouping variable (e.g. <code>runWilcoxon</code> output, recommended to be row
subset).</p></dd>


<dt>groupBy</dt>
<dd><p>The names of the columns in <code>cellMeta</code> slot storing
categorical variables. Expression data would be aggregated basing on these,
together with <code>splitBy</code>. Default <code>"leiden_cluster"</code>.</p></dd>


<dt>splitBy</dt>
<dd><p>The names of the columns in <code>cellMeta</code> slot storing
categorical variables. Dotplot panel splitting would be based on these.
Default <code>NULL</code>.</p></dd>


<dt>featureScaleFunc</dt>
<dd><p>A function object applied to normalized data for
scaling the value for better visualization. Default <code>function(x)
log2(10000*x + 1)</code></p></dd>


<dt>cellIdx</dt>
<dd><p>Valid cell subscription. See <code><a href="subsetLiger.html">subsetLiger</a></code>.
Default <code>NULL</code> for using all cells.</p></dd>


<dt>legendColorTitle</dt>
<dd><p>Title for colorbar legend. Default
<code>"Mean\nExpression"</code>.</p></dd>


<dt>legendSizeTitle</dt>
<dd><p>Title for size legend. Default
<code>"Percent\nExpressed"</code></p></dd>


<dt>viridisOption</dt>
<dd><p>Name of available viridis palette. See
<code><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></code>. Default <code>"magma"</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show progress information. Mainly when
subsetting data. Default <code>FALSE</code>.</p></dd>


<dt>...</dt>
<dd><p>Additional theme setting arguments passed to
<code><a href="dot-complexHeatmapDotPlot.html">.complexHeatmapDotPlot</a></code> and heatmap setting arguments passed to
<code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></code>. See Details.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p><code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapList.html" class="external-link">HeatmapList</a></code> object.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>For <code>...</code>, please notice that arguments <code>colorMat</code>,
<code>sizeMat</code>, <code>featureAnnDF</code>, <code>cellSplitVar</code>, <code>cellLabels</code>
and <code>viridisOption</code> from <code><a href="dot-complexHeatmapDotPlot.html">.complexHeatmapDotPlot</a></code> are
already occupied by this function internally. A lot of arguments from
<code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></code> have also been occupied: <code>matrix,
name, heatmap_legend_param, rect_gp, col, layer_fun, km, border, border_gp,
column_gap, row_gap, cluster_row_slices, cluster_rows, row_title_gp,
row_names_gp, row_split, row_labels, cluster_column_slices, cluster_columns,
column_split, column_title_gp, column_title, column_labels, column_names_gp,
top_annotation</code>.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pbmcPlot"</span>, package <span class="op">=</span> <span class="st">"rliger2"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Use character vector of genes</span></span></span>
<span class="r-in"><span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="liger-class.html">varFeatures</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu">plotClusterGeneDot</span><span class="op">(</span><span class="va">pbmcPlot</span>, features <span class="op">=</span> <span class="va">features</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotClusterGeneDot-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Use data.frame with grouping information, with more tweak on plot</span></span></span>
<span class="r-in"><span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">features</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">letters</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">plotClusterGeneDot</span><span class="op">(</span><span class="va">pbmcPlot</span>, features <span class="op">=</span> <span class="va">features</span>,</span></span>
<span class="r-in"><span>                   clusterFeature <span class="op">=</span> <span class="cn">TRUE</span>, clusterCell <span class="op">=</span> <span class="cn">TRUE</span>, maxDotSize <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotClusterGeneDot-2.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

