<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Plot Heatmap of Gene Expression or Factor Loading"><title>Plot Heatmap of Gene Expression or Factor Loading — plotGeneHeatmap • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Plot Heatmap of Gene Expression or Factor Loading — plotGeneHeatmap"><meta property="og:description" content="Plot Heatmap of Gene Expression or Factor Loading"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    <a class="dropdown-item" href="../articles/benchmark.html">Benchmarking integration quality</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Plot Heatmap of Gene Expression or Factor Loading</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/heatmap.R" class="external-link"><code>R/heatmap.R</code></a></small>
      <div class="d-none name"><code>plotHeatmap.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Plot Heatmap of Gene Expression or Factor Loading</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotGeneHeatmap</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  cellIdx <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  slot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"normData"</span>, <span class="st">"rawData"</span>, <span class="st">"scaleData"</span>, <span class="st">"scaleUnsharedData"</span><span class="op">)</span>,</span>
<span>  useCellMeta <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cellAnnotation <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  featureAnnotation <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cellSplitBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  featureSplitBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  viridisOption <span class="op">=</span> <span class="st">"C"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotFactorHeatmap</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  factors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cellIdx <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  slot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H.norm"</span>, <span class="st">"H"</span><span class="op">)</span>,</span>
<span>  useCellMeta <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cellAnnotation <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  factorAnnotation <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cellSplitBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  factorSplitBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  trim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.03</span><span class="op">)</span>,</span>
<span>  viridisOption <span class="op">=</span> <span class="st">"D"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object, with data to be plot available.</p></dd>


<dt>features, factors</dt>
<dd><p>Character vector of genes of interests or numeric
index of factor to be involved. <code>features</code> is required, while
<code>factors</code> is by default all the factors (reads object recorded k value
in <code>uns</code> slot).</p></dd>


<dt>cellIdx</dt>
<dd><p>Valid index to subscribe cells to be included. See
<code><a href="subsetLiger.html">subsetLiger</a></code>. Default <code>NULL</code> use all cells.</p></dd>


<dt>slot</dt>
<dd><p>Use the chosen matrix for heatmap. For <code>plotGeneHeatmap</code>,
default <code>"normData"</code>, alternatively <code>"rawData"</code>,
<code>"scaleData"</code> or <code>"scaleUnsharedData"</code>. For
<code>plotFactorHeatmap</code>, default <code>"H.norm"</code>, alternatively <code>"H"</code>.</p></dd>


<dt>useCellMeta</dt>
<dd><p>Character vector of available variable names in
<code>cellMeta</code>, variables will be added as annotation to the heatmap.
Default <code>NULL</code>.</p></dd>


<dt>cellAnnotation</dt>
<dd><p>data.frame object for using external annotation, with
each column a variable and each row is a cell. Row names of this data.frame
will be used for matching cells involved in heatmap. For cells not found in
this data.frame, <code>NA</code>s will be added with warning. Default <code>NULL</code>.</p></dd>


<dt>featureAnnotation, factorAnnotation</dt>
<dd><p>Similar as <code>cellAnnotation</code>,
while each row would be a gene or factor, respectively. Default <code>NULL</code>.</p></dd>


<dt>cellSplitBy</dt>
<dd><p>Character vector of variable names available in annotation
given by <code>useCellMeta</code> and <code>cellAnnotation</code>. This slices the
heatmap by specified variables. Default <code>NULL</code>.</p></dd>


<dt>featureSplitBy, factorSplitBy</dt>
<dd><p>Similar as <code>cellSplitBy</code>. Default
<code>NULL</code></p></dd>


<dt>viridisOption</dt>
<dd><p>See <code>option</code> argument of
<code><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></code>. Default <code>"C"</code> (plasma) for
<code>plotGeneHeatmap</code> and <code>"D"</code> (viridis) for <code>plotFactorHeatmap</code>.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed on to <code><a href="dot-plotHeatmap.html">.plotHeatmap</a></code></p><dl><dt><code>transpose</code></dt>
<dd><p>Logical, whether to "rotate" the heatmap by 90 degrees so
that cell information is displayed by row. Default <code>FALSE</code>.</p></dd>

    <dt><code>showCellLabel,showFeatureLabel</code></dt>
<dd><p>Logical, whether to show cell barcodes,
gene symbols or factor names. Default <code>TRUE</code> for gene/factors but
<code>FALSE</code> for cells.</p></dd>

    <dt><code>showCellLegend,showFeatureLegend</code></dt>
<dd><p>Logical, whether to show cell or
feature legends. Default <code>TRUE</code>. Can be a scalar for overall control
or a vector matching with each given annotation variable.</p></dd>

    <dt><code>cellAnnColList,featureAnnColList</code></dt>
<dd><p>List object, with each element a
named vector of R-interpretable color code. The names of the list elements
are used for matching the annotation variable names. The names of the colors
in the vectors are used for matching the levels of a variable (factor object,
categorical). Default <code>NULL</code> generates ggplot-flavor categorical colors.</p></dd>

    <dt><code>scale</code></dt>
<dd><p>Logical, whether to take z-score to scale and center gene
expression. Applied after <code>dataScaleFunc</code>. Default <code>FALSE</code>.</p></dd>

    <dt><code>baseSize</code></dt>
<dd><p>One-parameter control of all text sizes. Individual text
element sizes can be controlled by other size arguments. "Title" sizes are
2 points larger than "text" sizes when being controlled by this.</p></dd>

    <dt><code>cellTextSize,featureTextSize,legendTextSize</code></dt>
<dd><p>Size of cell barcode
labels, gene/factor labels, or legend values. Default <code>NULL</code>.</p></dd>

    <dt><code>cellTitleSize,featureTitleSize,legendTitleSize</code></dt>
<dd><p>Size of titles of the
cell slices, gene/factor slices, or the legends. Default <code>NULL</code>.</p></dd>

    <dt><code>RColorBrewerOption</code></dt>
<dd><p>When <code>scale = TRUE</code>, heatmap color will be
mapped with <code><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></code>. This is passed to
<code>name</code>. Default <code>"RdBu"</code>.</p></dd>

  
</dl></dd>


<dt>trim</dt>
<dd><p>Numeric vector of two numbers. Higher value limits the maximum
value and lower value limits the minimum value. Default <code>c(0, 0.03)</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p><code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapList-class.html" class="external-link">HeatmapList-class</a></code> object</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="fu">plotGeneHeatmap</span><span class="op">(</span><span class="va">pbmcPlot</span>, <span class="fu"><a href="liger-class.html">varFeatures</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Subsetting dataset: <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Subsetting dataset: <span style="color: #0000BB;">"stim"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Subsetting dataset: <span style="color: #0000BB;">"stim"</span> ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Subsetting dataset: <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Subsetting dataset: <span style="color: #0000BB;">"ctrl"</span> ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="plotHeatmap-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu">plotGeneHeatmap</span><span class="op">(</span><span class="va">pbmcPlot</span>, <span class="fu"><a href="liger-class.html">varFeatures</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                useCellMeta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"leiden_cluster"</span>, <span class="st">"dataset"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                cellSplitBy <span class="op">=</span> <span class="st">"leiden_cluster"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Subsetting dataset: <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Subsetting dataset: <span style="color: #0000BB;">"stim"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Subsetting dataset: <span style="color: #0000BB;">"stim"</span> ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Subsetting dataset: <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Subsetting dataset: <span style="color: #0000BB;">"ctrl"</span> ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="plotHeatmap-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">plotFactorHeatmap</span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotHeatmap-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu">plotFactorHeatmap</span><span class="op">(</span><span class="va">pbmcPlot</span>, cellIdx <span class="op">=</span> <span class="va">pbmcPlot</span><span class="op">$</span><span class="va">leiden_cluster</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,</span></span>
<span class="r-in"><span>                  useCellMeta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"leiden_cluster"</span>, <span class="st">"dataset"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                  cellSplitBy <span class="op">=</span> <span class="st">"leiden_cluster"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotHeatmap-4.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

