<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Quantile Align (Normalize) Factor Loadings — quantileNorm • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Quantile Align (Normalize) Factor Loadings — quantileNorm"><meta property="og:description" content="This process builds a shared factor neighborhood graph to
jointly cluster cells, then quantile normalizes corresponding clusters.
The first step, building the shared factor neighborhood graph, is performed
in SNF(), and produces a graph representation where edge weights between
cells (across all datasets) correspond to their similarity in the shared
factor neighborhood space. An important parameter here is nNeighbors,
the number of neighbors used to build the shared factor space.
Next we perform quantile alignment for each dataset, factor, and cluster (by
stretching/compressing datasets' quantiles to better match those of the
reference dataset)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/import_data.html">Import data</a>
    </li>
    <li>
      <a href="../articles/liger_object.html">Introduction to a liger object</a>
    </li>
    <li>
      <a href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Quantile Align (Normalize) Factor Loadings</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/integration.R" class="external-link"><code>R/integration.R</code></a></small>
    <div class="hidden name"><code>quantileNorm.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This process builds a shared factor neighborhood graph to
jointly cluster cells, then quantile normalizes corresponding clusters.</p>
<p>The first step, building the shared factor neighborhood graph, is performed
in SNF(), and produces a graph representation where edge weights between
cells (across all datasets) correspond to their similarity in the shared
factor neighborhood space. An important parameter here is <code>nNeighbors</code>,
the number of neighbors used to build the shared factor space.</p>
<p>Next we perform quantile alignment for each dataset, factor, and cluster (by
stretching/compressing datasets' quantiles to better match those of the
reference dataset).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">quantileNorm</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  quantiles <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  reference <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minCells <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  nNeighbors <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  useDims <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  maxSample <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  refineKNN <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  clusterName <span class="op">=</span> <span class="st">"H.norm_cluster"</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for liger</span></span>
<span><span class="fu">quantileNorm</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  quantiles <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  reference <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minCells <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  nNeighbors <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  useDims <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  maxSample <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  refineKNN <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  clusterName <span class="op">=</span> <span class="st">"quantileNormCluster"</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for Seurat</span></span>
<span><span class="fu">quantileNorm</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  reduction <span class="op">=</span> <span class="st">"inmf"</span>,</span>
<span>  quantiles <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  reference <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minCells <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  nNeighbors <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  useDims <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  maxSample <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  refineKNN <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  clusterName <span class="op">=</span> <span class="st">"quantileNormCluster"</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> or Seurat object with valid factorization
result available (i.e. <code><a href="runIntegration.html">runIntegration</a></code> performed in advance).</p></dd>


<dt>quantiles</dt>
<dd><p>Number of quantiles to use for quantile normalization.
Default <code>50</code>.</p></dd>


<dt>reference</dt>
<dd><p>Character, numeric or logical selection of one dataset, out
of all available datasets in <code>object</code>, to use as a "reference" for
normalization. Default <code>NULL</code> use the dataset with the largest number of
cells.</p></dd>


<dt>minCells</dt>
<dd><p>Minimum number of cells to consider a cluster shared across
datasets. Default <code>20</code>.</p></dd>


<dt>nNeighbors</dt>
<dd><p>Number of nearest neighbors for within-dataset knn graph.
Default <code>20</code>.</p></dd>


<dt>useDims</dt>
<dd><p>Indices of factors to use for shared nearest factor
determination. Default <code>NULL</code> uses all factors.</p></dd>


<dt>center</dt>
<dd><p>Whether to center the data when scaling factors. Could be
useful for less sparse modalities like methylation data. Default
<code>FALSE</code>.</p></dd>


<dt>maxSample</dt>
<dd><p>Maximum number of cells used for quantile normalization of
each cluster and factor. Default <code>1000</code>.</p></dd>


<dt>eps</dt>
<dd><p>The error bound of the nearest neighbor search. Lower values give
more accurate nearest neighbor graphs but take much longer to compute.
Default <code>0.9</code>.</p></dd>


<dt>refineKNN</dt>
<dd><p>whether to increase robustness of cluster assignments using
KNN graph. Default <code>TRUE</code>.</p></dd>


<dt>clusterName</dt>
<dd><p>Variable name that will store the clustering result
in metadata of a <a href="liger-class.html">liger</a> object or a <code>Seurat</code> object.</p></dd>


<dt>seed</dt>
<dd><p>Random seed to allow reproducible results. Default <code>1</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> which is <code>TRUE</code> if users have not set.</p></dd>


<dt>reduction</dt>
<dd><p>Name of the reduction where LIGER integration result is
stored. Default <code>"inmf"</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Updated input object</p><ul><li><p>liger method
 </p><ul><li><p>Update the <code>H.norm</code> slot for the alignment cell factor
         loading, ready for running graph based community detection
         clustering or dimensionality reduction for visualization.</p></li>
<li><p>Update the <code>cellMata</code> slot with a cluster assignment basing
         on cell factor loading</p></li>
</ul></li>
<li><p>Seurat method
 </p><ul><li><p>Update the <code>reductions</code> slot with a new <code>DimReduc</code>
         object containing the aligned cell factor loading.</p></li>
<li><p>Update the metadata with a cluster assignment basing on cell
         factor loading</p></li>
</ul></li>
</ul></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">quantileNorm</span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

