<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This is a deprecated function. Calling 'quantileNorm' instead."><title>Quantile align (normalize) factor loadings — quantileAlignSNF • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.8/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Quantile align (normalize) factor loadings — quantileAlignSNF"><meta property="og:description" content="This is a deprecated function. Calling 'quantileNorm' instead."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Quantile align (normalize) factor loadings</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/deprecated.R" class="external-link"><code>R/deprecated.R</code></a></small>
      <div class="d-none name"><code>quantileAlignSNF.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This is a deprecated function. Calling 'quantileNorm' instead.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">quantileAlignSNF</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  knn_k <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  k2 <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  prune.thresh <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  ref_dataset <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_cells <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  quantiles <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  nstart <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  resolution <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  dims.use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">object</span><span class="op">@</span><span class="va">H</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>  dist.use <span class="op">=</span> <span class="st">"CR"</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  small.clust.thresh <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  id.number <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  print.mod <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  print.align.summary <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p><code>liger</code> object. Should run optimizeALS before calling.</p></dd>


<dt>knn_k</dt>
<dd><p>Number of nearest neighbors for within-dataset knn graph (default 20).</p></dd>


<dt>k2</dt>
<dd><p>Horizon parameter for shared nearest factor graph. Distances to all but the k2 nearest
neighbors are set to 0 (cuts down on memory usage for very large graphs). (default 500)</p></dd>


<dt>prune.thresh</dt>
<dd><p>Minimum allowed edge weight. Any edges below this are removed (given weight
0) (default 0.2)</p></dd>


<dt>ref_dataset</dt>
<dd><p>Name of dataset to use as a "reference" for normalization. By default,
the dataset with the largest number of cells is used.</p></dd>


<dt>min_cells</dt>
<dd><p>Minimum number of cells to consider a cluster shared across datasets (default 2)</p></dd>


<dt>quantiles</dt>
<dd><p>Number of quantiles to use for quantile normalization (default 50).</p></dd>


<dt>nstart</dt>
<dd><p>Number of times to perform Louvain community detection with different random
starts (default 10).</p></dd>


<dt>resolution</dt>
<dd><p>Controls the number of communities detected. Higher resolution -&gt; more
communities. (default 1)</p></dd>


<dt>dims.use</dt>
<dd><p>Indices of factors to use for shared nearest factor determination (default
1:ncol(H[[1]])).</p></dd>


<dt>dist.use</dt>
<dd><p>Distance metric to use in calculating nearest neighbors (default "CR").</p></dd>


<dt>center</dt>
<dd><p>Centers the data when scaling factors (useful for less sparse modalities like
methylation data). (default FALSE)</p></dd>


<dt>small.clust.thresh</dt>
<dd><p>Extracts small clusters loading highly on single factor with fewer
cells than this before regular alignment (default 0 -- no small cluster extraction).</p></dd>


<dt>id.number</dt>
<dd><p>Number to use for identifying edge file (when running in parallel)
(generates random value by default).</p></dd>


<dt>print.mod</dt>
<dd><p>Print modularity output from clustering algorithm (default FALSE).</p></dd>


<dt>print.align.summary</dt>
<dd><p>Print summary of clusters which did not align normally (default FALSE).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p><code>liger</code> object with H.norm and cluster slots set.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This process builds a shared factor neighborhood graph to jointly cluster cells, then quantile
normalizes corresponding clusters.</p>
<p>The first step, building the shared factor neighborhood graph, is performed in SNF(), and
produces a graph representation where edge weights between cells (across all datasets)
correspond to their similarity in the shared factor neighborhood space. An important parameter
here is knn_k, the number of neighbors used to build the shared factor space (see SNF()). Afterwards,
modularity-based community detection is performed on this graph (Louvain clustering) in order
to identify shared clusters across datasets. The method was first developed by Waltman and van Eck
(2013) and source code is available at http://www.ludowaltman.nl/slm/. The most important parameter
here is resolution, which corresponds to the number of communities detected.</p>
<p>Next we perform quantile alignment for each dataset, factor, and cluster (by
stretching/compressing datasets' quantiles to better match those of the reference dataset). These
aligned factor loadings are combined into a single matrix and returned as H.norm.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># liger object, factorization complete</span></span></span>
<span class="r-in"><span><span class="va">ligerex</span></span></span>
<span class="r-in"><span><span class="co"># do basic quantile alignment</span></span></span>
<span class="r-in"><span><span class="va">ligerex</span> <span class="op">&lt;-</span> <span class="fu">quantileAlignSNF</span><span class="op">(</span><span class="va">ligerex</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># higher resolution for more clusters (note that SNF is conserved)</span></span></span>
<span class="r-in"><span><span class="va">ligerex</span> <span class="op">&lt;-</span> <span class="fu">quantileAlignSNF</span><span class="op">(</span><span class="va">ligerex</span>, resolution <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># change knn_k for more fine-grained local clustering</span></span></span>
<span class="r-in"><span><span class="va">ligerex</span> <span class="op">&lt;-</span> <span class="fu">quantileAlignSNF</span><span class="op">(</span><span class="va">ligerex</span>, knn_k <span class="op">=</span> <span class="fl">15</span>, resolution <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

