<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Load in data from 10X — read10X • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Load in data from 10X — read10X"><meta property="og:description" content='Enables easy loading of sparse data matrices provided by 10X genomics.
read10X works generally for 10X cellranger pipelines including:
CellRanger &amp;lt; 3.0 &amp;amp; &amp;gt;= 3.0 and CellRanger-ARC.
read10XRNA invokes read10X and takes the "Gene Expression" out,
so that the result can directly be used to construct a liger
object. See Examples for demonstration.
read10XATAC works for both cellRanger-ARC and cellRanger-ATAC
pipelines but needs user arguments for correct recognition. Similarly, the
returned value can directly be used for constructing a liger
object.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown-header">Installation</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Introductory Vignettes</li>
    <li class="dropdown-header">Integrating Multiple scRNAseq Data</li>
    <li class="dropdown-header">Integrating On-Disk (HDF5 Based) Data</li>
    <li class="divider">
    <li class="dropdown-header">Data structure and access</li>
    <li class="dropdown-header">Outer container - liger object</li>
    <li class="dropdown-header">Dataset container - ligerDataset object and subclasses</li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Import Data</li>
      </ul></li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Load in data from 10X</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/import.R" class="external-link"><code>R/import.R</code></a></small>
    <div class="hidden name"><code>read10X.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Enables easy loading of sparse data matrices provided by 10X genomics.</p>
<p><code>read10X</code> works generally for 10X cellranger pipelines including:
CellRanger &lt; 3.0 &amp; &gt;= 3.0 and CellRanger-ARC.</p>
<p><code>read10XRNA</code> invokes <code>read10X</code> and takes the "Gene Expression" out,
so that the result can directly be used to construct a <a href="liger-class.html">liger</a>
object. See Examples for demonstration.</p>
<p><code>read10XATAC</code> works for both cellRanger-ARC and cellRanger-ATAC
pipelines but needs user arguments for correct recognition. Similarly, the
returned value can directly be used for constructing a <a href="liger-class.html">liger</a>
object.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">read10X</span><span class="op">(</span></span>
<span>  <span class="va">path</span>,</span>
<span>  sampleNames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useFiltered <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reference <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  geneCol <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  cellCol <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  returnList <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span>,</span>
<span>  sample.dirs <span class="op">=</span> <span class="va">path</span>,</span>
<span>  sample.names <span class="op">=</span> <span class="va">sampleNames</span>,</span>
<span>  use.filtered <span class="op">=</span> <span class="va">useFiltered</span>,</span>
<span>  data.type <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  merge <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  num.cells <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min.umis <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read10XRNA</span><span class="op">(</span></span>
<span>  <span class="va">path</span>,</span>
<span>  sampleNames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useFiltered <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reference <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  returnList <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read10XATAC</span><span class="op">(</span></span>
<span>  <span class="va">path</span>,</span>
<span>  sampleNames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useFiltered <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pipeline <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"atac"</span>, <span class="st">"arc"</span><span class="op">)</span>,</span>
<span>  arcFeatureType <span class="op">=</span> <span class="st">"Peaks"</span>,</span>
<span>  returnList <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  geneCol <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  cellCol <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>path</dt>
<dd><p>[A.] A Directory containing the matrix.mtx, genes.tsv (or
features.tsv), and barcodes.tsv files provided by 10X. A vector, a named
vector, a list or a named list can be given in order to load several data
directories. [B.] The 10X root directory where subdirectories of per-sample
output folders can be found. Sample names will by default take the name of
the vector, list or subfolders.</p></dd>


<dt>sampleNames</dt>
<dd><p>A vector of names to override the detected or set sample
names for what is given to <code>path</code>. Default <code>NULL</code>. If no name
detected at all and multiple samples are given, will name them by numbers.</p></dd>


<dt>useFiltered</dt>
<dd><p>Logical, if <code>path</code> is given as case B, whether to use
the filtered feature barcode matrix instead of raw (unfiltered). Default
<code>TRUE</code>.</p></dd>


<dt>reference</dt>
<dd><p>In case of specifying a CellRanger&lt;3 root folder to
<code>path</code>, import the matrix from the output using which reference. Only
needed when multiple references present. Default <code>NULL</code>.</p></dd>


<dt>geneCol</dt>
<dd><p>Specify which column of genes.tsv or features.tsv to use for
gene names. Default <code>2</code>.</p></dd>


<dt>cellCol</dt>
<dd><p>Specify which column of barcodes.tsv to use for cell names.
Default <code>1</code>.</p></dd>


<dt>returnList</dt>
<dd><p>Logical, whether to still return a structured list instead
of a single matrix object, in the case where only one sample and only one
feature type can be found. Otherwise will always return a list. Default
<code>FALSE</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> which is <code>TRUE</code> if users have not set.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed to <code>read10X</code></p></dd>


<dt>pipeline</dt>
<dd><p>Which cellRanger pipeline type to find the ATAC data. Choose
<code>"atac"</code> to read the peak matrix from cellranger-atac pipeline output
folder(s), or <code>"arc"</code> to split the ATAC feature subset out from the
multiomic cellranger-arc pipeline output folder(s). Default <code>"atac"</code>.</p></dd>


<dt>arcFeatureType</dt>
<dd><p>When <code>pipeline = "arc"</code>, which feature type is
for the ATAC data of interests. Default <code>"Peaks"</code>. Other possible
feature types can be <code>"Chromatin Accessibility"</code>. Error message will
show available options if argument specification cannot be found.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<ul><li><p>When only one sample is given or detected, and only one feature type
 is detected or using CellRanger &lt; 3.0, and <code>returnList = FALSE</code>, a
 sparse matrix object (dgCMatrix class) will be returned.</p></li>
<li><p>When using <code>read10XRNA</code> or <code>read10XATAC</code>, which are modality
 specific, returns a list named by samples, and each element is the
 corresponding sparse matrix object (dgCMatrix class).</p></li>
<li><p><code>read10X</code> generally returns a list named by samples. Each sample
 element will be another list named by feature types even if only one feature
 type is detected (or using CellRanger &lt; 3.0) for data structure consistency.
 The feature type "Gene Expression" always comes as the first type if
 available.</p></li>
</ul></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># For output from CellRanger &lt; 3.0</span></span></span>
<span class="r-in"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="st">'path/to/data/directory'</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir</span><span class="op">)</span> <span class="co"># Should show barcodes.tsv, genes.tsv, and matrix.mtx</span></span></span>
<span class="r-in"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">read10X</span><span class="op">(</span><span class="va">dir</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="co"># Should show dgCMatrix</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># For root directory from CellRanger &lt; 3.0</span></span></span>
<span class="r-in"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="st">'path/to/root'</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.dirs</a></span><span class="op">(</span><span class="va">dir</span><span class="op">)</span> <span class="co"># Should show sample names</span></span></span>
<span class="r-in"><span><span class="va">matList</span> <span class="op">&lt;-</span> <span class="fu">read10X</span><span class="op">(</span><span class="va">dir</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">matList</span><span class="op">)</span> <span class="co"># Should show the sample names</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">matList</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"Gene Expression"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co"># Should show dgCMatrix</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># For output from CellRanger &gt;= 3.0 with multiple data types</span></span></span>
<span class="r-in"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="st">'path/to/data/directory'</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir</span><span class="op">)</span> <span class="co"># Should show barcodes.tsv.gz, features.tsv.gz, and matrix.mtx.gz</span></span></span>
<span class="r-in"><span><span class="va">matList</span> <span class="op">&lt;-</span> <span class="fu">read10X</span><span class="op">(</span><span class="va">dir</span>, sampleNames <span class="op">=</span> <span class="st">"tissue1"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">matList</span><span class="op">)</span> <span class="co"># Shoud show "tissue1"</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">matList</span><span class="op">$</span><span class="va">tissue1</span><span class="op">)</span> <span class="co"># Should show feature types, e.g. "Gene Expression" and etc.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># For root directory from CellRanger &gt;= 3.0 with multiple data types</span></span></span>
<span class="r-in"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="st">'path/to/root'</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.dirs</a></span><span class="op">(</span><span class="va">dir</span><span class="op">)</span> <span class="co"># Should show sample names, e.g. "rep1", "rep2", "rep3"</span></span></span>
<span class="r-in"><span><span class="va">matList</span> <span class="op">&lt;-</span> <span class="fu">read10X</span><span class="op">(</span><span class="va">dir</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">matList</span><span class="op">)</span> <span class="co"># Should show the sample names: "rep1", "rep2", "rep3"</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">matList</span><span class="op">$</span><span class="va">rep1</span><span class="op">)</span> <span class="co"># Should show the avalable feature types for rep1</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># For creating LIGER object from root directory of CellRanger &gt;= 3.0</span></span></span>
<span class="r-in"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="st">'path/to/root'</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.dirs</a></span><span class="op">(</span><span class="va">dir</span><span class="op">)</span> <span class="co"># Should show sample names, e.g. "rep1", "rep2", "rep3"</span></span></span>
<span class="r-in"><span><span class="va">matList</span> <span class="op">&lt;-</span> <span class="fu">read10XRNA</span><span class="op">(</span><span class="va">dir</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">matList</span><span class="op">)</span> <span class="co"># Should show the sample names: "rep1", "rep2", "rep3"</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">matList</span>, <span class="va">class</span><span class="op">)</span> <span class="co"># Should show matrix class all are "dgCMatrix"</span></span></span>
<span class="r-in"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu">createLigerObject</span><span class="op">(</span><span class="va">matList</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

