<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='Enables easy loading of sparse data matrices provided by 10X genomics.
read10X works generally for 10X cellranger pipelines including:
CellRanger &amp;lt; 3.0 &amp;amp; &amp;gt;= 3.0 and CellRanger-ARC.
read10XRNA invokes read10X and takes the "Gene Expression" out,
so that the result can directly be used to construct a liger
object. See Examples for demonstration.
read10XATAC works for both cellRanger-ARC and cellRanger-ATAC
pipelines but needs user arguments for correct recognition. Similarly, the
returned value can directly be used for constructing a liger
object.'><title>Load in data from 10X — read10X • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Load in data from 10X — read10X"><meta property="og:description" content='Enables easy loading of sparse data matrices provided by 10X genomics.
read10X works generally for 10X cellranger pipelines including:
CellRanger &amp;lt; 3.0 &amp;amp; &amp;gt;= 3.0 and CellRanger-ARC.
read10XRNA invokes read10X and takes the "Gene Expression" out,
so that the result can directly be used to construct a liger
object. See Examples for demonstration.
read10XATAC works for both cellRanger-ARC and cellRanger-ATAC
pipelines but needs user arguments for correct recognition. Similarly, the
returned value can directly be used for constructing a liger
object.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Load in data from 10X</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/import.R" class="external-link"><code>R/import.R</code></a></small>
      <div class="d-none name"><code>read10X.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Enables easy loading of sparse data matrices provided by 10X genomics.</p>
<p><code>read10X</code> works generally for 10X cellranger pipelines including:
CellRanger &lt; 3.0 &amp; &gt;= 3.0 and CellRanger-ARC.</p>
<p><code>read10XRNA</code> invokes <code>read10X</code> and takes the "Gene Expression" out,
so that the result can directly be used to construct a <a href="liger-class.html">liger</a>
object. See Examples for demonstration.</p>
<p><code>read10XATAC</code> works for both cellRanger-ARC and cellRanger-ATAC
pipelines but needs user arguments for correct recognition. Similarly, the
returned value can directly be used for constructing a <a href="liger-class.html">liger</a>
object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">read10X</span><span class="op">(</span></span>
<span>  <span class="va">path</span>,</span>
<span>  sampleNames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useFiltered <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reference <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  geneCol <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  cellCol <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  returnList <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span>, <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  sample.dirs <span class="op">=</span> <span class="va">path</span>,</span>
<span>  sample.names <span class="op">=</span> <span class="va">sampleNames</span>,</span>
<span>  use.filtered <span class="op">=</span> <span class="va">useFiltered</span>,</span>
<span>  data.type <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  merge <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  num.cells <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min.umis <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read10XRNA</span><span class="op">(</span></span>
<span>  <span class="va">path</span>,</span>
<span>  sampleNames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useFiltered <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reference <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  returnList <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">read10XATAC</span><span class="op">(</span></span>
<span>  <span class="va">path</span>,</span>
<span>  sampleNames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useFiltered <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pipeline <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"atac"</span>, <span class="st">"arc"</span><span class="op">)</span>,</span>
<span>  arcFeatureType <span class="op">=</span> <span class="st">"Peaks"</span>,</span>
<span>  returnList <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  geneCol <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  cellCol <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span>, <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>path</dt>
<dd><p>[A.] A Directory containing the matrix.mtx, genes.tsv (or
features.tsv), and barcodes.tsv files provided by 10X. A vector, a named
vector, a list or a named list can be given in order to load several data
directories. [B.] The 10X root directory where subdirectories of per-sample
output folders can be found. Sample names will by default take the name of
the vector, list or subfolders.</p></dd>


<dt>sampleNames</dt>
<dd><p>A vector of names to override the detected or set sample
names for what is given to <code>path</code>. Default <code>NULL</code>. If no name
detected at all and multiple samples are given, will name them by numbers.</p></dd>


<dt>useFiltered</dt>
<dd><p>Logical, if <code>path</code> is given as case B, whether to use
the filtered feature barcode matrix instead of raw (unfiltered). Default
<code>TRUE</code>.</p></dd>


<dt>reference</dt>
<dd><p>In case of specifying a CellRanger&lt;3 root folder to
<code>path</code>, import the matrix from the output using which reference. Only
needed when multiple references present. Default <code>NULL</code>.</p></dd>


<dt>geneCol</dt>
<dd><p>Specify which column of genes.tsv or features.tsv to use for
gene names. Default <code>2</code>.</p></dd>


<dt>cellCol</dt>
<dd><p>Specify which column of barcodes.tsv to use for cell names.
Default <code>1</code>.</p></dd>


<dt>returnList</dt>
<dd><p>Logical, whether to still return a structured list instead
of a single matrix object, in the case where only one sample and only one
feature type can be found. Otherwise will always return a list. Default
<code>FALSE</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> or <code>TRUE</code> if users have not set.</p></dd>


<dt>sample.dirs, sample.names, use.filtered</dt>
<dd><p>These arguments are renamed and
will be deprecated in the future. Please see usage for corresponding
arguments.</p></dd>


<dt>data.type, merge, num.cells, min.umis</dt>
<dd><p>These arguments are defuncted
because the functionality can/should be fulfilled with other functions.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed to <code>read10X</code></p></dd>


<dt>pipeline</dt>
<dd><p>Which cellRanger pipeline type to find the ATAC data. Choose
<code>"atac"</code> to read the peak matrix from cellranger-atac pipeline output
folder(s), or <code>"arc"</code> to split the ATAC feature subset out from the
multiomic cellranger-arc pipeline output folder(s). Default <code>"atac"</code>.</p></dd>


<dt>arcFeatureType</dt>
<dd><p>When <code>pipeline = "arc"</code>, which feature type is
for the ATAC data of interests. Default <code>"Peaks"</code>. Other possible
feature types can be <code>"Chromatin Accessibility"</code>. Error message will
show available options if argument specification cannot be found.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<ul><li><p>When only one sample is given or detected, and only one feature type
 is detected or using CellRanger &lt; 3.0, and <code>returnList = FALSE</code>, a
 sparse matrix object (dgCMatrix class) will be returned.</p></li>
<li><p>When using <code>read10XRNA</code> or <code>read10XATAC</code>, which are modality
 specific, returns a list named by samples, and each element is the
 corresponding sparse matrix object (dgCMatrix class).</p></li>
<li><p><code>read10X</code> generally returns a list named by samples. Each sample
 element will be another list named by feature types even if only one feature
 type is detected (or using CellRanger &lt; 3.0) for data structure consistency.
 The feature type "Gene Expression" always comes as the first type if
 available.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># For output from CellRanger &lt; 3.0</span></span></span>
<span class="r-in"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="st">'path/to/data/directory'</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir</span><span class="op">)</span> <span class="co"># Should show barcodes.tsv, genes.tsv, and matrix.mtx</span></span></span>
<span class="r-in"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">read10X</span><span class="op">(</span><span class="va">dir</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="co"># Should show dgCMatrix</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># For root directory from CellRanger &lt; 3.0</span></span></span>
<span class="r-in"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="st">'path/to/root'</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.dirs</a></span><span class="op">(</span><span class="va">dir</span><span class="op">)</span> <span class="co"># Should show sample names</span></span></span>
<span class="r-in"><span><span class="va">matList</span> <span class="op">&lt;-</span> <span class="fu">read10X</span><span class="op">(</span><span class="va">dir</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">matList</span><span class="op">)</span> <span class="co"># Should show the sample names</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">matList</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"Gene Expression"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co"># Should show dgCMatrix</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># For output from CellRanger &gt;= 3.0 with multiple data types</span></span></span>
<span class="r-in"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="st">'path/to/data/directory'</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir</span><span class="op">)</span> <span class="co"># Should show barcodes.tsv.gz, features.tsv.gz, and matrix.mtx.gz</span></span></span>
<span class="r-in"><span><span class="va">matList</span> <span class="op">&lt;-</span> <span class="fu">read10X</span><span class="op">(</span><span class="va">dir</span>, sampleNames <span class="op">=</span> <span class="st">"tissue1"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">matList</span><span class="op">)</span> <span class="co"># Shoud show "tissue1"</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">matList</span><span class="op">$</span><span class="va">tissue1</span><span class="op">)</span> <span class="co"># Should show feature types, e.g. "Gene Expression" and etc.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># For root directory from CellRanger &gt;= 3.0 with multiple data types</span></span></span>
<span class="r-in"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="st">'path/to/root'</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.dirs</a></span><span class="op">(</span><span class="va">dir</span><span class="op">)</span> <span class="co"># Should show sample names, e.g. "rep1", "rep2", "rep3"</span></span></span>
<span class="r-in"><span><span class="va">matList</span> <span class="op">&lt;-</span> <span class="fu">read10X</span><span class="op">(</span><span class="va">dir</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">matList</span><span class="op">)</span> <span class="co"># Should show the sample names: "rep1", "rep2", "rep3"</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">matList</span><span class="op">$</span><span class="va">rep1</span><span class="op">)</span> <span class="co"># Should show the avalable feature types for rep1</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># For creating LIGER object from root directory of CellRanger &gt;= 3.0</span></span></span>
<span class="r-in"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="st">'path/to/root'</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.dirs</a></span><span class="op">(</span><span class="va">dir</span><span class="op">)</span> <span class="co"># Should show sample names, e.g. "rep1", "rep2", "rep3"</span></span></span>
<span class="r-in"><span><span class="va">matList</span> <span class="op">&lt;-</span> <span class="fu">read10XRNA</span><span class="op">(</span><span class="va">dir</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">matList</span><span class="op">)</span> <span class="co"># Should show the sample names: "rep1", "rep2", "rep3"</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">matList</span>, <span class="va">class</span><span class="op">)</span> <span class="co"># Should show matrix class all are "dgCMatrix"</span></span></span>
<span class="r-in"><span><span class="va">lig</span> <span class="op">&lt;-</span> <span class="fu">createLigerObject</span><span class="op">(</span><span class="va">matList</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

