<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="After quantile normalization, users can additionally run the Leiden or
Louvain algorithm for community detection, which is widely used in
single-cell analysis and excels at merging small clusters into broad cell
classes.
While using quantile normalized factor loadings (result from
quantileNorm) is recommended, this function looks for
unnormalized factor loadings (result from runIntegration) when
the former is not available."><title>SNN Graph Based Community Detection — runCluster • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="SNN Graph Based Community Detection — runCluster"><meta property="og:description" content="After quantile normalization, users can additionally run the Leiden or
Louvain algorithm for community detection, which is widely used in
single-cell analysis and excels at merging small clusters into broad cell
classes.
While using quantile normalized factor loadings (result from
quantileNorm) is recommended, this function looks for
unnormalized factor loadings (result from runIntegration) when
the former is not available."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>SNN Graph Based Community Detection</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/clustering.R" class="external-link"><code>R/clustering.R</code></a></small>
      <div class="d-none name"><code>runCluster.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>After quantile normalization, users can additionally run the Leiden or
Louvain algorithm for community detection, which is widely used in
single-cell analysis and excels at merging small clusters into broad cell
classes.</p>
<p>While using quantile normalized factor loadings (result from
<code><a href="quantileNorm.html">quantileNorm</a></code>) is recommended, this function looks for
unnormalized factor loadings (result from <code><a href="runIntegration.html">runIntegration</a></code>) when
the former is not available.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runCluster</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  resolution <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  nNeighbors <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  prune <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">15</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  nRandomStarts <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  nIterations <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"leiden"</span>, <span class="st">"louvain"</span><span class="op">)</span>,</span>
<span>  useRaw <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useDims <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  groupSingletons <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  saveSNN <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  clusterName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">method</span>, <span class="st">"_cluster"</span><span class="op">)</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span>, <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object. Should have valid factorization
result available.</p></dd>


<dt>resolution</dt>
<dd><p>Numeric, value of the resolution parameter, a larger value
results in a larger number of communities with smaller sizes. Default
<code>1.0</code>.</p></dd>


<dt>nNeighbors</dt>
<dd><p>Integer, the maximum number of nearest neighbors to
compute. Default <code>20</code>.</p></dd>


<dt>prune</dt>
<dd><p>Numeric. Sets the cutoff for acceptable Jaccard index when
computing the neighborhood overlap for the SNN construction. Any edges with
values less than or equal to this will be set to 0 and removed from the SNN
graph. Essentially sets the stringency of pruning. <code>0</code> for no pruning,
while <code>1</code> prunes everything. Default <code>1/15</code>.</p></dd>


<dt>eps</dt>
<dd><p>Numeric, the error bound of the nearest neighbor search. Default
<code>0.1</code>.</p></dd>


<dt>nRandomStarts</dt>
<dd><p>Integer number of random starts. Will pick the
membership with highest quality to return. Default <code>10</code>.</p></dd>


<dt>nIterations</dt>
<dd><p>Integer, maximal number of iterations per random start.
Default <code>5</code>.</p></dd>


<dt>method</dt>
<dd><p>Community detection algorithm to use. Choose from
<code>"leiden"</code> or <code>"louvain"</code>. Default <code>"leiden"</code>.</p></dd>


<dt>useRaw</dt>
<dd><p>Whether to use un-aligned cell factor loadings (\(H\)
matrices). Default <code>NULL</code> search for quantile-normalized loadings first
and un-aligned loadings then.</p></dd>


<dt>useDims</dt>
<dd><p>Indices of factors to use for clustering. Default <code>NULL</code>
uses all available factors.</p></dd>


<dt>groupSingletons</dt>
<dd><p>Whether to group single cells that make up their own
cluster in with the cluster they are most connected to. Default <code>TRUE</code>,
if <code>FALSE</code>, assign all singletons to a <code>"singleton"</code> group.</p></dd>


<dt>saveSNN</dt>
<dd><p>Logical, whether to store the SNN graph, as a dgCMatrix
object, in the object. Default <code>FALSE</code>.</p></dd>


<dt>clusterName</dt>
<dd><p>Name of the variable that will store the clustering result
in <code>cellMeta</code> slot of <code>object</code>. Default <code>"leiden_cluster"</code> and
<code>"louvain_cluster"</code>.</p></dd>


<dt>seed</dt>
<dd><p>Seed of the random number generator. Default <code>1</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> or <code>TRUE</code> if users have not set.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p><code>object</code> with cluster assignment updated in <code>clusterName</code></p>


<p>variable in <code>cellMeta</code> slot. Can be fetched with
<code>object[[clusterName]]</code>. If <code>saveSNN = TRUE</code>, the SNN graph will
be stored at <code>object@uns$snn</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">pbmcPlot</span> <span class="op">&lt;-</span> <span class="fu">runCluster</span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> leiden clustering on quantile normalized cell factor loadings</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> leiden clustering on quantile normalized cell factor loadings ... Found 8 clu…</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> <span style="color: #00BB00;">cellMeta</span> variable <span style="color: #0000BB;">"leiden_cluster"</span> is now set as default.</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">$</span><span class="va">leiden_cluster</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AAACATACCTCGCT.1 ctrl_AAACGGCTCTTCGC.1 ctrl_AACACTCTAAGTAG.1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     0                     0                     0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AACCGCCTCAGGAG.1 ctrl_AACGTTCTTCCGTC.1 ctrl_AAGAACGAAACGAA.1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     3                     1                     0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Levels: 0 1 2 3 4 5 6 7</span>
<span class="r-in"><span><span class="va">pbmcPlot</span> <span class="op">&lt;-</span> <span class="fu">runCluster</span><span class="op">(</span><span class="va">pbmcPlot</span>, method <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> louvain clustering on quantile normalized cell factor loadings</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reading input file...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of nodes: 600</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of edges: 17528</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running Louvain algorithm...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Maximum modularity in 10 random starts: 0.7580</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of communities: 8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Elapsed time: 0 seconds</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> louvain clustering on quantile normalized cell factor loadings ... Found 8 cl…</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> <span style="color: #00BB00;">cellMeta</span> variable <span style="color: #0000BB;">"louvain_cluster"</span> is now set as default.</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">$</span><span class="va">louvain_cluster</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AAACATACCTCGCT.1 ctrl_AAACGGCTCTTCGC.1 ctrl_AACACTCTAAGTAG.1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     0                     0                     0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AACCGCCTCAGGAG.1 ctrl_AACGTTCTTCCGTC.1 ctrl_AAGAACGAAACGAA.1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     3                     1                     0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Levels: 0 1 2 3 4 5 6 7</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

