<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>SNN Graph Based Community Detection — runCluster • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="SNN Graph Based Community Detection — runCluster"><meta property="og:description" content="After quantile normalization, users can additionally run the
Leiden or Louvain algorithm for community detection, which is widely used in
single-cell analysis and excels at merging small clusters into broad cell
classes.
While using quantile normalized factor loadings (result from
quantileNorm) is recommended, this function looks for
unnormalized factor loadings (result from optimizeALS or
online_iNMF) when the former is not available."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/import_data.html">Import data</a>
    </li>
    <li>
      <a href="../articles/liger_object.html">Introduction to a liger object</a>
    </li>
    <li>
      <a href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>SNN Graph Based Community Detection</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/clustering.R" class="external-link"><code>R/clustering.R</code></a></small>
    <div class="hidden name"><code>runCluster.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>After quantile normalization, users can additionally run the
Leiden or Louvain algorithm for community detection, which is widely used in
single-cell analysis and excels at merging small clusters into broad cell
classes.</p>
<p>While using quantile normalized factor loadings (result from
<code><a href="quantileNorm.html">quantileNorm</a></code>) is recommended, this function looks for
unnormalized factor loadings (result from <code><a href="rliger2-deprecated.html">optimizeALS</a></code> or
<code><a href="rliger2-deprecated.html">online_iNMF</a></code>) when the former is not available.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runCluster</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  resolution <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  nNeighbors <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  prune <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">15</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  nRandomStarts <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  nIterations <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"leiden"</span>, <span class="st">"louvain"</span><span class="op">)</span>,</span>
<span>  useDims <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  groupSingletons <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  clusterName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">method</span>, <span class="st">"_cluster"</span><span class="op">)</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object. Should have valid factorization
result available.</p></dd>


<dt>resolution</dt>
<dd><p>Numeric, value of the resolution parameter, a larger value
results in a larger number of communities with smaller sizes. Default
<code>1.0</code>.</p></dd>


<dt>nNeighbors</dt>
<dd><p>Integer, the maximum number of nearest neighbors to
compute. Default <code>20</code>.</p></dd>


<dt>prune</dt>
<dd><p>Numeric. Sets the cutoff for acceptable Jaccard index when
computing the neighborhood overlap for the SNN construction. Any edges with
values less than or equal to this will be set to 0 and removed from the SNN
graph. Essentially sets the stringency of pruning. <code>0</code> for no pruning,
while <code>1</code> prunes everything. Default <code>1/15</code>.</p></dd>


<dt>eps</dt>
<dd><p>Numeric, the error bound of the nearest neighbor search. Default
<code>0.1</code>.</p></dd>


<dt>nRandomStarts</dt>
<dd><p>Integer number of random starts. Will pick the
membership with highest quality to return. Default <code>10</code>.</p></dd>


<dt>nIterations</dt>
<dd><p>Integer, maximal number of iterations per random start.
Default <code>5</code>.</p></dd>


<dt>useDims</dt>
<dd><p>Indices of factors to use for clustering. Default <code>NULL</code>
uses all available factors.</p></dd>


<dt>groupSingletons</dt>
<dd><p>Whether to group single cells that make up their own
cluster in with the cluster they are most connected to. Default <code>TRUE</code>,
if <code>FALSE</code>, assign all singletons to a <code>"singleton"</code> group.</p></dd>


<dt>clusterName</dt>
<dd><p>Name of the variable that will store the clustering result
in <code>cellMeta</code> slot of <code>object</code>. Default <code>"leiden_cluster"</code> and
<code>"louvain_cluster"</code>.</p></dd>


<dt>seed</dt>
<dd><p>Seed of the random number generator. Default <code>1</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> which is <code>TRUE</code> if users have not set.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p><code>object</code> with refined cluster assignment updated in
<code>clusterName</code> variable in <code>cellMeta</code> slot. Can be fetched
with <code>object[[clusterName]]</code></p>


    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">pbmcPlot</span> <span class="op">&lt;-</span> <span class="fu">runCluster</span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Dec 20 15:51:03 2023 ... leiden clustering on quantile normalized cell factor loadings...</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">$</span><span class="va">leiden_cluster</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AAACATACCTCGCT.1 ctrl_AAACGGCTCTTCGC.1 ctrl_AACACTCTAAGTAG.1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     4                     3                     7 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AACCGCCTCAGGAG.1 ctrl_AACGTTCTTCCGTC.1 ctrl_AAGAACGAAACGAA.1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     5                     2                     3 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Levels: 0 1 2 3 4 5 6 7 8 9 10</span>
<span class="r-in"><span><span class="va">pbmcPlot</span> <span class="op">&lt;-</span> <span class="fu">runCluster</span><span class="op">(</span><span class="va">pbmcPlot</span>, method <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Dec 20 15:51:04 2023 ... louvain clustering on quantile normalized cell factor loadings...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reading input file...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of nodes: 600</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of edges: 11200</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running Louvain algorithm...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Maximum modularity in 10 random starts: 0.8474</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of communities: 11</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Elapsed time: 0 seconds</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">$</span><span class="va">louvain_cluster</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AAACATACCTCGCT.1 ctrl_AAACGGCTCTTCGC.1 ctrl_AACACTCTAAGTAG.1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     4                     3                     6 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ctrl_AACCGCCTCAGGAG.1 ctrl_AACGTTCTTCCGTC.1 ctrl_AAGAACGAAACGAA.1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     5                     2                     3 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Levels: 0 1 2 3 4 5 6 7 8 9 10</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

