<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Identify the biological pathways (gene sets from Reactome) that
each metagene (factor) might belongs to."><title>Analyze biological interpretations of metagene — runGSEA • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.8/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Analyze biological interpretations of metagene — runGSEA"><meta property="og:description" content="Identify the biological pathways (gene sets from Reactome) that
each metagene (factor) might belongs to."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Analyze biological interpretations of metagene</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/GSEA.R" class="external-link"><code>R/GSEA.R</code></a></small>
      <div class="d-none name"><code>runGSEA.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Identify the biological pathways (gene sets from Reactome) that
each metagene (factor) might belongs to.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runGSEA</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  genesets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useW <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  useV <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  customGenesets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene_sets <span class="op">=</span> <span class="va">genesets</span>,</span>
<span>  mat_w <span class="op">=</span> <span class="va">useW</span>,</span>
<span>  mat_v <span class="op">=</span> <span class="va">useV</span>,</span>
<span>  custom_gene_sets <span class="op">=</span> <span class="va">customGenesets</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object with valid factorization result.</p></dd>


<dt>genesets</dt>
<dd><p>Character vector of the Reactome gene sets names to be
tested. Default <code>NULL</code> uses all the gene sets from the Reactome.</p></dd>


<dt>useW</dt>
<dd><p>Logical, whether to use the shared factor loadings (\(W\)).
Default <code>TRUE</code>.</p></dd>


<dt>useV</dt>
<dd><p>A character vector of the names, a numeric or logical
vector of the index of the datasets where the \(V\) matrices will be
included for analysis. Default <code>NULL</code> uses all datasets.</p></dd>


<dt>customGenesets</dt>
<dd><p>A named list of character vectors of entrez gene ids.
Default <code>NULL</code> uses all the gene symbols from the input matrix.</p></dd>


<dt>gene_sets, mat_w, mat_v, custom_gene_sets</dt>
<dd><p><b>Deprecated</b>. See Usage
section for replacement.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A list of matrices with GSEA analysis for each factor</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="fu">runGSEA</span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'select()' returned 1:1 mapping between keys and columns</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'select()' returned 1:many mapping between keys and columns</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'select()' returned 1:1 mapping between keys and columns</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (40.43% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (38.3% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (40.43% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (29.79% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (42.55% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (34.04% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (29.79% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (42.55% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (46.81% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (38.3% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (44.68% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (31.91% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (51.06% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (38.3% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (53.19% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (34.04% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (34.04% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (40.43% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (40.43% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (31.91% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway        pval        padj        ES      NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Viral Infection Pathways 0.001391118 0.005564473 0.6841373 1.848427</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.006055455 0.012110910 0.6324946 1.732690</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  Disease 0.035265547 0.047020730 0.5555992 1.539175</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.402103474 0.402103474 0.3869393 1.060003</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1           12   16 6232, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2           56   18 6232, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3          333   21 6232, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         3784   18 5473, 91....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway         pval         padj         ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       Infectious disease 0.0004224757 0.0008449514  0.7516368  2.048992</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Viral Infection Pathways 0.0002133561 0.0008449514  0.7945198  2.127748</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  Disease 0.0014753926 0.0019671901  0.6878732  1.890748</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.4189944134 0.4189944134 -0.2442762 -1.005201</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            3   18 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2            1   16 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3           13   21 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4          224   18 929, 362....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway      pval      padj        ES      NES nMoreExtreme</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                  Disease 0.1021156 0.2902135 0.5033034 1.363403          974</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.1451067 0.2902135 0.4849868 1.296973         1372</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Viral Infection Pathways 0.2918218 0.3890957 0.4347337 1.148921         2743</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.4154513 0.4154513 0.3933589 1.051938         3930</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   21 3553, 92....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   18 3553, 22....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   16 6202, 90....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   18 3553, 92....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway      pval      padj        ES       NES nMoreExtreme</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       Infectious disease 0.1959253 0.3918505 0.4457936 1.2418414         1855</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Viral Infection Pathways 0.1652452 0.3918505 0.4642226 1.2730006         1549</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  Disease 0.5654097 0.7538796 0.3355193 0.9424895         5367</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.9053098 0.9053098 0.2307424 0.6427760         8575</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   18 9636, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   16 9636, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   21 9636, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   18 91543, 9....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway       pval      padj        ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       Infectious disease 0.06196468 0.1239294 0.5286897 1.4474224</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Viral Infection Pathways 0.05807003 0.1239294 0.5352879 1.4438988</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  Disease 0.26604925 0.3547323 0.4263150 1.1795495</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.90440943 0.9044094 0.2309144 0.6321869</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1          585   18 3553, 96....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2          543   16 9636, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3         2527   21 3553, 96....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         8552   18 972, 355....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway       pval       padj        ES      NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            Immune System 0.02084869 0.08339476 0.5831827 1.576855</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.27787722 0.37050297 0.4286276 1.158957</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Viral Infection Pathways 0.24860011 0.37050297 0.4431857 1.186587</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                  Disease 0.40515615 0.40515615 0.3858369 1.055685</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1          197   18 2214, 51....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2         2638   18 2214, 96....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3         2352   16 9636, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         3865   21 2214, 96....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway        pval        padj        ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       Infectious disease 0.001468275 0.005873099 0.6880527 1.8578428</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                  Disease 0.012445095 0.016593460 0.6088180 1.6578106</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Viral Infection Pathways 0.008969083 0.016593460 0.6283282 1.6766993</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.965285789 0.965285789 0.2017394 0.5447259</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1           13   18 2214, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2          118   21 2214, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3           84   16 6228, 90....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         9203   18 10578, 2....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway         pval         padj        ES      NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       Infectious disease 0.0001065757 0.0002146614 0.7910399 2.175216</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Viral Infection Pathways 0.0001073307 0.0002146614 0.8328494 2.265038</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  Disease 0.0004227883 0.0005637177 0.7311131 2.035824</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.8661408931 0.8661408931 0.2444750 0.672262</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            0   18 6133, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2            0   16 6133, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3            3   21 6133, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         8126   18 10578, 6....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway       pval       padj        ES      NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            Immune System 0.01214187 0.04856747 0.6100554 1.659854</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                  Disease 0.37211274 0.51569892 0.3951049 1.085591</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Viral Infection Pathways 0.38677419 0.51569892 0.4006452 1.078623</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       Infectious disease 0.51869209 0.51869209 0.3590636 0.976949</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1          113   18 9636, 91....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2         3511   21 9636, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3         3596   16 9636, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         4869   18 9636, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway        pval       padj         ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Viral Infection Pathways 0.003203075 0.01281230  0.6689367  1.819988</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.009693225 0.01938645  0.6113634  1.680414</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  Disease 0.060532432 0.08070991  0.5244344  1.459576</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.288492707 0.28849271 -0.2735749 -1.155878</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1           29   16 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2           90   18 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3          572   21 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4          177   18 929, 355....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_11</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway      pval      padj        ES      NES nMoreExtreme</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       Infectious disease 0.1345868 0.2691736 0.4849411 1.318384         1266</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Viral Infection Pathways 0.1274247 0.2691736 0.4949309 1.326601         1188</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  Disease 0.4201841 0.5602454 0.3823529 1.052155         3971</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.9417888 0.9417888 0.2086250 0.567178         8865</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   18 2214, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   16 6133, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   21 2214, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   18 2207, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_12</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway       pval       padj        ES      NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                  Disease 0.02467465 0.09869859 0.5886130 1.573397</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.09727301 0.15975875 0.5168881 1.364840</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Viral Infection Pathways 0.11981906 0.15975875 0.5088522 1.330409</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.32305924 0.32305924 0.4253640 1.123171</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1          236   21 6279, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2          930   18 3553, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3         1138   16 6135, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         3091   18 6279, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_13</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway         pval        padj         ES        NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       Infectious disease 0.0005426525 0.001085305  0.7125681  1.9947014</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Viral Infection Pathways 0.0003276540 0.001085305  0.7171567  1.9818528</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  Disease 0.0062205062 0.008294008  0.6230769  1.7643122</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.9898605830 0.989860583 -0.1143908 -0.4973585</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            4   18 4869, 90....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2            2   16 4869, 90....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3           57   21 4869, 90....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4          780   18 929, 513....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway         pval         padj         ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                  Disease 0.0001050641 0.0001415879  0.8417415  2.308089</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.0001055632 0.0001415879  0.8800578  2.382872</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Viral Infection Pathways 0.0001061909 0.0001415879  0.9076006  2.432131</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.2687969925 0.2687969925 -0.2882457 -1.176610</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            0   21 6164, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2            0   18 6164, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3            0   16 6164, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4          142   18 929, 362....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_15</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway         pval         padj        ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Viral Infection Pathways 0.0002189381 0.0008757526 0.7639766 2.1129491</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.0006509004 0.0013018008 0.7089421 1.9856468</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  Disease 0.0050690250 0.0067587000 0.6304771 1.7843747</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.8229550879 0.8229550879 0.2552083 0.7148025</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            1   16 4869, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2            5   18 4869, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3           46   21 4869, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         7585   18   5473, 4869</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_16</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway         pval        padj         ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Viral Infection Pathways 0.0009634943 0.003853977  0.7242736  1.988848</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.0020129251 0.004025850  0.6729600  1.872986</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  Disease 0.0122647494 0.016352999  0.5949352  1.670171</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.3605683837 0.360568384 -0.2577690 -1.064250</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            8   16 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2           18   18 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3          115   21 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4          202   18 929, 513....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_17</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway         pval        padj         ES        NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       Infectious disease 0.0004203889 0.001681555  0.7231558  1.9422854</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Viral Infection Pathways 0.0011650074 0.002330015  0.7091217  1.8841903</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  Disease 0.0035553697 0.004740493  0.6524684  1.7706436</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.5881147541 0.588114754 -0.2288559 -0.9160614</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            3   18 6164, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2           10   16 6164, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3           33   21 6164, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4          286   18 929, 513....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_18</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway       pval      padj        ES      NES nMoreExtreme</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            Immune System 0.08980426 0.3078081 0.5036946 1.400687          834</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.15390407 0.3078081 0.4669226 1.298430         1430</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Viral Infection Pathways 0.33178929 0.4423857 0.4085270 1.124577         3079</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                  Disease 0.45788521 0.4578852 0.3603539 1.017387         4299</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   18 3627, 26....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   18 2214, 96....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   16 9636, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   21 2214, 96....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_19</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway         pval        padj        ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Viral Infection Pathways 0.0004265302 0.001706121 0.7083495 1.9105478</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.0021128248 0.004225650 0.6656027 1.8229746</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  Disease 0.0118784821 0.015837976 0.6023632 1.6642634</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.8160785971 0.816078597 0.2716142 0.7439058</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            3   16 6232, 78....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2           19   18 6232, 78....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3          112   21 6232, 78....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         7724   18 6402, 26....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_20</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway       pval       padj        ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       Infectious disease 0.03930039 0.07860078 0.5537159 1.4995176</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Viral Infection Pathways 0.03593005 0.07860078 0.5629286 1.5040992</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  Disease 0.17402243 0.23202991 0.4622005 1.2645045</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.60056896 0.60056896 0.3386207 0.9170184</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1          372   18 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2          338   16 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3         1659   21 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         5699   18 10578, 5....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

