<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Identify the biological pathways (gene sets from Reactome) that
each metagene (factor) might belongs to."><title>Analyze biological interpretations of metagene — runGSEA • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Analyze biological interpretations of metagene — runGSEA"><meta property="og:description" content="Identify the biological pathways (gene sets from Reactome) that
each metagene (factor) might belongs to."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    <a class="dropdown-item" href="../articles/benchmark.html">Benchmarking integration quality</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Analyze biological interpretations of metagene</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/GSEA.R" class="external-link"><code>R/GSEA.R</code></a></small>
      <div class="d-none name"><code>runGSEA.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Identify the biological pathways (gene sets from Reactome) that
each metagene (factor) might belongs to.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runGSEA</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  genesets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useW <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  useV <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  customGenesets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene_sets <span class="op">=</span> <span class="va">genesets</span>,</span>
<span>  mat_w <span class="op">=</span> <span class="va">useW</span>,</span>
<span>  mat_v <span class="op">=</span> <span class="va">useV</span>,</span>
<span>  custom_gene_sets <span class="op">=</span> <span class="va">customGenesets</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object with valid factorization result.</p></dd>


<dt>genesets</dt>
<dd><p>Character vector of the Reactome gene sets names to be
tested. Default <code>NULL</code> uses all the gene sets from the Reactome.</p></dd>


<dt>useW</dt>
<dd><p>Logical, whether to use the shared factor loadings (\(W\)).
Default <code>TRUE</code>.</p></dd>


<dt>useV</dt>
<dd><p>A character vector of the names, a numeric or logical
vector of the index of the datasets where the \(V\) matrices will be
included for analysis. Default <code>NULL</code> uses all datasets.</p></dd>


<dt>customGenesets</dt>
<dd><p>A named list of character vectors of entrez gene ids.
Default <code>NULL</code> uses all the gene symbols from the input matrix.</p></dd>


<dt>gene_sets, mat_w, mat_v, custom_gene_sets</dt>
<dd><p><b>Deprecated</b>. See Usage
section for replacement.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A list of matrices with GSEA analysis for each factor</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"org.Hs.eg.db"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&amp;&amp;</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"reactome.db"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&amp;&amp;</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"fgsea"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&amp;&amp;</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"AnnotationDbi"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>    <span class="fu">runGSEA</span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'select()' returned 1:1 mapping between keys and columns</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'select()' returned 1:many mapping between keys and columns</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'select()' returned 1:1 mapping between keys and columns</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (40.43% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (38.3% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (40.43% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (29.79% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (42.55% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (34.04% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (29.79% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (42.55% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (46.81% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (38.3% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (44.68% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (31.91% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (51.06% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (38.3% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (53.19% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (34.04% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (34.04% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (40.43% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (40.43% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (31.91% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway        pval       padj        ES      NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Viral Infection Pathways 0.002879693 0.01439846 0.6582108 1.796383</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    Developmental Biology 0.012806715 0.02134453 0.6102249 1.639093</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       Infectious disease 0.012114772 0.02134453 0.6068905 1.670470</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                  Disease 0.033611669 0.04201459 0.5555992 1.540023</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            Immune System 0.398703231 0.39870323 0.3869393 1.060113</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1           26   17 6232, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2          118   15 6232, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3          113   19 6232, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4          317   21 6232, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5         3750   18 5473, 91....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway         pval         padj         ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    Developmental Biology 0.0001066553 0.0005292125  0.8125000  2.161949</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Viral Infection Pathways 0.0002116850 0.0005292125  0.7729954  2.087306</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       Infectious disease 0.0004218519 0.0007030865  0.7303780  1.992890</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                  Disease 0.0012647555 0.0015809444  0.6878732  1.886114</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            Immune System 0.3978300181 0.3978300181 -0.2442762 -1.020686</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            0   15 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2            1   17 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3            3   19 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4           11   21 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5          219   18 929, 362....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway       pval      padj        ES      NES nMoreExtreme</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                  Disease 0.09827316 0.2456829 0.5033034 1.368443          938</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.09027705 0.2456829 0.5107349 1.377837          856</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3    Developmental Biology 0.25334189 0.3166774 0.4475847 1.179798         2368</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 Viral Infection Pathways 0.20220199 0.3166774 0.4607527 1.231663         1909</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            Immune System 0.41105719 0.4110572 0.3933589 1.058450         3895</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   21 3553, 92....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   19 3553, 92....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   15 6202, 90....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   17 929, 620....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   18 3553, 92....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway      pval      padj        ES       NES nMoreExtreme</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    Developmental Biology 0.3984009 0.6640014 0.3911585 1.0621787         3736</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.3076028 0.6640014 0.4098445 1.1395900         2908</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Viral Infection Pathways 0.2589248 0.6640014 0.4281271 1.1776923         2436</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                  Disease 0.5667122 0.7083903 0.3355193 0.9399953         5389</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            Immune System 0.9079949 0.9079949 0.2307424 0.6380570         8585</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   15 6189, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   19 9636, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   17 9636, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   21 9636, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   18 91543, 9....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway      pval      padj        ES       NES nMoreExtreme</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    Developmental Biology 0.1080645 0.1820968 0.5086699 1.3580984         1004</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.1078004 0.1820968 0.4947284 1.3542520         1012</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Viral Infection Pathways 0.1092581 0.1820968 0.4991228 1.3502781         1021</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                  Disease 0.2675253 0.3344067 0.4263150 1.1772368         2533</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            Immune System 0.9046351 0.9046351 0.2309144 0.6291364         8489</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   15 6228, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   19 3553, 96....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   17 9636, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   21 3553, 96....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   18 972, 355....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway       pval      padj        ES      NES nMoreExtreme</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            Immune System 0.02033505 0.1016753 0.5831827 1.577232          192</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    Developmental Biology 0.16108809 0.2507099 0.4795479 1.269607         1515</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       Infectious disease 0.20056794 0.2507099 0.4514197 1.226441         1906</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 Viral Infection Pathways 0.18656085 0.2507099 0.4623016 1.240334         1762</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                  Disease 0.40207809 0.4020781 0.3858369 1.057147         3830</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   18 2214, 51....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   15 51316, 6....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   19 2214, 96....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   17 9636, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   21 2214, 96....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway        pval        padj        ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    Developmental Biology 0.002119992 0.005299979 0.6833333 1.8098851</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.001676973 0.005299979 0.6660305 1.8065540</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  Disease 0.009916493 0.016527488 0.6088180 1.6635131</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 Viral Infection Pathways 0.014956815 0.018696019 0.6034146 1.6205789</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            Immune System 0.966235405 0.966235405 0.2017394 0.5445265</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1           19   15 6228, 90....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2           15   19 2214, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3           94   21 2214, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4          141   17 6228, 90....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5         9185   18 10578, 2....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway         pval         padj        ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                  Disease 0.0001056524 0.0001779486 0.7311131 2.0347979</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.0001062022 0.0001779486 0.7707551 2.1320177</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Viral Infection Pathways 0.0001067692 0.0001779486 0.8117175 2.2240811</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4    Developmental Biology 0.0002160994 0.0002701243 0.7432016 2.0051556</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            Immune System 0.8720520756 0.8720520756 0.2444750 0.6734521</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            0   21 6133, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2            0   19 6133, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3            0   17 6133, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            1   15 6133, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5         8171   18 10578, 6....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway       pval       padj        ES      NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            Immune System 0.01203792 0.06018962 0.6100554 1.660667</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    Developmental Biology 0.28440762 0.38382555 0.4366819 1.165137</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  Disease 0.37995979 0.38382555 0.3951049 1.086650</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       Infectious disease 0.38382555 0.38382555 0.3945865 1.080472</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 Viral Infection Pathways 0.27099808 0.38382555 0.4359909 1.180323</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1          112   18 9636, 91....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2         2642   15 51316, 6....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3         3590   21 9636, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         3625   19 9636, 92....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5         2535   17 9636, 92....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway        pval       padj         ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Viral Infection Pathways 0.005518412 0.02759206  0.6401089  1.744890</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    Developmental Biology 0.020137104 0.03356184  0.5955000  1.601290</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       Infectious disease 0.019694576 0.03356184  0.5825854  1.606885</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                  Disease 0.060752575 0.07594072  0.5244344  1.455608</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            Immune System 0.271217712 0.27121771 -0.2735749 -1.170263</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1           51   17 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2          187   15 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3          186   19 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4          577   21 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5          146   18 929, 355....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_11</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway      pval      padj        ES       NES nMoreExtreme</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    Developmental Biology 0.1183673 0.3593037 0.5055147 1.3508828         1101</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.2155822 0.3593037 0.4479718 1.2244411         2030</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Viral Infection Pathways 0.2041231 0.3593037 0.4555556 1.2352720         1910</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                  Disease 0.4193139 0.5241423 0.3823529 1.0511160         3959</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            Immune System 0.9427322 0.9427322 0.2086250 0.5678465         8839</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   15 6133, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   19 2214, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   17 6133, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   21 2214, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   18 2207, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_12</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway       pval       padj        ES      NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                  Disease 0.01976161 0.09880803 0.5886130 1.578153</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    Developmental Biology 0.08275058 0.10343823 0.5305796 1.381736</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       Infectious disease 0.05509671 0.10343823 0.5397122 1.443171</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 Viral Infection Pathways 0.06961361 0.10343823 0.5311018 1.405059</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            Immune System 0.31128812 0.31128812 0.4253640 1.131166</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1          188   21 6279, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2          780   15 6135, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3          526   19 929, 355....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4          662   17 929, 613....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5         2969   18 6279, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_13</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway         pval        padj         ES        NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       Infectious disease 0.0009673259 0.003800630  0.6824759  1.9287409</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Viral Infection Pathways 0.0015202519 0.003800630  0.6849741  1.9137847</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3    Developmental Biology 0.0036307625 0.006051271  0.6645914  1.8295170</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                  Disease 0.0057747834 0.007218479  0.6230769  1.7715767</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            Immune System 0.9948320413 0.994832041 -0.1143908 -0.5001018</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            8   19 4869, 90....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2           13   17 4869, 90....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3           32   15 9045, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4           53   21 4869, 90....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5          769   18 929, 513....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway         pval        padj         ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    Developmental Biology 0.0001064056 0.000133007  0.9062500  2.395471</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                  Disease 0.0001051193 0.000133007  0.8417415  2.295903</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       Infectious disease 0.0001054964 0.000133007  0.8669039  2.346979</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 Viral Infection Pathways 0.0001056747 0.000133007  0.8936170  2.398999</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            Immune System 0.2222222222 0.222222222 -0.2882457 -1.197634</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            0   15 6164, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2            0   21 6164, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3            0   19 6164, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            0   17 6164, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5          117   18 929, 362....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_15</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway         pval         padj        ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Viral Infection Pathways 0.0001087784 0.0005438921 0.7361111 2.0495122</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    Developmental Biology 0.0004392708 0.0010981770 0.7310739 2.0012432</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       Infectious disease 0.0007551241 0.0012585401 0.6823557 1.9212872</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                  Disease 0.0043958400 0.0054948000 0.6304771 1.7860334</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            Immune System 0.8202955237 0.8202955237 0.2552083 0.7145374</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            0   17 4869, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2            3   15 6232, 78....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3            6   19 4869, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4           40   21 4869, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5         7549   18   5473, 4869</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_16</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway         pval        padj         ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Viral Infection Pathways 0.0006393862 0.003196931  0.6986226  1.925168</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    Developmental Biology 0.0040882195 0.006813699  0.6577213  1.784370</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       Infectious disease 0.0030759440 0.006813699  0.6471931  1.804147</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                  Disease 0.0118293198 0.014786650  0.5949352  1.672219</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            Immune System 0.3481989708 0.348198971 -0.2577690 -1.065138</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            5   17 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2           37   15 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3           28   19 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4          111   21 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5          202   18 929, 513....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_17</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway        pval        padj         ES        NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    Developmental Biology 0.000741997 0.002990010  0.7078037  1.8673325</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.001261167 0.002990010  0.7037357  1.9036915</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Viral Infection Pathways 0.001794006 0.002990010  0.6881356  1.8384582</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                  Disease 0.003459119 0.004323899  0.6524684  1.7745925</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            Immune System 0.547169811 0.547169811 -0.2288559 -0.9203805</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            6   15 6164, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2           11   19 6164, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3           16   17 6164, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4           32   21 6164, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5          260   18 929, 513....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_18</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway       pval      padj        ES      NES nMoreExtreme</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            Immune System 0.08560768 0.4280384 0.5036946 1.397848          802</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    Developmental Biology 0.46387299 0.4645994 0.3766108 1.019211         4294</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  Disease 0.46459943 0.4645994 0.3603539 1.014640         4389</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       Infectious disease 0.23678356 0.4645994 0.4316227 1.204928         2234</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 Viral Infection Pathways 0.45952280 0.4645994 0.3700680 1.021795         4313</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   18 3627, 26....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   15 6228, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   21 2214, 96....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   19 2214, 96....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   17 9636, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_19</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway         pval        padj        ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    Developmental Biology 0.0002155172 0.001077586 0.7596713 2.0335976</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Viral Infection Pathways 0.0012802731 0.003200683 0.6868687 1.8746697</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       Infectious disease 0.0037194474 0.006199079 0.6444805 1.7699671</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                  Disease 0.0138492441 0.017311555 0.6023632 1.6657098</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            Immune System 0.8094122658 0.809412266 0.2716142 0.7433329</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            1   15 6232, 78....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2           11   17 6232, 78....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3           34   19 6232, 78....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4          130   21 6232, 78....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5         7601   18 6402, 26....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_20</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway       pval      padj        ES      NES nMoreExtreme</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    Developmental Biology 0.02433581 0.1211962 0.5925050 1.565952          228</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.07271773 0.1211962 0.5234712 1.413193          692</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Viral Infection Pathways 0.06922752 0.1211962 0.5318504 1.423238          655</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                  Disease 0.18054973 0.2256872 0.4622005 1.253422         1720</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            Immune System 0.60986877 0.6098688 0.3386207 0.911259         5808</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   15 51316, 7....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   19 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   17 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   21 7852, 62....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   18 10578, 5....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

