<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Analyze biological interpretations of metagene — runGSEA • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Analyze biological interpretations of metagene — runGSEA"><meta property="og:description" content="Identify the biological pathways (gene sets from Reactome) that
each metagene (factor) might belongs to."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/import_data.html">Import data</a>
    </li>
    <li>
      <a href="../articles/liger_object.html">Introduction to a liger object</a>
    </li>
    <li>
      <a href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Analyze biological interpretations of metagene</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/GSEA.R" class="external-link"><code>R/GSEA.R</code></a></small>
    <div class="hidden name"><code>runGSEA.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Identify the biological pathways (gene sets from Reactome) that
each metagene (factor) might belongs to.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runGSEA</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  genesets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useW <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  useDatasets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  customGenesets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene_sets <span class="op">=</span> <span class="va">genesets</span>,</span>
<span>  mat_w <span class="op">=</span> <span class="va">useW</span>,</span>
<span>  mat_v <span class="op">=</span> <span class="va">useDatasets</span>,</span>
<span>  custom_gene_sets <span class="op">=</span> <span class="va">customGenesets</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object with valid factorization result.</p></dd>


<dt>genesets</dt>
<dd><p>Character vector of the Reactome gene sets names to be
tested. Default <code>NULL</code> uses all the gene sets from the Reactome.</p></dd>


<dt>useW</dt>
<dd><p>Logical, whether to use the shared factor loadings (\(W\)).
Default <code>TRUE</code>.</p></dd>


<dt>useDatasets</dt>
<dd><p>A character vector of the names, a numeric or logical
vector of the index of the datasets where the \(V\) matrices will be
included for analysis. Default <code>NULL</code> uses all datasets.</p></dd>


<dt>customGenesets</dt>
<dd><p>A named list of character vectors of entrez gene ids.
Default <code>NULL</code> uses all the gene symbols from the input matrix.</p></dd>


<dt>gene_sets, mat_w, mat_v, custom_gene_sets</dt>
<dd><p><b>Deprecated</b>. See Usage
section for replacement.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list of matrices with GSEA analysis for each factor</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu">runGSEA</span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'select()' returned 1:1 mapping between keys and columns</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'select()' returned 1:many mapping between keys and columns</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'select()' returned 1:1 mapping between keys and columns</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (34.04% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You are trying to run fgseaSimple. It is recommended to use fgseaMultilevel. To run fgseaMultilevel, you need to remove the nperm argument in the fgsea function call.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There are ties in the preranked stats (40.43% of the list).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All values in the stats vector are greater than zero and scoreType is "std", maybe you should switch to scoreType = "pos".</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway      pval       padj        ES      NES nMoreExtreme</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            Immune System 0.0126327 0.05053079 0.6137536 1.637542          118</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                  Disease 0.2718995 0.27189953 0.4293680 1.172129         2597</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       Infectious disease 0.2074755 0.27189953 0.4570513 1.228547         1964</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 Viral Infection Pathways 0.1865290 0.27189953 0.4719430 1.250740         1752</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   17 91543, 9....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   21 9636, 92....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   18 9636, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   16 9636, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Factor_2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    pathway         pval         padj         ES       NES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                  Disease 0.0001051967 0.0001417384  0.7911480  2.166210</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       Infectious disease 0.0001056078 0.0001417384  0.8289855  2.248346</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Viral Infection Pathways 0.0001063038 0.0001417384  0.8613096  2.309901</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            Immune System 0.1652323580 0.1652323580 -0.3033898 -1.240585</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nMoreExtreme size  leadingEdge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            0   21 6164, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2            0   18 6164, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3            0   16 6164, 61....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4           95   17 929, 362....</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

