<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Perform online iNMF on scaled datasets — runOnlineINMF • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Perform online iNMF on scaled datasets — runOnlineINMF"><meta property="og:description" content="Perform online integrative non-negative matrix factorization to
represent multiple single-cell datasets in terms of \(H\), \(W\), and
\(V\) matrices. It optimizes the iNMF objective function (see
runINMF) using online learning (non-negative least squares for
\(H\) matrices, and hierarchical alternating least squares (HALS) for
\(V\) matrices and \(W\)), where the number of factors is set by
k. The function allows online learning in 3 scenarios:

Fully observed datasets;
Iterative refinement using continually arriving datasets;
Projection of new datasets without updating the existing factorization


All three scenarios require fixed memory independent of the number of cells.
For each dataset, this factorization produces an \(H\) matrix (k by cell),
a \(V\) matrix (genes by k), and a shared \(W\)
matrix (genes by k). The \(H\) matrices represent the cell factor loadings.
\(W\) is identical among all datasets, as it represents the shared
components of the metagenes across datasets. The \(V\) matrices represent
the dataset-specific components of the metagenes."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/import_data.html">Import data</a>
    </li>
    <li>
      <a href="../articles/liger_object.html">Introduction to a liger object</a>
    </li>
    <li>
      <a href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform online iNMF on scaled datasets</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/integration.R" class="external-link"><code>R/integration.R</code></a></small>
    <div class="hidden name"><code>runOnlineINMF.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Perform online integrative non-negative matrix factorization to
represent multiple single-cell datasets in terms of \(H\), \(W\), and
\(V\) matrices. It optimizes the iNMF objective function (see
<code><a href="runINMF.html">runINMF</a></code>) using online learning (non-negative least squares for
\(H\) matrices, and hierarchical alternating least squares (HALS) for
\(V\) matrices and \(W\)), where the number of factors is set by
<code>k</code>. The function allows online learning in 3 scenarios:</p>
<ol><li><p>Fully observed datasets;</p></li>
<li><p>Iterative refinement using continually arriving datasets;</p></li>
<li><p>Projection of new datasets without updating the existing factorization</p></li>
</ol><p>All three scenarios require fixed memory independent of the number of cells.</p>
<p>For each dataset, this factorization produces an \(H\) matrix (k by cell),
a \(V\) matrix (genes by k), and a shared \(W\)
matrix (genes by k). The \(H\) matrices represent the cell factor loadings.
\(W\) is identical among all datasets, as it represents the shared
components of the metagenes across datasets. The \(V\) matrices represent
the dataset-specific components of the metagenes.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runOnlineINMF</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  lambda <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  maxEpochs <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  HALSiter <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  miniBatchSize <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">123</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for liger</span></span>
<span><span class="fu">runOnlineINMF</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  lambda <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  newDatasets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  projection <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  maxEpochs <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  HALSiter <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  miniBatchSize <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  WInit <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  VInit <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  AInit <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  BInit <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for Seurat</span></span>
<span><span class="fu">runOnlineINMF</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  lambda <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  datasetVar <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>  useLayer <span class="op">=</span> <span class="st">"ligerScaleData"</span>,</span>
<span>  assay <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reduction <span class="op">=</span> <span class="st">"onlineINMF"</span>,</span>
<span>  maxEpochs <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  HALSiter <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  miniBatchSize <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p><a href="liger-class.html">liger</a> object. Scaled data required.</p></dd>


<dt>k</dt>
<dd><p>Inner dimension of factorization--number of metagenes. A value in
the range 20-50 works well for most analyses. Default <code>20</code>.</p></dd>


<dt>lambda</dt>
<dd><p>Regularization parameter. Larger values penalize
dataset-specific effects more strongly (i.e. alignment should increase as
lambda increases). We recommend always using the default value except
possibly for analyses with relatively small differences (biological
replicates, male/female comparisons, etc.) in which case a lower value such
as 1.0 may improve reconstruction quality. Default <code>5.0</code>.</p></dd>


<dt>maxEpochs</dt>
<dd><p>The number of epochs to iterate through. See detail.
Default <code>5</code>.</p></dd>


<dt>miniBatchSize</dt>
<dd><p>Total number of cells in each minibatch. See detail.
Default <code>5000</code>.</p></dd>


<dt>seed</dt>
<dd><p>Random seed to allow reproducible results. Default <code>123</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> which is <code>TRUE</code> if users have not set.</p></dd>


<dt>newDatasets</dt>
<dd><p>Named list of <a href="https://rdrr.io/pkg/Matrix/man/dgCMatrix-class.html" class="external-link">dgCMatrix</a>. New datasets for
scenario 2 or scenario 3. Default <code>NULL</code> triggers scenario 1.</p></dd>


<dt>projection</dt>
<dd><p>Whether to perform data integration with scenario 3 when
<code>newDatasets</code> is specified. See description. Default <code>FALSE</code>.</p></dd>


<dt>WInit, VInit, AInit, BInit</dt>
<dd><p>Optional initialization for \(W\), \(V\),
\(A\), and \(B\) matrices, respectively. Must be presented all together.
See detail. Default <code>NULL</code>.</p></dd>


<dt>datasetVar</dt>
<dd><p>Metadata variable name that stores the dataset source
annotation. Default <code>"orig.ident"</code>.</p></dd>


<dt>useLayer</dt>
<dd><p>For Seurat&gt;=4.9.9, the name of layer to retrieve input
non-negative scaled data. Default <code>"ligerScaleData"</code>. For older Seurat,
always retrieve from <code>scale.data</code> slot.</p></dd>


<dt>assay</dt>
<dd><p>Name of assay to use. Default <code>NULL</code> uses current active
assay.</p></dd>


<dt>reduction</dt>
<dd><p>Name of the reduction to store result. Also used as the
feature key. Default <code>"onlineINMF"</code>.</p></dd>


<dt>HALSiters</dt>
<dd><p>Maximum number of block coordinate descent (HALS
algorithm) iterations to perform for each update of \(W\) and \(V\).
Default <code>1</code>. Changing this parameter is not recommended.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<ul><li><p>liger method - Returns updated input <a href="liger-class.html">liger</a> object.
 </p><ul><li><p>A list of all \(H\) matrices can be accessed with
         <code>getMatrix(object, "H")</code></p></li>
<li><p>A list of all \(V\) matrices can be accessed with
         <code>getMatrix(object, "V")</code></p></li>
<li><p>The \(W\) matrix can be accessed with
         <code>getMatrix(object, "W")</code></p></li>
<li><p>Meanwhile, intermediate matrices \(A\) and \(B\) produced in
         HALS update can also be accessed similarly.</p></li>
</ul></li>
<li><p>Seurat method - Returns updated input Seurat object.
 </p><ul><li><p>\(H\) matrices for all datasets will be concatenated and
         transposed (all cells by k), and form a DimReduc object in the
         <code>reductions</code> slot named by argument <code>reduction</code>.</p></li>
<li><p>\(W\) matrix will be presented as <code>feature.loadings</code> in the
         same DimReduc object.</p></li>
<li><p>\(V\) matrices, \(A\) matrices, \(B\) matricesm an objective
         error value and the dataset variable used for the factorization is
         currently stored in <code>misc</code> slot of the same DimReduc object.</p></li>
</ul></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>For performing scenario 2 or 3, a complete set of factorization result from
a run of scenario 1 is required. Given the structure of a <a href="liger-class.html">liger</a>
object, all of the required information can be retrieved automatically.
Under the circumstance where users need customized information for existing
factorization, arguments <code>WInit</code>, <code>VInit</code>, <code>AInit</code> and
<code>BInit</code> are exposed. The requirements for these argument follows:</p><ul><li><p>WInit - A matrix object of size \(m \times k\). (see
     <code><a href="runINMF.html">runINMF</a></code> for notation)</p></li>
<li><p>VInit - A list object of matrices each of size \(m \times k\).
     Number of matrices should match with <code>newDatasets</code>.</p></li>
<li><p>AInit - A list object of matrices each of size \(k \times k\).
     Number of matrices should match with <code>newDatasets</code>.</p></li>
<li><p>BInit - A list object of matrices each of size \(m \times k\).
     Number of matrices should match with <code>newDatasets</code>.</p></li>
</ul><p>Minibatch iterations is performed on small subset of cells. The exact
minibatch size applied on each dataset is <code>miniBatchSize</code> multiplied by
the proportion of cells in this dataset out of all cells. In general,
<code>miniBatchSize</code> should be no larger than the number of cells in the
smallest dataset (considering both <code>object</code> and <code>newDatasets</code>).
Therefore, a smaller value may be necessary for analyzing very small
datasets.</p>
<p>An epoch is one completion of calculation on all cells after a number of
iterations of minibatches. Therefore, the total number of iterations is
determined by the setting of <code>maxEpochs</code>, total number of cells, and
<code>miniBatchSize</code>.</p>
<p>Currently, Seurat S3 method does not support working on Scenario 2 and 3,
because there is no simple solution for organizing a number of miscellaneous
matrices with a single Seurat object. We strongly recommend that users create
a <a href="liger-class.html">liger</a> object which has the specific structure.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Chao Gao and et al., Iterative single-cell multi-omic integration
using online learning, Nat Biotechnol., 2021</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="normalize.html">normalize</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Nov  6 16:34:17 2023 ... Normalizing dataset: ctrl</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Nov  6 16:34:17 2023 ... Normalizing dataset: stim</span>
<span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Nov  6 16:34:17 2023 ... Selecting variable features for dataset: ctrl</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Nov  6 16:34:17 2023 ...... 171 features selected out of 249 shared features</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Nov  6 16:34:17 2023 ... Selecting variable features for dataset: stim</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Nov  6 16:34:17 2023 ...... 249 features selected out of 249 shared features</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Nov  6 16:34:17 2023 ... Finally 171 shared variable features selected.</span>
<span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Nov  6 16:34:17 2023 ... Scaling dataset: ctrl</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Nov  6 16:34:17 2023 ... Scaling dataset: stim</span>
<span class="r-in"><span><span class="co"># Scenario 1</span></span></span>
<span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="rliger2-deprecated.html">online_iNMF</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Minibatch size larger than the smallest dataset involved.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Setting to the smallest dataset size: 300</span>
<span class="r-in"><span><span class="co"># Scenario 2</span></span></span>
<span class="r-in"><span><span class="co"># Fake new dataset by increasing all non-zero value in "ctrl" by 1</span></span></span>
<span class="r-in"><span><span class="va">ctrl2</span> <span class="op">&lt;-</span> <span class="fu"><a href="ligerDataset-class.html">rawData</a></span><span class="op">(</span><span class="fu"><a href="liger-class.html">dataset</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"ctrl"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ctrl2</span><span class="op">@</span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">ctrl2</span><span class="op">@</span><span class="va">x</span> <span class="op">+</span> <span class="fl">1</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">ctrl2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">ctrl2</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pbmc2</span> <span class="op">&lt;-</span> <span class="fu"><a href="rliger2-deprecated.html">online_iNMF</a></span><span class="op">(</span><span class="va">pbmc</span>, k <span class="op">=</span> <span class="fl">20</span>, X_new <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ctrl2 <span class="op">=</span> <span class="va">ctrl2</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filling in NAs to H.norm matrix</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Nov  6 16:34:17 2023 ... Normalizing dataset: ctrl2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Nov  6 16:34:17 2023 ... Scaling dataset: ctrl2</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Minibatch size larger than the smallest dataset involved.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Setting to the smallest dataset size: 300</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in names(res[[i]]) &lt;- names(objectList):</span> 'names' attribute [2] must be the same length as the vector [1]</span>
<span class="r-in"><span><span class="co"># Scenario 3</span></span></span>
<span class="r-in"><span><span class="va">pbmc3</span> <span class="op">&lt;-</span> <span class="fu"><a href="rliger2-deprecated.html">online_iNMF</a></span><span class="op">(</span><span class="va">pbmc</span>, k <span class="op">=</span> <span class="fl">20</span>, X_new <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ctrl2 <span class="op">=</span> <span class="va">ctrl2</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                     projection <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filling in NAs to H.norm matrix</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Nov  6 16:34:17 2023 ... Normalizing dataset: ctrl2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Nov  6 16:34:17 2023 ... Scaling dataset: ctrl2</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Minibatch size larger than the smallest dataset involved.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Setting to the smallest dataset size: 300</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in dimnames(res$V[[i]]) &lt;- list(features, factorNames):</span> attempt to set an attribute on NULL</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

