<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Run DESeq2 differential expression test on pseudo-bulk expression — runPseudoBulkDEG • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Run DESeq2 differential expression test on pseudo-bulk expression — runPseudoBulkDEG"><meta property="og:description" content="This function aggregate raw counts data of specified comparison groups of
cells basing on either given biological replicate annotation or pseudo
replicates. According to user prior knowledge, existing replicate annotation
should be used when it is true that the cells involved in comparison come
from multiple biological replicates, while when all cells in both groups
come from the same replicate, pseudo replicates should be created.
Known replicate annotation can be specified by useCellMetaVar to use
metadata stored in the cellMeta slot, or specify external annotation
through replicateAnn. To use pseudo replicates, users should leave the
two mentioned argument as default NULL."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown-header">Installation</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Import data</li>
    <li class="dropdown-header">Introduction to a liger object</li>
    <li class="dropdown-header">Interoperate with Seurat</li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run DESeq2 differential expression test on pseudo-bulk expression</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/pseudoBulkDEG.R" class="external-link"><code>R/pseudoBulkDEG.R</code></a></small>
    <div class="hidden name"><code>runPseudoBulkDEG.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function aggregate raw counts data of specified comparison groups of
cells basing on either given biological replicate annotation or pseudo
replicates. According to user prior knowledge, existing replicate annotation
should be used when it is true that the cells involved in comparison come
from multiple biological replicates, while when all cells in both groups
come from the same replicate, pseudo replicates should be created.</p>
<p>Known replicate annotation can be specified by <code>useCellMetaVar</code> to use
metadata stored in the <code>cellMeta</code> slot, or specify external annotation
through <code>replicateAnn</code>. To use pseudo replicates, users should leave the
two mentioned argument as default <code>NULL</code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runPseudoBulkDEG</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  groups <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  markerBy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DESeq2"</span>, <span class="st">"limma"</span>, <span class="st">"edgeR"</span><span class="op">)</span>,</span>
<span>  useCellMetaVar <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  replicateAnn <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nRep <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a> object.</p></dd>


<dt>groups</dt>
<dd><p>A named list with at least two elements, where each is a vector
of valid cell specification (numeric or logical indexing should be based on
the whole object). See Details. Default <code>NULL</code>.</p></dd>


<dt>markerBy</dt>
<dd><p><code>cellMeta</code> categorical variable name(s) to group cells.
See Details. Default <code>NULL</code>.</p></dd>


<dt>method</dt>
<dd><p>Choose from <code>"DESeq2"</code>, <code>"limma"</code> or <code>"edgeR"</code>.
Default <code>"DESeq2"</code></p></dd>


<dt>useCellMetaVar</dt>
<dd><p>Name(s) of <code>cellMeta</code> variable that specify the
replicate annotation. Default <code>NULL</code>.</p></dd>


<dt>replicateAnn</dt>
<dd><p>A data.frame or factor to pass external replicate
annotation. See Details for requirements. Default <code>NULL</code>.</p></dd>


<dt>nRep</dt>
<dd><p>Number of pseudo replicates per comparison group, applied only
when running pseudo replicate mode (i.e. both <code>useCellMetaVar</code> and
<code>replicateAnn</code> are <code>NULL</code>). Default <code>5</code>.</p></dd>


<dt>seed</dt>
<dd><p>Random seed to allow reproducible results. Default <code>1</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> which is <code>TRUE</code> if users have not set.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A data.frame as the output of DESeq2. Ordered by adjusted p-value
(<code>padj</code>).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The differential expression test can be run in two mode:</p><ul><li><p>Group wise test: Users need to explicitly specify the cell groups with
a named list through <code>groups</code>, and leave <code>markerBy</code> as <code>NULL</code>.
The DE test will be performed pairwisely between groups.</p></li>
<li><p>Biomarker test: Users need to pass the grouping (e.g. clustering)
variable(s) to <code>markerBy</code>, and leave <code>groups</code> as <code>NULL</code>. The
DE test will be performed by comparing one group against all the rest
iteratively.</p></li>
</ul><p>For using <code>replicateAnn</code>. When passing a data.frame, rownames have to be
available and it should cover all cells involved in both comparison group.
When using a named factor/vector, the names should as well cover all involved
cells. When using a un-named factor/vector, it is recommended to have its
length matched with the total number of cells of the whole object. When the
length of the un-named factor/vector equals to <code>length(groups[[1]])
+ length(groups[[2]])</code>, it is assumed that the order matches to
<code>c(groups[[1]], groups[[2]])</code>.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="ligerDataset-class.html">rawData</a></span><span class="op">(</span><span class="fu"><a href="liger-class.html">datasets</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="ligerDataset-class.html">rawData</a></span><span class="op">(</span><span class="fu"><a href="liger-class.html">dataset</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="ligerDataset-class.html">rawData</a></span><span class="op">(</span><span class="fu"><a href="liger-class.html">datasets</a></span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="ligerDataset-class.html">rawData</a></span><span class="op">(</span><span class="fu"><a href="liger-class.html">dataset</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">runPseudoBulkDEG</span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">pbmcPlot</span>,</span></span>
<span class="r-in"><span>    groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>        c1 <span class="op">=</span> <span class="va">pbmcPlot</span><span class="op">$</span><span class="va">leiden_cluster</span> <span class="op">==</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>        c2 <span class="op">=</span> <span class="va">pbmcPlot</span><span class="op">$</span><span class="va">leiden_cluster</span> <span class="op">==</span> <span class="fl">2</span></span></span>
<span class="r-in"><span>    <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:38:14 2023 ... No replicate annotation specified. Creating pseudo-replicates...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:38:14 2023 ... Subsetting dataset: ctrl</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:38:14 2023 ... Subsetting dataset: stim</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:38:14 2023 ... Generating pseudo-bulks for condition "c1":</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>                              c2.6, c2.7, c2.8, c2.9, c2.10</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:38:14 2023 ... Generating pseudo-bulks for condition "c2":</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>                              c1.1, c1.2, c1.3, c1.4, c1.5</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sun Nov  5 11:38:14 2023 ... Calling DESeq2 Wald test</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> converting counts to integer mode</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating size factors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating dispersions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gene-wise dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mean-dispersion relationship</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> final dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> fitting model and testing</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    feature     logFC          pval          padj</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8  S100A11 -4.866363 2.272674e-122 1.136337e-120</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7  S100A10 -4.563204 1.766943e-119 4.417356e-118</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11  S100A6 -4.442083  1.157403e-69  1.929005e-68</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 41   ANXA5 -6.564061  2.602630e-59  3.253288e-58</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12  S100A4 -6.355146  8.603406e-59  8.603406e-58</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14  FCER1G -7.250794  3.623027e-57  3.019189e-56</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

