<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='Runs t-SNE on the aligned cell factors (result from
alignFactors), or unaligned cell factors (result from
runIntegration)) to generate a 2D embedding for visualization.
By default Rtsne (Barnes-Hut implementation of t-SNE)
method is invoked, while alternative "fftRtsne" method (FFT-accelerated
Interpolation-based t-SNE, using Kluger Lab implementation) is also
supported. For very large datasets, it is recommended to use
method = "fftRtsne" due to its efficiency and scalability.
Extra external installation steps are required for using "fftRtsne" method.
Please consult
detailed guide.'><title>Perform t-SNE dimensionality reduction — runTSNE • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Perform t-SNE dimensionality reduction — runTSNE"><meta property="og:description" content='Runs t-SNE on the aligned cell factors (result from
alignFactors), or unaligned cell factors (result from
runIntegration)) to generate a 2D embedding for visualization.
By default Rtsne (Barnes-Hut implementation of t-SNE)
method is invoked, while alternative "fftRtsne" method (FFT-accelerated
Interpolation-based t-SNE, using Kluger Lab implementation) is also
supported. For very large datasets, it is recommended to use
method = "fftRtsne" due to its efficiency and scalability.
Extra external installation steps are required for using "fftRtsne" method.
Please consult
detailed guide.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    <a class="dropdown-item" href="../articles/benchmark.html">Benchmarking integration quality</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Perform t-SNE dimensionality reduction</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/embedding.R" class="external-link"><code>R/embedding.R</code></a></small>
      <div class="d-none name"><code>runTSNE.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Runs t-SNE on the aligned cell factors (result from
<code><a href="alignFactors.html">alignFactors</a></code>), or unaligned cell factors (result from
<code><a href="runIntegration.html">runIntegration</a></code>)) to generate a 2D embedding for visualization.
By default <code><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html" class="external-link">Rtsne</a></code> (Barnes-Hut implementation of t-SNE)
method is invoked, while alternative "fftRtsne" method (FFT-accelerated
Interpolation-based t-SNE, using Kluger Lab implementation) is also
supported. For very large datasets, it is recommended to use
<code>method = "fftRtsne"</code> due to its efficiency and scalability.</p>
<p>Extra external installation steps are required for using "fftRtsne" method.
Please consult
<a href="https://welch-lab.github.io/liger/articles/installation.html" class="external-link">detailed guide</a>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runTSNE</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  useRaw <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useDims <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nDims <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  usePCA <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  perplexity <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  theta <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Rtsne"</span>, <span class="st">"fftRtsne"</span><span class="op">)</span>,</span>
<span>  dimredName <span class="op">=</span> <span class="st">"TSNE"</span>,</span>
<span>  asDefault <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fitsnePath <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">42</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span>, <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  k <span class="op">=</span> <span class="va">nDims</span>,</span>
<span>  use.raw <span class="op">=</span> <span class="va">useRaw</span>,</span>
<span>  dims.use <span class="op">=</span> <span class="va">useDims</span>,</span>
<span>  use.pca <span class="op">=</span> <span class="va">usePCA</span>,</span>
<span>  fitsne.path <span class="op">=</span> <span class="va">fitsnePath</span>,</span>
<span>  rand.seed <span class="op">=</span> <span class="va">seed</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p><a href="liger-class.html">liger</a> object with factorization results.</p></dd>


<dt>useRaw</dt>
<dd><p>Whether to use un-aligned cell factor loadings (\(H\)
matrices). Default <code>NULL</code> search for aligned factor loadings first
and un-aligned loadings then.</p></dd>


<dt>useDims</dt>
<dd><p>Index of factors to use for computing the embedding. Default
<code>NULL</code> uses all factors.</p></dd>


<dt>nDims</dt>
<dd><p>Number of dimensions to reduce to. Default <code>2</code>.</p></dd>


<dt>usePCA</dt>
<dd><p>Whether to perform initial PCA step for Rtsne. Default
<code>FALSE</code>.</p></dd>


<dt>perplexity</dt>
<dd><p>Numeric parameter to pass to Rtsne (expected number of
neighbors). Default <code>30</code>.</p></dd>


<dt>theta</dt>
<dd><p>Speed/accuracy trade-off (increase for less accuracy), set to
<code>0.0</code> for exact TSNE. Default <code>0.5</code>.</p></dd>


<dt>method</dt>
<dd><p>Choose from <code>"Rtsne"</code> or <code>"fftRtsne"</code>. See
Description. Default <code>"Rtsne"</code>.</p></dd>


<dt>dimredName</dt>
<dd><p>Name of the variable in <code>cellMeta</code> slot to store the
result matrix. Default <code>"TSNE"</code>.</p></dd>


<dt>asDefault</dt>
<dd><p>Logical, whether to set the resulting dimRed as default for
visualization. Default <code>NULL</code> will set it when no default is set.</p></dd>


<dt>fitsnePath</dt>
<dd><p>Path to the cloned FIt-SNE directory (i.e.
<code>'/path/to/dir/FIt-SNE'</code>). Required only when first time using
<code>runTSNE</code> with <code>method = "fftRtsne"</code>. Default <code>NULL</code>.</p></dd>


<dt>seed</dt>
<dd><p>Random seed for reproducibility. Default <code>42</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> or <code>TRUE</code> if users have not set.</p></dd>


<dt>use.raw, dims.use, k, use.pca, fitsne.path, rand.seed</dt>
<dd><p><b>Deprecated</b>.
See Usage section for replacement.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>The <code>object</code> where a <code>"TSNE"</code> variable is updated in the
<code>cellMeta</code> slot with the whole 2D embedding matrix.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="runUMAP.html">runUMAP</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">runTSNE</span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Generating TSNE (Rtsne) on aligned cell factor loadings</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Generating TSNE (Rtsne) on aligned cell factor loadings ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> No rownames detected. Assume cells match to the same order as in the object.</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

