<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Perform Mosaic iNMF (UINMF) on scaled datasets with unshared features — runUINMF • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Perform Mosaic iNMF (UINMF) on scaled datasets with unshared features — runUINMF"><meta property="og:description" content="Performs mosaic integrative non-negative matrix factorization (UINMF) (A.R.
Kriebel, 2022) using block coordinate descent (alternating non-negative
least squares, ANLS) to return factorized \(H\), \(W\), \(V\) and
\(U\) matrices. The objective function is stated as
$$\arg\min_{H\ge0,W\ge0,V\ge0,U\ge0}\sum_{i}^{d}
||\begin{bmatrix}E_i \\ P_i \end{bmatrix} -
(\begin{bmatrix}W \\ 0 \end{bmatrix}+
\begin{bmatrix}V_i \\ U_i \end{bmatrix})Hi||^2_F+
\lambda_i\sum_{i}^{d}||\begin{bmatrix}V_i \\ U_i \end{bmatrix}H_i||_F^2$$
where \(E_i\) is the input non-negative matrix of the \(i\)'th dataset,
\(P_i\) is the input non-negative matrix for the unshared features,
\(d\) is the total number of datasets. \(E_i\) is of size
\(m \times n_i\) for \(m\) shared features and \(n_i\) cells, \(P_i\)
is of size \(u_i \times n_i\) for \(u_i\) unshared feaetures,
\(H_i\) is of size \(k \times n_i\), \(V_i\) is of size
\(m \times k\), \(W\) is of size \(m \times k\) and \(U_i\) is of
size \(u_i \times k\).
The factorization produces a shared \(W\) matrix (genes by k). For each
dataset, an \(H\) matrix (k by cells), a \(V\) matrix (genes by k) and
a \(U\) matrix (unshared genes by k). The \(H\) matrices represent the
cell factor loadings. \(W\) is held consistent among all datasets, as it
represents the shared components of the metagenes across datasets. The
\(V\) matrices represent the dataset-specific components of the metagenes,
\(U\) matrices are similar to \(V\)s but represents the loading
contributed by unshared features.
This function adopts highly optimized fast and memory efficient
implementation extended from Planc (Kannan, 2016). Pre-installation of
extension package RcppPlanc is required. The underlying algorithm
adopts the identical ANLS strategy as optimizeALS(unshared =
TRUE) in the old version of LIGER."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/import_data.html">Import data</a>
    </li>
    <li>
      <a href="../articles/liger_object.html">Introduction to a liger object</a>
    </li>
    <li>
      <a href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform Mosaic iNMF (UINMF) on scaled datasets with unshared features</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/integration.R" class="external-link"><code>R/integration.R</code></a></small>
    <div class="hidden name"><code>runUINMF.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Performs mosaic integrative non-negative matrix factorization (UINMF) (A.R.
Kriebel, 2022) using block coordinate descent (alternating non-negative
least squares, ANLS) to return factorized \(H\), \(W\), \(V\) and
\(U\) matrices. The objective function is stated as</p>
<p>$$\arg\min_{H\ge0,W\ge0,V\ge0,U\ge0}\sum_{i}^{d}
||\begin{bmatrix}E_i \\ P_i \end{bmatrix} -
(\begin{bmatrix}W \\ 0 \end{bmatrix}+
\begin{bmatrix}V_i \\ U_i \end{bmatrix})Hi||^2_F+
\lambda_i\sum_{i}^{d}||\begin{bmatrix}V_i \\ U_i \end{bmatrix}H_i||_F^2$$</p>
<p>where \(E_i\) is the input non-negative matrix of the \(i\)'th dataset,
\(P_i\) is the input non-negative matrix for the unshared features,
\(d\) is the total number of datasets. \(E_i\) is of size
\(m \times n_i\) for \(m\) shared features and \(n_i\) cells, \(P_i\)
is of size \(u_i \times n_i\) for \(u_i\) unshared feaetures,
\(H_i\) is of size \(k \times n_i\), \(V_i\) is of size
\(m \times k\), \(W\) is of size \(m \times k\) and \(U_i\) is of
size \(u_i \times k\).</p>
<p>The factorization produces a shared \(W\) matrix (genes by k). For each
dataset, an \(H\) matrix (k by cells), a \(V\) matrix (genes by k) and
a \(U\) matrix (unshared genes by k). The \(H\) matrices represent the
cell factor loadings. \(W\) is held consistent among all datasets, as it
represents the shared components of the metagenes across datasets. The
\(V\) matrices represent the dataset-specific components of the metagenes,
\(U\) matrices are similar to \(V\)s but represents the loading
contributed by unshared features.</p>
<p>This function adopts highly optimized fast and memory efficient
implementation extended from Planc (Kannan, 2016). Pre-installation of
extension package <code>RcppPlanc</code> is required. The underlying algorithm
adopts the identical ANLS strategy as <code><a href="rliger2-deprecated.html">optimizeALS</a>(unshared =
TRUE)</code> in the old version of LIGER.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runUINMF</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  lambda <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  nIteration <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  nRandomStarts <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for liger</span></span>
<span><span class="fu">runUINMF</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  lambda <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  nIteration <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  nRandomStarts <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p><a href="liger-class.html">liger</a> object. Should run
<code><a href="selectGenes.html">selectGenes</a></code> with <code>unshared = TRUE</code> and then run
<code><a href="scaleNotCenter.html">scaleNotCenter</a></code> in advance.</p></dd>


<dt>k</dt>
<dd><p>Inner dimension of factorization (number of factors). Generally, a
higher <code>k</code> will be needed for datasets with more sub-structure. Default
<code>20</code>.</p></dd>


<dt>lambda</dt>
<dd><p>Regularization parameter. Larger values penalize
dataset-specific effects more strongly (i.e. alignment should increase as
<code>lambda</code> increases). Default <code>5</code>.</p></dd>


<dt>nIteration</dt>
<dd><p>Total number of block coordinate descent iterations to
perform. Default <code>30</code>.</p></dd>


<dt>nRandomStarts</dt>
<dd><p>Number of restarts to perform (iNMF objective function
is non-convex, so taking the best objective from multiple successive
initialization is recommended). For easier reproducibility, this increments
the random seed by 1 for each consecutive restart, so future factorization
of the same dataset can be run with one rep if necessary. Default <code>1</code>.</p></dd>


<dt>seed</dt>
<dd><p>Random seed to allow reproducible results. Default <code>1</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> which is <code>TRUE</code> if users have not set.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed to other methods and wrapped functions.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<ul><li><p>liger method - Returns updated input <a href="liger-class.html">liger</a> object.
 </p><ul><li><p>A list of all \(H\) matrices can be accessed with
         <code>getMatrix(object, "H")</code></p></li>
<li><p>A list of all \(V\) matrices can be accessed with
         <code>getMatrix(object, "V")</code></p></li>
<li><p>The \(W\) matrix can be accessed with
         <code>getMatrix(object, "W")</code></p></li>
<li><p>A list of all \(U\) matrices can be accessed with
         <code>getMatrix(object, "U")</code></p></li>
</ul></li>
</ul></div>
    <div id="note">
    <h2>Note</h2>
    <p>Currently, Seurat S3 method is not supported for UINMF because there is no
simple solution for organizing a number of miscellaneous matrices with a
single Seurat object. We strongly recommend that users create a
<a href="liger-class.html">liger</a> object which has the specific structure.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>April R. Kriebel and Joshua D. Welch, UINMF performs mosaic
integration of single-cell multi-omic datasets using nonnegative matrix
factorization, Nat. Comm., 2022</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="normalize.html">normalize</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:11:34 2023 ... Normalizing dataset: ctrl</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:11:34 2023 ... Normalizing dataset: stim</span>
<span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="selectGenes.html">selectGenes</a></span><span class="op">(</span><span class="va">pbmc</span>, useUnsharedDatasets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ctrl"</span>, <span class="st">"stim"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:11:34 2023 ... Selecting variable features for dataset: ctrl</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:11:34 2023 ...... 168 features selected out of 249 shared features</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:11:34 2023 ...... 0 features selected out of 17 unshared features</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:11:34 2023 ... Selecting variable features for dataset: stim</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:11:34 2023 ...... 166 features selected out of 249 shared features</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:11:34 2023 ...... 0 features selected out of 13 unshared features</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:11:34 2023 ... Finally 173 shared variable features selected.</span>
<span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="scaleNotCenter.html">scaleNotCenter</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:11:34 2023 ... Scaling dataset: ctrl</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:11:34 2023 ... Scaling dataset: stim</span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="fu"><a href="ligerDataset-class.html">getMatrix</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"scaleUnsharedData"</span>, <span class="st">"ctrl"</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;&amp;</span></span></span>
<span class="r-in"><span>    <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="fu"><a href="ligerDataset-class.html">getMatrix</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"scaleUnsharedData"</span>, <span class="st">"stim"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>    <span class="co"># TODO: unshared variable features cannot be detected from this example</span></span></span>
<span class="r-in"><span>    <span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">runUINMF</span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

