<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Perform UMAP Dimensionality Reduction — runUMAP • rliger2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Perform UMAP Dimensionality Reduction — runUMAP"><meta property="og:description" content="Run UMAP on the quantile normalized cell factors (result from
quantileNorm), or unnormalized cell
factors (result from optimizeALS or online_iNMF))
to generate a 2D embedding for visualization (or general dimensionality
reduction). Has option to run on subset of factors. It is generally
recommended to use this method for dimensionality reduction with extremely
large datasets. The underlying UMAP calculation imports uwot
umap."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rliger2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    </li>
    <li>
      <a href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    </li>
    <li>
      <a href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    </li>
    <li>
      <a href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    </li>
    <li>
      <a href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    </li>
    <li>
      <a href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    </li>
    <li>
      <a href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    </li>
    <li>
      <a href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
    </li>
    <li>
      <a href="../articles/Parameter_selection.html">Parameter selection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Usage
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/import_data.html">Import data</a>
    </li>
    <li>
      <a href="../articles/liger_object.html">Introduction to a liger object</a>
    </li>
    <li>
      <a href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/welch-lab/liger" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform UMAP Dimensionality Reduction</h1>
    <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/embedding.R" class="external-link"><code>R/embedding.R</code></a></small>
    <div class="hidden name"><code>runUMAP.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Run UMAP on the quantile normalized cell factors (result from
<code><a href="quantileNorm.html">quantileNorm</a></code>), or unnormalized cell
factors (result from <code><a href="rliger2-deprecated.html">optimizeALS</a></code> or <code><a href="rliger2-deprecated.html">online_iNMF</a></code>))
to generate a 2D embedding for visualization (or general dimensionality
reduction). Has option to run on subset of factors. It is generally
recommended to use this method for dimensionality reduction with extremely
large datasets. The underlying UMAP calculation imports uwot
<code><a href="https://rdrr.io/pkg/uwot/man/umap.html" class="external-link">umap</a></code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runUMAP</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  useRaw <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  useDims <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nDims <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  distance <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cosine"</span>, <span class="st">"euclidean"</span>, <span class="st">"manhattan"</span>, <span class="st">"hamming"</span><span class="op">)</span>,</span>
<span>  nNeighbors <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  minDist <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  dimredName <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">42</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span><span class="op">)</span>,</span>
<span>  k <span class="op">=</span> <span class="va">nDims</span>,</span>
<span>  use.raw <span class="op">=</span> <span class="va">useRaw</span>,</span>
<span>  dims.use <span class="op">=</span> <span class="va">useDims</span>,</span>
<span>  n_neighbors <span class="op">=</span> <span class="va">nNeighbors</span>,</span>
<span>  min_dist <span class="op">=</span> <span class="va">minDist</span>,</span>
<span>  rand.seed <span class="op">=</span> <span class="va">seed</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p><a href="liger-class.html">liger</a> object with factorization results.</p></dd>


<dt>useRaw</dt>
<dd><p>Whether to use un-aligned cell factor loadings (\(H\)
matrices). Default <code>FALSE</code>.</p></dd>


<dt>useDims</dt>
<dd><p>Index of factors to use for computing UMAP embedding. Default
<code>NULL</code> uses all factors.</p></dd>


<dt>nDims</dt>
<dd><p>Number of dimensions to reduce to. Default <code>2</code>.</p></dd>


<dt>distance</dt>
<dd><p>Character. Metric used to measure distance in the input
space. Default <code>"cosine"</code>, alternative options include:
<code>"euclidean"</code>, <code>"manhattan"</code> and <code>"hamming"</code>.</p></dd>


<dt>nNeighbors</dt>
<dd><p>Number of neighboring points used in local approximations
of manifold structure. Default <code>10</code>.</p></dd>


<dt>minDist</dt>
<dd><p>Numeric. Controls how tightly the embedding is allowed
compress points together. Default <code>0.1</code>.</p></dd>


<dt>dimredName</dt>
<dd><p>Name of the variable in <code>cellMeta</code> slot to store the
result matrix. Default <code>"UMAP"</code>.</p></dd>


<dt>seed</dt>
<dd><p>Random seed for reproducibility. Default <code>42</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> which is <code>TRUE</code> if users have not set.</p></dd>


<dt>k, use.raw, dims.use, n_neighbors, min_dist, rand.seed</dt>
<dd><p><b>Deprecated</b>.
See Usage section for replacement.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>The <code>object</code> where a <code>"UMAP"</code> variable is updated in the
<code>cellMeta</code> slot with the whole 2D embedding matrix.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>For <code>nNeighbors</code>, larger values will result in more global
structure being preserved at the loss of detailed local structure. In general
this parameter should often be in the range 5 to 50, with a choice of 10 to
15 being a sensible default.</p>
<p>For <code>minDist</code>, larger values ensure embedded points are more evenly
distributed, while smaller values allow the algorithm to optimize more
accurately with regard to local structure. Sensible values are in the range
0.001 to 0.5, with 0.1 being a reasonable default.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="runTSNE.html">runTSNE</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">runUMAP</span><span class="op">(</span><span class="va">pbmcPlot</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Mon Dec 11 16:11:35 2023 ... Generating UMAP on quantile normalized cell factor loadings...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘spam’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘spam’</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

