<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function identifies highly variable genes from each dataset
and combines these gene sets (either by union or intersection) for use in
downstream analysis. Assuming that gene expression approximately follows a
Poisson distribution, this function identifies genes with gene expression
variance above a given variance threshold (relative to mean gene expression).
Alternatively, we allow selecting a desired number of genes for each dataset
by ranking the relative variance, and then take the combination."><title>Select a subset of informative genes — selectGenes • rliger</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.8/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Select a subset of informative genes — selectGenes"><meta property="og:description" content="This function identifies highly variable genes from each dataset
and combines these gene sets (either by union or intersection) for use in
downstream analysis. Assuming that gene expression approximately follows a
Poisson distribution, this function identifies genes with gene expression
variance above a given variance threshold (relative to mean gene expression).
Alternatively, we allow selecting a desired number of genes for each dataset
by ranking the relative variance, and then take the combination."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rliger</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/installation.html">Installation</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/Integrating_multi_scRNA_data.html">Integrating multiple scRNAseq data</a>
    <a class="dropdown-item" href="../articles/Integrating_scRNA_and_scATAC_data.html">Integrating scRNA and scATAC data</a>
    <a class="dropdown-item" href="../articles/online_iNMF_tutorial.html">Online iNMF with HDF5 based data</a>
    <a class="dropdown-item" href="../articles/UINMF_vignette.html">Integration using unshared features</a>
    <a class="dropdown-item" href="../articles/cross_species_vig.html">Cross-species Analysis with UINMF</a>
    <a class="dropdown-item" href="../articles/STARmap_dropviz_vig.html">UINMF with STARmap spatial and transcriptomic</a>
    <a class="dropdown-item" href="../articles/SNAREseq_walkthrough.html">UINMF with SNAREseq scATAC and scRNA</a>
    <a class="dropdown-item" href="../articles/rna_methylation.html">Integrating scRNAseq and DNA methylation</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-usage">Usage</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-usage">
    <a class="dropdown-item" href="../articles/import_data.html">Import data</a>
    <a class="dropdown-item" href="../articles/liger_object.html">Introduction to a liger object</a>
    <a class="dropdown-item" href="../articles/liger_with_seurat.html">Interoperate with Seurat</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">NEWS</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://welch-lab.github.io/">Welch Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/welch-lab/liger">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/LabWelch">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Select a subset of informative genes</h1>
      <small class="dont-index">Source: <a href="https://github.com/welch-lab/liger/blob/HEAD/R/preprocess.R" class="external-link"><code>R/preprocess.R</code></a></small>
      <div class="d-none name"><code>selectGenes.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function identifies highly variable genes from each dataset
and combines these gene sets (either by union or intersection) for use in
downstream analysis. Assuming that gene expression approximately follows a
Poisson distribution, this function identifies genes with gene expression
variance above a given variance threshold (relative to mean gene expression).
Alternatively, we allow selecting a desired number of genes for each dataset
by ranking the relative variance, and then take the combination.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">selectGenes</span><span class="op">(</span><span class="va">object</span>, thresh <span class="op">=</span> <span class="fl">0.1</span>, nGenes <span class="op">=</span> <span class="cn">NULL</span>, alpha <span class="op">=</span> <span class="fl">0.99</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for liger</span></span>
<span><span class="fu">selectGenes</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  thresh <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  nGenes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.99</span>,</span>
<span>  useDatasets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useUnsharedDatasets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  unsharedThresh <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  combine <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"union"</span>, <span class="st">"intersection"</span><span class="op">)</span>,</span>
<span>  chunk <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span>, <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  var.thresh <span class="op">=</span> <span class="va">thresh</span>,</span>
<span>  alpha.thresh <span class="op">=</span> <span class="va">alpha</span>,</span>
<span>  num.genes <span class="op">=</span> <span class="va">nGenes</span>,</span>
<span>  datasets.use <span class="op">=</span> <span class="va">useDatasets</span>,</span>
<span>  unshared.datasets <span class="op">=</span> <span class="va">useUnsharedDatasets</span>,</span>
<span>  unshared.thresh <span class="op">=</span> <span class="va">unsharedThresh</span>,</span>
<span>  tol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  do.plot <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cex.use <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  unshared <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for Seurat</span></span>
<span><span class="fu">selectGenes</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  thresh <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  nGenes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.99</span>,</span>
<span>  useDatasets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  layer <span class="op">=</span> <span class="st">"ligerNormData"</span>,</span>
<span>  assay <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  datasetVar <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>  combine <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"union"</span>, <span class="st">"intersection"</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"ligerVerbose"</span>, <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>A <a href="liger-class.html">liger</a>, <a href="ligerDataset-class.html">ligerDataset</a> or
<code>Seurat</code> object, with normalized data available (no scale factor
multipled nor log transformed).</p></dd>


<dt>thresh</dt>
<dd><p>Variance threshold used to identify variable genes. Higher
threshold results in fewer selected genes. Liger and Seurat S3 methods accept
a single value or a vector with specific threshold for each dataset in
<code>useDatasets</code>.* Default <code>0.1</code>.</p></dd>


<dt>nGenes</dt>
<dd><p>Number of genes to find for each dataset. By setting this,
we optimize the threshold used for each dataset so that we get <code>nGenes</code>
selected features for each dataset. Accepts single value or a vector for
dataset specific setting matching <code>useDataset</code>.* Default <code>NULL</code>
does not optimize.</p></dd>


<dt>alpha</dt>
<dd><p>Alpha threshold. Controls upper bound for expected mean gene
expression. Lower threshold means higher upper bound. Default <code>0.99</code>.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed to other methods.</p></dd>


<dt>useDatasets</dt>
<dd><p>A character vector of the names, a numeric or logical
vector of the index of the datasets to use for shared variable feature
selection. Default <code>NULL</code> uses all datasets.</p></dd>


<dt>useUnsharedDatasets</dt>
<dd><p>A character vector of the names, a numeric or
logical vector of the index of the datasets to use for finding unshared
variable features. Default <code>NULL</code> does not attempt to find unshared
features.</p></dd>


<dt>unsharedThresh</dt>
<dd><p>The same thing as <code>thresh</code> that is applied to test
unshared features. A single value for all datasets in
<code>useUnsharedDatasets</code> or a vector for dataset-specific setting.* Default
<code>0.1</code>.</p></dd>


<dt>combine</dt>
<dd><p>How to combine variable genes selected from all datasets.
Choose from <code>"union"</code> or <code>"intersection"</code>. Default <code>"union"</code>.</p></dd>


<dt>chunk</dt>
<dd><p>Integer. Number of maximum number of cells in each chunk, when
gene selection is applied to any HDF5 based dataset. Default <code>1000</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical. Whether to show information of the progress. Default
<code>getOption("ligerVerbose")</code> or <code>TRUE</code> if users have not set.</p></dd>


<dt>var.thresh, alpha.thresh, num.genes, datasets.use, unshared.datasets, unshared.thresh</dt>
<dd><p><b>Deprecated</b>.
These arguments are renamed and will be removed in the future. Please see
function usage for replacement.</p></dd>


<dt>tol, do.plot, cex.use, unshared</dt>
<dd><p><b>Deprecated</b>. Gene variability
metric is now visualized with separated function
<code><a href="plotVarFeatures.html">plotVarFeatures</a></code>. Users can now set none-NULL
<code>useUnsharedDatasets</code> to select unshared genes, instead of having to
switch <code>unshared</code> on.</p></dd>


<dt>layer</dt>
<dd><p>Where the input normalized counts should be from. Default
<code>"ligerNormData"</code>. For older Seurat, always retrieve from <code>data</code>
slot.</p></dd>


<dt>assay</dt>
<dd><p>Name of assay to use. Default <code>NULL</code> uses current active
assay.</p></dd>


<dt>datasetVar</dt>
<dd><p>Metadata variable name that stores the dataset source
annotation. Default <code>"orig.ident"</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Updated object</p><ul><li><p>liger method - Each involved dataset stored in
 <a href="ligerDataset-class.html">ligerDataset</a> is updated with its <code><a href="ligerDataset-class.html">featureMeta</a></code>
 slot and <code>varUnsharedFeatures</code> slot (if requested with
 <code>useUnsharedDatasets</code>), while <code><a href="liger-class.html">varFeatures</a>(object)</code> will be
 updated with the final combined gene set.</p></li>
<li><p>Seurat method - Final selection will be updated at
 <code>Seurat::VariableFeatures(object)</code>. Per-dataset information is
 stored in the <code>meta.features</code> slot of the chosen Assay.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="normalize.html">normalize</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Normalizing datasets <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Normalizing datasets <span style="color: #0000BB;">"stim"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Normalizing datasets <span style="color: #0000BB;">"stim"</span> ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Normalizing datasets <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Normalizing datasets <span style="color: #0000BB;">"ctrl"</span> ... done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co"># Select basing on thresholding the relative variance</span></span></span>
<span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">selectGenes</span><span class="op">(</span><span class="va">pbmc</span>, thresh <span class="op">=</span> <span class="fl">.1</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Selecting variable features for dataset <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> ... 168 features selected out of 249 shared features.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Selecting variable features for dataset <span style="color: #0000BB;">"stim"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> ... 166 features selected out of 249 shared features.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Finally 173 shared variable features are selected.</span>
<span class="r-in"><span><span class="co"># Select specified number for each dataset</span></span></span>
<span class="r-in"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">selectGenes</span><span class="op">(</span><span class="va">pbmc</span>, nGenes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">60</span>, <span class="fl">60</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Selecting variable features for dataset <span style="color: #0000BB;">"ctrl"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> ... 60 features selected out of 249 shared features.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Selecting variable features for dataset <span style="color: #0000BB;">"stim"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> ... 60 features selected out of 249 shared features.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Finally 80 shared variable features are selected.</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Joshua Welch, Yichen Wang, Chao Gao, Jialin Liu, Joshua Sodicoff, Velina Kozareva, Evan Macosko.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

