#' @importFrom Rcpp evalCpp
#' @importFrom Matrix colSums rowSums t summary
#' @importFrom foreach "%dopar%"
#' @importFrom magrittr %>%
#' @importFrom rlang .data
#' @importFrom methods new show
NULL


#' Read 10X alignment data (including V3)
#'
#' This function generates a sparse matrix (genes x cells) from the data generated by 10X's
#' cellranger count pipeline. It can process V2 and V3 data together, producing either a single
#' merged matrix or list of matrices. Also handles multiple data types produced by 10X V3 (Gene
#' Expression, Antibody Capture, CRISPR, CUSTOM).
#'
#' @param sample.dirs List of directories containing either matrix.mtx(.gz) file along with genes.tsv,
#'   (features.tsv), and barcodes.tsv, or outer level 10X output directory (containing outs directory).
#' @param sample.names Vector of names to use for samples (corresponding to sample.dirs)
#' @param merge Whether to merge all matrices of the same data type across samples or leave as list
#'   of matrices (default TRUE).
#' @param num.cells Optional limit on number of cells returned for each sample (only for Gene
#'   Expression data). Retains the cells with the highest numbers of transcripts (default NULL).
#' @param min.umis Minimum UMI threshold for cells (default 0).
#' @param use.filtered Whether to use 10X's filtered data (as opposed to raw). Only relevant for
#'   sample.dirs containing 10X outs directory (default FALSE).
#' @param reference For 10X V<3, specify which reference directory to use if sample.dir is outer
#'   level 10X directory (only necessary if more than one reference used for sequencing).
#'   (default NULL)
#' @param data.type Indicates the protocol of the input data. If not specified, input data will be
#' considered scRNA-seq data (default 'rna', alternatives: 'atac').
#' @param verbose Logical. Whether to show information of the progress. Default
#' \code{getOption("ligerVerbose")} which is \code{TRUE} if users have not set.
#'
#' @return List of merged matrices across data types (returns sparse matrix if only one data type
#'   detected), or nested list of matrices organized by sample if merge=F.
#'
#' @export
#' @examples
#' \dontrun{
#' # 10X output directory V2 -- contains outs/raw_gene_bc_matrices/<reference>/...
#' sample.dir1 <- "path/to/outer/dir1"
#' # 10X output directory V3 -- for two data types, Gene Expression and CUSTOM
#' sample.dir2 <- "path/to/outer/dir2"
#' dges1 <- read10X(list(sample.dir1, sample.dir2), c("sample1", "sample2"), min.umis = 50)
#' ligerex <- createLiger(expr = dges1[["Gene Expression"]], custom = dges1[["CUSTOM"]])
#' }

read10X <-
    function(sample.dirs,
             sample.names,
             merge = TRUE,
             num.cells = NULL,
             min.umis = 0,
             use.filtered = FALSE,
             reference = NULL,
             data.type = "rna",
             verbose = getOption("ligerVerbose")) {
        datalist <- list()
        datatypes <- c("Gene Expression")

        if (length(num.cells) == 1) {
            num.cells <- rep(num.cells, length(sample.dirs))
        }
        for (i in seq_along(sample.dirs)) {
            print(paste0("Processing sample ", sample.names[i]))
            sample.dir <- sample.dirs[[i]]
            inner1 <- paste0(sample.dir, "/outs")
            if (dir.exists(inner1)) {
                sample.dir <- inner1
                is_v3 <-
                    dir.exists(paste0(sample.dir, "/filtered_feature_bc_matrix"))
                matrix.prefix <- ifelse(use.filtered, "filtered", "raw")
                if (is_v3) {
                    sample.dir <-
                        paste0(sample.dir,
                               "/",
                               matrix.prefix,
                               "_feature_bc_matrix")
                } else {
                    if (is.null(reference)) {
                        references <- list.dirs(
                            paste0(sample.dir, "/raw_gene_bc_matrices"),
                            full.names = FALSE,
                            recursive = FALSE
                        )
                        if (length(references) > 1) {
                            stop(
                                "Multiple reference genomes found. Please specify a single one."
                            )
                        } else {
                            reference <- references[1]
                        }
                    }
                    sample.dir <-
                        paste0(sample.dir,
                               "/",
                               matrix.prefix,
                               "_gene_bc_matrices/",
                               reference)
                }
            } else {
                is_v3 <- file.exists(paste0(sample.dir, "/features.tsv.gz"))
            }
            suffix <- ifelse(is_v3, ".gz", "")
            if (data.type == "rna") {
                features.file <-
                    ifelse(
                        is_v3,
                        paste0(sample.dir, "/features.tsv.gz"),
                        paste0(sample.dir, "/genes.tsv")
                    )
            } else if (data.type == "atac") {
                features.file <- ifelse(
                    is_v3,
                    paste0(sample.dir, "/peaks.bed.gz"),
                    paste0(sample.dir, "/peaks.bed")
                )
            }
            matrix.file <- paste0(sample.dir, "/matrix.mtx", suffix)
            barcodes.file <- paste0(sample.dir, "/barcodes.tsv", suffix)

            rawdata <- Matrix::readMM(matrix.file)
            # convert to dgc matrix
            if (class(rawdata)[1] == "dgTMatrix") {
                rawdata <- methods::as(rawdata, "CsparseMatrix")
            }

            # filter for UMIs first to increase speed
            umi.pass <- which(colSums(rawdata) > min.umis)
            if (length(umi.pass) == 0) {
                message("No cells pass UMI cutoff. Please lower it.")
            }
            rawdata <- rawdata[, umi.pass, drop = FALSE]

            barcodes <- readLines(barcodes.file)[umi.pass]
            # Remove -1 tag from barcodes
            if (all(grepl(barcodes, pattern = "\\-1$"))) {
                barcodes <- as.vector(sapply(barcodes, function(x) {
                    strsplit(x, "-")[[1]][1]
                }))
            }
            if (data.type == "rna") {
                features <-
                    utils::read.delim(features.file,
                               header = FALSE,
                               stringsAsFactors = FALSE)
                rownames(rawdata) <- make.unique(features[, 2])
            } else if (data.type == "atac") {
                features <- utils::read.table(features.file, header = FALSE)
                features <-
                    paste0(features[, 1], ":", features[, 2], "-", features[, 3])
                rownames(rawdata) <- features
            }
            # since some genes are only differentiated by ENSMBL
            colnames(rawdata) <- barcodes

            # split based on 10X datatype -- V3 has Gene Expression, Antibody Capture, CRISPR, CUSTOM
            # V2 has only Gene Expression by default and just two columns
            if (is.null(ncol(features))) {
                samplelist <- list(rawdata)
                names(samplelist) <- c("Chromatin Accessibility")
            } else if (ncol(features) < 3) {
                samplelist <- list(rawdata)
                names(samplelist) <- c("Gene Expression")
            } else {
                sam.datatypes <- features[, 3]
                sam.datatypes.unique <- unique(sam.datatypes)
                # keep track of all unique datatypes
                datatypes <- union(datatypes, unique(sam.datatypes))
                samplelist <- lapply(sam.datatypes.unique, function(x) {
                    rawdata[which(sam.datatypes == x),]
                })
                names(samplelist) <- sam.datatypes.unique
            }

            # num.cells filter only for gene expression data
            if (!is.null(num.cells)) {
                if (names(samplelist) == "Gene Expression" |
                    names(samplelist) == "Chromatin Accessibility") {
                    data_label <- names(samplelist)
                    cs <- colSums(samplelist[[data_label]])
                    limit <- ncol(samplelist[[data_label]])
                    if (num.cells[i] > limit) {
                        if (verbose) {
                            message(
                                "You selected more cells than are in matrix ",
                                i,
                                ". Returning all ",
                                limit,
                                " cells."
                            )
                        }
                        num.cells[i] <- limit
                    }
                    samplelist[[data_label]] <-
                        samplelist[[data_label]][, order(cs, decreasing = TRUE)
                                                 [1:num.cells[i]]]
                }

                # cs <- colSums(samplelist[["Gene Expression"]])
                # limit <- ncol(samplelist[["Gene Expression"]])
                # if (num.cells[i] > limit) {
                #   print(paste0(
                #     "You selected more cells than are in matrix ", i,
                #     ". Returning all ", limit, " cells."
                #   ))
                #   num.cells[i] <- limit
                # }
                # samplelist[["Gene Expression"]] <- samplelist[["Gene Expression"]][, order(cs, decreasing = TRUE)
                #                                                                    [1:num.cells[i]]]
            }

            datalist[[i]] <- samplelist
        }
        if (merge) {
            if (verbose) {
                message("Merging samples")
            }
            return_dges <- lapply(datatypes, function(x) {
                mergelist <- lapply(datalist, function(d) {
                    d[[x]]
                })
                mergelist <- mergelist[!sapply(mergelist, is.null)]
                sample.names.x <-
                    sample.names[!sapply(mergelist, is.null)]
                mergeSparseAll(mergelist, sample.names)
            })
            names(return_dges) <- datatypes

            # if only one type of data present
            if (length(return_dges) == 1) {
                if (verbose) {
                    message("Returning ", datatypes, " data matrix")
                }
                return(return_dges[[1]])
            }
            return(return_dges)
        } else {
            names(datalist) <- sample.names
            return(datalist)
        }
    }

#' Read liger object from RDS file
#' @param filename Path to an RDS file of a \code{liger} object of old versions.
#' @param dimredName The name of variable in \code{cellMeta} slot to store the
#' dimensionality reduction matrix, which originally located in
#' \code{tsne.coords} slot. Default \code{"tsne.coords"}.
#' @param clusterName The name of variable in \code{cellMeta} slot to store the
#' clustering assignment, which originally located in \code{clusters} slot.
#' Default \code{"clusters"}.
#' @param update Logical, whether to update an old (<=1.0.0) \code{liger} object
#' to the currect version of structure. Default \code{TRUE}.
#' @return New version of \linkS4class{liger} object
#' @export
readLiger <- function(
        filename,
        dimredName = "tsne_coords",
        clusterName = "clusters",
        update = TRUE) {
    oldObj <- readRDS(filename)
    if (!inherits(oldObj, "liger"))
        stop("Object is not of class \"liger\".")
    oldVer <- oldObj@version
    if (oldVer >= package_version("1.99.0")) return(oldObj)
    .log("Older version (", oldVer, ") of liger object detected.")
    if (isTRUE(update)) {
        .log("Updating the object structure to make it compatible ",
             "with current version (", utils::packageVersion("rliger"), ")")
        return(convertOldLiger(oldObj, dimredName = dimredName,
                               clusterName = clusterName))
    } else {
        .log("`update = FALSE` specified. Returning the original object.")
        return(oldObj)
    }
}
